{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ggmsa v1.3.4  Document: http://yulab-smu.top/ggmsa/\n",
      "\n",
      "If you use ggmsa in published research, please cite:\n",
      "L Zhou, T Feng, S Xu, F Gao, TT Lam, Q Wang, T Wu, H Huang, L Zhan, L Li, Y Guan, Z Dai*, G Yu* ggmsa: a visual exploration tool for multiple sequence alignment and associated data. Briefings in Bioinformatics. DOI:10.1093/bib/bbac222\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(LymphoSeq2)\n",
    "library(forcats)\n",
    "library(lubridate)\n",
    "library(readxl)\n",
    "library(ggpubr)\n",
    "library(patchwork)\n",
    "library(vegan)\n",
    "library(tidygraph)\n",
    "library(ggraph)\n",
    "library(gtools)\n",
    "library(graphlayouts)\n",
    "library(Biostrings)\n",
    "library(corrgram)\n",
    "library(GGally)\n",
    "library(ggseqlogo)\n",
    "library(ggh4x)\n",
    "library(ggcorrplot)\n",
    "library(ggdendro)\n",
    "library(\"ltm\")\n",
    "library(\"extrafont\")\n",
    "library(svglite)\n",
    "library(UpSetR)\n",
    "library(patchwork)\n",
    "library(grid)\n",
    "library(gridExtra)\n",
    "library(ComplexUpset)\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(SeuratData)\n",
    "library(SingleCellExperiment)\n",
    "library(SeuratWrappers)\n",
    "library(org.Hs.eg.db)\n",
    "library(ggseqlogo)\n",
    "library(ggmsa)\n",
    "# Suppress summarise info\n",
    "options(dplyr.summarise.inform = FALSE)\n",
    "options(repr.matrix.max.rows=5, repr.matrix.max.columns=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Path configuration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Configure main paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "analysis_path <- \"/path/to/kstme/folder\" \n",
    "tcr_sequencing_path <- \"/path/to/ImmunoSEQ/folder\" \n",
    "tcr_database_path <- \"/path/to/TCR/database/folder\" \n",
    "tcr_v3_sequencing_path <- \"/path/to/Adaptive/v3/formated/ImmunoSEQ/data/\" \n",
    "scrna_sequencing_path <- \"/path/to/scRNASeq/results/folder\"\n",
    "results_path <- \"/path/to/store/figures/tables/and/checkpoints\" #You must create three sub-folders under path called figures, tables, checkpoints\n",
    "reference_path <- \"/path/to/reference/genomes\" "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata path configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "hippos_lab_metrics_path <- str_c(analysis_path, \"metadata/Hippos_titer_values.xls\", sep = \"/\")\n",
    "hippos_hiv_status_path <- str_c(analysis_path, \"metadata/Hippos_hiv_status.xls\", sep = \"/\")\n",
    "hippos_metadata_path <- str_c(analysis_path, \"metadata/KSTME_main_metadata.csv\", sep = \"/\")\n",
    "hippos_rnaseq_formating_path <- str_c(analysis_path, \"metadata/KSTME_RNASeq_formating_metadata.csv\", sep = \"/\")\n",
    "blghana_metadata_path <- str_c(analysis_path, \"metadata/BLGhana_metadata.csv\", sep = \"/\")\n",
    "bluganda_metadata_path <- str_c(analysis_path, \"metadata/BLUganda_metadata.csv\", sep = \"/\")\n",
    "public_rnaseq_metadata_path <- str_c(analysis_path, \"metadata/RNASeq_metadata.csv\", sep = \"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AIRR-Seq data paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 427,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "hippos_path <- str_c(tcr_v3_sequencing_path, \"ks\", sep = \"/\")\n",
    "blghana_path <- str_c(tcr_sequencing_path, \"TRB/Burkitt_Ghana\", sep = \"/\")\n",
    "bluganda_path <- str_c(tcr_sequencing_path, \"TRB/UCI\", sep = \"/\")\n",
    "kspbmc_trb_path <- str_c(tcr_sequencing_path, \"TRB/KS\", sep = \"/\")\n",
    "kskobs_trb_path <- str_c(tcr_sequencing_path, \"TRB/KOBS\", sep = \"/\")\n",
    "ksu035_trb_path <- str_c(tcr_sequencing_path, \"U035_KSHV\", sep = \"/\")\n",
    "arks_pbmc_path <- str_c(tcr_sequencing_path, \"ACSR_ARKS_2\", sep = \"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AIRR-Seq ananlysis results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# GLIPH analysis \n",
    "bl_gliph_trb_input_path <- str_c(analysis_path, \"gliph/bl_gliph_run/bl_gliph_trb_input_table.tsv\", sep = \"/\")\n",
    "bl_gliph_clusters_path <- str_c(analysis_path, \"gliph/bl_gliph_run/bl_gliph_cluster.csv\", sep = \"/\")\n",
    "arks_gliph_input_path <- str_c(analysis_path, \"gliph/arks_gliph_run/arks_gliph_run.tsv\", sep = \"/\")\n",
    "arks_gliph_clusters_path <- str_c(analysis_path, \"gliph/arks_gliph_run/arks_gliph_cluster.csv\", sep = \"/\")\n",
    "kstme_gliph_input_path <-  str_c(analysis_path, \"gliph/kstme_tumor_normal_final/gliph_trb_table.tsv\", sep = \"/\")\n",
    "kstme_gliph_hla_input_path <- str_c(analysis_path, \"gliph/kstme_tumor_normal_final/gliph_hla_table.tsv\", sep = \"/\")\n",
    "kstme_gliph_clusters_path <- str_c(analysis_path, \"gliph/kstme_tumor_normal_final/gliph_clusters_cluster.csv\", sep = \"/\") \n",
    "kstme_gliph_hlapred_path <- str_c(analysis_path, \"gliph/kstme_tumor_normal_final/gliph_clusters_HLA.csv\", sep = \"/\")\n",
    "# CompAIRR results\n",
    "comp_path <- str_c(tcr_sequencing_path, \"Preprocessing/tables/compairr/compairr_output.tsv\", sep = \"/\")\n",
    "# Public antigen paths\n",
    "vdjdb_slim_path <- str_c(tcr_database_path, \"VdjDB/vdjdb.slim.txt\", sep = \"/\")\n",
    "mcpas_tcr_path <- str_c(tcr_database_path, \"McPASDB/McPAS-TCR.csv\", sep = \"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RNA-Seq and Nanostring data paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "hiv_fasta_path <- str_c(reference_path, \"HIV_genomes/ncbi_dataset/data/GCF_000864765.1/protein.faa\", sep = \"/\")\n",
    "hhv8_fasta_path <- str_c(reference_path, \"KSHV_genomes/ncbi_dataset/data/GCF_000838265.1/protein.faa\", sep = \"/\")\n",
    "cibersortx_results_path <- str_c(analysis_path, \"cibersort_inputs/outputs\", sep = \"/\")\n",
    "hiv_gene_expression_path <- str_c(analysis_path, \"rnaseq/hiv/salmon/kstme_hiv_genes_counts.tsv\", sep = \"/\")\n",
    "hhv8_gene_expression_path <- str_c(analysis_path, \"rnaseq/hhv8/salmon/kstme_hhv8_genes_counts.tsv\", sep = \"/\")\n",
    "ilc_path <- str_c(analysis_path, \"CRI_iAtlas_Portal_end_epi.csv\", sep = \"/\") \n",
    "gtex_nses_expression_path <- str_c(analysis_path, \"cibersort_inputs/quant/gtex_nses/gene_tpm_2017-06-05_v8_skin_not_sun_exposed_suprapubic.gct\", sep = \"/\")\n",
    "gtex_ses_expression_path <- str_c(analysis_path, \"cibersort_inputs/quant/gtex_ses/gene_tpm_2017-06-05_v8_skin_sun_exposed_lower_leg.gct\", sep = \"/\")\n",
    "lidenge_batchone_expression_path <- str_c(analysis_path, \"rnaseq/human/cibersort_inputs/lidenge_batchone.tsv\", sep = \"/\")\n",
    "lidenge_batchtwo_expression_path <- str_c(analysis_path, \"rnaseq/human/cibersort_inputs/lidenge_batchtwo.tsv\", sep = \"/\")\n",
    "lidenge_batchthree_expression_path <- str_c(analysis_path, \"rnaseq/human/cibersort_inputs/lidenge_batchthree.tsv\", sep = \"/\")\n",
    "lidenge_batchfour_expression_path <- str_c(analysis_path, \"rnaseq/human/cibersort_inputs/lidenge_batchfour.tsv\", sep = \"/\")\n",
    "tso_expression_path <- str_c(analysis_path, \"rnaseq/human/cibersort_inputs/tso_epidemic.tsv\", sep = \"/\")\n",
    "hippos_epidemic_expression <- str_c(analysis_path, \"rnaseq/human/cibersort_inputs/hippos_epidemic_merged.tsv\", sep = \"/\")\n",
    "hippos_endemic_expression <- str_c(analysis_path, \"rnaseq/human/cibersort_inputs/hippos_endemic_merged.tsv\", sep = \"/\")\n",
    "# Targeted gene expression\n",
    "targerted_expression_path <- str_c(analysis_path, \"nanostring/nanostring_counts.xlsx\", sep = \"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### scRNA-Seq GEX + VDJ data paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:3:0: unexpected end of input\n1: seurat_input_path <- str_c(scrna_sequencing_path, \"KS/preprocess/object/TCRmerged\", sep = \n2: azimuth_reference_path <- str_c(analysis_path, \"pbmc_multimodal.h5seurat\", sep = \"/\")\n  ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:3:0: unexpected end of input\n1: seurat_input_path <- str_c(scrna_sequencing_path, \"KS/preprocess/object/TCRmerged\", sep = \n2: azimuth_reference_path <- str_c(analysis_path, \"pbmc_multimodal.h5seurat\", sep = \"/\")\n  ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "seurat_input_path <- str_c(scrna_sequencing_path, \"KS/preprocess/object/TCRmerged\", sep = \"/\")\n",
    "azimuth_reference_path <- str_c(analysis_path, \"pbmc_multimodal.h5seurat\", sep = \"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GLIPH2 results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kspbmc_gliph_clusters_path <- str_c(analysis_path, \"gliph/kspbmc_gliph_run/gliph_clusters_cluster.csv\", sep = \"/\")\n",
    "kspbmc_gliph_hlapreds_path <- str_c(analysis_path, \"gliph/kspbmc_gliph_run/gliph_clusters_HLA.csv\", sep = \"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "figures_path <- str_c(results_path, \"figures\", sep = \"/\")\n",
    "table_path <- str_c(results_path, \"tables\", sep = \"/\")\n",
    "checkpoint_path <- str_c(results_path, \"checkpoints\", sep = \"/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HIPPOS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load raw titer tables from the HIPPOS study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1147 × 65</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>repertoire_id</th><th scope=col>ptid</th><th scope=col>collected_date_idi</th><th scope=col>visitcode_idi</th><th scope=col>hivvlyn_idi</th><th scope=col>hivvl_lod_idi</th><th scope=col>hivvl_detected_idi</th><th scope=col>hivvl_detectedcat_idi</th><th scope=col>hivvl_idi</th><th scope=col>cd4abs_test_date_idi</th><th scope=col>⋯</th><th scope=col>titer_kshv_pl2</th><th scope=col>titer_kshv_pl3</th><th scope=col>titer_kshv_pl4</th><th scope=col>titer_kshv_pl5</th><th scope=col>titer_kshv_pl6</th><th scope=col>titer_kshv_pl7</th><th scope=col>visitcode_kshv_pl8</th><th scope=col>titer_kshv_pl8</th><th scope=col>titer_kshv_pl9</th><th scope=col>titer_kshv_pl10</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dttm&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dttm&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001_A</td><td>U008-001</td><td>2012-10-04</td><td>V00</td><td>NO </td><td>NA</td><td>NA </td><td>NA          </td><td>    NA</td><td>NA</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>008_001_A</td><td>U008-001</td><td>2012-10-08</td><td>V01</td><td>YES</td><td>20</td><td>YES</td><td>DETECTED&gt;LOD</td><td>276908</td><td>2012-10-08</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>008_001_A</td><td>U008-001</td><td>2012-12-31</td><td>V05</td><td>YES</td><td>20</td><td>YES</td><td>DETECTED&gt;LOD</td><td>347326</td><td>2012-12-31</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_261_V9_10_26_21</td><td>U008-261</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>008_264_V1_5_6_21  </td><td>U008-264</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1147 × 65\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " repertoire\\_id & ptid & collected\\_date\\_idi & visitcode\\_idi & hivvlyn\\_idi & hivvl\\_lod\\_idi & hivvl\\_detected\\_idi & hivvl\\_detectedcat\\_idi & hivvl\\_idi & cd4abs\\_test\\_date\\_idi & ⋯ & titer\\_kshv\\_pl2 & titer\\_kshv\\_pl3 & titer\\_kshv\\_pl4 & titer\\_kshv\\_pl5 & titer\\_kshv\\_pl6 & titer\\_kshv\\_pl7 & visitcode\\_kshv\\_pl8 & titer\\_kshv\\_pl8 & titer\\_kshv\\_pl9 & titer\\_kshv\\_pl10\\\\\n",
       " <chr> & <chr> & <dttm> & <chr> & <chr> & <dbl> & <chr> & <chr> & <dbl> & <dttm> & ⋯ & <dbl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t 008\\_001\\_A & U008-001 & 2012-10-04 & V00 & NO  & NA & NA  & NA           &     NA & NA & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 008\\_001\\_A & U008-001 & 2012-10-08 & V01 & YES & 20 & YES & DETECTED>LOD & 276908 & 2012-10-08 & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 008\\_001\\_A & U008-001 & 2012-12-31 & V05 & YES & 20 & YES & DETECTED>LOD & 347326 & 2012-12-31 & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_261\\_V9\\_10\\_26\\_21 & U008-261 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 008\\_264\\_V1\\_5\\_6\\_21   & U008-264 & NA & NA & NA & NA & NA & NA & NA & NA & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1147 × 65\n",
       "\n",
       "| repertoire_id &lt;chr&gt; | ptid &lt;chr&gt; | collected_date_idi &lt;dttm&gt; | visitcode_idi &lt;chr&gt; | hivvlyn_idi &lt;chr&gt; | hivvl_lod_idi &lt;dbl&gt; | hivvl_detected_idi &lt;chr&gt; | hivvl_detectedcat_idi &lt;chr&gt; | hivvl_idi &lt;dbl&gt; | cd4abs_test_date_idi &lt;dttm&gt; | ⋯ ⋯ | titer_kshv_pl2 &lt;dbl&gt; | titer_kshv_pl3 &lt;lgl&gt; | titer_kshv_pl4 &lt;lgl&gt; | titer_kshv_pl5 &lt;lgl&gt; | titer_kshv_pl6 &lt;lgl&gt; | titer_kshv_pl7 &lt;lgl&gt; | visitcode_kshv_pl8 &lt;lgl&gt; | titer_kshv_pl8 &lt;lgl&gt; | titer_kshv_pl9 &lt;lgl&gt; | titer_kshv_pl10 &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 008_001_A | U008-001 | 2012-10-04 | V00 | NO  | NA | NA  | NA           |     NA | NA | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 008_001_A | U008-001 | 2012-10-08 | V01 | YES | 20 | YES | DETECTED&gt;LOD | 276908 | 2012-10-08 | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 008_001_A | U008-001 | 2012-12-31 | V05 | YES | 20 | YES | DETECTED&gt;LOD | 347326 | 2012-12-31 | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_261_V9_10_26_21 | U008-261 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 008_264_V1_5_6_21   | U008-264 | NA | NA | NA | NA | NA | NA | NA | NA | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "     repertoire_id       ptid     collected_date_idi visitcode_idi hivvlyn_idi\n",
       "1    008_001_A           U008-001 2012-10-04         V00           NO         \n",
       "2    008_001_A           U008-001 2012-10-08         V01           YES        \n",
       "3    008_001_A           U008-001 2012-12-31         V05           YES        \n",
       "⋮    ⋮                   ⋮        ⋮                  ⋮             ⋮          \n",
       "1146 008_261_V9_10_26_21 U008-261 <NA>               NA            NA         \n",
       "1147 008_264_V1_5_6_21   U008-264 <NA>               NA            NA         \n",
       "     hivvl_lod_idi hivvl_detected_idi hivvl_detectedcat_idi hivvl_idi\n",
       "1    NA            NA                 NA                        NA   \n",
       "2    20            YES                DETECTED>LOD          276908   \n",
       "3    20            YES                DETECTED>LOD          347326   \n",
       "⋮    ⋮             ⋮                  ⋮                     ⋮        \n",
       "1146 NA            NA                 NA                    NA       \n",
       "1147 NA            NA                 NA                    NA       \n",
       "     cd4abs_test_date_idi ⋯ titer_kshv_pl2 titer_kshv_pl3 titer_kshv_pl4\n",
       "1    <NA>                 ⋯ NA             NA             NA            \n",
       "2    2012-10-08           ⋯ NA             NA             NA            \n",
       "3    2012-12-31           ⋯ NA             NA             NA            \n",
       "⋮    ⋮                    ⋱ ⋮              ⋮              ⋮             \n",
       "1146 <NA>                 ⋯ NA             NA             NA            \n",
       "1147 <NA>                 ⋯ NA             NA             NA            \n",
       "     titer_kshv_pl5 titer_kshv_pl6 titer_kshv_pl7 visitcode_kshv_pl8\n",
       "1    NA             NA             NA             NA                \n",
       "2    NA             NA             NA             NA                \n",
       "3    NA             NA             NA             NA                \n",
       "⋮    ⋮              ⋮              ⋮              ⋮                 \n",
       "1146 NA             NA             NA             NA                \n",
       "1147 NA             NA             NA             NA                \n",
       "     titer_kshv_pl8 titer_kshv_pl9 titer_kshv_pl10\n",
       "1    NA             NA             NA             \n",
       "2    NA             NA             NA             \n",
       "3    NA             NA             NA             \n",
       "⋮    ⋮              ⋮              ⋮              \n",
       "1146 NA             NA             NA             \n",
       "1147 NA             NA             NA             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hippos_titer_table <- readxl::read_excel(hippos_lab_metrics_path) |>\n",
    "  dplyr::select(repertoire_id, ptid, collected_date_idi,\n",
    "    matches(\"visitcode|hivvl|cd4abs|cd8abs|colldt_kshv_or\\\\d+|titer_kshv_or\\\\d+|\n",
    "    colldt_kshv_pl\\\\d+|titer_kshv_pl\\\\d+\"))\n",
    "kstme_titer_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we are going to create a table that shows the study entry dates for each of the individuals enrolled in the HIPPOS cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 146 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>patient_id</th><th scope=col>study_entry_date</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;date&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001</td><td>2012-10-04</td></tr>\n",
       "\t<tr><td>008_002</td><td>2012-10-15</td></tr>\n",
       "\t<tr><td>008_003</td><td>2012-10-22</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_263</td><td>2021-04-20</td></tr>\n",
       "\t<tr><td>008_264</td><td>2021-04-26</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 146 × 2\n",
       "\\begin{tabular}{ll}\n",
       " patient\\_id & study\\_entry\\_date\\\\\n",
       " <chr> & <date>\\\\\n",
       "\\hline\n",
       "\t 008\\_001 & 2012-10-04\\\\\n",
       "\t 008\\_002 & 2012-10-15\\\\\n",
       "\t 008\\_003 & 2012-10-22\\\\\n",
       "\t ⋮ & ⋮\\\\\n",
       "\t 008\\_263 & 2021-04-20\\\\\n",
       "\t 008\\_264 & 2021-04-26\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 146 × 2\n",
       "\n",
       "| patient_id &lt;chr&gt; | study_entry_date &lt;date&gt; |\n",
       "|---|---|\n",
       "| 008_001 | 2012-10-04 |\n",
       "| 008_002 | 2012-10-15 |\n",
       "| 008_003 | 2012-10-22 |\n",
       "| ⋮ | ⋮ |\n",
       "| 008_263 | 2021-04-20 |\n",
       "| 008_264 | 2021-04-26 |\n",
       "\n"
      ],
      "text/plain": [
       "    patient_id study_entry_date\n",
       "1   008_001    2012-10-04      \n",
       "2   008_002    2012-10-15      \n",
       "3   008_003    2012-10-22      \n",
       "⋮   ⋮          ⋮               \n",
       "145 008_263    2021-04-20      \n",
       "146 008_264    2021-04-26      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hippos_study_entry_table <- hippos_titer_table |> \n",
    "  dplyr::select(ptid, matches(\"date|colldt\"))  |> \n",
    "  pivot_longer(-ptid, names_to = \"group\", values_to = \"date\") |> \n",
    "  group_by(ptid) |> \n",
    "  summarize(min_date = min(date, na.rm = TRUE)) |>\n",
    "  ungroup() |> \n",
    "  dplyr::rename(patient_id = ptid, study_entry_date = min_date) |> \n",
    "  mutate(patient_id = str_replace(str_remove(patient_id, \"U\"), \"-\", \"_\"),\n",
    "    study_entry_date = as_date(study_entry_date)) \n",
    "hippos_study_entry_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### HIV titer values\n",
    "HIV titre values are just recorded at every entry for individuals with epidemic \n",
    "KS. PTIDs in the titre tables do not match the rest of our metadata this needs\n",
    "to standardized first. The visit codes need to be transformed into consistent \n",
    "factors. Dates need to be converted to relative time frames. When the viral \n",
    "loads are below LOD, titre values are set to NA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 667 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>patient_id</th><th scope=col>visit_code</th><th scope=col>titre</th><th scope=col>category</th><th scope=col>units</th><th scope=col>days_in_study</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001</td><td>V01</td><td>276908</td><td>HIV</td><td>copies/mL</td><td>  4</td></tr>\n",
       "\t<tr><td>008_001</td><td>V05</td><td>347326</td><td>HIV</td><td>copies/mL</td><td> 88</td></tr>\n",
       "\t<tr><td>008_001</td><td>V08</td><td>   685</td><td>HIV</td><td>copies/mL</td><td>174</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_265</td><td>V01</td><td>139653</td><td>HIV</td><td>copies/mL</td><td>NA</td></tr>\n",
       "\t<tr><td>008_265</td><td>V05</td><td>141524</td><td>HIV</td><td>copies/mL</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 667 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " patient\\_id & visit\\_code & titre & category & units & days\\_in\\_study\\\\\n",
       " <chr> & <fct> & <dbl> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 008\\_001 & V01 & 276908 & HIV & copies/mL &   4\\\\\n",
       "\t 008\\_001 & V05 & 347326 & HIV & copies/mL &  88\\\\\n",
       "\t 008\\_001 & V08 &    685 & HIV & copies/mL & 174\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_265 & V01 & 139653 & HIV & copies/mL & NA\\\\\n",
       "\t 008\\_265 & V05 & 141524 & HIV & copies/mL & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 667 × 6\n",
       "\n",
       "| patient_id &lt;chr&gt; | visit_code &lt;fct&gt; | titre &lt;dbl&gt; | category &lt;chr&gt; | units &lt;chr&gt; | days_in_study &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 008_001 | V01 | 276908 | HIV | copies/mL |   4 |\n",
       "| 008_001 | V05 | 347326 | HIV | copies/mL |  88 |\n",
       "| 008_001 | V08 |    685 | HIV | copies/mL | 174 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_265 | V01 | 139653 | HIV | copies/mL | NA |\n",
       "| 008_265 | V05 | 141524 | HIV | copies/mL | NA |\n",
       "\n"
      ],
      "text/plain": [
       "    patient_id visit_code titre  category units     days_in_study\n",
       "1   008_001    V01        276908 HIV      copies/mL   4          \n",
       "2   008_001    V05        347326 HIV      copies/mL  88          \n",
       "3   008_001    V08           685 HIV      copies/mL 174          \n",
       "⋮   ⋮          ⋮          ⋮      ⋮        ⋮         ⋮            \n",
       "666 008_265    V01        139653 HIV      copies/mL NA           \n",
       "667 008_265    V05        141524 HIV      copies/mL NA           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hippos_hiv_titer_table <- read_excel(hippos_hiv_status_path) |> #Just one sheet in dataset\n",
    "  dplyr::select(-c(ptid, visitcode, accession_no, hivrna_comments, \n",
    "    hivrna_detectedcat_bl, hivrna_dt_bl, hivrna_bl)) |> #We just need the HIV counts in blood by visit\n",
    "  dplyr::rename(patient_id = PTID, screen_date = hivscreendt, titre = hivrna,\n",
    "    collection_date = hivrna_dt, visit_code = visitname) |> #Rename variables \n",
    "  filter(!is.na(hiv)) |> #Remove missing information\n",
    "  mutate(patient_id = str_replace(str_extract(patient_id, \"008-\\\\d+\"), \n",
    "      \"-\", \"_\"), #Need to convert the PTIDs to the format compatible with sequencing data\n",
    "    collection_date = as_date(collection_date), \n",
    "    screen_date = as_date(screen_date),\n",
    "    visit_code = factor(visit_code, levels = c(\"V00\", \"V01\", \"V02\", \"V03\",\n",
    "      \"V04\", \"V05\", \"V06\", \"V07\", \"V08\", \"V09\", \"V10\", \"V11\")), #The max visit observed was V11\n",
    "    days_in_study = interval(screen_date, collection_date) %/% days(1),\n",
    "    category = \"HIV\", units = \"copies/mL\") |>\n",
    "  dplyr::select(patient_id, visit_code, titre, category, units, days_in_study) |>\n",
    "  arrange(patient_id, visit_code)\n",
    "hippos_hiv_titer_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CD4 values \n",
    "\n",
    "These follow the same table format as the CD8 table, so the code is \n",
    "essentially reused here and in the next code chunk. It is important to note that\n",
    "CD4+ T-cell counts were mostly collected for just the individuals with epidemic\n",
    "KS. Though a handful of individuals with endemic KS do have CD4+ counts\n",
    "associated with their case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 477 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>patient_id</th><th scope=col>visit_code</th><th scope=col>titre</th><th scope=col>category</th><th scope=col>units</th><th scope=col>days_in_study</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001</td><td>V01</td><td>127</td><td>CD4_count</td><td>cells/uL</td><td>  4</td></tr>\n",
       "\t<tr><td>008_001</td><td>V05</td><td>127</td><td>CD4_count</td><td>cells/uL</td><td> 88</td></tr>\n",
       "\t<tr><td>008_001</td><td>V08</td><td>225</td><td>CD4_count</td><td>cells/uL</td><td>174</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_263</td><td>V09</td><td>286</td><td>CD4_count</td><td>cells/uL</td><td>275</td></tr>\n",
       "\t<tr><td>008_263</td><td>V10</td><td>406</td><td>CD4_count</td><td>cells/uL</td><td>366</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 477 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " patient\\_id & visit\\_code & titre & category & units & days\\_in\\_study\\\\\n",
       " <chr> & <fct> & <dbl> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 008\\_001 & V01 & 127 & CD4\\_count & cells/uL &   4\\\\\n",
       "\t 008\\_001 & V05 & 127 & CD4\\_count & cells/uL &  88\\\\\n",
       "\t 008\\_001 & V08 & 225 & CD4\\_count & cells/uL & 174\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_263 & V09 & 286 & CD4\\_count & cells/uL & 275\\\\\n",
       "\t 008\\_263 & V10 & 406 & CD4\\_count & cells/uL & 366\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 477 × 6\n",
       "\n",
       "| patient_id &lt;chr&gt; | visit_code &lt;fct&gt; | titre &lt;dbl&gt; | category &lt;chr&gt; | units &lt;chr&gt; | days_in_study &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 008_001 | V01 | 127 | CD4_count | cells/uL |   4 |\n",
       "| 008_001 | V05 | 127 | CD4_count | cells/uL |  88 |\n",
       "| 008_001 | V08 | 225 | CD4_count | cells/uL | 174 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_263 | V09 | 286 | CD4_count | cells/uL | 275 |\n",
       "| 008_263 | V10 | 406 | CD4_count | cells/uL | 366 |\n",
       "\n"
      ],
      "text/plain": [
       "    patient_id visit_code titre category  units    days_in_study\n",
       "1   008_001    V01        127   CD4_count cells/uL   4          \n",
       "2   008_001    V05        127   CD4_count cells/uL  88          \n",
       "3   008_001    V08        225   CD4_count cells/uL 174          \n",
       "⋮   ⋮          ⋮          ⋮     ⋮         ⋮        ⋮            \n",
       "476 008_263    V09        286   CD4_count cells/uL 275          \n",
       "477 008_263    V10        406   CD4_count cells/uL 366          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hippos_cd4_titer_table <- hippos_titer_table |> \n",
    "  dplyr::select(ptid, visitcode_idi, cd4abs_test_date_idi, cd4abs_idi) |> \n",
    "  dplyr::rename(patient_id = ptid, visit_code = visitcode_idi, \n",
    "    collection_date = cd4abs_test_date_idi, titre = cd4abs_idi) |>\n",
    "  mutate(patient_id = str_replace(str_remove(patient_id, \"U\"), \"-\", \"_\"),\n",
    "    collection_date = as_date(collection_date),\n",
    "    visit_code = factor(visit_code, levels = c(\"V00\", \"V01\", \"V02\", \"V03\", \n",
    "      \"V04\", \"V05\", \"V06\", \"V07\", \"V08\", \"V09\", \"V10\", \"V11\")),\n",
    "    category = \"CD4_count\", units = \"cells/uL\") |>\n",
    "  left_join(hippos_study_entry_table, by = \"patient_id\") |>\n",
    "  mutate(days_in_study = interval(study_entry_date, collection_date) %/% days(1)) |> \n",
    "  ungroup() |> \n",
    "  dplyr::select(-collection_date, -study_entry_date) |>\n",
    "  filter(!is.na(titre))\n",
    "hippos_cd4_titer_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### KSHV plasma titer\n",
    "Now we repeat this process for KSHV titres in peripheral blood samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 831 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>patient_id</th><th scope=col>visit_code</th><th scope=col>titre</th><th scope=col>days_in_study</th><th scope=col>category</th><th scope=col>units</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001</td><td>V01</td><td>   0.00</td><td> 4</td><td>KSHV_plasma</td><td>copies/mL</td></tr>\n",
       "\t<tr><td>008_001</td><td>V02</td><td>5062.27</td><td>26</td><td>KSHV_plasma</td><td>copies/mL</td></tr>\n",
       "\t<tr><td>008_001</td><td>V03</td><td>2158.66</td><td>46</td><td>KSHV_plasma</td><td>copies/mL</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_263</td><td>V01</td><td>19.85711</td><td> 0</td><td>KSHV_plasma</td><td>copies/mL</td></tr>\n",
       "\t<tr><td>008_264</td><td>V01</td><td>23.17499</td><td>10</td><td>KSHV_plasma</td><td>copies/mL</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 831 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " patient\\_id & visit\\_code & titre & days\\_in\\_study & category & units\\\\\n",
       " <chr> & <fct> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_001 & V01 &    0.00 &  4 & KSHV\\_plasma & copies/mL\\\\\n",
       "\t 008\\_001 & V02 & 5062.27 & 26 & KSHV\\_plasma & copies/mL\\\\\n",
       "\t 008\\_001 & V03 & 2158.66 & 46 & KSHV\\_plasma & copies/mL\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_263 & V01 & 19.85711 &  0 & KSHV\\_plasma & copies/mL\\\\\n",
       "\t 008\\_264 & V01 & 23.17499 & 10 & KSHV\\_plasma & copies/mL\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 831 × 6\n",
       "\n",
       "| patient_id &lt;chr&gt; | visit_code &lt;fct&gt; | titre &lt;dbl&gt; | days_in_study &lt;dbl&gt; | category &lt;chr&gt; | units &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 008_001 | V01 |    0.00 |  4 | KSHV_plasma | copies/mL |\n",
       "| 008_001 | V02 | 5062.27 | 26 | KSHV_plasma | copies/mL |\n",
       "| 008_001 | V03 | 2158.66 | 46 | KSHV_plasma | copies/mL |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_263 | V01 | 19.85711 |  0 | KSHV_plasma | copies/mL |\n",
       "| 008_264 | V01 | 23.17499 | 10 | KSHV_plasma | copies/mL |\n",
       "\n"
      ],
      "text/plain": [
       "    patient_id visit_code titre    days_in_study category    units    \n",
       "1   008_001    V01           0.00   4            KSHV_plasma copies/mL\n",
       "2   008_001    V02        5062.27  26            KSHV_plasma copies/mL\n",
       "3   008_001    V03        2158.66  46            KSHV_plasma copies/mL\n",
       "⋮   ⋮          ⋮          ⋮        ⋮             ⋮           ⋮        \n",
       "830 008_263    V01        19.85711  0            KSHV_plasma copies/mL\n",
       "831 008_264    V01        23.17499 10            KSHV_plasma copies/mL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hippos_plasma_titre_dates <- hippos_titer_table |> \n",
    "  dplyr::select(ptid, visitcode_kshv_pl, matches(\"colldt_kshv_or1\")) |>\n",
    "  pivot_longer(cols = -c(ptid, visitcode_kshv_pl)) |>\n",
    "  dplyr::rename(patient_id = ptid, visit_code = visitcode_kshv_pl, observation = name,\n",
    "    collection_date = value) |> \n",
    "  mutate(patient_id = str_replace(str_remove(patient_id, \"U\"), \"-\", \"_\"),\n",
    "    collection_date = as_date(collection_date), \n",
    "    observation = str_remove(observation, \"colldt_kshv_\"),\n",
    "    observation = str_replace(observation, \"or\", \"pl\")) |>\n",
    "  filter(!is.na(collection_date)) |> \n",
    "  left_join(hippos_study_entry_table, by = \"patient_id\") |>\n",
    "  mutate(days_in_study = interval(study_entry_date, collection_date) %/% days(1))\n",
    "# Separate out titres for each swab at each time point\n",
    "hippos_kshv_plasma_titre_values <- hippos_titer_table |> \n",
    "  dplyr::select(ptid, visitcode_kshv_pl, matches(\"titer_kshv_pl1\")) |>\n",
    "  pivot_longer(cols = -c(ptid, visitcode_kshv_pl)) |>\n",
    "  dplyr::rename(patient_id = ptid, visit_code = visitcode_kshv_pl, observation = name,\n",
    "    titre = value) |> \n",
    "  mutate(patient_id = str_replace(str_remove(patient_id, \"U\"), \"-\", \"_\"),\n",
    "    observation = str_remove(observation, \"titer_kshv_\")) |>\n",
    "  filter(!is.na(titre)) \n",
    "# Merge the dates and titres on patient_id, visit_code, and swab number. As \n",
    "# mentioned earlier, for now, we are just keeping one titre value for each\n",
    "# visit. For each visit the lowest value of days_in_study and the highest value\n",
    "# of titre in recorded\n",
    "hippos_kshv_plasma_titre_table <- left_join(hippos_plasma_titre_dates,\n",
    "    hippos_kshv_plasma_titre_values,\n",
    "    by = c(\"patient_id\", \"visit_code\", \"observation\")) |>\n",
    "  dplyr::select(-study_entry_date) |>\n",
    "  ungroup() |> \n",
    "  group_by(patient_id, visit_code) |> \n",
    "  summarize(titre = max(titre), days_in_study = min(days_in_study)) |> \n",
    "  ungroup() |> \n",
    "  mutate(visit_code = factor(visit_code, levels = c(\"V00\", \"V01\", \"V02\", \"V03\", \n",
    "    \"V04\", \"V05\", \"V06\", \"V07\", \"V08\", \"V09\", \"V10\", \"V11\")),\n",
    "    category = \"KSHV_plasma\", units = \"copies/mL\")\n",
    "hippos_kshv_plasma_titre_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1975 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>patient_id</th><th scope=col>visit_code</th><th scope=col>titre</th><th scope=col>category</th><th scope=col>units</th><th scope=col>days_in_study</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001</td><td>V01</td><td>276908</td><td>HIV</td><td>copies/mL</td><td>  4</td></tr>\n",
       "\t<tr><td>008_001</td><td>V05</td><td>347326</td><td>HIV</td><td>copies/mL</td><td> 88</td></tr>\n",
       "\t<tr><td>008_001</td><td>V08</td><td>   685</td><td>HIV</td><td>copies/mL</td><td>174</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_263</td><td>V01</td><td>19.85711</td><td>KSHV_plasma</td><td>copies/mL</td><td> 0</td></tr>\n",
       "\t<tr><td>008_264</td><td>V01</td><td>23.17499</td><td>KSHV_plasma</td><td>copies/mL</td><td>10</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1975 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " patient\\_id & visit\\_code & titre & category & units & days\\_in\\_study\\\\\n",
       " <chr> & <fct> & <dbl> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 008\\_001 & V01 & 276908 & HIV & copies/mL &   4\\\\\n",
       "\t 008\\_001 & V05 & 347326 & HIV & copies/mL &  88\\\\\n",
       "\t 008\\_001 & V08 &    685 & HIV & copies/mL & 174\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_263 & V01 & 19.85711 & KSHV\\_plasma & copies/mL &  0\\\\\n",
       "\t 008\\_264 & V01 & 23.17499 & KSHV\\_plasma & copies/mL & 10\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1975 × 6\n",
       "\n",
       "| patient_id &lt;chr&gt; | visit_code &lt;fct&gt; | titre &lt;dbl&gt; | category &lt;chr&gt; | units &lt;chr&gt; | days_in_study &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 008_001 | V01 | 276908 | HIV | copies/mL |   4 |\n",
       "| 008_001 | V05 | 347326 | HIV | copies/mL |  88 |\n",
       "| 008_001 | V08 |    685 | HIV | copies/mL | 174 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_263 | V01 | 19.85711 | KSHV_plasma | copies/mL |  0 |\n",
       "| 008_264 | V01 | 23.17499 | KSHV_plasma | copies/mL | 10 |\n",
       "\n"
      ],
      "text/plain": [
       "     patient_id visit_code titre    category    units     days_in_study\n",
       "1    008_001    V01        276908   HIV         copies/mL   4          \n",
       "2    008_001    V05        347326   HIV         copies/mL  88          \n",
       "3    008_001    V08           685   HIV         copies/mL 174          \n",
       "⋮    ⋮          ⋮          ⋮        ⋮           ⋮         ⋮            \n",
       "1974 008_263    V01        19.85711 KSHV_plasma copies/mL  0           \n",
       "1975 008_264    V01        23.17499 KSHV_plasma copies/mL 10           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Now we can merge all six tables into one titre table for whole study\n",
    "hippos_titer_tables <- bind_rows(hippos_hiv_titer_table, hippos_cd4_titer_table, \n",
    "  hippos_kshv_plasma_titre_table)\n",
    "hippos_titer_tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 1141 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>patient_id</th><th scope=col>visit_code</th><th scope=col>HIV</th><th scope=col>CD4_count</th><th scope=col>KSHV_plasma</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001</td><td>V01</td><td>276908</td><td>127</td><td>   0.00</td></tr>\n",
       "\t<tr><td>008_001</td><td>V02</td><td>    NA</td><td> NA</td><td>5062.27</td></tr>\n",
       "\t<tr><td>008_001</td><td>V03</td><td>    NA</td><td> NA</td><td>2158.66</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_265</td><td>V01</td><td>139653</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>008_265</td><td>V05</td><td>141524</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 1141 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " patient\\_id & visit\\_code & HIV & CD4\\_count & KSHV\\_plasma\\\\\n",
       " <chr> & <fct> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 008\\_001 & V01 & 276908 & 127 &    0.00\\\\\n",
       "\t 008\\_001 & V02 &     NA &  NA & 5062.27\\\\\n",
       "\t 008\\_001 & V03 &     NA &  NA & 2158.66\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_265 & V01 & 139653 & NA & NA\\\\\n",
       "\t 008\\_265 & V05 & 141524 & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 1141 × 5\n",
       "\n",
       "| patient_id &lt;chr&gt; | visit_code &lt;fct&gt; | HIV &lt;dbl&gt; | CD4_count &lt;dbl&gt; | KSHV_plasma &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 008_001 | V01 | 276908 | 127 |    0.00 |\n",
       "| 008_001 | V02 |     NA |  NA | 5062.27 |\n",
       "| 008_001 | V03 |     NA |  NA | 2158.66 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_265 | V01 | 139653 | NA | NA |\n",
       "| 008_265 | V05 | 141524 | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "     patient_id visit_code HIV    CD4_count KSHV_plasma\n",
       "1    008_001    V01        276908 127          0.00    \n",
       "2    008_001    V02            NA  NA       5062.27    \n",
       "3    008_001    V03            NA  NA       2158.66    \n",
       "⋮    ⋮          ⋮          ⋮      ⋮         ⋮          \n",
       "1140 008_265    V01        139653 NA        NA         \n",
       "1141 008_265    V05        141524 NA        NA         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# By pivoting wide, we can generate a table with one row for each of the\n",
    "# subjects in the study. \n",
    "hippos_titre_table_wide <- hippos_titer_tables |> \n",
    "  pivot_wider(id_cols = c(patient_id, visit_code), \n",
    "    names_from = category, values_from = titre) |>\n",
    "  group_by(patient_id, visit_code) |>\n",
    "  summarize_all(max)\n",
    "hippos_titre_table_wide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequencing metadata\n",
    "The current study was concieved with a subset of the Hippos datasets where we\n",
    "had collected many modalities of sequencing data. This dataset will be reffered\n",
    "to as the KSTME dataset. A total of 622 TRB repertoires from the 130 individuals in the study. With a total of 436 repertoires (74 PBMCs, 84 NAT, 268 Tumor, and 10 scREP) from individuals with epidemic KS and 152 repertoires (42 PBMCs, 26 NAT, 80 Tumor, and 4 scREP) from individuals with endemic KS. This is accompanied with 27 nanostring datasets and 52 bulk RNA-Seq datasets from Endemic and Epidemic KS lesions. For each of the individuals with sequencing data, we also have accompanying HLA typing information. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A spec_tbl_df: 682 × 42</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>tra_repertoire_id</th><th scope=col>trb_repertoire_id</th><th scope=col>rna_libraries</th><th scope=col>nanostring_libraries</th><th scope=col>gex_library</th><th scope=col>vdj_library</th><th scope=col>sc_library_count</th><th scope=col>⋯</th><th scope=col>DPB1allele2</th><th scope=col>DQA1allele1</th><th scope=col>DQA1allele2</th><th scope=col>DQB1allele1</th><th scope=col>DQB1allele2</th><th scope=col>DRB1allele1</th><th scope=col>DRB1allele2</th><th scope=col>DRB345allele1</th><th scope=col>DRB345allele2</th><th scope=col>timestamp</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_A</td><td>NA       </td><td>008_001_A</td><td>NA          </td><td>001-A</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*04:01:01</td><td>DQB1*04:02:01</td><td>DQB1*05:01:01</td><td>DRB1*03:02:01</td><td>DRB1*11:01:02</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td><td>1/12/23</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_B</td><td>008_001_B</td><td>008_001_B</td><td>008-001-B-12</td><td>001-B</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*04:01:01</td><td>DQB1*04:02:01</td><td>DQB1*05:01:01</td><td>DRB1*03:02:01</td><td>DRB1*11:01:02</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td><td>1/12/23</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_C</td><td>NA       </td><td>NA       </td><td>008-001-C-13</td><td>NA   </td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*04:01:01</td><td>DQB1*04:02:01</td><td>DQB1*05:01:01</td><td>DRB1*03:02:01</td><td>DRB1*11:01:02</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td><td>1/12/23</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>TsoEtAl</td><td>SRR5787176</td><td>SRR5787176</td><td>NA</td><td>NA</td><td>p83</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>TsoEtAl</td><td>SRR5787177</td><td>SRR5787177</td><td>NA</td><td>NA</td><td>p23</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A spec\\_tbl\\_df: 682 × 42\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " cohort & patient\\_id & repertoire\\_id & tra\\_repertoire\\_id & trb\\_repertoire\\_id & rna\\_libraries & nanostring\\_libraries & gex\\_library & vdj\\_library & sc\\_library\\_count & ⋯ & DPB1allele2 & DQA1allele1 & DQA1allele2 & DQB1allele1 & DQB1allele2 & DRB1allele1 & DRB1allele2 & DRB345allele1 & DRB345allele2 & timestamp\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_A & NA        & 008\\_001\\_A & NA           & 001-A & NA & NA & NA & ⋯ & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*04:01:01 & DQB1*04:02:01 & DQB1*05:01:01 & DRB1*03:02:01 & DRB1*11:01:02 & DRB3*01:01:02 & DRB3*03:01:01 & 1/12/23\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_B & 008\\_001\\_B & 008\\_001\\_B & 008-001-B-12 & 001-B & NA & NA & NA & ⋯ & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*04:01:01 & DQB1*04:02:01 & DQB1*05:01:01 & DRB1*03:02:01 & DRB1*11:01:02 & DRB3*01:01:02 & DRB3*03:01:01 & 1/12/23\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_C & NA        & NA        & 008-001-C-13 & NA    & NA & NA & NA & ⋯ & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*04:01:01 & DQB1*04:02:01 & DQB1*05:01:01 & DRB1*03:02:01 & DRB1*11:01:02 & DRB3*01:01:02 & DRB3*03:01:01 & 1/12/23\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t TsoEtAl & SRR5787176 & SRR5787176 & NA & NA & p83 & NA & NA & NA & NA & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t TsoEtAl & SRR5787177 & SRR5787177 & NA & NA & p23 & NA & NA & NA & NA & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A spec_tbl_df: 682 × 42\n",
       "\n",
       "| cohort &lt;chr&gt; | patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | tra_repertoire_id &lt;chr&gt; | trb_repertoire_id &lt;chr&gt; | rna_libraries &lt;chr&gt; | nanostring_libraries &lt;chr&gt; | gex_library &lt;chr&gt; | vdj_library &lt;chr&gt; | sc_library_count &lt;dbl&gt; | ⋯ ⋯ | DPB1allele2 &lt;chr&gt; | DQA1allele1 &lt;chr&gt; | DQA1allele2 &lt;chr&gt; | DQB1allele1 &lt;chr&gt; | DQB1allele2 &lt;chr&gt; | DRB1allele1 &lt;chr&gt; | DRB1allele2 &lt;chr&gt; | DRB345allele1 &lt;chr&gt; | DRB345allele2 &lt;chr&gt; | timestamp &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Hippos | 008_001 | 008_001_A | NA        | 008_001_A | NA           | 001-A | NA | NA | NA | ⋯ | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*04:01:01 | DQB1*04:02:01 | DQB1*05:01:01 | DRB1*03:02:01 | DRB1*11:01:02 | DRB3*01:01:02 | DRB3*03:01:01 | 1/12/23 |\n",
       "| Hippos | 008_001 | 008_001_B | 008_001_B | 008_001_B | 008-001-B-12 | 001-B | NA | NA | NA | ⋯ | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*04:01:01 | DQB1*04:02:01 | DQB1*05:01:01 | DRB1*03:02:01 | DRB1*11:01:02 | DRB3*01:01:02 | DRB3*03:01:01 | 1/12/23 |\n",
       "| Hippos | 008_001 | 008_001_C | NA        | NA        | 008-001-C-13 | NA    | NA | NA | NA | ⋯ | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*04:01:01 | DQB1*04:02:01 | DQB1*05:01:01 | DRB1*03:02:01 | DRB1*11:01:02 | DRB3*01:01:02 | DRB3*03:01:01 | 1/12/23 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| TsoEtAl | SRR5787176 | SRR5787176 | NA | NA | p83 | NA | NA | NA | NA | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| TsoEtAl | SRR5787177 | SRR5787177 | NA | NA | p23 | NA | NA | NA | NA | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "    cohort  patient_id repertoire_id tra_repertoire_id trb_repertoire_id\n",
       "1   Hippos  008_001    008_001_A     NA                008_001_A        \n",
       "2   Hippos  008_001    008_001_B     008_001_B         008_001_B        \n",
       "3   Hippos  008_001    008_001_C     NA                NA               \n",
       "⋮   ⋮       ⋮          ⋮             ⋮                 ⋮                \n",
       "681 TsoEtAl SRR5787176 SRR5787176    NA                NA               \n",
       "682 TsoEtAl SRR5787177 SRR5787177    NA                NA               \n",
       "    rna_libraries nanostring_libraries gex_library vdj_library sc_library_count\n",
       "1   NA            001-A                NA          NA          NA              \n",
       "2   008-001-B-12  001-B                NA          NA          NA              \n",
       "3   008-001-C-13  NA                   NA          NA          NA              \n",
       "⋮   ⋮             ⋮                    ⋮           ⋮           ⋮               \n",
       "681 p83           NA                   NA          NA          NA              \n",
       "682 p23           NA                   NA          NA          NA              \n",
       "    ⋯ DPB1allele2    DQA1allele1   DQA1allele2   DQB1allele1   DQB1allele2  \n",
       "1   ⋯ DPB1*105:01:01 DQA1*01:02:01 DQA1*04:01:01 DQB1*04:02:01 DQB1*05:01:01\n",
       "2   ⋯ DPB1*105:01:01 DQA1*01:02:01 DQA1*04:01:01 DQB1*04:02:01 DQB1*05:01:01\n",
       "3   ⋯ DPB1*105:01:01 DQA1*01:02:01 DQA1*04:01:01 DQB1*04:02:01 DQB1*05:01:01\n",
       "⋮   ⋱ ⋮              ⋮             ⋮             ⋮             ⋮            \n",
       "681 ⋯ NA             NA            NA            NA            NA           \n",
       "682 ⋯ NA             NA            NA            NA            NA           \n",
       "    DRB1allele1   DRB1allele2   DRB345allele1 DRB345allele2 timestamp\n",
       "1   DRB1*03:02:01 DRB1*11:01:02 DRB3*01:01:02 DRB3*03:01:01 1/12/23  \n",
       "2   DRB1*03:02:01 DRB1*11:01:02 DRB3*01:01:02 DRB3*03:01:01 1/12/23  \n",
       "3   DRB1*03:02:01 DRB1*11:01:02 DRB3*01:01:02 DRB3*03:01:01 1/12/23  \n",
       "⋮   ⋮             ⋮             ⋮             ⋮             ⋮        \n",
       "681 NA            NA            NA            NA            NA       \n",
       "682 NA            NA            NA            NA            NA       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hippos_sequencing_table <- readr::read_csv(hippos_metadata_path, \n",
    "  show_col_types = FALSE)\n",
    "hippos_sequencing_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge laboratory and sequencing metadata\n",
    "The sequencing data was obtained from a subset of tissue that was used to \n",
    "quantify the laboratory metrics. Therefore we can merge the two tables by \n",
    "patient ID and visit code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 682 × 46</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>tra_repertoire_id</th><th scope=col>trb_repertoire_id</th><th scope=col>rna_libraries</th><th scope=col>nanostring_libraries</th><th scope=col>gex_library</th><th scope=col>vdj_library</th><th scope=col>sc_library_count</th><th scope=col>⋯</th><th scope=col>DQB1allele2</th><th scope=col>DRB1allele1</th><th scope=col>DRB1allele2</th><th scope=col>DRB345allele1</th><th scope=col>DRB345allele2</th><th scope=col>timestamp</th><th scope=col>HIV</th><th scope=col>CD4_count</th><th scope=col>KSHV_plasma</th><th scope=col>phenotype</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_A</td><td>NA       </td><td>008_001_A</td><td>NA          </td><td>001-A</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DQB1*05:01:01</td><td>DRB1*03:02:01</td><td>DRB1*11:01:02</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td><td>1/12/23</td><td>276908</td><td>127</td><td>0</td><td>Epidemic KS</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_B</td><td>008_001_B</td><td>008_001_B</td><td>008-001-B-12</td><td>001-B</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DQB1*05:01:01</td><td>DRB1*03:02:01</td><td>DRB1*11:01:02</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td><td>1/12/23</td><td>276908</td><td>127</td><td>0</td><td>Epidemic KS</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_C</td><td>NA       </td><td>NA       </td><td>008-001-C-13</td><td>NA   </td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DQB1*05:01:01</td><td>DRB1*03:02:01</td><td>DRB1*11:01:02</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td><td>1/12/23</td><td>276908</td><td>127</td><td>0</td><td>Epidemic KS</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>TsoEtAl</td><td>SRR5787176</td><td>SRR5787176</td><td>NA</td><td>NA</td><td>p83</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>Epidemic KS</td></tr>\n",
       "\t<tr><td>TsoEtAl</td><td>SRR5787177</td><td>SRR5787177</td><td>NA</td><td>NA</td><td>p23</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>Epidemic KS</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 682 × 46\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " cohort & patient\\_id & repertoire\\_id & tra\\_repertoire\\_id & trb\\_repertoire\\_id & rna\\_libraries & nanostring\\_libraries & gex\\_library & vdj\\_library & sc\\_library\\_count & ⋯ & DQB1allele2 & DRB1allele1 & DRB1allele2 & DRB345allele1 & DRB345allele2 & timestamp & HIV & CD4\\_count & KSHV\\_plasma & phenotype\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_A & NA        & 008\\_001\\_A & NA           & 001-A & NA & NA & NA & ⋯ & DQB1*05:01:01 & DRB1*03:02:01 & DRB1*11:01:02 & DRB3*01:01:02 & DRB3*03:01:01 & 1/12/23 & 276908 & 127 & 0 & Epidemic KS\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_B & 008\\_001\\_B & 008\\_001\\_B & 008-001-B-12 & 001-B & NA & NA & NA & ⋯ & DQB1*05:01:01 & DRB1*03:02:01 & DRB1*11:01:02 & DRB3*01:01:02 & DRB3*03:01:01 & 1/12/23 & 276908 & 127 & 0 & Epidemic KS\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_C & NA        & NA        & 008-001-C-13 & NA    & NA & NA & NA & ⋯ & DQB1*05:01:01 & DRB1*03:02:01 & DRB1*11:01:02 & DRB3*01:01:02 & DRB3*03:01:01 & 1/12/23 & 276908 & 127 & 0 & Epidemic KS\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t TsoEtAl & SRR5787176 & SRR5787176 & NA & NA & p83 & NA & NA & NA & NA & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & Epidemic KS\\\\\n",
       "\t TsoEtAl & SRR5787177 & SRR5787177 & NA & NA & p23 & NA & NA & NA & NA & ⋯ & NA & NA & NA & NA & NA & NA & NA & NA & NA & Epidemic KS\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 682 × 46\n",
       "\n",
       "| cohort &lt;chr&gt; | patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | tra_repertoire_id &lt;chr&gt; | trb_repertoire_id &lt;chr&gt; | rna_libraries &lt;chr&gt; | nanostring_libraries &lt;chr&gt; | gex_library &lt;chr&gt; | vdj_library &lt;chr&gt; | sc_library_count &lt;dbl&gt; | ⋯ ⋯ | DQB1allele2 &lt;chr&gt; | DRB1allele1 &lt;chr&gt; | DRB1allele2 &lt;chr&gt; | DRB345allele1 &lt;chr&gt; | DRB345allele2 &lt;chr&gt; | timestamp &lt;chr&gt; | HIV &lt;dbl&gt; | CD4_count &lt;dbl&gt; | KSHV_plasma &lt;dbl&gt; | phenotype &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Hippos | 008_001 | 008_001_A | NA        | 008_001_A | NA           | 001-A | NA | NA | NA | ⋯ | DQB1*05:01:01 | DRB1*03:02:01 | DRB1*11:01:02 | DRB3*01:01:02 | DRB3*03:01:01 | 1/12/23 | 276908 | 127 | 0 | Epidemic KS |\n",
       "| Hippos | 008_001 | 008_001_B | 008_001_B | 008_001_B | 008-001-B-12 | 001-B | NA | NA | NA | ⋯ | DQB1*05:01:01 | DRB1*03:02:01 | DRB1*11:01:02 | DRB3*01:01:02 | DRB3*03:01:01 | 1/12/23 | 276908 | 127 | 0 | Epidemic KS |\n",
       "| Hippos | 008_001 | 008_001_C | NA        | NA        | 008-001-C-13 | NA    | NA | NA | NA | ⋯ | DQB1*05:01:01 | DRB1*03:02:01 | DRB1*11:01:02 | DRB3*01:01:02 | DRB3*03:01:01 | 1/12/23 | 276908 | 127 | 0 | Epidemic KS |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| TsoEtAl | SRR5787176 | SRR5787176 | NA | NA | p83 | NA | NA | NA | NA | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | Epidemic KS |\n",
       "| TsoEtAl | SRR5787177 | SRR5787177 | NA | NA | p23 | NA | NA | NA | NA | ⋯ | NA | NA | NA | NA | NA | NA | NA | NA | NA | Epidemic KS |\n",
       "\n"
      ],
      "text/plain": [
       "    cohort  patient_id repertoire_id tra_repertoire_id trb_repertoire_id\n",
       "1   Hippos  008_001    008_001_A     NA                008_001_A        \n",
       "2   Hippos  008_001    008_001_B     008_001_B         008_001_B        \n",
       "3   Hippos  008_001    008_001_C     NA                NA               \n",
       "⋮   ⋮       ⋮          ⋮             ⋮                 ⋮                \n",
       "681 TsoEtAl SRR5787176 SRR5787176    NA                NA               \n",
       "682 TsoEtAl SRR5787177 SRR5787177    NA                NA               \n",
       "    rna_libraries nanostring_libraries gex_library vdj_library sc_library_count\n",
       "1   NA            001-A                NA          NA          NA              \n",
       "2   008-001-B-12  001-B                NA          NA          NA              \n",
       "3   008-001-C-13  NA                   NA          NA          NA              \n",
       "⋮   ⋮             ⋮                    ⋮           ⋮           ⋮               \n",
       "681 p83           NA                   NA          NA          NA              \n",
       "682 p23           NA                   NA          NA          NA              \n",
       "    ⋯ DQB1allele2   DRB1allele1   DRB1allele2   DRB345allele1 DRB345allele2\n",
       "1   ⋯ DQB1*05:01:01 DRB1*03:02:01 DRB1*11:01:02 DRB3*01:01:02 DRB3*03:01:01\n",
       "2   ⋯ DQB1*05:01:01 DRB1*03:02:01 DRB1*11:01:02 DRB3*01:01:02 DRB3*03:01:01\n",
       "3   ⋯ DQB1*05:01:01 DRB1*03:02:01 DRB1*11:01:02 DRB3*01:01:02 DRB3*03:01:01\n",
       "⋮   ⋱ ⋮             ⋮             ⋮             ⋮             ⋮            \n",
       "681 ⋯ NA            NA            NA            NA            NA           \n",
       "682 ⋯ NA            NA            NA            NA            NA           \n",
       "    timestamp HIV    CD4_count KSHV_plasma phenotype  \n",
       "1   1/12/23   276908 127       0           Epidemic KS\n",
       "2   1/12/23   276908 127       0           Epidemic KS\n",
       "3   1/12/23   276908 127       0           Epidemic KS\n",
       "⋮   ⋮         ⋮      ⋮         ⋮           ⋮          \n",
       "681 NA        NA     NA        NA          Epidemic KS\n",
       "682 NA        NA     NA        NA          Epidemic KS"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hippos_metadata_table <- left_join(hippos_sequencing_table, hippos_titre_table_wide,\n",
    "    by = c(\"patient_id\", \"visit_code\")) |>\n",
    "  mutate(phenotype = case_when(\n",
    "    hiv_status == \"Positive\" & ks_status == \"Positive\" ~ \"Epidemic KS\",\n",
    "    hiv_status == \"Negative\" & ks_status == \"Positive\" ~ \"Endemic KS\",\n",
    "    hiv_status == \"Negative\" & ks_status == \"Negative\" ~ \"Control\"))\n",
    "hippos_metadata_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BL metadata\n",
    "As a way to benchmark the T-cell responses in KS, we compare the diversity of \n",
    "the T-cell repertoire of the KS TME with the repertoire of other Pathogen\n",
    "associated malingnancies like BL. Here we use two datasets of pediatric BL from\n",
    "Ghana and Uganda. But before we load the datasets, lets format the metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>trb_repertoire_id</th><th scope=col>tissue_type</th><th scope=col>patient_id</th><th scope=col>cohort</th><th scope=col>phenotype</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>H003840_01</td><td>Tumor</td><td>H003840_01</td><td>BL - Ghana</td><td>BL</td></tr>\n",
       "\t<tr><td>H003840_02</td><td>Tumor</td><td>H003840_02</td><td>BL - Ghana</td><td>BL</td></tr>\n",
       "\t<tr><td>H031201_02</td><td>Tumor</td><td>H031201_02</td><td>BL - Ghana</td><td>BL</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>H057518_02</td><td>Tumor</td><td>H057518_02</td><td>BL - Ghana</td><td>BL</td></tr>\n",
       "\t<tr><td>H057519_01</td><td>Tumor</td><td>H057519_01</td><td>BL - Ghana</td><td>BL</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " trb\\_repertoire\\_id & tissue\\_type & patient\\_id & cohort & phenotype\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t H003840\\_01 & Tumor & H003840\\_01 & BL - Ghana & BL\\\\\n",
       "\t H003840\\_02 & Tumor & H003840\\_02 & BL - Ghana & BL\\\\\n",
       "\t H031201\\_02 & Tumor & H031201\\_02 & BL - Ghana & BL\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t H057518\\_02 & Tumor & H057518\\_02 & BL - Ghana & BL\\\\\n",
       "\t H057519\\_01 & Tumor & H057519\\_01 & BL - Ghana & BL\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 5\n",
       "\n",
       "| trb_repertoire_id &lt;chr&gt; | tissue_type &lt;chr&gt; | patient_id &lt;chr&gt; | cohort &lt;chr&gt; | phenotype &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| H003840_01 | Tumor | H003840_01 | BL - Ghana | BL |\n",
       "| H003840_02 | Tumor | H003840_02 | BL - Ghana | BL |\n",
       "| H031201_02 | Tumor | H031201_02 | BL - Ghana | BL |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| H057518_02 | Tumor | H057518_02 | BL - Ghana | BL |\n",
       "| H057519_01 | Tumor | H057519_01 | BL - Ghana | BL |\n",
       "\n"
      ],
      "text/plain": [
       "  trb_repertoire_id tissue_type patient_id cohort     phenotype\n",
       "1 H003840_01        Tumor       H003840_01 BL - Ghana BL       \n",
       "2 H003840_02        Tumor       H003840_02 BL - Ghana BL       \n",
       "3 H031201_02        Tumor       H031201_02 BL - Ghana BL       \n",
       "⋮ ⋮                 ⋮           ⋮          ⋮          ⋮        \n",
       "5 H057518_02        Tumor       H057518_02 BL - Ghana BL       \n",
       "6 H057519_01        Tumor       H057519_01 BL - Ghana BL       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "blghana_metadata_table <- read_csv(blghana_metadata_path, show_col_types = FALSE) |>\n",
    "  dplyr::select(fileName, tissueType) |> \n",
    "  dplyr::rename(trb_repertoire_id = fileName, tissue_type = tissueType) |>\n",
    "  mutate(patient_id = str_extract(trb_repertoire_id, \"\\\\w+\"),\n",
    "    cohort = \"BL - Ghana\",\n",
    "    phenotype = \"BL\") \n",
    "head(blghana_metadata_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 29</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>trb_repertoire_id</th><th scope=col>patient_id</th><th scope=col>tissue_type</th><th scope=col>aAllele1</th><th scope=col>aAllele2</th><th scope=col>bAllele1</th><th scope=col>bAllele2</th><th scope=col>cAllele1</th><th scope=col>cAllele2</th><th scope=col>DPA1allele1</th><th scope=col>⋯</th><th scope=col>DRB345allele1</th><th scope=col>DRB345allele2</th><th scope=col>hiv_status</th><th scope=col>ebv_status</th><th scope=col>bl_status</th><th scope=col>died</th><th scope=col>gender</th><th scope=col>age</th><th scope=col>cohort</th><th scope=col>phenotype</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>MVQ80482a</td><td>009_0103</td><td>Tumor</td><td>A*02:05:01</td><td>A*74:01   </td><td>B*45:01:01</td><td>B*58:02   </td><td>C*06:02:01</td><td>C*06:02:01</td><td>DPA1*01:03:01</td><td>⋯</td><td>DRB3*03:01:01</td><td>DRB4*01:03:01</td><td>Negative</td><td>Positive</td><td>Positive</td><td>No </td><td>Male  </td><td>10</td><td>BL - Uganda</td><td>BL</td></tr>\n",
       "\t<tr><td>MVQ81040A</td><td>009_0109</td><td>Tumor</td><td>A*02:01:01</td><td>A*30:01:01</td><td>B*39:10:01</td><td>B*45:01:01</td><td>C*12:03:01</td><td>C*16:01:01</td><td>DPA1*01:03:01</td><td>⋯</td><td>NA           </td><td>NA           </td><td>Negative</td><td>Positive</td><td>Positive</td><td>No </td><td>Male  </td><td>12</td><td>BL - Uganda</td><td>BL</td></tr>\n",
       "\t<tr><td>MVQ81231A</td><td>009_0112</td><td>Tumor</td><td>A*01:01:01</td><td>A*74:01   </td><td>B*45:01:01</td><td>B*58:01:01</td><td>C*03:02:02</td><td>C*06:02:01</td><td>DPA1*01:03:01</td><td>⋯</td><td>DRB3*02:02:01</td><td>DRB3*03:01:01</td><td>Negative</td><td>Negative</td><td>Positive</td><td>Yes</td><td>Female</td><td> 7</td><td>BL - Uganda</td><td>BL</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>MVQ90143A</td><td>009_0128</td><td>Tumor</td><td>A*31:04</td><td>A*66:02</td><td>B*15:03:01</td><td>B*81:01</td><td>C*08:02:01</td><td>C*08:04:01</td><td>DPA1*01:03:01</td><td>⋯</td><td>DRB4*01:01:01</td><td>DRB5*01:01:01</td><td>Positive</td><td>Negative</td><td>Positive</td><td>Yes</td><td>Male  </td><td>5</td><td>BL - Uganda</td><td>BL</td></tr>\n",
       "\t<tr><td>MVQ91163A</td><td>009_0145</td><td>Tumor</td><td>NA     </td><td>NA     </td><td>NA        </td><td>NA     </td><td>NA        </td><td>NA        </td><td>NA           </td><td>⋯</td><td>NA           </td><td>NA           </td><td>Negative</td><td>Negative</td><td>Negative</td><td>Yes</td><td>Female</td><td>8</td><td>BL - Uganda</td><td>BL</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 29\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " trb\\_repertoire\\_id & patient\\_id & tissue\\_type & aAllele1 & aAllele2 & bAllele1 & bAllele2 & cAllele1 & cAllele2 & DPA1allele1 & ⋯ & DRB345allele1 & DRB345allele2 & hiv\\_status & ebv\\_status & bl\\_status & died & gender & age & cohort & phenotype\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t MVQ80482a & 009\\_0103 & Tumor & A*02:05:01 & A*74:01    & B*45:01:01 & B*58:02    & C*06:02:01 & C*06:02:01 & DPA1*01:03:01 & ⋯ & DRB3*03:01:01 & DRB4*01:03:01 & Negative & Positive & Positive & No  & Male   & 10 & BL - Uganda & BL\\\\\n",
       "\t MVQ81040A & 009\\_0109 & Tumor & A*02:01:01 & A*30:01:01 & B*39:10:01 & B*45:01:01 & C*12:03:01 & C*16:01:01 & DPA1*01:03:01 & ⋯ & NA            & NA            & Negative & Positive & Positive & No  & Male   & 12 & BL - Uganda & BL\\\\\n",
       "\t MVQ81231A & 009\\_0112 & Tumor & A*01:01:01 & A*74:01    & B*45:01:01 & B*58:01:01 & C*03:02:02 & C*06:02:01 & DPA1*01:03:01 & ⋯ & DRB3*02:02:01 & DRB3*03:01:01 & Negative & Negative & Positive & Yes & Female &  7 & BL - Uganda & BL\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t MVQ90143A & 009\\_0128 & Tumor & A*31:04 & A*66:02 & B*15:03:01 & B*81:01 & C*08:02:01 & C*08:04:01 & DPA1*01:03:01 & ⋯ & DRB4*01:01:01 & DRB5*01:01:01 & Positive & Negative & Positive & Yes & Male   & 5 & BL - Uganda & BL\\\\\n",
       "\t MVQ91163A & 009\\_0145 & Tumor & NA      & NA      & NA         & NA      & NA         & NA         & NA            & ⋯ & NA            & NA            & Negative & Negative & Negative & Yes & Female & 8 & BL - Uganda & BL\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 29\n",
       "\n",
       "| trb_repertoire_id &lt;chr&gt; | patient_id &lt;chr&gt; | tissue_type &lt;chr&gt; | aAllele1 &lt;chr&gt; | aAllele2 &lt;chr&gt; | bAllele1 &lt;chr&gt; | bAllele2 &lt;chr&gt; | cAllele1 &lt;chr&gt; | cAllele2 &lt;chr&gt; | DPA1allele1 &lt;chr&gt; | ⋯ ⋯ | DRB345allele1 &lt;chr&gt; | DRB345allele2 &lt;chr&gt; | hiv_status &lt;chr&gt; | ebv_status &lt;chr&gt; | bl_status &lt;chr&gt; | died &lt;chr&gt; | gender &lt;chr&gt; | age &lt;dbl&gt; | cohort &lt;chr&gt; | phenotype &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| MVQ80482a | 009_0103 | Tumor | A*02:05:01 | A*74:01    | B*45:01:01 | B*58:02    | C*06:02:01 | C*06:02:01 | DPA1*01:03:01 | ⋯ | DRB3*03:01:01 | DRB4*01:03:01 | Negative | Positive | Positive | No  | Male   | 10 | BL - Uganda | BL |\n",
       "| MVQ81040A | 009_0109 | Tumor | A*02:01:01 | A*30:01:01 | B*39:10:01 | B*45:01:01 | C*12:03:01 | C*16:01:01 | DPA1*01:03:01 | ⋯ | NA            | NA            | Negative | Positive | Positive | No  | Male   | 12 | BL - Uganda | BL |\n",
       "| MVQ81231A | 009_0112 | Tumor | A*01:01:01 | A*74:01    | B*45:01:01 | B*58:01:01 | C*03:02:02 | C*06:02:01 | DPA1*01:03:01 | ⋯ | DRB3*02:02:01 | DRB3*03:01:01 | Negative | Negative | Positive | Yes | Female |  7 | BL - Uganda | BL |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| MVQ90143A | 009_0128 | Tumor | A*31:04 | A*66:02 | B*15:03:01 | B*81:01 | C*08:02:01 | C*08:04:01 | DPA1*01:03:01 | ⋯ | DRB4*01:01:01 | DRB5*01:01:01 | Positive | Negative | Positive | Yes | Male   | 5 | BL - Uganda | BL |\n",
       "| MVQ91163A | 009_0145 | Tumor | NA      | NA      | NA         | NA      | NA         | NA         | NA            | ⋯ | NA            | NA            | Negative | Negative | Negative | Yes | Female | 8 | BL - Uganda | BL |\n",
       "\n"
      ],
      "text/plain": [
       "  trb_repertoire_id patient_id tissue_type aAllele1   aAllele2   bAllele1  \n",
       "1 MVQ80482a         009_0103   Tumor       A*02:05:01 A*74:01    B*45:01:01\n",
       "2 MVQ81040A         009_0109   Tumor       A*02:01:01 A*30:01:01 B*39:10:01\n",
       "3 MVQ81231A         009_0112   Tumor       A*01:01:01 A*74:01    B*45:01:01\n",
       "⋮ ⋮                 ⋮          ⋮           ⋮          ⋮          ⋮         \n",
       "5 MVQ90143A         009_0128   Tumor       A*31:04    A*66:02    B*15:03:01\n",
       "6 MVQ91163A         009_0145   Tumor       NA         NA         NA        \n",
       "  bAllele2   cAllele1   cAllele2   DPA1allele1   ⋯ DRB345allele1 DRB345allele2\n",
       "1 B*58:02    C*06:02:01 C*06:02:01 DPA1*01:03:01 ⋯ DRB3*03:01:01 DRB4*01:03:01\n",
       "2 B*45:01:01 C*12:03:01 C*16:01:01 DPA1*01:03:01 ⋯ NA            NA           \n",
       "3 B*58:01:01 C*03:02:02 C*06:02:01 DPA1*01:03:01 ⋯ DRB3*02:02:01 DRB3*03:01:01\n",
       "⋮ ⋮          ⋮          ⋮          ⋮             ⋱ ⋮             ⋮            \n",
       "5 B*81:01    C*08:02:01 C*08:04:01 DPA1*01:03:01 ⋯ DRB4*01:01:01 DRB5*01:01:01\n",
       "6 NA         NA         NA         NA            ⋯ NA            NA           \n",
       "  hiv_status ebv_status bl_status died gender age cohort      phenotype\n",
       "1 Negative   Positive   Positive  No   Male   10  BL - Uganda BL       \n",
       "2 Negative   Positive   Positive  No   Male   12  BL - Uganda BL       \n",
       "3 Negative   Negative   Positive  Yes  Female  7  BL - Uganda BL       \n",
       "⋮ ⋮          ⋮          ⋮         ⋮    ⋮      ⋮   ⋮           ⋮        \n",
       "5 Positive   Negative   Positive  Yes  Male   5   BL - Uganda BL       \n",
       "6 Negative   Negative   Negative  Yes  Female 8   BL - Uganda BL       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bluganda_metadata_table <- read_csv(bluganda_metadata_path,\n",
    "  show_col_types = FALSE, name_repair = \"unique_quiet\") |>\n",
    "  dplyr::select(samples, patientID, tissueType, aAllele1:DRB345allele2, HIVStatus,\n",
    "    EBVstatus, BL, vitalStatus, sex, age) |> \n",
    "  dplyr::rename(trb_repertoire_id = samples, tissue_type = tissueType,\n",
    "    hiv_status = HIVStatus, ebv_status = EBVstatus, bl_status = BL,\n",
    "    died = vitalStatus, gender = sex, patient_id = patientID) |>\n",
    "  mutate(bl_status = recode(bl_status, \"Yes\" = \"Positive\", \"No\" = \"Negative\"),\n",
    "    died = recode(died, \"Alive\" = \"No\", \"Dead\" = \"Yes\"),\n",
    "    patient_id = as.character(patient_id),\n",
    "    cohort = \"BL - Uganda\",\n",
    "    phenotype = \"BL\") \n",
    "head(bluganda_metadata_table)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge metadata\n",
    "\n",
    "In the present study we will use multiple sources of data:\n",
    "1. AIRR-Seq, RNA-Seq and Nanostring data from the HIPPOS cohort\n",
    "2. RNA-Seq data from Tso et al.,\n",
    "3. RNA-Seq data from Lidenge et al.,\n",
    "4. AIRR-Seq data from pediatric BL from Ghana and Uganda\n",
    "\n",
    "In the next few codeblock we will compile a joint `study_metdata` and summarize\n",
    "the samples counts of the study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 748 × 49</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>tra_repertoire_id</th><th scope=col>trb_repertoire_id</th><th scope=col>rna_libraries</th><th scope=col>nanostring_libraries</th><th scope=col>gex_library</th><th scope=col>vdj_library</th><th scope=col>sc_library_count</th><th scope=col>⋯</th><th scope=col>DRB345allele1</th><th scope=col>DRB345allele2</th><th scope=col>timestamp</th><th scope=col>HIV</th><th scope=col>CD4_count</th><th scope=col>KSHV_plasma</th><th scope=col>phenotype</th><th scope=col>DPBallele2</th><th scope=col>ebv_status</th><th scope=col>bl_status</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_A</td><td>NA       </td><td>008_001_A</td><td>NA          </td><td>001-A</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td><td>1/12/23</td><td>276908</td><td>127</td><td>0</td><td>Epidemic KS</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_B</td><td>008_001_B</td><td>008_001_B</td><td>008-001-B-12</td><td>001-B</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td><td>1/12/23</td><td>276908</td><td>127</td><td>0</td><td>Epidemic KS</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_C</td><td>NA       </td><td>NA       </td><td>008-001-C-13</td><td>NA   </td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td><td>1/12/23</td><td>276908</td><td>127</td><td>0</td><td>Epidemic KS</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>BL - Uganda</td><td>009_0232</td><td>MVQ108912A</td><td>NA</td><td>MVQ108912A</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DRB3*02:02:01</td><td>DRB3*02:02:01</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>BL</td><td>DPB1*105:01</td><td>Positive</td><td>Positive</td></tr>\n",
       "\t<tr><td>BL - Uganda</td><td>009_0249</td><td>MVQ194745A</td><td>NA</td><td>MVQ194745A</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>DRB3*01:01:02</td><td>DRB3*02:02:01</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>BL</td><td>DPB1*105:01</td><td>Positive</td><td>Positive</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 748 × 49\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " cohort & patient\\_id & repertoire\\_id & tra\\_repertoire\\_id & trb\\_repertoire\\_id & rna\\_libraries & nanostring\\_libraries & gex\\_library & vdj\\_library & sc\\_library\\_count & ⋯ & DRB345allele1 & DRB345allele2 & timestamp & HIV & CD4\\_count & KSHV\\_plasma & phenotype & DPBallele2 & ebv\\_status & bl\\_status\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & ⋯ & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_A & NA        & 008\\_001\\_A & NA           & 001-A & NA & NA & NA & ⋯ & DRB3*01:01:02 & DRB3*03:01:01 & 1/12/23 & 276908 & 127 & 0 & Epidemic KS & NA & NA & NA\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_B & 008\\_001\\_B & 008\\_001\\_B & 008-001-B-12 & 001-B & NA & NA & NA & ⋯ & DRB3*01:01:02 & DRB3*03:01:01 & 1/12/23 & 276908 & 127 & 0 & Epidemic KS & NA & NA & NA\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_C & NA        & NA        & 008-001-C-13 & NA    & NA & NA & NA & ⋯ & DRB3*01:01:02 & DRB3*03:01:01 & 1/12/23 & 276908 & 127 & 0 & Epidemic KS & NA & NA & NA\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t BL - Uganda & 009\\_0232 & MVQ108912A & NA & MVQ108912A & NA & NA & NA & NA & NA & ⋯ & DRB3*02:02:01 & DRB3*02:02:01 & NA & NA & NA & NA & BL & DPB1*105:01 & Positive & Positive\\\\\n",
       "\t BL - Uganda & 009\\_0249 & MVQ194745A & NA & MVQ194745A & NA & NA & NA & NA & NA & ⋯ & DRB3*01:01:02 & DRB3*02:02:01 & NA & NA & NA & NA & BL & DPB1*105:01 & Positive & Positive\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 748 × 49\n",
       "\n",
       "| cohort &lt;chr&gt; | patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | tra_repertoire_id &lt;chr&gt; | trb_repertoire_id &lt;chr&gt; | rna_libraries &lt;chr&gt; | nanostring_libraries &lt;chr&gt; | gex_library &lt;chr&gt; | vdj_library &lt;chr&gt; | sc_library_count &lt;dbl&gt; | ⋯ ⋯ | DRB345allele1 &lt;chr&gt; | DRB345allele2 &lt;chr&gt; | timestamp &lt;chr&gt; | HIV &lt;dbl&gt; | CD4_count &lt;dbl&gt; | KSHV_plasma &lt;dbl&gt; | phenotype &lt;chr&gt; | DPBallele2 &lt;chr&gt; | ebv_status &lt;chr&gt; | bl_status &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Hippos | 008_001 | 008_001_A | NA        | 008_001_A | NA           | 001-A | NA | NA | NA | ⋯ | DRB3*01:01:02 | DRB3*03:01:01 | 1/12/23 | 276908 | 127 | 0 | Epidemic KS | NA | NA | NA |\n",
       "| Hippos | 008_001 | 008_001_B | 008_001_B | 008_001_B | 008-001-B-12 | 001-B | NA | NA | NA | ⋯ | DRB3*01:01:02 | DRB3*03:01:01 | 1/12/23 | 276908 | 127 | 0 | Epidemic KS | NA | NA | NA |\n",
       "| Hippos | 008_001 | 008_001_C | NA        | NA        | 008-001-C-13 | NA    | NA | NA | NA | ⋯ | DRB3*01:01:02 | DRB3*03:01:01 | 1/12/23 | 276908 | 127 | 0 | Epidemic KS | NA | NA | NA |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| BL - Uganda | 009_0232 | MVQ108912A | NA | MVQ108912A | NA | NA | NA | NA | NA | ⋯ | DRB3*02:02:01 | DRB3*02:02:01 | NA | NA | NA | NA | BL | DPB1*105:01 | Positive | Positive |\n",
       "| BL - Uganda | 009_0249 | MVQ194745A | NA | MVQ194745A | NA | NA | NA | NA | NA | ⋯ | DRB3*01:01:02 | DRB3*02:02:01 | NA | NA | NA | NA | BL | DPB1*105:01 | Positive | Positive |\n",
       "\n"
      ],
      "text/plain": [
       "    cohort      patient_id repertoire_id tra_repertoire_id trb_repertoire_id\n",
       "1   Hippos      008_001    008_001_A     NA                008_001_A        \n",
       "2   Hippos      008_001    008_001_B     008_001_B         008_001_B        \n",
       "3   Hippos      008_001    008_001_C     NA                NA               \n",
       "⋮   ⋮           ⋮          ⋮             ⋮                 ⋮                \n",
       "747 BL - Uganda 009_0232   MVQ108912A    NA                MVQ108912A       \n",
       "748 BL - Uganda 009_0249   MVQ194745A    NA                MVQ194745A       \n",
       "    rna_libraries nanostring_libraries gex_library vdj_library sc_library_count\n",
       "1   NA            001-A                NA          NA          NA              \n",
       "2   008-001-B-12  001-B                NA          NA          NA              \n",
       "3   008-001-C-13  NA                   NA          NA          NA              \n",
       "⋮   ⋮             ⋮                    ⋮           ⋮           ⋮               \n",
       "747 NA            NA                   NA          NA          NA              \n",
       "748 NA            NA                   NA          NA          NA              \n",
       "    ⋯ DRB345allele1 DRB345allele2 timestamp HIV    CD4_count KSHV_plasma\n",
       "1   ⋯ DRB3*01:01:02 DRB3*03:01:01 1/12/23   276908 127       0          \n",
       "2   ⋯ DRB3*01:01:02 DRB3*03:01:01 1/12/23   276908 127       0          \n",
       "3   ⋯ DRB3*01:01:02 DRB3*03:01:01 1/12/23   276908 127       0          \n",
       "⋮   ⋱ ⋮             ⋮             ⋮         ⋮      ⋮         ⋮          \n",
       "747 ⋯ DRB3*02:02:01 DRB3*02:02:01 NA        NA     NA        NA         \n",
       "748 ⋯ DRB3*01:01:02 DRB3*02:02:01 NA        NA     NA        NA         \n",
       "    phenotype   DPBallele2  ebv_status bl_status\n",
       "1   Epidemic KS NA          NA         NA       \n",
       "2   Epidemic KS NA          NA         NA       \n",
       "3   Epidemic KS NA          NA         NA       \n",
       "⋮   ⋮           ⋮           ⋮          ⋮        \n",
       "747 BL          DPB1*105:01 Positive   Positive \n",
       "748 BL          DPB1*105:01 Positive   Positive "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "study_metadata <- bind_rows(hippos_metadata_table, blghana_metadata_table,\n",
    "  bluganda_metadata_table) |> \n",
    "  mutate(repertoire_id = if_else(is.na(repertoire_id), trb_repertoire_id, \n",
    "    repertoire_id))\n",
    "study_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 1: Viral gene expression, and immune cell type abundance in KS tumor and control tissue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare HIV gene expression table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>NP_057849.4</dt><dd>'Gag-Pol'</dd><dt>NP_057850.1</dt><dd>'Pr55(Gag)'</dd><dt>NP_057851.1</dt><dd>'Vif'</dd><dt>NP_057852.2</dt><dd>'Vpr'</dd><dt>NP_057853.1</dt><dd>'Tat'</dd><dt>NP_057854.1</dt><dd>'Rev'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[NP\\textbackslash{}\\_057849.4] 'Gag-Pol'\n",
       "\\item[NP\\textbackslash{}\\_057850.1] 'Pr55(Gag)'\n",
       "\\item[NP\\textbackslash{}\\_057851.1] 'Vif'\n",
       "\\item[NP\\textbackslash{}\\_057852.2] 'Vpr'\n",
       "\\item[NP\\textbackslash{}\\_057853.1] 'Tat'\n",
       "\\item[NP\\textbackslash{}\\_057854.1] 'Rev'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "NP_057849.4\n",
       ":   'Gag-Pol'NP_057850.1\n",
       ":   'Pr55(Gag)'NP_057851.1\n",
       ":   'Vif'NP_057852.2\n",
       ":   'Vpr'NP_057853.1\n",
       ":   'Tat'NP_057854.1\n",
       ":   'Rev'\n",
       "\n"
      ],
      "text/plain": [
       "NP_057849.4 NP_057850.1 NP_057851.1 NP_057852.2 NP_057853.1 NP_057854.1 \n",
       "  \"Gag-Pol\" \"Pr55(Gag)\"       \"Vif\"       \"Vpr\"       \"Tat\"       \"Rev\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "suppressWarnings({\n",
    "hiv_dictionary <- read.table(hiv_fasta_path, sep=\"\\n\")$V1[grep(\">\", readLines(hiv_fasta_path))] |>\n",
    "  as_tibble() |>\n",
    "  separate(value, into = c(\"pid\", \"gene_name\", \"org\"), sep = \"\\\\s\") |>\n",
    "  mutate(pid = str_remove(pid, \">\"),\n",
    "    gene_name = if_else(gene_name == \"Envelope\", \"Env\", gene_name)) |>\n",
    "    pull(gene_name)\n",
    "names(hiv_dictionary) <- read.table(hiv_fasta_path, sep=\"\\n\")$V1[grep(\">\", readLines(hiv_fasta_path))] |>\n",
    "  as_tibble() |>\n",
    "  separate(value, into = c(\"pid\", \"gene_name\", \"org\"), sep = \"\\\\s\") |>\n",
    "  mutate(pid = str_remove(pid, \">\"),\n",
    "    gene_name = if_else(gene_name == \"Envelope\", \"Env\", gene_name)) |>\n",
    "    pull(pid)\n",
    "head(hiv_dictionary)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 44</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Name</th><th scope=col>008_001_B</th><th scope=col>008_001_C</th><th scope=col>008_001_I</th><th scope=col>008_003_C</th><th scope=col>008_004_D</th><th scope=col>008_005_C</th><th scope=col>008_006_B</th><th scope=col>008_006_C</th><th scope=col>008_007_B</th><th scope=col>⋯</th><th scope=col>008_029_C</th><th scope=col>008_030_B</th><th scope=col>008_030_C</th><th scope=col>008_032_B</th><th scope=col>008_034_B</th><th scope=col>008_037_B</th><th scope=col>008_122_B</th><th scope=col>008_140_B</th><th scope=col>008_175_B</th><th scope=col>008_211_D</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Vpr</td><td>    0.0</td><td>0</td><td>109427.0</td><td>     0</td><td>     0.0</td><td>0</td><td>     0</td><td>146631</td><td>108206.0</td><td>⋯</td><td>146745</td><td> 36466.7</td><td>532094</td><td> 77129.1</td><td>     0</td><td>73775.5</td><td>     0</td><td>821708</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>Tat</td><td>57184.7</td><td>0</td><td> 55645.2</td><td>658517</td><td>392418.0</td><td>0</td><td>     0</td><td>     0</td><td>     0.0</td><td>⋯</td><td>227184</td><td>142175.0</td><td>     0</td><td>142841.0</td><td>     0</td><td>20955.2</td><td>     0</td><td>     0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>Vif</td><td>    0.0</td><td>0</td><td> 57443.8</td><td>137771</td><td> 52443.8</td><td>0</td><td>226270</td><td>203311</td><td> 88059.1</td><td>⋯</td><td> 42024</td><td> 92065.8</td><td>     0</td><td> 41340.5</td><td>292284</td><td>53127.4</td><td>436469</td><td>     0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Nef</td><td>    0.0</td><td>0</td><td>285122</td><td>62371.1</td><td>430348</td><td>     0</td><td>0</td><td>92812.1</td><td>258860</td><td>⋯</td><td>280027</td><td>337332</td><td>191108</td><td>354976</td><td>     0</td><td>395778</td><td>387058.0</td><td>     0</td><td>567126</td><td>     0</td></tr>\n",
       "\t<tr><td>Env</td><td>40654.9</td><td>0</td><td>105637</td><td>80992.5</td><td> 56292</td><td>634090</td><td>0</td><td>80359.4</td><td>125382</td><td>⋯</td><td>126295</td><td>116803</td><td>159696</td><td>144293</td><td>113451</td><td>107418</td><td> 61869.2</td><td>178292</td><td>201805</td><td>777618</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 44\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Name & 008\\_001\\_B & 008\\_001\\_C & 008\\_001\\_I & 008\\_003\\_C & 008\\_004\\_D & 008\\_005\\_C & 008\\_006\\_B & 008\\_006\\_C & 008\\_007\\_B & ⋯ & 008\\_029\\_C & 008\\_030\\_B & 008\\_030\\_C & 008\\_032\\_B & 008\\_034\\_B & 008\\_037\\_B & 008\\_122\\_B & 008\\_140\\_B & 008\\_175\\_B & 008\\_211\\_D\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Vpr &     0.0 & 0 & 109427.0 &      0 &      0.0 & 0 &      0 & 146631 & 108206.0 & ⋯ & 146745 &  36466.7 & 532094 &  77129.1 &      0 & 73775.5 &      0 & 821708 & 0 & 0\\\\\n",
       "\t Tat & 57184.7 & 0 &  55645.2 & 658517 & 392418.0 & 0 &      0 &      0 &      0.0 & ⋯ & 227184 & 142175.0 &      0 & 142841.0 &      0 & 20955.2 &      0 &      0 & 0 & 0\\\\\n",
       "\t Vif &     0.0 & 0 &  57443.8 & 137771 &  52443.8 & 0 & 226270 & 203311 &  88059.1 & ⋯ &  42024 &  92065.8 &      0 &  41340.5 & 292284 & 53127.4 & 436469 &      0 & 0 & 0\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t Nef &     0.0 & 0 & 285122 & 62371.1 & 430348 &      0 & 0 & 92812.1 & 258860 & ⋯ & 280027 & 337332 & 191108 & 354976 &      0 & 395778 & 387058.0 &      0 & 567126 &      0\\\\\n",
       "\t Env & 40654.9 & 0 & 105637 & 80992.5 &  56292 & 634090 & 0 & 80359.4 & 125382 & ⋯ & 126295 & 116803 & 159696 & 144293 & 113451 & 107418 &  61869.2 & 178292 & 201805 & 777618\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 44\n",
       "\n",
       "| Name &lt;chr&gt; | 008_001_B &lt;dbl&gt; | 008_001_C &lt;dbl&gt; | 008_001_I &lt;dbl&gt; | 008_003_C &lt;dbl&gt; | 008_004_D &lt;dbl&gt; | 008_005_C &lt;dbl&gt; | 008_006_B &lt;dbl&gt; | 008_006_C &lt;dbl&gt; | 008_007_B &lt;dbl&gt; | ⋯ ⋯ | 008_029_C &lt;dbl&gt; | 008_030_B &lt;dbl&gt; | 008_030_C &lt;dbl&gt; | 008_032_B &lt;dbl&gt; | 008_034_B &lt;dbl&gt; | 008_037_B &lt;dbl&gt; | 008_122_B &lt;dbl&gt; | 008_140_B &lt;dbl&gt; | 008_175_B &lt;dbl&gt; | 008_211_D &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Vpr |     0.0 | 0 | 109427.0 |      0 |      0.0 | 0 |      0 | 146631 | 108206.0 | ⋯ | 146745 |  36466.7 | 532094 |  77129.1 |      0 | 73775.5 |      0 | 821708 | 0 | 0 |\n",
       "| Tat | 57184.7 | 0 |  55645.2 | 658517 | 392418.0 | 0 |      0 |      0 |      0.0 | ⋯ | 227184 | 142175.0 |      0 | 142841.0 |      0 | 20955.2 |      0 |      0 | 0 | 0 |\n",
       "| Vif |     0.0 | 0 |  57443.8 | 137771 |  52443.8 | 0 | 226270 | 203311 |  88059.1 | ⋯ |  42024 |  92065.8 |      0 |  41340.5 | 292284 | 53127.4 | 436469 |      0 | 0 | 0 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| Nef |     0.0 | 0 | 285122 | 62371.1 | 430348 |      0 | 0 | 92812.1 | 258860 | ⋯ | 280027 | 337332 | 191108 | 354976 |      0 | 395778 | 387058.0 |      0 | 567126 |      0 |\n",
       "| Env | 40654.9 | 0 | 105637 | 80992.5 |  56292 | 634090 | 0 | 80359.4 | 125382 | ⋯ | 126295 | 116803 | 159696 | 144293 | 113451 | 107418 |  61869.2 | 178292 | 201805 | 777618 |\n",
       "\n"
      ],
      "text/plain": [
       "  Name 008_001_B 008_001_C 008_001_I 008_003_C 008_004_D 008_005_C 008_006_B\n",
       "1 Vpr      0.0   0         109427.0       0         0.0  0              0   \n",
       "2 Tat  57184.7   0          55645.2  658517    392418.0  0              0   \n",
       "3 Vif      0.0   0          57443.8  137771     52443.8  0         226270   \n",
       "⋮ ⋮    ⋮         ⋮         ⋮         ⋮         ⋮         ⋮         ⋮        \n",
       "5 Nef      0.0   0         285122    62371.1   430348         0    0        \n",
       "6 Env  40654.9   0         105637    80992.5    56292    634090    0        \n",
       "  008_006_C 008_007_B ⋯ 008_029_C 008_030_B 008_030_C 008_032_B 008_034_B\n",
       "1 146631    108206.0  ⋯ 146745     36466.7  532094     77129.1       0   \n",
       "2      0         0.0  ⋯ 227184    142175.0       0    142841.0       0   \n",
       "3 203311     88059.1  ⋯  42024     92065.8       0     41340.5  292284   \n",
       "⋮ ⋮         ⋮         ⋱ ⋮         ⋮         ⋮         ⋮         ⋮        \n",
       "5 92812.1   258860    ⋯ 280027    337332    191108    354976         0   \n",
       "6 80359.4   125382    ⋯ 126295    116803    159696    144293    113451   \n",
       "  008_037_B 008_122_B 008_140_B 008_175_B 008_211_D\n",
       "1 73775.5        0    821708    0         0        \n",
       "2 20955.2        0         0    0         0        \n",
       "3 53127.4   436469         0    0         0        \n",
       "⋮ ⋮         ⋮         ⋮         ⋮         ⋮        \n",
       "5 395778    387058.0       0    567126         0   \n",
       "6 107418     61869.2  178292    201805    777618   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Prepare HIV gene expression table\n",
    "hiv_expression_table <- read_tsv(hiv_gene_expression_path,\n",
    "    show_col_types = FALSE) |>\n",
    "  mutate(Name = str_remove_all(Name, \"lcl\\\\|NC_001802.1_cds_|_\\\\d+$\"),\n",
    "    Name = hiv_dictionary[Name]) \n",
    "hiv_gene_names <- hiv_expression_table |>\n",
    "  pull(Name)\n",
    "head(hiv_expression_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare HHV8 gene expression tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>YP_001129350.1</dt><dd>'K1'</dd><dt>YP_001129351.1</dt><dd>'ORF4'</dd><dt>YP_001129352.1</dt><dd>'ORF6'</dd><dt>YP_001129353.1</dt><dd>'ORF7'</dd><dt>YP_001129354.1</dt><dd>'ORF8'</dd><dt>YP_001129355.1</dt><dd>'ORF9'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[YP\\textbackslash{}\\_001129350.1] 'K1'\n",
       "\\item[YP\\textbackslash{}\\_001129351.1] 'ORF4'\n",
       "\\item[YP\\textbackslash{}\\_001129352.1] 'ORF6'\n",
       "\\item[YP\\textbackslash{}\\_001129353.1] 'ORF7'\n",
       "\\item[YP\\textbackslash{}\\_001129354.1] 'ORF8'\n",
       "\\item[YP\\textbackslash{}\\_001129355.1] 'ORF9'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "YP_001129350.1\n",
       ":   'K1'YP_001129351.1\n",
       ":   'ORF4'YP_001129352.1\n",
       ":   'ORF6'YP_001129353.1\n",
       ":   'ORF7'YP_001129354.1\n",
       ":   'ORF8'YP_001129355.1\n",
       ":   'ORF9'\n",
       "\n"
      ],
      "text/plain": [
       "YP_001129350.1 YP_001129351.1 YP_001129352.1 YP_001129353.1 YP_001129354.1 \n",
       "          \"K1\"         \"ORF4\"         \"ORF6\"         \"ORF7\"         \"ORF8\" \n",
       "YP_001129355.1 \n",
       "        \"ORF9\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Prepare HHV8 gene dictionary\n",
    "suppressWarnings({\n",
    "hhv8_dictionary <- read.table(hhv8_fasta_path, sep=\"\\n\")$V1[grep(\">\", readLines(hhv8_fasta_path))] |>\n",
    "  as_tibble() |>\n",
    "  separate(value, into = c(\"pid\", \"gene_name\", \"org\"), sep = \"\\\\s\") |>\n",
    "  mutate(pid = str_remove(pid, \">\"),\n",
    "    gene_name = if_else(gene_name == \"Envelope\", \"Env\", gene_name)) |>\n",
    "    pull(gene_name)\n",
    "names(hhv8_dictionary) <- read.table(hhv8_fasta_path, sep=\"\\n\")$V1[grep(\">\", readLines(hhv8_fasta_path))] |>\n",
    "  as_tibble() |>\n",
    "  separate(value, into = c(\"pid\", \"gene_name\", \"org\"), sep = \"\\\\s\") |>\n",
    "  mutate(pid = str_remove(pid, \">\"),\n",
    "    gene_name = if_else(gene_name == \"Envelope\", \"Env\", gene_name)) |>\n",
    "    pull(pid)\n",
    "head(hhv8_dictionary)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 52</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Name</th><th scope=col>008_001_B</th><th scope=col>008_001_C</th><th scope=col>008_001_I</th><th scope=col>008_003_C</th><th scope=col>008_004_D</th><th scope=col>008_005_C</th><th scope=col>008_006_B</th><th scope=col>008_006_C</th><th scope=col>008_007_B</th><th scope=col>⋯</th><th scope=col>008_098_C</th><th scope=col>008_122_B</th><th scope=col>008_140_B</th><th scope=col>008_140_D</th><th scope=col>008_175_B</th><th scope=col>008_189_B</th><th scope=col>008_196_C</th><th scope=col>008_211_D</th><th scope=col>008_216_D</th><th scope=col>008_234_D</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ORF43</td><td>313.259</td><td>22.1425</td><td>1172.3</td><td>194.571</td><td> 618.727</td><td>0</td><td>  15.0339</td><td>  69.0059</td><td>  0.000</td><td>⋯</td><td>  0.000</td><td>0</td><td> 163.988</td><td>   0.00</td><td>791.9940</td><td>  28.4459</td><td> 98.8281</td><td>  0.000</td><td>296.225</td><td>640.2350</td></tr>\n",
       "\t<tr><td>ORF4 </td><td>116.006</td><td>24.6127</td><td>   0.0</td><td>648.511</td><td>1721.300</td><td>0</td><td>4563.1200</td><td>3313.0000</td><td>344.554</td><td>⋯</td><td>729.133</td><td>0</td><td>1569.200</td><td>2361.93</td><td> 84.9846</td><td>1699.8700</td><td>278.5380</td><td>239.734</td><td>  0.000</td><td>385.9460</td></tr>\n",
       "\t<tr><td>ORF19</td><td>116.241</td><td> 9.8651</td><td>1306.5</td><td>  0.000</td><td>1724.840</td><td>0</td><td>  16.7403</td><td>   0.0000</td><td>  0.000</td><td>⋯</td><td>  0.000</td><td>0</td><td>   0.000</td><td>   0.00</td><td>  0.0000</td><td>  32.1476</td><td>  0.0000</td><td>  0.000</td><td>  0.000</td><td> 77.3667</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>ORF36</td><td> 295.124</td><td>150.500</td><td>0</td><td>1100.03</td><td>3079.28</td><td>0</td><td> 21.2659</td><td> 48.8927</td><td> 2192.8</td><td>⋯</td><td>  0.0</td><td>6102.28</td><td>  0.000</td><td>  0.000</td><td> 618.423</td><td> 297.66</td><td>   0.00</td><td>  0.00</td><td> 583.544</td><td> 713.919</td></tr>\n",
       "\t<tr><td>K8.1 </td><td>5628.400</td><td>837.321</td><td>0</td><td>9254.41</td><td>8110.30</td><td>0</td><td>335.3740</td><td>444.0080</td><td>22716.7</td><td>⋯</td><td>690.7</td><td>5818.84</td><td>333.623</td><td>713.613</td><td>3145.210</td><td>4220.06</td><td>1020.26</td><td>215.76</td><td>1569.710</td><td>9186.370</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 52\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Name & 008\\_001\\_B & 008\\_001\\_C & 008\\_001\\_I & 008\\_003\\_C & 008\\_004\\_D & 008\\_005\\_C & 008\\_006\\_B & 008\\_006\\_C & 008\\_007\\_B & ⋯ & 008\\_098\\_C & 008\\_122\\_B & 008\\_140\\_B & 008\\_140\\_D & 008\\_175\\_B & 008\\_189\\_B & 008\\_196\\_C & 008\\_211\\_D & 008\\_216\\_D & 008\\_234\\_D\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t ORF43 & 313.259 & 22.1425 & 1172.3 & 194.571 &  618.727 & 0 &   15.0339 &   69.0059 &   0.000 & ⋯ &   0.000 & 0 &  163.988 &    0.00 & 791.9940 &   28.4459 &  98.8281 &   0.000 & 296.225 & 640.2350\\\\\n",
       "\t ORF4  & 116.006 & 24.6127 &    0.0 & 648.511 & 1721.300 & 0 & 4563.1200 & 3313.0000 & 344.554 & ⋯ & 729.133 & 0 & 1569.200 & 2361.93 &  84.9846 & 1699.8700 & 278.5380 & 239.734 &   0.000 & 385.9460\\\\\n",
       "\t ORF19 & 116.241 &  9.8651 & 1306.5 &   0.000 & 1724.840 & 0 &   16.7403 &    0.0000 &   0.000 & ⋯ &   0.000 & 0 &    0.000 &    0.00 &   0.0000 &   32.1476 &   0.0000 &   0.000 &   0.000 &  77.3667\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t ORF36 &  295.124 & 150.500 & 0 & 1100.03 & 3079.28 & 0 &  21.2659 &  48.8927 &  2192.8 & ⋯ &   0.0 & 6102.28 &   0.000 &   0.000 &  618.423 &  297.66 &    0.00 &   0.00 &  583.544 &  713.919\\\\\n",
       "\t K8.1  & 5628.400 & 837.321 & 0 & 9254.41 & 8110.30 & 0 & 335.3740 & 444.0080 & 22716.7 & ⋯ & 690.7 & 5818.84 & 333.623 & 713.613 & 3145.210 & 4220.06 & 1020.26 & 215.76 & 1569.710 & 9186.370\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 52\n",
       "\n",
       "| Name &lt;chr&gt; | 008_001_B &lt;dbl&gt; | 008_001_C &lt;dbl&gt; | 008_001_I &lt;dbl&gt; | 008_003_C &lt;dbl&gt; | 008_004_D &lt;dbl&gt; | 008_005_C &lt;dbl&gt; | 008_006_B &lt;dbl&gt; | 008_006_C &lt;dbl&gt; | 008_007_B &lt;dbl&gt; | ⋯ ⋯ | 008_098_C &lt;dbl&gt; | 008_122_B &lt;dbl&gt; | 008_140_B &lt;dbl&gt; | 008_140_D &lt;dbl&gt; | 008_175_B &lt;dbl&gt; | 008_189_B &lt;dbl&gt; | 008_196_C &lt;dbl&gt; | 008_211_D &lt;dbl&gt; | 008_216_D &lt;dbl&gt; | 008_234_D &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ORF43 | 313.259 | 22.1425 | 1172.3 | 194.571 |  618.727 | 0 |   15.0339 |   69.0059 |   0.000 | ⋯ |   0.000 | 0 |  163.988 |    0.00 | 791.9940 |   28.4459 |  98.8281 |   0.000 | 296.225 | 640.2350 |\n",
       "| ORF4  | 116.006 | 24.6127 |    0.0 | 648.511 | 1721.300 | 0 | 4563.1200 | 3313.0000 | 344.554 | ⋯ | 729.133 | 0 | 1569.200 | 2361.93 |  84.9846 | 1699.8700 | 278.5380 | 239.734 |   0.000 | 385.9460 |\n",
       "| ORF19 | 116.241 |  9.8651 | 1306.5 |   0.000 | 1724.840 | 0 |   16.7403 |    0.0000 |   0.000 | ⋯ |   0.000 | 0 |    0.000 |    0.00 |   0.0000 |   32.1476 |   0.0000 |   0.000 |   0.000 |  77.3667 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| ORF36 |  295.124 | 150.500 | 0 | 1100.03 | 3079.28 | 0 |  21.2659 |  48.8927 |  2192.8 | ⋯ |   0.0 | 6102.28 |   0.000 |   0.000 |  618.423 |  297.66 |    0.00 |   0.00 |  583.544 |  713.919 |\n",
       "| K8.1  | 5628.400 | 837.321 | 0 | 9254.41 | 8110.30 | 0 | 335.3740 | 444.0080 | 22716.7 | ⋯ | 690.7 | 5818.84 | 333.623 | 713.613 | 3145.210 | 4220.06 | 1020.26 | 215.76 | 1569.710 | 9186.370 |\n",
       "\n"
      ],
      "text/plain": [
       "  Name  008_001_B 008_001_C 008_001_I 008_003_C 008_004_D 008_005_C 008_006_B\n",
       "1 ORF43 313.259   22.1425   1172.3    194.571    618.727  0           15.0339\n",
       "2 ORF4  116.006   24.6127      0.0    648.511   1721.300  0         4563.1200\n",
       "3 ORF19 116.241    9.8651   1306.5      0.000   1724.840  0           16.7403\n",
       "⋮ ⋮     ⋮         ⋮         ⋮         ⋮         ⋮         ⋮         ⋮        \n",
       "5 ORF36  295.124  150.500   0         1100.03   3079.28   0          21.2659 \n",
       "6 K8.1  5628.400  837.321   0         9254.41   8110.30   0         335.3740 \n",
       "  008_006_C 008_007_B ⋯ 008_098_C 008_122_B 008_140_B 008_140_D 008_175_B\n",
       "1   69.0059   0.000   ⋯   0.000   0          163.988     0.00   791.9940 \n",
       "2 3313.0000 344.554   ⋯ 729.133   0         1569.200  2361.93    84.9846 \n",
       "3    0.0000   0.000   ⋯   0.000   0            0.000     0.00     0.0000 \n",
       "⋮ ⋮         ⋮         ⋱ ⋮         ⋮         ⋮         ⋮         ⋮        \n",
       "5  48.8927   2192.8   ⋯   0.0     6102.28     0.000     0.000    618.423 \n",
       "6 444.0080  22716.7   ⋯ 690.7     5818.84   333.623   713.613   3145.210 \n",
       "  008_189_B 008_196_C 008_211_D 008_216_D 008_234_D\n",
       "1   28.4459  98.8281    0.000   296.225   640.2350 \n",
       "2 1699.8700 278.5380  239.734     0.000   385.9460 \n",
       "3   32.1476   0.0000    0.000     0.000    77.3667 \n",
       "⋮ ⋮         ⋮         ⋮         ⋮         ⋮        \n",
       "5  297.66      0.00     0.00     583.544   713.919 \n",
       "6 4220.06   1020.26   215.76    1569.710  9186.370 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Prepare HHV8 gene expression data\n",
    "hhv8_expression_table <- read_tsv(hhv8_gene_expression_path,\n",
    "    show_col_types = FALSE) |>\n",
    "  dplyr::mutate(Name = str_remove_all(Name, \"lcl\\\\|NC_009333.1_cds_|_\\\\d+$\"),\n",
    "    Name = hhv8_dictionary[Name])\n",
    "# Break up KSHV genes into groups based on stage of expression\n",
    "#https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1001013#s5\n",
    "#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3894221/\n",
    "kshv_groups <- c(\"vIRF-3\" = \"La\", \"ORF73\" = \"La\", \"K4.2\" = \"IE\", \"ORF45\" = \"IE\",\n",
    "  \"ORF48\" = \"IE\", \"ORF50\" = \"IE\", \"K8\" = \"IE\", \"ORF70\" = \"E1\", \"ORF10\" = \"E1\",\n",
    "  \"ORF56\" = \"E1\", \"ORF11\" = \"E1\", \"K3\" = \"E1\", \"ORF40\" = \"E1\", \"ORF59\" = \"E1\",\n",
    "  \"K15\" = \"E1\", \"ORF54\" = \"E1\", \"ORF2\" = \"E2\", \"K1\" = \"E2\", \"K2\" = \"E2\",\n",
    "  \"ORF46\" = \"E2\", \"ORF9\" = \"E2\", \"K5\" = \"E2\", \"K7\" = \"E2\", \"K6\" = \"E2\",\n",
    "  \"vIRF-1\" = \"E2\", \"K14\" = \"E2\", \"vIRF-2\" = \"E2\", \"ORF49\" = \"E2\", \"ORF74\" = \"E2\",\n",
    "  \"ORF7\" = \"E2\", \"ORF6\" = \"E2\", \"ORF44\" = \"E3\", \"ORF31\" = \"E3\", \"ORF19\" = \"E3\",\n",
    "  \"ORF37\" = \"E3\",  \"ORF57\" = \"E3\", \"ORF17\" = \"E3\", \"ORF36\" = \"E3\", \"ORF29\" = \"E3\",\n",
    "  \"ORF21\" = \"E3\", \"ORF61\" = \"E3\", \"ORF60\" = \"E3\", \"ORF66\" = \"E3\", \"ORF16\" = \"E3\",\n",
    "  \"vIRF-4\" = \"E3\", \"ORF69\" = \"E3\", \"ORF23\" = \"L4\", \"ORF22\" = \"L4\", \"ORF25\" = \"L4\",\n",
    "  \"ORF20\" = \"L4\", \"ORF26\" = \"L4\", \"ORF24\" = \"L4\", \"ORF27\" = \"L4\", \"ORF30\" = \"L4\",\n",
    "  \"ORF28\" = \"L4\", \"ORF34\" = \"L4\", \"ORF18\" = \"L4\", \"ORF35\" = \"L4\", \"ORF63\" = \"L4\",\n",
    "  \"ORF62\" = \"L4\", \"ORF65\" = \"L4\", \"ORF32\" = \"L4\", \"ORF38\" = \"L4\", \"ORF33\" = \"L4\", \n",
    "  \"ORF43\" = \"L4\", \"ORF64\" = \"L4\", \"ORF67\" = \"L4\", \"ORF68\" = \"L4\", \"ORF42\" = \"L5\",\n",
    "  \"ORF4\" = \"L5\", \"ORF39\" = \"L5\", \"K8.1\" = \"L5\", \"ORF8\" = \"L5\", \"ORF47\" = \"L5\",\n",
    "  \"ORF75\" = \"L5\", \"ORF58\" = \"L5\", \"ORF52\" = \"L5\", \"ORF53\" = \"L5\", \"ORF55\" = \"L5\",\n",
    "  \"K12\" = \"La\", \"K4\" = \"E1\", \"K4.1\" = \"E1\", \"ORF17.5\" = \"E2\", \"ORF67A\" = \"L5\",\n",
    "  \"ORF71\" = \"La\", \"ORF72\" = \"La\" )\n",
    "  head(hhv8_expression_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare CibersortX results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 51 × 23</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Mixture</th><th scope=col>B cells naive</th><th scope=col>B cells memory</th><th scope=col>Plasma cells</th><th scope=col>T cells CD8</th><th scope=col>T cells CD4 naive</th><th scope=col>T cells CD4 memory resting</th><th scope=col>T cells CD4 memory activated</th><th scope=col>T cells follicular helper</th><th scope=col>T cells regulatory (Tregs)</th><th scope=col>⋯</th><th scope=col>Monocytes</th><th scope=col>Macrophages M0</th><th scope=col>Macrophages M1</th><th scope=col>Macrophages M2</th><th scope=col>Dendritic cells resting</th><th scope=col>Dendritic cells activated</th><th scope=col>Mast cells resting</th><th scope=col>Mast cells activated</th><th scope=col>Eosinophils</th><th scope=col>Neutrophils</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_061_C</td><td>7.912886</td><td>0</td><td>3.921065</td><td>19.70740</td><td>0</td><td>3.996909</td><td>1.182755</td><td>5.857817</td><td>0.5966646</td><td>⋯</td><td>1.1638893</td><td>10.835613</td><td>11.778665</td><td>23.80936</td><td>1.9529614</td><td>0</td><td>2.60010266</td><td>0.0000000</td><td>0</td><td>0.00000000</td></tr>\n",
       "\t<tr><td>008_098_B</td><td>6.596161</td><td>0</td><td>3.678067</td><td>21.50593</td><td>0</td><td>0.000000</td><td>9.657895</td><td>3.845360</td><td>0.9626662</td><td>⋯</td><td>0.0000000</td><td> 1.311762</td><td> 6.150924</td><td>39.87146</td><td>0.0000000</td><td>0</td><td>0.00000000</td><td>0.0000000</td><td>0</td><td>0.23090733</td></tr>\n",
       "\t<tr><td>008_098_C</td><td>5.366013</td><td>0</td><td>2.474363</td><td>42.23846</td><td>0</td><td>0.000000</td><td>4.580758</td><td>3.435627</td><td>6.1273057</td><td>⋯</td><td>0.5155276</td><td> 0.000000</td><td> 4.586556</td><td>24.21384</td><td>0.1408064</td><td>0</td><td>0.03316393</td><td>0.2484124</td><td>0</td><td>0.03748412</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_034_B</td><td>0.000000</td><td>0.3521525</td><td> 4.483397</td><td>25.12626</td><td>0</td><td>0</td><td>0</td><td>16.59834</td><td>0.0000000</td><td>⋯</td><td>0.1082690</td><td>0.00000</td><td>22.656619</td><td>14.69550</td><td>0.3275799</td><td>0</td><td>8.387082</td><td>0</td><td>0</td><td>1.730945</td></tr>\n",
       "\t<tr><td>008_037_B</td><td>7.057123</td><td>0.0000000</td><td>14.341140</td><td>18.42728</td><td>0</td><td>0</td><td>0</td><td>14.77712</td><td>0.3843394</td><td>⋯</td><td>0.5192566</td><td>8.19816</td><td> 7.489275</td><td>20.11378</td><td>0.0000000</td><td>0</td><td>3.648959</td><td>0</td><td>0</td><td>0.000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 51 × 23\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Mixture & B cells naive & B cells memory & Plasma cells & T cells CD8 & T cells CD4 naive & T cells CD4 memory resting & T cells CD4 memory activated & T cells follicular helper & T cells regulatory (Tregs) & ⋯ & Monocytes & Macrophages M0 & Macrophages M1 & Macrophages M2 & Dendritic cells resting & Dendritic cells activated & Mast cells resting & Mast cells activated & Eosinophils & Neutrophils\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 008\\_061\\_C & 7.912886 & 0 & 3.921065 & 19.70740 & 0 & 3.996909 & 1.182755 & 5.857817 & 0.5966646 & ⋯ & 1.1638893 & 10.835613 & 11.778665 & 23.80936 & 1.9529614 & 0 & 2.60010266 & 0.0000000 & 0 & 0.00000000\\\\\n",
       "\t 008\\_098\\_B & 6.596161 & 0 & 3.678067 & 21.50593 & 0 & 0.000000 & 9.657895 & 3.845360 & 0.9626662 & ⋯ & 0.0000000 &  1.311762 &  6.150924 & 39.87146 & 0.0000000 & 0 & 0.00000000 & 0.0000000 & 0 & 0.23090733\\\\\n",
       "\t 008\\_098\\_C & 5.366013 & 0 & 2.474363 & 42.23846 & 0 & 0.000000 & 4.580758 & 3.435627 & 6.1273057 & ⋯ & 0.5155276 &  0.000000 &  4.586556 & 24.21384 & 0.1408064 & 0 & 0.03316393 & 0.2484124 & 0 & 0.03748412\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_034\\_B & 0.000000 & 0.3521525 &  4.483397 & 25.12626 & 0 & 0 & 0 & 16.59834 & 0.0000000 & ⋯ & 0.1082690 & 0.00000 & 22.656619 & 14.69550 & 0.3275799 & 0 & 8.387082 & 0 & 0 & 1.730945\\\\\n",
       "\t 008\\_037\\_B & 7.057123 & 0.0000000 & 14.341140 & 18.42728 & 0 & 0 & 0 & 14.77712 & 0.3843394 & ⋯ & 0.5192566 & 8.19816 &  7.489275 & 20.11378 & 0.0000000 & 0 & 3.648959 & 0 & 0 & 0.000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 51 × 23\n",
       "\n",
       "| Mixture &lt;chr&gt; | B cells naive &lt;dbl&gt; | B cells memory &lt;dbl&gt; | Plasma cells &lt;dbl&gt; | T cells CD8 &lt;dbl&gt; | T cells CD4 naive &lt;dbl&gt; | T cells CD4 memory resting &lt;dbl&gt; | T cells CD4 memory activated &lt;dbl&gt; | T cells follicular helper &lt;dbl&gt; | T cells regulatory (Tregs) &lt;dbl&gt; | ⋯ ⋯ | Monocytes &lt;dbl&gt; | Macrophages M0 &lt;dbl&gt; | Macrophages M1 &lt;dbl&gt; | Macrophages M2 &lt;dbl&gt; | Dendritic cells resting &lt;dbl&gt; | Dendritic cells activated &lt;dbl&gt; | Mast cells resting &lt;dbl&gt; | Mast cells activated &lt;dbl&gt; | Eosinophils &lt;dbl&gt; | Neutrophils &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 008_061_C | 7.912886 | 0 | 3.921065 | 19.70740 | 0 | 3.996909 | 1.182755 | 5.857817 | 0.5966646 | ⋯ | 1.1638893 | 10.835613 | 11.778665 | 23.80936 | 1.9529614 | 0 | 2.60010266 | 0.0000000 | 0 | 0.00000000 |\n",
       "| 008_098_B | 6.596161 | 0 | 3.678067 | 21.50593 | 0 | 0.000000 | 9.657895 | 3.845360 | 0.9626662 | ⋯ | 0.0000000 |  1.311762 |  6.150924 | 39.87146 | 0.0000000 | 0 | 0.00000000 | 0.0000000 | 0 | 0.23090733 |\n",
       "| 008_098_C | 5.366013 | 0 | 2.474363 | 42.23846 | 0 | 0.000000 | 4.580758 | 3.435627 | 6.1273057 | ⋯ | 0.5155276 |  0.000000 |  4.586556 | 24.21384 | 0.1408064 | 0 | 0.03316393 | 0.2484124 | 0 | 0.03748412 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_034_B | 0.000000 | 0.3521525 |  4.483397 | 25.12626 | 0 | 0 | 0 | 16.59834 | 0.0000000 | ⋯ | 0.1082690 | 0.00000 | 22.656619 | 14.69550 | 0.3275799 | 0 | 8.387082 | 0 | 0 | 1.730945 |\n",
       "| 008_037_B | 7.057123 | 0.0000000 | 14.341140 | 18.42728 | 0 | 0 | 0 | 14.77712 | 0.3843394 | ⋯ | 0.5192566 | 8.19816 |  7.489275 | 20.11378 | 0.0000000 | 0 | 3.648959 | 0 | 0 | 0.000000 |\n",
       "\n"
      ],
      "text/plain": [
       "   Mixture   B cells naive B cells memory Plasma cells T cells CD8\n",
       "1  008_061_C 7.912886      0              3.921065     19.70740   \n",
       "2  008_098_B 6.596161      0              3.678067     21.50593   \n",
       "3  008_098_C 5.366013      0              2.474363     42.23846   \n",
       "⋮  ⋮         ⋮             ⋮              ⋮            ⋮          \n",
       "50 008_034_B 0.000000      0.3521525       4.483397    25.12626   \n",
       "51 008_037_B 7.057123      0.0000000      14.341140    18.42728   \n",
       "   T cells CD4 naive T cells CD4 memory resting T cells CD4 memory activated\n",
       "1  0                 3.996909                   1.182755                    \n",
       "2  0                 0.000000                   9.657895                    \n",
       "3  0                 0.000000                   4.580758                    \n",
       "⋮  ⋮                 ⋮                          ⋮                           \n",
       "50 0                 0                          0                           \n",
       "51 0                 0                          0                           \n",
       "   T cells follicular helper T cells regulatory (Tregs) ⋯ Monocytes\n",
       "1  5.857817                  0.5966646                  ⋯ 1.1638893\n",
       "2  3.845360                  0.9626662                  ⋯ 0.0000000\n",
       "3  3.435627                  6.1273057                  ⋯ 0.5155276\n",
       "⋮  ⋮                         ⋮                          ⋱ ⋮        \n",
       "50 16.59834                  0.0000000                  ⋯ 0.1082690\n",
       "51 14.77712                  0.3843394                  ⋯ 0.5192566\n",
       "   Macrophages M0 Macrophages M1 Macrophages M2 Dendritic cells resting\n",
       "1  10.835613      11.778665      23.80936       1.9529614              \n",
       "2   1.311762       6.150924      39.87146       0.0000000              \n",
       "3   0.000000       4.586556      24.21384       0.1408064              \n",
       "⋮  ⋮              ⋮              ⋮              ⋮                      \n",
       "50 0.00000        22.656619      14.69550       0.3275799              \n",
       "51 8.19816         7.489275      20.11378       0.0000000              \n",
       "   Dendritic cells activated Mast cells resting Mast cells activated\n",
       "1  0                         2.60010266         0.0000000           \n",
       "2  0                         0.00000000         0.0000000           \n",
       "3  0                         0.03316393         0.2484124           \n",
       "⋮  ⋮                         ⋮                  ⋮                   \n",
       "50 0                         8.387082           0                   \n",
       "51 0                         3.648959           0                   \n",
       "   Eosinophils Neutrophils\n",
       "1  0           0.00000000 \n",
       "2  0           0.23090733 \n",
       "3  0           0.03748412 \n",
       "⋮  ⋮           ⋮          \n",
       "50 0           1.730945   \n",
       "51 0           0.000000   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load and format CibersortX results\n",
    "cibersort_table <- list.files(cibersortx_results_path, pattern = \"hippos\",\n",
    "    full.names = TRUE) |>\n",
    "    map_df(~read_csv(.x, show_col_types = FALSE)) |>\n",
    "    mutate(Mixture = str_remove(Mixture, \"_quant\"),\n",
    "        Mixture = str_replace_all(Mixture, \"-\", \"_\"),\n",
    "        Mixture = str_extract(Mixture, \"008_\\\\d+_\\\\w\")) |>\n",
    "    dplyr::select(-`P-value`, -Correlation, -RMSE) |> \n",
    "    pivot_longer(c(-Mixture, -`Absolute score (sig.score)`), names_to = \"cell_type\", values_to = \"score\") |>\n",
    "    mutate(percentage = (score*100)/`Absolute score (sig.score)`) |>\n",
    "    pivot_wider(id_cols = Mixture, names_from = cell_type, values_from = percentage)\n",
    "cibersort_table "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Format HIV table to include all samples form the study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>samples</th><th scope=col>Name</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_061_C</td><td>Vpr</td></tr>\n",
       "\t<tr><td>008_061_C</td><td>Tat</td></tr>\n",
       "\t<tr><td>008_061_C</td><td>Vif</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_061_C</td><td>Nef</td></tr>\n",
       "\t<tr><td>008_061_C</td><td>Env</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 2\n",
       "\\begin{tabular}{ll}\n",
       " samples & Name\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_061\\_C & Vpr\\\\\n",
       "\t 008\\_061\\_C & Tat\\\\\n",
       "\t 008\\_061\\_C & Vif\\\\\n",
       "\t ⋮ & ⋮\\\\\n",
       "\t 008\\_061\\_C & Nef\\\\\n",
       "\t 008\\_061\\_C & Env\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 2\n",
       "\n",
       "| samples &lt;chr&gt; | Name &lt;chr&gt; |\n",
       "|---|---|\n",
       "| 008_061_C | Vpr |\n",
       "| 008_061_C | Tat |\n",
       "| 008_061_C | Vif |\n",
       "| ⋮ | ⋮ |\n",
       "| 008_061_C | Nef |\n",
       "| 008_061_C | Env |\n",
       "\n"
      ],
      "text/plain": [
       "  samples   Name\n",
       "1 008_061_C Vpr \n",
       "2 008_061_C Tat \n",
       "3 008_061_C Vif \n",
       "⋮ ⋮         ⋮   \n",
       "5 008_061_C Nef \n",
       "6 008_061_C Env "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generate HIV sample list table\n",
    "# We will need this table to generate the final heatmap. \n",
    "# Since many samples do not show expression of HIV, this table will help\n",
    "# generate a full table with 51 samples\n",
    "sample_list <- cibersort_table |> \n",
    "  pull(Mixture)\n",
    "hiv_gene_list <- hiv_expression_table |>\n",
    "  pull(Name)\n",
    "\n",
    "hiv_sample_table <- expand_grid(sample_list, hiv_gene_list) |>\n",
    "  mutate(tpm = NULL) |>\n",
    "  dplyr::rename(samples = sample_list, Name = hiv_gene_list)\n",
    "sample_table <- cibersort_table |> \n",
    "  pull(Mixture) |> \n",
    "  as_tibble() |> \n",
    "  dplyr::rename(samples = value)\n",
    "head(hiv_sample_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 110 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>repertoire_id</th><th scope=col>rna_libraries</th><th scope=col>tissue_type</th><th scope=col>phenotype</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>008_001_B</td><td>008-001-B-12 </td><td>Tumor</td><td>Epidemic KS</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001_C</td><td>008-001-C-13 </td><td>Tumor</td><td>Epidemic KS</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001_I</td><td>008-001-I-282</td><td>Tumor</td><td>Epidemic KS</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>TsoEtAl</td><td>SRR5787176</td><td>p83</td><td>Tumor</td><td>Epidemic KS</td></tr>\n",
       "\t<tr><td>TsoEtAl</td><td>SRR5787177</td><td>p23</td><td>Tumor</td><td>Epidemic KS</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 110 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " cohort & repertoire\\_id & rna\\_libraries & tissue\\_type & phenotype\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t Hippos & 008\\_001\\_B & 008-001-B-12  & Tumor & Epidemic KS\\\\\n",
       "\t Hippos & 008\\_001\\_C & 008-001-C-13  & Tumor & Epidemic KS\\\\\n",
       "\t Hippos & 008\\_001\\_I & 008-001-I-282 & Tumor & Epidemic KS\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t TsoEtAl & SRR5787176 & p83 & Tumor & Epidemic KS\\\\\n",
       "\t TsoEtAl & SRR5787177 & p23 & Tumor & Epidemic KS\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 110 × 5\n",
       "\n",
       "| cohort &lt;chr&gt; | repertoire_id &lt;chr&gt; | rna_libraries &lt;chr&gt; | tissue_type &lt;chr&gt; | phenotype &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| Hippos | 008_001_B | 008-001-B-12  | Tumor | Epidemic KS |\n",
       "| Hippos | 008_001_C | 008-001-C-13  | Tumor | Epidemic KS |\n",
       "| Hippos | 008_001_I | 008-001-I-282 | Tumor | Epidemic KS |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| TsoEtAl | SRR5787176 | p83 | Tumor | Epidemic KS |\n",
       "| TsoEtAl | SRR5787177 | p23 | Tumor | Epidemic KS |\n",
       "\n"
      ],
      "text/plain": [
       "    cohort  repertoire_id rna_libraries tissue_type phenotype  \n",
       "1   Hippos  008_001_B     008-001-B-12  Tumor       Epidemic KS\n",
       "2   Hippos  008_001_C     008-001-C-13  Tumor       Epidemic KS\n",
       "3   Hippos  008_001_I     008-001-I-282 Tumor       Epidemic KS\n",
       "⋮   ⋮       ⋮             ⋮             ⋮           ⋮          \n",
       "109 TsoEtAl SRR5787176    p83           Tumor       Epidemic KS\n",
       "110 TsoEtAl SRR5787177    p23           Tumor       Epidemic KS"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Prep RNA-Seq metadata\n",
    "rna_seq_metadata <-  study_metadata |> \n",
    "  filter(!is.na(rna_libraries)) |> \n",
    "  dplyr::select(cohort, repertoire_id, rna_libraries, tissue_type, phenotype)\n",
    "rna_seq_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "plot_order <- cibersort_table |> \n",
    "    left_join(rna_seq_metadata, by = c(\"Mixture\" = \"repertoire_id\")) |>\n",
    "    dplyr::arrange(phenotype, `Macrophages M2`) |>\n",
    "    pull(Mixture)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Plot cibersortX panel\n",
    "cibersort_plot <- cibersort_table |>\n",
    "  pivot_longer(cols = `B cells naive`:Neutrophils, names_to = \"cell_type\", \n",
    "    values_to = \"abundance\" ) |> \n",
    "  left_join(rna_seq_metadata, by = c(\"Mixture\" = \"repertoire_id\")) |>\n",
    "  mutate(Mixture = factor(Mixture, levels = plot_order)) |>\n",
    "  ggplot(aes(x = cell_type, y = Mixture, fill = abundance)) +\n",
    "  geom_tile(color = \"white\", lwd = 0.25, linetype = 1) +\n",
    "  labs(y = \"TRB repertoire ID\", x = \"Cell type\", fill = \"Cell type\\nabundance\") +\n",
    "  ggtitle(\"Immune cell-type\\ncomposition\") +\n",
    "  #scale_fill_gradientn(colors = rev(c('#b2182b','#d6604d','#d1e5f0'))) +\n",
    "  scale_fill_viridis_b(option = \"plasma\") + \n",
    "  theme_classic(base_size = 20) +\n",
    "  facet_grid(rows = vars(phenotype),scales = \"free\", space = \"free\", margins = F) +\n",
    "  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face=\"bold\"),  \n",
    "    axis.title.x = element_blank(), \n",
    "    axis.text.y = element_blank(), axis.line.y = element_blank(), \n",
    "    axis.title.y = element_blank(), axis.ticks.y = element_blank(), \n",
    "    strip.background.x = element_blank(), \n",
    "    strip.background.y = element_rect(linewidth = 2, color = \"black\", linetype = 1), \n",
    "    strip.text.y.right = element_text(face = \"bold\"),\n",
    "    strip.placement = \"inside\", plot.title = element_text(face = \"bold\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(Name, samples)`\n"
     ]
    }
   ],
   "source": [
    "# Plot HIV gene expression panel\n",
    "hiv_plot <- hiv_expression_table |>\n",
    "  pivot_longer(-Name, names_to = \"samples\",\n",
    "    values_to = \"tpm\") |>\n",
    "  full_join(hiv_sample_table) |> \n",
    "  left_join(rna_seq_metadata, by = c(\"samples\" = \"repertoire_id\")) |>\n",
    "  mutate(tpm = if_else(tpm == 0, NA_integer_, tpm),\n",
    "    samples = factor(samples, levels = plot_order)) |>\n",
    "  ggplot(aes(x = Name, y = samples, fill = tpm)) +\n",
    "  geom_tile(color = \"white\", lwd = 0.25, linetype = 1) +\n",
    "  labs(x = \"Gene\", fill = \"TPM\") +\n",
    "  ggtitle(\"HIV gene\\nexpression\") +\n",
    "  scale_fill_viridis_c(option = \"viridis\", \n",
    "    trans = scales::pseudo_log_trans(sigma = 0.001),\n",
    "    breaks = c(0, 10, 100, 1000, 10000, 100000, 1000000)) +\n",
    "  theme_classic(base_size = 20) +\n",
    "  facet_grid(rows = vars(phenotype), scales = \"free\", space = \"free\", \n",
    "    margins = F) +\n",
    "  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face=\"bold\"), \n",
    "    axis.title.x = element_blank(), legend.position = \"none\", \n",
    "    axis.text.y = element_blank(), axis.line.y = element_blank(), \n",
    "    axis.title.y = element_blank(), axis.ticks.y = element_blank(),  \n",
    "    strip.background = element_blank(), strip.text.y = element_blank(),\n",
    "    strip.placement = \"inside\", plot.title = element_text(face = \"bold\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(samples)`\n"
     ]
    }
   ],
   "source": [
    "# Plot KSHV gene expression\n",
    "kshv_plot <- hhv8_expression_table |>\n",
    "  pivot_longer(col = `008_001_B`:`008_234_D`, names_to = \"samples\",\n",
    "    values_to = \"tpm\") |>\n",
    "  mutate(tpm = if_else(tpm == 0, NA_integer_, tpm)) |>\n",
    "  right_join(sample_table) |> \n",
    "  left_join(rna_seq_metadata, by = c(\"samples\" = \"repertoire_id\")) |>\n",
    "  mutate(stage = factor(kshv_groups[Name], \n",
    "    levels =c(\"La\", \"IE\", \"E1\", \"E2\", \"E3\", \"L4\", \"L5\", NA)),\n",
    "    samples = factor(samples, levels = plot_order)) |>\n",
    "  ggplot(aes(x = Name, y = samples, fill = tpm)) +\n",
    "  geom_tile(color = \"white\", lwd = 0.25, linetype = 1) +\n",
    "  labs(x = \"Gene\", fill = \"TPM\", y = \"Sample ID\") +\n",
    "  ggtitle(\"KSHV gene\\nexpression\") +\n",
    "  scale_fill_viridis_c(option = \"viridis\", \n",
    "    trans = scales::pseudo_log_trans(sigma = 0.001),\n",
    "    breaks = c(0, 10, 100, 1000, 10000, 100000, 1000000)) +\n",
    "  theme_classic(base_size = 20) +\n",
    "  facet_grid(rows = vars(phenotype), cols = vars(stage), scales = \"free\", \n",
    "    space = \"free\", margins = F)  +\n",
    "  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face=\"bold\"),\n",
    "    axis.text.y = element_text(face = \"bold\"),\n",
    "    axis.title.x = element_blank(),panel.spacing.x = unit(0.25, \"lines\"),\n",
    "    strip.text.y = element_blank(), strip.text.x = element_text(face=\"bold\"),\n",
    "    strip.placement = \"inside\", axis.title.y = element_text(face = \"bold\"),\n",
    "    plot.title = element_text(face = \"bold\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Plot cell type boxplot\n",
    "cohort_palette <- c(\"Epidemic KS\" = \"#e41a1c\", \"Endemic KS\" = \"#377eb8\")\n",
    "cibersort_boxplots <- cibersort_table |>\n",
    "  pivot_longer(`B cells naive`:Neutrophils, names_to = \"cell_type\", \n",
    "    values_to = \"abundance\") |>\n",
    "  filter(cell_type %in% c(\"Macrophages M2\", \n",
    "    \"T cells CD4 memory activated\", \"T cells CD4 memory resting\",\n",
    "    \"T cells CD8\", \"T cells follicular helper\")) |>\n",
    "  left_join(rna_seq_metadata, by = c(\"Mixture\" = \"repertoire_id\")) |>\n",
    "  mutate(cell_type = case_when(cell_type == \"T cells CD4 memory activated\" ~ \"T cells CD4\\nmemory activated\",\n",
    "    cell_type == \"T cells CD4 memory resting\" ~ \"T cells CD4\\nmemory resting\",\n",
    "    cell_type == \"T cells follicular helper\" ~ \"T cells\\nfollicular helper\",\n",
    "    TRUE ~ cell_type)) |>\n",
    "  ggboxplot(x = \"phenotype\", y = \"abundance\", color = \"phenotype\", \n",
    "    facet.by = \"cell_type\", nrow = 1) |>\n",
    "  ggpar(xlab = \"Cohort\", ylab = \"Abundance\", legend.title = \"Cohort\",\n",
    "    ggtheme = theme_classic(base_size = 20), ylim = c(0, 60),\n",
    "    legend = \"right\", palette = cohort_palette,\n",
    "    panel.labs.font = list(face = \"bold\")) +\n",
    "  stat_compare_means(comparisons = list(c(\"Endemic KS\", \"Epidemic KS\")),\n",
    "    size = 5, label = \"p.signif\") +\n",
    "  theme(strip.text.x = element_text(face=\"bold\"),\n",
    "    axis.text.x = element_blank(), axis.ticks.x = element_blank(),\n",
    "    axis.title = element_text(face = \"bold\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Plot correlation scatter plot for M2 macrophages and CD8 T-cells\n",
    "tcell_mac_corr_plot <- cibersort_table |>\n",
    "    left_join(rna_seq_metadata, by = c(\"Mixture\" = \"repertoire_id\")) |>\n",
    "    ggplot(aes(x = `Macrophages M2`, y = `T cells CD8`, color = phenotype)) +\n",
    "    geom_point() +\n",
    "    labs(ylab = \"T cells CD8 (%)\",\n",
    "        xlab = \"Macrophage M2 (%)\") +\n",
    "    geom_smooth(method = \"glm\", aes(fill=phenotype)) +\n",
    "    ylim(0, 50) + \n",
    "    stat_cor(p.accuracy = 0.001, r.accuracy = 0.01, size = 8)+\n",
    "    scale_color_manual(values = cohort_palette) + \n",
    "    scale_fill_manual(values = cohort_palette) + \n",
    "    theme_classic(base_size = 20) +\n",
    "    theme(axis.title = element_text(face = \"bold\"),\n",
    "        legend.position = \"none\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n",
      "\u001b[1m\u001b[22m`geom_smooth()` using formula = 'y ~ x'\n"
     ]
    }
   ],
   "source": [
    "# Assemble all panels of Figure 1\n",
    "figure_one_layout <- \"\n",
    "1111111111111111112233333\n",
    "1111111111111111112233333\n",
    "1111111111111111112233333\n",
    "1111111111111111112233333\n",
    "4444444444444444444455555\"\n",
    "suppressWarnings({\n",
    "figure_one <- kshv_plot +  hiv_plot +\n",
    "  cibersort_plot  +  cibersort_boxplots + tcell_mac_corr_plot +\n",
    "  plot_layout(design = figure_one_layout, guides = \"collect\",\n",
    "    widths =c(9, 1, 2.5, 10, 2.5)) +\n",
    "  plot_annotation(tag_levels = 'A') & theme(text = element_text('NimbusSan'))\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Figure_one.pdf\", sep = \"/\"), figure_one, \n",
    "  limitsize = FALSE, width = 30, height = 24, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Figure_one.svg\", sep = \"/\"), figure_one,\n",
    "  limitsize = FALSE, width = 30, height = 24, device = \"svg\")\n",
    "ggsave(str_c(figures_path, \"png\", \"Figure_one.png\", sep = \"/\"), figure_one,\n",
    "  limitsize = FALSE, width = 30, height = 24, device = \"png\")})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary figure 1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "phenotype_table <- study_metadata |> \n",
    "    filter(cohort == \"Hippos\") |>\n",
    "    dplyr::select(patient_id, phenotype) |>\n",
    "    mutate(phenotype = replace_na(phenotype, \"Epidemic KS\")) |>\n",
    "    distinct()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "targeted_viral_expression <- read_excel(targerted_expression_path, sheet = 4)\n",
    "supplementary_figure_one <- targeted_viral_expression  |>\n",
    "    pivot_longer(c(-Target, -`Probe Name`), names_to = \"sample\", values_to = \"l2fc\") |>\n",
    "    mutate(sample = case_when(sample == \"8020\" ~ \"008_020\",\n",
    "        sample == \"030-B\" ~ \"008_030\", sample == \"028-B\" ~ \"008_028\",\n",
    "        sample == \"024-B\" ~ \"008_024\", sample == \"013-B\" ~ \"008_013\",\n",
    "        sample == \"008-C\" ~ \"008_008\", sample == \"001-B\" ~ \"008_001\",\n",
    "        TRUE ~ str_replace(sample, \"-\", \"_\")),\n",
    "      sample = factor(sample, levels = c(\"008_001\", \"008_008\", \"008_013\",\n",
    "          \"008_020\", \"008_024\", \"008_028\", \"008_030\", \"008_036\", \"008_037\",\n",
    "          \"008_052\", \"008_059\", \"008_061\", \"008_062\", \"008_075\", \"008_085\",\n",
    "          \"008_088\", \"008_092\", \"008_093\", \"008_097\", \"008_099\", \"008_101\",\n",
    "          \"008_106\"))) |>\n",
    "    left_join(phenotype_table, by = c(\"sample\" = \"patient_id\")) |>\n",
    "    arrange(sample) |>\n",
    "    ggplot(aes(y = `sample`, x = `Probe Name`, fill = l2fc)) +\n",
    "    geom_tile(color = \"white\", lwd = 0.5, linetype = 1) +\n",
    "    labs(x = \"Probe name\", y = \"Sample\", fill = \"Log2FC\") +\n",
    "    #scale_fill_gradientn(colors = rev(c('#b2182b','#d6604d','#d1e5f0'))) +\n",
    "    scale_fill_viridis_b(breaks = c(2,4,6,8,10)) +\n",
    "    theme_classic(base_size = 20) +\n",
    "    facet_grid(cols = vars(Target), rows = vars(phenotype), \n",
    "      scales = \"free\", space = \"free\", margins = F) +\n",
    "    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, face = \"bold\", family = 'NimbusSan'),\n",
    "        axis.text.y = element_text(face = \"bold\", family = 'NimbusSan'),\n",
    "         strip.text.x = element_text(face=\"bold\", family = 'NimbusSan'),\n",
    "         strip.text.y = element_text(face = \"bold\", angle = 0, family = 'NimbusSan'),\n",
    "        axis.title.x = element_blank(), \n",
    "        axis.line.y = element_blank(), legend.key.size = unit(1, \"cm\"),\n",
    "        axis.title.y = element_blank(), axis.ticks.y = element_blank()) \n",
    "ggsave(str_c(figures_path, \"pdf\", \"Supplementary_figure_one.pdf\", sep = \"/\"), supplementary_figure_one, \n",
    "  limitsize = FALSE, width = 20, height = 12, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Supplementary_figure_one.svg\", sep = \"/\"), supplementary_figure_one,\n",
    "  limitsize = FALSE, width = 20, height = 12, device = \"svg\")\n",
    "ggsave(str_c(figures_path, \"png\", \"Supplementary_figure_one.png\", sep = \"/\"), supplementary_figure_one,\n",
    "  limitsize = FALSE, width = 20, height = 12, device = \"png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary figure 2:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 119 × 24</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Mixture</th><th scope=col>cohort</th><th scope=col>B cells naive</th><th scope=col>B cells memory</th><th scope=col>Plasma cells</th><th scope=col>T cells CD8</th><th scope=col>T cells CD4 naive</th><th scope=col>T cells CD4 memory resting</th><th scope=col>T cells CD4 memory activated</th><th scope=col>T cells follicular helper</th><th scope=col>⋯</th><th scope=col>Monocytes</th><th scope=col>Macrophages M0</th><th scope=col>Macrophages M1</th><th scope=col>Macrophages M2</th><th scope=col>Dendritic cells resting</th><th scope=col>Dendritic cells activated</th><th scope=col>Mast cells resting</th><th scope=col>Mast cells activated</th><th scope=col>Eosinophils</th><th scope=col>Neutrophils</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GTEX-1117F-2926-SM-5GZYI</td><td>Non sun-exposed skin</td><td>0.000000</td><td>0.0000000</td><td>0</td><td>7.4767371</td><td>0</td><td> 5.5761521</td><td>0</td><td>3.842084</td><td>⋯</td><td>0.000000</td><td>5.678447</td><td>0.000000</td><td>18.45082</td><td> 0.000000</td><td>0.000000</td><td>55.45436</td><td>0</td><td>0</td><td>0.000000</td></tr>\n",
       "\t<tr><td>GTEX-111CU-1126-SM-5EGIM</td><td>Non sun-exposed skin</td><td>3.686808</td><td>0.0000000</td><td>0</td><td>1.6520993</td><td>0</td><td> 0.8580771</td><td>0</td><td>5.497539</td><td>⋯</td><td>0.000000</td><td>6.602357</td><td>0.000000</td><td>16.72780</td><td>11.188591</td><td>1.285114</td><td>47.10898</td><td>0</td><td>0</td><td>1.528191</td></tr>\n",
       "\t<tr><td>GTEX-111FC-2526-SM-5GZXU</td><td>Non sun-exposed skin</td><td>0.000000</td><td>0.7061475</td><td>0</td><td>0.7025772</td><td>0</td><td>10.7521579</td><td>0</td><td>9.144440</td><td>⋯</td><td>2.019597</td><td>0.000000</td><td>0.549818</td><td>14.75316</td><td> 1.070596</td><td>0.000000</td><td>54.85222</td><td>0</td><td>0</td><td>0.000000</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>SRR5787176</td><td>Tso et al.</td><td>0.000000</td><td>1.838458</td><td>9.234666</td><td> 7.635842</td><td>0</td><td>0.000000</td><td>0</td><td>3.198196</td><td>⋯</td><td>0.08701132</td><td>18.28287</td><td> 4.904041</td><td>36.51076</td><td>3.438229</td><td>0</td><td>5.2369415</td><td>0.000000</td><td>0</td><td>0.08552486</td></tr>\n",
       "\t<tr><td>SRR5787177</td><td>Tso et al.</td><td>2.546005</td><td>0.000000</td><td>6.736545</td><td>11.449475</td><td>0</td><td>1.000833</td><td>0</td><td>8.065807</td><td>⋯</td><td>0.00000000</td><td>13.29125</td><td>10.929159</td><td>27.07443</td><td>1.822308</td><td>0</td><td>0.4075915</td><td>2.379944</td><td>0</td><td>0.00000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 119 × 24\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Mixture & cohort & B cells naive & B cells memory & Plasma cells & T cells CD8 & T cells CD4 naive & T cells CD4 memory resting & T cells CD4 memory activated & T cells follicular helper & ⋯ & Monocytes & Macrophages M0 & Macrophages M1 & Macrophages M2 & Dendritic cells resting & Dendritic cells activated & Mast cells resting & Mast cells activated & Eosinophils & Neutrophils\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t GTEX-1117F-2926-SM-5GZYI & Non sun-exposed skin & 0.000000 & 0.0000000 & 0 & 7.4767371 & 0 &  5.5761521 & 0 & 3.842084 & ⋯ & 0.000000 & 5.678447 & 0.000000 & 18.45082 &  0.000000 & 0.000000 & 55.45436 & 0 & 0 & 0.000000\\\\\n",
       "\t GTEX-111CU-1126-SM-5EGIM & Non sun-exposed skin & 3.686808 & 0.0000000 & 0 & 1.6520993 & 0 &  0.8580771 & 0 & 5.497539 & ⋯ & 0.000000 & 6.602357 & 0.000000 & 16.72780 & 11.188591 & 1.285114 & 47.10898 & 0 & 0 & 1.528191\\\\\n",
       "\t GTEX-111FC-2526-SM-5GZXU & Non sun-exposed skin & 0.000000 & 0.7061475 & 0 & 0.7025772 & 0 & 10.7521579 & 0 & 9.144440 & ⋯ & 2.019597 & 0.000000 & 0.549818 & 14.75316 &  1.070596 & 0.000000 & 54.85222 & 0 & 0 & 0.000000\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t SRR5787176 & Tso et al. & 0.000000 & 1.838458 & 9.234666 &  7.635842 & 0 & 0.000000 & 0 & 3.198196 & ⋯ & 0.08701132 & 18.28287 &  4.904041 & 36.51076 & 3.438229 & 0 & 5.2369415 & 0.000000 & 0 & 0.08552486\\\\\n",
       "\t SRR5787177 & Tso et al. & 2.546005 & 0.000000 & 6.736545 & 11.449475 & 0 & 1.000833 & 0 & 8.065807 & ⋯ & 0.00000000 & 13.29125 & 10.929159 & 27.07443 & 1.822308 & 0 & 0.4075915 & 2.379944 & 0 & 0.00000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 119 × 24\n",
       "\n",
       "| Mixture &lt;chr&gt; | cohort &lt;chr&gt; | B cells naive &lt;dbl&gt; | B cells memory &lt;dbl&gt; | Plasma cells &lt;dbl&gt; | T cells CD8 &lt;dbl&gt; | T cells CD4 naive &lt;dbl&gt; | T cells CD4 memory resting &lt;dbl&gt; | T cells CD4 memory activated &lt;dbl&gt; | T cells follicular helper &lt;dbl&gt; | ⋯ ⋯ | Monocytes &lt;dbl&gt; | Macrophages M0 &lt;dbl&gt; | Macrophages M1 &lt;dbl&gt; | Macrophages M2 &lt;dbl&gt; | Dendritic cells resting &lt;dbl&gt; | Dendritic cells activated &lt;dbl&gt; | Mast cells resting &lt;dbl&gt; | Mast cells activated &lt;dbl&gt; | Eosinophils &lt;dbl&gt; | Neutrophils &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| GTEX-1117F-2926-SM-5GZYI | Non sun-exposed skin | 0.000000 | 0.0000000 | 0 | 7.4767371 | 0 |  5.5761521 | 0 | 3.842084 | ⋯ | 0.000000 | 5.678447 | 0.000000 | 18.45082 |  0.000000 | 0.000000 | 55.45436 | 0 | 0 | 0.000000 |\n",
       "| GTEX-111CU-1126-SM-5EGIM | Non sun-exposed skin | 3.686808 | 0.0000000 | 0 | 1.6520993 | 0 |  0.8580771 | 0 | 5.497539 | ⋯ | 0.000000 | 6.602357 | 0.000000 | 16.72780 | 11.188591 | 1.285114 | 47.10898 | 0 | 0 | 1.528191 |\n",
       "| GTEX-111FC-2526-SM-5GZXU | Non sun-exposed skin | 0.000000 | 0.7061475 | 0 | 0.7025772 | 0 | 10.7521579 | 0 | 9.144440 | ⋯ | 2.019597 | 0.000000 | 0.549818 | 14.75316 |  1.070596 | 0.000000 | 54.85222 | 0 | 0 | 0.000000 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| SRR5787176 | Tso et al. | 0.000000 | 1.838458 | 9.234666 |  7.635842 | 0 | 0.000000 | 0 | 3.198196 | ⋯ | 0.08701132 | 18.28287 |  4.904041 | 36.51076 | 3.438229 | 0 | 5.2369415 | 0.000000 | 0 | 0.08552486 |\n",
       "| SRR5787177 | Tso et al. | 2.546005 | 0.000000 | 6.736545 | 11.449475 | 0 | 1.000833 | 0 | 8.065807 | ⋯ | 0.00000000 | 13.29125 | 10.929159 | 27.07443 | 1.822308 | 0 | 0.4075915 | 2.379944 | 0 | 0.00000000 |\n",
       "\n"
      ],
      "text/plain": [
       "    Mixture                  cohort               B cells naive B cells memory\n",
       "1   GTEX-1117F-2926-SM-5GZYI Non sun-exposed skin 0.000000      0.0000000     \n",
       "2   GTEX-111CU-1126-SM-5EGIM Non sun-exposed skin 3.686808      0.0000000     \n",
       "3   GTEX-111FC-2526-SM-5GZXU Non sun-exposed skin 0.000000      0.7061475     \n",
       "⋮   ⋮                        ⋮                    ⋮             ⋮             \n",
       "118 SRR5787176               Tso et al.           0.000000      1.838458      \n",
       "119 SRR5787177               Tso et al.           2.546005      0.000000      \n",
       "    Plasma cells T cells CD8 T cells CD4 naive T cells CD4 memory resting\n",
       "1   0            7.4767371   0                  5.5761521                \n",
       "2   0            1.6520993   0                  0.8580771                \n",
       "3   0            0.7025772   0                 10.7521579                \n",
       "⋮   ⋮            ⋮           ⋮                 ⋮                         \n",
       "118 9.234666      7.635842   0                 0.000000                  \n",
       "119 6.736545     11.449475   0                 1.000833                  \n",
       "    T cells CD4 memory activated T cells follicular helper ⋯ Monocytes \n",
       "1   0                            3.842084                  ⋯ 0.000000  \n",
       "2   0                            5.497539                  ⋯ 0.000000  \n",
       "3   0                            9.144440                  ⋯ 2.019597  \n",
       "⋮   ⋮                            ⋮                         ⋱ ⋮         \n",
       "118 0                            3.198196                  ⋯ 0.08701132\n",
       "119 0                            8.065807                  ⋯ 0.00000000\n",
       "    Macrophages M0 Macrophages M1 Macrophages M2 Dendritic cells resting\n",
       "1   5.678447       0.000000       18.45082        0.000000              \n",
       "2   6.602357       0.000000       16.72780       11.188591              \n",
       "3   0.000000       0.549818       14.75316        1.070596              \n",
       "⋮   ⋮              ⋮              ⋮              ⋮                      \n",
       "118 18.28287        4.904041      36.51076       3.438229               \n",
       "119 13.29125       10.929159      27.07443       1.822308               \n",
       "    Dendritic cells activated Mast cells resting Mast cells activated\n",
       "1   0.000000                  55.45436           0                   \n",
       "2   1.285114                  47.10898           0                   \n",
       "3   0.000000                  54.85222           0                   \n",
       "⋮   ⋮                         ⋮                  ⋮                   \n",
       "118 0                         5.2369415          0.000000            \n",
       "119 0                         0.4075915          2.379944            \n",
       "    Eosinophils Neutrophils\n",
       "1   0           0.000000   \n",
       "2   0           1.528191   \n",
       "3   0           0.000000   \n",
       "⋮   ⋮           ⋮          \n",
       "118 0           0.08552486 \n",
       "119 0           0.00000000 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "readCiberSort <- function(path) {\n",
    "    study = basename(path)\n",
    "    cib_table <- read_csv(path, show_col_types = FALSE) |>\n",
    "        mutate(Mixture = str_remove(Mixture, \"_quant\"),\n",
    "            cohort = study,\n",
    "            cohort = case_when(cohort == \"gtex_nes.csv\" ~ \"Non sun-exposed skin\",\n",
    "                cohort == \"gtex_ses.csv\" ~ \"Sun-exposed skin\",\n",
    "                str_detect(cohort, \"lid\") ~ \"Lidenge et al.\",\n",
    "                str_detect(cohort, \"tso\") ~ \"Tso et al.\"))\n",
    "    return(cib_table)\n",
    "}\n",
    "\n",
    "\n",
    "public_cibersort_table <- list.files(cibersortx_results_path,\n",
    "        pattern = \"lid|tso|gtex\", full.names = TRUE) |>\n",
    "    map_df(readCiberSort)  |>\n",
    "    dplyr::select(-`P-value`, -Correlation, -RMSE) |> \n",
    "    pivot_longer(c(-Mixture, -cohort,  -`Absolute score (sig.score)`),\n",
    "        names_to = \"cell_type\", values_to = \"score\") |>\n",
    "    mutate(percentage = (score*100)/`Absolute score (sig.score)`) |>\n",
    "    pivot_wider(id_cols = c(Mixture, cohort), names_from = cell_type,\n",
    "        values_from = percentage)\n",
    "public_cibersort_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "supplementary_figure_two_heatmap <- public_cibersort_table |>\n",
    "  pivot_longer(cols = `B cells naive`:Neutrophils, names_to = \"cell_type\", \n",
    "    values_to = \"abundance\" ) |>\n",
    "  mutate(Mixture = if_else(cohort == \"Lidenge et al.\", \n",
    "    str_extract(Mixture, \"SRR\\\\d+$\"), Mixture)) |>\n",
    "  left_join(rna_seq_metadata, by = c(\"Mixture\" = \"repertoire_id\")) |>\n",
    "  mutate(phenotype = case_when(cohort.x %in% c(\"Non sun-exposed skin\", \"Sun-exposed skin\") & is.na(tissue_type) ~ \"Control\", \n",
    "    cohort.x == \"Lidenge et al.\" & tissue_type == \"Normal\" ~ \"Con-\\ntrol\",\n",
    "    tissue_type == \"Tumor\" & phenotype == \"Endemic KS\" ~ \"Ende-\\nmic KS\", \n",
    "    cohort.x == \"Tso et al.\" & tissue_type == \"Tumor\" & phenotype == \"Epidemic KS\" ~ \"Epide-\\nmic KS\", \n",
    "    tissue_type == \"Tumor\" & phenotype == \"Epidemic KS\" ~ \"Epidemic KS\",\n",
    "    TRUE ~ tissue_type),\n",
    "    cohort.x = factor(cohort.x, levels = c(\"Lidenge et al.\", \"Tso et al.\", \"Non sun-exposed skin\", \"Sun-exposed skin\")),\n",
    "    phenotype = factor(phenotype, levels = c(\"Con-\\ntrol\", \"Control\", \"NAT\", \"Ende-\\nmic KS\", \"Endemic KS\", \"Epide-\\nmic KS\", \"Epidemic KS\"))) |>\n",
    "  ggplot(aes(y = cell_type, x = Mixture, fill = abundance)) +\n",
    "  geom_tile(color = \"white\", lwd = 0.5, linetype = 1) +\n",
    "  labs(x = \"Sample\", y = \"Cell type\", fill = \"Cell type\\nabundance\") +\n",
    "  #scale_fill_gradientn(colors = rev(c('#b2182b','#d6604d','#d1e5f0'))) +\n",
    "  scale_fill_viridis_b(option = \"plasma\") + \n",
    "  theme_classic(base_size = 20) +\n",
    "  facet_nested(cols = vars(cohort.x, phenotype),scales = \"free\", space = \"free\", margins = F) +\n",
    "  theme(axis.text.x = element_blank(), strip.background = element_blank(),\n",
    "    axis.title.x = element_blank(), strip.text = element_text(face = \"bold\"),\n",
    "    axis.text.y = element_text(face = \"bold\"), axis.line = element_blank(), \n",
    "    axis.ticks = element_blank(),\n",
    "    text = element_text(family = 'NimbusSan')) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "supp_cohort_palatte <- c(\"Epidemic KS - NAT\" = \"#fb8072\", \n",
    "  \"Epidemic KS - Tumor\" = \"#e41a1c\", \"Endemic KS - NAT\" = \"#80b1d3\",\n",
    "  \"Endemic KS - Tumor\" = \"#377eb8\", \"Control\" = \"#67a628\")\n",
    "\n",
    "\n",
    "supplementary_figure_two_corrplot <- public_cibersort_table |>\n",
    "    mutate(Mixture = if_else(cohort == \"Lidenge et al.\", \n",
    "    str_extract(Mixture, \"SRR\\\\d+$\"), Mixture)) |>\n",
    "    left_join(rna_seq_metadata, by = c(\"Mixture\" = \"repertoire_id\")) |>\n",
    "    mutate(phenotype = case_when(cohort.x %in% c(\"Non sun-exposed skin\", \"Sun-exposed skin\") & is.na(tissue_type) ~ \"Control\", \n",
    "    cohort.x == \"Lidenge et al.\" & tissue_type == \"Normal\" ~ \"Control\",\n",
    "    tissue_type == \"Tumor\" & phenotype == \"Endemic KS\" ~ \"Endemic KS - Tumor\", \n",
    "    tissue_type == \"Tumor\" & phenotype == \"Epidemic KS\" ~ \"Epidemic KS - Tumor\", \n",
    "    tissue_type == \"NAT\" & phenotype == \"Epidemic KS\" ~ \"Epidemic KS - NAT\", \n",
    "    tissue_type == \"NAT\" & phenotype == \"Endemic KS\" ~ \"Endemic KS - NAT\"),\n",
    "    cohort.x = factor(cohort.x, levels = c(\"Lidenge et al.\", \"Tso et al.\", \"Non sun-exposed skin\", \"Sun-exposed skin\")),\n",
    "    phenotype = factor(phenotype)) |>\n",
    "    ggscatter(x = \"Macrophages M2\", y = \"T cells CD8\", color = \"phenotype\", \n",
    "        facet.by = \"cohort.x\", nrow = 1,\n",
    "    add = \"reg.line\")  + \n",
    "    stat_cor(aes(color = phenotype), p.accuracy = 0.001, r.accuracy = 0.01, size = 6) +\n",
    "    theme_classic(base_size = 20) + \n",
    "    labs(y = \"T cells CD8 (%)\", x = \"Macrophages M2 (%)\", color = \"Tissue type\") +\n",
    "    ylim(0,50) +\n",
    "    scale_color_manual(values = supp_cohort_palatte) +\n",
    "    scale_fill_manual(values = supp_cohort_palatte) +\n",
    "    theme(strip.text.x = element_text(face=\"bold\"),\n",
    "      axis.title = element_text(face = \"bold\"),\n",
    "      legend.position = \"right\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    " \n",
    "supplementary_figure_two <- supplementary_figure_two_heatmap / supplementary_figure_two_corrplot\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Supplementary_figure_two.pdf\", sep = \"/\"), supplementary_figure_two, \n",
    "  limitsize = FALSE, width = 30, height = 14, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Supplementary_figure_two.svg\", sep = \"/\"), supplementary_figure_two,\n",
    "  limitsize = FALSE, width = 30, height = 14, device = \"svg\")\n",
    "ggsave(str_c(figures_path, \"png\", \"Supplementary_figure_two.png\", sep = \"/\"), supplementary_figure_two,\n",
    "  limitsize = FALSE, width = 30, height = 14, device = \"png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary figure 3:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 798 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Function</th><th scope=col>Probe Name</th><th scope=col>sample</th><th scope=col>counts</th><th scope=col>tissue_type</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Macrophage</td><td>CD14</td><td>001-B</td><td>903.94</td><td>Lesion</td></tr>\n",
       "\t<tr><td>Macrophage</td><td>CD14</td><td>008-C</td><td>734.67</td><td>Lesion</td></tr>\n",
       "\t<tr><td>Macrophage</td><td>CD14</td><td>013-B</td><td>377.48</td><td>Lesion</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>T-cell inhibition</td><td>TIGIT</td><td>KS Norm</td><td>1.789021</td><td>NA</td></tr>\n",
       "\t<tr><td>T-cell inhibition</td><td>TIGIT</td><td>KS + HC</td><td>1.438920</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 798 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " Function & Probe Name & sample & counts & tissue\\_type\\\\\n",
       " <chr> & <chr> & <chr> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t Macrophage & CD14 & 001-B & 903.94 & Lesion\\\\\n",
       "\t Macrophage & CD14 & 008-C & 734.67 & Lesion\\\\\n",
       "\t Macrophage & CD14 & 013-B & 377.48 & Lesion\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t T-cell inhibition & TIGIT & KS Norm & 1.789021 & NA\\\\\n",
       "\t T-cell inhibition & TIGIT & KS + HC & 1.438920 & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 798 × 5\n",
       "\n",
       "| Function &lt;chr&gt; | Probe Name &lt;chr&gt; | sample &lt;chr&gt; | counts &lt;dbl&gt; | tissue_type &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| Macrophage | CD14 | 001-B | 903.94 | Lesion |\n",
       "| Macrophage | CD14 | 008-C | 734.67 | Lesion |\n",
       "| Macrophage | CD14 | 013-B | 377.48 | Lesion |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| T-cell inhibition | TIGIT | KS Norm | 1.789021 | NA |\n",
       "| T-cell inhibition | TIGIT | KS + HC | 1.438920 | NA |\n",
       "\n"
      ],
      "text/plain": [
       "    Function          Probe Name sample  counts   tissue_type\n",
       "1   Macrophage        CD14       001-B   903.94   Lesion     \n",
       "2   Macrophage        CD14       008-C   734.67   Lesion     \n",
       "3   Macrophage        CD14       013-B   377.48   Lesion     \n",
       "⋮   ⋮                 ⋮          ⋮       ⋮        ⋮          \n",
       "797 T-cell inhibition TIGIT      KS Norm 1.789021 NA         \n",
       "798 T-cell inhibition TIGIT      KS + HC 1.438920 NA         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nanostring_matrix <- read_excel(targerted_expression_path, sheet=1) |>\n",
    "  pivot_longer(cols = c(-`Probe Name`, -Function), names_to = \"sample\", values_to = \"counts\")\n",
    "nanostring_meta <- read_excel(targerted_expression_path, sheet=2) |>\n",
    "  pivot_longer(cols = `001-B`:His9DP, names_to = \"sample\", values_to = \"tissue_type\") |> \n",
    "  dplyr::select(sample, tissue_type) |> \n",
    "  distinct()\n",
    "\n",
    "nanostring_matrix <- left_join(nanostring_matrix, nanostring_meta, by = \"sample\")\n",
    "my_breaks <- c(0, 0.01, 0.1, 1, 10, 100, 1000, 10000)\n",
    "nanostring_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "nanostring_plot <- nanostring_matrix |> \n",
    "  dplyr::filter(!is.na(tissue_type) & sample != \"KS + HC\") |>\n",
    "  mutate(tissue_type = case_when(str_detect(sample, \"A$\") ~ \"NAT\",\n",
    "        str_detect(sample, \"^His\") ~ \"Histology\",\n",
    "        TRUE ~ \"Tumor\"),\n",
    "    Function = case_when(Function == \"Tumor proliferation\" ~ \"Proliferation\",\n",
    "      Function == \"T-cell proliferation\" ~ \"T-cell\",\n",
    "      Function == \"Tumor inhibition\" ~ \"M1 Macrophage\",\n",
    "      TRUE ~ Function),\n",
    "    Function = factor(Function, levels = c(\"T-cell\", \"CD4 T-cell\", \"CD8 T-cell\", \n",
    "      \"T-reg\", \"T-cell inhibition\", \"NK cell\", \n",
    "      \"Macrophage\", \"M1 Macrophage\", \"M2 Macrophage\", \n",
    "      \"Proliferation\"))) |>\n",
    "  filter(tissue_type != \"Histology\") |>\n",
    "  ggplot(aes(x = sample, y = `Probe Name`, fill = counts)) +\n",
    "  geom_tile(aes(width=0.9, height=0.9), color = \"black\", linewidth=0.1) +\n",
    "  labs(x = \"Samples\", y = \"Probe\", fill = \"Counts\") +\n",
    "  scale_fill_viridis_c(breaks = my_breaks, labels = my_breaks, \n",
    "                       trans = scales::pseudo_log_trans(sigma = 0.001)) +\n",
    "  theme_classic(base_size = 26) +\n",
    "  facet_grid(cols = vars(tissue_type), rows = vars(Function), scales = \"free\", space = \"free\", margins = F) +\n",
    "  theme(axis.text.x = element_blank(),  axis.title.x = element_blank(),\n",
    "    axis.line.y = element_blank(), axis.title.y = element_blank(), \n",
    "    axis.text.y = element_text(face = \"bold\"),\n",
    "    strip.placement = \"inside\", axis.ticks.y = element_blank(),\n",
    "    strip.text.y = element_blank(), strip.text.x = element_text(face = \"bold\"),\n",
    "    legend.position = \"bottom\",\n",
    "    legend.key.size = unit(2, \"cm\"),\n",
    "    text = element_text(family = \"NimbusSan\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "nanostringfc_plot <- nanostring_matrix |> \n",
    "  dplyr::filter(is.na(tissue_type) & sample != \"KS + HC\") |>\n",
    "  mutate(tissue_type = if_else(sample == \"KS + HC\", \"Lesion\\nvs\\nNormal/HC\", \"FC\"),\n",
    "    Function = case_when(Function == \"Tumor proliferation\" ~ \"Proliferation\",\n",
    "      Function == \"T-cell proliferation\" ~ \"T-cell\",\n",
    "      Function == \"Tumor inhibition\" ~ \"M1 Macrophage\",\n",
    "      TRUE ~ Function),\n",
    "    Function = factor(Function, levels = c(\"T-cell\", \"CD4 T-cell\", \"CD8 T-cell\", \n",
    "      \"T-reg\", \"T-cell inhibition\", \"NK cell\", \n",
    "      \"Macrophage\", \"M1 Macrophage\", \"M2 Macrophage\", \n",
    "      \"Proliferation\"))) |>\n",
    "  ggplot(aes(x = sample, y = `Probe Name`, fill = counts)) +\n",
    "  geom_tile(aes(width=0.9, height=0.9), color = \"black\", linewidth=0.1) +\n",
    "  labs(x = \"Samples\", y = \"Probe\", fill = \"Log2FC\") +\n",
    "  scale_fill_viridis_b(option = \"plasma\") +\n",
    "  theme_classic(base_size = 26) +\n",
    "  facet_grid(cols = vars(tissue_type), rows = vars(Function), scales = \"free\", space = \"free\", margins = F) +\n",
    "  theme(axis.text.x = element_blank(),  axis.title.x = element_blank(),\n",
    "    axis.text.y = element_blank(),  \n",
    "    axis.line.y = element_blank(), axis.title.y = element_blank(), \n",
    "    strip.placement = \"inside\", axis.ticks.y = element_blank(),\n",
    "    strip.text.x = element_text(face = \"bold\"),\n",
    "    strip.text.y = element_text(angle = 0, face = \"bold\"), legend.position = \"bottom\",\n",
    "    text = element_text(family = \"NimbusSan\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "plot_design <- \"11111111111111111111111111112\"\n",
    "supplementary_figure_three <- nanostring_plot + nanostringfc_plot +plot_layout(design = plot_design)\n",
    "\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Supplementary_figure_three.pdf\", sep = \"/\"), supplementary_figure_three, \n",
    "  limitsize = FALSE, width = 25, height = 15, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Supplementary_figure_three.svg\", sep = \"/\"), supplementary_figure_three,\n",
    "  limitsize = FALSE, width = 25, height = 15, device = \"svg\")\n",
    "ggsave(str_c(figures_path, \"png\", \"Supplementary_figure_three.png\", sep = \"/\"), supplementary_figure_three,\n",
    "  limitsize = FALSE, width = 25, height = 15, device = \"png\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 2: KS TIL repertoires are diverse and poorly correlated with viral loads "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AIRR-Seq data preparation\n",
    "The coming sections will outline the steps taken to analyze the following groups \n",
    "of TCR datasets \n",
    "1. Study entry samples with one NAT and one tumor sample from the Hippos cohort \n",
    "2. Pediatric BL samples from Ghana and Uganda\n",
    "\n",
    "Only datasets with more than 1000 sequences will be considered for the study\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading AIRR-Seq files [----------------------------------------] 2/816 (  0%) eta:  2m elapsed:  0s\n",
      "\n",
      "Reading AIRR-Seq files [----------------------------------------] 3/816 (  0%) eta:  3m elapsed:  1s\n",
      "\n",
      "Reading AIRR-Seq files [----------------------------------------] 4/816 (  0%) eta:  4m elapsed:  1s\n",
      "\n",
      "Reading AIRR-Seq files [----------------------------------------] 5/816 (  1%) eta:  5m elapsed:  2s\n",
      "\n",
      "Reading AIRR-Seq files [----------------------------------------] 6/816 (  1%) eta:  4m elapsed:  2s\n",
      "\n",
      "Reading AIRR-Seq files [----------------------------------------] 7/816 (  1%) eta:  5m elapsed:  3s\n",
      "\n",
      "Reading AIRR-Seq files [----------------------------------------] 8/816 (  1%) eta:  5m elapsed:  3s\n",
      "\n",
      "Reading AIRR-Seq files [----------------------------------------] 9/816 (  1%) eta:  6m elapsed:  4s\n",
      "\n",
      "Reading AIRR-Seq files [---------------------------------------] 10/816 (  1%) eta:  6m elapsed:  4s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 11/816 (  1%) eta:  6m elapsed:  5s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 12/816 (  1%) eta:  6m elapsed:  5s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 13/816 (  2%) eta:  6m elapsed:  5s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 14/816 (  2%) eta:  6m elapsed:  6s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 15/816 (  2%) eta:  5m elapsed:  6s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 16/816 (  2%) eta:  5m elapsed:  7s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 17/816 (  2%) eta:  5m elapsed:  7s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 18/816 (  2%) eta:  6m elapsed:  8s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 19/816 (  2%) eta:  6m elapsed:  8s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 20/816 (  2%) eta:  6m elapsed:  8s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 21/816 (  3%) eta:  6m elapsed:  9s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 22/816 (  3%) eta:  6m elapsed:  9s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 23/816 (  3%) eta:  6m elapsed: 10s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 24/816 (  3%) eta:  5m elapsed: 10s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 25/816 (  3%) eta:  6m elapsed: 11s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 26/816 (  3%) eta:  6m elapsed: 11s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 27/816 (  3%) eta:  5m elapsed: 11s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 28/816 (  3%) eta:  5m elapsed: 12s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 29/816 (  4%) eta:  6m elapsed: 12s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 30/816 (  4%) eta:  6m elapsed: 13s\n",
      "\n",
      "Reading AIRR-Seq files [>--------------------------------------] 31/816 (  4%) eta:  5m elapsed: 13s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 32/816 (  4%) eta:  5m elapsed: 13s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 33/816 (  4%) eta:  6m elapsed: 14s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 34/816 (  4%) eta:  6m elapsed: 15s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 35/816 (  4%) eta:  6m elapsed: 15s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 36/816 (  4%) eta:  6m elapsed: 16s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 37/816 (  5%) eta:  6m elapsed: 16s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 38/816 (  5%) eta:  6m elapsed: 17s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 39/816 (  5%) eta:  6m elapsed: 17s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 40/816 (  5%) eta:  6m elapsed: 17s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 41/816 (  5%) eta:  6m elapsed: 19s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 42/816 (  5%) eta:  6m elapsed: 19s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 43/816 (  5%) eta:  6m elapsed: 20s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 44/816 (  5%) eta:  6m elapsed: 20s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 45/816 (  6%) eta:  6m elapsed: 20s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 46/816 (  6%) eta:  6m elapsed: 21s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 47/816 (  6%) eta:  6m elapsed: 21s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 48/816 (  6%) eta:  6m elapsed: 22s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 49/816 (  6%) eta:  6m elapsed: 22s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 50/816 (  6%) eta:  6m elapsed: 22s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 51/816 (  6%) eta:  6m elapsed: 23s\n",
      "\n",
      "Reading AIRR-Seq files [=>-------------------------------------] 52/816 (  6%) eta:  6m elapsed: 23s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 53/816 (  6%) eta:  6m elapsed: 24s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 54/816 (  7%) eta:  6m elapsed: 24s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 55/816 (  7%) eta:  6m elapsed: 24s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 56/816 (  7%) eta:  6m elapsed: 25s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 57/816 (  7%) eta:  6m elapsed: 26s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 58/816 (  7%) eta:  6m elapsed: 26s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 59/816 (  7%) eta:  6m elapsed: 27s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 60/816 (  7%) eta:  6m elapsed: 27s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 61/816 (  7%) eta:  6m elapsed: 27s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 62/816 (  8%) eta:  6m elapsed: 28s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 63/816 (  8%) eta:  6m elapsed: 29s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 64/816 (  8%) eta:  6m elapsed: 29s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 65/816 (  8%) eta:  6m elapsed: 30s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 66/816 (  8%) eta:  6m elapsed: 30s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 67/816 (  8%) eta:  6m elapsed: 31s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 68/816 (  8%) eta:  6m elapsed: 32s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 69/816 (  8%) eta:  6m elapsed: 32s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 70/816 (  9%) eta:  6m elapsed: 33s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 71/816 (  9%) eta:  6m elapsed: 33s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 72/816 (  9%) eta:  6m elapsed: 33s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 73/816 (  9%) eta:  6m elapsed: 34s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 74/816 (  9%) eta:  6m elapsed: 34s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 75/816 (  9%) eta:  6m elapsed: 35s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 76/816 (  9%) eta:  6m elapsed: 35s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 77/816 (  9%) eta:  6m elapsed: 36s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 78/816 ( 10%) eta:  6m elapsed: 36s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 79/816 ( 10%) eta:  6m elapsed: 36s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 80/816 ( 10%) eta:  6m elapsed: 37s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 81/816 ( 10%) eta:  6m elapsed: 37s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 82/816 ( 10%) eta:  6m elapsed: 38s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 83/816 ( 10%) eta:  6m elapsed: 38s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 84/816 ( 10%) eta:  6m elapsed: 38s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 85/816 ( 10%) eta:  6m elapsed: 39s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 86/816 ( 11%) eta:  6m elapsed: 39s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 87/816 ( 11%) eta:  6m elapsed: 40s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 88/816 ( 11%) eta:  6m elapsed: 40s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 89/816 ( 11%) eta:  6m elapsed: 41s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 90/816 ( 11%) eta:  6m elapsed: 41s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 91/816 ( 11%) eta:  6m elapsed: 41s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 92/816 ( 11%) eta:  5m elapsed: 42s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 93/816 ( 11%) eta:  5m elapsed: 42s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 94/816 ( 12%) eta:  6m elapsed: 43s\n",
      "\n",
      "Reading AIRR-Seq files [====>----------------------------------] 95/816 ( 12%) eta:  6m elapsed: 44s\n",
      "\n",
      "Reading AIRR-Seq files [====>----------------------------------] 96/816 ( 12%) eta:  6m elapsed: 44s\n",
      "\n",
      "Reading AIRR-Seq files [====>----------------------------------] 97/816 ( 12%) eta:  6m elapsed: 45s\n",
      "\n",
      "Reading AIRR-Seq files [====>----------------------------------] 98/816 ( 12%) eta:  5m elapsed: 45s\n",
      "\n",
      "Reading AIRR-Seq files [====>----------------------------------] 99/816 ( 12%) eta:  5m elapsed: 45s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 100/816 ( 12%) eta:  5m elapsed: 46s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 101/816 ( 12%) eta:  5m elapsed: 46s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 102/816 ( 12%) eta:  5m elapsed: 46s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 103/816 ( 13%) eta:  5m elapsed: 47s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 104/816 ( 13%) eta:  5m elapsed: 47s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 105/816 ( 13%) eta:  5m elapsed: 48s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 106/816 ( 13%) eta:  5m elapsed: 48s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 107/816 ( 13%) eta:  5m elapsed: 49s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 108/816 ( 13%) eta:  5m elapsed: 49s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 109/816 ( 13%) eta:  5m elapsed: 49s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 110/816 ( 13%) eta:  5m elapsed: 50s\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 111/816 ( 14%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 112/816 ( 14%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 113/816 ( 14%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 114/816 ( 14%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 115/816 ( 14%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 116/816 ( 14%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 117/816 ( 14%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [====>---------------------------------] 118/816 ( 14%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 119/816 ( 15%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 120/816 ( 15%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 121/816 ( 15%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 122/816 ( 15%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 123/816 ( 15%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 124/816 ( 15%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 125/816 ( 15%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 126/816 ( 15%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 127/816 ( 16%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 128/816 ( 16%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 129/816 ( 16%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 130/816 ( 16%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 131/816 ( 16%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 132/816 ( 16%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 133/816 ( 16%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 134/816 ( 16%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 135/816 ( 17%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 136/816 ( 17%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 137/816 ( 17%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 138/816 ( 17%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=====>--------------------------------] 139/816 ( 17%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 140/816 ( 17%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 141/816 ( 17%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 142/816 ( 17%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 143/816 ( 18%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 144/816 ( 18%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 145/816 ( 18%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 146/816 ( 18%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 147/816 ( 18%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 148/816 ( 18%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 149/816 ( 18%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 150/816 ( 18%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 151/816 ( 19%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 152/816 ( 19%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 153/816 ( 19%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 154/816 ( 19%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 155/816 ( 19%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 156/816 ( 19%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 157/816 ( 19%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 158/816 ( 19%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 159/816 ( 19%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 160/816 ( 20%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [======>-------------------------------] 161/816 ( 20%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 162/816 ( 20%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 163/816 ( 20%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 164/816 ( 20%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 165/816 ( 20%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 166/816 ( 20%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 167/816 ( 20%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 168/816 ( 21%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 169/816 ( 21%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 170/816 ( 21%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 171/816 ( 21%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 172/816 ( 21%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 173/816 ( 21%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 174/816 ( 21%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 175/816 ( 21%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 176/816 ( 22%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 177/816 ( 22%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 178/816 ( 22%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 179/816 ( 22%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 180/816 ( 22%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 181/816 ( 22%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=======>------------------------------] 182/816 ( 22%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 183/816 ( 22%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 184/816 ( 23%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 185/816 ( 23%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 186/816 ( 23%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 187/816 ( 23%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 188/816 ( 23%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 189/816 ( 23%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 190/816 ( 23%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 191/816 ( 23%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 192/816 ( 24%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 193/816 ( 24%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 194/816 ( 24%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 195/816 ( 24%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 196/816 ( 24%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 197/816 ( 24%) eta:  5m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 198/816 ( 24%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 199/816 ( 24%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 200/816 ( 25%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 201/816 ( 25%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 202/816 ( 25%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========>-----------------------------] 203/816 ( 25%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 204/816 ( 25%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 205/816 ( 25%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 206/816 ( 25%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 207/816 ( 25%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 208/816 ( 25%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 209/816 ( 26%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 210/816 ( 26%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 211/816 ( 26%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 212/816 ( 26%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 213/816 ( 26%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 214/816 ( 26%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 215/816 ( 26%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 216/816 ( 26%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 217/816 ( 27%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 218/816 ( 27%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 219/816 ( 27%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 220/816 ( 27%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 221/816 ( 27%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 222/816 ( 27%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 223/816 ( 27%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 224/816 ( 27%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========>----------------------------] 225/816 ( 28%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 226/816 ( 28%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 227/816 ( 28%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 228/816 ( 28%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 229/816 ( 28%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 230/816 ( 28%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 231/816 ( 28%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 232/816 ( 28%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 233/816 ( 29%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 234/816 ( 29%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 235/816 ( 29%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 236/816 ( 29%) eta:  5m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 237/816 ( 29%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 238/816 ( 29%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 239/816 ( 29%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 240/816 ( 29%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 241/816 ( 30%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 242/816 ( 30%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 243/816 ( 30%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 244/816 ( 30%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 245/816 ( 30%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==========>---------------------------] 246/816 ( 30%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 247/816 ( 30%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 248/816 ( 30%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 249/816 ( 31%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 250/816 ( 31%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 251/816 ( 31%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 252/816 ( 31%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 253/816 ( 31%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 254/816 ( 31%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 255/816 ( 31%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 256/816 ( 31%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 257/816 ( 31%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 258/816 ( 32%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 259/816 ( 32%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 260/816 ( 32%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 261/816 ( 32%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 262/816 ( 32%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 263/816 ( 32%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 264/816 ( 32%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 265/816 ( 32%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 266/816 ( 33%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 267/816 ( 33%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===========>--------------------------] 268/816 ( 33%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 269/816 ( 33%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 270/816 ( 33%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 271/816 ( 33%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 272/816 ( 33%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 273/816 ( 33%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 274/816 ( 34%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 275/816 ( 34%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 276/816 ( 34%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 277/816 ( 34%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 278/816 ( 34%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 279/816 ( 34%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 280/816 ( 34%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 281/816 ( 34%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 282/816 ( 35%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 283/816 ( 35%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 284/816 ( 35%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 285/816 ( 35%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 286/816 ( 35%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 287/816 ( 35%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 288/816 ( 35%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [============>-------------------------] 289/816 ( 35%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 290/816 ( 36%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 291/816 ( 36%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 292/816 ( 36%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 293/816 ( 36%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 294/816 ( 36%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 295/816 ( 36%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 296/816 ( 36%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 297/816 ( 36%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 298/816 ( 37%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 299/816 ( 37%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 300/816 ( 37%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 301/816 ( 37%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 302/816 ( 37%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 303/816 ( 37%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 304/816 ( 37%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 305/816 ( 37%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 306/816 ( 38%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 307/816 ( 38%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 308/816 ( 38%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 309/816 ( 38%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 310/816 ( 38%) eta:  4m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=============>------------------------] 311/816 ( 38%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 312/816 ( 38%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 313/816 ( 38%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 314/816 ( 38%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 315/816 ( 39%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 316/816 ( 39%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 317/816 ( 39%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 318/816 ( 39%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 319/816 ( 39%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 320/816 ( 39%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 321/816 ( 39%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 322/816 ( 39%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 323/816 ( 40%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 324/816 ( 40%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 325/816 ( 40%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 326/816 ( 40%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 327/816 ( 40%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 328/816 ( 40%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 329/816 ( 40%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 330/816 ( 40%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 331/816 ( 41%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============>-----------------------] 332/816 ( 41%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 333/816 ( 41%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 334/816 ( 41%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 335/816 ( 41%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 336/816 ( 41%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 337/816 ( 41%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 338/816 ( 41%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 339/816 ( 42%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 340/816 ( 42%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 341/816 ( 42%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 342/816 ( 42%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 343/816 ( 42%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 344/816 ( 42%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 345/816 ( 42%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 346/816 ( 42%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 347/816 ( 43%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 348/816 ( 43%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 349/816 ( 43%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 350/816 ( 43%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 351/816 ( 43%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 352/816 ( 43%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 353/816 ( 43%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============>----------------------] 354/816 ( 43%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 355/816 ( 44%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 356/816 ( 44%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 357/816 ( 44%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 358/816 ( 44%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 359/816 ( 44%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 360/816 ( 44%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 361/816 ( 44%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 362/816 ( 44%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 363/816 ( 44%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 364/816 ( 45%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 365/816 ( 45%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 366/816 ( 45%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 367/816 ( 45%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 368/816 ( 45%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 369/816 ( 45%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 370/816 ( 45%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 371/816 ( 45%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 372/816 ( 46%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 373/816 ( 46%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 374/816 ( 46%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [================>---------------------] 375/816 ( 46%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 376/816 ( 46%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 377/816 ( 46%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 378/816 ( 46%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 379/816 ( 46%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 380/816 ( 47%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 381/816 ( 47%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 382/816 ( 47%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 383/816 ( 47%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 384/816 ( 47%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 385/816 ( 47%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 386/816 ( 47%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 387/816 ( 47%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 388/816 ( 48%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 389/816 ( 48%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 390/816 ( 48%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 391/816 ( 48%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 392/816 ( 48%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 393/816 ( 48%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 394/816 ( 48%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 395/816 ( 48%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 396/816 ( 49%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=================>--------------------] 397/816 ( 49%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 398/816 ( 49%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 399/816 ( 49%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 400/816 ( 49%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 401/816 ( 49%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 402/816 ( 49%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 403/816 ( 49%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 404/816 ( 50%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 405/816 ( 50%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 406/816 ( 50%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 407/816 ( 50%) eta:  4m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 408/816 ( 50%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 409/816 ( 50%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 410/816 ( 50%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 411/816 ( 50%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 412/816 ( 50%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 413/816 ( 51%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 414/816 ( 51%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 415/816 ( 51%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 416/816 ( 51%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 417/816 ( 51%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================>-------------------] 418/816 ( 51%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 419/816 ( 51%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 420/816 ( 51%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 421/816 ( 52%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 422/816 ( 52%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 423/816 ( 52%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 424/816 ( 52%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 425/816 ( 52%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 426/816 ( 52%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 427/816 ( 52%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 428/816 ( 52%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 429/816 ( 53%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 430/816 ( 53%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 431/816 ( 53%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 432/816 ( 53%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 433/816 ( 53%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 434/816 ( 53%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 435/816 ( 53%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 436/816 ( 53%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 437/816 ( 54%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 438/816 ( 54%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 439/816 ( 54%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================>------------------] 440/816 ( 54%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 441/816 ( 54%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 442/816 ( 54%) eta:  4m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 443/816 ( 54%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 444/816 ( 54%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 445/816 ( 55%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 446/816 ( 55%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 447/816 ( 55%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 448/816 ( 55%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 449/816 ( 55%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 450/816 ( 55%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 451/816 ( 55%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 452/816 ( 55%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 453/816 ( 56%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 454/816 ( 56%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 455/816 ( 56%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 456/816 ( 56%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 457/816 ( 56%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 458/816 ( 56%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 459/816 ( 56%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 460/816 ( 56%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [====================>-----------------] 461/816 ( 56%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 462/816 ( 57%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 463/816 ( 57%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 464/816 ( 57%) eta:  3m elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 465/816 ( 57%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 466/816 ( 57%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 467/816 ( 57%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 468/816 ( 57%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 469/816 ( 57%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 470/816 ( 58%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 471/816 ( 58%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 472/816 ( 58%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 473/816 ( 58%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 474/816 ( 58%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 475/816 ( 58%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 476/816 ( 58%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 477/816 ( 58%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 478/816 ( 59%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 479/816 ( 59%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 480/816 ( 59%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 481/816 ( 59%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 482/816 ( 59%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>----------------] 483/816 ( 59%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 484/816 ( 59%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 485/816 ( 59%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 486/816 ( 60%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 487/816 ( 60%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 488/816 ( 60%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 489/816 ( 60%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 490/816 ( 60%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 491/816 ( 60%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 492/816 ( 60%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 493/816 ( 60%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 494/816 ( 61%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 495/816 ( 61%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 496/816 ( 61%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 497/816 ( 61%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 498/816 ( 61%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 499/816 ( 61%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 500/816 ( 61%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 501/816 ( 61%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 502/816 ( 62%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 503/816 ( 62%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================>---------------] 504/816 ( 62%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 505/816 ( 62%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 506/816 ( 62%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 507/816 ( 62%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 508/816 ( 62%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 509/816 ( 62%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 510/816 ( 62%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 511/816 ( 63%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 512/816 ( 63%) eta:  3m elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 513/816 ( 63%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 514/816 ( 63%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 515/816 ( 63%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 516/816 ( 63%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 517/816 ( 63%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 518/816 ( 63%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 519/816 ( 64%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 520/816 ( 64%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 521/816 ( 64%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 522/816 ( 64%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 523/816 ( 64%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 524/816 ( 64%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 525/816 ( 64%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>--------------] 526/816 ( 64%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 527/816 ( 65%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 528/816 ( 65%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 529/816 ( 65%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 530/816 ( 65%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 531/816 ( 65%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 532/816 ( 65%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 533/816 ( 65%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 534/816 ( 65%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 535/816 ( 66%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 536/816 ( 66%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 537/816 ( 66%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 538/816 ( 66%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 539/816 ( 66%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 540/816 ( 66%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 541/816 ( 66%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 542/816 ( 66%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 543/816 ( 67%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 544/816 ( 67%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 545/816 ( 67%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 546/816 ( 67%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 547/816 ( 67%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 548/816 ( 67%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 549/816 ( 67%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 550/816 ( 67%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 551/816 ( 68%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 552/816 ( 68%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 553/816 ( 68%) eta:  3m elapsed:  6m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 554/816 ( 68%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 555/816 ( 68%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 556/816 ( 68%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 557/816 ( 68%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 558/816 ( 68%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 559/816 ( 69%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 560/816 ( 69%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 561/816 ( 69%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 562/816 ( 69%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 563/816 ( 69%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 564/816 ( 69%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 565/816 ( 69%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 566/816 ( 69%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 567/816 ( 69%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 568/816 ( 70%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 569/816 ( 70%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 570/816 ( 70%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 571/816 ( 70%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 572/816 ( 70%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 573/816 ( 70%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 574/816 ( 70%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 575/816 ( 70%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 576/816 ( 71%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 577/816 ( 71%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 578/816 ( 71%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 579/816 ( 71%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 580/816 ( 71%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 581/816 ( 71%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 582/816 ( 71%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 583/816 ( 71%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 584/816 ( 72%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 585/816 ( 72%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 586/816 ( 72%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 587/816 ( 72%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 588/816 ( 72%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 589/816 ( 72%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 590/816 ( 72%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 591/816 ( 72%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 592/816 ( 73%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 593/816 ( 73%) eta:  3m elapsed:  7m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 594/816 ( 73%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 595/816 ( 73%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 596/816 ( 73%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 597/816 ( 73%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 598/816 ( 73%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 599/816 ( 73%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 600/816 ( 74%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 601/816 ( 74%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 602/816 ( 74%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 603/816 ( 74%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 604/816 ( 74%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 605/816 ( 74%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 606/816 ( 74%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 607/816 ( 74%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 608/816 ( 75%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 609/816 ( 75%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 610/816 ( 75%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 611/816 ( 75%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 612/816 ( 75%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 613/816 ( 75%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 614/816 ( 75%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 615/816 ( 75%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 616/816 ( 75%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 617/816 ( 76%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 618/816 ( 76%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 619/816 ( 76%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 620/816 ( 76%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 621/816 ( 76%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 622/816 ( 76%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 623/816 ( 76%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 624/816 ( 76%) eta:  3m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 625/816 ( 77%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 626/816 ( 77%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 627/816 ( 77%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 628/816 ( 77%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 629/816 ( 77%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 630/816 ( 77%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 631/816 ( 77%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 632/816 ( 77%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 633/816 ( 78%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 634/816 ( 78%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 635/816 ( 78%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 636/816 ( 78%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 637/816 ( 78%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 638/816 ( 78%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 639/816 ( 78%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 640/816 ( 78%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 641/816 ( 79%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 642/816 ( 79%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 643/816 ( 79%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 644/816 ( 79%) eta:  2m elapsed:  8m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 645/816 ( 79%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 646/816 ( 79%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 647/816 ( 79%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 648/816 ( 79%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 649/816 ( 80%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 650/816 ( 80%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 651/816 ( 80%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 652/816 ( 80%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 653/816 ( 80%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 654/816 ( 80%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 655/816 ( 80%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 656/816 ( 80%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 657/816 ( 81%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 658/816 ( 81%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 659/816 ( 81%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 660/816 ( 81%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 661/816 ( 81%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 662/816 ( 81%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 663/816 ( 81%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 664/816 ( 81%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 665/816 ( 81%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 666/816 ( 82%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 667/816 ( 82%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 668/816 ( 82%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 669/816 ( 82%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 670/816 ( 82%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 671/816 ( 82%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 672/816 ( 82%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 673/816 ( 82%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 674/816 ( 83%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 675/816 ( 83%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 676/816 ( 83%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 677/816 ( 83%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 678/816 ( 83%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 679/816 ( 83%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 680/816 ( 83%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 681/816 ( 83%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 682/816 ( 84%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 683/816 ( 84%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 684/816 ( 84%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 685/816 ( 84%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 686/816 ( 84%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 687/816 ( 84%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 688/816 ( 84%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 689/816 ( 84%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 690/816 ( 85%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 691/816 ( 85%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 692/816 ( 85%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 693/816 ( 85%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 694/816 ( 85%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 695/816 ( 85%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 696/816 ( 85%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 697/816 ( 85%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 698/816 ( 86%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 699/816 ( 86%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 700/816 ( 86%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 701/816 ( 86%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 702/816 ( 86%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 703/816 ( 86%) eta:  1m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 704/816 ( 86%) eta:  2m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 705/816 ( 86%) eta:  1m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 706/816 ( 87%) eta:  1m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 707/816 ( 87%) eta:  1m elapsed:  9m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 708/816 ( 87%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 709/816 ( 87%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 710/816 ( 87%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 711/816 ( 87%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 712/816 ( 87%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 713/816 ( 87%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 714/816 ( 88%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 715/816 ( 88%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 716/816 ( 88%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 717/816 ( 88%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 718/816 ( 88%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 719/816 ( 88%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 720/816 ( 88%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 721/816 ( 88%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 722/816 ( 88%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 723/816 ( 89%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 724/816 ( 89%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 725/816 ( 89%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 726/816 ( 89%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 727/816 ( 89%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 728/816 ( 89%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 729/816 ( 89%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 730/816 ( 89%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 731/816 ( 90%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 732/816 ( 90%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 733/816 ( 90%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 734/816 ( 90%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 735/816 ( 90%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 736/816 ( 90%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 737/816 ( 90%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 738/816 ( 90%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 739/816 ( 91%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 740/816 ( 91%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 741/816 ( 91%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 742/816 ( 91%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 743/816 ( 91%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 744/816 ( 91%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 745/816 ( 91%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 746/816 ( 91%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 747/816 ( 92%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 748/816 ( 92%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 749/816 ( 92%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 750/816 ( 92%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 751/816 ( 92%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 752/816 ( 92%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 753/816 ( 92%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 754/816 ( 92%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 755/816 ( 93%) eta:  1m elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 756/816 ( 93%) eta: 50s elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 757/816 ( 93%) eta: 49s elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 758/816 ( 93%) eta: 48s elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 759/816 ( 93%) eta: 47s elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 760/816 ( 93%) eta: 46s elapsed: 10m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 761/816 ( 93%) eta: 46s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 762/816 ( 93%) eta: 45s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 763/816 ( 94%) eta: 44s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 764/816 ( 94%) eta: 43s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 765/816 ( 94%) eta: 42s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 766/816 ( 94%) eta: 41s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 767/816 ( 94%) eta: 40s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 768/816 ( 94%) eta: 40s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 769/816 ( 94%) eta: 39s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 770/816 ( 94%) eta: 38s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 771/816 ( 94%) eta: 37s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 772/816 ( 95%) eta: 36s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 773/816 ( 95%) eta: 35s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 774/816 ( 95%) eta: 35s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 775/816 ( 95%) eta: 34s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 776/816 ( 95%) eta: 33s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 777/816 ( 95%) eta: 32s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 778/816 ( 95%) eta: 31s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 779/816 ( 95%) eta: 30s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 780/816 ( 96%) eta: 30s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 781/816 ( 96%) eta: 29s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 782/816 ( 96%) eta: 28s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 783/816 ( 96%) eta: 27s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 784/816 ( 96%) eta: 26s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 785/816 ( 96%) eta: 25s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 786/816 ( 96%) eta: 24s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 787/816 ( 96%) eta: 24s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 788/816 ( 97%) eta: 23s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 789/816 ( 97%) eta: 22s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 790/816 ( 97%) eta: 21s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 791/816 ( 97%) eta: 20s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 792/816 ( 97%) eta: 20s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 793/816 ( 97%) eta: 19s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 794/816 ( 97%) eta: 18s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 795/816 ( 97%) eta: 17s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 796/816 ( 98%) eta: 16s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 797/816 ( 98%) eta: 15s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 798/816 ( 98%) eta: 15s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 799/816 ( 98%) eta: 14s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 800/816 ( 98%) eta: 13s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 801/816 ( 98%) eta: 12s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 802/816 ( 98%) eta: 11s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 803/816 ( 98%) eta: 11s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 804/816 ( 99%) eta: 10s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 805/816 ( 99%) eta:  9s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 806/816 ( 99%) eta:  8s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 807/816 ( 99%) eta:  7s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 808/816 ( 99%) eta:  6s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 809/816 ( 99%) eta:  6s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 810/816 ( 99%) eta:  5s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 811/816 ( 99%) eta:  4s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 812/816 (100%) eta:  3s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 813/816 (100%) eta:  2s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 814/816 (100%) eta:  2s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 815/816 (100%) eta:  1s elapsed: 11m\n",
      "\n",
      "Reading AIRR-Seq files [======================================] 816/816 (100%) eta:  0s elapsed: 11m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Subsetting productive sequences [------------]   0% eta:  4h\n",
      "\n",
      "Subsetting productive sequences [------------]   0% eta:  2h\n",
      "\n",
      "Subsetting productive sequences [------------]   0% eta:  1h\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta:  1h\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 45m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 46m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 42m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 39m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 46m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 43m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 40m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 37m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 35m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 33m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 31m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 32m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 30m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 29m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 28m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 27m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 26m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 25m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 24m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 23m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 22m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 21m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 24m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 23m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   4% eta: 23m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 22m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 21m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   6% eta: 20m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   6% eta: 19m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   6% eta: 18m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 18m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 17m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 18m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 17m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 16m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   8% eta: 17m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   8% eta: 16m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   9% eta: 16m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  10% eta: 15m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  10% eta: 14m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  11% eta: 14m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  11% eta: 15m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  11% eta: 14m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  12% eta: 14m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  12% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  12% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  13% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  13% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  14% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  15% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  16% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  17% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  19% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  19% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  20% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  21% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  21% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  22% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  23% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  24% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  25% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  26% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  26% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  27% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  28% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  29% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  29% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  30% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  31% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  32% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  32% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  33% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  34% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  35% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  36% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  37% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  38% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  38% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  39% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  40% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  41% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  42% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  43% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  44% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  45% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  46% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  46% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  47% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  48% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  49% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  49% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  50% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  51% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  51% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  52% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  53% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  54% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  54% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  55% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  56% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  57% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  57% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  57% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  58% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  58% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  59% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  59% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  60% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  61% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  61% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  61% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  62% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  62% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  62% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  62% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  62% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  63% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  64% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  64% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  65% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  66% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  66% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  67% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  67% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  67% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  68% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  69% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  70% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  71% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [========>---]  71% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [========>---]  72% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [========>---]  73% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [========>---]  74% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [========>---]  75% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [========>---]  76% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [========>---]  76% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  77% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  78% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  79% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  79% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  80% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  80% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  81% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  82% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  83% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  84% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  84% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  85% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  86% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  87% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  87% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  88% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  88% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  89% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  90% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  90% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  91% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  92% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  93% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  94% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  95% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  96% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [===========>]  96% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [===========>]  96% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===========>]  97% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 49s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 46s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 43s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 41s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 38s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 35s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 32s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 30s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 27s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 24s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 22s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 19s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 16s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 13s\n",
      "\n",
      "Subsetting productive sequences [===========>] 100% eta: 11s\n",
      "\n",
      "Subsetting productive sequences [===========>] 100% eta:  8s\n",
      "\n",
      "Subsetting productive sequences [===========>] 100% eta:  5s\n",
      "\n",
      "Subsetting productive sequences [===========>] 100% eta:  3s\n",
      "\n",
      "Subsetting productive sequences [============] 100% eta:  0s\n",
      "\n",
      "\n",
      "\n",
      "Subsetting productive sequences [------------]   0% eta:  2h\n",
      "\n",
      "Subsetting productive sequences [------------]   0% eta:  1h\n",
      "\n",
      "Subsetting productive sequences [------------]   0% eta: 49m\n",
      "\n",
      "Subsetting productive sequences [------------]   0% eta:  1h\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 43m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 36m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 40m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 36m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 33m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 30m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 39m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 36m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 33m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 31m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 29m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 27m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 26m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 24m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 25m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 24m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 23m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 22m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 21m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 20m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 21m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 20m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 19m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 18m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 18m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 17m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 16m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 17m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 16m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   4% eta: 16m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   4% eta: 17m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 16m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 15m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 16m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 15m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   6% eta: 15m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   6% eta: 14m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   6% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   8% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   8% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   8% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   9% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  10% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  10% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  11% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  12% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  12% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  12% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  13% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  14% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  14% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  14% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  14% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  15% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  15% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  15% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  16% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  17% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  17% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  19% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  20% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  21% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  21% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  22% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  23% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  24% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  25% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  26% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  27% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  27% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  27% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  28% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  28% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  29% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  29% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  29% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  29% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  30% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  31% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  32% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  33% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  34% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  35% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  36% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  37% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  38% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  38% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  39% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  40% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  41% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  42% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  43% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  44% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  45% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  46% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  46% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  47% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  48% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  49% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  49% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  50% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  51% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  51% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  51% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  51% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  52% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  53% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  54% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  54% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  55% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  56% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  57% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  58% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  59% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  59% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  59% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  60% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  61% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  62% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  62% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  63% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  64% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  64% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  64% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  64% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  65% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  66% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  66% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  67% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  68% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  69% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  70% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  71% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  71% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  72% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  73% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  74% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  75% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  76% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [========>---]  76% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [========>---]  77% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [========>---]  78% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [========>---]  79% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  79% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  80% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  81% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  81% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  82% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  83% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  84% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  85% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  85% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  86% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  87% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  88% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  88% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  89% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  89% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  90% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  91% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  92% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  93% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  94% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  95% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  95% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  96% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===========>]  96% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===========>]  97% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 48s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 45s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 43s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 40s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 38s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 36s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 33s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 31s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 28s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 26s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 24s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 21s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 19s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 16s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 14s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 12s\n",
      "\n",
      "Subsetting productive sequences [===========>] 100% eta:  9s\n",
      "\n",
      "Subsetting productive sequences [===========>] 100% eta:  7s\n",
      "\n",
      "Subsetting productive sequences [===========>] 100% eta:  5s\n",
      "\n",
      "Subsetting productive sequences [===========>] 100% eta:  2s\n",
      "\n",
      "Subsetting productive sequences [============] 100% eta:  0s\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Load AIRR-Seq data\n",
    "trb_paths <- list.files(c(hippos_path, blghana_path, bluganda_path, arks_path), \n",
    "    pattern = \"tsv\", recursive = TRUE, full.names = TRUE)\n",
    "study_raw_table <- readImmunoSeq(trb_paths)\n",
    "study_nprod_table <- productiveSeq(study_raw_table, aggregate = \"junction\")\n",
    "study_amino_acid_table <- productiveSeq(study_raw_table, aggregate = \"junction_aa\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Raw repertoire counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Total number of repertoires : 656\"\n",
      "[1] \"Total number of ptids : 215\"\n"
     ]
    }
   ],
   "source": [
    "annotated_nprod_table <- study_metadata |>\n",
    "  inner_join(study_nprod_table, by = c(\"trb_repertoire_id\" = \"repertoire_id\"))\n",
    "print(str_c(\"Total number of repertoires\", \n",
    "  annotated_nprod_table |> pull(trb_repertoire_id) |> unique() |> length(),\n",
    "  sep = \" : \"))\n",
    "print(str_c(\"Total number of ptids\", \n",
    "  annotated_nprod_table |> pull(patient_id) |> unique() |> length(),\n",
    "  sep = \" : \"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Hippos TRB count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "annotated_summary <- clonality(annotated_nprod_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 10 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>phenotype</th><th scope=col>tissue_type</th><th scope=col>number_patient</th><th scope=col>number_repertoires</th><th scope=col>number_trb_repertoires</th><th scope=col>number_rna_libraries</th><th scope=col>number_nanostring</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>BL - Ghana </td><td>BL         </td><td>Tumor</td><td> 37</td><td> 37</td><td> 37</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><td>BL - Uganda</td><td>BL         </td><td>Tumor</td><td> 12</td><td> 12</td><td> 12</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Endemic KS </td><td>NAT  </td><td> 25</td><td> 25</td><td> 25</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Endemic KS </td><td>PBMC </td><td> 15</td><td> 42</td><td> 42</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Endemic KS </td><td>Tumor</td><td> 36</td><td> 78</td><td> 78</td><td>12</td><td> 1</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Endemic KS </td><td>scREP</td><td>  1</td><td>  1</td><td>  1</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Epidemic KS</td><td>NAT  </td><td> 39</td><td> 39</td><td> 39</td><td> 0</td><td> 3</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Epidemic KS</td><td>PBMC </td><td> 33</td><td> 73</td><td> 73</td><td> 0</td><td> 0</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Epidemic KS</td><td>Tumor</td><td>101</td><td>221</td><td>221</td><td>18</td><td>17</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Epidemic KS</td><td>scREP</td><td>  5</td><td>  6</td><td>  7</td><td> 0</td><td> 0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 10 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " cohort & phenotype & tissue\\_type & number\\_patient & number\\_repertoires & number\\_trb\\_repertoires & number\\_rna\\_libraries & number\\_nanostring\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t BL - Ghana  & BL          & Tumor &  37 &  37 &  37 &  0 &  0\\\\\n",
       "\t BL - Uganda & BL          & Tumor &  12 &  12 &  12 &  0 &  0\\\\\n",
       "\t Hippos      & Endemic KS  & NAT   &  25 &  25 &  25 &  0 &  0\\\\\n",
       "\t Hippos      & Endemic KS  & PBMC  &  15 &  42 &  42 &  0 &  0\\\\\n",
       "\t Hippos      & Endemic KS  & Tumor &  36 &  78 &  78 & 12 &  1\\\\\n",
       "\t Hippos      & Endemic KS  & scREP &   1 &   1 &   1 &  0 &  0\\\\\n",
       "\t Hippos      & Epidemic KS & NAT   &  39 &  39 &  39 &  0 &  3\\\\\n",
       "\t Hippos      & Epidemic KS & PBMC  &  33 &  73 &  73 &  0 &  0\\\\\n",
       "\t Hippos      & Epidemic KS & Tumor & 101 & 221 & 221 & 18 & 17\\\\\n",
       "\t Hippos      & Epidemic KS & scREP &   5 &   6 &   7 &  0 &  0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 10 × 8\n",
       "\n",
       "| cohort &lt;chr&gt; | phenotype &lt;chr&gt; | tissue_type &lt;chr&gt; | number_patient &lt;int&gt; | number_repertoires &lt;int&gt; | number_trb_repertoires &lt;int&gt; | number_rna_libraries &lt;int&gt; | number_nanostring &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| BL - Ghana  | BL          | Tumor |  37 |  37 |  37 |  0 |  0 |\n",
       "| BL - Uganda | BL          | Tumor |  12 |  12 |  12 |  0 |  0 |\n",
       "| Hippos      | Endemic KS  | NAT   |  25 |  25 |  25 |  0 |  0 |\n",
       "| Hippos      | Endemic KS  | PBMC  |  15 |  42 |  42 |  0 |  0 |\n",
       "| Hippos      | Endemic KS  | Tumor |  36 |  78 |  78 | 12 |  1 |\n",
       "| Hippos      | Endemic KS  | scREP |   1 |   1 |   1 |  0 |  0 |\n",
       "| Hippos      | Epidemic KS | NAT   |  39 |  39 |  39 |  0 |  3 |\n",
       "| Hippos      | Epidemic KS | PBMC  |  33 |  73 |  73 |  0 |  0 |\n",
       "| Hippos      | Epidemic KS | Tumor | 101 | 221 | 221 | 18 | 17 |\n",
       "| Hippos      | Epidemic KS | scREP |   5 |   6 |   7 |  0 |  0 |\n",
       "\n"
      ],
      "text/plain": [
       "   cohort      phenotype   tissue_type number_patient number_repertoires\n",
       "1  BL - Ghana  BL          Tumor        37             37               \n",
       "2  BL - Uganda BL          Tumor        12             12               \n",
       "3  Hippos      Endemic KS  NAT          25             25               \n",
       "4  Hippos      Endemic KS  PBMC         15             42               \n",
       "5  Hippos      Endemic KS  Tumor        36             78               \n",
       "6  Hippos      Endemic KS  scREP         1              1               \n",
       "7  Hippos      Epidemic KS NAT          39             39               \n",
       "8  Hippos      Epidemic KS PBMC         33             73               \n",
       "9  Hippos      Epidemic KS Tumor       101            221               \n",
       "10 Hippos      Epidemic KS scREP         5              6               \n",
       "   number_trb_repertoires number_rna_libraries number_nanostring\n",
       "1   37                     0                    0               \n",
       "2   12                     0                    0               \n",
       "3   25                     0                    0               \n",
       "4   42                     0                    0               \n",
       "5   78                    12                    1               \n",
       "6    1                     0                    0               \n",
       "7   39                     0                    3               \n",
       "8   73                     0                    0               \n",
       "9  221                    18                   17               \n",
       "10   7                     0                    0               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Study datasets summary table\n",
    "# Displays all modalities of data avaiable from each dataset \n",
    "# Datasets are filtered out if their TRB sequencing contains less than 1000 sequences\n",
    "options(repr.matrix.max.rows=10, repr.matrix.max.columns=5)\n",
    "study_metadata |> \n",
    "    inner_join(annotated_summary, by = c(\"repertoire_id\" = \"repertoire_id\")) |>\n",
    "    filter(total_sequences >= 1000) |>\n",
    "    group_by(cohort, phenotype, tissue_type) |> \n",
    "    summarize(number_patient =  length(unique(patient_id)),\n",
    "        number_repertoires = length(unique(repertoire_id)),\n",
    "        number_trb_repertoires = length(unique(trb_repertoire_id)),\n",
    "        number_rna_libraries = length(unique(na.omit(rna_libraries))),\n",
    "        number_nanostring = length(unique(na.omit(nanostring_libraries))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "  1.000   1.000   1.000   1.871   3.000   7.000 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Tumors per individual\n",
    "\n",
    "study_metadata |> \n",
    "    inner_join(annotated_summary, by = c(\"repertoire_id\" = \"repertoire_id\")) |>\n",
    "    filter(total_sequences >= 1000 & tissue_type == \"Tumor\") |>\n",
    "    group_by(patient_id) |>\n",
    "    summarize(number_of_repertoires = length(unique(trb_repertoire_id))) |>\n",
    "    pull(number_of_repertoires) |>\n",
    "    summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 4 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>phenotype</th><th scope=col>number_of_repertoires</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>BL - Ghana </td><td>BL         </td><td> 37</td></tr>\n",
       "\t<tr><td>BL - Uganda</td><td>BL         </td><td> 12</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Endemic KS </td><td> 38</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Epidemic KS</td><td>106</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 4 × 3\n",
       "\\begin{tabular}{lll}\n",
       " cohort & phenotype & number\\_of\\_repertoires\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t BL - Ghana  & BL          &  37\\\\\n",
       "\t BL - Uganda & BL          &  12\\\\\n",
       "\t Hippos      & Endemic KS  &  38\\\\\n",
       "\t Hippos      & Epidemic KS & 106\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 4 × 3\n",
       "\n",
       "| cohort &lt;chr&gt; | phenotype &lt;chr&gt; | number_of_repertoires &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| BL - Ghana  | BL          |  37 |\n",
       "| BL - Uganda | BL          |  12 |\n",
       "| Hippos      | Endemic KS  |  38 |\n",
       "| Hippos      | Epidemic KS | 106 |\n",
       "\n"
      ],
      "text/plain": [
       "  cohort      phenotype   number_of_repertoires\n",
       "1 BL - Ghana  BL           37                  \n",
       "2 BL - Uganda BL           12                  \n",
       "3 Hippos      Endemic KS   38                  \n",
       "4 Hippos      Epidemic KS 106                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Repertoires per study\n",
    "options(repr.matrix.max.rows=5)\n",
    "study_metadata |> \n",
    "    inner_join(annotated_summary, by = c(\"repertoire_id\" = \"repertoire_id\")) |>\n",
    "    filter(total_sequences >= 1000) |>\n",
    "    group_by(cohort, phenotype) |>\n",
    "    summarize(number_of_repertoires = length(unique(patient_id))) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>hiv_status</th><th scope=col>count</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Negative</td><td>12</td></tr>\n",
       "\t<tr><td>Positive</td><td>39</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " hiv\\_status & count\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t Negative & 12\\\\\n",
       "\t Positive & 39\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2\n",
       "\n",
       "| hiv_status &lt;chr&gt; | count &lt;int&gt; |\n",
       "|---|---|\n",
       "| Negative | 12 |\n",
       "| Positive | 39 |\n",
       "\n"
      ],
      "text/plain": [
       "  hiv_status count\n",
       "1 Negative   12   \n",
       "2 Positive   39   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# RNA-Seq samples by HIV status\n",
    "study_metadata |> \n",
    "    filter(!is.na(rna_libraries) & cohort == \"Hippos\") |>\n",
    "    group_by(hiv_status) |>\n",
    "    summarize(count = length(unique(rna_libraries)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>tissue_type</th><th scope=col>count</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>NAT  </td><td> 5</td></tr>\n",
       "\t<tr><td>Tumor</td><td>27</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " tissue\\_type & count\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t NAT   &  5\\\\\n",
       "\t Tumor & 27\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2\n",
       "\n",
       "| tissue_type &lt;chr&gt; | count &lt;int&gt; |\n",
       "|---|---|\n",
       "| NAT   |  5 |\n",
       "| Tumor | 27 |\n",
       "\n"
      ],
      "text/plain": [
       "  tissue_type count\n",
       "1 NAT          5   \n",
       "2 Tumor       27   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Nanostring sample summary \n",
    "study_metadata |> \n",
    "    filter(!is.na(nanostring_libraries) & cohort == \"Hippos\") |>\n",
    "    group_by(tissue_type) |>\n",
    "    summarize(count = length(unique(nanostring_libraries)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 4 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>phenotype</th><th scope=col>number_patient</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>BL - Ghana </td><td>BL         </td><td> 37</td></tr>\n",
       "\t<tr><td>BL - Uganda</td><td>BL         </td><td> 12</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Endemic KS </td><td> 38</td></tr>\n",
       "\t<tr><td>Hippos     </td><td>Epidemic KS</td><td>106</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 4 × 3\n",
       "\\begin{tabular}{lll}\n",
       " cohort & phenotype & number\\_patient\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t BL - Ghana  & BL          &  37\\\\\n",
       "\t BL - Uganda & BL          &  12\\\\\n",
       "\t Hippos      & Endemic KS  &  38\\\\\n",
       "\t Hippos      & Epidemic KS & 106\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 4 × 3\n",
       "\n",
       "| cohort &lt;chr&gt; | phenotype &lt;chr&gt; | number_patient &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| BL - Ghana  | BL          |  37 |\n",
       "| BL - Uganda | BL          |  12 |\n",
       "| Hippos      | Endemic KS  |  38 |\n",
       "| Hippos      | Epidemic KS | 106 |\n",
       "\n"
      ],
      "text/plain": [
       "  cohort      phenotype   number_patient\n",
       "1 BL - Ghana  BL           37           \n",
       "2 BL - Uganda BL           12           \n",
       "3 Hippos      Endemic KS   38           \n",
       "4 Hippos      Epidemic KS 106           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "study_metadata |> \n",
    "    inner_join(annotated_summary, by = c(\"repertoire_id\" = \"repertoire_id\")) |>\n",
    "    filter(total_sequences >= 1000 ) |>\n",
    "    group_by(cohort, phenotype) |> \n",
    "    summarize(number_patient =  length(unique(patient_id)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get summarized Renyi numbers\n",
    "Vegan has a wonderful utility to calculate the Renyi numbers accumulated over \n",
    "100 iterations. Since vegan will sample down to the smallest sample, we will \n",
    "need to see how many of these samlpes have at least 5000 sequences, as well as\n",
    "how many of them have 1000 sequences so that we can pick a sample diversity \n",
    "threshold based on depth of sequencing. Using LymphoSeq2 we can get a summary of\n",
    "repertoire characteristics for all 656 re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>phenotype</th><th scope=col>number_patient</th><th scope=col>number_repertoires</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>Endemic KS </td><td> 38</td><td> 99</td></tr>\n",
       "\t<tr><td>Hippos</td><td>Epidemic KS</td><td>105</td><td>214</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 4\n",
       "\\begin{tabular}{llll}\n",
       " cohort & phenotype & number\\_patient & number\\_repertoires\\\\\n",
       " <chr> & <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t Hippos & Endemic KS  &  38 &  99\\\\\n",
       "\t Hippos & Epidemic KS & 105 & 214\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 4\n",
       "\n",
       "| cohort &lt;chr&gt; | phenotype &lt;chr&gt; | number_patient &lt;int&gt; | number_repertoires &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| Hippos | Endemic KS  |  38 |  99 |\n",
       "| Hippos | Epidemic KS | 105 | 214 |\n",
       "\n"
      ],
      "text/plain": [
       "  cohort phenotype   number_patient number_repertoires\n",
       "1 Hippos Endemic KS   38             99               \n",
       "2 Hippos Epidemic KS 105            214               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "study_metadata |> \n",
    "    inner_join(annotated_summary, by = c(\"repertoire_id\" = \"repertoire_id\")) |>\n",
    "    filter(total_sequences >= 1000 & visit_code == \"V01\") |>\n",
    "    group_by(cohort, phenotype) |> \n",
    "    summarize(number_patient =  length(unique(patient_id)),\n",
    "        number_repertoires = length(unique(repertoire_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 4521328 × 60</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>tra_repertoire_id</th><th scope=col>trb_repertoire_id</th><th scope=col>rna_libraries</th><th scope=col>nanostring_libraries</th><th scope=col>gex_library</th><th scope=col>vdj_library</th><th scope=col>sc_library_count</th><th scope=col>⋯</th><th scope=col>junction_aa</th><th scope=col>v_call</th><th scope=col>d_call</th><th scope=col>j_call</th><th scope=col>v_family</th><th scope=col>d_family</th><th scope=col>j_family</th><th scope=col>reading_frame</th><th scope=col>duplicate_count</th><th scope=col>duplicate_frequency</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_B</td><td>008_001_B</td><td>008_001_B</td><td>008-001-B-12</td><td>001-B</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>PNDPTNEKLFF    </td><td>TRBV19-1</td><td>TRBD2-1</td><td>TRBJ1-4</td><td>TRBV19</td><td>TRBD2</td><td>TRBJ1</td><td>in-frame</td><td> 1</td><td>3.218953e-05</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_B</td><td>008_001_B</td><td>008_001_B</td><td>008-001-B-12</td><td>001-B</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>RASSPPLGLNYGYTF</td><td>TRBV7-3 </td><td>TRBD1-1</td><td>TRBJ1-2</td><td>TRBV7 </td><td>TRBD1</td><td>TRBJ1</td><td>in-frame</td><td>10</td><td>3.218953e-04</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_B</td><td>008_001_B</td><td>008_001_B</td><td>008-001-B-12</td><td>001-B</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>RASSSPGTGCGYTF </td><td>TRBV7-3 </td><td>TRBD1-1</td><td>TRBJ1-2</td><td>TRBV7 </td><td>TRBD1</td><td>TRBJ1</td><td>in-frame</td><td> 1</td><td>3.218953e-05</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>BL - Uganda</td><td>009_0249</td><td>MVQ194745A</td><td>NA</td><td>MVQ194745A</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>CASNLDSREGGGAEAFF</td><td>TRBV5-5</td><td>NA     </td><td>TRBJ1-1</td><td>TRBV5</td><td>NA   </td><td>TRBJ1</td><td>in-frame</td><td>1</td><td>0.0004258944</td></tr>\n",
       "\t<tr><td>BL - Uganda</td><td>009_0249</td><td>MVQ194745A</td><td>NA</td><td>MVQ194745A</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>⋯</td><td>CASSLHPVGGTSYEQYF</td><td>TRBV5-5</td><td>TRBD2-1</td><td>TRBJ2-7</td><td>TRBV5</td><td>TRBD2</td><td>TRBJ2</td><td>in-frame</td><td>1</td><td>0.0004258944</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 4521328 × 60\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " cohort & patient\\_id & repertoire\\_id & tra\\_repertoire\\_id & trb\\_repertoire\\_id & rna\\_libraries & nanostring\\_libraries & gex\\_library & vdj\\_library & sc\\_library\\_count & ⋯ & junction\\_aa & v\\_call & d\\_call & j\\_call & v\\_family & d\\_family & j\\_family & reading\\_frame & duplicate\\_count & duplicate\\_frequency\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_B & 008\\_001\\_B & 008\\_001\\_B & 008-001-B-12 & 001-B & NA & NA & NA & ⋯ & PNDPTNEKLFF     & TRBV19-1 & TRBD2-1 & TRBJ1-4 & TRBV19 & TRBD2 & TRBJ1 & in-frame &  1 & 3.218953e-05\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_B & 008\\_001\\_B & 008\\_001\\_B & 008-001-B-12 & 001-B & NA & NA & NA & ⋯ & RASSPPLGLNYGYTF & TRBV7-3  & TRBD1-1 & TRBJ1-2 & TRBV7  & TRBD1 & TRBJ1 & in-frame & 10 & 3.218953e-04\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_B & 008\\_001\\_B & 008\\_001\\_B & 008-001-B-12 & 001-B & NA & NA & NA & ⋯ & RASSSPGTGCGYTF  & TRBV7-3  & TRBD1-1 & TRBJ1-2 & TRBV7  & TRBD1 & TRBJ1 & in-frame &  1 & 3.218953e-05\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t BL - Uganda & 009\\_0249 & MVQ194745A & NA & MVQ194745A & NA & NA & NA & NA & NA & ⋯ & CASNLDSREGGGAEAFF & TRBV5-5 & NA      & TRBJ1-1 & TRBV5 & NA    & TRBJ1 & in-frame & 1 & 0.0004258944\\\\\n",
       "\t BL - Uganda & 009\\_0249 & MVQ194745A & NA & MVQ194745A & NA & NA & NA & NA & NA & ⋯ & CASSLHPVGGTSYEQYF & TRBV5-5 & TRBD2-1 & TRBJ2-7 & TRBV5 & TRBD2 & TRBJ2 & in-frame & 1 & 0.0004258944\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 4521328 × 60\n",
       "\n",
       "| cohort &lt;chr&gt; | patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | tra_repertoire_id &lt;chr&gt; | trb_repertoire_id &lt;chr&gt; | rna_libraries &lt;chr&gt; | nanostring_libraries &lt;chr&gt; | gex_library &lt;chr&gt; | vdj_library &lt;chr&gt; | sc_library_count &lt;dbl&gt; | ⋯ ⋯ | junction_aa &lt;chr&gt; | v_call &lt;chr&gt; | d_call &lt;chr&gt; | j_call &lt;chr&gt; | v_family &lt;chr&gt; | d_family &lt;chr&gt; | j_family &lt;chr&gt; | reading_frame &lt;chr&gt; | duplicate_count &lt;dbl&gt; | duplicate_frequency &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Hippos | 008_001 | 008_001_B | 008_001_B | 008_001_B | 008-001-B-12 | 001-B | NA | NA | NA | ⋯ | PNDPTNEKLFF     | TRBV19-1 | TRBD2-1 | TRBJ1-4 | TRBV19 | TRBD2 | TRBJ1 | in-frame |  1 | 3.218953e-05 |\n",
       "| Hippos | 008_001 | 008_001_B | 008_001_B | 008_001_B | 008-001-B-12 | 001-B | NA | NA | NA | ⋯ | RASSPPLGLNYGYTF | TRBV7-3  | TRBD1-1 | TRBJ1-2 | TRBV7  | TRBD1 | TRBJ1 | in-frame | 10 | 3.218953e-04 |\n",
       "| Hippos | 008_001 | 008_001_B | 008_001_B | 008_001_B | 008-001-B-12 | 001-B | NA | NA | NA | ⋯ | RASSSPGTGCGYTF  | TRBV7-3  | TRBD1-1 | TRBJ1-2 | TRBV7  | TRBD1 | TRBJ1 | in-frame |  1 | 3.218953e-05 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| BL - Uganda | 009_0249 | MVQ194745A | NA | MVQ194745A | NA | NA | NA | NA | NA | ⋯ | CASNLDSREGGGAEAFF | TRBV5-5 | NA      | TRBJ1-1 | TRBV5 | NA    | TRBJ1 | in-frame | 1 | 0.0004258944 |\n",
       "| BL - Uganda | 009_0249 | MVQ194745A | NA | MVQ194745A | NA | NA | NA | NA | NA | ⋯ | CASSLHPVGGTSYEQYF | TRBV5-5 | TRBD2-1 | TRBJ2-7 | TRBV5 | TRBD2 | TRBJ2 | in-frame | 1 | 0.0004258944 |\n",
       "\n"
      ],
      "text/plain": [
       "        cohort      patient_id repertoire_id tra_repertoire_id\n",
       "1       Hippos      008_001    008_001_B     008_001_B        \n",
       "2       Hippos      008_001    008_001_B     008_001_B        \n",
       "3       Hippos      008_001    008_001_B     008_001_B        \n",
       "⋮       ⋮           ⋮          ⋮             ⋮                \n",
       "4521327 BL - Uganda 009_0249   MVQ194745A    NA               \n",
       "4521328 BL - Uganda 009_0249   MVQ194745A    NA               \n",
       "        trb_repertoire_id rna_libraries nanostring_libraries gex_library\n",
       "1       008_001_B         008-001-B-12  001-B                NA         \n",
       "2       008_001_B         008-001-B-12  001-B                NA         \n",
       "3       008_001_B         008-001-B-12  001-B                NA         \n",
       "⋮       ⋮                 ⋮             ⋮                    ⋮          \n",
       "4521327 MVQ194745A        NA            NA                   NA         \n",
       "4521328 MVQ194745A        NA            NA                   NA         \n",
       "        vdj_library sc_library_count ⋯ junction_aa       v_call   d_call \n",
       "1       NA          NA               ⋯ PNDPTNEKLFF       TRBV19-1 TRBD2-1\n",
       "2       NA          NA               ⋯ RASSPPLGLNYGYTF   TRBV7-3  TRBD1-1\n",
       "3       NA          NA               ⋯ RASSSPGTGCGYTF    TRBV7-3  TRBD1-1\n",
       "⋮       ⋮           ⋮                ⋱ ⋮                 ⋮        ⋮      \n",
       "4521327 NA          NA               ⋯ CASNLDSREGGGAEAFF TRBV5-5  NA     \n",
       "4521328 NA          NA               ⋯ CASSLHPVGGTSYEQYF TRBV5-5  TRBD2-1\n",
       "        j_call  v_family d_family j_family reading_frame duplicate_count\n",
       "1       TRBJ1-4 TRBV19   TRBD2    TRBJ1    in-frame       1             \n",
       "2       TRBJ1-2 TRBV7    TRBD1    TRBJ1    in-frame      10             \n",
       "3       TRBJ1-2 TRBV7    TRBD1    TRBJ1    in-frame       1             \n",
       "⋮       ⋮       ⋮        ⋮        ⋮        ⋮             ⋮              \n",
       "4521327 TRBJ1-1 TRBV5    NA       TRBJ1    in-frame      1              \n",
       "4521328 TRBJ2-7 TRBV5    TRBD2    TRBJ2    in-frame      1              \n",
       "        duplicate_frequency\n",
       "1       3.218953e-05       \n",
       "2       3.218953e-04       \n",
       "3       3.218953e-05       \n",
       "⋮       ⋮                  \n",
       "4521327 0.0004258944       \n",
       "4521328 0.0004258944       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "study_sequenced_fitlered_sample_list <- annotated_summary |> filter(total_sequences > 1000)\n",
    "study_annotated_nprod_table <- annotated_nprod_table |>\n",
    "  filter(repertoire_id %in% study_sequenced_fitlered_sample_list$repertoire_id)\n",
    "study_annotated_nprod_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Sampled Renyi numbers\n",
    "Given that the sequencing depth across all the samples vary quite a lot due to\n",
    "various factors, we will need to normalize the diversity metrics by depth. One\n",
    "way to do this is to iteratively sample all the datasets to the lowest possible\n",
    "depth of sequencing and then calculate the Renyi metrics for each iteration. \n",
    "This can be then average to get normalized Renyi metrics for each repertoire"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# This function randomly samples 1000 TRB sequences from each repertoire and \n",
    "# iteratively calls the sampledRenyi function based on user input. Finally\n",
    "# it calculates the mean Renyi numbers for a sample across all the iterations\n",
    "iterativeRenyi <- function(study_table, iterations, min_count = 1000) {\n",
    "    uncount_table <- study_table %>% \n",
    "        tidyr::uncount(weights = duplicate_count) \n",
    "    progress_bar <- progress::progress_bar$new(\n",
    "        format = \"Iteration [:bar] :current/:total (:percent) eta: :eta elapsed: :elapsed\",\n",
    "        total = iterations, clear = FALSE, width = 100)\n",
    "    progress_bar$tick(0)\n",
    "    srenyi_table <- purrr::map(1:iterations, \n",
    "                                \\(i)sampledRenyi(uncount_table, min_count, progress_bar)) %>%\n",
    "        dplyr::bind_rows() %>% \n",
    "        dplyr::group_by(repertoire_id) %>%\n",
    "        dplyr::summarise_all(mean)\n",
    "    return(srenyi_table)\n",
    "}\n",
    "# Given an uncounted table of TCR sequences with each sequence in the repertoire\n",
    "# represented as a unique row of count 1. This function will first collapse \n",
    "# all rows by junction_aa, calculate the total count and frequency of each\n",
    "# unique CDR3 sequence and then use the vegan package to calculate the Renyi \n",
    "# numbers for the sampled repertoire across multiple values of alpha\n",
    "sampledRenyi <- function(study_table, min_count, progress) {\n",
    "    progress$tick()\n",
    "    study_table <- study_table %>% \n",
    "        dplyr::group_by(repertoire_id) |>\n",
    "        dplyr::sample_n(min_count) %>%\n",
    "        dplyr::select(-duplicate_frequency) %>% \n",
    "        dplyr::group_by(repertoire_id, junction_aa) %>% \n",
    "        dplyr::summarise(duplicate_count = dplyr::n()) %>%\n",
    "        dplyr::group_by(repertoire_id) |>\n",
    "        dplyr::mutate(duplicate_frequency = duplicate_count/sum(duplicate_count)) %>%\n",
    "        dplyr::ungroup() \n",
    "    study_matrix <- study_table |>\n",
    "      pivot_wider(id_cols = repertoire_id, names_from = junction_aa,\n",
    "        values_from = duplicate_frequency, values_fn = sum, values_fill = 0)\n",
    "    rep_names <- study_matrix |>\n",
    "      pull(repertoire_id)\n",
    "    study_matrix <- study_matrix |>\n",
    "      select(-repertoire_id) |>\n",
    "      as.matrix()\n",
    "    rownames(study_matrix) <- rep_names\n",
    "    if (length(rep_names) == 1) {\n",
    "      renyi_table <- renyi(study_matrix) |>\n",
    "        as_tibble_row() |>\n",
    "        mutate(repertoire_id = rownames(study_matrix))\n",
    "    } else {\n",
    "      renyi_table <- renyi(study_matrix) |>\n",
    "        as_tibble() |>\n",
    "        mutate(repertoire_id = rownames(study_matrix))\n",
    "    }\n",
    "    return(renyi_table)\n",
    "}\n",
    "study_renyi_table <- study_annotated_nprod_table |>\n",
    "  iterativeRenyi(100, min_count = 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### KS TME visit one sample list\n",
    "\n",
    "Since the evaluation of the TME is primarily looking at the visit one tumor-NAT\n",
    "samples. Lets now identify these samples tumor-NAT samples that pass our \n",
    "sequencing filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"number of PTID with complete visitone data:59\"\n",
      "[1] \"number of repertoires with complete visitone data:162\"\n",
      "[1] \"number of repertoires with complete visitone data:162\"\n"
     ]
    }
   ],
   "source": [
    "ks_visitone <- study_metadata |> \n",
    "  filter(repertoire_id %in% study_sequenced_fitlered_sample_list$repertoire_id &\n",
    "    phenotype %in% c(\"Epidemic KS\", \"Endemic KS\") & \n",
    "    tissue_type %in% c(\"Tumor\", \"NAT\")) |> \n",
    "  filter(visit_code == \"V01\") |> \n",
    "  group_by(phenotype, patient_id, tissue_type) |> \n",
    "  summarize(reps =length(unique(repertoire_id))) |> \n",
    "  pivot_wider(id_cols = patient_id, names_from = tissue_type, \n",
    "    values_from = reps) |> \n",
    "  filter(!is.na(NAT) & !is.na(Tumor)) |>\n",
    "  mutate(total = sum(NAT, Tumor)) \n",
    "print(str_c(\"number of PTID with complete visitone data\", length(ks_visitone$patient_id),\n",
    "  sep = \":\"))\n",
    "print(str_c(\"number of repertoires with complete visitone data\", sum(ks_visitone$total),\n",
    "  sep = \":\"))\n",
    "print(str_c(\"number of repertoires with complete visitone data\", sum(ks_visitone$total),\n",
    "  sep = \":\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Checkpoint 3:\n",
    "\n",
    "At this point lets create our first checkpoint, we will store all the AIRR-Seq\n",
    "data, the metadata and summary metrics in RDA files for ease of access"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Save raw study table, and annotated nucleotide table\n",
    "save(study_raw_table, study_annotated_nprod_table, study_amino_acid_table,\n",
    "    file = str_c(checkpoint_path, \"kstme_paper_raw_tables.rda\", sep = \"/\"))\n",
    "# Save Renyi summary metrics, clonality table, and a list of samples that passed\n",
    "# the seqeuencing threshold of 1000 total sequences\n",
    "save(annotated_summary, study_renyi_table, study_sequenced_fitlered_sample_list, \n",
    "    file = str_c(checkpoint_path, \"kstme_paper_tcr_tables.rda\", sep = \"/\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "load(str_c(checkpoint_path, \"kstme_paper_tcr_tables.rda\", sep = \"/\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Subset tumor and NAT samples from the Renyi table and merge with metadata\n",
    "til_normal_renyi <- study_renyi_table |> \n",
    "    dplyr::rename(trb_repertoire_id = repertoire_id) |>\n",
    "    pivot_longer(cols = `0`:`Inf`, names_to = \"alpha\", \n",
    "        values_to = \"renyi_number\") |>\n",
    "    inner_join(study_metadata, by = \"trb_repertoire_id\") |>\n",
    "    filter(cohort %in% c(\"Hippos\", \"BL - Ghana\", \"BL - Uganda\") & \n",
    "        tissue_type %in% c(\"NAT\", \"Tumor\") & \n",
    "        (visit_code == \"V01\" | is.na(visit_code)) & !is.na(trb_repertoire_id)) |>\n",
    "    mutate(cohort = case_when(\n",
    "        cohort == \"Hippos\"  ~  str_c(phenotype, tissue_type, sep = \" - \"),\n",
    "        cohort != \"Hippos\" ~ cohort),\n",
    "        cohort = factor(cohort,  \n",
    "            levels = c(\"Endemic KS - NAT\", \"Endemic KS - Tumor\",\n",
    "                \"Epidemic KS - NAT\", \"Epidemic KS - Tumor\", \n",
    "                \"BL - Ghana\", \"BL - Uganda\")),\n",
    "        alpha = str_replace(alpha, \"Inf\", \"\\u221e\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>ptid_count</th><th scope=col>repertoire_count</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Endemic KS - NAT   </td><td>23</td><td> 23</td></tr>\n",
       "\t<tr><td>Endemic KS - Tumor </td><td>21</td><td> 45</td></tr>\n",
       "\t<tr><td>Epidemic KS - NAT  </td><td>34</td><td> 34</td></tr>\n",
       "\t<tr><td>Epidemic KS - Tumor</td><td>70</td><td>109</td></tr>\n",
       "\t<tr><td>BL - Ghana         </td><td>37</td><td> 37</td></tr>\n",
       "\t<tr><td>BL - Uganda        </td><td>12</td><td> 12</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 3\n",
       "\\begin{tabular}{lll}\n",
       " cohort & ptid\\_count & repertoire\\_count\\\\\n",
       " <fct> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t Endemic KS - NAT    & 23 &  23\\\\\n",
       "\t Endemic KS - Tumor  & 21 &  45\\\\\n",
       "\t Epidemic KS - NAT   & 34 &  34\\\\\n",
       "\t Epidemic KS - Tumor & 70 & 109\\\\\n",
       "\t BL - Ghana          & 37 &  37\\\\\n",
       "\t BL - Uganda         & 12 &  12\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 3\n",
       "\n",
       "| cohort &lt;fct&gt; | ptid_count &lt;int&gt; | repertoire_count &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| Endemic KS - NAT    | 23 |  23 |\n",
       "| Endemic KS - Tumor  | 21 |  45 |\n",
       "| Epidemic KS - NAT   | 34 |  34 |\n",
       "| Epidemic KS - Tumor | 70 | 109 |\n",
       "| BL - Ghana          | 37 |  37 |\n",
       "| BL - Uganda         | 12 |  12 |\n",
       "\n"
      ],
      "text/plain": [
       "  cohort              ptid_count repertoire_count\n",
       "1 Endemic KS - NAT    23          23             \n",
       "2 Endemic KS - Tumor  21          45             \n",
       "3 Epidemic KS - NAT   34          34             \n",
       "4 Epidemic KS - Tumor 70         109             \n",
       "5 BL - Ghana          37          37             \n",
       "6 BL - Uganda         12          12             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Summary of all ptids and repertoires with tumor and normal data from the three studies\n",
    "options(repr.matrix.max.rows=7)\n",
    "til_normal_renyi |>\n",
    "  group_by(cohort) |>\n",
    "  summarize(ptid_count = length(unique(patient_id)),\n",
    "    repertoire_count = length(unique(trb_repertoire_id)))\n",
    "options(repr.matrix.max.rows=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "comparisons <- list(c(\"Endemic KS - Tumor\", \"Epidemic KS - Tumor\"),\n",
    "   c(\"Epidemic KS - NAT\", \"Epidemic KS - Tumor\"), \n",
    "   c(\"Endemic KS - Tumor\", \"BL - Ghana\"),\n",
    "  c(\"Epidemic KS - Tumor\", \"BL - Ghana\"),\n",
    "  c(\"Epidemic KS - Tumor\", \"BL - Uganda\"))\n",
    "comparisons_clin <- list(c(\"Endemic KS\", \"Epidemic KS\"))\n",
    "cohort_palette <- c(\"Epidemic KS - NAT\" = \"#fb8072\", \n",
    "  \"Epidemic KS - Tumor\" = \"#e41a1c\", \"Endemic KS - NAT\" = \"#80b1d3\",\n",
    "  \"Endemic KS - Tumor\" = \"#377eb8\", \"BL - Uganda\" = \"#a65628\",\n",
    "  \"BL - Ghana\" = \"#fdb462\")\n",
    "\n",
    "alpha_zero <- til_normal_renyi |>\n",
    "  filter(alpha == \"0\") |>\n",
    "  ggboxplot(x = \"cohort\", y = \"renyi_number\", color = \"cohort\") |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Species richness\\n(\\u03b1 = 0)\", legend.title = \"Cohort\",\n",
    "    ggtheme = theme_classic(base_size = 24),\n",
    "    legend = \"none\", palette = cohort_palette) +\n",
    "  rremove(\"x.text\") +\n",
    "  stat_compare_means(comparisons = list(c(\"Endemic KS - Tumor\", \"Epidemic KS - Tumor\"),\n",
    "   c(\"Endemic KS - Tumor\", \"BL - Ghana\"),\n",
    "  c(\"Epidemic KS - Tumor\", \"BL - Ghana\"),\n",
    "  c(\"Epidemic KS - Tumor\", \"BL - Uganda\")), \n",
    "  hide.ns = TRUE, size = 5, label = \"p.signif\", show.legend = T)\n",
    "alpha_one <- til_normal_renyi |>\n",
    "  filter(alpha == \"1\") |>\n",
    "  ggboxplot(x = \"cohort\", y = \"renyi_number\", color = \"cohort\") |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Shannon entropy\\n(\\u03b1 = 1)\", legend.title = \"Cohort\",\n",
    "    ggtheme = theme_classic(base_size = 24),\n",
    "    legend = \"none\", palette = cohort_palette) + \n",
    "  rremove(\"x.text\") +\n",
    "  stat_compare_means(comparisons = list(c(\"Endemic KS - Tumor\", \"Epidemic KS - Tumor\"),\n",
    "   c(\"Endemic KS - Tumor\", \"BL - Ghana\"),\n",
    "  c(\"Epidemic KS - Tumor\", \"BL - Uganda\")), hide.ns = TRUE, size = 5, label = \"p.signif\", show.legend = T)\n",
    "alpha_two <- til_normal_renyi |>\n",
    "  filter(alpha == \"2\") |>\n",
    "  ggboxplot(x = \"cohort\", y = \"renyi_number\", color = \"cohort\") |>\n",
    "  ggpar(xlab = \"Cohort\", ylab = \"Simpson's diversity\\n(\\u03b1 = 2)\", legend.title = \"Cohort\",\n",
    "    ggtheme = theme_classic(base_size = 24), \n",
    "    legend = \"none\", palette = cohort_palette, x.text.angle = 45) +\n",
    "  stat_compare_means(comparisons = list(c(\"Endemic KS - Tumor\", \"Epidemic KS - Tumor\"),\n",
    "   c(\"Epidemic KS - NAT\", \"Epidemic KS - Tumor\"), \n",
    "   c(\"Endemic KS - Tumor\", \"BL - Ghana\"),\n",
    "  c(\"Epidemic KS - Tumor\", \"BL - Uganda\")), \n",
    "  hide.ns = TRUE, size = 5, label = \"p.signif\", show.legend = T)\n",
    "alpha_inf <- til_normal_renyi |>\n",
    "  filter(alpha == \"\\u221e\") |>\n",
    "  ggboxplot(x = \"cohort\", y = \"renyi_number\", color = \"cohort\") |>\n",
    "  ggpar(xlab = \"Cohort\", ylab = \"Berger-Parker index\\n(\\u03b1 = \\u221e)\", legend.title = \"Cohort\",\n",
    "    ggtheme = theme_classic(base_size = 24), \n",
    "    legend = \"right\", palette = cohort_palette, x.text.angle = 45) +\n",
    "  stat_compare_means(comparisons = list(\n",
    "   c(\"Epidemic KS - NAT\", \"Epidemic KS - Tumor\"), \n",
    "  c(\"Epidemic KS - Tumor\", \"BL - Uganda\")), \n",
    "  hide.ns = TRUE, size = 5, label = \"p.signif\", show.legend = T)\n",
    "renyi_plot <- ggline(til_normal_renyi, x = \"alpha\", y = \"renyi_number\",\n",
    "    color = \"cohort\", add = \"mean_se\") |> \n",
    "  ggpar(xlab = \"\\u03b1\", ylab = \"Renyi entropy\", legend.title = \"Cohort\",\n",
    "    ggtheme = theme_classic(base_size = 24), \n",
    "    legend = \"none\", palette = cohort_palette)\n",
    "figure_two_layout <- \"\n",
    "1122\n",
    "1122\n",
    "3344\n",
    "3344\n",
    "5555\n",
    "5555\"\n",
    "figure_two <- alpha_zero + alpha_one + alpha_two + \n",
    "  alpha_inf + renyi_plot + \n",
    "  plot_layout(design = figure_two_layout, guides = \"collect\") +\n",
    "  plot_annotation(tag_levels = 'A') & theme(text = element_text('NimbusSan'))\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Figure_two.pdf\", sep = \"/\"), figure_two, width = 16, \n",
    "    height = 18, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Figure_two.svg\", sep = \"/\"), figure_two, width = 16, \n",
    "    height = 18)\n",
    "ggsave(str_c(figures_path, \"png\", \"Figure_two.png\", sep = \"/\"), figure_two, width = 16, \n",
    "    height = 18)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary Figure 4:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "plot_titer_tables <- study_metadata |>\n",
    "  filter(trb_repertoire_id %in% til_normal_renyi$repertoire_id) |>\n",
    "  dplyr::select(patient_id, visit_code, hiv_status, ks_status,\n",
    "    HIV, CD4_count, KSHV_plasma) |>\n",
    "  distinct()\n",
    "kshv_titer_plot <- plot_titer_tables |>\n",
    "  filter(!is.na(KSHV_plasma)) |>\n",
    "  mutate(cohort = if_else(hiv_status == \"Positive\", \"Epidemic KS\",\n",
    "    \"Endemic KS\")) |>\n",
    "  ggboxplot(x = \"cohort\", y = \"KSHV_plasma\", color = \"cohort\", \n",
    "    add = \"jitter\") |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Plasma KSHV titer\\n(copies/mL)\", \n",
    "    legend.title = \"Cohort\",\n",
    "    ggtheme = theme_classic(base_size = 24),\n",
    "    legend = \"none\", palette = c(\"Epidemic KS\" = \"#e41a1c\",\n",
    "      \"Endemic KS\" = \"#377eb8\")) +\n",
    "  stat_compare_means(comparisons = comparisons_clin, hide.ns = TRUE, size = 5, label = \"p.signif\", show.legend = T)\n",
    "cd4_plot <- plot_titer_tables |>\n",
    "  filter(!is.na(CD4_count)) |>\n",
    "  mutate(cohort = if_else(hiv_status == \"Positive\", \"Epidemic KS\",\n",
    "    \"Endemic KS\")) |>\n",
    "  ggboxplot(x = \"cohort\", y = \"CD4_count\", color = \"cohort\", \n",
    "    add = \"jitter\") |>\n",
    "  ggpar(xlab = FALSE, ylab = str_c(\"CD4+ count\\n(cells/\\u03BC\", \"L)\", sep = \"\"), \n",
    "    legend.title = \"Cohort\",\n",
    "    ggtheme = theme_classic(base_size = 24),\n",
    "    legend = \"none\", palette = c(\"Epidemic KS\" = \"#e41a1c\",\n",
    "      \"Endemic KS\" = \"#377eb8\")) +\n",
    "  geom_hline(yintercept = 200, linetype = 2) +\n",
    "  stat_compare_means(comparisons = comparisons_clin, hide.ns = TRUE, size = 5, label = \"p.signif\", show.legend = T)\n",
    "supp_figure_four_layout <- \"\n",
    "1122\n",
    "1122\"\n",
    "supplementary_figure_four <- kshv_titer_plot + cd4_plot +\n",
    "  plot_layout(design = supp_figure_four_layout, guides = \"collect\") +\n",
    "  plot_annotation(tag_levels = 'A') & theme(text = element_text('NimbusSan'))\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Supplementarty_figure_four.pdf\", sep = \"/\"), \n",
    "    supplementary_figure_four, width = 16, \n",
    "    height = 6, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Supplementarty_figure_four.svg\", sep = \"/\"), \n",
    "    supplementary_figure_four, width = 16, \n",
    "    height = 6)\n",
    "ggsave(str_c(figures_path, \"png\", \"Supplementarty_figure_four.png\", sep = \"/\"), \n",
    "    supplementary_figure_four, width = 16, \n",
    "    height = 6)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary Figure 5:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 174 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>HIV\n",
       "VL</th><th scope=col>CD4+\n",
       "count</th><th scope=col>KSHV\n",
       "Plasma\n",
       "VL</th><th scope=col>0</th><th scope=col>0.25</th><th scope=col>0.5</th><th scope=col>1</th><th scope=col>2</th><th scope=col>4</th><th scope=col>8</th><th scope=col>16</th><th scope=col>32</th><th scope=col>64</th><th scope=col>∞</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>008_001_B</th><td>276908</td><td>127</td><td>    0.00</td><td>5.936510</td><td>5.798321</td><td>5.634872</td><td>5.244745</td><td>4.502215</td><td>3.826136</td><td>3.438104</td><td>3.240889</td><td>3.143218</td><td>3.094753</td><td>3.046694</td></tr>\n",
       "\t<tr><th scope=row>008_001_D</th><td>276908</td><td>127</td><td>    0.00</td><td>6.002887</td><td>5.869040</td><td>5.709425</td><td>5.323547</td><td>4.581217</td><td>3.931988</td><td>3.589271</td><td>3.413928</td><td>3.321192</td><td>3.272770</td><td>3.222954</td></tr>\n",
       "\t<tr><th scope=row>008_002_B</th><td>  4658</td><td>418</td><td>10805.01</td><td>6.185585</td><td>6.086141</td><td>5.967807</td><td>5.674548</td><td>5.020768</td><td>4.293126</td><td>3.846466</td><td>3.617506</td><td>3.505721</td><td>3.451238</td><td>3.397933</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>008_262_D</th><td>100493</td><td>342</td><td>50.17757</td><td>6.463758</td><td>6.394088</td><td>6.305485</td><td>6.051294</td><td>5.263333</td><td>4.246675</td><td>3.707140</td><td>3.463881</td><td>3.352327</td><td>3.299121</td><td>3.247573</td></tr>\n",
       "\t<tr><th scope=row>008_263_D</th><td>   313</td><td>231</td><td>19.85711</td><td>6.124420</td><td>6.017465</td><td>5.891833</td><td>5.590408</td><td>4.973164</td><td>4.351676</td><td>3.987935</td><td>3.792147</td><td>3.688173</td><td>3.634169</td><td>3.579372</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 174 × 14\n",
       "\\begin{tabular}{r|llllllllllllll}\n",
       "  & HIV\n",
       "VL & CD4+\n",
       "count & KSHV\n",
       "Plasma\n",
       "VL & 0 & 0.25 & 0.5 & 1 & 2 & 4 & 8 & 16 & 32 & 64 & ∞\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t008\\_001\\_B & 276908 & 127 &     0.00 & 5.936510 & 5.798321 & 5.634872 & 5.244745 & 4.502215 & 3.826136 & 3.438104 & 3.240889 & 3.143218 & 3.094753 & 3.046694\\\\\n",
       "\t008\\_001\\_D & 276908 & 127 &     0.00 & 6.002887 & 5.869040 & 5.709425 & 5.323547 & 4.581217 & 3.931988 & 3.589271 & 3.413928 & 3.321192 & 3.272770 & 3.222954\\\\\n",
       "\t008\\_002\\_B &   4658 & 418 & 10805.01 & 6.185585 & 6.086141 & 5.967807 & 5.674548 & 5.020768 & 4.293126 & 3.846466 & 3.617506 & 3.505721 & 3.451238 & 3.397933\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t008\\_262\\_D & 100493 & 342 & 50.17757 & 6.463758 & 6.394088 & 6.305485 & 6.051294 & 5.263333 & 4.246675 & 3.707140 & 3.463881 & 3.352327 & 3.299121 & 3.247573\\\\\n",
       "\t008\\_263\\_D &    313 & 231 & 19.85711 & 6.124420 & 6.017465 & 5.891833 & 5.590408 & 4.973164 & 4.351676 & 3.987935 & 3.792147 & 3.688173 & 3.634169 & 3.579372\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 174 × 14\n",
       "\n",
       "| <!--/--> | HIV\n",
       "VL &lt;dbl&gt; | CD4+\n",
       "count &lt;dbl&gt; | KSHV\n",
       "Plasma\n",
       "VL &lt;dbl&gt; | 0 &lt;dbl&gt; | 0.25 &lt;dbl&gt; | 0.5 &lt;dbl&gt; | 1 &lt;dbl&gt; | 2 &lt;dbl&gt; | 4 &lt;dbl&gt; | 8 &lt;dbl&gt; | 16 &lt;dbl&gt; | 32 &lt;dbl&gt; | 64 &lt;dbl&gt; | ∞ &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 008_001_B | 276908 | 127 |     0.00 | 5.936510 | 5.798321 | 5.634872 | 5.244745 | 4.502215 | 3.826136 | 3.438104 | 3.240889 | 3.143218 | 3.094753 | 3.046694 |\n",
       "| 008_001_D | 276908 | 127 |     0.00 | 6.002887 | 5.869040 | 5.709425 | 5.323547 | 4.581217 | 3.931988 | 3.589271 | 3.413928 | 3.321192 | 3.272770 | 3.222954 |\n",
       "| 008_002_B |   4658 | 418 | 10805.01 | 6.185585 | 6.086141 | 5.967807 | 5.674548 | 5.020768 | 4.293126 | 3.846466 | 3.617506 | 3.505721 | 3.451238 | 3.397933 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_262_D | 100493 | 342 | 50.17757 | 6.463758 | 6.394088 | 6.305485 | 6.051294 | 5.263333 | 4.246675 | 3.707140 | 3.463881 | 3.352327 | 3.299121 | 3.247573 |\n",
       "| 008_263_D |    313 | 231 | 19.85711 | 6.124420 | 6.017465 | 5.891833 | 5.590408 | 4.973164 | 4.351676 | 3.987935 | 3.792147 | 3.688173 | 3.634169 | 3.579372 |\n",
       "\n"
      ],
      "text/plain": [
       "          HIV\\nVL CD4+\\ncount KSHV\\nPlasma\\nVL 0        0.25     0.5     \n",
       "008_001_B 276908  127             0.00         5.936510 5.798321 5.634872\n",
       "008_001_D 276908  127             0.00         6.002887 5.869040 5.709425\n",
       "008_002_B   4658  418         10805.01         6.185585 6.086141 5.967807\n",
       "⋮         ⋮       ⋮           ⋮                ⋮        ⋮        ⋮       \n",
       "008_262_D 100493  342         50.17757         6.463758 6.394088 6.305485\n",
       "008_263_D    313  231         19.85711         6.124420 6.017465 5.891833\n",
       "          1        2        4        8        16       32       64      \n",
       "008_001_B 5.244745 4.502215 3.826136 3.438104 3.240889 3.143218 3.094753\n",
       "008_001_D 5.323547 4.581217 3.931988 3.589271 3.413928 3.321192 3.272770\n",
       "008_002_B 5.674548 5.020768 4.293126 3.846466 3.617506 3.505721 3.451238\n",
       "⋮         ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮       \n",
       "008_262_D 6.051294 5.263333 4.246675 3.707140 3.463881 3.352327 3.299121\n",
       "008_263_D 5.590408 4.973164 4.351676 3.987935 3.792147 3.688173 3.634169\n",
       "          ∞       \n",
       "008_001_B 3.046694\n",
       "008_001_D 3.222954\n",
       "008_002_B 3.397933\n",
       "⋮         ⋮       \n",
       "008_262_D 3.247573\n",
       "008_263_D 3.579372"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "correlation_data_frame <- study_metadata |>\n",
    "    filter(cohort == \"Hippos\" & tissue_type %in% c(\"NAT\", \"Tumor\") & visit_code == \"V01\") |>\n",
    "    filter(!is.na(HIV) & !is.na(CD4_count) & !is.na(KSHV_plasma)) |>\n",
    "    dplyr::select(patient_id, repertoire_id, HIV, CD4_count, KSHV_plasma, phenotype) |> \n",
    "    inner_join(study_renyi_table, by = \"repertoire_id\") |>\n",
    "    dplyr::select(repertoire_id, HIV:`Inf`, -phenotype)\n",
    "corr_row_names <- correlation_data_frame |>\n",
    "    pull(repertoire_id)\n",
    "correlation_data_frame <- correlation_data_frame |>\n",
    "    dplyr::select(-repertoire_id) |>\n",
    "    as.data.frame()\n",
    "rownames(correlation_data_frame) <- corr_row_names\n",
    "colnames(correlation_data_frame) <- c(\"HIV\\nVL\", \"CD4+\\ncount\", \"KSHV\\nPlasma\\nVL\", \"0\", \"0.25\", \"0.5\", \"1\", \"2\", \"4\", \"8\", \"16\", \"32\", \"64\", \"\\u221e\")\n",
    "\n",
    "correlation_data_frame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 14 × 14 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>HIV\n",
       "VL</th><th scope=col>CD4+\n",
       "count</th><th scope=col>KSHV\n",
       "Plasma\n",
       "VL</th><th scope=col>0</th><th scope=col>0.25</th><th scope=col>0.5</th><th scope=col>1</th><th scope=col>2</th><th scope=col>4</th><th scope=col>8</th><th scope=col>16</th><th scope=col>32</th><th scope=col>64</th><th scope=col>∞</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>HIV\n",
       "VL</th><td> 1.0</td><td>-0.2</td><td>0.1</td><td> 0.0</td><td> 0.0</td><td> 0.0</td><td> 0.0</td><td> 0.0</td><td>0.0</td><td>0.0</td><td>0.0</td><td>0.0</td><td>0.0</td><td>0.0</td></tr>\n",
       "\t<tr><th scope=row>CD4+\n",
       "count</th><td>-0.2</td><td> 1.0</td><td>0.0</td><td>-0.1</td><td>-0.1</td><td>-0.1</td><td>-0.1</td><td>-0.1</td><td>0.0</td><td>0.0</td><td>0.0</td><td>0.0</td><td>0.0</td><td>0.0</td></tr>\n",
       "\t<tr><th scope=row>KSHV\n",
       "Plasma\n",
       "VL</th><td> 0.1</td><td> 0.0</td><td>1.0</td><td> 0.0</td><td> 0.0</td><td> 0.0</td><td> 0.0</td><td> 0.0</td><td>0.1</td><td>0.1</td><td>0.1</td><td>0.1</td><td>0.1</td><td>0.1</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>64</th><td>0</td><td>0</td><td>0.1</td><td>0.5</td><td>0.5</td><td>0.6</td><td>0.7</td><td>0.9</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>∞</th><td>0</td><td>0</td><td>0.1</td><td>0.5</td><td>0.5</td><td>0.6</td><td>0.7</td><td>0.9</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 14 × 14 of type dbl\n",
       "\\begin{tabular}{r|llllllllllllll}\n",
       "  & HIV\n",
       "VL & CD4+\n",
       "count & KSHV\n",
       "Plasma\n",
       "VL & 0 & 0.25 & 0.5 & 1 & 2 & 4 & 8 & 16 & 32 & 64 & ∞\\\\\n",
       "\\hline\n",
       "\tHIV\n",
       "VL &  1.0 & -0.2 & 0.1 &  0.0 &  0.0 &  0.0 &  0.0 &  0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 0.0\\\\\n",
       "\tCD4+\n",
       "count & -0.2 &  1.0 & 0.0 & -0.1 & -0.1 & -0.1 & -0.1 & -0.1 & 0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 0.0\\\\\n",
       "\tKSHV\n",
       "Plasma\n",
       "VL &  0.1 &  0.0 & 1.0 &  0.0 &  0.0 &  0.0 &  0.0 &  0.0 & 0.1 & 0.1 & 0.1 & 0.1 & 0.1 & 0.1\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t64 & 0 & 0 & 0.1 & 0.5 & 0.5 & 0.6 & 0.7 & 0.9 & 1 & 1 & 1 & 1 & 1 & 1\\\\\n",
       "\t∞ & 0 & 0 & 0.1 & 0.5 & 0.5 & 0.6 & 0.7 & 0.9 & 1 & 1 & 1 & 1 & 1 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 14 × 14 of type dbl\n",
       "\n",
       "| <!--/--> | HIV\n",
       "VL | CD4+\n",
       "count | KSHV\n",
       "Plasma\n",
       "VL | 0 | 0.25 | 0.5 | 1 | 2 | 4 | 8 | 16 | 32 | 64 | ∞ |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HIV\n",
       "VL |  1.0 | -0.2 | 0.1 |  0.0 |  0.0 |  0.0 |  0.0 |  0.0 | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 |\n",
       "| CD4+\n",
       "count | -0.2 |  1.0 | 0.0 | -0.1 | -0.1 | -0.1 | -0.1 | -0.1 | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 |\n",
       "| KSHV\n",
       "Plasma\n",
       "VL |  0.1 |  0.0 | 1.0 |  0.0 |  0.0 |  0.0 |  0.0 |  0.0 | 0.1 | 0.1 | 0.1 | 0.1 | 0.1 | 0.1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 64 | 0 | 0 | 0.1 | 0.5 | 0.5 | 0.6 | 0.7 | 0.9 | 1 | 1 | 1 | 1 | 1 | 1 |\n",
       "| ∞ | 0 | 0 | 0.1 | 0.5 | 0.5 | 0.6 | 0.7 | 0.9 | 1 | 1 | 1 | 1 | 1 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "                 HIV\\nVL CD4+\\ncount KSHV\\nPlasma\\nVL 0    0.25 0.5  1    2   \n",
       "HIV\\nVL           1.0    -0.2        0.1               0.0  0.0  0.0  0.0  0.0\n",
       "CD4+\\ncount      -0.2     1.0        0.0              -0.1 -0.1 -0.1 -0.1 -0.1\n",
       "KSHV\\nPlasma\\nVL  0.1     0.0        1.0               0.0  0.0  0.0  0.0  0.0\n",
       "⋮                ⋮       ⋮           ⋮                ⋮    ⋮    ⋮    ⋮    ⋮   \n",
       "64               0       0           0.1              0.5  0.5  0.6  0.7  0.9 \n",
       "∞                0       0           0.1              0.5  0.5  0.6  0.7  0.9 \n",
       "                 4   8   16  32  64  ∞  \n",
       "HIV\\nVL          0.0 0.0 0.0 0.0 0.0 0.0\n",
       "CD4+\\ncount      0.0 0.0 0.0 0.0 0.0 0.0\n",
       "KSHV\\nPlasma\\nVL 0.1 0.1 0.1 0.1 0.1 0.1\n",
       "⋮                ⋮   ⋮   ⋮   ⋮   ⋮   ⋮  \n",
       "64               1   1   1   1   1   1  \n",
       "∞                1   1   1   1   1   1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "correlation_matrix <- round(cor(correlation_data_frame), 1)\n",
    "correlation_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "supplementary_figure_five <- ggcorrplot(correlation_matrix,  outline.col = \"white\",  lab = TRUE) +\n",
    "  theme_classic() +\n",
    "  theme(axis.text = element_text(size = 12, face = \"bold\"), axis.title = element_blank(),\n",
    "    legend.key.size = unit(1, \"cm\"), text = element_text(family = \"NimbusSan\"))\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Supplementary_figure_five.pdf\", sep = \"/\"), supplementary_figure_five, \n",
    "  limitsize = FALSE, width = 14, height = 10, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Supplementary_figure_five.svg\", sep = \"/\"), supplementary_figure_five,\n",
    "  limitsize = FALSE, width = 14, height = 10, device = \"svg\")\n",
    "ggsave(str_c(figures_path, \"png\", \"Supplementary_figure_five.png\", sep = \"/\"), supplementary_figure_five,\n",
    "  limitsize = FALSE, width = 14, height = 10, device = \"png\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary analysis 1:\n",
    "Calculate the biserial correlation between renyi metrics "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.16597248923765"
      ],
      "text/latex": [
       "0.16597248923765"
      ],
      "text/markdown": [
       "0.16597248923765"
      ],
      "text/plain": [
       "[1] 0.1659725"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.130131307331331"
      ],
      "text/latex": [
       "0.130131307331331"
      ],
      "text/markdown": [
       "0.130131307331331"
      ],
      "text/plain": [
       "[1] 0.1301313"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.120131520980057"
      ],
      "text/latex": [
       "0.120131520980057"
      ],
      "text/markdown": [
       "0.120131520980057"
      ],
      "text/plain": [
       "[1] 0.1201315"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.0940116663710002"
      ],
      "text/latex": [
       "0.0940116663710002"
      ],
      "text/markdown": [
       "0.0940116663710002"
      ],
      "text/plain": [
       "[1] 0.09401167"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clin_correlation_data_frame <- study_metadata |>\n",
    "    filter(cohort == \"Hippos\" & tissue_type %in% c(\"NAT\", \"Tumor\") & visit_code == \"V01\") |>\n",
    "    dplyr::select(patient_id, repertoire_id, phenotype, response) |> \n",
    "    inner_join(study_renyi_table, by = \"repertoire_id\") |>\n",
    "    filter(!is.na(response)) |>\n",
    "    mutate(response = if_else(response == \"CR\" | response == \"PR\", 1, 0)) \n",
    "# Endemic KS\n",
    "biserial.cor(clin_correlation_data_frame |> filter(phenotype == \"Endemic KS\") |> pull(`0`), clin_correlation_data_frame |> filter(phenotype == \"Endemic KS\") |> pull(response), use = c(\"all.obs\"), level = 2)\n",
    "biserial.cor(clin_correlation_data_frame |> filter(phenotype == \"Endemic KS\") |> pull(`1`), clin_correlation_data_frame |> filter(phenotype == \"Endemic KS\") |> pull(response), use = c(\"all.obs\"), level = 2)\n",
    "biserial.cor(clin_correlation_data_frame |> filter(phenotype == \"Endemic KS\") |> pull(`2`), clin_correlation_data_frame |> filter(phenotype == \"Endemic KS\") |> pull(response), use = c(\"all.obs\"), level = 2)\n",
    "biserial.cor(clin_correlation_data_frame |> filter(phenotype == \"Endemic KS\") |> pull(`Inf`), clin_correlation_data_frame |> filter(phenotype == \"Endemic KS\") |> pull(response), use = c(\"all.obs\"), level = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.163749316191277"
      ],
      "text/latex": [
       "0.163749316191277"
      ],
      "text/markdown": [
       "0.163749316191277"
      ],
      "text/plain": [
       "[1] 0.1637493"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.160294360640217"
      ],
      "text/latex": [
       "0.160294360640217"
      ],
      "text/markdown": [
       "0.160294360640217"
      ],
      "text/plain": [
       "[1] 0.1602944"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.149130104816308"
      ],
      "text/latex": [
       "0.149130104816308"
      ],
      "text/markdown": [
       "0.149130104816308"
      ],
      "text/plain": [
       "[1] 0.1491301"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.164872059049478"
      ],
      "text/latex": [
       "0.164872059049478"
      ],
      "text/markdown": [
       "0.164872059049478"
      ],
      "text/plain": [
       "[1] 0.1648721"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Epidemic KS\n",
    "biserial.cor(clin_correlation_data_frame |> filter(phenotype == \"Epidemic KS\") |> pull(`0`), clin_correlation_data_frame |> filter(phenotype == \"Epidemic KS\") |> pull(response), use = c(\"all.obs\"), level = 2)\n",
    "biserial.cor(clin_correlation_data_frame |> filter(phenotype == \"Epidemic KS\") |> pull(`1`), clin_correlation_data_frame |> filter(phenotype == \"Epidemic KS\") |> pull(response), use = c(\"all.obs\"), level = 2)\n",
    "biserial.cor(clin_correlation_data_frame |> filter(phenotype == \"Epidemic KS\") |> pull(`2`), clin_correlation_data_frame |> filter(phenotype == \"Epidemic KS\") |> pull(response), use = c(\"all.obs\"), level = 2)\n",
    "biserial.cor(clin_correlation_data_frame |> filter(phenotype == \"Epidemic KS\") |> pull(`Inf`), clin_correlation_data_frame |> filter(phenotype == \"Epidemic KS\") |> pull(response), use = c(\"all.obs\"), level = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.241866449379336"
      ],
      "text/latex": [
       "0.241866449379336"
      ],
      "text/markdown": [
       "0.241866449379336"
      ],
      "text/plain": [
       "[1] 0.2418664"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.22980048141952"
      ],
      "text/latex": [
       "0.22980048141952"
      ],
      "text/markdown": [
       "0.22980048141952"
      ],
      "text/plain": [
       "[1] 0.2298005"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.198949221100268"
      ],
      "text/latex": [
       "0.198949221100268"
      ],
      "text/markdown": [
       "0.198949221100268"
      ],
      "text/plain": [
       "[1] 0.1989492"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.180414401877767"
      ],
      "text/latex": [
       "0.180414401877767"
      ],
      "text/markdown": [
       "0.180414401877767"
      ],
      "text/plain": [
       "[1] 0.1804144"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Epidemic and Endemic KS\n",
    "biserial.cor(clin_correlation_data_frame$`0`, clin_correlation_data_frame$response, use = c(\"all.obs\"), level = 2)\n",
    "biserial.cor(clin_correlation_data_frame$`1`, clin_correlation_data_frame$response, use = c(\"all.obs\"), level = 2)\n",
    "biserial.cor(clin_correlation_data_frame$`2`, clin_correlation_data_frame$response, use = c(\"all.obs\"), level = 2)\n",
    "biserial.cor(clin_correlation_data_frame$`Inf`, clin_correlation_data_frame$response, use = c(\"all.obs\"), level = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 3: TIL repertoires of KS tumors are largely private to each individual"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Comparison of seqeunce sharing at a individual TCR level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Subsetting productive sequences [=>----------]  12% eta:  2s\n",
      "\n",
      "Subsetting productive sequences [=>----------]  19% eta: 13s\n",
      "\n",
      "Subsetting productive sequences [==>---------]  25% eta: 10s\n",
      "\n",
      "Subsetting productive sequences [===>--------]  31% eta:  8s\n",
      "\n",
      "Subsetting productive sequences [===>--------]  38% eta:  7s\n",
      "\n",
      "Subsetting productive sequences [====>-------]  44% eta:  9s\n",
      "\n",
      "Subsetting productive sequences [=====>------]  50% eta:  7s\n",
      "\n",
      "Subsetting productive sequences [======>-----]  56% eta:  6s\n",
      "\n",
      "Subsetting productive sequences [=======>----]  62% eta:  5s\n",
      "\n",
      "Subsetting productive sequences [=======>----]  69% eta:  5s\n",
      "\n",
      "Subsetting productive sequences [========>---]  75% eta:  4s\n",
      "\n",
      "Subsetting productive sequences [=========>--]  81% eta:  3s\n",
      "\n",
      "Subsetting productive sequences [=========>--]  88% eta:  2s\n",
      "\n",
      "Subsetting productive sequences [==========>-]  94% eta:  1s\n",
      "\n",
      "Subsetting productive sequences [============] 100% eta:  0s\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "rep_dict <- study_metadata |>\n",
    "  filter(!is.na(trb_repertoire_id)) |>\n",
    "  pull(repertoire_id)\n",
    "names(rep_dict) <- study_metadata |>\n",
    "  filter(!is.na(trb_repertoire_id)) |>\n",
    "  pull(trb_repertoire_id)\n",
    "ptid_dict <- study_metadata |>\n",
    "  filter(!is.na(patient_id)) |>\n",
    "  pull(phenotype)\n",
    "names(ptid_dict) <- study_metadata |>\n",
    "  filter(!is.na(patient_id)) |>\n",
    "  pull(patient_id)\n",
    "ptid_dict[\"008_265\"] <- \"Epidemic KS\"\n",
    "compairr_table <- read_tsv(comp_path, show_col_types = FALSE) |>\n",
    "  dplyr::rename(rep_one = `#`) \n",
    "comp_names <- compairr_table |>\n",
    "  pull(rep_one)\n",
    "compairr_matrix <- compairr_table |>\n",
    "  dplyr::select(-rep_one) |>\n",
    "  as.matrix() \n",
    "rownames(compairr_matrix) <- comp_names\n",
    "compairr_matrix[lower.tri(compairr_matrix, diag = FALSE)] <- NA \n",
    "compairr_table <- compairr_matrix |>\n",
    "  as_tibble(rownames = \"rep_one\") |>\n",
    "  pivot_longer(-rep_one, names_to = \"rep_two\", values_to = \"overlap\") |>\n",
    "  filter(str_detect(rep_one, \"008_\\\\d+\") & str_detect(rep_two, \"008_\\\\d+\") & \n",
    "    !is.na(overlap) & \n",
    "    str_extract(rep_one, \"[A-Z]+$\") %in% c(\"A\",\"B\", \"C\", \"D\") &\n",
    "    str_extract(rep_two, \"[A-Z]+$\") %in% c(\"A\",\"B\", \"C\", \"D\")) |>\n",
    "  mutate(rep_one = rep_dict[rep_one], rep_two = rep_dict[rep_two],\n",
    "    sample_group = if_else(str_extract(rep_one, \"008_\\\\d+\") ==\n",
    "      str_extract(rep_two, \"008_\\\\d+\"), \"Intra-subject\", \"Inter-subject\"),\n",
    "    tumor_group = case_when(\n",
    "      str_extract(rep_one, \"[A-Z]+$\") == \"A\" &\n",
    "        str_extract(rep_two, \"[A-Z]+$\") == \"A\" ~ \"NAT - NAT\",\n",
    "      str_extract(rep_one, \"[A-Z]+$\") == \"A\" &\n",
    "        str_extract(rep_two, \"[A-Z]+$\") %in% c(\"B\", \"C\", \"D\") ~ \"NAT - Tumor\",\n",
    "      str_extract(rep_two, \"[A-Z]+$\") == \"A\" &\n",
    "        str_extract(rep_one, \"[A-Z]+$\") %in% c(\"B\", \"C\", \"D\") ~ \"NAT - Tumor\",\n",
    "      str_extract(rep_two, \"[A-Z]+$\") %in% c(\"B\", \"C\", \"D\") &\n",
    "        str_extract(rep_one, \"[A-Z]+$\") %in% c(\"B\", \"C\", \"D\") ~ \"Tumor - Tumor\"),\n",
    "    ks_group = str_c(ptid_dict[str_extract(rep_one, \"008_\\\\d+\")],\n",
    "      ptid_dict[str_extract(rep_two, \"008_\\\\d+\")], sep = \" - \"),\n",
    "    ks_group = str_replace(ks_group, \"Epidemic KS - Endemic KS\",\n",
    "      \"Endemic KS - Epidemic KS\")) |>\n",
    "  filter(!(sample_group == \"Inter-subject\" & tumor_group == \"NAT - Tumor\") &\n",
    "    !(rep_one == rep_two))\n",
    "\n",
    "alluvial_data <- study_annotated_nprod_table |>\n",
    "  filter(patient_id %in% c(\"008_057\", \"008_061\", \"008_021\", \"008_048\") & \n",
    "    visit_code == \"V01\") |>\n",
    "  productiveSeq()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Understanding the prevelance of TCRs with known antigenic specificity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generate database of TCRs with known antigenic specificity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1711 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>epitope</th><th scope=col>pathology</th><th scope=col>antigen</th><th scope=col>mhc_allele</th><th scope=col>trb_cdr3_aa</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GLCTLVAML</td><td>EBV        </td><td>BMLF1  </td><td>A*02   </td><td>CASSEGRISPGELFF  </td></tr>\n",
       "\t<tr><td>VLEETSVML</td><td>CMV        </td><td>IE1    </td><td>A*02:01</td><td>CASSPDSQSSGNTIYTF</td></tr>\n",
       "\t<tr><td>VAANIVLTV</td><td>HomoSapiens</td><td>SLC30A8</td><td>A*02   </td><td>CASSSVGVDTQYF    </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>VAANIVLTV</td><td>HomoSapiens</td><td>SLC30A8</td><td>A*02   </td><td>CASSPFLTGSNTEAFF</td></tr>\n",
       "\t<tr><td>RLARLALVL</td><td>HomoSapiens</td><td>5T4    </td><td>A*02:01</td><td>CASSYMGPEAFF    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1711 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " epitope & pathology & antigen & mhc\\_allele & trb\\_cdr3\\_aa\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t GLCTLVAML & EBV         & BMLF1   & A*02    & CASSEGRISPGELFF  \\\\\n",
       "\t VLEETSVML & CMV         & IE1     & A*02:01 & CASSPDSQSSGNTIYTF\\\\\n",
       "\t VAANIVLTV & HomoSapiens & SLC30A8 & A*02    & CASSSVGVDTQYF    \\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t VAANIVLTV & HomoSapiens & SLC30A8 & A*02    & CASSPFLTGSNTEAFF\\\\\n",
       "\t RLARLALVL & HomoSapiens & 5T4     & A*02:01 & CASSYMGPEAFF    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1711 × 5\n",
       "\n",
       "| epitope &lt;chr&gt; | pathology &lt;chr&gt; | antigen &lt;chr&gt; | mhc_allele &lt;chr&gt; | trb_cdr3_aa &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| GLCTLVAML | EBV         | BMLF1   | A*02    | CASSEGRISPGELFF   |\n",
       "| VLEETSVML | CMV         | IE1     | A*02:01 | CASSPDSQSSGNTIYTF |\n",
       "| VAANIVLTV | HomoSapiens | SLC30A8 | A*02    | CASSSVGVDTQYF     |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| VAANIVLTV | HomoSapiens | SLC30A8 | A*02    | CASSPFLTGSNTEAFF |\n",
       "| RLARLALVL | HomoSapiens | 5T4     | A*02:01 | CASSYMGPEAFF     |\n",
       "\n"
      ],
      "text/plain": [
       "     epitope   pathology   antigen mhc_allele trb_cdr3_aa      \n",
       "1    GLCTLVAML EBV         BMLF1   A*02       CASSEGRISPGELFF  \n",
       "2    VLEETSVML CMV         IE1     A*02:01    CASSPDSQSSGNTIYTF\n",
       "3    VAANIVLTV HomoSapiens SLC30A8 A*02       CASSSVGVDTQYF    \n",
       "⋮    ⋮         ⋮           ⋮       ⋮          ⋮                \n",
       "1710 VAANIVLTV HomoSapiens SLC30A8 A*02       CASSPFLTGSNTEAFF \n",
       "1711 RLARLALVL HomoSapiens 5T4     A*02:01    CASSYMGPEAFF     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create a database of VDJdb and McPAS-TCR sequences\n",
    "vdjdb_table <- read_tsv(vdjdb_slim_path, show_col_types = FALSE) |>\n",
    "    dplyr::filter(vdjdb.score > 1 & gene == \"TRB\" & species == \"HomoSapiens\") |>\n",
    "    dplyr::rename(epitope = antigen.epitope, pathology = antigen.species, \n",
    "        mhc_allele = mhc.a, trb_cdr3_aa = cdr3, antigen = antigen.gene) |>\n",
    "    dplyr::select(epitope, pathology, antigen, mhc_allele, trb_cdr3_aa) |>\n",
    "    separate_rows(mhc_allele, sep = \",\") |>\n",
    "    mutate(mhc_allele = str_remove(mhc_allele, \"HLA-\"))\n",
    "vdjdb_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 7894 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>trb_cdr3_aa</th><th scope=col>pathology</th><th scope=col>antigen</th><th scope=col>epitope</th><th scope=col>mhc_allele</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>CASSLGSSYEQYF  </td><td>Other</td><td>gp100 </td><td>YLEPGPVTA</td><td>A*02</td></tr>\n",
       "\t<tr><td>CASSESAGGYYNEQF</td><td>Other</td><td>EBNA-4</td><td>IVTDFSVIK</td><td>A*11</td></tr>\n",
       "\t<tr><td>CASSFGFGSSYGYTF</td><td>Other</td><td>EBNA-4</td><td>IVTDFSVIK</td><td>A*11</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>CASSQGVNTGELFF</td><td>Other</td><td>NA</td><td>FLCMKALLL</td><td>A*02:01</td></tr>\n",
       "\t<tr><td>CSAETGLSYEQYF </td><td>Other</td><td>NA</td><td>FLCMKALLL</td><td>A*02:01</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 7894 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " trb\\_cdr3\\_aa & pathology & antigen & epitope & mhc\\_allele\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t CASSLGSSYEQYF   & Other & gp100  & YLEPGPVTA & A*02\\\\\n",
       "\t CASSESAGGYYNEQF & Other & EBNA-4 & IVTDFSVIK & A*11\\\\\n",
       "\t CASSFGFGSSYGYTF & Other & EBNA-4 & IVTDFSVIK & A*11\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t CASSQGVNTGELFF & Other & NA & FLCMKALLL & A*02:01\\\\\n",
       "\t CSAETGLSYEQYF  & Other & NA & FLCMKALLL & A*02:01\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 7894 × 5\n",
       "\n",
       "| trb_cdr3_aa &lt;chr&gt; | pathology &lt;chr&gt; | antigen &lt;chr&gt; | epitope &lt;chr&gt; | mhc_allele &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| CASSLGSSYEQYF   | Other | gp100  | YLEPGPVTA | A*02 |\n",
       "| CASSESAGGYYNEQF | Other | EBNA-4 | IVTDFSVIK | A*11 |\n",
       "| CASSFGFGSSYGYTF | Other | EBNA-4 | IVTDFSVIK | A*11 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| CASSQGVNTGELFF | Other | NA | FLCMKALLL | A*02:01 |\n",
       "| CSAETGLSYEQYF  | Other | NA | FLCMKALLL | A*02:01 |\n",
       "\n"
      ],
      "text/plain": [
       "     trb_cdr3_aa     pathology antigen epitope   mhc_allele\n",
       "1    CASSLGSSYEQYF   Other     gp100   YLEPGPVTA A*02      \n",
       "2    CASSESAGGYYNEQF Other     EBNA-4  IVTDFSVIK A*11      \n",
       "3    CASSFGFGSSYGYTF Other     EBNA-4  IVTDFSVIK A*11      \n",
       "⋮    ⋮               ⋮         ⋮       ⋮         ⋮         \n",
       "7893 CASSQGVNTGELFF  Other     NA      FLCMKALLL A*02:01   \n",
       "7894 CSAETGLSYEQYF   Other     NA      FLCMKALLL A*02:01   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mcpas_table <- read_csv(mcpas_tcr_path, show_col_types = FALSE) |> \n",
    "    mutate(Pathology = case_when(Pathology == \"Epstein Barr Virus (EBV)\" ~ \"EBV\",\n",
    "            Pathology == \"Cytomegalovirus (CMV)\" ~ \"CMV\",\n",
    "            Pathology == \"COVID-19\" ~ \"SARS-CoV-2\", \n",
    "            Pathology == \"Hepatitis C virus (HCV)\" ~ \"HCV\",\n",
    "            Pathology == \"Human immunodeficiency virus (HIV)\" ~ \"HIV-1\",\n",
    "            Pathology == \"Human immunodeficiency virus (HIV) related\" ~ \"HIV-1\",\n",
    "            Pathology == \"Influenza\" ~ \"InfluenzaA\",\n",
    "            Pathology == \"M.tuberculosis\" ~ \"M.tuberculosis\",\n",
    "            Pathology == \"Herpes simplex virus 2 (HSV2)\" ~ \"HSV-2\",\n",
    "            Pathology == \"HTLV-1\" ~ \"HTLV-1\",\n",
    "            Pathology == \"HTLV-1 (Chronic\" ~ \"HTLV-1\",\n",
    "            Pathology == \"Human herpes virus 1\" ~ \"HHV\",\n",
    "            TRUE ~ \"Other\")) |> \n",
    "    filter(Species == \"Human\" & Antigen.identification.method < 3 & !is.na(Epitope.peptide)) |>\n",
    "    dplyr::select(CDR3.beta.aa, Pathology, Antigen.protein,\n",
    "        Epitope.peptide, MHC) |>\n",
    "    dplyr::rename(epitope = Epitope.peptide, pathology = Pathology, \n",
    "        mhc_allele = MHC, trb_cdr3_aa = CDR3.beta.aa, antigen = Antigen.protein) |>\n",
    "    separate_rows(mhc_allele, sep = \",\") |>\n",
    "    mutate(mhc_allele = str_remove(mhc_allele, \"HLA-\"),\n",
    "        mhc_allele = case_when(mhc_allele == \"A2\" ~ \"A*02\",\n",
    "            mhc_allele == \"Cw*16:01\" ~ \"C*16:01\",\n",
    "            mhc_allele == \"A*011\" ~ \"A*11\",\n",
    "            mhc_allele == \"A*2:01\" ~ \"A*02:01\",\n",
    "            mhc_allele == \"B*8\" ~ \"B*08\",\n",
    "            mhc_allele == \"Cw*16:01\" ~ \"C*16:01\",\n",
    "            mhc_allele == \"A1\" ~ \"A*01\",\n",
    "            mhc_allele == \"B7\" ~ \"B*07\",\n",
    "            mhc_allele == \"DRB1*04-01\" ~ \"DRB1*04:01\",\n",
    "            mhc_allele == \"A2:01\" ~ \"A*02:01\")) |>\n",
    "    filter(!is.na(mhc_allele))\n",
    "mcpas_table  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 9605 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>epitope</th><th scope=col>pathology</th><th scope=col>antigen</th><th scope=col>mhc_allele</th><th scope=col>trb_cdr3_aa</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>GLCTLVAML</td><td>EBV        </td><td>BMLF1  </td><td>A*02   </td><td>CASSEGRISPGELFF  </td></tr>\n",
       "\t<tr><td>VLEETSVML</td><td>CMV        </td><td>IE1    </td><td>A*02:01</td><td>CASSPDSQSSGNTIYTF</td></tr>\n",
       "\t<tr><td>VAANIVLTV</td><td>HomoSapiens</td><td>SLC30A8</td><td>A*02   </td><td>CASSSVGVDTQYF    </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>FLCMKALLL</td><td>Other</td><td>NA</td><td>A*02:01</td><td>CASSQGVNTGELFF</td></tr>\n",
       "\t<tr><td>FLCMKALLL</td><td>Other</td><td>NA</td><td>A*02:01</td><td>CSAETGLSYEQYF </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 9605 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " epitope & pathology & antigen & mhc\\_allele & trb\\_cdr3\\_aa\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t GLCTLVAML & EBV         & BMLF1   & A*02    & CASSEGRISPGELFF  \\\\\n",
       "\t VLEETSVML & CMV         & IE1     & A*02:01 & CASSPDSQSSGNTIYTF\\\\\n",
       "\t VAANIVLTV & HomoSapiens & SLC30A8 & A*02    & CASSSVGVDTQYF    \\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t FLCMKALLL & Other & NA & A*02:01 & CASSQGVNTGELFF\\\\\n",
       "\t FLCMKALLL & Other & NA & A*02:01 & CSAETGLSYEQYF \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 9605 × 5\n",
       "\n",
       "| epitope &lt;chr&gt; | pathology &lt;chr&gt; | antigen &lt;chr&gt; | mhc_allele &lt;chr&gt; | trb_cdr3_aa &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| GLCTLVAML | EBV         | BMLF1   | A*02    | CASSEGRISPGELFF   |\n",
       "| VLEETSVML | CMV         | IE1     | A*02:01 | CASSPDSQSSGNTIYTF |\n",
       "| VAANIVLTV | HomoSapiens | SLC30A8 | A*02    | CASSSVGVDTQYF     |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| FLCMKALLL | Other | NA | A*02:01 | CASSQGVNTGELFF |\n",
       "| FLCMKALLL | Other | NA | A*02:01 | CSAETGLSYEQYF  |\n",
       "\n"
      ],
      "text/plain": [
       "     epitope   pathology   antigen mhc_allele trb_cdr3_aa      \n",
       "1    GLCTLVAML EBV         BMLF1   A*02       CASSEGRISPGELFF  \n",
       "2    VLEETSVML CMV         IE1     A*02:01    CASSPDSQSSGNTIYTF\n",
       "3    VAANIVLTV HomoSapiens SLC30A8 A*02       CASSSVGVDTQYF    \n",
       "⋮    ⋮         ⋮           ⋮       ⋮          ⋮                \n",
       "9604 FLCMKALLL Other       NA      A*02:01    CASSQGVNTGELFF   \n",
       "9605 FLCMKALLL Other       NA      A*02:01    CSAETGLSYEQYF    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "antigen_db <- bind_rows(vdjdb_table, mcpas_table)\n",
    "antigen_db"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Identify all TCRs with known antigenic specificities from KSTME cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2186 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>junction</th><th scope=col>junction_aa</th><th scope=col>pathology</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001</td><td>008_001_D</td><td>TGTGCCACCGGGACAGGGGATAGCAATCAGCCCCAGCATTTT</td><td>CATGTGDSNQPQHF</td><td>EBV</td></tr>\n",
       "\t<tr><td>008_001</td><td>008_001_D</td><td>TGTGCCAGCAGTACAGGGGATAGCAATCAGCCCCAGCATTTT</td><td>CASSTGDSNQPQHF</td><td>EBV</td></tr>\n",
       "\t<tr><td>008_001</td><td>008_001_D</td><td>TGTGCCAGCGGGACAGGTGATAGCAATCAGCCCCAGCATTTT</td><td>CASGTGDSNQPQHF</td><td>EBV</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Unk2</td><td>008_037_B</td><td>TGTGCCAGCAGTTTGAACACTGAAGCTTTCTTT   </td><td>CASSLNTEAFF </td><td>Other</td></tr>\n",
       "\t<tr><td>Unk2</td><td>008_037_B</td><td>TGTGCCAGCAGTTTGGGGAACACTGAAGCTTTCTTT</td><td>CASSLGNTEAFF</td><td>Other</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2186 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " patient\\_id & repertoire\\_id & junction & junction\\_aa & pathology\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_001 & 008\\_001\\_D & TGTGCCACCGGGACAGGGGATAGCAATCAGCCCCAGCATTTT & CATGTGDSNQPQHF & EBV\\\\\n",
       "\t 008\\_001 & 008\\_001\\_D & TGTGCCAGCAGTACAGGGGATAGCAATCAGCCCCAGCATTTT & CASSTGDSNQPQHF & EBV\\\\\n",
       "\t 008\\_001 & 008\\_001\\_D & TGTGCCAGCGGGACAGGTGATAGCAATCAGCCCCAGCATTTT & CASGTGDSNQPQHF & EBV\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t Unk2 & 008\\_037\\_B & TGTGCCAGCAGTTTGAACACTGAAGCTTTCTTT    & CASSLNTEAFF  & Other\\\\\n",
       "\t Unk2 & 008\\_037\\_B & TGTGCCAGCAGTTTGGGGAACACTGAAGCTTTCTTT & CASSLGNTEAFF & Other\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2186 × 5\n",
       "\n",
       "| patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | junction &lt;chr&gt; | junction_aa &lt;chr&gt; | pathology &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 008_001 | 008_001_D | TGTGCCACCGGGACAGGGGATAGCAATCAGCCCCAGCATTTT | CATGTGDSNQPQHF | EBV |\n",
       "| 008_001 | 008_001_D | TGTGCCAGCAGTACAGGGGATAGCAATCAGCCCCAGCATTTT | CASSTGDSNQPQHF | EBV |\n",
       "| 008_001 | 008_001_D | TGTGCCAGCGGGACAGGTGATAGCAATCAGCCCCAGCATTTT | CASGTGDSNQPQHF | EBV |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| Unk2 | 008_037_B | TGTGCCAGCAGTTTGAACACTGAAGCTTTCTTT    | CASSLNTEAFF  | Other |\n",
       "| Unk2 | 008_037_B | TGTGCCAGCAGTTTGGGGAACACTGAAGCTTTCTTT | CASSLGNTEAFF | Other |\n",
       "\n"
      ],
      "text/plain": [
       "     patient_id repertoire_id junction                                  \n",
       "1    008_001    008_001_D     TGTGCCACCGGGACAGGGGATAGCAATCAGCCCCAGCATTTT\n",
       "2    008_001    008_001_D     TGTGCCAGCAGTACAGGGGATAGCAATCAGCCCCAGCATTTT\n",
       "3    008_001    008_001_D     TGTGCCAGCGGGACAGGTGATAGCAATCAGCCCCAGCATTTT\n",
       "⋮    ⋮          ⋮             ⋮                                         \n",
       "2185 Unk2       008_037_B     TGTGCCAGCAGTTTGAACACTGAAGCTTTCTTT         \n",
       "2186 Unk2       008_037_B     TGTGCCAGCAGTTTGGGGAACACTGAAGCTTTCTTT      \n",
       "     junction_aa    pathology\n",
       "1    CATGTGDSNQPQHF EBV      \n",
       "2    CASSTGDSNQPQHF EBV      \n",
       "3    CASGTGDSNQPQHF EBV      \n",
       "⋮    ⋮              ⋮        \n",
       "2185 CASSLNTEAFF    Other    \n",
       "2186 CASSLGNTEAFF   Other    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hla_matched_tcr <- annotated_nprod_table |>\n",
    "    mutate(visit_code = if_else(is.na(visit_code), \"V01\", visit_code)) |>\n",
    "    filter(cohort %in% c(\"Hippos\", \"ARKS\") & visit_code == \"V01\" &\n",
    "        tissue_type %in% c(\"NAT\", \"Tumor\")) |>\n",
    "    inner_join(antigen_db, by = c(\"junction_aa\" = \"trb_cdr3_aa\"), \n",
    "        relationship = \"many-to-many\") |>\n",
    "    dplyr::select(patient_id, repertoire_id, junction, junction_aa, epitope, pathology,\n",
    "        antigen, mhc_allele, aAllele1:DRB345allele2) |> \n",
    "    pivot_longer(cols = aAllele1:DRB345allele2, names_to = \"allele\",\n",
    "        values_to = \"mhc_value\") |> \n",
    "    mutate(mhc_allele = str_remove(mhc_allele, \"HLA-\"),\n",
    "        hla_matched = if_else(mhc_allele == mhc_value |\n",
    "            mhc_allele == str_extract(mhc_value, \"\\\\w+\\\\*\\\\d+:\\\\d+\") |\n",
    "            mhc_allele == str_extract(mhc_value, \"\\\\w+\\\\*\\\\d+\"), TRUE, FALSE)) |>\n",
    "    filter(hla_matched) |> \n",
    "    dplyr::select(patient_id:junction_aa, epitope:antigen) |>\n",
    "    distinct() |> \n",
    "    group_by(patient_id, repertoire_id, junction, junction_aa) |>\n",
    "    summarise(npath = length(unique(pathology)),\n",
    "        pathology = str_c(unique(pathology), collapse = \";\")) |>\n",
    "    ungroup() |>\n",
    "    mutate(pathology = if_else(npath > 1, \"Multi-pathogen\", pathology)) |>\n",
    "    dplyr::select(patient_id, repertoire_id, junction, junction_aa, pathology) |>\n",
    "    distinct() |>\n",
    "    mutate(pathology = recode(pathology, \n",
    "        `HomoSapiens` = \"Other\", \"Homo sapiens\" = \"Other\", \n",
    "        \"SelaginellaMoellendorffii\" = \"Other\", \"TriticumAestivum\" = \"Other\", \n",
    "        \"synthetic\" = \"Other\"))\n",
    "hla_matched_tcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 856 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>repertoire_id</th><th scope=col>pathology</th><th scope=col>phenotype</th><th scope=col>tissue_type</th><th scope=col>frequency</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Epidemic KS - NAT  </td><td>008_001_A</td><td>Unknown</td><td>Epidemic KS</td><td>NAT  </td><td>1.0000000000</td></tr>\n",
       "\t<tr><td>Epidemic KS - Tumor</td><td>008_001_B</td><td>Unknown</td><td>Epidemic KS</td><td>Tumor</td><td>1.0000000000</td></tr>\n",
       "\t<tr><td>Epidemic KS - Tumor</td><td>008_001_D</td><td>EBV    </td><td>Epidemic KS</td><td>Tumor</td><td>0.0001045779</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Epidemic KS - Tumor</td><td>008_265_D</td><td>Multi-pathogen</td><td>Epidemic KS</td><td>Tumor</td><td>0.0002486449</td></tr>\n",
       "\t<tr><td>Epidemic KS - Tumor</td><td>008_265_D</td><td>Unknown       </td><td>Epidemic KS</td><td>Tumor</td><td>0.9972151773</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 856 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " cohort & repertoire\\_id & pathology & phenotype & tissue\\_type & frequency\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Epidemic KS - NAT   & 008\\_001\\_A & Unknown & Epidemic KS & NAT   & 1.0000000000\\\\\n",
       "\t Epidemic KS - Tumor & 008\\_001\\_B & Unknown & Epidemic KS & Tumor & 1.0000000000\\\\\n",
       "\t Epidemic KS - Tumor & 008\\_001\\_D & EBV     & Epidemic KS & Tumor & 0.0001045779\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t Epidemic KS - Tumor & 008\\_265\\_D & Multi-pathogen & Epidemic KS & Tumor & 0.0002486449\\\\\n",
       "\t Epidemic KS - Tumor & 008\\_265\\_D & Unknown        & Epidemic KS & Tumor & 0.9972151773\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 856 × 6\n",
       "\n",
       "| cohort &lt;chr&gt; | repertoire_id &lt;chr&gt; | pathology &lt;chr&gt; | phenotype &lt;chr&gt; | tissue_type &lt;chr&gt; | frequency &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| Epidemic KS - NAT   | 008_001_A | Unknown | Epidemic KS | NAT   | 1.0000000000 |\n",
       "| Epidemic KS - Tumor | 008_001_B | Unknown | Epidemic KS | Tumor | 1.0000000000 |\n",
       "| Epidemic KS - Tumor | 008_001_D | EBV     | Epidemic KS | Tumor | 0.0001045779 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| Epidemic KS - Tumor | 008_265_D | Multi-pathogen | Epidemic KS | Tumor | 0.0002486449 |\n",
       "| Epidemic KS - Tumor | 008_265_D | Unknown        | Epidemic KS | Tumor | 0.9972151773 |\n",
       "\n"
      ],
      "text/plain": [
       "    cohort              repertoire_id pathology      phenotype   tissue_type\n",
       "1   Epidemic KS - NAT   008_001_A     Unknown        Epidemic KS NAT        \n",
       "2   Epidemic KS - Tumor 008_001_B     Unknown        Epidemic KS Tumor      \n",
       "3   Epidemic KS - Tumor 008_001_D     EBV            Epidemic KS Tumor      \n",
       "⋮   ⋮                   ⋮             ⋮              ⋮           ⋮          \n",
       "855 Epidemic KS - Tumor 008_265_D     Multi-pathogen Epidemic KS Tumor      \n",
       "856 Epidemic KS - Tumor 008_265_D     Unknown        Epidemic KS Tumor      \n",
       "    frequency   \n",
       "1   1.0000000000\n",
       "2   1.0000000000\n",
       "3   0.0001045779\n",
       "⋮   ⋮           \n",
       "855 0.0002486449\n",
       "856 0.9972151773"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "path_annotated_nprod_table <- annotated_nprod_table |> \n",
    "    mutate(visit_code = if_else(is.na(visit_code), \"V01\", visit_code)) |>\n",
    "    filter(cohort %in% c(\"Hippos\", \"ARKS\") & visit_code == \"V01\" &\n",
    "        tissue_type %in% c(\"NAT\", \"Tumor\")) |>\n",
    "    left_join(hla_matched_tcr, by = c(\"patient_id\", \"repertoire_id\", \"junction\",\n",
    "        \"junction_aa\")) |>\n",
    "    dplyr::select(cohort, patient_id, repertoire_id, tissue_type, phenotype, pathology, junction_aa, duplicate_frequency) |>\n",
    "    mutate(pathology = str_replace_na(pathology, \"Unknown\"),\n",
    "        phenotype = if_else(cohort == \"ARKS\", \"ARKS\", phenotype)) |>\n",
    "    group_by(cohort, repertoire_id, pathology, phenotype, tissue_type) |>\n",
    "    summarize(frequency = sum(duplicate_frequency)) |>\n",
    "    ungroup() |>\n",
    "    mutate(cohort = if_else(cohort == \"Hippos\", \n",
    "            str_c(phenotype, tissue_type, sep = \" - \"), \"ARKS\"))\n",
    "path_annotated_nprod_table "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare GLIPH inputs for KSTME datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1069462 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>junction_aa</th><th scope=col>v_call</th><th scope=col>j_call</th><th scope=col>cdra</th><th scope=col>gliph_id</th><th scope=col>duplicate_count</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>YSSAATIGQSRANVLTF</td><td>TRBV17-1</td><td>TRBJ2-6</td><td>NA</td><td>008_001_A:Epidemic KS - NAT</td><td>1</td></tr>\n",
       "\t<tr><td>CSVTPGLSYEQYF    </td><td>TRBV29-1</td><td>TRBJ2-7</td><td>NA</td><td>008_001_A:Epidemic KS - NAT</td><td>1</td></tr>\n",
       "\t<tr><td>CSAYKGQTGVNTEAFF </td><td>NA      </td><td>TRBJ1-1</td><td>NA</td><td>008_001_A:Epidemic KS - NAT</td><td>1</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>CVALTRGTEAFF</td><td>TRBV2-1</td><td>TRBJ1-1</td><td>NA</td><td>008_265_D:Epidemic KS - Tumor</td><td>9</td></tr>\n",
       "\t<tr><td>CYGALAKNIQYF</td><td>TRBV5-4</td><td>TRBJ2-4</td><td>NA</td><td>008_265_D:Epidemic KS - Tumor</td><td>6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1069462 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " junction\\_aa & v\\_call & j\\_call & cdra & gliph\\_id & duplicate\\_count\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t YSSAATIGQSRANVLTF & TRBV17-1 & TRBJ2-6 & NA & 008\\_001\\_A:Epidemic KS - NAT & 1\\\\\n",
       "\t CSVTPGLSYEQYF     & TRBV29-1 & TRBJ2-7 & NA & 008\\_001\\_A:Epidemic KS - NAT & 1\\\\\n",
       "\t CSAYKGQTGVNTEAFF  & NA       & TRBJ1-1 & NA & 008\\_001\\_A:Epidemic KS - NAT & 1\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t CVALTRGTEAFF & TRBV2-1 & TRBJ1-1 & NA & 008\\_265\\_D:Epidemic KS - Tumor & 9\\\\\n",
       "\t CYGALAKNIQYF & TRBV5-4 & TRBJ2-4 & NA & 008\\_265\\_D:Epidemic KS - Tumor & 6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1069462 × 6\n",
       "\n",
       "| junction_aa &lt;chr&gt; | v_call &lt;chr&gt; | j_call &lt;chr&gt; | cdra &lt;chr&gt; | gliph_id &lt;chr&gt; | duplicate_count &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| YSSAATIGQSRANVLTF | TRBV17-1 | TRBJ2-6 | NA | 008_001_A:Epidemic KS - NAT | 1 |\n",
       "| CSVTPGLSYEQYF     | TRBV29-1 | TRBJ2-7 | NA | 008_001_A:Epidemic KS - NAT | 1 |\n",
       "| CSAYKGQTGVNTEAFF  | NA       | TRBJ1-1 | NA | 008_001_A:Epidemic KS - NAT | 1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| CVALTRGTEAFF | TRBV2-1 | TRBJ1-1 | NA | 008_265_D:Epidemic KS - Tumor | 9 |\n",
       "| CYGALAKNIQYF | TRBV5-4 | TRBJ2-4 | NA | 008_265_D:Epidemic KS - Tumor | 6 |\n",
       "\n"
      ],
      "text/plain": [
       "        junction_aa       v_call   j_call  cdra gliph_id                     \n",
       "1       YSSAATIGQSRANVLTF TRBV17-1 TRBJ2-6 NA   008_001_A:Epidemic KS - NAT  \n",
       "2       CSVTPGLSYEQYF     TRBV29-1 TRBJ2-7 NA   008_001_A:Epidemic KS - NAT  \n",
       "3       CSAYKGQTGVNTEAFF  NA       TRBJ1-1 NA   008_001_A:Epidemic KS - NAT  \n",
       "⋮       ⋮                 ⋮        ⋮       ⋮    ⋮                            \n",
       "1069461 CVALTRGTEAFF      TRBV2-1  TRBJ1-1 NA   008_265_D:Epidemic KS - Tumor\n",
       "1069462 CYGALAKNIQYF      TRBV5-4  TRBJ2-4 NA   008_265_D:Epidemic KS - Tumor\n",
       "        duplicate_count\n",
       "1       1              \n",
       "2       1              \n",
       "3       1              \n",
       "⋮       ⋮              \n",
       "1069461 9              \n",
       "1069462 6              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# KSTME GLIPH table\n",
    "kstme_gliph_trb_input <- annotated_nprod_table |>\n",
    "    mutate(visit_code = if_else(is.na(visit_code), \"V01\", visit_code)) |>\n",
    "    filter(cohort %in% c(\"Hippos\") & visit_code == \"V01\" &\n",
    "        tissue_type %in% c(\"NAT\", \"Tumor\")) |>\n",
    "    mutate(phenotype = if_else(is.na(phenotype), \"Epidemic KS\", phenotype)) |>\n",
    "    dplyr::select(junction_aa, v_call, j_call, repertoire_id, tissue_type, phenotype,\n",
    "        duplicate_count) |>\n",
    "    mutate(gliph_id = str_c(repertoire_id,\n",
    "            str_c(phenotype, tissue_type, sep = \" - \"), sep = \":\"),\n",
    "            cdra = NA_character_) |>\n",
    "    dplyr::select(junction_aa, v_call, j_call, cdra, gliph_id, duplicate_count)\n",
    "write_tsv(kstme_gliph_trb_input, file = kstme_gliph_input_path, col_names = FALSE)\n",
    "kstme_gliph_trb_input "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 342 × 19</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>repertoire_id</th><th scope=col>aAllele1</th><th scope=col>aAllele2</th><th scope=col>bAllele1</th><th scope=col>bAllele2</th><th scope=col>cAllele1</th><th scope=col>cAllele2</th><th scope=col>DPA1allele1</th><th scope=col>DPA1allele2</th><th scope=col>DPB1allele1</th><th scope=col>DPB1allele2</th><th scope=col>DQA1allele1</th><th scope=col>DQA1allele2</th><th scope=col>DQB1allele1</th><th scope=col>DQB1allele2</th><th scope=col>DRB1allele1</th><th scope=col>DRB1allele2</th><th scope=col>DRB345allele1</th><th scope=col>DRB345allele2</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001_A</td><td>A*23:17:01</td><td>A*29:02:01</td><td>B*35:01:01</td><td>B*42:01:01</td><td>C*04:01:01</td><td>C*17:01:01</td><td>DPA1*02:02:02</td><td>DPA1*03:01:01</td><td>DPB1*01:01:01</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*04:01:01</td><td>DQB1*04:02:01</td><td>DQB1*05:01:01</td><td>DRB1*03:02:01</td><td>DRB1*11:01:02</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td></tr>\n",
       "\t<tr><td>008_001_B</td><td>A*23:17:01</td><td>A*29:02:01</td><td>B*35:01:01</td><td>B*42:01:01</td><td>C*04:01:01</td><td>C*17:01:01</td><td>DPA1*02:02:02</td><td>DPA1*03:01:01</td><td>DPB1*01:01:01</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*04:01:01</td><td>DQB1*04:02:01</td><td>DQB1*05:01:01</td><td>DRB1*03:02:01</td><td>DRB1*11:01:02</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td></tr>\n",
       "\t<tr><td>008_001_D</td><td>A*23:17:01</td><td>A*29:02:01</td><td>B*35:01:01</td><td>B*42:01:01</td><td>C*04:01:01</td><td>C*17:01:01</td><td>DPA1*02:02:02</td><td>DPA1*03:01:01</td><td>DPB1*01:01:01</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*04:01:01</td><td>DQB1*04:02:01</td><td>DQB1*05:01:01</td><td>DRB1*03:02:01</td><td>DRB1*11:01:02</td><td>DRB3*01:01:02</td><td>DRB3*03:01:01</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_264_D</td><td>A*02:01:01</td><td>A*74:01:01</td><td>B*08:01:01</td><td>B*53:01:01</td><td>C*04:01:01</td><td>C*07:01:02</td><td>DPA1*02:01:01</td><td>DPA1*03:01:01</td><td>DPB1*105:01:01</td><td>DPB1*13:01:01 </td><td>DQA1*01:02:01</td><td>DQA1*04:01:02</td><td>DQB1*03:19:01</td><td>DQB1*06:09:01</td><td>DRB1*08:04:01</td><td>DRB1*13:02:01</td><td>DRB3*03:01:01</td><td>DRB345*Not_Present</td></tr>\n",
       "\t<tr><td>008_265_D</td><td>A*30:02:01</td><td>A*32:01:01</td><td>B*07:02:01</td><td>B*18:01:01</td><td>C*07:04:01</td><td>C*07:18:01</td><td>DPA1*03:01:01</td><td>DPA1*03:01:01</td><td>DPB1*105:01:01</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*01:03:01</td><td>DQB1*06:02:01</td><td>DQB1*06:03:01</td><td>DRB1*11:01:02</td><td>DRB1*13:01:01</td><td>DRB3*02:02:01</td><td>DRB3*03:01:01     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 342 × 19\n",
       "\\begin{tabular}{lllllllllllllllllll}\n",
       " repertoire\\_id & aAllele1 & aAllele2 & bAllele1 & bAllele2 & cAllele1 & cAllele2 & DPA1allele1 & DPA1allele2 & DPB1allele1 & DPB1allele2 & DQA1allele1 & DQA1allele2 & DQB1allele1 & DQB1allele2 & DRB1allele1 & DRB1allele2 & DRB345allele1 & DRB345allele2\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_001\\_A & A*23:17:01 & A*29:02:01 & B*35:01:01 & B*42:01:01 & C*04:01:01 & C*17:01:01 & DPA1*02:02:02 & DPA1*03:01:01 & DPB1*01:01:01 & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*04:01:01 & DQB1*04:02:01 & DQB1*05:01:01 & DRB1*03:02:01 & DRB1*11:01:02 & DRB3*01:01:02 & DRB3*03:01:01\\\\\n",
       "\t 008\\_001\\_B & A*23:17:01 & A*29:02:01 & B*35:01:01 & B*42:01:01 & C*04:01:01 & C*17:01:01 & DPA1*02:02:02 & DPA1*03:01:01 & DPB1*01:01:01 & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*04:01:01 & DQB1*04:02:01 & DQB1*05:01:01 & DRB1*03:02:01 & DRB1*11:01:02 & DRB3*01:01:02 & DRB3*03:01:01\\\\\n",
       "\t 008\\_001\\_D & A*23:17:01 & A*29:02:01 & B*35:01:01 & B*42:01:01 & C*04:01:01 & C*17:01:01 & DPA1*02:02:02 & DPA1*03:01:01 & DPB1*01:01:01 & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*04:01:01 & DQB1*04:02:01 & DQB1*05:01:01 & DRB1*03:02:01 & DRB1*11:01:02 & DRB3*01:01:02 & DRB3*03:01:01\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_264\\_D & A*02:01:01 & A*74:01:01 & B*08:01:01 & B*53:01:01 & C*04:01:01 & C*07:01:02 & DPA1*02:01:01 & DPA1*03:01:01 & DPB1*105:01:01 & DPB1*13:01:01  & DQA1*01:02:01 & DQA1*04:01:02 & DQB1*03:19:01 & DQB1*06:09:01 & DRB1*08:04:01 & DRB1*13:02:01 & DRB3*03:01:01 & DRB345*Not\\_Present\\\\\n",
       "\t 008\\_265\\_D & A*30:02:01 & A*32:01:01 & B*07:02:01 & B*18:01:01 & C*07:04:01 & C*07:18:01 & DPA1*03:01:01 & DPA1*03:01:01 & DPB1*105:01:01 & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*01:03:01 & DQB1*06:02:01 & DQB1*06:03:01 & DRB1*11:01:02 & DRB1*13:01:01 & DRB3*02:02:01 & DRB3*03:01:01     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 342 × 19\n",
       "\n",
       "| repertoire_id &lt;chr&gt; | aAllele1 &lt;chr&gt; | aAllele2 &lt;chr&gt; | bAllele1 &lt;chr&gt; | bAllele2 &lt;chr&gt; | cAllele1 &lt;chr&gt; | cAllele2 &lt;chr&gt; | DPA1allele1 &lt;chr&gt; | DPA1allele2 &lt;chr&gt; | DPB1allele1 &lt;chr&gt; | DPB1allele2 &lt;chr&gt; | DQA1allele1 &lt;chr&gt; | DQA1allele2 &lt;chr&gt; | DQB1allele1 &lt;chr&gt; | DQB1allele2 &lt;chr&gt; | DRB1allele1 &lt;chr&gt; | DRB1allele2 &lt;chr&gt; | DRB345allele1 &lt;chr&gt; | DRB345allele2 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 008_001_A | A*23:17:01 | A*29:02:01 | B*35:01:01 | B*42:01:01 | C*04:01:01 | C*17:01:01 | DPA1*02:02:02 | DPA1*03:01:01 | DPB1*01:01:01 | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*04:01:01 | DQB1*04:02:01 | DQB1*05:01:01 | DRB1*03:02:01 | DRB1*11:01:02 | DRB3*01:01:02 | DRB3*03:01:01 |\n",
       "| 008_001_B | A*23:17:01 | A*29:02:01 | B*35:01:01 | B*42:01:01 | C*04:01:01 | C*17:01:01 | DPA1*02:02:02 | DPA1*03:01:01 | DPB1*01:01:01 | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*04:01:01 | DQB1*04:02:01 | DQB1*05:01:01 | DRB1*03:02:01 | DRB1*11:01:02 | DRB3*01:01:02 | DRB3*03:01:01 |\n",
       "| 008_001_D | A*23:17:01 | A*29:02:01 | B*35:01:01 | B*42:01:01 | C*04:01:01 | C*17:01:01 | DPA1*02:02:02 | DPA1*03:01:01 | DPB1*01:01:01 | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*04:01:01 | DQB1*04:02:01 | DQB1*05:01:01 | DRB1*03:02:01 | DRB1*11:01:02 | DRB3*01:01:02 | DRB3*03:01:01 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_264_D | A*02:01:01 | A*74:01:01 | B*08:01:01 | B*53:01:01 | C*04:01:01 | C*07:01:02 | DPA1*02:01:01 | DPA1*03:01:01 | DPB1*105:01:01 | DPB1*13:01:01  | DQA1*01:02:01 | DQA1*04:01:02 | DQB1*03:19:01 | DQB1*06:09:01 | DRB1*08:04:01 | DRB1*13:02:01 | DRB3*03:01:01 | DRB345*Not_Present |\n",
       "| 008_265_D | A*30:02:01 | A*32:01:01 | B*07:02:01 | B*18:01:01 | C*07:04:01 | C*07:18:01 | DPA1*03:01:01 | DPA1*03:01:01 | DPB1*105:01:01 | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*01:03:01 | DQB1*06:02:01 | DQB1*06:03:01 | DRB1*11:01:02 | DRB1*13:01:01 | DRB3*02:02:01 | DRB3*03:01:01      |\n",
       "\n"
      ],
      "text/plain": [
       "    repertoire_id aAllele1   aAllele2   bAllele1   bAllele2   cAllele1  \n",
       "1   008_001_A     A*23:17:01 A*29:02:01 B*35:01:01 B*42:01:01 C*04:01:01\n",
       "2   008_001_B     A*23:17:01 A*29:02:01 B*35:01:01 B*42:01:01 C*04:01:01\n",
       "3   008_001_D     A*23:17:01 A*29:02:01 B*35:01:01 B*42:01:01 C*04:01:01\n",
       "⋮   ⋮             ⋮          ⋮          ⋮          ⋮          ⋮         \n",
       "341 008_264_D     A*02:01:01 A*74:01:01 B*08:01:01 B*53:01:01 C*04:01:01\n",
       "342 008_265_D     A*30:02:01 A*32:01:01 B*07:02:01 B*18:01:01 C*07:04:01\n",
       "    cAllele2   DPA1allele1   DPA1allele2   DPB1allele1    DPB1allele2   \n",
       "1   C*17:01:01 DPA1*02:02:02 DPA1*03:01:01 DPB1*01:01:01  DPB1*105:01:01\n",
       "2   C*17:01:01 DPA1*02:02:02 DPA1*03:01:01 DPB1*01:01:01  DPB1*105:01:01\n",
       "3   C*17:01:01 DPA1*02:02:02 DPA1*03:01:01 DPB1*01:01:01  DPB1*105:01:01\n",
       "⋮   ⋮          ⋮             ⋮             ⋮              ⋮             \n",
       "341 C*07:01:02 DPA1*02:01:01 DPA1*03:01:01 DPB1*105:01:01 DPB1*13:01:01 \n",
       "342 C*07:18:01 DPA1*03:01:01 DPA1*03:01:01 DPB1*105:01:01 DPB1*105:01:01\n",
       "    DQA1allele1   DQA1allele2   DQB1allele1   DQB1allele2   DRB1allele1  \n",
       "1   DQA1*01:02:01 DQA1*04:01:01 DQB1*04:02:01 DQB1*05:01:01 DRB1*03:02:01\n",
       "2   DQA1*01:02:01 DQA1*04:01:01 DQB1*04:02:01 DQB1*05:01:01 DRB1*03:02:01\n",
       "3   DQA1*01:02:01 DQA1*04:01:01 DQB1*04:02:01 DQB1*05:01:01 DRB1*03:02:01\n",
       "⋮   ⋮             ⋮             ⋮             ⋮             ⋮            \n",
       "341 DQA1*01:02:01 DQA1*04:01:02 DQB1*03:19:01 DQB1*06:09:01 DRB1*08:04:01\n",
       "342 DQA1*01:02:01 DQA1*01:03:01 DQB1*06:02:01 DQB1*06:03:01 DRB1*11:01:02\n",
       "    DRB1allele2   DRB345allele1 DRB345allele2     \n",
       "1   DRB1*11:01:02 DRB3*01:01:02 DRB3*03:01:01     \n",
       "2   DRB1*11:01:02 DRB3*01:01:02 DRB3*03:01:01     \n",
       "3   DRB1*11:01:02 DRB3*01:01:02 DRB3*03:01:01     \n",
       "⋮   ⋮             ⋮             ⋮                 \n",
       "341 DRB1*13:02:01 DRB3*03:01:01 DRB345*Not_Present\n",
       "342 DRB1*13:01:01 DRB3*02:02:01 DRB3*03:01:01     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# KSTME GLIPH HLA table\n",
    "kstme_gliph_hla_input <- annotated_nprod_table |>\n",
    "    mutate(visit_code = if_else(is.na(visit_code), \"V01\", visit_code)) |>\n",
    "    filter(cohort %in% c(\"Hippos\") & visit_code == \"V01\" &\n",
    "        tissue_type %in% c(\"NAT\", \"Tumor\")) |>\n",
    "    dplyr::select(repertoire_id, aAllele1:DRB345allele2) |>\n",
    "    distinct()\n",
    "write_tsv(kstme_gliph_hla_input, file = kstme_gliph_hla_input_path,\n",
    "    col_names = FALSE)\n",
    "kstme_gliph_hla_input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Format GLIPH results and annotate with antigenic specificity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kstme_gliph_cluster_pred_table <- read_csv(kstme_gliph_clusters_path,\n",
    "        show_col_types = FALSE) \n",
    "kstme_gliph_hla_pred_table <- read_csv(kstme_gliph_hlapred_path, \n",
    "    show_col_types = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A spec_tbl_df: 1440808 × 29</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>index</th><th scope=col>pattern</th><th scope=col>Fisher_score</th><th scope=col>number_subject</th><th scope=col>number_unique_cdr3</th><th scope=col>final_score</th><th scope=col>hla_score</th><th scope=col>vb_score</th><th scope=col>expansion_score</th><th scope=col>length_score</th><th scope=col>⋯</th><th scope=col>HLA-B</th><th scope=col>HLA-C</th><th scope=col>HLA-DPA1</th><th scope=col>HLA-DPB1</th><th scope=col>HLA-DQA1</th><th scope=col>HLA-DQB1</th><th scope=col>HLA-DRB1</th><th scope=col>HLA-DRB3</th><th scope=col>HLA-DRB4</th><th scope=col>HLA-DRB5</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>KTYG</td><td>0</td><td>226</td><td>374</td><td>2.6e-08</td><td>0.0027</td><td>1</td><td>1</td><td>0.97</td><td>⋯</td><td>B*15:03:01/B*41:02:01 </td><td>C*02:10:01/C*17:01:01 </td><td>DPA1*01:03:01/DPA1*03:01:01</td><td>DPB1*105:01:01/DPB1*61:01N </td><td>DQA1*01:02:01/DQA1*04:01:02</td><td>DQB1*03:19:01/DQB1*06:09:01</td><td>DRB1*08:04:01/DRB1*13:02:01</td><td>DRB3*03:01:01/DRB345*Not_Present</td><td>-</td><td>-</td></tr>\n",
       "\t<tr><td>1</td><td>KTYG</td><td>0</td><td>226</td><td>374</td><td>2.6e-08</td><td>0.0027</td><td>1</td><td>1</td><td>0.97</td><td>⋯</td><td>B*49:01:01!/B*58:01:01</td><td>C*03:02:02/C*07:01:01!</td><td>DPA1*01:03:01/DPA1*01:03:01</td><td>DPB1*30:01:01/DPB1*39:01:01</td><td>DQA1*01:01:02/DQA1*01:02:01</td><td>DQB1*05:01:01/DQB1*06:04:01</td><td>DRB1*01:02:01/DRB1*13:02:01</td><td>DRB3*03:01:01/DRB345*Not_Present</td><td>-</td><td>-</td></tr>\n",
       "\t<tr><td>1</td><td>KTYG</td><td>0</td><td>226</td><td>374</td><td>2.6e-08</td><td>0.0027</td><td>1</td><td>1</td><td>0.97</td><td>⋯</td><td>B*49:01:01!/B*58:01:01</td><td>C*03:02:02/C*07:01:01!</td><td>DPA1*01:03:01/DPA1*01:03:01</td><td>DPB1*30:01:01/DPB1*39:01:01</td><td>DQA1*01:01:02/DQA1*01:02:01</td><td>DQB1*05:01:01/DQB1*06:04:01</td><td>DRB1*01:02:01/DRB1*13:02:01</td><td>DRB3*03:01:01/DRB345*Not_Present</td><td>-</td><td>-</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>873940</td><td>RKTY</td><td>1</td><td>6</td><td>4</td><td>2.5e-08</td><td>0.021</td><td>0.001</td><td>1</td><td>0.51</td><td>⋯</td><td>B*51:01:01/B*58:02:01</td><td>C*06:02:01/C*16:46   </td><td>DPA1*02:01:01/DPA1*02:01:01</td><td>DPB1*17:01:01!/DPB1*17:01:01!</td><td>DQA1*01:01:02/DQA1*01:02:01</td><td>DQB1*05:01:01/DQB1*06:02:01</td><td>DRB1*01:02:01/DRB1*11:01:02</td><td>DRB3*02:02:01/DRB345*Not_Present</td><td>-               </td><td>-</td></tr>\n",
       "\t<tr><td>873940</td><td>RKTY</td><td>1</td><td>6</td><td>4</td><td>2.5e-08</td><td>0.021</td><td>0.001</td><td>1</td><td>0.51</td><td>⋯</td><td>B*18:01:01/B*53:01:01</td><td>C*04:01:01/C*07:04:01</td><td>DPA1*02:01:08/DPA1*03:01:01</td><td>DPB1*01:01:01/DPB1*105:01:01 </td><td>DQA1*01:02:01/DQA1*02:01:01</td><td>DQB1*02:02:01/DQB1*06:02:01</td><td>DRB1*07:01:01/DRB1*11:01:02</td><td>DRB3*02:02:01                   </td><td>DRB4*01:01:01:01</td><td>-</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A spec\\_tbl\\_df: 1440808 × 29\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " index & pattern & Fisher\\_score & number\\_subject & number\\_unique\\_cdr3 & final\\_score & hla\\_score & vb\\_score & expansion\\_score & length\\_score & ⋯ & HLA-B & HLA-C & HLA-DPA1 & HLA-DPB1 & HLA-DQA1 & HLA-DQB1 & HLA-DRB1 & HLA-DRB3 & HLA-DRB4 & HLA-DRB5\\\\\n",
       " <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & KTYG & 0 & 226 & 374 & 2.6e-08 & 0.0027 & 1 & 1 & 0.97 & ⋯ & B*15:03:01/B*41:02:01  & C*02:10:01/C*17:01:01  & DPA1*01:03:01/DPA1*03:01:01 & DPB1*105:01:01/DPB1*61:01N  & DQA1*01:02:01/DQA1*04:01:02 & DQB1*03:19:01/DQB1*06:09:01 & DRB1*08:04:01/DRB1*13:02:01 & DRB3*03:01:01/DRB345*Not\\_Present & - & -\\\\\n",
       "\t 1 & KTYG & 0 & 226 & 374 & 2.6e-08 & 0.0027 & 1 & 1 & 0.97 & ⋯ & B*49:01:01!/B*58:01:01 & C*03:02:02/C*07:01:01! & DPA1*01:03:01/DPA1*01:03:01 & DPB1*30:01:01/DPB1*39:01:01 & DQA1*01:01:02/DQA1*01:02:01 & DQB1*05:01:01/DQB1*06:04:01 & DRB1*01:02:01/DRB1*13:02:01 & DRB3*03:01:01/DRB345*Not\\_Present & - & -\\\\\n",
       "\t 1 & KTYG & 0 & 226 & 374 & 2.6e-08 & 0.0027 & 1 & 1 & 0.97 & ⋯ & B*49:01:01!/B*58:01:01 & C*03:02:02/C*07:01:01! & DPA1*01:03:01/DPA1*01:03:01 & DPB1*30:01:01/DPB1*39:01:01 & DQA1*01:01:02/DQA1*01:02:01 & DQB1*05:01:01/DQB1*06:04:01 & DRB1*01:02:01/DRB1*13:02:01 & DRB3*03:01:01/DRB345*Not\\_Present & - & -\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 873940 & RKTY & 1 & 6 & 4 & 2.5e-08 & 0.021 & 0.001 & 1 & 0.51 & ⋯ & B*51:01:01/B*58:02:01 & C*06:02:01/C*16:46    & DPA1*02:01:01/DPA1*02:01:01 & DPB1*17:01:01!/DPB1*17:01:01! & DQA1*01:01:02/DQA1*01:02:01 & DQB1*05:01:01/DQB1*06:02:01 & DRB1*01:02:01/DRB1*11:01:02 & DRB3*02:02:01/DRB345*Not\\_Present & -                & -\\\\\n",
       "\t 873940 & RKTY & 1 & 6 & 4 & 2.5e-08 & 0.021 & 0.001 & 1 & 0.51 & ⋯ & B*18:01:01/B*53:01:01 & C*04:01:01/C*07:04:01 & DPA1*02:01:08/DPA1*03:01:01 & DPB1*01:01:01/DPB1*105:01:01  & DQA1*01:02:01/DQA1*02:01:01 & DQB1*02:02:01/DQB1*06:02:01 & DRB1*07:01:01/DRB1*11:01:02 & DRB3*02:02:01                    & DRB4*01:01:01:01 & -\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A spec_tbl_df: 1440808 × 29\n",
       "\n",
       "| index &lt;dbl&gt; | pattern &lt;chr&gt; | Fisher_score &lt;dbl&gt; | number_subject &lt;dbl&gt; | number_unique_cdr3 &lt;dbl&gt; | final_score &lt;dbl&gt; | hla_score &lt;dbl&gt; | vb_score &lt;dbl&gt; | expansion_score &lt;dbl&gt; | length_score &lt;dbl&gt; | ⋯ ⋯ | HLA-B &lt;chr&gt; | HLA-C &lt;chr&gt; | HLA-DPA1 &lt;chr&gt; | HLA-DPB1 &lt;chr&gt; | HLA-DQA1 &lt;chr&gt; | HLA-DQB1 &lt;chr&gt; | HLA-DRB1 &lt;chr&gt; | HLA-DRB3 &lt;chr&gt; | HLA-DRB4 &lt;chr&gt; | HLA-DRB5 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | KTYG | 0 | 226 | 374 | 2.6e-08 | 0.0027 | 1 | 1 | 0.97 | ⋯ | B*15:03:01/B*41:02:01  | C*02:10:01/C*17:01:01  | DPA1*01:03:01/DPA1*03:01:01 | DPB1*105:01:01/DPB1*61:01N  | DQA1*01:02:01/DQA1*04:01:02 | DQB1*03:19:01/DQB1*06:09:01 | DRB1*08:04:01/DRB1*13:02:01 | DRB3*03:01:01/DRB345*Not_Present | - | - |\n",
       "| 1 | KTYG | 0 | 226 | 374 | 2.6e-08 | 0.0027 | 1 | 1 | 0.97 | ⋯ | B*49:01:01!/B*58:01:01 | C*03:02:02/C*07:01:01! | DPA1*01:03:01/DPA1*01:03:01 | DPB1*30:01:01/DPB1*39:01:01 | DQA1*01:01:02/DQA1*01:02:01 | DQB1*05:01:01/DQB1*06:04:01 | DRB1*01:02:01/DRB1*13:02:01 | DRB3*03:01:01/DRB345*Not_Present | - | - |\n",
       "| 1 | KTYG | 0 | 226 | 374 | 2.6e-08 | 0.0027 | 1 | 1 | 0.97 | ⋯ | B*49:01:01!/B*58:01:01 | C*03:02:02/C*07:01:01! | DPA1*01:03:01/DPA1*01:03:01 | DPB1*30:01:01/DPB1*39:01:01 | DQA1*01:01:02/DQA1*01:02:01 | DQB1*05:01:01/DQB1*06:04:01 | DRB1*01:02:01/DRB1*13:02:01 | DRB3*03:01:01/DRB345*Not_Present | - | - |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 873940 | RKTY | 1 | 6 | 4 | 2.5e-08 | 0.021 | 0.001 | 1 | 0.51 | ⋯ | B*51:01:01/B*58:02:01 | C*06:02:01/C*16:46    | DPA1*02:01:01/DPA1*02:01:01 | DPB1*17:01:01!/DPB1*17:01:01! | DQA1*01:01:02/DQA1*01:02:01 | DQB1*05:01:01/DQB1*06:02:01 | DRB1*01:02:01/DRB1*11:01:02 | DRB3*02:02:01/DRB345*Not_Present | -                | - |\n",
       "| 873940 | RKTY | 1 | 6 | 4 | 2.5e-08 | 0.021 | 0.001 | 1 | 0.51 | ⋯ | B*18:01:01/B*53:01:01 | C*04:01:01/C*07:04:01 | DPA1*02:01:08/DPA1*03:01:01 | DPB1*01:01:01/DPB1*105:01:01  | DQA1*01:02:01/DQA1*02:01:01 | DQB1*02:02:01/DQB1*06:02:01 | DRB1*07:01:01/DRB1*11:01:02 | DRB3*02:02:01                    | DRB4*01:01:01:01 | - |\n",
       "\n"
      ],
      "text/plain": [
       "        index  pattern Fisher_score number_subject number_unique_cdr3\n",
       "1       1      KTYG    0            226            374               \n",
       "2       1      KTYG    0            226            374               \n",
       "3       1      KTYG    0            226            374               \n",
       "⋮       ⋮      ⋮       ⋮            ⋮              ⋮                 \n",
       "1440807 873940 RKTY    1            6              4                 \n",
       "1440808 873940 RKTY    1            6              4                 \n",
       "        final_score hla_score vb_score expansion_score length_score ⋯\n",
       "1       2.6e-08     0.0027    1        1               0.97         ⋯\n",
       "2       2.6e-08     0.0027    1        1               0.97         ⋯\n",
       "3       2.6e-08     0.0027    1        1               0.97         ⋯\n",
       "⋮       ⋮           ⋮         ⋮        ⋮               ⋮            ⋱\n",
       "1440807 2.5e-08     0.021     0.001    1               0.51         ⋯\n",
       "1440808 2.5e-08     0.021     0.001    1               0.51         ⋯\n",
       "        HLA-B                  HLA-C                 \n",
       "1       B*15:03:01/B*41:02:01  C*02:10:01/C*17:01:01 \n",
       "2       B*49:01:01!/B*58:01:01 C*03:02:02/C*07:01:01!\n",
       "3       B*49:01:01!/B*58:01:01 C*03:02:02/C*07:01:01!\n",
       "⋮       ⋮                      ⋮                     \n",
       "1440807 B*51:01:01/B*58:02:01  C*06:02:01/C*16:46    \n",
       "1440808 B*18:01:01/B*53:01:01  C*04:01:01/C*07:04:01 \n",
       "        HLA-DPA1                    HLA-DPB1                     \n",
       "1       DPA1*01:03:01/DPA1*03:01:01 DPB1*105:01:01/DPB1*61:01N   \n",
       "2       DPA1*01:03:01/DPA1*01:03:01 DPB1*30:01:01/DPB1*39:01:01  \n",
       "3       DPA1*01:03:01/DPA1*01:03:01 DPB1*30:01:01/DPB1*39:01:01  \n",
       "⋮       ⋮                           ⋮                            \n",
       "1440807 DPA1*02:01:01/DPA1*02:01:01 DPB1*17:01:01!/DPB1*17:01:01!\n",
       "1440808 DPA1*02:01:08/DPA1*03:01:01 DPB1*01:01:01/DPB1*105:01:01 \n",
       "        HLA-DQA1                    HLA-DQB1                   \n",
       "1       DQA1*01:02:01/DQA1*04:01:02 DQB1*03:19:01/DQB1*06:09:01\n",
       "2       DQA1*01:01:02/DQA1*01:02:01 DQB1*05:01:01/DQB1*06:04:01\n",
       "3       DQA1*01:01:02/DQA1*01:02:01 DQB1*05:01:01/DQB1*06:04:01\n",
       "⋮       ⋮                           ⋮                          \n",
       "1440807 DQA1*01:01:02/DQA1*01:02:01 DQB1*05:01:01/DQB1*06:02:01\n",
       "1440808 DQA1*01:02:01/DQA1*02:01:01 DQB1*02:02:01/DQB1*06:02:01\n",
       "        HLA-DRB1                    HLA-DRB3                        \n",
       "1       DRB1*08:04:01/DRB1*13:02:01 DRB3*03:01:01/DRB345*Not_Present\n",
       "2       DRB1*01:02:01/DRB1*13:02:01 DRB3*03:01:01/DRB345*Not_Present\n",
       "3       DRB1*01:02:01/DRB1*13:02:01 DRB3*03:01:01/DRB345*Not_Present\n",
       "⋮       ⋮                           ⋮                               \n",
       "1440807 DRB1*01:02:01/DRB1*11:01:02 DRB3*02:02:01/DRB345*Not_Present\n",
       "1440808 DRB1*07:01:01/DRB1*11:01:02 DRB3*02:02:01                   \n",
       "        HLA-DRB4         HLA-DRB5\n",
       "1       -                -       \n",
       "2       -                -       \n",
       "3       -                -       \n",
       "⋮       ⋮                ⋮       \n",
       "1440807 -                -       \n",
       "1440808 DRB4*01:01:01:01 -       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kstme_gliph_cluster_pred_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Total number of GLIPH groups in KSTME datasets : 215850\"\n",
      "[1] \"Number of GLIPH groups of length >= 3 : 215816\"\n",
      "[1] \"Number of GLIPH groups of length >= 3, vb_score <= 0.01, unique CDR3s >= 3, and unique PTIDs >= 3 : 43545\"\n"
     ]
    }
   ],
   "source": [
    "options(repr.matrix.max.rows=5, repr.matrix.max.columns=5)\n",
    "# Total number of GLIPH groups identified from the KSTME dataset\n",
    "print(str_c(\"Total number of GLIPH groups in KSTME datasets\",\n",
    "    length(unique(kstme_gliph_cluster_pred_table$pattern)), sep = \" : \"))\n",
    "print(str_c(\"Number of GLIPH groups of length >= 3\",\n",
    "    kstme_gliph_cluster_pred_table |> \n",
    "        filter(pattern != \"single\" & nchar(pattern) > 2) |>\n",
    "        pull(pattern) |>\n",
    "        unique() |>\n",
    "        length(), sep = \" : \")) \n",
    "print(str_c(\"Number of GLIPH groups of length >= 3, vb_score <= 0.01, unique CDR3s >= 3, and unique PTIDs >= 3\",\n",
    "    kstme_gliph_cluster_pred_table |> \n",
    "        filter(pattern != \"single\" & nchar(pattern) > 2) |>\n",
    "        separate(Sample, into = c(\"repertoire_id\", \"cohort\"), sep = \":\") |>\n",
    "        mutate(patient_id = str_extract(repertoire_id, \"008_\\\\d+\"),\n",
    "            pattern = if_else(pattern == \"single\", TcRb, pattern)) |>\n",
    "        group_by(pattern) |>\n",
    "        mutate(number_of_unique_cdr = length(unique(TcRb)),\n",
    "                number_of_unique_ptid = length(unique(patient_id))) |>\n",
    "        filter(vb_score <= 0.01 & number_of_unique_cdr >= 3 &\n",
    "                number_of_unique_ptid >= 3) |>\n",
    "        pull(pattern) |>\n",
    "        unique() |>\n",
    "        length(), sep = \" : \")) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A spec_tbl_df: 1516816 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>index</th><th scope=col>pattern</th><th scope=col>'HLA allele with lowest Fisher Score'</th><th scope=col>Pvalue</th><th scope=col>'Number of subjects in this cluster with this allele'</th><th scope=col>'Number of subjects in this cluster with HLA'</th><th scope=col>'Number of subjects with this allele in total'</th><th scope=col>'Number of subjects with HLA in total'</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>lKTYG</td><td>B*81</td><td>0.0250</td><td>25</td><td>226</td><td>30</td><td>342</td></tr>\n",
       "\t<tr><td>1</td><td>lKTYG</td><td>C*18</td><td>0.0027</td><td>44</td><td>226</td><td>53</td><td>342</td></tr>\n",
       "\t<tr><td>1</td><td>lKTYG</td><td>B*18</td><td>0.0580</td><td>55</td><td>226</td><td>74</td><td>342</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>873940</td><td>lRKTY</td><td>A*31   </td><td>0.021</td><td>2</td><td>6</td><td>14</td><td>342</td></tr>\n",
       "\t<tr><td>873940</td><td>lRKTY</td><td>DPB1*17</td><td>0.089</td><td>2</td><td>6</td><td>30</td><td>342</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A spec\\_tbl\\_df: 1516816 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " index & pattern & 'HLA allele with lowest Fisher Score' & Pvalue & 'Number of subjects in this cluster with this allele' & 'Number of subjects in this cluster with HLA' & 'Number of subjects with this allele in total' & 'Number of subjects with HLA in total'\\\\\n",
       " <dbl> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1 & lKTYG & B*81 & 0.0250 & 25 & 226 & 30 & 342\\\\\n",
       "\t 1 & lKTYG & C*18 & 0.0027 & 44 & 226 & 53 & 342\\\\\n",
       "\t 1 & lKTYG & B*18 & 0.0580 & 55 & 226 & 74 & 342\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 873940 & lRKTY & A*31    & 0.021 & 2 & 6 & 14 & 342\\\\\n",
       "\t 873940 & lRKTY & DPB1*17 & 0.089 & 2 & 6 & 30 & 342\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A spec_tbl_df: 1516816 × 8\n",
       "\n",
       "| index &lt;dbl&gt; | pattern &lt;chr&gt; | 'HLA allele with lowest Fisher Score' &lt;chr&gt; | Pvalue &lt;dbl&gt; | 'Number of subjects in this cluster with this allele' &lt;dbl&gt; | 'Number of subjects in this cluster with HLA' &lt;dbl&gt; | 'Number of subjects with this allele in total' &lt;dbl&gt; | 'Number of subjects with HLA in total' &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | lKTYG | B*81 | 0.0250 | 25 | 226 | 30 | 342 |\n",
       "| 1 | lKTYG | C*18 | 0.0027 | 44 | 226 | 53 | 342 |\n",
       "| 1 | lKTYG | B*18 | 0.0580 | 55 | 226 | 74 | 342 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 873940 | lRKTY | A*31    | 0.021 | 2 | 6 | 14 | 342 |\n",
       "| 873940 | lRKTY | DPB1*17 | 0.089 | 2 | 6 | 30 | 342 |\n",
       "\n"
      ],
      "text/plain": [
       "        index  pattern 'HLA allele with lowest Fisher Score' Pvalue\n",
       "1       1      lKTYG   B*81                                  0.0250\n",
       "2       1      lKTYG   C*18                                  0.0027\n",
       "3       1      lKTYG   B*18                                  0.0580\n",
       "⋮       ⋮      ⋮       ⋮                                     ⋮     \n",
       "1516815 873940 lRKTY   A*31                                  0.021 \n",
       "1516816 873940 lRKTY   DPB1*17                               0.089 \n",
       "        'Number of subjects in this cluster with this allele'\n",
       "1       25                                                   \n",
       "2       44                                                   \n",
       "3       55                                                   \n",
       "⋮       ⋮                                                    \n",
       "1516815 2                                                    \n",
       "1516816 2                                                    \n",
       "        'Number of subjects in this cluster with HLA'\n",
       "1       226                                          \n",
       "2       226                                          \n",
       "3       226                                          \n",
       "⋮       ⋮                                            \n",
       "1516815 6                                            \n",
       "1516816 6                                            \n",
       "        'Number of subjects with this allele in total'\n",
       "1       30                                            \n",
       "2       53                                            \n",
       "3       74                                            \n",
       "⋮       ⋮                                             \n",
       "1516815 14                                            \n",
       "1516816 30                                            \n",
       "        'Number of subjects with HLA in total'\n",
       "1       342                                   \n",
       "2       342                                   \n",
       "3       342                                   \n",
       "⋮       ⋮                                     \n",
       "1516815 342                                   \n",
       "1516816 342                                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kstme_gliph_hla_pred_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Total number of HLA restrictions predicted by GLIPH : 1516816\"\n",
      "[1] \"Total number of HLA restrictions for GLIPH gorups of length >= 3 : 993612\"\n",
      "[1] \"Total number of significant HLA restrictions for GLIPH gorups of length >= 3 : 640223\"\n"
     ]
    }
   ],
   "source": [
    "print(str_c(\"Total number of HLA restrictions predicted by GLIPH\", \n",
    "    nrow(kstme_gliph_hla_pred_table), sep = \" : \"))\n",
    "print(str_c(\"Total number of HLA restrictions for GLIPH gorups of length >= 3\",\n",
    "    kstme_gliph_hla_pred_table |> \n",
    "        mutate(pattern = str_remove(pattern, \"^[lg]\")) |>\n",
    "        filter(pattern != \"single\" & nchar(pattern) > 2) |>\n",
    "        nrow(), sep = \" : \")) \n",
    "print(str_c(\"Total number of significant HLA restrictions for GLIPH gorups of length >= 3\",\n",
    "    kstme_gliph_hla_pred_table |> \n",
    "        mutate(pattern = str_remove(pattern, \"^[lg]\")) |>\n",
    "        filter(pattern != \"single\" & nchar(pattern) > 2 & Pvalue <= 0.05) |>\n",
    "        nrow(), sep = \" : \")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 521659 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pattern</th><th scope=col>vb_score</th><th scope=col>hla_score</th><th scope=col>expansion_score</th><th scope=col>TcRb</th><th scope=col>repertoire_id</th><th scope=col>cohort</th><th scope=col>V</th><th scope=col>J</th><th scope=col>Freq</th><th scope=col>patient_id</th><th scope=col>number_of_unique_cdr</th><th scope=col>number_of_unique_ptid</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>HHQP</td><td>0.001</td><td>0.023</td><td>1</td><td>CASSALIHFWHHQPQHF</td><td>008_074_D</td><td>Endemic KS - Tumor</td><td>NA</td><td>TRBJ1-5</td><td>1</td><td>008_074</td><td>155</td><td>89</td></tr>\n",
       "\t<tr><td>HHQP</td><td>0.001</td><td>0.023</td><td>1</td><td>CASSLNLRHHQPQHF  </td><td>008_143_D</td><td>Endemic KS - Tumor</td><td>NA</td><td>TRBJ1-5</td><td>1</td><td>008_143</td><td>155</td><td>89</td></tr>\n",
       "\t<tr><td>HHQP</td><td>0.001</td><td>0.023</td><td>1</td><td>CASSECLPIFHHQPQHF</td><td>008_189_B</td><td>Endemic KS - Tumor</td><td>NA</td><td>TRBJ1-5</td><td>1</td><td>008_189</td><td>155</td><td>89</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>RKTY</td><td>0.001</td><td>0.021</td><td>1</td><td>CASGSTRAARLATLLDPRKTYGYTF</td><td>008_008_C</td><td>Epidemic KS - Tumor</td><td>NA     </td><td>TRBJ1-2</td><td>1</td><td>008_008</td><td>243</td><td>99</td></tr>\n",
       "\t<tr><td>RKTY</td><td>0.001</td><td>0.021</td><td>1</td><td>CTSRLRSLCERPMVRKTYGYTF   </td><td>008_175_B</td><td>Endemic KS - Tumor </td><td>TRBV1-1</td><td>TRBJ1-2</td><td>1</td><td>008_175</td><td>243</td><td>99</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 521659 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " pattern & vb\\_score & hla\\_score & expansion\\_score & TcRb & repertoire\\_id & cohort & V & J & Freq & patient\\_id & number\\_of\\_unique\\_cdr & number\\_of\\_unique\\_ptid\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t HHQP & 0.001 & 0.023 & 1 & CASSALIHFWHHQPQHF & 008\\_074\\_D & Endemic KS - Tumor & NA & TRBJ1-5 & 1 & 008\\_074 & 155 & 89\\\\\n",
       "\t HHQP & 0.001 & 0.023 & 1 & CASSLNLRHHQPQHF   & 008\\_143\\_D & Endemic KS - Tumor & NA & TRBJ1-5 & 1 & 008\\_143 & 155 & 89\\\\\n",
       "\t HHQP & 0.001 & 0.023 & 1 & CASSECLPIFHHQPQHF & 008\\_189\\_B & Endemic KS - Tumor & NA & TRBJ1-5 & 1 & 008\\_189 & 155 & 89\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t RKTY & 0.001 & 0.021 & 1 & CASGSTRAARLATLLDPRKTYGYTF & 008\\_008\\_C & Epidemic KS - Tumor & NA      & TRBJ1-2 & 1 & 008\\_008 & 243 & 99\\\\\n",
       "\t RKTY & 0.001 & 0.021 & 1 & CTSRLRSLCERPMVRKTYGYTF    & 008\\_175\\_B & Endemic KS - Tumor  & TRBV1-1 & TRBJ1-2 & 1 & 008\\_175 & 243 & 99\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 521659 × 13\n",
       "\n",
       "| pattern &lt;chr&gt; | vb_score &lt;dbl&gt; | hla_score &lt;dbl&gt; | expansion_score &lt;dbl&gt; | TcRb &lt;chr&gt; | repertoire_id &lt;chr&gt; | cohort &lt;chr&gt; | V &lt;chr&gt; | J &lt;chr&gt; | Freq &lt;dbl&gt; | patient_id &lt;chr&gt; | number_of_unique_cdr &lt;int&gt; | number_of_unique_ptid &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HHQP | 0.001 | 0.023 | 1 | CASSALIHFWHHQPQHF | 008_074_D | Endemic KS - Tumor | NA | TRBJ1-5 | 1 | 008_074 | 155 | 89 |\n",
       "| HHQP | 0.001 | 0.023 | 1 | CASSLNLRHHQPQHF   | 008_143_D | Endemic KS - Tumor | NA | TRBJ1-5 | 1 | 008_143 | 155 | 89 |\n",
       "| HHQP | 0.001 | 0.023 | 1 | CASSECLPIFHHQPQHF | 008_189_B | Endemic KS - Tumor | NA | TRBJ1-5 | 1 | 008_189 | 155 | 89 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| RKTY | 0.001 | 0.021 | 1 | CASGSTRAARLATLLDPRKTYGYTF | 008_008_C | Epidemic KS - Tumor | NA      | TRBJ1-2 | 1 | 008_008 | 243 | 99 |\n",
       "| RKTY | 0.001 | 0.021 | 1 | CTSRLRSLCERPMVRKTYGYTF    | 008_175_B | Endemic KS - Tumor  | TRBV1-1 | TRBJ1-2 | 1 | 008_175 | 243 | 99 |\n",
       "\n"
      ],
      "text/plain": [
       "       pattern vb_score hla_score expansion_score TcRb                     \n",
       "1      HHQP    0.001    0.023     1               CASSALIHFWHHQPQHF        \n",
       "2      HHQP    0.001    0.023     1               CASSLNLRHHQPQHF          \n",
       "3      HHQP    0.001    0.023     1               CASSECLPIFHHQPQHF        \n",
       "⋮      ⋮       ⋮        ⋮         ⋮               ⋮                        \n",
       "521658 RKTY    0.001    0.021     1               CASGSTRAARLATLLDPRKTYGYTF\n",
       "521659 RKTY    0.001    0.021     1               CTSRLRSLCERPMVRKTYGYTF   \n",
       "       repertoire_id cohort              V       J       Freq patient_id\n",
       "1      008_074_D     Endemic KS - Tumor  NA      TRBJ1-5 1    008_074   \n",
       "2      008_143_D     Endemic KS - Tumor  NA      TRBJ1-5 1    008_143   \n",
       "3      008_189_B     Endemic KS - Tumor  NA      TRBJ1-5 1    008_189   \n",
       "⋮      ⋮             ⋮                   ⋮       ⋮       ⋮    ⋮         \n",
       "521658 008_008_C     Epidemic KS - Tumor NA      TRBJ1-2 1    008_008   \n",
       "521659 008_175_B     Endemic KS - Tumor  TRBV1-1 TRBJ1-2 1    008_175   \n",
       "       number_of_unique_cdr number_of_unique_ptid\n",
       "1      155                  89                   \n",
       "2      155                  89                   \n",
       "3      155                  89                   \n",
       "⋮      ⋮                    ⋮                    \n",
       "521658 243                  99                   \n",
       "521659 243                  99                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Filter out high confidence GLIPH groups with a pattern length of >2 where, \n",
    "# the pattern has at 3 unique CDR3s associated with it, at least 3 PTIDs \n",
    "# associated with it and with a vb_score <= 0.01\n",
    "kstme_hc_gliph_cluster_table <- kstme_gliph_cluster_pred_table |>\n",
    "  dplyr::select(pattern, vb_score, hla_score, expansion_score, TcRb, Sample, V, J,\n",
    "    Freq) |>\n",
    "  separate(Sample, into = c(\"repertoire_id\", \"cohort\"), sep = \":\") |>\n",
    "  mutate(patient_id = str_extract(repertoire_id, \"008_\\\\d+\")) |>\n",
    "  group_by(pattern) |>\n",
    "  mutate(number_of_unique_cdr = length(unique(TcRb)),\n",
    "    number_of_unique_ptid = length(unique(patient_id))) |>\n",
    "  ungroup() |>\n",
    "  filter(vb_score <= 0.01 & number_of_unique_cdr >= 3 &\n",
    "    number_of_unique_ptid >= 3 & pattern != \"single\" & nchar(pattern) > 2)\n",
    "kstme_hc_gliph_cluster_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 220850 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pattern</th><th scope=col>sig_allele</th><th scope=col>Pvalue</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>RKTY</td><td>B*81</td><td>0.04500</td></tr>\n",
       "\t<tr><td>RKTY</td><td>C*18</td><td>0.00016</td></tr>\n",
       "\t<tr><td>RKTY</td><td>B*18</td><td>0.04900</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>SLGLAGG%NE</td><td>A*74</td><td>0.021</td></tr>\n",
       "\t<tr><td>RKTY      </td><td>A*31</td><td>0.021</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 220850 × 3\n",
       "\\begin{tabular}{lll}\n",
       " pattern & sig\\_allele & Pvalue\\\\\n",
       " <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t RKTY & B*81 & 0.04500\\\\\n",
       "\t RKTY & C*18 & 0.00016\\\\\n",
       "\t RKTY & B*18 & 0.04900\\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t SLGLAGG\\%NE & A*74 & 0.021\\\\\n",
       "\t RKTY       & A*31 & 0.021\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 220850 × 3\n",
       "\n",
       "| pattern &lt;chr&gt; | sig_allele &lt;chr&gt; | Pvalue &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| RKTY | B*81 | 0.04500 |\n",
       "| RKTY | C*18 | 0.00016 |\n",
       "| RKTY | B*18 | 0.04900 |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| SLGLAGG%NE | A*74 | 0.021 |\n",
       "| RKTY       | A*31 | 0.021 |\n",
       "\n"
      ],
      "text/plain": [
       "       pattern    sig_allele Pvalue \n",
       "1      RKTY       B*81       0.04500\n",
       "2      RKTY       C*18       0.00016\n",
       "3      RKTY       B*18       0.04900\n",
       "⋮      ⋮          ⋮          ⋮      \n",
       "220849 SLGLAGG%NE A*74       0.021  \n",
       "220850 RKTY       A*31       0.021  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Filter out HLA predictions for high confidence GLIPH groups identified in the\n",
    "# previous step\n",
    "kstme_hc_gliph_hla_table <- kstme_gliph_hla_pred_table |>\n",
    "  dplyr::rename(sig_allele = `'HLA allele with lowest Fisher Score'`) |>\n",
    "  mutate(pattern = str_remove(pattern, \"^[lg]\")) |>\n",
    "  filter(pattern %in% kstme_hc_gliph_cluster_table$pattern & Pvalue <= 0.05) |>\n",
    "  dplyr::select(pattern, sig_allele, Pvalue)\n",
    "kstme_hc_gliph_hla_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 3236411 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pattern</th><th scope=col>vb_score</th><th scope=col>hla_score</th><th scope=col>expansion_score</th><th scope=col>TcRb</th><th scope=col>repertoire_id</th><th scope=col>cohort</th><th scope=col>V</th><th scope=col>J</th><th scope=col>Freq</th><th scope=col>patient_id</th><th scope=col>number_of_unique_cdr</th><th scope=col>number_of_unique_ptid</th><th scope=col>sig_allele</th><th scope=col>Pvalue</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>HHQP</td><td>0.001</td><td>0.023</td><td>1</td><td>CASSALIHFWHHQPQHF</td><td>008_074_D</td><td>Endemic KS - Tumor</td><td>NA</td><td>TRBJ1-5</td><td>1</td><td>008_074</td><td>155</td><td>89</td><td>DPB1*03</td><td>0.023</td></tr>\n",
       "\t<tr><td>HHQP</td><td>0.001</td><td>0.023</td><td>1</td><td>CASSLNLRHHQPQHF  </td><td>008_143_D</td><td>Endemic KS - Tumor</td><td>NA</td><td>TRBJ1-5</td><td>1</td><td>008_143</td><td>155</td><td>89</td><td>DPB1*03</td><td>0.023</td></tr>\n",
       "\t<tr><td>HHQP</td><td>0.001</td><td>0.023</td><td>1</td><td>CASSECLPIFHHQPQHF</td><td>008_189_B</td><td>Endemic KS - Tumor</td><td>NA</td><td>TRBJ1-5</td><td>1</td><td>008_189</td><td>155</td><td>89</td><td>DPB1*03</td><td>0.023</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>RKTY</td><td>0.001</td><td>0.021</td><td>1</td><td>CTSRLRSLCERPMVRKTYGYTF</td><td>008_175_B</td><td>Endemic KS - Tumor</td><td>TRBV1-1</td><td>TRBJ1-2</td><td>1</td><td>008_175</td><td>243</td><td>99</td><td>DRB1*04</td><td>0.037</td></tr>\n",
       "\t<tr><td>RKTY</td><td>0.001</td><td>0.021</td><td>1</td><td>CTSRLRSLCERPMVRKTYGYTF</td><td>008_175_B</td><td>Endemic KS - Tumor</td><td>TRBV1-1</td><td>TRBJ1-2</td><td>1</td><td>008_175</td><td>243</td><td>99</td><td>A*31   </td><td>0.021</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 3236411 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " pattern & vb\\_score & hla\\_score & expansion\\_score & TcRb & repertoire\\_id & cohort & V & J & Freq & patient\\_id & number\\_of\\_unique\\_cdr & number\\_of\\_unique\\_ptid & sig\\_allele & Pvalue\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <int> & <int> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t HHQP & 0.001 & 0.023 & 1 & CASSALIHFWHHQPQHF & 008\\_074\\_D & Endemic KS - Tumor & NA & TRBJ1-5 & 1 & 008\\_074 & 155 & 89 & DPB1*03 & 0.023\\\\\n",
       "\t HHQP & 0.001 & 0.023 & 1 & CASSLNLRHHQPQHF   & 008\\_143\\_D & Endemic KS - Tumor & NA & TRBJ1-5 & 1 & 008\\_143 & 155 & 89 & DPB1*03 & 0.023\\\\\n",
       "\t HHQP & 0.001 & 0.023 & 1 & CASSECLPIFHHQPQHF & 008\\_189\\_B & Endemic KS - Tumor & NA & TRBJ1-5 & 1 & 008\\_189 & 155 & 89 & DPB1*03 & 0.023\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t RKTY & 0.001 & 0.021 & 1 & CTSRLRSLCERPMVRKTYGYTF & 008\\_175\\_B & Endemic KS - Tumor & TRBV1-1 & TRBJ1-2 & 1 & 008\\_175 & 243 & 99 & DRB1*04 & 0.037\\\\\n",
       "\t RKTY & 0.001 & 0.021 & 1 & CTSRLRSLCERPMVRKTYGYTF & 008\\_175\\_B & Endemic KS - Tumor & TRBV1-1 & TRBJ1-2 & 1 & 008\\_175 & 243 & 99 & A*31    & 0.021\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 3236411 × 15\n",
       "\n",
       "| pattern &lt;chr&gt; | vb_score &lt;dbl&gt; | hla_score &lt;dbl&gt; | expansion_score &lt;dbl&gt; | TcRb &lt;chr&gt; | repertoire_id &lt;chr&gt; | cohort &lt;chr&gt; | V &lt;chr&gt; | J &lt;chr&gt; | Freq &lt;dbl&gt; | patient_id &lt;chr&gt; | number_of_unique_cdr &lt;int&gt; | number_of_unique_ptid &lt;int&gt; | sig_allele &lt;chr&gt; | Pvalue &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HHQP | 0.001 | 0.023 | 1 | CASSALIHFWHHQPQHF | 008_074_D | Endemic KS - Tumor | NA | TRBJ1-5 | 1 | 008_074 | 155 | 89 | DPB1*03 | 0.023 |\n",
       "| HHQP | 0.001 | 0.023 | 1 | CASSLNLRHHQPQHF   | 008_143_D | Endemic KS - Tumor | NA | TRBJ1-5 | 1 | 008_143 | 155 | 89 | DPB1*03 | 0.023 |\n",
       "| HHQP | 0.001 | 0.023 | 1 | CASSECLPIFHHQPQHF | 008_189_B | Endemic KS - Tumor | NA | TRBJ1-5 | 1 | 008_189 | 155 | 89 | DPB1*03 | 0.023 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| RKTY | 0.001 | 0.021 | 1 | CTSRLRSLCERPMVRKTYGYTF | 008_175_B | Endemic KS - Tumor | TRBV1-1 | TRBJ1-2 | 1 | 008_175 | 243 | 99 | DRB1*04 | 0.037 |\n",
       "| RKTY | 0.001 | 0.021 | 1 | CTSRLRSLCERPMVRKTYGYTF | 008_175_B | Endemic KS - Tumor | TRBV1-1 | TRBJ1-2 | 1 | 008_175 | 243 | 99 | A*31    | 0.021 |\n",
       "\n"
      ],
      "text/plain": [
       "        pattern vb_score hla_score expansion_score TcRb                  \n",
       "1       HHQP    0.001    0.023     1               CASSALIHFWHHQPQHF     \n",
       "2       HHQP    0.001    0.023     1               CASSLNLRHHQPQHF       \n",
       "3       HHQP    0.001    0.023     1               CASSECLPIFHHQPQHF     \n",
       "⋮       ⋮       ⋮        ⋮         ⋮               ⋮                     \n",
       "3236410 RKTY    0.001    0.021     1               CTSRLRSLCERPMVRKTYGYTF\n",
       "3236411 RKTY    0.001    0.021     1               CTSRLRSLCERPMVRKTYGYTF\n",
       "        repertoire_id cohort             V       J       Freq patient_id\n",
       "1       008_074_D     Endemic KS - Tumor NA      TRBJ1-5 1    008_074   \n",
       "2       008_143_D     Endemic KS - Tumor NA      TRBJ1-5 1    008_143   \n",
       "3       008_189_B     Endemic KS - Tumor NA      TRBJ1-5 1    008_189   \n",
       "⋮       ⋮             ⋮                  ⋮       ⋮       ⋮    ⋮         \n",
       "3236410 008_175_B     Endemic KS - Tumor TRBV1-1 TRBJ1-2 1    008_175   \n",
       "3236411 008_175_B     Endemic KS - Tumor TRBV1-1 TRBJ1-2 1    008_175   \n",
       "        number_of_unique_cdr number_of_unique_ptid sig_allele Pvalue\n",
       "1       155                  89                    DPB1*03    0.023 \n",
       "2       155                  89                    DPB1*03    0.023 \n",
       "3       155                  89                    DPB1*03    0.023 \n",
       "⋮       ⋮                    ⋮                     ⋮          ⋮     \n",
       "3236410 243                  99                    DRB1*04    0.037 \n",
       "3236411 243                  99                    A*31       0.021 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Merge GLIPH clusters with significant HLA association\n",
    "kstme_high_confidence_gliph_groups <- kstme_hc_gliph_cluster_table |>\n",
    "  left_join(kstme_hc_gliph_hla_table, by = \"pattern\",\n",
    "    relationship = \"many-to-many\") \n",
    "kstme_high_confidence_gliph_groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 4776 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>repertoire_id</th><th scope=col>pathology</th><th scope=col>pattern</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001_D</td><td>EBV</td><td>%TGDSNQP</td></tr>\n",
       "\t<tr><td>008_001_D</td><td>EBV</td><td>GT%DSNQP</td></tr>\n",
       "\t<tr><td>008_001_D</td><td>EBV</td><td>GTGD%NQP</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_037_B</td><td>Other</td><td>SL%NTE</td></tr>\n",
       "\t<tr><td>008_037_B</td><td>Other</td><td>SLG%TE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 4776 × 3\n",
       "\\begin{tabular}{lll}\n",
       " repertoire\\_id & pathology & pattern\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_001\\_D & EBV & \\%TGDSNQP\\\\\n",
       "\t 008\\_001\\_D & EBV & GT\\%DSNQP\\\\\n",
       "\t 008\\_001\\_D & EBV & GTGD\\%NQP\\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_037\\_B & Other & SL\\%NTE\\\\\n",
       "\t 008\\_037\\_B & Other & SLG\\%TE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 4776 × 3\n",
       "\n",
       "| repertoire_id &lt;chr&gt; | pathology &lt;chr&gt; | pattern &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 008_001_D | EBV | %TGDSNQP |\n",
       "| 008_001_D | EBV | GT%DSNQP |\n",
       "| 008_001_D | EBV | GTGD%NQP |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| 008_037_B | Other | SL%NTE |\n",
       "| 008_037_B | Other | SLG%TE |\n",
       "\n"
      ],
      "text/plain": [
       "     repertoire_id pathology pattern \n",
       "1    008_001_D     EBV       %TGDSNQP\n",
       "2    008_001_D     EBV       GT%DSNQP\n",
       "3    008_001_D     EBV       GTGD%NQP\n",
       "⋮    ⋮             ⋮         ⋮       \n",
       "4775 008_037_B     Other     SL%NTE  \n",
       "4776 008_037_B     Other     SLG%TE  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Annotate GLIPH groups with antigenic specificity based on the CDR3 sequences\n",
    "# they contain\n",
    "path_annotated_gliph_clusters <- hla_matched_tcr |>\n",
    "  distinct() |>\n",
    "  inner_join(kstme_high_confidence_gliph_groups, by = c(\"repertoire_id\", \"junction_aa\" = \"TcRb\"),\n",
    "    relationship = \"many-to-many\") |>\n",
    "  dplyr::select(repertoire_id, pathology, pattern) |>\n",
    "  distinct()\n",
    "path_annotated_gliph_clusters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    " kstme_ck_gliph_table <- kstme_high_confidence_gliph_groups |>\n",
    "  dplyr::select(repertoire_id, pattern, TcRb, V, J) |>\n",
    "  left_join(path_annotated_gliph_clusters, by = c(\"repertoire_id\", \"pattern\"), relationship = \"many-to-many\") |>\n",
    "  distinct() |>\n",
    "  filter(is.na(pathology)) |>\n",
    "  mutate(pathology = if_else(is.na(pathology), \"Clustered\\nUnknown\", pathology)) |>\n",
    "  group_by(TcRb, V, J) |>\n",
    "  mutate(npath = length(unique(pathology)),\n",
    "    pathology = str_c(unique(pathology), collapse =\";\")) |>\n",
    "  ungroup() |>\n",
    "  mutate(pathology = if_else(npath > 1, \"Multi-pathogen\", pathology)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 11 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pathology</th><th scope=col>count</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>CMV               </td><td>    65</td></tr>\n",
       "\t<tr><td>Clustered\n",
       "Unknown</td><td>151771</td></tr>\n",
       "\t<tr><td>EBV               </td><td>    17</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Other     </td><td>275</td></tr>\n",
       "\t<tr><td>SARS-CoV-2</td><td>  6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 11 × 2\n",
       "\\begin{tabular}{ll}\n",
       " pathology & count\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t CMV                &     65\\\\\n",
       "\t Clustered\n",
       "Unknown & 151771\\\\\n",
       "\t EBV                &     17\\\\\n",
       "\t ⋮ & ⋮\\\\\n",
       "\t Other      & 275\\\\\n",
       "\t SARS-CoV-2 &   6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 11 × 2\n",
       "\n",
       "| pathology &lt;chr&gt; | count &lt;int&gt; |\n",
       "|---|---|\n",
       "| CMV                |     65 |\n",
       "| Clustered\n",
       "Unknown | 151771 |\n",
       "| EBV                |     17 |\n",
       "| ⋮ | ⋮ |\n",
       "| Other      | 275 |\n",
       "| SARS-CoV-2 |   6 |\n",
       "\n"
      ],
      "text/plain": [
       "   pathology          count \n",
       "1  CMV                    65\n",
       "2  Clustered\\nUnknown 151771\n",
       "3  EBV                    17\n",
       "⋮  ⋮                  ⋮     \n",
       "10 Other              275   \n",
       "11 SARS-CoV-2           6   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Annotate GLIPH groups by antigenic specificity of CDR3s goruped within\n",
    "kstme_path_annotated_gliph_groups <- kstme_high_confidence_gliph_groups |>\n",
    "  dplyr::select(repertoire_id, pattern, TcRb, V, J) |>\n",
    "  left_join(path_annotated_gliph_clusters, by = c(\"repertoire_id\", \"pattern\"), relationship = \"many-to-many\") |>\n",
    "  distinct() |>\n",
    "  mutate(pathology = if_else(is.na(pathology), \"Clustered\\nUnknown\", pathology)) |>\n",
    "  group_by(TcRb, V, J) |>\n",
    "  mutate(npath = length(unique(pathology)),\n",
    "    pathology = str_c(unique(pathology), collapse =\";\")) |>\n",
    "  ungroup() |>\n",
    "  mutate(pathology = if_else(npath > 1, \"Multi-pathogen\", pathology)) \n",
    "kstme_path_annotated_gliph_groups |>\n",
    "  group_by(pathology) |> \n",
    "  summarize(count = length(unique(TcRb))) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1064269 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>tissue_type</th><th scope=col>phenotype</th><th scope=col>pathology</th><th scope=col>junction_aa</th><th scope=col>v_call</th><th scope=col>j_call</th><th scope=col>duplicate_frequency</th><th scope=col>npath</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_A</td><td>NAT</td><td>Epidemic KS</td><td>Unclustered\n",
       "Unknown</td><td>YSSAATIGQSRANVLTF</td><td>TRBV17-1</td><td>TRBJ2-6</td><td>0.01190476</td><td>1</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_A</td><td>NAT</td><td>Epidemic KS</td><td>Clustered\n",
       "Unknown  </td><td>CSVTPGLSYEQYF    </td><td>TRBV29-1</td><td>TRBJ2-7</td><td>0.01190476</td><td>1</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_A</td><td>NAT</td><td>Epidemic KS</td><td>Unclustered\n",
       "Unknown</td><td>CSAYKGQTGVNTEAFF </td><td>NA      </td><td>TRBJ1-1</td><td>0.01190476</td><td>1</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_265</td><td>008_265_D</td><td>Tumor</td><td>Epidemic KS</td><td>Unclustered\n",
       "Unknown</td><td>CVALTRGTEAFF</td><td>TRBV2-1</td><td>TRBJ1-1</td><td>0.0004475608</td><td>1</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_265</td><td>008_265_D</td><td>Tumor</td><td>Epidemic KS</td><td>Clustered\n",
       "Unknown  </td><td>CYGALAKNIQYF</td><td>TRBV5-4</td><td>TRBJ2-4</td><td>0.0002983739</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1064269 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " cohort & patient\\_id & repertoire\\_id & tissue\\_type & phenotype & pathology & junction\\_aa & v\\_call & j\\_call & duplicate\\_frequency & npath\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_A & NAT & Epidemic KS & Unclustered\n",
       "Unknown & YSSAATIGQSRANVLTF & TRBV17-1 & TRBJ2-6 & 0.01190476 & 1\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_A & NAT & Epidemic KS & Clustered\n",
       "Unknown   & CSVTPGLSYEQYF     & TRBV29-1 & TRBJ2-7 & 0.01190476 & 1\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_A & NAT & Epidemic KS & Unclustered\n",
       "Unknown & CSAYKGQTGVNTEAFF  & NA       & TRBJ1-1 & 0.01190476 & 1\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t Hippos & 008\\_265 & 008\\_265\\_D & Tumor & Epidemic KS & Unclustered\n",
       "Unknown & CVALTRGTEAFF & TRBV2-1 & TRBJ1-1 & 0.0004475608 & 1\\\\\n",
       "\t Hippos & 008\\_265 & 008\\_265\\_D & Tumor & Epidemic KS & Clustered\n",
       "Unknown   & CYGALAKNIQYF & TRBV5-4 & TRBJ2-4 & 0.0002983739 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1064269 × 11\n",
       "\n",
       "| cohort &lt;chr&gt; | patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | tissue_type &lt;chr&gt; | phenotype &lt;chr&gt; | pathology &lt;chr&gt; | junction_aa &lt;chr&gt; | v_call &lt;chr&gt; | j_call &lt;chr&gt; | duplicate_frequency &lt;dbl&gt; | npath &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Hippos | 008_001 | 008_001_A | NAT | Epidemic KS | Unclustered\n",
       "Unknown | YSSAATIGQSRANVLTF | TRBV17-1 | TRBJ2-6 | 0.01190476 | 1 |\n",
       "| Hippos | 008_001 | 008_001_A | NAT | Epidemic KS | Clustered\n",
       "Unknown   | CSVTPGLSYEQYF     | TRBV29-1 | TRBJ2-7 | 0.01190476 | 1 |\n",
       "| Hippos | 008_001 | 008_001_A | NAT | Epidemic KS | Unclustered\n",
       "Unknown | CSAYKGQTGVNTEAFF  | NA       | TRBJ1-1 | 0.01190476 | 1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| Hippos | 008_265 | 008_265_D | Tumor | Epidemic KS | Unclustered\n",
       "Unknown | CVALTRGTEAFF | TRBV2-1 | TRBJ1-1 | 0.0004475608 | 1 |\n",
       "| Hippos | 008_265 | 008_265_D | Tumor | Epidemic KS | Clustered\n",
       "Unknown   | CYGALAKNIQYF | TRBV5-4 | TRBJ2-4 | 0.0002983739 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "        cohort patient_id repertoire_id tissue_type phenotype  \n",
       "1       Hippos 008_001    008_001_A     NAT         Epidemic KS\n",
       "2       Hippos 008_001    008_001_A     NAT         Epidemic KS\n",
       "3       Hippos 008_001    008_001_A     NAT         Epidemic KS\n",
       "⋮       ⋮      ⋮          ⋮             ⋮           ⋮          \n",
       "1064268 Hippos 008_265    008_265_D     Tumor       Epidemic KS\n",
       "1064269 Hippos 008_265    008_265_D     Tumor       Epidemic KS\n",
       "        pathology            junction_aa       v_call   j_call \n",
       "1       Unclustered\\nUnknown YSSAATIGQSRANVLTF TRBV17-1 TRBJ2-6\n",
       "2       Clustered\\nUnknown   CSVTPGLSYEQYF     TRBV29-1 TRBJ2-7\n",
       "3       Unclustered\\nUnknown CSAYKGQTGVNTEAFF  NA       TRBJ1-1\n",
       "⋮       ⋮                    ⋮                 ⋮        ⋮      \n",
       "1064268 Unclustered\\nUnknown CVALTRGTEAFF      TRBV2-1  TRBJ1-1\n",
       "1064269 Clustered\\nUnknown   CYGALAKNIQYF      TRBV5-4  TRBJ2-4\n",
       "        duplicate_frequency npath\n",
       "1       0.01190476          1    \n",
       "2       0.01190476          1    \n",
       "3       0.01190476          1    \n",
       "⋮       ⋮                   ⋮    \n",
       "1064268 0.0004475608        1    \n",
       "1064269 0.0002983739        1    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kstme_gliph_nprod_table <- annotated_nprod_table |>\n",
    "    mutate(visit_code = if_else(is.na(visit_code), \"V01\", visit_code)) |>\n",
    "    filter(cohort %in% c(\"Hippos\") & visit_code == \"V01\" &\n",
    "        tissue_type %in% c(\"NAT\", \"Tumor\")) |>\n",
    "    left_join(kstme_path_annotated_gliph_groups, by = c(\"repertoire_id\", \n",
    "        \"junction_aa\" = \"TcRb\", \"v_call\" = \"V\",\n",
    "        \"j_call\" = \"J\"), relationship = \"many-to-many\") |>\n",
    "    distinct() |>\n",
    "    dplyr::select(cohort, patient_id, repertoire_id, tissue_type, phenotype, pathology, junction_aa, v_call, j_call, duplicate_frequency) |>\n",
    "    mutate(pathology = if_else(is.na(pathology), \"Unclustered\\nUnknown\", pathology),\n",
    "        phenotype = if_else(is.na(phenotype) , \"Epidemic KS\", phenotype)) |>\n",
    "    group_by(junction_aa, v_call, j_call ) |>\n",
    "    mutate(npath = length(unique(pathology)),\n",
    "        pathology = str_c(unique(pathology), collapse =\";\")) |>\n",
    "    ungroup() |>\n",
    "    mutate(pathology = if_else(npath > 1, \"Multi-pathogen\", pathology)) |>\n",
    "    distinct()\n",
    "kstme_gliph_nprod_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1315 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>repertoire_id</th><th scope=col>pathology</th><th scope=col>phenotype</th><th scope=col>tissue_type</th><th scope=col>frequency</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Epidemic KS - NAT  </td><td>008_001_A</td><td>Clustered\n",
       "Unknown  </td><td>Epidemic KS</td><td>NAT  </td><td>0.2380952</td></tr>\n",
       "\t<tr><td>Epidemic KS - NAT  </td><td>008_001_A</td><td>Unclustered\n",
       "Unknown</td><td>Epidemic KS</td><td>NAT  </td><td>0.7619048</td></tr>\n",
       "\t<tr><td>Epidemic KS - Tumor</td><td>008_001_B</td><td>Clustered\n",
       "Unknown  </td><td>Epidemic KS</td><td>Tumor</td><td>0.2932144</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Epidemic KS - Tumor</td><td>008_265_D</td><td>Multi-pathogen      </td><td>Epidemic KS</td><td>Tumor</td><td>0.002585907</td></tr>\n",
       "\t<tr><td>Epidemic KS - Tumor</td><td>008_265_D</td><td>Unclustered\n",
       "Unknown</td><td>Epidemic KS</td><td>Tumor</td><td>0.718683177</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1315 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " cohort & repertoire\\_id & pathology & phenotype & tissue\\_type & frequency\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Epidemic KS - NAT   & 008\\_001\\_A & Clustered\n",
       "Unknown   & Epidemic KS & NAT   & 0.2380952\\\\\n",
       "\t Epidemic KS - NAT   & 008\\_001\\_A & Unclustered\n",
       "Unknown & Epidemic KS & NAT   & 0.7619048\\\\\n",
       "\t Epidemic KS - Tumor & 008\\_001\\_B & Clustered\n",
       "Unknown   & Epidemic KS & Tumor & 0.2932144\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t Epidemic KS - Tumor & 008\\_265\\_D & Multi-pathogen       & Epidemic KS & Tumor & 0.002585907\\\\\n",
       "\t Epidemic KS - Tumor & 008\\_265\\_D & Unclustered\n",
       "Unknown & Epidemic KS & Tumor & 0.718683177\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1315 × 6\n",
       "\n",
       "| cohort &lt;chr&gt; | repertoire_id &lt;chr&gt; | pathology &lt;chr&gt; | phenotype &lt;chr&gt; | tissue_type &lt;chr&gt; | frequency &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| Epidemic KS - NAT   | 008_001_A | Clustered\n",
       "Unknown   | Epidemic KS | NAT   | 0.2380952 |\n",
       "| Epidemic KS - NAT   | 008_001_A | Unclustered\n",
       "Unknown | Epidemic KS | NAT   | 0.7619048 |\n",
       "| Epidemic KS - Tumor | 008_001_B | Clustered\n",
       "Unknown   | Epidemic KS | Tumor | 0.2932144 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| Epidemic KS - Tumor | 008_265_D | Multi-pathogen       | Epidemic KS | Tumor | 0.002585907 |\n",
       "| Epidemic KS - Tumor | 008_265_D | Unclustered\n",
       "Unknown | Epidemic KS | Tumor | 0.718683177 |\n",
       "\n"
      ],
      "text/plain": [
       "     cohort              repertoire_id pathology            phenotype  \n",
       "1    Epidemic KS - NAT   008_001_A     Clustered\\nUnknown   Epidemic KS\n",
       "2    Epidemic KS - NAT   008_001_A     Unclustered\\nUnknown Epidemic KS\n",
       "3    Epidemic KS - Tumor 008_001_B     Clustered\\nUnknown   Epidemic KS\n",
       "⋮    ⋮                   ⋮             ⋮                    ⋮          \n",
       "1314 Epidemic KS - Tumor 008_265_D     Multi-pathogen       Epidemic KS\n",
       "1315 Epidemic KS - Tumor 008_265_D     Unclustered\\nUnknown Epidemic KS\n",
       "     tissue_type frequency  \n",
       "1    NAT         0.2380952  \n",
       "2    NAT         0.7619048  \n",
       "3    Tumor       0.2932144  \n",
       "⋮    ⋮           ⋮          \n",
       "1314 Tumor       0.002585907\n",
       "1315 Tumor       0.718683177"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "path_annotated_gliph_nprod_table <- kstme_gliph_nprod_table |>\n",
    "    group_by(cohort, repertoire_id, pathology, phenotype, tissue_type) |>\n",
    "    summarize(frequency = sum(duplicate_frequency)) |>\n",
    "    ungroup() |>\n",
    "    mutate(cohort = if_else(cohort == \"Hippos\", \n",
    "            str_c(phenotype, tissue_type, sep = \" - \"), \"ARKS\"))\n",
    "path_annotated_gliph_nprod_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in left_join(arrange(filter(dplyr::select(left_join(filter(mutate(annotated_nprod_table, :\n",
      "“\u001b[1m\u001b[22mDetected an unexpected many-to-many relationship between `x` and `y`.\n",
      "\u001b[36mℹ\u001b[39m Row 3 of `x` matches multiple rows in `y`.\n",
      "\u001b[36mℹ\u001b[39m Row 3399 of `y` matches multiple rows in `x`.\n",
      "\u001b[36mℹ\u001b[39m If a many-to-many relationship is expected, set `relationship = \"many-to-many\"` to silence this warning.”\n"
     ]
    }
   ],
   "source": [
    "annotated_nprod_table |> \n",
    "    mutate(visit_code = if_else(is.na(visit_code), \"V01\", visit_code)) |>\n",
    "    filter(cohort %in% c(\"Hippos\", \"ARKS\") & visit_code == \"V01\" &\n",
    "        tissue_type %in% c(\"NAT\", \"Tumor\")) |>\n",
    "    left_join(hla_matched_tcr, by = c(\"patient_id\", \"repertoire_id\", \"junction\",\n",
    "        \"junction_aa\")) |>\n",
    "    dplyr::select(cohort, patient_id, repertoire_id, tissue_type, phenotype, pathology, junction_aa, duplicate_frequency)|> \n",
    "    filter(pathology %in% c(\"CMV\", \"DENV1\", \"DENV3/4\", \"EBV\", \"HCV\", \"HIV-1\",\n",
    "       \"HSV-2\", \"InfluenzaA\", \"M.tuberculosis\")) |>\n",
    "    arrange(desc(duplicate_frequency)) |> \n",
    "    left_join(antigen_db, by = c(\"pathology\", \"junction_aa\" = \"trb_cdr3_aa\")) |>\n",
    "    dplyr::select(-mhc_allele) |>\n",
    "    distinct() |>\n",
    "    write_csv(str_c(table_path, \"Top_known_sequences.csv\", sep = \"/\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in left_join(arrange(filter(kstme_gliph_nprod_table, pathology %in% :\n",
      "“\u001b[1m\u001b[22mDetected an unexpected many-to-many relationship between `x` and `y`.\n",
      "\u001b[36mℹ\u001b[39m Row 6 of `x` matches multiple rows in `y`.\n",
      "\u001b[36mℹ\u001b[39m Row 3399 of `y` matches multiple rows in `x`.\n",
      "\u001b[36mℹ\u001b[39m If a many-to-many relationship is expected, set `relationship = \"many-to-many\"` to silence this warning.”\n"
     ]
    }
   ],
   "source": [
    "kstme_gliph_nprod_table |> filter(pathology %in% c(\"CMV\", \"DENV1\", \"DENV3/4\", \"EBV\", \"HCV\", \"HIV-1\",\n",
    "       \"HSV-2\", \"InfluenzaA\", \"M.tuberculosis\")) |>\n",
    "    arrange(desc(duplicate_frequency)) |>\n",
    "    left_join(antigen_db, by = c(\"pathology\", \"junction_aa\" = \"trb_cdr3_aa\")) |>\n",
    "    write_csv(str_c(table_path, \"Top_gliph_confirmed_known_sequences.csv\", sep = \"/\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Identify overlapping Clustered unknown GLIPH groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kstme_gliph_tpack_table <- annotated_nprod_table |>\n",
    "    mutate(visit_code = if_else(is.na(visit_code), \"V01\", visit_code)) |>\n",
    "    filter(cohort %in% c(\"Hippos\") & visit_code == \"V01\" &\n",
    "        tissue_type %in% c(\"NAT\", \"Tumor\")) |>\n",
    "    left_join(kstme_path_annotated_gliph_groups, by = c(\"repertoire_id\", \n",
    "        \"junction_aa\" = \"TcRb\", \"v_call\" = \"V\",\n",
    "        \"j_call\" = \"J\"), relationship = \"many-to-many\") |>\n",
    "    distinct() |>\n",
    "    dplyr::select(cohort, patient_id, repertoire_id, tissue_type, phenotype, pathology, \n",
    "        junction_aa, v_call, j_call, duplicate_count, duplicate_frequency, pattern) |>\n",
    "    mutate(pathology = if_else(is.na(pathology), \"Unclustered\\nUnknown\", pathology),\n",
    "        phenotype = if_else(is.na(phenotype) , \"Epidemic KS\", phenotype)) |>\n",
    "    group_by(junction_aa, v_call, j_call ) |>\n",
    "    mutate(npath = length(unique(pathology)),\n",
    "        pathology = str_c(unique(pathology), collapse =\";\")) |>\n",
    "    ungroup() |>\n",
    "    mutate(pathology = if_else(npath > 1, \"Multi-pathogen\", pathology)) |>\n",
    "    distinct()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 492445 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>tissue_type</th><th scope=col>phenotype</th><th scope=col>pathology</th><th scope=col>junction_aa</th><th scope=col>v_call</th><th scope=col>j_call</th><th scope=col>duplicate_count</th><th scope=col>duplicate_frequency</th><th scope=col>pattern</th><th scope=col>npath</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_A</td><td>NAT</td><td>Epidemic KS</td><td>Clustered\n",
       "Unknown</td><td>CSVTPGLSYEQYF   </td><td>TRBV29-1</td><td>TRBJ2-7</td><td>1</td><td>0.01190476</td><td>TPG%SYE   </td><td>1</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_A</td><td>NAT</td><td>Epidemic KS</td><td>Clustered\n",
       "Unknown</td><td>CSVTPGLSYEQYF   </td><td>TRBV29-1</td><td>TRBJ2-7</td><td>1</td><td>0.01190476</td><td>%PGLSYE   </td><td>1</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_001</td><td>008_001_A</td><td>NAT</td><td>Epidemic KS</td><td>Clustered\n",
       "Unknown</td><td>CASSQEGGGRGQPQHF</td><td>TRBV4-1 </td><td>TRBJ1-5</td><td>2</td><td>0.02380952</td><td>SQEGGG%GQP</td><td>1</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_265</td><td>008_265_D</td><td>Tumor</td><td>Epidemic KS</td><td>Clustered\n",
       "Unknown</td><td>CASGSRVHEQYF</td><td>TRBV12-5</td><td>TRBJ2-7</td><td>12</td><td>0.0005967477</td><td>GSRV%E</td><td>1</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_265</td><td>008_265_D</td><td>Tumor</td><td>Epidemic KS</td><td>Clustered\n",
       "Unknown</td><td>CYGALAKNIQYF</td><td>TRBV5-4 </td><td>TRBJ2-4</td><td> 6</td><td>0.0002983739</td><td>%LAKNI</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 492445 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " cohort & patient\\_id & repertoire\\_id & tissue\\_type & phenotype & pathology & junction\\_aa & v\\_call & j\\_call & duplicate\\_count & duplicate\\_frequency & pattern & npath\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_A & NAT & Epidemic KS & Clustered\n",
       "Unknown & CSVTPGLSYEQYF    & TRBV29-1 & TRBJ2-7 & 1 & 0.01190476 & TPG\\%SYE    & 1\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_A & NAT & Epidemic KS & Clustered\n",
       "Unknown & CSVTPGLSYEQYF    & TRBV29-1 & TRBJ2-7 & 1 & 0.01190476 & \\%PGLSYE    & 1\\\\\n",
       "\t Hippos & 008\\_001 & 008\\_001\\_A & NAT & Epidemic KS & Clustered\n",
       "Unknown & CASSQEGGGRGQPQHF & TRBV4-1  & TRBJ1-5 & 2 & 0.02380952 & SQEGGG\\%GQP & 1\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t Hippos & 008\\_265 & 008\\_265\\_D & Tumor & Epidemic KS & Clustered\n",
       "Unknown & CASGSRVHEQYF & TRBV12-5 & TRBJ2-7 & 12 & 0.0005967477 & GSRV\\%E & 1\\\\\n",
       "\t Hippos & 008\\_265 & 008\\_265\\_D & Tumor & Epidemic KS & Clustered\n",
       "Unknown & CYGALAKNIQYF & TRBV5-4  & TRBJ2-4 &  6 & 0.0002983739 & \\%LAKNI & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 492445 × 13\n",
       "\n",
       "| cohort &lt;chr&gt; | patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | tissue_type &lt;chr&gt; | phenotype &lt;chr&gt; | pathology &lt;chr&gt; | junction_aa &lt;chr&gt; | v_call &lt;chr&gt; | j_call &lt;chr&gt; | duplicate_count &lt;dbl&gt; | duplicate_frequency &lt;dbl&gt; | pattern &lt;chr&gt; | npath &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Hippos | 008_001 | 008_001_A | NAT | Epidemic KS | Clustered\n",
       "Unknown | CSVTPGLSYEQYF    | TRBV29-1 | TRBJ2-7 | 1 | 0.01190476 | TPG%SYE    | 1 |\n",
       "| Hippos | 008_001 | 008_001_A | NAT | Epidemic KS | Clustered\n",
       "Unknown | CSVTPGLSYEQYF    | TRBV29-1 | TRBJ2-7 | 1 | 0.01190476 | %PGLSYE    | 1 |\n",
       "| Hippos | 008_001 | 008_001_A | NAT | Epidemic KS | Clustered\n",
       "Unknown | CASSQEGGGRGQPQHF | TRBV4-1  | TRBJ1-5 | 2 | 0.02380952 | SQEGGG%GQP | 1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| Hippos | 008_265 | 008_265_D | Tumor | Epidemic KS | Clustered\n",
       "Unknown | CASGSRVHEQYF | TRBV12-5 | TRBJ2-7 | 12 | 0.0005967477 | GSRV%E | 1 |\n",
       "| Hippos | 008_265 | 008_265_D | Tumor | Epidemic KS | Clustered\n",
       "Unknown | CYGALAKNIQYF | TRBV5-4  | TRBJ2-4 |  6 | 0.0002983739 | %LAKNI | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "       cohort patient_id repertoire_id tissue_type phenotype  \n",
       "1      Hippos 008_001    008_001_A     NAT         Epidemic KS\n",
       "2      Hippos 008_001    008_001_A     NAT         Epidemic KS\n",
       "3      Hippos 008_001    008_001_A     NAT         Epidemic KS\n",
       "⋮      ⋮      ⋮          ⋮             ⋮           ⋮          \n",
       "492444 Hippos 008_265    008_265_D     Tumor       Epidemic KS\n",
       "492445 Hippos 008_265    008_265_D     Tumor       Epidemic KS\n",
       "       pathology          junction_aa      v_call   j_call  duplicate_count\n",
       "1      Clustered\\nUnknown CSVTPGLSYEQYF    TRBV29-1 TRBJ2-7 1              \n",
       "2      Clustered\\nUnknown CSVTPGLSYEQYF    TRBV29-1 TRBJ2-7 1              \n",
       "3      Clustered\\nUnknown CASSQEGGGRGQPQHF TRBV4-1  TRBJ1-5 2              \n",
       "⋮      ⋮                  ⋮                ⋮        ⋮       ⋮              \n",
       "492444 Clustered\\nUnknown CASGSRVHEQYF     TRBV12-5 TRBJ2-7 12             \n",
       "492445 Clustered\\nUnknown CYGALAKNIQYF     TRBV5-4  TRBJ2-4  6             \n",
       "       duplicate_frequency pattern    npath\n",
       "1      0.01190476          TPG%SYE    1    \n",
       "2      0.01190476          %PGLSYE    1    \n",
       "3      0.02380952          SQEGGG%GQP 1    \n",
       "⋮      ⋮                   ⋮          ⋮    \n",
       "492444 0.0005967477        GSRV%E     1    \n",
       "492445 0.0002983739        %LAKNI     1    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kstme_clustered_unknown_table <- kstme_gliph_tpack_table |>\n",
    "    filter(pathology == \"Clustered\\nUnknown\") \n",
    "kstme_clustered_unknown_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 436726 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>repertoire_id</th><th scope=col>pattern</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001_A</td><td>TPG%SYE   </td></tr>\n",
       "\t<tr><td>008_001_A</td><td>%PGLSYE   </td></tr>\n",
       "\t<tr><td>008_001_A</td><td>SQEGGG%GQP</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_265_D</td><td>GSRV%E</td></tr>\n",
       "\t<tr><td>008_265_D</td><td>%LAKNI</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 436726 × 2\n",
       "\\begin{tabular}{ll}\n",
       " repertoire\\_id & pattern\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_001\\_A & TPG\\%SYE   \\\\\n",
       "\t 008\\_001\\_A & \\%PGLSYE   \\\\\n",
       "\t 008\\_001\\_A & SQEGGG\\%GQP\\\\\n",
       "\t ⋮ & ⋮\\\\\n",
       "\t 008\\_265\\_D & GSRV\\%E\\\\\n",
       "\t 008\\_265\\_D & \\%LAKNI\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 436726 × 2\n",
       "\n",
       "| repertoire_id &lt;chr&gt; | pattern &lt;chr&gt; |\n",
       "|---|---|\n",
       "| 008_001_A | TPG%SYE    |\n",
       "| 008_001_A | %PGLSYE    |\n",
       "| 008_001_A | SQEGGG%GQP |\n",
       "| ⋮ | ⋮ |\n",
       "| 008_265_D | GSRV%E |\n",
       "| 008_265_D | %LAKNI |\n",
       "\n"
      ],
      "text/plain": [
       "       repertoire_id pattern   \n",
       "1      008_001_A     TPG%SYE   \n",
       "2      008_001_A     %PGLSYE   \n",
       "3      008_001_A     SQEGGG%GQP\n",
       "⋮      ⋮             ⋮         \n",
       "436725 008_265_D     GSRV%E    \n",
       "436726 008_265_D     %LAKNI    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kstme_gliph_pattern_table <- kstme_clustered_unknown_table |>\n",
    "    dplyr::select(repertoire_id, pattern) |>\n",
    "    filter(!is.na(repertoire_id) & !is.na(pattern)) |>\n",
    "    distinct()\n",
    "kstme_gliph_pattern_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 342 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>repertoire_id</th><th scope=col>gliph_groups</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;list&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001_A</td><td>%PGLSYE,....</td></tr>\n",
       "\t<tr><td>008_001_B</td><td>%AGE, %A....</td></tr>\n",
       "\t<tr><td>008_001_D</td><td>%ADNE, %....</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_264_D</td><td>%ADNE, %....</td></tr>\n",
       "\t<tr><td>008_265_D</td><td>%AGGE, %....</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 342 × 2\n",
       "\\begin{tabular}{ll}\n",
       " repertoire\\_id & gliph\\_groups\\\\\n",
       " <chr> & <list>\\\\\n",
       "\\hline\n",
       "\t 008\\_001\\_A & \\%PGLSYE,....\\\\\n",
       "\t 008\\_001\\_B & \\%AGE, \\%A....\\\\\n",
       "\t 008\\_001\\_D & \\%ADNE, \\%....\\\\\n",
       "\t ⋮ & ⋮\\\\\n",
       "\t 008\\_264\\_D & \\%ADNE, \\%....\\\\\n",
       "\t 008\\_265\\_D & \\%AGGE, \\%....\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 342 × 2\n",
       "\n",
       "| repertoire_id &lt;chr&gt; | gliph_groups &lt;list&gt; |\n",
       "|---|---|\n",
       "| 008_001_A | %PGLSYE,.... |\n",
       "| 008_001_B | %AGE, %A.... |\n",
       "| 008_001_D | %ADNE, %.... |\n",
       "| ⋮ | ⋮ |\n",
       "| 008_264_D | %ADNE, %.... |\n",
       "| 008_265_D | %AGGE, %.... |\n",
       "\n"
      ],
      "text/plain": [
       "    repertoire_id gliph_groups\n",
       "1   008_001_A     %PGLSYE,....\n",
       "2   008_001_B     %AGE, %A....\n",
       "3   008_001_D     %ADNE, %....\n",
       "⋮   ⋮             ⋮           \n",
       "341 008_264_D     %ADNE, %....\n",
       "342 008_265_D     %AGGE, %...."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kstme_gliph_dist_table <- kstme_gliph_pattern_table |>\n",
    "    group_by(repertoire_id) |>\n",
    "    summarize(gliph_groups = list(unique(pattern))) |>\n",
    "    mutate(gliph_groups = map(gliph_groups, ~sort(as.character(.)))) |>\n",
    "    ungroup()\n",
    "kstme_gliph_dist_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "43539"
      ],
      "text/latex": [
       "43539"
      ],
      "text/markdown": [
       "43539"
      ],
      "text/plain": [
       "[1] 43539"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kstme_clustered_unknown_table |> pull(pattern) |> unique() |> length()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the same individual : 26509'"
      ],
      "text/latex": [
       "'Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the same individual : 26509'"
      ],
      "text/markdown": [
       "'Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the same individual : 26509'"
      ],
      "text/plain": [
       "[1] \"Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the same individual : 26509\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "intra_tumor_sharing_count <- kstme_clustered_unknown_table |>\n",
    "    filter(!str_detect(repertoire_id, \"008_\\\\d+_A\")) |>\n",
    "    group_by(pattern, patient_id) |>\n",
    "    summarise(nreps = length(unique(repertoire_id))) |>\n",
    "    filter(nreps > 1) |>\n",
    "    pull(pattern) |>\n",
    "    unique() |>\n",
    "    length()\n",
    "str_c(\"Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the same individual\", intra_tumor_sharing_count, sep = \" : \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the different individuals : 42747'"
      ],
      "text/latex": [
       "'Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the different individuals : 42747'"
      ],
      "text/markdown": [
       "'Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the different individuals : 42747'"
      ],
      "text/plain": [
       "[1] \"Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the different individuals : 42747\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "inter_tumor_sharing_count <- kstme_clustered_unknown_table |>\n",
    "    filter(!str_detect(repertoire_id, \"008_\\\\d+_A\")) |>\n",
    "    group_by(pattern) |>\n",
    "    summarise(nptids = length(unique(patient_id))) |>\n",
    "    filter(nptids > 1) |>\n",
    "    pull(pattern) |>\n",
    "    unique() |>\n",
    "    length()\n",
    "str_c(\"Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the different individuals\", inter_tumor_sharing_count, sep = \" : \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>phenotype</th><th scope=col>ngliphs</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Endemic KS </td><td>26735</td></tr>\n",
       "\t<tr><td>Epidemic KS</td><td>33469</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " phenotype & ngliphs\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t Endemic KS  & 26735\\\\\n",
       "\t Epidemic KS & 33469\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2\n",
       "\n",
       "| phenotype &lt;chr&gt; | ngliphs &lt;int&gt; |\n",
       "|---|---|\n",
       "| Endemic KS  | 26735 |\n",
       "| Epidemic KS | 33469 |\n",
       "\n"
      ],
      "text/plain": [
       "  phenotype   ngliphs\n",
       "1 Endemic KS  26735  \n",
       "2 Epidemic KS 33469  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "inter_by_cond_tumor_sharing_count <- kstme_clustered_unknown_table |>\n",
    "    filter(!str_detect(repertoire_id, \"008_\\\\d+_A\")) |>\n",
    "    group_by(pattern,phenotype) |>\n",
    "    summarise(nptids = length(unique(patient_id))) |>\n",
    "    filter(nptids > 1) |>\n",
    "    ungroup() |>\n",
    "    group_by(phenotype) |>\n",
    "    summarize(ngliphs = length(unique(pattern)))\n",
    "inter_by_cond_tumor_sharing_count "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "'Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the epidemic and endemic individuals : 35731'"
      ],
      "text/latex": [
       "'Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the epidemic and endemic individuals : 35731'"
      ],
      "text/markdown": [
       "'Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the epidemic and endemic individuals : 35731'"
      ],
      "text/plain": [
       "[1] \"Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the epidemic and endemic individuals : 35731\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "inter_cond_tumor_sharing_count <- kstme_clustered_unknown_table |>\n",
    "    filter(!str_detect(repertoire_id, \"008_\\\\d+_A\")) |>\n",
    "    group_by(pattern) |>\n",
    "    summarise(nptids = length(unique(phenotype))) |>\n",
    "    filter(nptids > 1) |>\n",
    "    pull(pattern) |>\n",
    "    unique() |>\n",
    "    length()\n",
    "str_c(\"Number of specificity groups shared by at least two tumors derieved from spatially distinct sites from the epidemic and endemic individuals\", inter_cond_tumor_sharing_count, sep = \" : \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `repertoire_id` -> `repertoire_id...1`\n",
      "\u001b[36m•\u001b[39m `gliph_groups` -> `gliph_groups...2`\n",
      "\u001b[36m•\u001b[39m `repertoire_id` -> `repertoire_id...3`\n",
      "\u001b[36m•\u001b[39m `gliph_groups` -> `gliph_groups...4`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 65734 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>rep_one</th><th scope=col>rep_two</th><th scope=col>jaccard</th><th scope=col>sample_group</th><th scope=col>tumor_group</th><th scope=col>ks_group</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001_A</td><td>008_001_B</td><td>0.014041514</td><td>Intra-sample</td><td>NAT - Tumor</td><td>Epidemic KS - Epidemic KS</td></tr>\n",
       "\t<tr><td>008_001_A</td><td>008_001_D</td><td>0.008120179</td><td>Intra-sample</td><td>NAT - Tumor</td><td>Epidemic KS - Epidemic KS</td></tr>\n",
       "\t<tr><td>008_001_A</td><td>008_002_A</td><td>0.000000000</td><td>Inter-sample</td><td>NAT - NAT  </td><td>Epidemic KS - Epidemic KS</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_265_D</td><td>008_263_D</td><td>0.0375829</td><td>Inter-sample</td><td>Tumor - Tumor</td><td>Epidemic KS - Epidemic KS</td></tr>\n",
       "\t<tr><td>008_265_D</td><td>008_264_D</td><td>0.0606463</td><td>Inter-sample</td><td>Tumor - Tumor</td><td>Endemic KS - Epidemic KS </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 65734 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " rep\\_one & rep\\_two & jaccard & sample\\_group & tumor\\_group & ks\\_group\\\\\n",
       " <chr> & <chr> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_001\\_A & 008\\_001\\_B & 0.014041514 & Intra-sample & NAT - Tumor & Epidemic KS - Epidemic KS\\\\\n",
       "\t 008\\_001\\_A & 008\\_001\\_D & 0.008120179 & Intra-sample & NAT - Tumor & Epidemic KS - Epidemic KS\\\\\n",
       "\t 008\\_001\\_A & 008\\_002\\_A & 0.000000000 & Inter-sample & NAT - NAT   & Epidemic KS - Epidemic KS\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_265\\_D & 008\\_263\\_D & 0.0375829 & Inter-sample & Tumor - Tumor & Epidemic KS - Epidemic KS\\\\\n",
       "\t 008\\_265\\_D & 008\\_264\\_D & 0.0606463 & Inter-sample & Tumor - Tumor & Endemic KS - Epidemic KS \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 65734 × 6\n",
       "\n",
       "| rep_one &lt;chr&gt; | rep_two &lt;chr&gt; | jaccard &lt;dbl&gt; | sample_group &lt;chr&gt; | tumor_group &lt;chr&gt; | ks_group &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 008_001_A | 008_001_B | 0.014041514 | Intra-sample | NAT - Tumor | Epidemic KS - Epidemic KS |\n",
       "| 008_001_A | 008_001_D | 0.008120179 | Intra-sample | NAT - Tumor | Epidemic KS - Epidemic KS |\n",
       "| 008_001_A | 008_002_A | 0.000000000 | Inter-sample | NAT - NAT   | Epidemic KS - Epidemic KS |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_265_D | 008_263_D | 0.0375829 | Inter-sample | Tumor - Tumor | Epidemic KS - Epidemic KS |\n",
       "| 008_265_D | 008_264_D | 0.0606463 | Inter-sample | Tumor - Tumor | Endemic KS - Epidemic KS  |\n",
       "\n"
      ],
      "text/plain": [
       "      rep_one   rep_two   jaccard     sample_group tumor_group  \n",
       "1     008_001_A 008_001_B 0.014041514 Intra-sample NAT - Tumor  \n",
       "2     008_001_A 008_001_D 0.008120179 Intra-sample NAT - Tumor  \n",
       "3     008_001_A 008_002_A 0.000000000 Inter-sample NAT - NAT    \n",
       "⋮     ⋮         ⋮         ⋮           ⋮            ⋮            \n",
       "65733 008_265_D 008_263_D 0.0375829   Inter-sample Tumor - Tumor\n",
       "65734 008_265_D 008_264_D 0.0606463   Inter-sample Tumor - Tumor\n",
       "      ks_group                 \n",
       "1     Epidemic KS - Epidemic KS\n",
       "2     Epidemic KS - Epidemic KS\n",
       "3     Epidemic KS - Epidemic KS\n",
       "⋮     ⋮                        \n",
       "65733 Epidemic KS - Epidemic KS\n",
       "65734 Endemic KS - Epidemic KS "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kstme_gliph_dist_tibble <- expand_grid(kstme_gliph_dist_table, \n",
    "    kstme_gliph_dist_table, .name_repair = \"unique\") |>\n",
    "  dplyr::rename(rep_one = repertoire_id...1,\n",
    "    glist_one = gliph_groups...2,\n",
    "    rep_two = repertoire_id...3,\n",
    "    glist_two = gliph_groups...4) |>\n",
    "  rowwise() |>\n",
    "  mutate(jaccard = length(intersect(glist_one, glist_two))/length(union(glist_one, glist_two)))  |>\n",
    "  dplyr::select(rep_one, rep_two, jaccard) |>\n",
    "  filter(str_detect(rep_one, \"008_\\\\d+\") & str_detect(rep_two, \"008_\\\\d+\") & \n",
    "    !is.na(jaccard) & \n",
    "    str_extract(rep_one, \"[A-Z]+$\") %in% c(\"A\",\"B\", \"C\", \"D\") &\n",
    "    str_extract(rep_two, \"[A-Z]+$\") %in% c(\"A\",\"B\", \"C\", \"D\")) |>\n",
    "  ungroup() |> \n",
    "  mutate(sample_group = if_else(str_extract(rep_one, \"008_\\\\d+\") ==\n",
    "      str_extract(rep_two, \"008_\\\\d+\"), \"Intra-sample\", \"Inter-sample\"),\n",
    "    tumor_group = case_when(\n",
    "      str_extract(rep_one, \"[A-Z]+$\") == \"A\" &\n",
    "        str_extract(rep_two, \"[A-Z]+$\") == \"A\" ~ \"NAT - NAT\",\n",
    "      str_extract(rep_one, \"[A-Z]+$\") == \"A\" &\n",
    "        str_extract(rep_two, \"[A-Z]+$\") %in% c(\"B\", \"C\", \"D\") ~ \"NAT - Tumor\",\n",
    "      str_extract(rep_two, \"[A-Z]+$\") == \"A\" &\n",
    "        str_extract(rep_one, \"[A-Z]+$\") %in% c(\"B\", \"C\", \"D\") ~ \"NAT - Tumor\",\n",
    "      str_extract(rep_two, \"[A-Z]+$\") %in% c(\"B\", \"C\", \"D\") &\n",
    "        str_extract(rep_one, \"[A-Z]+$\") %in% c(\"B\", \"C\", \"D\") ~ \"Tumor - Tumor\"),\n",
    "    ks_group = str_c(ptid_dict[str_extract(rep_one, \"008_\\\\d+\")],\n",
    "      ptid_dict[str_extract(rep_two, \"008_\\\\d+\")], sep = \" - \"),\n",
    "    ks_group = str_replace(ks_group, \"Epidemic KS - Endemic KS\",\n",
    "      \"Endemic KS - Epidemic KS\")) |>\n",
    "  filter(!(sample_group == \"Inter-sample\" & tumor_group == \"NAT - Tumor\") &\n",
    "    !(rep_one == rep_two))\n",
    "kstme_gliph_dist_tibble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 3:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in wilcox.test.default(c(0.1583104587, 0.162004413, 0.1557680306, :\n",
      "“cannot compute exact p-value with ties”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mComputation failed in `stat_signif()`\n",
      "Caused by error in `if (scales$x$map(comp[1]) == data$group[1] | manual) ...`:\n",
      "\u001b[33m!\u001b[39m missing value where TRUE/FALSE needed”\n",
      "Warning message in wilcox.test.default(c(0.1583104587, 0.162004413, 0.1557680306, :\n",
      "“cannot compute exact p-value with ties”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mComputation failed in `stat_signif()`\n",
      "Caused by error in `if (scales$x$map(comp[1]) == data$group[1] | manual) ...`:\n",
      "\u001b[33m!\u001b[39m missing value where TRUE/FALSE needed”\n",
      "Warning message in wilcox.test.default(c(0.1583104587, 0.162004413, 0.1557680306, :\n",
      "“cannot compute exact p-value with ties”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mComputation failed in `stat_signif()`\n",
      "Caused by error in `if (scales$x$map(comp[1]) == data$group[1] | manual) ...`:\n",
      "\u001b[33m!\u001b[39m missing value where TRUE/FALSE needed”\n"
     ]
    }
   ],
   "source": [
    "### Overlapping TCRs section\n",
    "endemic_pal <- alluvial_data |>\n",
    "  filter(str_detect(repertoire_id,  \"008_057|008_061\")) |>\n",
    "  topSeqs(top = 500) |>\n",
    "  cloneTrack() |> \n",
    "  group_by(junction_aa) |>\n",
    "  summarize(reps = str_c(unique(str_extract(repertoire_id, \"[ABCD]$\")), collapse= \"\"),\n",
    "    samples = length(unique(str_extract(repertoire_id, \"\\\\d+_\\\\d+\"))),\n",
    "    seen = max(seen)) |>\n",
    "  mutate(pal = case_when(seen == 1 ~ \"#f0f0f0\",\n",
    "    samples > 1 & reps != \"A\" ~ \"#e7298a\",\n",
    "    samples > 1 & reps == \"A\" ~ \"#1b9e77\",\n",
    "    samples == 1 & seen > 1 & str_detect(reps, \"A\") & \n",
    "      str_detect(reps, \"B|C|D\") ~ \"#d95f02\",\n",
    "    samples == 1 & seen > 1 & !str_detect(reps, \"A\") ~ \"#7570b3\"))\n",
    "endemic_alluvial <- alluvial_data |>\n",
    "  filter(str_detect(repertoire_id, \"008_057|008_061\")) |>\n",
    "  topSeqs(top = 500) |>\n",
    "  cloneTrack() |>\n",
    "  plotTrack(alist = endemic_pal$junction_aa, apal = endemic_pal$pal) +\n",
    "  scale_x_discrete(limits = c(\"008_061_A\", \"008_061_B\", \"008_061_C\", \"008_061_D\",\n",
    "    \"008_057_D\", \"008_057_C\", \"008_057_B\", \"008_057_A\"),\n",
    "    labels = c(\"NAT\", \"Tumor B\", \"Tumor C\", \"Tumor D\", \"Tumor D\", \"Tumor C\",\n",
    "      \"Tumor B\", \"NAT\")) +\n",
    "  annotate(\"text\", label = \"Endemic KS\\nStudy Entry\", y = 0.95, size = 12, x = 2.5) +\n",
    "  annotate(\"text\", label = \"Endemic KS\\nStudy Entry\", y = 0.95, size = 12, x = 6.5) +\n",
    "  geom_vline(xintercept = 4.5, linetype=\"dotted\") + \n",
    "  ylim(0,1) +\n",
    "  theme_classic(base_size = 30) +\n",
    "  theme(legend.position = \"none\")\n",
    "\n",
    "\n",
    "epidemic_pal <- alluvial_data |>\n",
    "  filter(str_detect(repertoire_id, \"008_021|008_048\")) |>\n",
    "  topSeqs(top = 500) |>\n",
    "  cloneTrack() |> \n",
    "  group_by(junction_aa) |>\n",
    "  summarize(reps = str_c(unique(str_extract(repertoire_id, \"[ABCD]$\")), collapse= \"\"),\n",
    "    samples = length(unique(str_extract(repertoire_id, \"\\\\d+_\\\\d+\"))),\n",
    "    seen = max(seen)) |>\n",
    "  mutate(pal = case_when(seen == 1 ~ \"#f0f0f0\",\n",
    "    samples > 1 & reps != \"A\" ~ \"#e7298a\",\n",
    "    samples > 1 & reps == \"A\" ~ \"#1b9e77\",\n",
    "    samples == 1 & seen > 1 & str_detect(reps, \"A\") & \n",
    "      str_detect(reps, \"B|C|D\") ~ \"#d95f02\",\n",
    "    samples == 1 & seen > 1 & !str_detect(reps, \"A\") ~ \"#7570b3\"))\n",
    "\n",
    "epidemic_alluvial <- alluvial_data |>\n",
    "  filter(str_detect(repertoire_id, \"008_021|008_048\")) |>\n",
    "  topSeqs(top = 500) |>\n",
    "  cloneTrack() |>\n",
    "  plotTrack(alist = epidemic_pal$junction_aa, apal = epidemic_pal$pal) +\n",
    "  scale_x_discrete(limits = c(\"008_021_A\", \"008_021_B\", \"008_021_C\", \"008_021_D\",\n",
    "    \"008_048_D\", \"008_048_C\", \"008_048_B\", \"008_048_A\"),\n",
    "    labels = c(\"NAT\", \"Tumor B\", \"Tumor C\", \"Tumor D\", \"Tumor D\", \"Tumor C\",\n",
    "      \"Tumor B\", \"NAT\")) +\n",
    "  geom_vline(xintercept = 4.5, linetype=\"dotted\") + \n",
    "  annotate(\"text\", label = \"Epidemic KS\\nStudy Entry\", y = 0.95, size = 12, x = 2.5) +\n",
    "  annotate(\"text\", label = \"Epidemic KS\\nStudy Entry\", y = 0.95, size = 12, x = 6.5) +\n",
    "  ylim(0,1) +\n",
    "  theme_classic(base_size = 30) +\n",
    "  theme(legend.position = \"none\")\n",
    "\n",
    "endepi_pal <- alluvial_data |>\n",
    "  filter(str_detect(repertoire_id,  \"008_021|008_057\")) |>\n",
    "  topSeqs(top = 500) |>\n",
    "  cloneTrack() |> \n",
    "  group_by(junction_aa) |>\n",
    "  summarize(reps = str_c(unique(str_extract(repertoire_id, \"[ABCD]$\")), collapse= \"\"),\n",
    "    samples = length(unique(str_extract(repertoire_id, \"\\\\d+_\\\\d+\"))),\n",
    "    seen = max(seen)) |>\n",
    "  mutate(pal = case_when(seen == 1 ~ \"#f0f0f0\",\n",
    "    samples > 1 & reps != \"A\" ~ \"#e7298a\",\n",
    "    samples > 1 & reps == \"A\" ~ \"#1b9e77\",\n",
    "    samples == 1 & seen > 1 & str_detect(reps, \"A\") & \n",
    "      str_detect(reps, \"B|C|D\") ~ \"#d95f02\",\n",
    "    samples == 1 & seen > 1 & !str_detect(reps, \"A\") ~ \"#7570b3\"))\n",
    "endepi_alluvial <- alluvial_data |>\n",
    "  filter(str_detect(repertoire_id, \"008_021|008_057\")) |>\n",
    "  topSeqs(top = 500) |>\n",
    "  cloneTrack() |>\n",
    "  plotTrack(alist = endepi_pal$junction_aa, apal = endepi_pal$pal) +\n",
    "  scale_x_discrete(limits = c(\"008_021_A\", \"008_021_B\", \"008_021_C\", \"008_021_D\",\n",
    "    \"008_057_D\", \"008_057_C\", \"008_057_B\", \"008_057_A\"),\n",
    "    labels = c(\"NAT\", \"Tumor B\", \"Tumor C\", \"Tumor D\", \"Tumor D\", \"Tumor C\",\n",
    "      \"Tumor B\", \"NAT\")) +\n",
    "  geom_vline(xintercept = 4.5, linetype=\"dotted\") + \n",
    "  annotate(\"text\", label = \"Epidemic KS\\nStudy Entry\", y = 0.95, size = 12, x = 2.5) +\n",
    "  annotate(\"text\", label = \"Endemic KS\\nStudy Entry\", y = 0.95, size = 12, x = 6.5) +\n",
    "  ylim(0,1) +\n",
    "  theme_classic(base_size = 30) +\n",
    "  theme(legend.position = \"none\")\n",
    "\n",
    "endemic_compairr <- compairr_table |>\n",
    "  mutate(color_group = str_c(sample_group, tumor_group, sep = \"\\n\"),\n",
    "    overlap = if_else(overlap > 1, 1, overlap)) |>\n",
    "  filter(ks_group == \"Endemic KS - Endemic KS\") |>\n",
    "  ggboxplot(x = \"color_group\", y = \"overlap\", fill = \"color_group\",\n",
    "    order = c(\"Inter-subject\\nNAT - NAT\", \"Intra-subject\\nNAT - Tumor\",\n",
    "      \"Intra-subject\\nTumor - Tumor\", \"Inter-subject\\nTumor - Tumor\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Morisita-Horn index\", \n",
    "    ggtheme = theme_classic(base_size = 30), legend = \"none\",\n",
    "    legend.title = \"Group\", palette = c(\"Inter-subject\\nNAT - NAT\" = \"#1b9e77\", \n",
    "      \"Intra-subject\\nNAT - Tumor\" = \"#d95f02\",\n",
    "      \"Intra-subject\\nTumor - Tumor\" = \"#7570b3\", \n",
    "      \"Inter-subject\\nTumor - Tumor\" = \"#e7298a\")) +\n",
    "  stat_compare_means(comparisons = list(c(\"Intra-subject\\nNAT - Tumor\",\n",
    "      \"Intra-subject\\nTumor - Tumor\")), size = 6,  label = \"p.signif\") +\n",
    "  scale_y_continuous(expand = expansion(mult = c(0.05, 0.25)))\n",
    "\n",
    "epidemic_compairr <- compairr_table |>\n",
    "  mutate(color_group = str_c(sample_group, tumor_group, sep = \"\\n\"),\n",
    "    overlap = if_else(overlap > 1, 1, overlap)) |>\n",
    "  filter(ks_group == \"Epidemic KS - Epidemic KS\") |>\n",
    "  ggboxplot(x = \"color_group\", y = \"overlap\", fill = \"color_group\",\n",
    "    order = c(\"Inter-subject\\nNAT - NAT\", \"Intra-subject\\nNAT - Tumor\",\n",
    "      \"Intra-subject\\nTumor - Tumor\", \"Inter-subject\\nTumor - Tumor\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Morisita-Horn index\", \n",
    "    ggtheme = theme_classic(base_size = 30), legend = \"right\",\n",
    "    legend.title = \"Group\", palette = c(\"Inter-subject\\nNAT - NAT\" = \"#1b9e77\", \n",
    "      \"Intra-subject\\nNAT - Tumor\" = \"#d95f02\",\n",
    "      \"Intra-subject\\nTumor - Tumor\" = \"#7570b3\", \n",
    "      \"Inter-subject\\nTumor - Tumor\" = \"#e7298a\")) +\n",
    "  stat_compare_means(comparisons = list(c(\"Intra-subject\\nNAT - Tumor\",\n",
    "      \"Intra-subject\\nTumor - Tumor\")), size = 6,  label = \"p.signif\") +\n",
    "  scale_y_continuous(expand = expansion(mult = c(0.05, 0.25)))\n",
    "\n",
    "endemic_epidemic_compairr <- compairr_table |>\n",
    "  mutate(color_group = str_c(sample_group, tumor_group, sep = \"\\n\"),\n",
    "    overlap = if_else(overlap > 1, 1, overlap)) |>\n",
    "  filter(ks_group == \"Endemic KS - Epidemic KS\") |>\n",
    "  ggboxplot(x = \"color_group\", y = \"overlap\", fill = \"color_group\",\n",
    "    order = c(\"Inter-subject\\nNAT - NAT\", \"Inter-subject\\nTumor - Tumor\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Morisita-Horn index\", ylim = c(0,1),\n",
    "    ggtheme = theme_classic(base_size = 30),\n",
    "    legend = \"none\", palette = c(\"Inter-subject\\nNAT - NAT\" = \"#1b9e77\", \n",
    "      \"Intra-subject\\nNAT - Tumor\" = \"#d95f02\",\n",
    "      \"Intra-subject\\nTumor - Tumor\" = \"#7570b3\", \n",
    "      \"Inter-subject\\nTumor - Tumor\" = \"#e7298a\"), size = 6, label = \"p.signif\")\n",
    "### Public TCRs section\n",
    "public_tcrs_plot <- path_annotated_nprod_table |>\n",
    "  mutate(frequency = frequency * 100) |>\n",
    "  filter(cohort != \"ARKS\") |>\n",
    "  ggboxplot(x = \"pathology\", y = \"frequency\", color = \"cohort\", \n",
    "    add = \"jitter\", order = c(\"CMV\", \"DENV1\", \"DENV3/4\", \"EBV\", \"HCV\", \"HIV-1\",\n",
    "       \"HSV-2\",\n",
    "      \"InfluenzaA\", \"M.tuberculosis\", \"SARS-CoV-2\", \"Other\", \"Multi-pathogen\",\n",
    "      \"Unknown\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Cummulative frequency\", \n",
    "    legend.title = \"Cohort\", \n",
    "    format.scale = T, legend = \"right\",\n",
    "    ggtheme = theme_classic(base_size = 30),\n",
    "    palette = c(\"Epidemic KS - NAT\" = \"#fb8072\", \n",
    "  \"Epidemic KS - Tumor\" = \"#e41a1c\", \"Endemic KS - NAT\" = \"#80b1d3\",\n",
    "  \"Endemic KS - Tumor\" = \"#377eb8\")) +\n",
    "  geom_hline(yintercept = 1, linetype = 2) + \n",
    "  geom_hline(yintercept = 25, linetype = 2) + \n",
    "  geom_hline(yintercept = 75, linetype = 2) + \n",
    "  geom_hline(yintercept = 90, linetype = 2) + \n",
    "  scale_y_log10(breaks = scales::trans_breaks(\"log10\", function(x) 10^x),\n",
    "    labels = scales::trans_format(\"log10\", scales::math_format(10^.x))) +\n",
    "  annotation_logticks(sides = \"l\", short = unit(2,\"mm\"),\n",
    "    mid = unit(4,\"mm\"),long = unit(8,\"mm\"), size = 1) +\n",
    "  theme_classic(base_size = 30) +\n",
    "  theme(axis.text = element_text(face = \"bold\"),\n",
    "    axis.title.x = element_blank())\n",
    "\n",
    "public_gliph_plot <- path_annotated_gliph_nprod_table |>\n",
    "  mutate(frequency = frequency * 100) |>\n",
    "  ggboxplot(x = \"pathology\", y = \"frequency\", color = \"cohort\", \n",
    "    add = \"jitter\", order = c(\"CMV\", \"EBV\", \"HCV\", \"HIV-1\",\n",
    "       \"HSV-2\", \"InfluenzaA\", \"M.tuberculosis\", \"SARS-CoV-2\", \"Other\", \n",
    "       \"Multi-pathogen\",   \"Clustered\\nUnknown\", \"Unclustered\\nUnknown\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Cummulative frequency\", \n",
    "    legend.title = \"Cohort\", \n",
    "    format.scale = T, legend = \"right\",\n",
    "    ggtheme = theme_classic(base_size = 30),\n",
    "    palette = c(\"Epidemic KS - NAT\" = \"#fb8072\", \n",
    "  \"Epidemic KS - Tumor\" = \"#e41a1c\", \"Endemic KS - NAT\" = \"#80b1d3\",\n",
    "  \"Endemic KS - Tumor\" = \"#377eb8\")) +\n",
    "  geom_hline(yintercept = 1, linetype = 2) +\n",
    "  geom_hline(yintercept = 25, linetype = 2) +\n",
    "  geom_hline(yintercept = 75, linetype = 2) +  \n",
    "  geom_hline(yintercept = 90, linetype = 2) +  \n",
    "  scale_y_log10(breaks = scales::trans_breaks(\"log10\", function(x) 10^x),\n",
    "    labels = scales::trans_format(\"log10\", scales::math_format(10^.x))) +\n",
    "  annotation_logticks(sides = \"l\", short = unit(2,\"mm\"),\n",
    "    mid = unit(4,\"mm\"),long = unit(8,\"mm\"), size = 1) + \n",
    "  theme_classic(base_size = 30) +\n",
    "  theme(axis.text = element_text(face = \"bold\"),\n",
    "    axis.title.x = element_blank())\n",
    "\n",
    "### Overlapping clustered unknown section\n",
    "endemic_gliph_compairr <- kstme_gliph_dist_tibble |>\n",
    "  mutate(color_group = str_c(str_replace(sample_group, \"sample\", \"subject\"), tumor_group, sep = \"\\n\"),\n",
    "    jaccard = if_else(jaccard > 1, 1, jaccard)) |>\n",
    "  filter(ks_group == \"Endemic KS - Endemic KS\") |>\n",
    "  ggboxplot(x = \"color_group\", y = \"jaccard\", fill = \"color_group\",\n",
    "    order = c(\"Inter-subject\\nNAT - NAT\", \"Intra-subject\\nNAT - Tumor\",\n",
    "      \"Intra-subject\\nTumor - Tumor\", \"Inter-subject\\nTumor - Tumor\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Jaccard index\",  #title = \"Endemic KS - Endemic KS\",\n",
    "    ggtheme = theme_classic(base_size = 30), legend = \"none\",\n",
    "    legend.title = \"Group\", palette = c(\"Inter-subject\\nNAT - NAT\" = \"#1b9e77\", \n",
    "      \"Intra-subject\\nNAT - Tumor\" = \"#d95f02\",\n",
    "      \"Intra-subject\\nTumor - Tumor\" = \"#7570b3\", \n",
    "      \"Inter-subject\\nTumor - Tumor\" = \"#e7298a\"),\n",
    "    ylim = c(0,1)) +  stat_compare_means(comparisons = list(c(\"Intra-subject\\nNAT - Tumor\",\n",
    "      \"Intra-subject\\nTumor - Tumor\")), size = 6)\n",
    "\n",
    "epidemic_gliph_compairr <- kstme_gliph_dist_tibble |>\n",
    "  mutate(color_group = str_c(str_replace(sample_group, \"sample\", \"subject\"), tumor_group, sep = \"\\n\"),\n",
    "    jaccard = if_else(jaccard > 1, 1, jaccard)) |>\n",
    "  filter(ks_group == \"Epidemic KS - Epidemic KS\") |>\n",
    "  ggboxplot(x = \"color_group\", y = \"jaccard\", fill = \"color_group\",\n",
    "    order = c(\"Inter-subject\\nNAT - NAT\", \"Intra-subject\\nNAT - Tumor\",\n",
    "      \"Intra-subject\\nTumor - Tumor\", \"Inter-subject\\nTumor - Tumor\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Jaccard index\",  #title = \"Epidemic KS - Epidemic KS\",\n",
    "    ggtheme = theme_classic(base_size = 30), legend = \"right\",\n",
    "    legend.title = \"Group\", palette = c(\"Inter-subject\\nNAT - NAT\" = \"#1b9e77\", \n",
    "      \"Intra-subject\\nNAT - Tumor\" = \"#d95f02\",\n",
    "      \"Intra-subject\\nTumor - Tumor\" = \"#7570b3\", \n",
    "      \"Inter-subject\\nTumor - Tumor\" = \"#e7298a\"),\n",
    "    ylim = c(0,1)) +\n",
    "  stat_compare_means(comparisons = list(c(\"Intra-subject\\nNAT - Tumor\",\n",
    "      \"Intra-subject\\nTumor - Tumor\")), size = 6) \n",
    "\n",
    "endemic_epidemic_gliph_compairr <- kstme_gliph_dist_tibble |>\n",
    "  mutate(color_group = str_c(str_replace(sample_group, \"sample\", \"subject\"), tumor_group, sep = \"\\n\"),\n",
    "    jaccard = if_else(jaccard > 1, 1, jaccard)) |>\n",
    "  filter(ks_group == \"Endemic KS - Epidemic KS\") |>\n",
    "  ggboxplot(x = \"color_group\", y = \"jaccard\", fill = \"color_group\",\n",
    "    order = c(\"Inter-subject\\nNAT - NAT\", \"Inter-subject\\nTumor - Tumor\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Jaccard index\", ylim = c(0,1),\n",
    "    ggtheme = theme_classic(base_size = 30), #title = \"Endemic KS - Epidemic KS\",\n",
    "    legend = \"none\", palette = c(\"Inter-sample\\nNAT - NAT\" = \"#1b9e77\", \n",
    "      \"Intra-subject\\nNAT - Tumor\" = \"#d95f02\",\n",
    "      \"Intra-subject\\nTumor - Tumor\" = \"#7570b3\", \n",
    "      \"Inter-subject\\nTumor - Tumor\" = \"#e7298a\")) +\n",
    "  stat_compare_means(comparisons = list(c(\"Intet-subject\\nNAT - NAT\",\n",
    "      \"Intra-subject\\nTumor - Tumor\")), size = 6) \n",
    "\n",
    "\n",
    "figure_three_layout <- \"\n",
    "AAAABBBBCCCC\n",
    "AAAABBBBCCCC\n",
    "AAAABBBBCCCC\n",
    "DDDDEEEEFFFF\n",
    "DDDDEEEEFFFF\n",
    "GGGGGGGGGGGG\n",
    "GGGGGGGGGGGG\n",
    "HHHHHHHHHHHH\n",
    "HHHHHHHHHHHH\n",
    "IIIIJJJJKKKK\n",
    "IIIIJJJJKKKK\n",
    "\"\n",
    "\n",
    "figure_three <- endemic_alluvial + \n",
    "  epidemic_alluvial + \n",
    "  endepi_alluvial +\n",
    "  endemic_compairr +\n",
    "  epidemic_compairr +\n",
    "  endemic_epidemic_compairr + \n",
    "  public_tcrs_plot + \n",
    "  public_gliph_plot +\n",
    "  endemic_gliph_compairr +\n",
    "  epidemic_gliph_compairr +\n",
    "  endemic_epidemic_gliph_compairr + \n",
    "  plot_layout(design = figure_three_layout, guides = \"collect\") +\n",
    "  plot_annotation(tag_levels = \"A\") \n",
    "\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Figure_three.pdf\", sep = \"/\"), figure_three, \n",
    "  width = 48, height = 42, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Figure_three.svg\", sep = \"/\"), figure_three, \n",
    "  width = 48, height = 42)\n",
    "ggsave(str_c(figures_path, \"png\", \"Figure_three.png\", sep = \"/\"), figure_three, \n",
    "  width = 48, height = 42)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary figure 6:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Prep the BL datasets for GLIPH\n",
    "bl_gliph_trb_input_table <- study_annotated_nprod_table |> \n",
    "    filter(cohort %in% c(\"BL - Ghana\", \"BL - Uganda\")) |>\n",
    "    dplyr::select(cohort, repertoire_id, junction_aa, v_call, j_call, duplicate_count) |>\n",
    "    mutate(gliph_id = str_c(repertoire_id, \"BL\", sep = \":\"),\n",
    "        cdra = NA_character_) |>\n",
    "    dplyr::select(junction_aa, v_call, j_call, cdra, gliph_id, duplicate_count) |>\n",
    "    write_tsv(bl_gliph_trb_input_path, col_names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 335 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pattern</th><th scope=col>vb_score</th><th scope=col>hla_score</th><th scope=col>expansion_score</th><th scope=col>TcRb</th><th scope=col>repertoire_id</th><th scope=col>cohort</th><th scope=col>V</th><th scope=col>J</th><th scope=col>Freq</th><th scope=col>patient_id</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SSL%GSNQP</td><td>0.006</td><td>1</td><td>0.14</td><td>CASSSLDGSNQPQHF</td><td>HU10496_0001</td><td>BL</td><td>TRBV6-6 </td><td>TRBJ1-5</td><td> 3</td><td>HU10496_0001</td></tr>\n",
       "\t<tr><td>SSL%GSNQP</td><td>0.006</td><td>1</td><td>0.14</td><td>CATSSLTGSNQPQHF</td><td>HU10496_0001</td><td>BL</td><td>TRBV15-1</td><td>TRBJ1-5</td><td> 9</td><td>HU10496_0001</td></tr>\n",
       "\t<tr><td>SSL%GSNQP</td><td>0.006</td><td>1</td><td>0.14</td><td>CASSSLSGSNQPQHF</td><td>HU10496_0001</td><td>BL</td><td>TRBV6-6 </td><td>TRBJ1-5</td><td>20</td><td>HU10496_0001</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>S%GQAYE</td><td>0.01</td><td>1</td><td>0.21</td><td>CASSRGQAYEQYF</td><td>H003840_02  </td><td>BL</td><td>TRBV11-2</td><td>TRBJ2-7</td><td>1</td><td>H003840_02  </td></tr>\n",
       "\t<tr><td>S%GQAYE</td><td>0.01</td><td>1</td><td>0.21</td><td>CASSRGQAYEQYF</td><td>HU11595_0008</td><td>BL</td><td>TRBV19-1</td><td>TRBJ2-7</td><td>2</td><td>HU11595_0008</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 335 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " pattern & vb\\_score & hla\\_score & expansion\\_score & TcRb & repertoire\\_id & cohort & V & J & Freq & patient\\_id\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t SSL\\%GSNQP & 0.006 & 1 & 0.14 & CASSSLDGSNQPQHF & HU10496\\_0001 & BL & TRBV6-6  & TRBJ1-5 &  3 & HU10496\\_0001\\\\\n",
       "\t SSL\\%GSNQP & 0.006 & 1 & 0.14 & CATSSLTGSNQPQHF & HU10496\\_0001 & BL & TRBV15-1 & TRBJ1-5 &  9 & HU10496\\_0001\\\\\n",
       "\t SSL\\%GSNQP & 0.006 & 1 & 0.14 & CASSSLSGSNQPQHF & HU10496\\_0001 & BL & TRBV6-6  & TRBJ1-5 & 20 & HU10496\\_0001\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t S\\%GQAYE & 0.01 & 1 & 0.21 & CASSRGQAYEQYF & H003840\\_02   & BL & TRBV11-2 & TRBJ2-7 & 1 & H003840\\_02  \\\\\n",
       "\t S\\%GQAYE & 0.01 & 1 & 0.21 & CASSRGQAYEQYF & HU11595\\_0008 & BL & TRBV19-1 & TRBJ2-7 & 2 & HU11595\\_0008\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 335 × 11\n",
       "\n",
       "| pattern &lt;chr&gt; | vb_score &lt;dbl&gt; | hla_score &lt;dbl&gt; | expansion_score &lt;dbl&gt; | TcRb &lt;chr&gt; | repertoire_id &lt;chr&gt; | cohort &lt;chr&gt; | V &lt;chr&gt; | J &lt;chr&gt; | Freq &lt;dbl&gt; | patient_id &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| SSL%GSNQP | 0.006 | 1 | 0.14 | CASSSLDGSNQPQHF | HU10496_0001 | BL | TRBV6-6  | TRBJ1-5 |  3 | HU10496_0001 |\n",
       "| SSL%GSNQP | 0.006 | 1 | 0.14 | CATSSLTGSNQPQHF | HU10496_0001 | BL | TRBV15-1 | TRBJ1-5 |  9 | HU10496_0001 |\n",
       "| SSL%GSNQP | 0.006 | 1 | 0.14 | CASSSLSGSNQPQHF | HU10496_0001 | BL | TRBV6-6  | TRBJ1-5 | 20 | HU10496_0001 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| S%GQAYE | 0.01 | 1 | 0.21 | CASSRGQAYEQYF | H003840_02   | BL | TRBV11-2 | TRBJ2-7 | 1 | H003840_02   |\n",
       "| S%GQAYE | 0.01 | 1 | 0.21 | CASSRGQAYEQYF | HU11595_0008 | BL | TRBV19-1 | TRBJ2-7 | 2 | HU11595_0008 |\n",
       "\n"
      ],
      "text/plain": [
       "    pattern   vb_score hla_score expansion_score TcRb            repertoire_id\n",
       "1   SSL%GSNQP 0.006    1         0.14            CASSSLDGSNQPQHF HU10496_0001 \n",
       "2   SSL%GSNQP 0.006    1         0.14            CATSSLTGSNQPQHF HU10496_0001 \n",
       "3   SSL%GSNQP 0.006    1         0.14            CASSSLSGSNQPQHF HU10496_0001 \n",
       "⋮   ⋮         ⋮        ⋮         ⋮               ⋮               ⋮            \n",
       "334 S%GQAYE   0.01     1         0.21            CASSRGQAYEQYF   H003840_02   \n",
       "335 S%GQAYE   0.01     1         0.21            CASSRGQAYEQYF   HU11595_0008 \n",
       "    cohort V        J       Freq patient_id  \n",
       "1   BL     TRBV6-6  TRBJ1-5  3   HU10496_0001\n",
       "2   BL     TRBV15-1 TRBJ1-5  9   HU10496_0001\n",
       "3   BL     TRBV6-6  TRBJ1-5 20   HU10496_0001\n",
       "⋮   ⋮      ⋮        ⋮       ⋮    ⋮           \n",
       "334 BL     TRBV11-2 TRBJ2-7 1    H003840_02  \n",
       "335 BL     TRBV19-1 TRBJ2-7 2    HU11595_0008"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bl_ebv_groups <- read_csv(bl_gliph_clusters_path, show_col_types = FALSE) |>\n",
    "    dplyr::select(pattern, vb_score, hla_score, expansion_score, TcRb, Sample, V, J,\n",
    "    Freq) |>\n",
    "    separate(Sample, into = c(\"repertoire_id\", \"cohort\"), sep = \":\") |>\n",
    "    mutate(patient_id = repertoire_id) |>\n",
    "    group_by(pattern) |>\n",
    "    mutate(number_of_unique_cdr = length(unique(TcRb)),\n",
    "        number_of_unique_ptid = length(unique(patient_id))) |>\n",
    "    ungroup() |>\n",
    "    filter(vb_score <= 0.01 & number_of_unique_cdr >= 3 &\n",
    "        number_of_unique_ptid >= 3 & pattern != \"single\" & nchar(pattern) > 2) |>\n",
    "    left_join(antigen_db, by = c(\"TcRb\" = \"trb_cdr3_aa\"), relationship = \"many-to-many\") |>\n",
    "    filter(pathology == \"EBV\")\n",
    "    \n",
    "bl_gliph_cluster_table <- read_csv(bl_gliph_clusters_path, show_col_types = FALSE) |>\n",
    "    dplyr::select(pattern, vb_score, hla_score, expansion_score, TcRb, Sample, V, J,\n",
    "    Freq) |>\n",
    "    separate(Sample, into = c(\"repertoire_id\", \"cohort\"), sep = \":\") |>\n",
    "    mutate(patient_id = repertoire_id) |>\n",
    "    filter(pattern %in% bl_ebv_groups$pattern)\n",
    "\n",
    "bl_gliph_cluster_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 362803 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>tissue_type</th><th scope=col>phenotype</th><th scope=col>junction_aa</th><th scope=col>v_call</th><th scope=col>j_call</th><th scope=col>duplicate_count</th><th scope=col>duplicate_frequency</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>BL - Ghana</td><td>H003840_01</td><td>H003840_01</td><td>Tumor</td><td>BL - Ghana</td><td>CYQQFF       </td><td>TRBV21-1</td><td>TRBJ2-1</td><td>3</td><td>2.215445e-05</td></tr>\n",
       "\t<tr><td>BL - Ghana</td><td>H003840_01</td><td>H003840_01</td><td>Tumor</td><td>BL - Ghana</td><td>CASSTDNSYGYTF</td><td>TRBV9-1 </td><td>TRBJ1-2</td><td>1</td><td>7.384815e-06</td></tr>\n",
       "\t<tr><td>BL - Ghana</td><td>H003840_01</td><td>H003840_01</td><td>Tumor</td><td>BL - Ghana</td><td>CASSVEQGDEQYF</td><td>TRBV9-1 </td><td>TRBJ2-7</td><td>1</td><td>7.384815e-06</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>BL - Uganda</td><td>009_0249</td><td>MVQ194745A</td><td>Tumor</td><td>BL - Uganda</td><td>CASNLDSREGGGAEAFF</td><td>TRBV5-5</td><td>TRBJ1-1</td><td>1</td><td>0.0004258944</td></tr>\n",
       "\t<tr><td>BL - Uganda</td><td>009_0249</td><td>MVQ194745A</td><td>Tumor</td><td>BL - Uganda</td><td>CASSLHPVGGTSYEQYF</td><td>TRBV5-5</td><td>TRBJ2-7</td><td>1</td><td>0.0004258944</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 362803 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " cohort & patient\\_id & repertoire\\_id & tissue\\_type & phenotype & junction\\_aa & v\\_call & j\\_call & duplicate\\_count & duplicate\\_frequency\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t BL - Ghana & H003840\\_01 & H003840\\_01 & Tumor & BL - Ghana & CYQQFF        & TRBV21-1 & TRBJ2-1 & 3 & 2.215445e-05\\\\\n",
       "\t BL - Ghana & H003840\\_01 & H003840\\_01 & Tumor & BL - Ghana & CASSTDNSYGYTF & TRBV9-1  & TRBJ1-2 & 1 & 7.384815e-06\\\\\n",
       "\t BL - Ghana & H003840\\_01 & H003840\\_01 & Tumor & BL - Ghana & CASSVEQGDEQYF & TRBV9-1  & TRBJ2-7 & 1 & 7.384815e-06\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t BL - Uganda & 009\\_0249 & MVQ194745A & Tumor & BL - Uganda & CASNLDSREGGGAEAFF & TRBV5-5 & TRBJ1-1 & 1 & 0.0004258944\\\\\n",
       "\t BL - Uganda & 009\\_0249 & MVQ194745A & Tumor & BL - Uganda & CASSLHPVGGTSYEQYF & TRBV5-5 & TRBJ2-7 & 1 & 0.0004258944\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 362803 × 10\n",
       "\n",
       "| cohort &lt;chr&gt; | patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | tissue_type &lt;chr&gt; | phenotype &lt;chr&gt; | junction_aa &lt;chr&gt; | v_call &lt;chr&gt; | j_call &lt;chr&gt; | duplicate_count &lt;dbl&gt; | duplicate_frequency &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| BL - Ghana | H003840_01 | H003840_01 | Tumor | BL - Ghana | CYQQFF        | TRBV21-1 | TRBJ2-1 | 3 | 2.215445e-05 |\n",
       "| BL - Ghana | H003840_01 | H003840_01 | Tumor | BL - Ghana | CASSTDNSYGYTF | TRBV9-1  | TRBJ1-2 | 1 | 7.384815e-06 |\n",
       "| BL - Ghana | H003840_01 | H003840_01 | Tumor | BL - Ghana | CASSVEQGDEQYF | TRBV9-1  | TRBJ2-7 | 1 | 7.384815e-06 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| BL - Uganda | 009_0249 | MVQ194745A | Tumor | BL - Uganda | CASNLDSREGGGAEAFF | TRBV5-5 | TRBJ1-1 | 1 | 0.0004258944 |\n",
       "| BL - Uganda | 009_0249 | MVQ194745A | Tumor | BL - Uganda | CASSLHPVGGTSYEQYF | TRBV5-5 | TRBJ2-7 | 1 | 0.0004258944 |\n",
       "\n"
      ],
      "text/plain": [
       "       cohort      patient_id repertoire_id tissue_type phenotype  \n",
       "1      BL - Ghana  H003840_01 H003840_01    Tumor       BL - Ghana \n",
       "2      BL - Ghana  H003840_01 H003840_01    Tumor       BL - Ghana \n",
       "3      BL - Ghana  H003840_01 H003840_01    Tumor       BL - Ghana \n",
       "⋮      ⋮           ⋮          ⋮             ⋮           ⋮          \n",
       "362802 BL - Uganda 009_0249   MVQ194745A    Tumor       BL - Uganda\n",
       "362803 BL - Uganda 009_0249   MVQ194745A    Tumor       BL - Uganda\n",
       "       junction_aa       v_call   j_call  duplicate_count duplicate_frequency\n",
       "1      CYQQFF            TRBV21-1 TRBJ2-1 3               2.215445e-05       \n",
       "2      CASSTDNSYGYTF     TRBV9-1  TRBJ1-2 1               7.384815e-06       \n",
       "3      CASSVEQGDEQYF     TRBV9-1  TRBJ2-7 1               7.384815e-06       \n",
       "⋮      ⋮                 ⋮        ⋮       ⋮               ⋮                  \n",
       "362802 CASNLDSREGGGAEAFF TRBV5-5  TRBJ1-1 1               0.0004258944       \n",
       "362803 CASSLHPVGGTSYEQYF TRBV5-5  TRBJ2-7 1               0.0004258944       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bl_nprod_table <- study_annotated_nprod_table |>\n",
    "    filter(cohort %in% c(\"BL - Ghana\", \"BL - Uganda\")) |>\n",
    "    dplyr::select(cohort, patient_id, repertoire_id, tissue_type, phenotype, junction_aa, v_call, j_call, duplicate_count, duplicate_frequency) |>\n",
    "    mutate(phenotype = cohort)\n",
    "bl_nprod_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kstme_clustered_unknown_summary_table <- study_annotated_nprod_table |>\n",
    "    filter(cohort == \"Hippos\" & junction_aa %in% kstme_clustered_unknown_table$junction_aa) |>\n",
    "    clonality()\n",
    "kstme_clustered_unknown_summary_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 485 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>repertoire_id</th><th scope=col>total_sequences</th><th scope=col>unique_productive_sequences</th><th scope=col>total_count</th><th scope=col>clonality</th><th scope=col>gini_coefficient</th><th scope=col>top_productive_sequence</th><th scope=col>convergence</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_001_B</td><td> 972</td><td> 972</td><td> 9132</td><td>0.351524</td><td>0.8388728</td><td>15.82348</td><td>1.000000</td></tr>\n",
       "\t<tr><td>008_001_D</td><td>1517</td><td>1517</td><td>10932</td><td>0.336478</td><td>0.8188119</td><td>12.35821</td><td>1.000000</td></tr>\n",
       "\t<tr><td>008_001_I</td><td> 553</td><td> 553</td><td> 2155</td><td>0.168007</td><td>0.6487356</td><td> 5.75406</td><td>1.010101</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_264_V09</td><td>5358</td><td>5358</td><td>6884</td><td>0.028300</td><td>0.2070225</td><td>0.7553748</td><td>1.010101</td></tr>\n",
       "\t<tr><td>008_265_D  </td><td>1013</td><td>1013</td><td>5638</td><td>0.229788</td><td>0.7400670</td><td>5.0549840</td><td>1.020408</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 485 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " repertoire\\_id & total\\_sequences & unique\\_productive\\_sequences & total\\_count & clonality & gini\\_coefficient & top\\_productive\\_sequence & convergence\\\\\n",
       " <chr> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 008\\_001\\_B &  972 &  972 &  9132 & 0.351524 & 0.8388728 & 15.82348 & 1.000000\\\\\n",
       "\t 008\\_001\\_D & 1517 & 1517 & 10932 & 0.336478 & 0.8188119 & 12.35821 & 1.000000\\\\\n",
       "\t 008\\_001\\_I &  553 &  553 &  2155 & 0.168007 & 0.6487356 &  5.75406 & 1.010101\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_264\\_V09 & 5358 & 5358 & 6884 & 0.028300 & 0.2070225 & 0.7553748 & 1.010101\\\\\n",
       "\t 008\\_265\\_D   & 1013 & 1013 & 5638 & 0.229788 & 0.7400670 & 5.0549840 & 1.020408\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 485 × 8\n",
       "\n",
       "| repertoire_id &lt;chr&gt; | total_sequences &lt;int&gt; | unique_productive_sequences &lt;int&gt; | total_count &lt;dbl&gt; | clonality &lt;dbl&gt; | gini_coefficient &lt;dbl&gt; | top_productive_sequence &lt;dbl&gt; | convergence &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 008_001_B |  972 |  972 |  9132 | 0.351524 | 0.8388728 | 15.82348 | 1.000000 |\n",
       "| 008_001_D | 1517 | 1517 | 10932 | 0.336478 | 0.8188119 | 12.35821 | 1.000000 |\n",
       "| 008_001_I |  553 |  553 |  2155 | 0.168007 | 0.6487356 |  5.75406 | 1.010101 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_264_V09 | 5358 | 5358 | 6884 | 0.028300 | 0.2070225 | 0.7553748 | 1.010101 |\n",
       "| 008_265_D   | 1013 | 1013 | 5638 | 0.229788 | 0.7400670 | 5.0549840 | 1.020408 |\n",
       "\n"
      ],
      "text/plain": [
       "    repertoire_id total_sequences unique_productive_sequences total_count\n",
       "1   008_001_B      972             972                         9132      \n",
       "2   008_001_D     1517            1517                        10932      \n",
       "3   008_001_I      553             553                         2155      \n",
       "⋮   ⋮             ⋮               ⋮                           ⋮          \n",
       "484 008_264_V09   5358            5358                        6884       \n",
       "485 008_265_D     1013            1013                        5638       \n",
       "    clonality gini_coefficient top_productive_sequence convergence\n",
       "1   0.351524  0.8388728        15.82348                1.000000   \n",
       "2   0.336478  0.8188119        12.35821                1.000000   \n",
       "3   0.168007  0.6487356         5.75406                1.010101   \n",
       "⋮   ⋮         ⋮                ⋮                       ⋮          \n",
       "484 0.028300  0.2070225        0.7553748               1.010101   \n",
       "485 0.229788  0.7400670        5.0549840               1.020408   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kstme_clustered_unknown_summary_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "    \n",
    "supplementary_figure_six <- kstme_gliph_tpack_table |>\n",
    "    filter(pathology == \"Clustered\\nUnknown\") |>\n",
    "    mutate(duplicate_frequency = duplicate_frequency * 100,\n",
    "        group = if_else(str_detect(repertoire_id, \"008_\\\\d+\"), \n",
    "          str_c(phenotype, tissue_type, sep = \" - \"), phenotype),\n",
    "        group = factor(group, levels = c(\"Epidemic KS - NAT\", \n",
    "          \"Epidemic KS - Tumor\", \"Endemic KS - NAT\",\n",
    "          \"Endemic KS - Tumor\", \"BL- Ghana\", \"BL - Uganda\"))) |>\n",
    "    ggboxplot(x = \"group\", y = \"duplicate_frequency\", color = \"group\") |>\n",
    "    ggpar(xlab = FALSE, ylab = \"Cummulative frequency\", \n",
    "    legend.title = \"Cohort\", \n",
    "    format.scale = T, legend = \"right\",\n",
    "    ggtheme = theme_classic(base_size = 32),\n",
    "    palette = c(\"Epidemic KS - NAT\" = \"#fb8072\", \n",
    "  \"Epidemic KS - Tumor\" = \"#e41a1c\", \"Endemic KS - NAT\" = \"#80b1d3\",\n",
    "  \"Endemic KS - Tumor\" = \"#377eb8\")) +\n",
    "  geom_hline(yintercept = 1, linetype = 2) +\n",
    "  geom_hline(yintercept = 10, linetype = 2) +\n",
    "  geom_hline(yintercept = 75, linetype = 2) +  \n",
    "  scale_y_log10(breaks = scales::trans_breaks(\"log10\", function(x) 10^x),\n",
    "    labels = scales::trans_format(\"log10\", scales::math_format(10^.x))) +\n",
    "  annotation_logticks(sides = \"l\", short = unit(2,\"mm\"),\n",
    "    mid = unit(4,\"mm\"),long = unit(8,\"mm\"), size = 1) +\n",
    "  theme(axis.text =  element_text(face = \"bold\"))\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Supplementary_figure_six.pdf\", sep = \"/\"), supplementary_figure_six, \n",
    "  width = 24, height = 10)\n",
    "ggsave(str_c(figures_path, \"svg\", \"Supplementary_figure_six.svg\", sep = \"/\"), supplementary_figure_six, \n",
    "  width = 24, height = 10)\n",
    "ggsave(str_c(figures_path, \"png\", \"Supplementary_figure_six.png\", sep = \"/\"), supplementary_figure_six, \n",
    "  width = 24, height = 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary figure 7:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 41921 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>repertoire_id</th><th scope=col>junction_aa</th><th scope=col>v_call</th><th scope=col>d_call</th><th scope=col>j_call</th><th scope=col>v_family</th><th scope=col>d_family</th><th scope=col>j_family</th><th scope=col>reading_frame</th><th scope=col>duplicate_count</th><th scope=col>duplicate_frequency</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_057_A</td><td>CAGGGRGAEDTQYF </td><td>TRBV7-8</td><td>TRBD2-1</td><td>TRBJ2-3</td><td>TRBV7</td><td>TRBD2</td><td>TRBJ2</td><td>in-frame</td><td>1</td><td>0.0007429421</td></tr>\n",
       "\t<tr><td>008_057_A</td><td>CAGRREETQYF    </td><td>TRBV2-1</td><td>TRBD1-1</td><td>TRBJ2-5</td><td>TRBV2</td><td>TRBD1</td><td>TRBJ2</td><td>in-frame</td><td>1</td><td>0.0007429421</td></tr>\n",
       "\t<tr><td>008_057_A</td><td>CAGSDRGRVDYGYTF</td><td>TRBV2-1</td><td>TRBD1-1</td><td>TRBJ1-2</td><td>TRBV2</td><td>TRBD1</td><td>TRBJ1</td><td>in-frame</td><td>1</td><td>0.0007429421</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_061_D</td><td>TLESCTGELFF      </td><td>TRBV7-5</td><td>TRBD2-1</td><td>TRBJ2-2</td><td>TRBV7</td><td>TRBD2</td><td>TRBJ2</td><td>in-frame</td><td>1</td><td>6.928566e-05</td></tr>\n",
       "\t<tr><td>008_061_D</td><td>WPAAATIPFHSDNEQFF</td><td>NA     </td><td>TRBD2-1</td><td>TRBJ2-1</td><td>NA   </td><td>TRBD2</td><td>TRBJ2</td><td>in-frame</td><td>1</td><td>6.928566e-05</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 41921 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " repertoire\\_id & junction\\_aa & v\\_call & d\\_call & j\\_call & v\\_family & d\\_family & j\\_family & reading\\_frame & duplicate\\_count & duplicate\\_frequency\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 008\\_057\\_A & CAGGGRGAEDTQYF  & TRBV7-8 & TRBD2-1 & TRBJ2-3 & TRBV7 & TRBD2 & TRBJ2 & in-frame & 1 & 0.0007429421\\\\\n",
       "\t 008\\_057\\_A & CAGRREETQYF     & TRBV2-1 & TRBD1-1 & TRBJ2-5 & TRBV2 & TRBD1 & TRBJ2 & in-frame & 1 & 0.0007429421\\\\\n",
       "\t 008\\_057\\_A & CAGSDRGRVDYGYTF & TRBV2-1 & TRBD1-1 & TRBJ1-2 & TRBV2 & TRBD1 & TRBJ1 & in-frame & 1 & 0.0007429421\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_061\\_D & TLESCTGELFF       & TRBV7-5 & TRBD2-1 & TRBJ2-2 & TRBV7 & TRBD2 & TRBJ2 & in-frame & 1 & 6.928566e-05\\\\\n",
       "\t 008\\_061\\_D & WPAAATIPFHSDNEQFF & NA      & TRBD2-1 & TRBJ2-1 & NA    & TRBD2 & TRBJ2 & in-frame & 1 & 6.928566e-05\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 41921 × 11\n",
       "\n",
       "| repertoire_id &lt;chr&gt; | junction_aa &lt;chr&gt; | v_call &lt;chr&gt; | d_call &lt;chr&gt; | j_call &lt;chr&gt; | v_family &lt;chr&gt; | d_family &lt;chr&gt; | j_family &lt;chr&gt; | reading_frame &lt;chr&gt; | duplicate_count &lt;dbl&gt; | duplicate_frequency &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 008_057_A | CAGGGRGAEDTQYF  | TRBV7-8 | TRBD2-1 | TRBJ2-3 | TRBV7 | TRBD2 | TRBJ2 | in-frame | 1 | 0.0007429421 |\n",
       "| 008_057_A | CAGRREETQYF     | TRBV2-1 | TRBD1-1 | TRBJ2-5 | TRBV2 | TRBD1 | TRBJ2 | in-frame | 1 | 0.0007429421 |\n",
       "| 008_057_A | CAGSDRGRVDYGYTF | TRBV2-1 | TRBD1-1 | TRBJ1-2 | TRBV2 | TRBD1 | TRBJ1 | in-frame | 1 | 0.0007429421 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_061_D | TLESCTGELFF       | TRBV7-5 | TRBD2-1 | TRBJ2-2 | TRBV7 | TRBD2 | TRBJ2 | in-frame | 1 | 6.928566e-05 |\n",
       "| 008_061_D | WPAAATIPFHSDNEQFF | NA      | TRBD2-1 | TRBJ2-1 | NA    | TRBD2 | TRBJ2 | in-frame | 1 | 6.928566e-05 |\n",
       "\n"
      ],
      "text/plain": [
       "      repertoire_id junction_aa       v_call  d_call  j_call  v_family d_family\n",
       "1     008_057_A     CAGGGRGAEDTQYF    TRBV7-8 TRBD2-1 TRBJ2-3 TRBV7    TRBD2   \n",
       "2     008_057_A     CAGRREETQYF       TRBV2-1 TRBD1-1 TRBJ2-5 TRBV2    TRBD1   \n",
       "3     008_057_A     CAGSDRGRVDYGYTF   TRBV2-1 TRBD1-1 TRBJ1-2 TRBV2    TRBD1   \n",
       "⋮     ⋮             ⋮                 ⋮       ⋮       ⋮       ⋮        ⋮       \n",
       "41920 008_061_D     TLESCTGELFF       TRBV7-5 TRBD2-1 TRBJ2-2 TRBV7    TRBD2   \n",
       "41921 008_061_D     WPAAATIPFHSDNEQFF NA      TRBD2-1 TRBJ2-1 NA       TRBD2   \n",
       "      j_family reading_frame duplicate_count duplicate_frequency\n",
       "1     TRBJ2    in-frame      1               0.0007429421       \n",
       "2     TRBJ2    in-frame      1               0.0007429421       \n",
       "3     TRBJ1    in-frame      1               0.0007429421       \n",
       "⋮     ⋮        ⋮             ⋮               ⋮                  \n",
       "41920 TRBJ2    in-frame      1               6.928566e-05       \n",
       "41921 TRBJ2    in-frame      1               6.928566e-05       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "endemic_upset_input_table <- alluvial_data |>\n",
    "  filter(str_detect(repertoire_id,  \"008_057|008_061\"))\n",
    "endemic_upset_input_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upset_query(intersect = c('008_061_B','008_061_C','008_061_D','008_057_B','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_061_D','008_057_B','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_D','008_057_B','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_057_B','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_061_D','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_061_D','008_057_B','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_061_D','008_057_B','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_D','008_057_B','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_057_B','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_061_D','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_061_D','008_057_B','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_061_D','008_057_B','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_057_B','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_D','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_D','008_057_B','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_D','008_057_B','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_057_B','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_057_B','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_061_D','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_061_D','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_061_D','008_057_B') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_057_B','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_D','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_D','008_057_B','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_D','008_057_B','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_057_B','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_057_B','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_061_D','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_061_D','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_061_D','008_057_B') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_057_B','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_057_B','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_D','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_D','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_D','008_057_B') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_057_B') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C','008_061_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_057_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_057_B','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_057_B','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_D','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_D','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_D','008_057_B') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_057_B') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C','008_061_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_057_B') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B','008_061_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_057_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_057_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_057_B') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_D') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_C') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
      "upset_query(intersect = c('008_061_B') color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')), \n"
     ]
    }
   ],
   "source": [
    "# Define your list\n",
    "my_list <- c(\"008_061_B\", \"008_061_C\", \"008_061_D\", \"008_057_B\", \"008_057_C\", \"008_057_D\")\n",
    "\n",
    "# Function to generate all subsets\n",
    "all_subsets <- function(lst) {\n",
    "  all_combinations <- list()\n",
    "  for (i in 1:length(lst)) {\n",
    "    all_combinations <- c(all_combinations, combn(lst, i, simplify = FALSE))\n",
    "  }\n",
    "  return(all_combinations)\n",
    "}\n",
    "\n",
    "# Get all subsets\n",
    "subsets <- all_subsets(my_list)\n",
    "\n",
    "formatter_list <- c()\n",
    "# Print the formatted subsets\n",
    "for (formatted_subset in subsets) {\n",
    "  #print(str_c(\"upset_query(intersect = c(\", str_c(\"'\", formatted_subset, \"'\", collapse = \",\"), \") color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')\", sep = \"\"))\n",
    "  formatter_list <- c(str_c(\"upset_query(intersect = c(\", str_c(\"'\", formatted_subset, \"'\", collapse = \",\"), \") color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\", sep = \"\"), formatter_list)\n",
    "}\n",
    "collapsed_string <- paste(formatter_list, collapse = \"\\n\")\n",
    "cat(collapsed_string, '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 411,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# UpSetR plots for endemic KS comparisons\n",
    "\n",
    "endemic_upset_input_tibble <- alluvial_data |>\n",
    "    filter(str_detect(repertoire_id,  \"008_057|008_061\"))  |>\n",
    "    left_join(study_metadata, by = \"repertoire_id\") |>\n",
    "    dplyr::select(repertoire_id, junction_aa, tissue_type, phenotype) |>\n",
    "    mutate(present = 1,\n",
    "        row_id = str_c(repertoire_id, tissue_type, phenotype, sep = \" \")) |>\n",
    "    pivot_wider(id_cols = junction_aa, names_from = repertoire_id, values_from = present, values_fill = 0) |>\n",
    "    dplyr::select(junction_aa, \"008_057_A\", \"008_057_B\", \"008_057_C\", \"008_057_D\", \"008_061_D\",\n",
    "        \"008_061_C\", \"008_061_B\", \"008_061_A\")\n",
    "\n",
    "endemic_upset_rownames <- endemic_upset_input_tibble |>\n",
    "    pull(junction_aa)\n",
    "endemic_upset_input_table <- endemic_upset_input_tibble |>\n",
    "    dplyr::select(-junction_aa) |>\n",
    "    as.data.frame() \n",
    "rownames(endemic_upset_input_table) <- endemic_upset_rownames\n",
    "\n",
    "endemic_upset <- upset(endemic_upset_input_table, \n",
    "    intersect = c(\"008_057_A\", \"008_057_B\", \"008_057_C\", \"008_057_D\", \"008_061_D\", \"008_061_C\", \"008_061_B\", \"008_061_A\"),\n",
    "    queries = list(\n",
    "    upset_query(set = '008_061_A', fill = '#80b1d3'),\n",
    "    upset_query(set = '008_061_B', fill = '#377eb8'),\n",
    "    upset_query(set = '008_061_C', fill = '#377eb8'),\n",
    "    upset_query(set = '008_061_D', fill = '#377eb8'),\n",
    "    upset_query(set = '008_057_A', fill = '#80b1d3'),\n",
    "    upset_query(set = '008_057_B', fill = '#377eb8'),\n",
    "    upset_query(set = '008_057_C', fill = '#377eb8'),\n",
    "    upset_query(set = '008_057_D', fill = '#377eb8'),\n",
    "    upset_query(intersect = c('008_061_A'), color = '#80b1d3', fill = '#80b1d3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_A'), color = '#80b1d3', fill = '#80b1d3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B'), color = '#377eb8', fill = '#377eb8', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_C'), color = '#377eb8', fill = '#377eb8', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_D'), color = '#377eb8', fill = '#377eb8', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_B'), color = '#377eb8', fill = '#377eb8', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_C'), color = '#377eb8', fill = '#377eb8', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_D'), color = '#377eb8', fill = '#377eb8', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_061_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_B','008_057_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_B','008_057_C', '008_057_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),    \n",
    "    upset_query(intersect = c('008_061_B','008_061_C', '008_061_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),    \n",
    "    upset_query(intersect = c('008_057_B','008_057_C'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_C','008_057_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_061_C'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_C','008_061_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_A','008_057_B', '008_057_C', '008_057_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_B', '008_061_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_B'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_B', '008_061_C', '008_061_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_A','008_057_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_A','008_057_B'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_A','008_057_B', '008_057_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_A','008_057_C'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_A','008_057_C', '008_057_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_B', '008_061_C'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_C'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_057_A','008_057_B', '008_057_C'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_D','008_057_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_057_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_057_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_C', '008_061_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_D','008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_057_D', '008_057_C', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_057_D'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_D','008_057_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_D','008_057_D', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_057_D', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_061_C', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_061_C', '008_057_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_C','008_057_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_061_C', '008_061_D', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_061_D', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_C','008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_057_B'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_057_D','008_057_C','008_057_B'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_061_D', '008_057_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_061_C', '008_061_D', '008_057_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_C','008_061_D', '008_057_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),    \n",
    "    upset_query(intersect = c('008_057_C','008_061_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_057_D','008_057_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_B', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_C','008_061_D', '008_057_D', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_D', '008_057_D', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_C','008_057_D', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_C','008_061_D', '008_057_C', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_B', '008_061_D', '008_057_C', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_057_C', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_D','008_057_D', '008_057_C', '008_057_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_061_D','008_057_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_057_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_057_A'), color = '#1b9e77', fill = '#1b9e77', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_057_D','008_057_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_B','008_061_C','008_061_D', '008_057_C', '008_057_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_D','008_057_B','008_057_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_057_B','008_057_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_057_D','008_057_B','008_057_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_D','008_057_D','008_057_C','008_057_B','008_057_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_C','008_057_D','008_057_C','008_057_B','008_057_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "    upset_query(intersect = c('008_061_A','008_057_D','008_057_C','008_057_B','008_057_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size'))),\n",
    "    intersections = list('008_061_B', '008_057_B','008_057_D', '008_061_D',\n",
    "        '008_057_C', '008_061_A', '008_061_C', '008_057_A', \n",
    "        c('008_061_B','008_061_D'),\n",
    "        c('008_061_B','008_061_C','008_061_D'),\n",
    "        c('008_061_B','008_061_C'),\n",
    "        c('008_061_D','008_061_C'),\n",
    "        c('008_057_D','008_057_B'),\n",
    "        c('008_057_D','008_057_C','008_057_B'), \n",
    "        c('008_057_C','008_057_B'),\n",
    "        c('008_057_D','008_057_C'),\n",
    "        c('008_061_A','008_061_B','008_061_D'),\n",
    "        c('008_061_A','008_061_B'),\n",
    "        c('008_061_A','008_061_B','008_061_C','008_061_D'),\n",
    "        c('008_061_A','008_061_D'),\n",
    "        c('008_061_A','008_061_B','008_061_C'),\n",
    "        c('008_061_A','008_061_C'),\n",
    "        c('008_061_A','008_061_C','008_061_D'),\n",
    "        c('008_057_D','008_057_C','008_057_B','008_057_A'),\n",
    "        c('008_057_D','008_057_A'),\n",
    "        c('008_057_B','008_057_A'),\n",
    "        c('008_057_D','008_057_B','008_057_A'),\n",
    "        c('008_057_C','008_057_A'),\n",
    "        c('008_057_D','008_057_C','008_057_A'),\n",
    "        c('008_057_C','008_057_B','008_057_A'),\n",
    "        c('008_057_B','008_061_B'),\n",
    "        c('008_057_D','008_061_D'),\n",
    "        c('008_061_B','008_057_C'),\n",
    "        c('008_061_B','008_057_D'),\n",
    "        c('008_061_D','008_057_B'),\n",
    "        c('008_061_B','008_057_D','008_057_C','008_057_B'),\n",
    "        c('008_061_D','008_057_C'),\n",
    "        c('008_061_D','008_057_D','008_057_B'),\n",
    "        c('008_061_B','008_057_D','008_057_B'),\n",
    "        c('008_061_B','008_061_C','008_057_B'),\n",
    "        c('008_061_B','008_061_C','008_057_C'),\n",
    "        c('008_061_C','008_057_C'),\n",
    "        c('008_061_B','008_061_C','008_061_D','008_057_B'),\n",
    "        c('008_061_B','008_061_D','008_057_B'),\n",
    "        c('008_061_C','008_057_B'),\n",
    "        c('008_061_B','008_061_D','008_057_D'),\n",
    "        c('008_061_B','008_061_C','008_061_D','008_057_C'),\n",
    "        c('008_061_C','008_061_D','008_057_C'),\n",
    "        c('008_061_B','008_057_D','008_057_C'),\n",
    "        c('008_061_A','008_061_B','008_057_B'),\n",
    "        c('008_061_C','008_061_D','008_057_D','008_057_B'),\n",
    "        c('008_061_A','008_061_D','008_057_D','008_057_B'),\n",
    "        c('008_061_C','008_057_D','008_057_B'),\n",
    "        c('008_061_C','008_061_D','008_057_C','008_057_B'),\n",
    "        c('008_061_A','008_061_B','008_061_D','008_057_C','008_057_B'),\n",
    "        c('008_061_B','008_057_C','008_057_B'),\n",
    "        c('008_061_D','008_057_D','008_057_C','008_057_B'),\n",
    "        c('008_061_B','008_057_D','008_057_A'),\n",
    "        c('008_061_B','008_061_C','008_061_D','008_057_C','008_057_A'),\n",
    "        c('008_061_D','008_057_B','008_057_A'),\n",
    "        c('008_061_D','008_057_D','008_057_C','008_057_B','008_057_A'),\n",
    "        c('008_061_C','008_057_D','008_057_C','008_057_B','008_057_A'),\n",
    "        c('008_061_A','008_057_A'),\n",
    "        c('008_061_A','008_057_D'),    \n",
    "        c('008_061_A','008_057_B'),\n",
    "        c('008_061_A','008_057_D','008_057_C','008_057_B'),\n",
    "        c('008_061_A','008_057_C'),\n",
    "        c('008_061_A','008_061_D','008_057_A'),\n",
    "        c('008_061_B','008_057_A'),\n",
    "        c('008_061_A','008_057_B','008_057_A'),\n",
    "        c('008_061_A','008_057_D','008_057_B','008_057_A'),\n",
    "        c('008_061_A','008_057_D','008_057_C','008_057_B','008_057_A')),\n",
    "    sort_sets=FALSE,\n",
    "    sort_intersections=FALSE,\n",
    "    width_ratio=0.1,\n",
    "    name = \"Overlap between TCR repertoires\",\n",
    "    base_annotations = list(\n",
    "            'Intersection size'=intersection_size(\n",
    "                text = list(size = 7, angle = 90, hjust = 0, vjust = 0.5)\n",
    "            ) + ylab('Number of overlapping TCRs') \n",
    "            + ggtitle(\"Endemic KS - Endemic KS\")\n",
    "            + theme_classic() \n",
    "            + labs(tag = \"A\")\n",
    "            + theme(axis.title.x = element_blank(),\n",
    "                plot.title = element_text(size = 30, face = \"bold\", family = \"NimbusSan\", hjust = 0.5),\n",
    "                axis.text.x = element_blank(),\n",
    "                axis.ticks.x = element_blank(),\n",
    "                axis.title.y = element_text(size = 24, face = \"bold\", family = \"NimbusSan\"),\n",
    "                axis.text.y = element_text(size = 22, face = \"bold\", family = \"NimbusSan\"),\n",
    "                plot.tag = element_text(size = 28, face = \"bold\", family = \"NimbusSan\"))),\n",
    "    set_sizes=(upset_set_size()+ ylab('Number of\\nunique TCRs') \n",
    "            + theme_void() \n",
    "            + theme(axis.title.y = element_blank(),\n",
    "                axis.text.y = element_blank(),\n",
    "                axis.ticks.y = element_blank(),\n",
    "                axis.title.x = element_blank(),\n",
    "                axis.text.x = element_text(size = 20, face = \"bold\", family = \"NimbusSan\", angle = 90, vjust = 0.5),\n",
    "                axis.ticks.x = element_line(linewidth = 2, colour = 'black' ))\n",
    "    )) +\n",
    "    theme(axis.title.x = element_blank(),\n",
    "        axis.text.y = element_text(size = 24, face = \"bold\", family = \"NimbusSan\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 412,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# UpSetR plots for epidemic KS comparisons\n",
    "epidemic_upset_input_tibble <- alluvial_data |>\n",
    "    filter(str_detect(repertoire_id, \"008_021|008_048\"))  |>\n",
    "    left_join(study_metadata, by = \"repertoire_id\") |>\n",
    "    dplyr::select(repertoire_id, junction_aa, tissue_type, phenotype) |>\n",
    "    mutate(present = 1,\n",
    "        row_id = str_c(repertoire_id, tissue_type, phenotype, sep = \" \")) |>\n",
    "    pivot_wider(id_cols = junction_aa, names_from = repertoire_id, values_from = present, values_fill = 0)\n",
    "\n",
    "epidemic_upset_rownames <- epidemic_upset_input_tibble |>\n",
    "    pull(junction_aa)\n",
    "epidemic_upset_input_table <- epidemic_upset_input_tibble |>\n",
    "    dplyr::select(-junction_aa) |>\n",
    "    as.data.frame() \n",
    "rownames(epidemic_upset_input_table) <- epidemic_upset_rownames\n",
    "\n",
    "epidemic_upset <- upset(epidemic_upset_input_tibble, \n",
    "    intersect = c(\"008_021_A\", \"008_021_B\", \"008_021_C\", \"008_021_D\", \"008_048_D\", \"008_048_C\", \"008_048_B\", \"008_048_A\"),\n",
    "    queries = list(\n",
    "        upset_query(set = '008_021_A', fill = '#fb8072'),\n",
    "        upset_query(set = '008_021_B', fill = '#e41a1c'),\n",
    "        upset_query(set = '008_021_C', fill = '#e41a1c'),\n",
    "        upset_query(set = '008_021_D', fill = '#e41a1c'),\n",
    "        upset_query(set = '008_048_A', fill = '#fb8072'),\n",
    "        upset_query(set = '008_048_B', fill = '#e41a1c'),\n",
    "        upset_query(set = '008_048_C', fill = '#e41a1c'),\n",
    "        upset_query(set = '008_048_D', fill = '#e41a1c'),\n",
    "        upset_query(intersect = c('008_048_A','008_048_B','008_048_C','008_048_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_048_C','008_048_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_C','008_048_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_C','008_048_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_B','008_048_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_048_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D'), color = '#e41a1c', fill = '#e41a1c', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_B','008_048_C'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_048_C'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_C'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_C'), color = '#e41a1c', fill = '#e41a1c', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_B'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B'), color = '#e41a1c', fill = '#e41a1c', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A'), color = '#fb8072', fill = '#fb8072', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_048_D','008_021_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D','008_021_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_021_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D'), color = '#e41a1c', fill = '#e41a1c', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_B','008_048_C','008_048_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_048_C','008_048_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_C','008_048_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_048_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_C','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_B','008_048_C','008_048_D','008_021_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_048_C','008_048_D','008_021_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_048_D','008_021_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D','008_021_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_C'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_C'), color = '#e41a1c', fill = '#e41a1c', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_B','008_048_C','008_048_D','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_048_D','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_021_B'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D','008_021_D','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_B'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D','008_021_C','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_C','008_021_C','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_C','008_021_B'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_C','008_021_B'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_B'), color = '#e41a1c', fill = '#e41a1c', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D','008_021_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_021_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D','008_021_D','008_021_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_C','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_C','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_B','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D','008_021_C','008_021_B','008_021_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_048_C','008_048_D','008_021_D','008_021_C','008_021_B','008_021_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_D','008_021_D','008_021_C','008_021_B','008_021_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_B','008_021_D','008_021_C','008_021_B','008_021_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_048_A','008_021_D','008_021_C','008_021_B','008_021_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_C','008_021_B','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_C','008_021_B','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_B','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_A'), color = '#fb8072', fill = '#fb8072', only_components=c('intersections_matrix', 'Intersection size'))\n",
    "    ),\n",
    "    intersections = list(\n",
    "        '008_048_B','008_048_D','008_021_C','008_021_D',\n",
    "        '008_048_A','008_021_B','008_048_C','008_021_A',\n",
    "        c('008_048_B','008_048_D'),\n",
    "        c('008_048_B','008_048_C','008_048_D'),\n",
    "        c('008_048_C','008_048_D'),\n",
    "        c('008_048_B','008_048_C'),\n",
    "        c('008_021_D','008_021_C'),\n",
    "        c('008_021_D','008_021_C','008_021_B'),\n",
    "        c('008_021_C','008_021_B'),\n",
    "        c('008_021_D','008_021_B'),\n",
    "        c('008_048_A','008_048_B','008_048_C','008_048_D'),\n",
    "        c('008_048_A','008_048_B','008_048_D'),\n",
    "        c('008_048_A','008_048_B'),\n",
    "        c('008_048_A','008_048_D'),\n",
    "        c('008_048_A','008_048_B','008_048_C'),\n",
    "        c('008_048_A','008_048_C','008_048_D'),\n",
    "        c('008_048_A','008_048_C'),\n",
    "        c('008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_021_D','008_021_C','008_021_A'),\n",
    "        c('008_021_B','008_021_A'),\n",
    "        c('008_021_C','008_021_A'),\n",
    "        c('008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_021_D','008_021_B','008_021_A'),\n",
    "        c('008_021_D','008_021_A'),\n",
    "        c('008_048_B','008_021_C'),\n",
    "        c('008_048_D','008_021_D'),\n",
    "        c('008_048_B','008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_048_D','008_021_B'),\n",
    "        c('008_048_A','008_048_B','008_048_C','008_048_D','008_021_D','008_021_C'),\n",
    "        c('008_048_B','008_048_D','008_021_C'),\n",
    "        c('008_048_B','008_048_D','008_021_B'),\n",
    "        c('008_048_B','008_021_B'),\n",
    "        c('008_048_B','008_048_D','008_021_D'),\n",
    "        c('008_048_B','008_021_D'),\n",
    "        c('008_048_A','008_048_B','008_048_C','008_048_D','008_021_C'),\n",
    "        c('008_048_B','008_048_C','008_048_D','008_021_C'),\n",
    "        c('008_048_C','008_048_D','008_021_C'),\n",
    "        c('008_048_C','008_021_C'),\n",
    "        c('008_048_B','008_048_C','008_048_D','008_021_D','008_021_C'),\n",
    "        c('008_048_B','008_048_D','008_021_D','008_021_C'),\n",
    "        c('008_048_D','008_021_D','008_021_C'),\n",
    "        c('008_048_A','008_048_B','008_048_C','008_048_D','008_021_B'),\n",
    "        c('008_048_D','008_021_D','008_021_B'),\n",
    "        c('008_048_D','008_021_C','008_021_B'),\n",
    "        c('008_048_C','008_021_C','008_021_B'),\n",
    "        c('008_048_D','008_021_D','008_021_A'),\n",
    "        c('008_048_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_048_A','008_048_C','008_048_D','008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_048_D','008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_048_A','008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_048_A','008_021_B'),\n",
    "        c('008_048_D','008_021_A'),\n",
    "        c('008_048_B','008_021_A')),\n",
    "    sort_sets=FALSE,\n",
    "    sort_intersections=FALSE,\n",
    "    width_ratio=0.1,\n",
    "    name = \"Overlap between TCR repertoires\",\n",
    "    base_annotations = list(\n",
    "            'Intersection size'=intersection_size(\n",
    "                text = list(size = 7, angle = 90, hjust = 0, vjust = 0.5)\n",
    "            ) + ylab('Number of overlapping TCRs') \n",
    "            + ggtitle(\"Epidemic KS - Epidemic KS\")\n",
    "            + theme_classic() \n",
    "            + labs(tag = \"B\")\n",
    "            + theme(axis.title.x = element_blank(),\n",
    "                plot.title = element_text(size = 30, face = \"bold\", family = \"NimbusSan\", hjust = 0.5),\n",
    "                axis.text.x = element_blank(),\n",
    "                axis.ticks.x = element_blank(),\n",
    "                axis.title.y = element_text(size = 24, face = \"bold\", family = \"NimbusSan\"),\n",
    "                axis.text.y = element_text(size = 22, face = \"bold\", family = \"NimbusSan\"),\n",
    "                plot.tag = element_text(size = 28, face = \"bold\", family = \"NimbusSan\"))),\n",
    "    set_sizes=(upset_set_size()+ ylab('Number of\\nunique TCRs') \n",
    "            + theme_void() \n",
    "            + theme(axis.title.y = element_blank(),\n",
    "                axis.text.y = element_blank(),\n",
    "                axis.ticks.y = element_blank(),\n",
    "                axis.title.x = element_blank(),\n",
    "                axis.text.x = element_text(size = 20, face = \"bold\", family = \"NimbusSan\", angle = 90, vjust = 0.5),\n",
    "                axis.ticks.x = element_line(linewidth = 2, colour = 'black' ))\n",
    "    )) +\n",
    "    theme(axis.title.x = element_blank(),\n",
    "        axis.text.y = element_text(size = 24, face = \"bold\", family = \"NimbusSan\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "epiend_upset_input_tibble <- alluvial_data |>\n",
    "    filter(str_detect(repertoire_id, \"008_021|008_057\"))  |>\n",
    "    left_join(study_metadata, by = \"repertoire_id\") |>\n",
    "    dplyr::select(repertoire_id, junction_aa, tissue_type, phenotype) |>\n",
    "    mutate(present = 1,\n",
    "        row_id = str_c(repertoire_id, tissue_type, phenotype, sep = \" \")) |>\n",
    "    pivot_wider(id_cols = junction_aa, names_from = repertoire_id, values_from = present, values_fill = 0)\n",
    "\n",
    "epiend_upset_rownames <- epiend_upset_input_tibble |>\n",
    "    pull(junction_aa)\n",
    "epiend_upset_input_table <- epiend_upset_input_tibble |>\n",
    "    dplyr::select(-junction_aa) |>\n",
    "    as.data.frame() \n",
    "rownames(epiend_upset_input_table) <- epiend_upset_rownames\n",
    "\n",
    "# UpSetR plots for epidemic - endemic KS data comparisons\n",
    "epiend_upset_data <- upset_data(epiend_upset_input_table, \n",
    "    intersect = c(\"008_021_A\", \"008_021_B\", \"008_021_C\", \"008_021_D\", \"008_057_D\", \"008_057_C\", \"008_057_B\", \"008_057_A\"))\n",
    "\n",
    "combined_pal <- c(\"Epidemic KS - NAT\" = \"#fb8072\", \n",
    "  \"Epidemic KS - Tumor\" = \"#e41a1c\", \"Endemic KS - NAT\" = \"#80b1d3\",\n",
    "  \"Endemic KS - Tumor\" = \"#377eb8\", \"Inter-subject\\nNAT - NAT\" = \"#1b9e77\", \n",
    "  \"Intra-subject\\nNAT - Tumor\" = \"#d95f02\",\n",
    "  \"Intra-subject\\nTumor - Tumor\" = \"#7570b3\", \n",
    "  \"Inter-subject\\nTumor - Tumor\" = \"#e7298a\",\n",
    "  \"Other\" = \"#606060\")\n",
    "epiend_pal_table <- epiend_upset_data$plot_intersections_subset |>\n",
    "    as_tibble() |> \n",
    "    dplyr::rename(intersections = value) |>\n",
    "    mutate(group = case_when(intersections == \"008_021_A\" ~ \"Epidemic KS - NAT\",\n",
    "        intersections == \"008_057_A\" ~ \"Endemic KS - NAT\",\n",
    "        intersections %in% c(\"008_057_B\",\"008_057_C\",\"008_057_D\")~ \"Endemic KS - Tumor\",\n",
    "        intersections %in% c(\"008_021_B\",\"008_021_C\",\"008_021_D\")~ \"Epidemic KS - Tumor\",\n",
    "        str_detect(intersections, \"008_021\") & !str_detect(intersections,\"008_057\") & str_detect(intersections,\"A\") & str_detect(intersections,'-') ~ \"Intra-subject\\nNAT - Tumor\",\n",
    "        !str_detect(intersections, \"008_021\") & str_detect(intersections,\"008_057\") & str_detect(intersections,\"A\") & str_detect(intersections,'[BCD]') ~ \"Intra-subject\\nNAT - Tumor\",\n",
    "        str_detect(intersections, \"008_021\") & !str_detect(intersections,\"008_057\") & !str_detect(intersections,\"A\") & str_detect(intersections,'[BCD]') ~ \"Intra-subject\\nTumor - Tumor\",\n",
    "        !str_detect(intersections, \"008_021\") & str_detect(intersections,\"008_057\") & !str_detect(intersections,\"A\") & str_detect(intersections,'[BCD]') ~ \"Intra-subject\\nTumor - Tumor\",\n",
    "        str_detect(intersections, \"008_021\") & str_detect(intersections,\"008_057\") & str_detect(intersections,\"008_021_B|008_021_C|008_021_D\") & str_detect(intersections,\"008_057_B|008_057_C|008_057_D\") ~ \"Inter-subject\\nTumor - Tumor\",\n",
    "        intersections == \"008_021_A-008_057_A\" ~ \"Inter-subject\\nNAT - NAT\",\n",
    "        str_detect(intersections,\"008_021\") & str_detect(intersections,\"008_057\") & !str_detect(intersections,\"008_021_B\") & !str_detect(intersections,\"008_021_C\") & !str_detect(intersections,\"008_021_D\") ~ \"Other\",\n",
    "        str_detect(intersections,\"008_021\") & str_detect(intersections,\"008_057\") & !str_detect(intersections,\"008_057_B\") & !str_detect(intersections,\"008_057_C\") & !str_detect(intersections,\"008_057_D\") ~ \"Other\"),\n",
    "        color = combined_pal[group])\n",
    "epiend_pal <- epiend_pal_table |>\n",
    "  pull(color) \n",
    "names(epiend_pal) <- epiend_pal_table |>\n",
    "  pull(intersections) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "epiend_upset <- upset(epiend_upset_input_table, \n",
    "    intersect = c(\"008_021_A\", \"008_021_B\", \"008_021_C\", \"008_021_D\", \"008_057_D\", \"008_057_C\", \"008_057_B\", \"008_057_A\"),\n",
    "    queries = list(\n",
    "        upset_query(set = '008_021_A', fill = '#fb8072'),\n",
    "        upset_query(set = '008_021_B', fill = '#e41a1c'),\n",
    "        upset_query(set = '008_021_C', fill = '#e41a1c'),\n",
    "        upset_query(set = '008_021_D', fill = '#e41a1c'),\n",
    "        upset_query(set = '008_057_A', fill = '#80b1d3'),\n",
    "        upset_query(set = '008_057_B', fill = '#377eb8'),\n",
    "        upset_query(set = '008_057_C', fill = '#377eb8'),\n",
    "        upset_query(set = '008_057_D', fill = '#377eb8'),\n",
    "        upset_query(intersect = c('008_057_A','008_057_B','008_057_C','008_057_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_057_C','008_057_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_057_C','008_057_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_C','008_057_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_057_B','008_057_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_057_D'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_057_D'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_D'), color = '#377eb8', fill = '#377eb8', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_057_B','008_057_C'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_057_C'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_057_C'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_C'), color = '#377eb8', fill = '#377eb8', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_057_B'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B'), color = '#377eb8', fill = '#377eb8', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A'), color = '#80b1d3', fill = '#80b1d3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_057_C','008_057_D','008_021_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_D','008_021_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_C','008_021_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_021_D'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_021_D'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D'), color = '#e41a1c', fill = '#e41a1c', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_C','008_057_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_057_B','008_057_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_057_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_057_C','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_C','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_021_C'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_057_D','008_021_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_D','008_021_D','008_021_C'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_C'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_C'), color = '#e41a1c', fill = '#e41a1c', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_057_D','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_D','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_C','008_021_D','008_021_B'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_B'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_021_C','008_021_B'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_C','008_021_B'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_C','008_021_B'), color = '#7570b3', fill = '#7570b3', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_B'), color = '#e41a1c', fill = '#e41a1c', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_D','008_021_A'), color = '#606060', fill = '#606060', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_C','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_C','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_B','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_057_B','008_057_C','008_057_D','008_021_D','008_021_C','008_021_B','008_021_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_A','008_057_B','008_057_D','008_021_D','008_021_C','008_021_B','008_021_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_D','008_021_D','008_021_C','008_021_B','008_021_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_057_B','008_021_D','008_021_C','008_021_B','008_021_A'), color = '#e7298a', fill = '#e7298a', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_D','008_021_C','008_021_B','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_C','008_021_B','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_B','008_021_A'), color = '#d95f02', fill = '#d95f02', only_components=c('intersections_matrix', 'Intersection size')),\n",
    "        upset_query(intersect = c('008_021_A'), color = '#fb8072', fill = '#fb8072', only_components=c('intersections_matrix', 'Intersection size'))),\n",
    "    intersections = list(\n",
    "        '008_057_B','008_057_D','008_057_C','008_057_A',\n",
    "        '008_021_C','008_021_D','008_021_B','008_021_A',\n",
    "        c('008_057_B','008_057_D'),\n",
    "        c('008_057_B','008_057_C','008_057_D'),\n",
    "        c('008_057_B','008_057_C'),\n",
    "        c('008_057_C','008_057_D'),\n",
    "        c('008_021_D','008_021_C'),\n",
    "        c('008_021_D','008_021_C','008_021_B'),\n",
    "        c('008_021_C','008_021_B'),\n",
    "        c('008_021_D','008_021_B'),\n",
    "        c('008_057_A','008_057_B','008_057_C','008_057_D'),\n",
    "        c('008_057_A','008_057_D'),\n",
    "        c('008_057_A','008_057_B'),\n",
    "        c('008_057_A','008_057_C'),\n",
    "        c('008_057_A','008_057_B','008_057_D'),\n",
    "        c('008_057_A','008_057_C','008_057_D'),\n",
    "        c('008_057_A','008_057_B','008_057_C'),\n",
    "        c('008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_021_D','008_021_C','008_021_A'),\n",
    "        c('008_021_B','008_021_A'),\n",
    "        c('008_021_C','008_021_A'),\n",
    "        c('008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_021_D','008_021_A'),\n",
    "        c('008_021_D','008_021_B','008_021_A'),\n",
    "        c('008_057_B','008_057_D','008_021_C'),\n",
    "        c('008_057_B','008_021_C'),\n",
    "        c('008_057_D','008_021_D'),\n",
    "        c('008_057_B','008_021_D'),\n",
    "        c('008_057_D','008_021_C'),\n",
    "        c('008_057_D','008_021_B'),\n",
    "        c('008_057_B','008_021_B'),\n",
    "        c('008_057_B','008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_057_B','008_057_C','008_057_D','008_021_D'),\n",
    "        c('008_057_C','008_021_D'),\n",
    "        c('008_057_C','008_057_D','008_021_C'),\n",
    "        c('008_057_A','008_057_B','008_057_D','008_021_C'),\n",
    "        c('008_057_B','008_057_C','008_021_C'),\n",
    "        c('008_057_C','008_021_C'),\n",
    "        c('008_057_B','008_057_D','008_021_D','008_021_C'),\n",
    "        c('008_057_D','008_021_D','008_021_C'),\n",
    "        c('008_057_B','008_057_D','008_021_B'),\n",
    "        c('008_057_C','008_021_D','008_021_B'),\n",
    "        c('008_057_A','008_057_B','008_057_C','008_057_D','008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_057_A','008_057_B','008_057_D','008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_057_D','008_021_D','008_021_C','008_021_B','008_021_A'),\n",
    "        c('008_057_A','008_021_D'),\n",
    "        c('008_057_A','008_021_C'),\n",
    "        c('008_057_A','008_021_C','008_021_B'),\n",
    "        c('008_057_D','008_021_A')\n",
    "\n",
    "    ),\n",
    "    sort_sets=FALSE,\n",
    "    sort_intersections=FALSE,\n",
    "    width_ratio=0.1,\n",
    "    name = \"Overlap between TCR repertoires\",\n",
    "    base_annotations = list(\n",
    "            'Intersection size'=intersection_size(\n",
    "                aes(fill = intersection),\n",
    "                text = list(size = 7, angle = 90, hjust = 0, vjust = 0.5)\n",
    "            ) + ylab('Number of overlapping TCRs') \n",
    "            + theme_classic() \n",
    "            + ggtitle(\"Endemic KS - Epidemic KS\") \n",
    "            + labs(tag = \"C\") \n",
    "            + scale_fill_manual(values = epiend_pal)\n",
    "            + theme(axis.title.x = element_blank(),\n",
    "                plot.title = element_text(size = 30, face = \"bold\", family = \"NimbusSan\", hjust = 0.5),\n",
    "                axis.text.x = element_blank(),\n",
    "                axis.ticks.x = element_blank(),\n",
    "                axis.title.y = element_text(size = 24, face = \"bold\", family = \"NimbusSan\"),\n",
    "                axis.text.y = element_text(size = 22, face = \"bold\", family = \"NimbusSan\"),\n",
    "                plot.tag = element_text(size = 28, face = \"bold\", family = \"NimbusSan\"),\n",
    "                legend.position = \"right\")),\n",
    "    set_sizes=(upset_set_size()+ ylab('Number of\\nunique TCRs') \n",
    "            + theme_void() \n",
    "            + theme(axis.title.y = element_blank(),\n",
    "                axis.text.y = element_blank(),\n",
    "                axis.ticks.y = element_blank(),\n",
    "                axis.title.x = element_text(size = 24, face = \"bold\", family = \"NimbusSan\"),\n",
    "                axis.text.x = element_text(size = 20, face = \"bold\", family = \"NimbusSan\", angle = 90, vjust = 0.5),\n",
    "                axis.ticks.x = element_line(linewidth = 2, colour = 'black' ))\n",
    "    ), guides = \"collect\") +\n",
    "    scale_fill_manual(values = epiend_pal) +\n",
    "    theme(axis.title.x = element_text(size = 24, face = \"bold\", family = \"NimbusSan\"),\n",
    "        axis.text.y = element_text(size = 24, face = \"bold\", family = \"NimbusSan\")) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "supplementary_figure_seven <- endemic_upset / epidemic_upset / epiend_upset + plot_layout(guides = \"collect\")  & theme(text = element_text('NimbusSan'))\n",
    "\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Supplementary_figure_seven.pdf\", sep = \"/\"), supplementary_figure_seven, width = 32, \n",
    "    height = 30, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"png\", \"Supplementary_figure_seven.png\", sep = \"/\"), supplementary_figure_seven, width = 32, \n",
    "    height = 30)\n",
    "ggsave(str_c(figures_path, \"svg\", \"Supplementary_figure_seven.svg\", sep = \"/\"), supplementary_figure_seven, width = 32, \n",
    "    height = 30)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 4: Temporal overlap between KS tumors\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "temporal_map_table <- study_metadata |>\n",
    "    filter(patient_id %in% c(\"008_141\", \"008_008\")) |>\n",
    "    dplyr::select(patient_id, repertoire_id, phenotype, tissue_type, tumor_code, visit_code) |>\n",
    "    mutate(visit_code = str_c(\"Visit\", str_extract(visit_code, \"\\\\d+\"), sep = \" \"),\n",
    "    label = if_else(tissue_type == \"NAT\", \n",
    "        str_c(tissue_type, \"\\n\", visit_code, sep = \"\"),\n",
    "        str_c(tissue_type, \" \", tumor_code, \"\\n\", visit_code, sep = \"\"))) |>\n",
    "    dplyr::select(repertoire_id, label)\n",
    "temporal_map_dict <- temporal_map_table |>\n",
    "    pull(label)\n",
    "names(temporal_map_dict) <- temporal_map_table |>\n",
    "    pull(repertoire_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Subsetting productive sequences [=>----------]  14% eta:  3s\n",
      "\n",
      "Subsetting productive sequences [==>---------]  21% eta:  3s\n",
      "\n",
      "Subsetting productive sequences [==>---------]  29% eta:  3s\n",
      "\n",
      "Subsetting productive sequences [===>--------]  36% eta: 15s\n",
      "\n",
      "Subsetting productive sequences [====>-------]  43% eta: 12s\n",
      "\n",
      "Subsetting productive sequences [=====>------]  50% eta:  9s\n",
      "\n",
      "Subsetting productive sequences [======>-----]  57% eta:  7s\n",
      "\n",
      "Subsetting productive sequences [=======>----]  64% eta:  6s\n",
      "\n",
      "Subsetting productive sequences [========>---]  71% eta:  4s\n",
      "\n",
      "Subsetting productive sequences [========>---]  79% eta:  3s\n",
      "\n",
      "Subsetting productive sequences [=========>--]  86% eta:  2s\n",
      "\n",
      "Subsetting productive sequences [==========>-]  93% eta:  1s\n",
      "\n",
      "Subsetting productive sequences [============] 100% eta:  0s\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "temopral_example_table <- study_annotated_nprod_table |>\n",
    "  filter(patient_id %in% c(\"008_141\", \"008_008\")) |>\n",
    "  productiveSeq(aggregate = \"junction_aa\") \n",
    "\n",
    "temporal_overlap_samples <- study_metadata |> \n",
    "  filter(tissue_type %in% c( \"Tumor\") & \n",
    "    str_detect(repertoire_id, \"008_\\\\d+\")) |>  \n",
    "  group_by(patient_id, hiv_status, visit_code) |> \n",
    "  summarize(count = length(unique(repertoire_id))) |> \n",
    "  pivot_wider(names_from = visit_code, values_from = count, values_fill = 0) |> \n",
    "  filter(V01 != 0 & (V05 != 0 | V08 != 0)) |>\n",
    "  pull(patient_id)\n",
    "\n",
    "visit_map <- study_metadata |>\n",
    "  filter(tissue_type %in% c( \"Tumor\") & \n",
    "    str_detect(repertoire_id, \"008_\\\\d+\")) |>\n",
    "  pull(visit_code)\n",
    "names(visit_map) <- study_metadata |>\n",
    "  filter(tissue_type %in% c( \"Tumor\") & \n",
    "    str_detect(repertoire_id, \"008_\\\\d+\")) |>\n",
    "  pull(trb_repertoire_id)\n",
    "\n",
    "endemic_ug <- temopral_example_table |> \n",
    "  filter(junction_aa %in% kstme_clustered_unknown_table$junction_aa & \n",
    "    str_detect(repertoire_id, \"008_141\")) |>\n",
    "  mutate(color = \"#377eb8\") |>\n",
    "  pull(color) \n",
    "\n",
    "endemic_alist <- temopral_example_table |> \n",
    "  filter(junction_aa %in% kstme_clustered_unknown_table$junction_aa & \n",
    "    str_detect(repertoire_id, \"008_141\")) |>\n",
    "  mutate(color = \"#377eb8\") |>\n",
    "  pull(junction_aa) \n",
    "names(endemic_ug) <- temopral_example_table |> \n",
    "  filter(junction_aa %in% kstme_clustered_unknown_table$junction_aa & \n",
    "    str_detect(repertoire_id, \"008_141\")) |>\n",
    "  mutate(color = \"#377eb8\") |>\n",
    "  pull(junction_aa) \n",
    "\n",
    "endemic_alluvial <- temopral_example_table |>\n",
    "  filter(str_detect(repertoire_id, \"008_141\")) |>\n",
    "  topSeqs(top = 500) |>\n",
    "  cloneTrack() |>\n",
    "  plotTrack(alist = endemic_alist, apal = endemic_ug) +\n",
    "  scale_x_discrete(labels = temporal_map_dict) +\n",
    "  labs(x = \"Sample\") + \n",
    "  theme_classic(base_size = 30) +\n",
    "  theme(legend.position = \"none\", \n",
    "    axis.text.x = element_text(face = \"bold\", color = \"black\"))\n",
    "\n",
    "epidemic_ug <- temopral_example_table |> \n",
    "  filter(junction_aa %in% kstme_clustered_unknown_table$junction_aa & \n",
    "    str_detect(repertoire_id, \"008_008\")) |>\n",
    "  mutate(color = \"#e41a1c\") |>\n",
    "  pull(color) \n",
    "epidemic_alist <- temopral_example_table |> \n",
    "  filter(junction_aa %in% kstme_clustered_unknown_table$junction_aa & \n",
    "    str_detect(repertoire_id, \"008_008\")) |>\n",
    "  mutate(color = \"#e41a1c\") |>\n",
    "  pull(junction_aa) \n",
    "names(epidemic_ug) <- temopral_example_table |> \n",
    "  filter(junction_aa %in% kstme_clustered_unknown_table$junction_aa & \n",
    "    str_detect(repertoire_id, \"008_008\")) |>\n",
    "  mutate(color = \"#e41a1c\") |>\n",
    "  pull(junction_aa) \n",
    "\n",
    "epidemic_alluvial <- temopral_example_table |>\n",
    "  filter(str_detect(repertoire_id, \"008_008\")) |>\n",
    "  topSeqs(top = 500) |>\n",
    "  cloneTrack() |>\n",
    "  plotTrack(alist = epidemic_alist, apal = epidemic_ug) +\n",
    "  scale_x_discrete(labels = temporal_map_dict) +\n",
    "  labs(x = \"Sample\") + \n",
    "  theme_classic(base_size = 30)  +\n",
    "  theme(legend.position = \"none\",\n",
    "   axis.text.x = element_text(face = \"bold\", color = \"black\"))\n",
    "\n",
    "figure_four_layout <- \"\n",
    "11112222\n",
    "11112222\n",
    "11112222\"\n",
    "figure_four <- (endemic_alluvial + epidemic_alluvial +plot_layout(guides = \"auto\"))  +\n",
    "  plot_layout(design = figure_four_layout) +\n",
    "  plot_annotation(tag_levels = 'A') & theme(text = element_text('NimbusSan', face = \"bold\"))\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Figure_four.pdf\", sep = \"/\"), figure_four, \n",
    "  width = 30, height = 30)\n",
    "ggsave(str_c(figures_path, \"svg\", \"Figure_four.svg\", sep = \"/\"), figure_four, \n",
    "  width = 30, height = 30)\n",
    "ggsave(str_c(figures_path, \"png\", \"Figure_four.png\", sep = \"/\"), figure_four, \n",
    "  width =  30, height = 30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 5: Temporal sharing of TCRs in single-cell sequencing data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load 10X PBMC dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Validating h5Seurat file\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding feature-level metadata for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding reduction PCA\n",
      "\n",
      "Adding cell embeddings for PCA\n",
      "\n",
      "Adding miscellaneous information for PCA\n",
      "\n",
      "Adding reduction TSNE\n",
      "\n",
      "Adding cell embeddings for TSNE\n",
      "\n",
      "Adding miscellaneous information for TSNE\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 23 features requested have zero variance (running reduction without them): ENSG00000169877, ENSG00000211720, ENSG00000277856, ENSG00000242766, ENSG00000196565, ENSG00000211976, ENSG00000211654, ENSG00000243063, ENSG00000276775, ENSG00000211672, ENSG00000241566, ENSG00000278857, ENSG00000211623, ENSG00000204866, ENSG00000286848, ENSG00000254056, ENSG00000259995, ENSG00000229613, ENSG00000251363, ENSG00000211961, ENSG00000224607, ENSG00000169760, ENSG00000112212”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000185885, ENSG00000077984, ENSG00000271503, ENSG00000105374, ENSG00000145649, ENSG00000180644, ENSG00000167286, ENSG00000115523, ENSG00000100450 \n",
      "\t   ENSG00000160654, ENSG00000203896, ENSG00000173762, ENSG00000137441, ENSG00000100453, ENSG00000211751, ENSG00000069667, ENSG00000078596, ENSG00000110848, ENSG00000172543 \n",
      "\t   ENSG00000172673, ENSG00000134539, ENSG00000213809, ENSG00000205336, ENSG00000181036, ENSG00000107742, ENSG00000139187, ENSG00000153563, ENSG00000111716, ENSG00000136732 \n",
      "Negative:  ENSG00000204472, ENSG00000216490, ENSG00000066336, ENSG00000101439, ENSG00000197249, ENSG00000204482, ENSG00000085265, ENSG00000197766, ENSG00000126759, ENSG00000101336 \n",
      "\t   ENSG00000129226, ENSG00000090382, ENSG00000136826, ENSG00000105835, ENSG00000116701, ENSG00000085514, ENSG00000165168, ENSG00000131042, ENSG00000172322, ENSG00000025708 \n",
      "\t   ENSG00000168461, ENSG00000120594, ENSG00000172243, ENSG00000204103, ENSG00000254087, ENSG00000183023, ENSG00000166927, ENSG00000163131, ENSG00000197405, ENSG00000163563 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000203747, ENSG00000105374, ENSG00000108518, ENSG00000185201, ENSG00000122862, ENSG00000180644, ENSG00000077984, ENSG00000115523, ENSG00000168329, ENSG00000145649 \n",
      "\t   ENSG00000100453, ENSG00000137441, ENSG00000172543, ENSG00000271503, ENSG00000100450, ENSG00000075624, ENSG00000196154, ENSG00000185885, ENSG00000155366, ENSG00000134539 \n",
      "\t   ENSG00000205336, ENSG00000108798, ENSG00000117984, ENSG00000159674, ENSG00000142089, ENSG00000180739, ENSG00000135046, ENSG00000171476, ENSG00000181036, ENSG00000184009 \n",
      "Negative:  ENSG00000105369, ENSG00000153064, ENSG00000156738, ENSG00000144218, ENSG00000167483, ENSG00000116191, ENSG00000247982, ENSG00000211899, ENSG00000196092, ENSG00000136573 \n",
      "\t   ENSG00000132185, ENSG00000082438, ENSG00000176533, ENSG00000177455, ENSG00000095585, ENSG00000163534, ENSG00000012124, ENSG00000102445, ENSG00000100721, ENSG00000269404 \n",
      "\t   ENSG00000241106, ENSG00000164330, ENSG00000211898, ENSG00000042980, ENSG00000128218, ENSG00000110777, ENSG00000211592, ENSG00000082293, ENSG00000104921, ENSG00000112182 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000182853, ENSG00000142512, ENSG00000203747, ENSG00000129757, ENSG00000224397, ENSG00000166165, ENSG00000287682, ENSG00000170873, ENSG00000007312, ENSG00000198682 \n",
      "\t   ENSG00000197353, ENSG00000148737, ENSG00000188290, ENSG00000135047, ENSG00000155366, ENSG00000104972, ENSG00000182578, ENSG00000261222, ENSG00000166927, ENSG00000170956 \n",
      "\t   ENSG00000259225, ENSG00000174837, ENSG00000105371, ENSG00000161921, ENSG00000233392, ENSG00000133789, ENSG00000104974, ENSG00000175489, ENSG00000166428, ENSG00000125898 \n",
      "Negative:  ENSG00000038427, ENSG00000163221, ENSG00000143546, ENSG00000170458, ENSG00000110077, ENSG00000100079, ENSG00000119535, ENSG00000186431, ENSG00000166523, ENSG00000135218 \n",
      "\t   ENSG00000124731, ENSG00000163220, ENSG00000146592, ENSG00000257764, ENSG00000249437, ENSG00000125538, ENSG00000162711, ENSG00000087903, ENSG00000145819, ENSG00000171051 \n",
      "\t   ENSG00000106546, ENSG00000169403, ENSG00000177575, ENSG00000182621, ENSG00000163823, ENSG00000121552, ENSG00000161944, ENSG00000008394, ENSG00000276085, ENSG00000090382 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000168685, ENSG00000152495, ENSG00000227507, ENSG00000165272, ENSG00000081059, ENSG00000109452, ENSG00000138795, ENSG00000172005, ENSG00000167286, ENSG00000178562 \n",
      "\t   ENSG00000139193, ENSG00000126353, ENSG00000111716, ENSG00000164300, ENSG00000026025, ENSG00000151150, ENSG00000182463, ENSG00000160654, ENSG00000169508, ENSG00000107742 \n",
      "\t   ENSG00000163519, ENSG00000089157, ENSG00000152270, ENSG00000165591, ENSG00000010610, ENSG00000170293, ENSG00000182809, ENSG00000102245, ENSG00000163600, ENSG00000079335 \n",
      "Negative:  ENSG00000100453, ENSG00000172543, ENSG00000180644, ENSG00000134539, ENSG00000105374, ENSG00000159674, ENSG00000115523, ENSG00000137441, ENSG00000077984, ENSG00000163453 \n",
      "\t   ENSG00000075234, ENSG00000171476, ENSG00000150045, ENSG00000169583, ENSG00000180739, ENSG00000205336, ENSG00000145649, ENSG00000136250, ENSG00000198574, ENSG00000205809 \n",
      "\t   ENSG00000036448, ENSG00000211829, ENSG00000135404, ENSG00000100385, ENSG00000168329, ENSG00000203747, ENSG00000181036, ENSG00000150687, ENSG00000169554, ENSG00000100450 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000034510, ENSG00000196154, ENSG00000197956, ENSG00000160654, ENSG00000100450, ENSG00000167286, ENSG00000197747, ENSG00000223865, ENSG00000231389, ENSG00000196126 \n",
      "\t   ENSG00000139187, ENSG00000176083, ENSG00000135046, ENSG00000137818, ENSG00000198502, ENSG00000100097, ENSG00000105374, ENSG00000172116, ENSG00000153563, ENSG00000089157 \n",
      "\t   ENSG00000172673, ENSG00000125148, ENSG00000213145, ENSG00000179344, ENSG00000100906, ENSG00000026025, ENSG00000019582, ENSG00000115523, ENSG00000008517, ENSG00000145649 \n",
      "Negative:  ENSG00000163736, ENSG00000168497, ENSG00000101162, ENSG00000113140, ENSG00000169704, ENSG00000204420, ENSG00000127920, ENSG00000163737, ENSG00000101335, ENSG00000154146 \n",
      "\t   ENSG00000111644, ENSG00000120885, ENSG00000171611, ENSG00000005961, ENSG00000161911, ENSG00000109272, ENSG00000095303, ENSG00000162722, ENSG00000184113, ENSG00000166681 \n",
      "\t   ENSG00000162366, ENSG00000259207, ENSG00000088726, ENSG00000187699, ENSG00000250334, ENSG00000105711, ENSG00000166091, ENSG00000180573, ENSG00000137801, ENSG00000138449 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 41 features requested have zero variance (running reduction without them): ENSG00000253239, ENSG00000242580, ENSG00000211637, ENSG00000204010, ENSG00000211652, ENSG00000211655, ENSG00000242766, ENSG00000004939, ENSG00000282122, ENSG00000211654, ENSG00000243063, ENSG00000166947, ENSG00000232216, ENSG00000253818, ENSG00000211650, ENSG00000211658, ENSG00000211632, ENSG00000163734, ENSG00000276775, ENSG00000275243, ENSG00000211672, ENSG00000241566, ENSG00000250361, ENSG00000254395, ENSG00000163554, ENSG00000253709, ENSG00000231912, ENSG00000254056, ENSG00000127533, ENSG00000115008, ENSG00000226097, ENSG00000186185, ENSG00000185897, ENSG00000259995, ENSG00000163518, ENSG00000236320, ENSG00000229613, ENSG00000112077, ENSG00000211961, ENSG00000224607, ENSG00000112212”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000185885, ENSG00000167286, ENSG00000271503, ENSG00000077984, ENSG00000160654, ENSG00000105374, ENSG00000203896, ENSG00000145649, ENSG00000173762 \n",
      "\t   ENSG00000100450, ENSG00000110848, ENSG00000180644, ENSG00000168685, ENSG00000115523, ENSG00000153563, ENSG00000078596, ENSG00000111716, ENSG00000121966, ENSG00000136732 \n",
      "\t   ENSG00000172116, ENSG00000211751, ENSG00000107742, ENSG00000172673, ENSG00000139187, ENSG00000137441, ENSG00000213809, ENSG00000269028, ENSG00000172543, ENSG00000069667 \n",
      "Negative:  ENSG00000216490, ENSG00000066336, ENSG00000101439, ENSG00000204472, ENSG00000085265, ENSG00000204482, ENSG00000090382, ENSG00000197249, ENSG00000101336, ENSG00000197766 \n",
      "\t   ENSG00000129226, ENSG00000126759, ENSG00000172322, ENSG00000163220, ENSG00000254087, ENSG00000121552, ENSG00000011600, ENSG00000168461, ENSG00000131042, ENSG00000158869 \n",
      "\t   ENSG00000085514, ENSG00000120594, ENSG00000172243, ENSG00000214212, ENSG00000116701, ENSG00000025708, ENSG00000165168, ENSG00000166927, ENSG00000182578, ENSG00000143546 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000227507, ENSG00000126353, ENSG00000188404, ENSG00000105369, ENSG00000081059, ENSG00000138795, ENSG00000172005, ENSG00000165272, ENSG00000144218, ENSG00000153064 \n",
      "\t   ENSG00000168685, ENSG00000211899, ENSG00000156738, ENSG00000116191, ENSG00000112182, ENSG00000247982, ENSG00000152495, ENSG00000111716, ENSG00000167642, ENSG00000136997 \n",
      "\t   ENSG00000132185, ENSG00000269404, ENSG00000177455, ENSG00000139193, ENSG00000142546, ENSG00000136573, ENSG00000176533, ENSG00000137818, ENSG00000169508, ENSG00000089157 \n",
      "Negative:  ENSG00000105374, ENSG00000077984, ENSG00000115523, ENSG00000100450, ENSG00000180644, ENSG00000271503, ENSG00000137441, ENSG00000100453, ENSG00000145649, ENSG00000172543 \n",
      "\t   ENSG00000108518, ENSG00000203747, ENSG00000168329, ENSG00000185201, ENSG00000205336, ENSG00000134539, ENSG00000150687, ENSG00000181036, ENSG00000275302, ENSG00000122862 \n",
      "\t   ENSG00000196154, ENSG00000171476, ENSG00000075624, ENSG00000180739, ENSG00000159674, ENSG00000169554, ENSG00000153563, ENSG00000100097, ENSG00000036448, ENSG00000197956 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000105369, ENSG00000153064, ENSG00000211899, ENSG00000156738, ENSG00000247982, ENSG00000132185, ENSG00000269404, ENSG00000177455, ENSG00000144218, ENSG00000211592 \n",
      "\t   ENSG00000136573, ENSG00000163534, ENSG00000167483, ENSG00000116191, ENSG00000231389, ENSG00000196735, ENSG00000042980, ENSG00000128218, ENSG00000176533, ENSG00000012124 \n",
      "\t   ENSG00000223865, ENSG00000102445, ENSG00000095585, ENSG00000110777, ENSG00000100721, ENSG00000007312, ENSG00000204287, ENSG00000198502, ENSG00000104921, ENSG00000196092 \n",
      "Negative:  ENSG00000168685, ENSG00000081059, ENSG00000165272, ENSG00000172005, ENSG00000138795, ENSG00000152495, ENSG00000111716, ENSG00000109452, ENSG00000142546, ENSG00000026025 \n",
      "\t   ENSG00000164300, ENSG00000170293, ENSG00000182463, ENSG00000008517, ENSG00000178562, ENSG00000170345, ENSG00000126353, ENSG00000102760, ENSG00000010610, ENSG00000139193 \n",
      "\t   ENSG00000163519, ENSG00000227507, ENSG00000182809, ENSG00000169756, ENSG00000237943, ENSG00000152270, ENSG00000188404, ENSG00000072110, ENSG00000160712, ENSG00000184384 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000188290, ENSG00000129757, ENSG00000224397, ENSG00000182853, ENSG00000166165, ENSG00000142512, ENSG00000166927, ENSG00000148737, ENSG00000197353, ENSG00000100292 \n",
      "\t   ENSG00000104974, ENSG00000233392, ENSG00000248810, ENSG00000175489, ENSG00000135047, ENSG00000161921, ENSG00000287682, ENSG00000123685, ENSG00000259225, ENSG00000182578 \n",
      "\t   ENSG00000261222, ENSG00000105371, ENSG00000170956, ENSG00000131042, ENSG00000110079, ENSG00000138119, ENSG00000107736, ENSG00000160326, ENSG00000125898, ENSG00000104972 \n",
      "Negative:  ENSG00000038427, ENSG00000163221, ENSG00000143546, ENSG00000170458, ENSG00000119535, ENSG00000100079, ENSG00000110077, ENSG00000163220, ENSG00000135218, ENSG00000145819 \n",
      "\t   ENSG00000163694, ENSG00000249437, ENSG00000148655, ENSG00000182621, ENSG00000124731, ENSG00000090382, ENSG00000120594, ENSG00000185215, ENSG00000158517, ENSG00000010327 \n",
      "\t   ENSG00000257764, ENSG00000146592, ENSG00000169403, ENSG00000141505, ENSG00000221869, ENSG00000166523, ENSG00000111275, ENSG00000162444, ENSG00000163563, ENSG00000125810 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000034510, ENSG00000137818, ENSG00000111640, ENSG00000184009, ENSG00000089157, ENSG00000213145, ENSG00000075624, ENSG00000196262, ENSG00000163191, ENSG00000108518 \n",
      "\t   ENSG00000196154, ENSG00000100097, ENSG00000197956, ENSG00000010610, ENSG00000163453, ENSG00000165272, ENSG00000158710, ENSG00000197747, ENSG00000186827, ENSG00000130429 \n",
      "\t   ENSG00000111716, ENSG00000102245, ENSG00000111796, ENSG00000036448, ENSG00000026025, ENSG00000171476, ENSG00000170296, ENSG00000142546, ENSG00000166794, ENSG00000132341 \n",
      "Negative:  ENSG00000177606, ENSG00000170345, ENSG00000120129, ENSG00000121966, ENSG00000113088, ENSG00000171223, ENSG00000110848, ENSG00000172116, ENSG00000128016, ENSG00000118503 \n",
      "\t   ENSG00000153563, ENSG00000158050, ENSG00000100906, ENSG00000125740, ENSG00000183508, ENSG00000213809, ENSG00000172673, ENSG00000136250, ENSG00000130844, ENSG00000187118 \n",
      "\t   ENSG00000170624, ENSG00000145819, ENSG00000143878, ENSG00000169554, ENSG00000141682, ENSG00000271503, ENSG00000198695, ENSG00000143669, ENSG00000111537, ENSG00000269028 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 3 features requested have zero variance (running reduction without them): ENSG00000275243, ENSG00000251363, ENSG00000186081”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000204472, ENSG00000101439, ENSG00000216490, ENSG00000197249, ENSG00000204482, ENSG00000197766, ENSG00000090382, ENSG00000158869, ENSG00000011600, ENSG00000163131 \n",
      "\t   ENSG00000105835, ENSG00000085265, ENSG00000066336, ENSG00000163220, ENSG00000129226, ENSG00000197746, ENSG00000166927, ENSG00000130066, ENSG00000025708, ENSG00000204287 \n",
      "\t   ENSG00000254087, ENSG00000126759, ENSG00000076944, ENSG00000198502, ENSG00000172322, ENSG00000120594, ENSG00000143546, ENSG00000131042, ENSG00000163563, ENSG00000183023 \n",
      "Negative:  ENSG00000008517, ENSG00000271503, ENSG00000105374, ENSG00000185885, ENSG00000115523, ENSG00000077984, ENSG00000168685, ENSG00000167286, ENSG00000145649, ENSG00000100450 \n",
      "\t   ENSG00000110848, ENSG00000137441, ENSG00000180644, ENSG00000173762, ENSG00000100453, ENSG00000160654, ENSG00000134539, ENSG00000203896, ENSG00000111796, ENSG00000153563 \n",
      "\t   ENSG00000172673, ENSG00000078596, ENSG00000069667, ENSG00000111716, ENSG00000172543, ENSG00000211751, ENSG00000213809, ENSG00000172116, ENSG00000113088, ENSG00000121966 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000105374, ENSG00000196154, ENSG00000115523, ENSG00000271503, ENSG00000077984, ENSG00000008517, ENSG00000185885, ENSG00000100450, ENSG00000145649, ENSG00000137441 \n",
      "\t   ENSG00000180644, ENSG00000203747, ENSG00000100453, ENSG00000197956, ENSG00000135046, ENSG00000108518, ENSG00000075624, ENSG00000122862, ENSG00000011600, ENSG00000134539 \n",
      "\t   ENSG00000142089, ENSG00000172543, ENSG00000185201, ENSG00000158869, ENSG00000163191, ENSG00000197249, ENSG00000159674, ENSG00000155366, ENSG00000204482, ENSG00000204472 \n",
      "Negative:  ENSG00000105369, ENSG00000156738, ENSG00000153064, ENSG00000196735, ENSG00000204287, ENSG00000116191, ENSG00000144218, ENSG00000211899, ENSG00000179344, ENSG00000019582 \n",
      "\t   ENSG00000136573, ENSG00000164330, ENSG00000211592, ENSG00000058091, ENSG00000198502, ENSG00000247982, ENSG00000134532, ENSG00000196092, ENSG00000223865, ENSG00000007312 \n",
      "\t   ENSG00000167483, ENSG00000231389, ENSG00000163534, ENSG00000176533, ENSG00000269404, ENSG00000082438, ENSG00000196628, ENSG00000177455, ENSG00000132185, ENSG00000196126 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000143546, ENSG00000038427, ENSG00000163220, ENSG00000090382, ENSG00000170458, ENSG00000110077, ENSG00000100079, ENSG00000163221, ENSG00000125538, ENSG00000276085 \n",
      "\t   ENSG00000085265, ENSG00000163563, ENSG00000119535, ENSG00000135218, ENSG00000125740, ENSG00000169429, ENSG00000166523, ENSG00000145819, ENSG00000158517, ENSG00000121552 \n",
      "\t   ENSG00000162711, ENSG00000148655, ENSG00000146592, ENSG00000171051, ENSG00000249437, ENSG00000137818, ENSG00000120738, ENSG00000087903, ENSG00000100906, ENSG00000106546 \n",
      "Negative:  ENSG00000203747, ENSG00000182853, ENSG00000197353, ENSG00000224397, ENSG00000166927, ENSG00000148737, ENSG00000155366, ENSG00000204482, ENSG00000175489, ENSG00000104972 \n",
      "\t   ENSG00000185201, ENSG00000131042, ENSG00000170873, ENSG00000142089, ENSG00000166165, ENSG00000287682, ENSG00000261371, ENSG00000129757, ENSG00000100292, ENSG00000142512 \n",
      "\t   ENSG00000198682, ENSG00000197249, ENSG00000111679, ENSG00000197405, ENSG00000182578, ENSG00000125898, ENSG00000187116, ENSG00000248810, ENSG00000158869, ENSG00000259225 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000105374, ENSG00000115523, ENSG00000100453, ENSG00000077984, ENSG00000180644, ENSG00000137441, ENSG00000271503, ENSG00000134539, ENSG00000172543, ENSG00000100450 \n",
      "\t   ENSG00000159674, ENSG00000145649, ENSG00000169583, ENSG00000171476, ENSG00000169554, ENSG00000205809, ENSG00000036448, ENSG00000150045, ENSG00000275302, ENSG00000136250 \n",
      "\t   ENSG00000187118, ENSG00000150687, ENSG00000075234, ENSG00000211829, ENSG00000213809, ENSG00000174482, ENSG00000135404, ENSG00000180739, ENSG00000181036, ENSG00000163453 \n",
      "Negative:  ENSG00000227507, ENSG00000168685, ENSG00000137818, ENSG00000089157, ENSG00000111716, ENSG00000171223, ENSG00000152495, ENSG00000109452, ENSG00000172005, ENSG00000165272 \n",
      "\t   ENSG00000081059, ENSG00000138795, ENSG00000026025, ENSG00000142546, ENSG00000164300, ENSG00000182463, ENSG00000151150, ENSG00000170345, ENSG00000139193, ENSG00000126353 \n",
      "\t   ENSG00000178562, ENSG00000115687, ENSG00000211747, ENSG00000110848, ENSG00000163600, ENSG00000121966, ENSG00000079335, ENSG00000182853, ENSG00000152270, ENSG00000165591 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000188536, ENSG00000206172, ENSG00000223609, ENSG00000158578, ENSG00000244734, ENSG00000169877, ENSG00000145335, ENSG00000206177, ENSG00000133742, ENSG00000147454 \n",
      "\t   ENSG00000143416, ENSG00000162722, ENSG00000013306, ENSG00000198876, ENSG00000082146, ENSG00000090013, ENSG00000004939, ENSG00000104765, ENSG00000066926, ENSG00000158856 \n",
      "\t   ENSG00000171552, ENSG00000162366, ENSG00000130830, ENSG00000170315, ENSG00000235169, ENSG00000172331, ENSG00000131981, ENSG00000136929, ENSG00000159346, ENSG00000086506 \n",
      "Negative:  ENSG00000137818, ENSG00000034510, ENSG00000089157, ENSG00000075624, ENSG00000196262, ENSG00000108518, ENSG00000170345, ENSG00000008517, ENSG00000111640, ENSG00000184009 \n",
      "\t   ENSG00000185885, ENSG00000177606, ENSG00000026025, ENSG00000171223, ENSG00000213145, ENSG00000168685, ENSG00000196154, ENSG00000111716, ENSG00000122862, ENSG00000160888 \n",
      "\t   ENSG00000240972, ENSG00000197747, ENSG00000189403, ENSG00000118503, ENSG00000113387, ENSG00000166794, ENSG00000100906, ENSG00000135046, ENSG00000110848, ENSG00000167286 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 1 features requested have zero variance (running reduction without them): ENSG00000282122”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000101439, ENSG00000216490, ENSG00000204472, ENSG00000204482, ENSG00000197249, ENSG00000197766, ENSG00000066336, ENSG00000090382, ENSG00000158869, ENSG00000129226 \n",
      "\t   ENSG00000105835, ENSG00000163131, ENSG00000011600, ENSG00000085265, ENSG00000166927, ENSG00000204103, ENSG00000025708, ENSG00000197746, ENSG00000163220, ENSG00000204287 \n",
      "\t   ENSG00000254087, ENSG00000136826, ENSG00000126759, ENSG00000224397, ENSG00000131042, ENSG00000130066, ENSG00000101336, ENSG00000198502, ENSG00000076944, ENSG00000197405 \n",
      "Negative:  ENSG00000008517, ENSG00000105374, ENSG00000271503, ENSG00000185885, ENSG00000077984, ENSG00000115523, ENSG00000145649, ENSG00000100450, ENSG00000110848, ENSG00000167286 \n",
      "\t   ENSG00000180644, ENSG00000137441, ENSG00000100453, ENSG00000173762, ENSG00000168685, ENSG00000153563, ENSG00000134539, ENSG00000203896, ENSG00000211751, ENSG00000160654 \n",
      "\t   ENSG00000172543, ENSG00000111796, ENSG00000121966, ENSG00000137818, ENSG00000078596, ENSG00000111716, ENSG00000069667, ENSG00000177606, ENSG00000118503, ENSG00000172116 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000105374, ENSG00000115523, ENSG00000077984, ENSG00000271503, ENSG00000108518, ENSG00000075624, ENSG00000180644, ENSG00000196154, ENSG00000137441, ENSG00000145649 \n",
      "\t   ENSG00000100453, ENSG00000203747, ENSG00000100450, ENSG00000185885, ENSG00000185201, ENSG00000008517, ENSG00000122862, ENSG00000134539, ENSG00000172543, ENSG00000011600 \n",
      "\t   ENSG00000159674, ENSG00000155366, ENSG00000184009, ENSG00000142089, ENSG00000197956, ENSG00000135046, ENSG00000168329, ENSG00000275302, ENSG00000158869, ENSG00000171476 \n",
      "Negative:  ENSG00000105369, ENSG00000156738, ENSG00000153064, ENSG00000196735, ENSG00000144218, ENSG00000116191, ENSG00000204287, ENSG00000179344, ENSG00000211899, ENSG00000211592 \n",
      "\t   ENSG00000019582, ENSG00000247982, ENSG00000167483, ENSG00000176533, ENSG00000269404, ENSG00000198502, ENSG00000164330, ENSG00000136573, ENSG00000163534, ENSG00000196092 \n",
      "\t   ENSG00000007312, ENSG00000100721, ENSG00000132185, ENSG00000227507, ENSG00000138639, ENSG00000196628, ENSG00000196126, ENSG00000112182, ENSG00000177455, ENSG00000042980 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000143546, ENSG00000038427, ENSG00000163220, ENSG00000100079, ENSG00000170458, ENSG00000090382, ENSG00000163221, ENSG00000110077, ENSG00000276085, ENSG00000125538 \n",
      "\t   ENSG00000085265, ENSG00000119535, ENSG00000137818, ENSG00000135218, ENSG00000163563, ENSG00000145819, ENSG00000120594, ENSG00000168685, ENSG00000118515, ENSG00000162711 \n",
      "\t   ENSG00000087903, ENSG00000124731, ENSG00000125740, ENSG00000106546, ENSG00000026025, ENSG00000148655, ENSG00000153823, ENSG00000121552, ENSG00000160789, ENSG00000182621 \n",
      "Negative:  ENSG00000203747, ENSG00000182853, ENSG00000197353, ENSG00000224397, ENSG00000155366, ENSG00000166927, ENSG00000105369, ENSG00000185201, ENSG00000156738, ENSG00000075624 \n",
      "\t   ENSG00000166165, ENSG00000007312, ENSG00000100292, ENSG00000131042, ENSG00000129757, ENSG00000104972, ENSG00000170873, ENSG00000198682, ENSG00000287682, ENSG00000153064 \n",
      "\t   ENSG00000175489, ENSG00000148737, ENSG00000125898, ENSG00000142512, ENSG00000111679, ENSG00000231389, ENSG00000108518, ENSG00000261371, ENSG00000142089, ENSG00000105374 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000105374, ENSG00000115523, ENSG00000077984, ENSG00000100453, ENSG00000180644, ENSG00000137441, ENSG00000271503, ENSG00000134539, ENSG00000172543, ENSG00000145649 \n",
      "\t   ENSG00000159674, ENSG00000100450, ENSG00000277632, ENSG00000169554, ENSG00000171476, ENSG00000169583, ENSG00000275302, ENSG00000038427, ENSG00000036448, ENSG00000143546 \n",
      "\t   ENSG00000136250, ENSG00000075234, ENSG00000223865, ENSG00000205809, ENSG00000150045, ENSG00000180739, ENSG00000100079, ENSG00000135404, ENSG00000276085, ENSG00000075624 \n",
      "Negative:  ENSG00000168685, ENSG00000227507, ENSG00000171223, ENSG00000111716, ENSG00000081059, ENSG00000165272, ENSG00000152495, ENSG00000089157, ENSG00000137818, ENSG00000172005 \n",
      "\t   ENSG00000026025, ENSG00000109452, ENSG00000138795, ENSG00000182853, ENSG00000197353, ENSG00000224397, ENSG00000142546, ENSG00000164300, ENSG00000170345, ENSG00000139193 \n",
      "\t   ENSG00000182463, ENSG00000166927, ENSG00000121966, ENSG00000178562, ENSG00000126353, ENSG00000204482, ENSG00000151150, ENSG00000118503, ENSG00000152270, ENSG00000131042 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000143546, ENSG00000038427, ENSG00000163220, ENSG00000090382, ENSG00000276085, ENSG00000125538, ENSG00000170458, ENSG00000100079, ENSG00000277632, ENSG00000085265 \n",
      "\t   ENSG00000163221, ENSG00000110077, ENSG00000177606, ENSG00000011600, ENSG00000245532, ENSG00000112149, ENSG00000143878, ENSG00000136250, ENSG00000087903, ENSG00000163563 \n",
      "\t   ENSG00000120594, ENSG00000118515, ENSG00000151726, ENSG00000119535, ENSG00000145819, ENSG00000162711, ENSG00000216490, ENSG00000106066, ENSG00000105835, ENSG00000170345 \n",
      "Negative:  ENSG00000163736, ENSG00000101162, ENSG00000168497, ENSG00000169704, ENSG00000113140, ENSG00000163737, ENSG00000154146, ENSG00000111644, ENSG00000204420, ENSG00000127920 \n",
      "\t   ENSG00000120885, ENSG00000101335, ENSG00000180573, ENSG00000161911, ENSG00000109272, ENSG00000102804, ENSG00000124491, ENSG00000095303, ENSG00000171611, ENSG00000005961 \n",
      "\t   ENSG00000162722, ENSG00000150681, ENSG00000162366, ENSG00000278828, ENSG00000166091, ENSG00000266412, ENSG00000156265, ENSG00000088726, ENSG00000187699, ENSG00000010278 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 26 features requested have zero variance (running reduction without them): ENSG00000282651, ENSG00000211951, ENSG00000196565, ENSG00000211654, ENSG00000070031, ENSG00000253818, ENSG00000211658, ENSG00000211972, ENSG00000211611, ENSG00000211625, ENSG00000222037, ENSG00000241244, ENSG00000242534, ENSG00000241566, ENSG00000173578, ENSG00000211623, ENSG00000244116, ENSG00000106341, ENSG00000248991, ENSG00000254056, ENSG00000115008, ENSG00000225523, ENSG00000253755, ENSG00000166265, ENSG00000112212, ENSG00000169575”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000085265, ENSG00000197249, ENSG00000066336, ENSG00000101439, ENSG00000216490, ENSG00000101336, ENSG00000090382, ENSG00000129226, ENSG00000158869, ENSG00000197766 \n",
      "\t   ENSG00000131042, ENSG00000085514, ENSG00000165168, ENSG00000163220, ENSG00000170458, ENSG00000204482, ENSG00000011600, ENSG00000143546, ENSG00000163563, ENSG00000140749 \n",
      "\t   ENSG00000128383, ENSG00000030582, ENSG00000121552, ENSG00000168461, ENSG00000172243, ENSG00000116701, ENSG00000110077, ENSG00000018280, ENSG00000204472, ENSG00000120594 \n",
      "Negative:  ENSG00000008517, ENSG00000167286, ENSG00000185885, ENSG00000160654, ENSG00000110848, ENSG00000203896, ENSG00000173762, ENSG00000111716, ENSG00000121966, ENSG00000077984 \n",
      "\t   ENSG00000118503, ENSG00000136732, ENSG00000271503, ENSG00000172543, ENSG00000145649, ENSG00000180644, ENSG00000211751, ENSG00000153563, ENSG00000162894, ENSG00000177606 \n",
      "\t   ENSG00000078596, ENSG00000100450, ENSG00000105374, ENSG00000172673, ENSG00000172116, ENSG00000172183, ENSG00000152495, ENSG00000168685, ENSG00000109452, ENSG00000115523 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000227507, ENSG00000168685, ENSG00000138795, ENSG00000126353, ENSG00000081059, ENSG00000172005, ENSG00000112182, ENSG00000189283, ENSG00000152495, ENSG00000164300 \n",
      "\t   ENSG00000170345, ENSG00000188404, ENSG00000171223, ENSG00000182463, ENSG00000171791, ENSG00000151150, ENSG00000125740, ENSG00000110848, ENSG00000120129, ENSG00000184384 \n",
      "\t   ENSG00000177606, ENSG00000237943, ENSG00000165272, ENSG00000184613, ENSG00000136997, ENSG00000072110, ENSG00000163519, ENSG00000107742, ENSG00000152270, ENSG00000196352 \n",
      "Negative:  ENSG00000105374, ENSG00000077984, ENSG00000271503, ENSG00000180644, ENSG00000145649, ENSG00000100450, ENSG00000172543, ENSG00000115523, ENSG00000137441, ENSG00000100453 \n",
      "\t   ENSG00000100097, ENSG00000134539, ENSG00000153563, ENSG00000108518, ENSG00000223865, ENSG00000205336, ENSG00000196154, ENSG00000185201, ENSG00000196126, ENSG00000171476 \n",
      "\t   ENSG00000203747, ENSG00000168329, ENSG00000150687, ENSG00000231389, ENSG00000213809, ENSG00000275302, ENSG00000075624, ENSG00000181036, ENSG00000276070, ENSG00000172116 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000129757, ENSG00000137818, ENSG00000142512, ENSG00000188290, ENSG00000034510, ENSG00000089157, ENSG00000166927, ENSG00000161921, ENSG00000166165, ENSG00000167996 \n",
      "\t   ENSG00000224397, ENSG00000196262, ENSG00000110079, ENSG00000135047, ENSG00000010671, ENSG00000188820, ENSG00000100292, ENSG00000173372, ENSG00000166428, ENSG00000175489 \n",
      "\t   ENSG00000104894, ENSG00000075624, ENSG00000261222, ENSG00000227507, ENSG00000204482, ENSG00000007312, ENSG00000233392, ENSG00000129007, ENSG00000161929, ENSG00000104904 \n",
      "Negative:  ENSG00000148655, ENSG00000038427, ENSG00000182621, ENSG00000145819, ENSG00000135636, ENSG00000163694, ENSG00000120594, ENSG00000075420, ENSG00000163625, ENSG00000119535 \n",
      "\t   ENSG00000146592, ENSG00000136250, ENSG00000033327, ENSG00000169554, ENSG00000151726, ENSG00000151948, ENSG00000135218, ENSG00000188641, ENSG00000119522, ENSG00000179104 \n",
      "\t   ENSG00000203804, ENSG00000137177, ENSG00000136205, ENSG00000183023, ENSG00000008083, ENSG00000197548, ENSG00000132334, ENSG00000138639, ENSG00000176463, ENSG00000130449 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000008517, ENSG00000185885, ENSG00000196154, ENSG00000197956, ENSG00000034510, ENSG00000167286, ENSG00000168685, ENSG00000026025, ENSG00000111640, ENSG00000204472 \n",
      "\t   ENSG00000170345, ENSG00000111716, ENSG00000137818, ENSG00000108518, ENSG00000197747, ENSG00000075624, ENSG00000163191, ENSG00000160654, ENSG00000172005, ENSG00000102265 \n",
      "\t   ENSG00000010610, ENSG00000165272, ENSG00000171223, ENSG00000100906, ENSG00000204482, ENSG00000138795, ENSG00000224397, ENSG00000118503, ENSG00000173762, ENSG00000081059 \n",
      "Negative:  ENSG00000156738, ENSG00000105369, ENSG00000153064, ENSG00000177455, ENSG00000211899, ENSG00000269404, ENSG00000132185, ENSG00000247982, ENSG00000136573, ENSG00000012124 \n",
      "\t   ENSG00000116191, ENSG00000196092, ENSG00000137101, ENSG00000042980, ENSG00000163534, ENSG00000211592, ENSG00000143297, ENSG00000211898, ENSG00000241106, ENSG00000081189 \n",
      "\t   ENSG00000164330, ENSG00000140968, ENSG00000211897, ENSG00000128815, ENSG00000110777, ENSG00000095585, ENSG00000196735, ENSG00000072694, ENSG00000179344, ENSG00000145088 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000128016, ENSG00000171223, ENSG00000120129, ENSG00000142512, ENSG00000177606, ENSG00000129757, ENSG00000188290, ENSG00000166927, ENSG00000143878, ENSG00000170345 \n",
      "\t   ENSG00000166165, ENSG00000224397, ENSG00000160888, ENSG00000203747, ENSG00000148737, ENSG00000167552, ENSG00000123146, ENSG00000104972, ENSG00000118503, ENSG00000121966 \n",
      "\t   ENSG00000110079, ENSG00000110848, ENSG00000185885, ENSG00000156738, ENSG00000127507, ENSG00000100292, ENSG00000160326, ENSG00000173372, ENSG00000197872, ENSG00000233392 \n",
      "Negative:  ENSG00000163736, ENSG00000168497, ENSG00000113140, ENSG00000101162, ENSG00000163737, ENSG00000120885, ENSG00000154146, ENSG00000127920, ENSG00000278828, ENSG00000204420 \n",
      "\t   ENSG00000180573, ENSG00000124491, ENSG00000156265, ENSG00000109272, ENSG00000111644, ENSG00000161911, ENSG00000101335, ENSG00000095303, ENSG00000169704, ENSG00000254614 \n",
      "\t   ENSG00000162722, ENSG00000259207, ENSG00000124635, ENSG00000108960, ENSG00000088826, ENSG00000145335, ENSG00000005961, ENSG00000283633, ENSG00000166091, ENSG00000171611 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 27 features requested have zero variance (running reduction without them): ENSG00000282651, ENSG00000211652, ENSG00000211655, ENSG00000211942, ENSG00000211654, ENSG00000243063, ENSG00000211639, ENSG00000198178, ENSG00000211665, ENSG00000211623, ENSG00000211974, ENSG00000225698, ENSG00000106341, ENSG00000248991, ENSG00000115718, ENSG00000077522, ENSG00000259995, ENSG00000108244, ENSG00000163092, ENSG00000225523, ENSG00000184292, ENSG00000253755, ENSG00000224607, ENSG00000166265, ENSG00000169760, ENSG00000186081, ENSG00000169575”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000167286, ENSG00000160654, ENSG00000185885, ENSG00000077984, ENSG00000271503, ENSG00000172543, ENSG00000173762, ENSG00000203896, ENSG00000153563 \n",
      "\t   ENSG00000145649, ENSG00000180644, ENSG00000105374, ENSG00000213809, ENSG00000110848, ENSG00000100450, ENSG00000172116, ENSG00000115523, ENSG00000111716, ENSG00000211751 \n",
      "\t   ENSG00000137441, ENSG00000136732, ENSG00000134539, ENSG00000100453, ENSG00000162894, ENSG00000121966, ENSG00000172673, ENSG00000078596, ENSG00000172183, ENSG00000205336 \n",
      "Negative:  ENSG00000216490, ENSG00000101439, ENSG00000085265, ENSG00000066336, ENSG00000197249, ENSG00000129226, ENSG00000090382, ENSG00000165168, ENSG00000197766, ENSG00000101336 \n",
      "\t   ENSG00000163220, ENSG00000158869, ENSG00000140749, ENSG00000163563, ENSG00000131042, ENSG00000172243, ENSG00000204482, ENSG00000120594, ENSG00000085514, ENSG00000116701 \n",
      "\t   ENSG00000143546, ENSG00000030582, ENSG00000126759, ENSG00000136826, ENSG00000204472, ENSG00000104093, ENSG00000121552, ENSG00000011600, ENSG00000105835, ENSG00000170458 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000105374, ENSG00000196154, ENSG00000077984, ENSG00000271503, ENSG00000180644, ENSG00000100450, ENSG00000145649, ENSG00000115523, ENSG00000100097, ENSG00000108518 \n",
      "\t   ENSG00000172543, ENSG00000122862, ENSG00000135046, ENSG00000100453, ENSG00000137441, ENSG00000134539, ENSG00000185201, ENSG00000203747, ENSG00000075624, ENSG00000205336 \n",
      "\t   ENSG00000197956, ENSG00000197747, ENSG00000185885, ENSG00000150687, ENSG00000168329, ENSG00000008517, ENSG00000153563, ENSG00000171476, ENSG00000184009, ENSG00000181036 \n",
      "Negative:  ENSG00000105369, ENSG00000156738, ENSG00000153064, ENSG00000116191, ENSG00000247982, ENSG00000144218, ENSG00000211899, ENSG00000269404, ENSG00000012124, ENSG00000196092 \n",
      "\t   ENSG00000177455, ENSG00000227507, ENSG00000132185, ENSG00000112182, ENSG00000163534, ENSG00000136573, ENSG00000164330, ENSG00000167483, ENSG00000095585, ENSG00000176533 \n",
      "\t   ENSG00000042980, ENSG00000110777, ENSG00000007312, ENSG00000100721, ENSG00000211898, ENSG00000110987, ENSG00000145088, ENSG00000126353, ENSG00000241106, ENSG00000104921 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000223865, ENSG00000196126, ENSG00000231389, ENSG00000156738, ENSG00000105369, ENSG00000153064, ENSG00000196735, ENSG00000019582, ENSG00000177455, ENSG00000247982 \n",
      "\t   ENSG00000269404, ENSG00000105374, ENSG00000198502, ENSG00000211899, ENSG00000196092, ENSG00000132185, ENSG00000012124, ENSG00000204287, ENSG00000203747, ENSG00000179344 \n",
      "\t   ENSG00000116191, ENSG00000007312, ENSG00000077984, ENSG00000164330, ENSG00000136573, ENSG00000163534, ENSG00000115523, ENSG00000180644, ENSG00000100450, ENSG00000271503 \n",
      "Negative:  ENSG00000168685, ENSG00000138795, ENSG00000172005, ENSG00000081059, ENSG00000126353, ENSG00000152495, ENSG00000164300, ENSG00000237943, ENSG00000227507, ENSG00000184613 \n",
      "\t   ENSG00000072110, ENSG00000152270, ENSG00000111716, ENSG00000139193, ENSG00000189283, ENSG00000165272, ENSG00000142546, ENSG00000171791, ENSG00000184384, ENSG00000188404 \n",
      "\t   ENSG00000163519, ENSG00000148908, ENSG00000107742, ENSG00000114737, ENSG00000182463, ENSG00000166681, ENSG00000198520, ENSG00000151150, ENSG00000154229, ENSG00000136997 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000034510, ENSG00000137818, ENSG00000089157, ENSG00000177606, ENSG00000196262, ENSG00000171223, ENSG00000026025, ENSG00000170345, ENSG00000163221, ENSG00000160888 \n",
      "\t   ENSG00000158517, ENSG00000120129, ENSG00000038427, ENSG00000188404, ENSG00000170458, ENSG00000100079, ENSG00000110077, ENSG00000042493, ENSG00000100906, ENSG00000125740 \n",
      "\t   ENSG00000104894, ENSG00000119535, ENSG00000166523, ENSG00000008394, ENSG00000197956, ENSG00000019582, ENSG00000143546, ENSG00000169385, ENSG00000197747, ENSG00000121101 \n",
      "Negative:  ENSG00000163736, ENSG00000113140, ENSG00000168497, ENSG00000101162, ENSG00000154146, ENSG00000095303, ENSG00000163737, ENSG00000127920, ENSG00000120885, ENSG00000158856 \n",
      "\t   ENSG00000111644, ENSG00000162722, ENSG00000180573, ENSG00000061918, ENSG00000169704, ENSG00000204420, ENSG00000259207, ENSG00000166091, ENSG00000149564, ENSG00000184113 \n",
      "\t   ENSG00000278828, ENSG00000185245, ENSG00000101335, ENSG00000124635, ENSG00000047648, ENSG00000161911, ENSG00000108846, ENSG00000108960, ENSG00000088826, ENSG00000150681 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000129757, ENSG00000166165, ENSG00000142512, ENSG00000135047, ENSG00000110079, ENSG00000188290, ENSG00000166927, ENSG00000261222, ENSG00000233392, ENSG00000224397 \n",
      "\t   ENSG00000161921, ENSG00000170956, ENSG00000137757, ENSG00000173372, ENSG00000182578, ENSG00000287682, ENSG00000171954, ENSG00000175445, ENSG00000175489, ENSG00000198682 \n",
      "\t   ENSG00000129007, ENSG00000173369, ENSG00000100292, ENSG00000182853, ENSG00000123685, ENSG00000104974, ENSG00000213889, ENSG00000160326, ENSG00000134072, ENSG00000109103 \n",
      "Negative:  ENSG00000163221, ENSG00000038427, ENSG00000135218, ENSG00000170458, ENSG00000008394, ENSG00000169385, ENSG00000119535, ENSG00000182621, ENSG00000166523, ENSG00000143546 \n",
      "\t   ENSG00000135636, ENSG00000100079, ENSG00000177575, ENSG00000124491, ENSG00000110077, ENSG00000135929, ENSG00000183019, ENSG00000169413, ENSG00000186431, ENSG00000115828 \n",
      "\t   ENSG00000088827, ENSG00000146592, ENSG00000159339, ENSG00000121316, ENSG00000169403, ENSG00000075420, ENSG00000249437, ENSG00000135046, ENSG00000163220, ENSG00000145819 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 6 features requested have zero variance (running reduction without them): ENSG00000282651, ENSG00000282639, ENSG00000154277, ENSG00000106341, ENSG00000229613, ENSG00000253755”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000271503, ENSG00000105374, ENSG00000115523, ENSG00000077984, ENSG00000185885, ENSG00000167286, ENSG00000145649, ENSG00000100450, ENSG00000153563 \n",
      "\t   ENSG00000172543, ENSG00000137441, ENSG00000110848, ENSG00000100453, ENSG00000180644, ENSG00000134539, ENSG00000160654, ENSG00000213809, ENSG00000203896, ENSG00000172116 \n",
      "\t   ENSG00000173762, ENSG00000211751, ENSG00000171476, ENSG00000121966, ENSG00000111537, ENSG00000136732, ENSG00000078596, ENSG00000169583, ENSG00000111716, ENSG00000168685 \n",
      "Negative:  ENSG00000101439, ENSG00000216490, ENSG00000090382, ENSG00000011600, ENSG00000085265, ENSG00000163220, ENSG00000143546, ENSG00000197249, ENSG00000158869, ENSG00000204472 \n",
      "\t   ENSG00000197766, ENSG00000204482, ENSG00000066336, ENSG00000197746, ENSG00000105835, ENSG00000129226, ENSG00000025708, ENSG00000163131, ENSG00000087086, ENSG00000163563 \n",
      "\t   ENSG00000038427, ENSG00000167996, ENSG00000119655, ENSG00000136826, ENSG00000170458, ENSG00000204103, ENSG00000100079, ENSG00000130066, ENSG00000120594, ENSG00000030582 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000244734, ENSG00000188536, ENSG00000206172, ENSG00000158578, ENSG00000273149, ENSG00000223609, ENSG00000147454, ENSG00000145335, ENSG00000269028, ENSG00000206177 \n",
      "\t   ENSG00000198695, ENSG00000162722, ENSG00000133742, ENSG00000169877, ENSG00000168298, ENSG00000198876, ENSG00000013306, ENSG00000143416, ENSG00000104765, ENSG00000082146 \n",
      "\t   ENSG00000163220, ENSG00000004939, ENSG00000267598, ENSG00000090382, ENSG00000143546, ENSG00000158856, ENSG00000038427, ENSG00000066926, ENSG00000090013, ENSG00000153064 \n",
      "Negative:  ENSG00000008517, ENSG00000145649, ENSG00000077984, ENSG00000105374, ENSG00000185885, ENSG00000100450, ENSG00000271503, ENSG00000075624, ENSG00000196154, ENSG00000180644 \n",
      "\t   ENSG00000100453, ENSG00000137441, ENSG00000185201, ENSG00000167286, ENSG00000100097, ENSG00000172543, ENSG00000184009, ENSG00000134539, ENSG00000115523, ENSG00000153563 \n",
      "\t   ENSG00000196262, ENSG00000213145, ENSG00000122862, ENSG00000034510, ENSG00000111640, ENSG00000108518, ENSG00000160654, ENSG00000213809, ENSG00000163041, ENSG00000171476 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000105369, ENSG00000156738, ENSG00000211899, ENSG00000153064, ENSG00000116191, ENSG00000144218, ENSG00000247982, ENSG00000196735, ENSG00000007312, ENSG00000100721 \n",
      "\t   ENSG00000227507, ENSG00000204287, ENSG00000196092, ENSG00000211592, ENSG00000179344, ENSG00000176533, ENSG00000211898, ENSG00000164330, ENSG00000112182, ENSG00000167483 \n",
      "\t   ENSG00000019582, ENSG00000163534, ENSG00000104894, ENSG00000136573, ENSG00000012124, ENSG00000269404, ENSG00000223865, ENSG00000081189, ENSG00000196126, ENSG00000231389 \n",
      "Negative:  ENSG00000115523, ENSG00000244734, ENSG00000105374, ENSG00000188536, ENSG00000206172, ENSG00000271503, ENSG00000158578, ENSG00000196154, ENSG00000163220, ENSG00000143546 \n",
      "\t   ENSG00000145335, ENSG00000223609, ENSG00000147454, ENSG00000077984, ENSG00000090382, ENSG00000135046, ENSG00000100450, ENSG00000137441, ENSG00000038427, ENSG00000206177 \n",
      "\t   ENSG00000162722, ENSG00000108518, ENSG00000133742, ENSG00000169877, ENSG00000090013, ENSG00000011600, ENSG00000198876, ENSG00000100453, ENSG00000169554, ENSG00000075624 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000137818, ENSG00000143546, ENSG00000168685, ENSG00000170345, ENSG00000177606, ENSG00000163221, ENSG00000170458, ENSG00000038427, ENSG00000171223, ENSG00000227507 \n",
      "\t   ENSG00000163220, ENSG00000090382, ENSG00000089157, ENSG00000110077, ENSG00000100079, ENSG00000276085, ENSG00000100906, ENSG00000188404, ENSG00000125740, ENSG00000111716 \n",
      "\t   ENSG00000160888, ENSG00000119535, ENSG00000163563, ENSG00000026025, ENSG00000081059, ENSG00000158517, ENSG00000172005, ENSG00000287979, ENSG00000059804, ENSG00000121101 \n",
      "Negative:  ENSG00000163736, ENSG00000101162, ENSG00000169704, ENSG00000111644, ENSG00000168497, ENSG00000163737, ENSG00000180573, ENSG00000127920, ENSG00000154146, ENSG00000204420 \n",
      "\t   ENSG00000113140, ENSG00000120885, ENSG00000109272, ENSG00000278828, ENSG00000101335, ENSG00000166091, ENSG00000171611, ENSG00000161911, ENSG00000102804, ENSG00000095303 \n",
      "\t   ENSG00000088726, ENSG00000254614, ENSG00000124635, ENSG00000108960, ENSG00000156265, ENSG00000150681, ENSG00000005961, ENSG00000259207, ENSG00000266412, ENSG00000010278 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000166927, ENSG00000203747, ENSG00000224397, ENSG00000129757, ENSG00000197353, ENSG00000204482, ENSG00000155366, ENSG00000173372, ENSG00000148737, ENSG00000175489 \n",
      "\t   ENSG00000131042, ENSG00000166165, ENSG00000142512, ENSG00000135047, ENSG00000170873, ENSG00000233392, ENSG00000197249, ENSG00000173369, ENSG00000104972, ENSG00000198682 \n",
      "\t   ENSG00000161921, ENSG00000182853, ENSG00000197405, ENSG00000100292, ENSG00000182578, ENSG00000085514, ENSG00000197766, ENSG00000204472, ENSG00000261222, ENSG00000101336 \n",
      "Negative:  ENSG00000163736, ENSG00000101162, ENSG00000143546, ENSG00000169704, ENSG00000168497, ENSG00000163221, ENSG00000038427, ENSG00000163737, ENSG00000111644, ENSG00000180573 \n",
      "\t   ENSG00000113140, ENSG00000154146, ENSG00000204420, ENSG00000109272, ENSG00000170458, ENSG00000127920, ENSG00000120885, ENSG00000163220, ENSG00000278828, ENSG00000168685 \n",
      "\t   ENSG00000166091, ENSG00000137818, ENSG00000090382, ENSG00000166681, ENSG00000177606, ENSG00000135218, ENSG00000095303, ENSG00000102804, ENSG00000101335, ENSG00000227507 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 27 features requested have zero variance (running reduction without them): ENSG00000179639, ENSG00000070031, ENSG00000131203, ENSG00000211968, ENSG00000275243, ENSG00000214548, ENSG00000138755, ENSG00000173578, ENSG00000184709, ENSG00000167768, ENSG00000163751, ENSG00000163554, ENSG00000248991, ENSG00000158485, ENSG00000163106, ENSG00000115008, ENSG00000017427, ENSG00000253188, ENSG00000276600, ENSG00000181634, ENSG00000102024, ENSG00000078114, ENSG00000089199, ENSG00000112902, ENSG00000184292, ENSG00000166265, ENSG00000112212”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000077984, ENSG00000145649, ENSG00000172543, ENSG00000271503, ENSG00000105374, ENSG00000180644, ENSG00000100450, ENSG00000134539, ENSG00000115523 \n",
      "\t   ENSG00000100453, ENSG00000167286, ENSG00000173762, ENSG00000160654, ENSG00000213809, ENSG00000137441, ENSG00000203896, ENSG00000169583, ENSG00000171476, ENSG00000139187 \n",
      "\t   ENSG00000275302, ENSG00000132965, ENSG00000181036, ENSG00000110848, ENSG00000153563, ENSG00000211751, ENSG00000078596, ENSG00000159674, ENSG00000205336, ENSG00000089692 \n",
      "Negative:  ENSG00000197249, ENSG00000204472, ENSG00000085265, ENSG00000131042, ENSG00000090382, ENSG00000126759, ENSG00000204103, ENSG00000204482, ENSG00000197405, ENSG00000187116 \n",
      "\t   ENSG00000197766, ENSG00000140749, ENSG00000175489, ENSG00000128383, ENSG00000129226, ENSG00000216490, ENSG00000110079, ENSG00000172243, ENSG00000224397, ENSG00000171051 \n",
      "\t   ENSG00000101439, ENSG00000177575, ENSG00000163220, ENSG00000166927, ENSG00000110077, ENSG00000155465, ENSG00000120594, ENSG00000239998, ENSG00000066336, ENSG00000085514 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000105369, ENSG00000156738, ENSG00000144218, ENSG00000153064, ENSG00000116191, ENSG00000211899, ENSG00000112182, ENSG00000007312, ENSG00000204287, ENSG00000136573 \n",
      "\t   ENSG00000100721, ENSG00000167483, ENSG00000163534, ENSG00000164330, ENSG00000196735, ENSG00000196092, ENSG00000158517, ENSG00000179344, ENSG00000083454, ENSG00000019582 \n",
      "\t   ENSG00000269404, ENSG00000176533, ENSG00000145416, ENSG00000211898, ENSG00000272398, ENSG00000132185, ENSG00000012124, ENSG00000247982, ENSG00000227507, ENSG00000095585 \n",
      "Negative:  ENSG00000122862, ENSG00000105374, ENSG00000008517, ENSG00000271503, ENSG00000172543, ENSG00000077984, ENSG00000196154, ENSG00000145649, ENSG00000180644, ENSG00000100450 \n",
      "\t   ENSG00000115523, ENSG00000185885, ENSG00000108518, ENSG00000100453, ENSG00000134539, ENSG00000135046, ENSG00000111640, ENSG00000075624, ENSG00000197747, ENSG00000203747 \n",
      "\t   ENSG00000184009, ENSG00000173762, ENSG00000197956, ENSG00000163191, ENSG00000167286, ENSG00000137441, ENSG00000125148, ENSG00000136250, ENSG00000117984, ENSG00000115738 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000034510, ENSG00000137818, ENSG00000177606, ENSG00000089157, ENSG00000170345, ENSG00000185885, ENSG00000026025, ENSG00000171223, ENSG00000160932, ENSG00000198502 \n",
      "\t   ENSG00000019582, ENSG00000197956, ENSG00000128016, ENSG00000163191, ENSG00000231389, ENSG00000160888, ENSG00000120129, ENSG00000196126, ENSG00000104894, ENSG00000196154 \n",
      "\t   ENSG00000100911, ENSG00000223865, ENSG00000172183, ENSG00000203747, ENSG00000166794, ENSG00000163131, ENSG00000121966, ENSG00000197747, ENSG00000196262, ENSG00000084207 \n",
      "Negative:  ENSG00000088726, ENSG00000171611, ENSG00000165682, ENSG00000101162, ENSG00000111644, ENSG00000169704, ENSG00000166091, ENSG00000149564, ENSG00000088826, ENSG00000127920 \n",
      "\t   ENSG00000162722, ENSG00000150681, ENSG00000204323, ENSG00000168497, ENSG00000102804, ENSG00000204420, ENSG00000170271, ENSG00000123405, ENSG00000166681, ENSG00000162614 \n",
      "\t   ENSG00000065534, ENSG00000158856, ENSG00000108960, ENSG00000095303, ENSG00000124635, ENSG00000187699, ENSG00000259207, ENSG00000185340, ENSG00000113140, ENSG00000254614 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000168685, ENSG00000172673, ENSG00000152495, ENSG00000113088, ENSG00000081059, ENSG00000172116, ENSG00000107742, ENSG00000109452, ENSG00000139193, ENSG00000153563 \n",
      "\t   ENSG00000164300, ENSG00000165272, ENSG00000144290, ENSG00000079335, ENSG00000172005, ENSG00000113319, ENSG00000184613, ENSG00000227507, ENSG00000089012, ENSG00000163519 \n",
      "\t   ENSG00000138795, ENSG00000178562, ENSG00000184384, ENSG00000172215, ENSG00000151150, ENSG00000152270, ENSG00000160593, ENSG00000182621, ENSG00000184588, ENSG00000143365 \n",
      "Negative:  ENSG00000203747, ENSG00000137441, ENSG00000100453, ENSG00000100097, ENSG00000205809, ENSG00000198502, ENSG00000185201, ENSG00000075624, ENSG00000159674, ENSG00000211829 \n",
      "\t   ENSG00000011600, ENSG00000115523, ENSG00000000938, ENSG00000075234, ENSG00000134532, ENSG00000171476, ENSG00000223865, ENSG00000180644, ENSG00000135404, ENSG00000231389 \n",
      "\t   ENSG00000150687, ENSG00000205336, ENSG00000168329, ENSG00000196126, ENSG00000108518, ENSG00000100450, ENSG00000240403, ENSG00000145287, ENSG00000213145, ENSG00000143297 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000134532, ENSG00000143297, ENSG00000211896, ENSG00000211893, ENSG00000211895, ENSG00000067715, ENSG00000105492, ENSG00000253364, ENSG00000184293, ENSG00000105967 \n",
      "\t   ENSG00000105926, ENSG00000240505, ENSG00000211897, ENSG00000011590, ENSG00000161929, ENSG00000151150, ENSG00000114013, ENSG00000226571, ENSG00000260186, ENSG00000123096 \n",
      "\t   ENSG00000172264, ENSG00000167641, ENSG00000163568, ENSG00000070190, ENSG00000176014, ENSG00000160856, ENSG00000042980, ENSG00000105246, ENSG00000133424, ENSG00000198846 \n",
      "Negative:  ENSG00000100721, ENSG00000128218, ENSG00000257275, ENSG00000132465, ENSG00000167483, ENSG00000272398, ENSG00000034510, ENSG00000188404, ENSG00000106537, ENSG00000211899 \n",
      "\t   ENSG00000211898, ENSG00000184305, ENSG00000211978, ENSG00000227507, ENSG00000163534, ENSG00000188820, ENSG00000104921, ENSG00000102445, ENSG00000251513, ENSG00000082293 \n",
      "\t   ENSG00000147535, ENSG00000151090, ENSG00000173801, ENSG00000109458, ENSG00000169413, ENSG00000112232, ENSG00000170476, ENSG00000160683, ENSG00000197705, ENSG00000136237 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 38 features requested have zero variance (running reduction without them): ENSG00000158578, ENSG00000133742, ENSG00000204010, ENSG00000143416, ENSG00000004939, ENSG00000281990, ENSG00000211654, ENSG00000070031, ENSG00000169245, ENSG00000158488, ENSG00000275243, ENSG00000222037, ENSG00000278857, ENSG00000250361, ENSG00000214548, ENSG00000211974, ENSG00000225698, ENSG00000167768, ENSG00000163751, ENSG00000163554, ENSG00000253709, ENSG00000231912, ENSG00000248991, ENSG00000147488, ENSG00000070159, ENSG00000115718, ENSG00000276600, ENSG00000130300, ENSG00000130720, ENSG00000108244, ENSG00000114948, ENSG00000196104, ENSG00000225523, ENSG00000184292, ENSG00000253755, ENSG00000224607, ENSG00000112212, ENSG00000169575”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000197249, ENSG00000204472, ENSG00000066336, ENSG00000158869, ENSG00000216490, ENSG00000163220, ENSG00000143546, ENSG00000171051, ENSG00000101439, ENSG00000170458 \n",
      "\t   ENSG00000101336, ENSG00000018280, ENSG00000038427, ENSG00000187116, ENSG00000204482, ENSG00000085265, ENSG00000129226, ENSG00000131042, ENSG00000163563, ENSG00000090382 \n",
      "\t   ENSG00000150337, ENSG00000126759, ENSG00000110079, ENSG00000197766, ENSG00000085514, ENSG00000110077, ENSG00000175489, ENSG00000186818, ENSG00000224397, ENSG00000123384 \n",
      "Negative:  ENSG00000008517, ENSG00000177606, ENSG00000167286, ENSG00000110848, ENSG00000180644, ENSG00000077984, ENSG00000145649, ENSG00000271503, ENSG00000160654, ENSG00000173762 \n",
      "\t   ENSG00000185885, ENSG00000100453, ENSG00000172543, ENSG00000100450, ENSG00000153563, ENSG00000121966, ENSG00000172116, ENSG00000089692, ENSG00000203896, ENSG00000128016 \n",
      "\t   ENSG00000172673, ENSG00000172183, ENSG00000078596, ENSG00000213809, ENSG00000115523, ENSG00000183508, ENSG00000171223, ENSG00000118503, ENSG00000139187, ENSG00000132965 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000101162, ENSG00000169704, ENSG00000165682, ENSG00000111644, ENSG00000166091, ENSG00000149564, ENSG00000168497, ENSG00000127920, ENSG00000088726, ENSG00000204420 \n",
      "\t   ENSG00000171611, ENSG00000095303, ENSG00000259207, ENSG00000150681, ENSG00000158856, ENSG00000108839, ENSG00000166681, ENSG00000254614, ENSG00000163737, ENSG00000162614 \n",
      "\t   ENSG00000170271, ENSG00000074416, ENSG00000161911, ENSG00000102804, ENSG00000109272, ENSG00000278828, ENSG00000145335, ENSG00000005961, ENSG00000176783, ENSG00000184702 \n",
      "Negative:  ENSG00000034510, ENSG00000026025, ENSG00000196154, ENSG00000197956, ENSG00000105374, ENSG00000163191, ENSG00000171223, ENSG00000134333, ENSG00000197747, ENSG00000185885 \n",
      "\t   ENSG00000166794, ENSG00000089157, ENSG00000160888, ENSG00000161960, ENSG00000170345, ENSG00000128016, ENSG00000120129, ENSG00000100911, ENSG00000213145, ENSG00000008517 \n",
      "\t   ENSG00000135046, ENSG00000137818, ENSG00000108518, ENSG00000125148, ENSG00000182718, ENSG00000240972, ENSG00000179218, ENSG00000167286, ENSG00000177606, ENSG00000145649 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000122862, ENSG00000105374, ENSG00000196154, ENSG00000077984, ENSG00000008517, ENSG00000271503, ENSG00000145649, ENSG00000172543, ENSG00000167286, ENSG00000180644 \n",
      "\t   ENSG00000075624, ENSG00000100450, ENSG00000100453, ENSG00000185885, ENSG00000125148, ENSG00000111640, ENSG00000108518, ENSG00000135046, ENSG00000197747, ENSG00000160654 \n",
      "\t   ENSG00000185201, ENSG00000197956, ENSG00000117984, ENSG00000173762, ENSG00000184009, ENSG00000153563, ENSG00000089692, ENSG00000137441, ENSG00000159840, ENSG00000134333 \n",
      "Negative:  ENSG00000105369, ENSG00000156738, ENSG00000153064, ENSG00000144218, ENSG00000177455, ENSG00000116191, ENSG00000211899, ENSG00000269404, ENSG00000132185, ENSG00000012124 \n",
      "\t   ENSG00000136573, ENSG00000247982, ENSG00000137101, ENSG00000095585, ENSG00000196092, ENSG00000110777, ENSG00000176533, ENSG00000211898, ENSG00000042980, ENSG00000204287 \n",
      "\t   ENSG00000100721, ENSG00000167483, ENSG00000241106, ENSG00000122986, ENSG00000163534, ENSG00000272398, ENSG00000128218, ENSG00000007312, ENSG00000158517, ENSG00000140968 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000151150, ENSG00000245532, ENSG00000125740, ENSG00000079335, ENSG00000069667, ENSG00000145819, ENSG00000188641, ENSG00000152270, ENSG00000075420, ENSG00000123066 \n",
      "\t   ENSG00000172673, ENSG00000154229, ENSG00000118503, ENSG00000109452, ENSG00000152495, ENSG00000130449, ENSG00000171791, ENSG00000198695, ENSG00000164300, ENSG00000184384 \n",
      "\t   ENSG00000269028, ENSG00000156299, ENSG00000149177, ENSG00000182463, ENSG00000176463, ENSG00000113448, ENSG00000139083, ENSG00000119522, ENSG00000147202, ENSG00000153317 \n",
      "Negative:  ENSG00000108518, ENSG00000196262, ENSG00000075624, ENSG00000137818, ENSG00000184009, ENSG00000111640, ENSG00000231389, ENSG00000198502, ENSG00000019582, ENSG00000034510 \n",
      "\t   ENSG00000198830, ENSG00000089157, ENSG00000166794, ENSG00000196230, ENSG00000223865, ENSG00000104904, ENSG00000163041, ENSG00000185201, ENSG00000240972, ENSG00000132341 \n",
      "\t   ENSG00000084207, ENSG00000196126, ENSG00000117632, ENSG00000005022, ENSG00000100911, ENSG00000087086, ENSG00000158710, ENSG00000160932, ENSG00000164032, ENSG00000130429 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000176890, ENSG00000171848, ENSG00000166803, ENSG00000148773, ENSG00000089685, ENSG00000175063, ENSG00000157456, ENSG00000145386, ENSG00000170312, ENSG00000117632 \n",
      "\t   ENSG00000122952, ENSG00000100526, ENSG00000167900, ENSG00000126787, ENSG00000088325, ENSG00000117399, ENSG00000127564, ENSG00000137804, ENSG00000117724, ENSG00000131747 \n",
      "\t   ENSG00000146670, ENSG00000237649, ENSG00000072571, ENSG00000178562, ENSG00000105011, ENSG00000166851, ENSG00000075218, ENSG00000066279, ENSG00000165272, ENSG00000121211 \n",
      "Negative:  ENSG00000105374, ENSG00000271503, ENSG00000100450, ENSG00000172543, ENSG00000180644, ENSG00000100453, ENSG00000077984, ENSG00000145649, ENSG00000137441, ENSG00000134539 \n",
      "\t   ENSG00000181036, ENSG00000205336, ENSG00000203747, ENSG00000168329, ENSG00000089692, ENSG00000223865, ENSG00000198502, ENSG00000150687, ENSG00000153563, ENSG00000231389 \n",
      "\t   ENSG00000169554, ENSG00000213809, ENSG00000180739, ENSG00000275302, ENSG00000111537, ENSG00000115523, ENSG00000155366, ENSG00000075234, ENSG00000196126, ENSG00000108798 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 28 features requested have zero variance (running reduction without them): ENSG00000211967, ENSG00000243264, ENSG00000211654, ENSG00000243063, ENSG00000239571, ENSG00000211658, ENSG00000211972, ENSG00000211611, ENSG00000241244, ENSG00000211672, ENSG00000242534, ENSG00000211665, ENSG00000241566, ENSG00000211819, ENSG00000184709, ENSG00000254395, ENSG00000244116, ENSG00000282600, ENSG00000211715, ENSG00000254056, ENSG00000127074, ENSG00000226097, ENSG00000230817, ENSG00000026559, ENSG00000229613, ENSG00000251129, ENSG00000225523, ENSG00000211667”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000271503, ENSG00000172673, ENSG00000269028, ENSG00000069667, ENSG00000110848, ENSG00000115523, ENSG00000008517, ENSG00000151150, ENSG00000100351, ENSG00000109452 \n",
      "\t   ENSG00000180644, ENSG00000198846, ENSG00000152495, ENSG00000104490, ENSG00000183508, ENSG00000203896, ENSG00000234663, ENSG00000152270, ENSG00000171791, ENSG00000213809 \n",
      "\t   ENSG00000134539, ENSG00000276241, ENSG00000107742, ENSG00000106415, ENSG00000100453, ENSG00000198695, ENSG00000172116, ENSG00000167286, ENSG00000033170, ENSG00000153563 \n",
      "Negative:  ENSG00000216490, ENSG00000101439, ENSG00000011600, ENSG00000204472, ENSG00000158869, ENSG00000197249, ENSG00000163131, ENSG00000163191, ENSG00000085265, ENSG00000204482 \n",
      "\t   ENSG00000119655, ENSG00000129226, ENSG00000197766, ENSG00000197746, ENSG00000066336, ENSG00000142089, ENSG00000025708, ENSG00000100300, ENSG00000087086, ENSG00000030582 \n",
      "\t   ENSG00000126759, ENSG00000167996, ENSG00000163220, ENSG00000127951, ENSG00000170296, ENSG00000197956, ENSG00000100097, ENSG00000187116, ENSG00000163931, ENSG00000101336 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000185885, ENSG00000196262, ENSG00000089157, ENSG00000184009, ENSG00000108518, ENSG00000137818, ENSG00000167286, ENSG00000008517, ENSG00000111716, ENSG00000240972 \n",
      "\t   ENSG00000111640, ENSG00000160654, ENSG00000166794, ENSG00000132341, ENSG00000243678, ENSG00000075624, ENSG00000180879, ENSG00000160932, ENSG00000145649, ENSG00000078596 \n",
      "\t   ENSG00000185201, ENSG00000005022, ENSG00000164032, ENSG00000113387, ENSG00000105374, ENSG00000077984, ENSG00000100911, ENSG00000034510, ENSG00000158710, ENSG00000198830 \n",
      "Negative:  ENSG00000183023, ENSG00000254087, ENSG00000245532, ENSG00000148655, ENSG00000120594, ENSG00000139132, ENSG00000163694, ENSG00000169554, ENSG00000130449, ENSG00000100077 \n",
      "\t   ENSG00000119522, ENSG00000123066, ENSG00000149177, ENSG00000104093, ENSG00000133812, ENSG00000033327, ENSG00000197555, ENSG00000269028, ENSG00000182247, ENSG00000147459 \n",
      "\t   ENSG00000188641, ENSG00000196782, ENSG00000145819, ENSG00000176463, ENSG00000156052, ENSG00000059377, ENSG00000067208, ENSG00000175471, ENSG00000112531, ENSG00000139083 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000153064, ENSG00000156738, ENSG00000247982, ENSG00000167483, ENSG00000144218, ENSG00000105369, ENSG00000116191, ENSG00000196092, ENSG00000163534, ENSG00000164330 \n",
      "\t   ENSG00000136573, ENSG00000176533, ENSG00000211899, ENSG00000100721, ENSG00000211898, ENSG00000112182, ENSG00000012124, ENSG00000095585, ENSG00000269404, ENSG00000076554 \n",
      "\t   ENSG00000110987, ENSG00000184226, ENSG00000082293, ENSG00000144645, ENSG00000128218, ENSG00000177455, ENSG00000132185, ENSG00000241106, ENSG00000102445, ENSG00000112232 \n",
      "Negative:  ENSG00000271503, ENSG00000136250, ENSG00000105374, ENSG00000115523, ENSG00000180644, ENSG00000172673, ENSG00000008517, ENSG00000069667, ENSG00000115738, ENSG00000134539 \n",
      "\t   ENSG00000203747, ENSG00000122862, ENSG00000100450, ENSG00000100453, ENSG00000245532, ENSG00000077984, ENSG00000276241, ENSG00000145649, ENSG00000168329, ENSG00000244734 \n",
      "\t   ENSG00000196154, ENSG00000109452, ENSG00000131042, ENSG00000188536, ENSG00000167286, ENSG00000158578, ENSG00000160654, ENSG00000148737, ENSG00000172116, ENSG00000206172 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000110203, ENSG00000163221, ENSG00000038427, ENSG00000110077, ENSG00000100079, ENSG00000257764, ENSG00000170458, ENSG00000119535, ENSG00000158517, ENSG00000166523 \n",
      "\t   ENSG00000135218, ENSG00000143546, ENSG00000169403, ENSG00000169385, ENSG00000042493, ENSG00000008394, ENSG00000090382, ENSG00000135046, ENSG00000124731, ENSG00000146592 \n",
      "\t   ENSG00000183019, ENSG00000169413, ENSG00000115828, ENSG00000163220, ENSG00000111275, ENSG00000157168, ENSG00000163563, ENSG00000161944, ENSG00000106546, ENSG00000182621 \n",
      "Negative:  ENSG00000203747, ENSG00000148737, ENSG00000129757, ENSG00000104972, ENSG00000142512, ENSG00000170873, ENSG00000182853, ENSG00000155366, ENSG00000224397, ENSG00000156113 \n",
      "\t   ENSG00000137393, ENSG00000137757, ENSG00000131042, ENSG00000166927, ENSG00000104974, ENSG00000213889, ENSG00000160326, ENSG00000007312, ENSG00000161921, ENSG00000135047 \n",
      "\t   ENSG00000110079, ENSG00000287682, ENSG00000239998, ENSG00000233392, ENSG00000168329, ENSG00000157827, ENSG00000172594, ENSG00000127507, ENSG00000100292, ENSG00000186480 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000269028, ENSG00000245532, ENSG00000008517, ENSG00000188641, ENSG00000203896, ENSG00000123146, ENSG00000167286, ENSG00000123066, ENSG00000160654, ENSG00000136250 \n",
      "\t   ENSG00000145819, ENSG00000149177, ENSG00000231607, ENSG00000172673, ENSG00000176463, ENSG00000145649, ENSG00000105374, ENSG00000148773, ENSG00000171848, ENSG00000153563 \n",
      "\t   ENSG00000169554, ENSG00000069667, ENSG00000062716, ENSG00000166803, ENSG00000173762, ENSG00000172116, ENSG00000184384, ENSG00000118503, ENSG00000147202, ENSG00000117632 \n",
      "Negative:  ENSG00000162722, ENSG00000158856, ENSG00000158578, ENSG00000206172, ENSG00000188536, ENSG00000145335, ENSG00000143416, ENSG00000169877, ENSG00000223609, ENSG00000130830 \n",
      "\t   ENSG00000137267, ENSG00000244734, ENSG00000133742, ENSG00000206177, ENSG00000086506, ENSG00000004939, ENSG00000166947, ENSG00000013306, ENSG00000104765, ENSG00000159346 \n",
      "\t   ENSG00000147454, ENSG00000066926, ENSG00000008441, ENSG00000101162, ENSG00000138449, ENSG00000198876, ENSG00000123405, ENSG00000162366, ENSG00000235169, ENSG00000198336 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 33 features requested have zero variance (running reduction without them): ENSG00000282651, ENSG00000204010, ENSG00000211655, ENSG00000211951, ENSG00000004939, ENSG00000282639, ENSG00000211611, ENSG00000138755, ENSG00000198336, ENSG00000154277, ENSG00000184709, ENSG00000211974, ENSG00000204866, ENSG00000188290, ENSG00000129596, ENSG00000254056, ENSG00000253188, ENSG00000185897, ENSG00000164825, ENSG00000236481, ENSG00000269927, ENSG00000078114, ENSG00000187800, ENSG00000234300, ENSG00000229613, ENSG00000225523, ENSG00000184292, ENSG00000112077, ENSG00000211961, ENSG00000211667, ENSG00000112212, ENSG00000204710, ENSG00000186081”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000115523, ENSG00000069667, ENSG00000271503, ENSG00000203896, ENSG00000110848, ENSG00000100351, ENSG00000173762, ENSG00000168685, ENSG00000172673 \n",
      "\t   ENSG00000180644, ENSG00000198846, ENSG00000167286, ENSG00000109452, ENSG00000213809, ENSG00000151150, ENSG00000160654, ENSG00000152495, ENSG00000269028, ENSG00000081059 \n",
      "\t   ENSG00000185885, ENSG00000104490, ENSG00000171791, ENSG00000211751, ENSG00000077984, ENSG00000134539, ENSG00000100450, ENSG00000074966, ENSG00000107742, ENSG00000100453 \n",
      "Negative:  ENSG00000216490, ENSG00000101439, ENSG00000197249, ENSG00000158869, ENSG00000204472, ENSG00000197766, ENSG00000129226, ENSG00000066336, ENSG00000101336, ENSG00000204482 \n",
      "\t   ENSG00000085265, ENSG00000116701, ENSG00000163563, ENSG00000139318, ENSG00000131042, ENSG00000090382, ENSG00000085514, ENSG00000163131, ENSG00000140749, ENSG00000120594 \n",
      "\t   ENSG00000106066, ENSG00000165168, ENSG00000126759, ENSG00000197746, ENSG00000025708, ENSG00000121552, ENSG00000018280, ENSG00000011600, ENSG00000166927, ENSG00000175489 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000008517, ENSG00000196154, ENSG00000185885, ENSG00000105374, ENSG00000122862, ENSG00000135046, ENSG00000271503, ENSG00000115523, ENSG00000077984, ENSG00000173762 \n",
      "\t   ENSG00000197747, ENSG00000180644, ENSG00000136250, ENSG00000145649, ENSG00000163191, ENSG00000100450, ENSG00000211751, ENSG00000125148, ENSG00000197956, ENSG00000203896 \n",
      "\t   ENSG00000100453, ENSG00000134539, ENSG00000167286, ENSG00000213809, ENSG00000213145, ENSG00000069667, ENSG00000137441, ENSG00000185201, ENSG00000160654, ENSG00000100097 \n",
      "Negative:  ENSG00000153064, ENSG00000156738, ENSG00000105369, ENSG00000167483, ENSG00000144218, ENSG00000116191, ENSG00000247982, ENSG00000196092, ENSG00000163534, ENSG00000211899 \n",
      "\t   ENSG00000082293, ENSG00000136573, ENSG00000100721, ENSG00000164330, ENSG00000104921, ENSG00000176533, ENSG00000138639, ENSG00000042980, ENSG00000007312, ENSG00000012124 \n",
      "\t   ENSG00000095585, ENSG00000128218, ENSG00000081189, ENSG00000211898, ENSG00000177455, ENSG00000102445, ENSG00000082438, ENSG00000144645, ENSG00000241106, ENSG00000269404 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000137818, ENSG00000089157, ENSG00000196262, ENSG00000111716, ENSG00000184009, ENSG00000108518, ENSG00000227507, ENSG00000034510, ENSG00000185885, ENSG00000026025 \n",
      "\t   ENSG00000075624, ENSG00000240972, ENSG00000008517, ENSG00000167286, ENSG00000132341, ENSG00000104894, ENSG00000111640, ENSG00000168685, ENSG00000158710, ENSG00000180879 \n",
      "\t   ENSG00000160654, ENSG00000089220, ENSG00000163041, ENSG00000166794, ENSG00000170315, ENSG00000196154, ENSG00000087086, ENSG00000104904, ENSG00000243678, ENSG00000167996 \n",
      "Negative:  ENSG00000136250, ENSG00000169554, ENSG00000269028, ENSG00000145819, ENSG00000149177, ENSG00000245532, ENSG00000170873, ENSG00000038382, ENSG00000198846, ENSG00000174482 \n",
      "\t   ENSG00000143669, ENSG00000254087, ENSG00000121101, ENSG00000104490, ENSG00000075420, ENSG00000265451, ENSG00000270022, ENSG00000198695, ENSG00000152217, ENSG00000163590 \n",
      "\t   ENSG00000115523, ENSG00000123066, ENSG00000007237, ENSG00000170962, ENSG00000123146, ENSG00000008083, ENSG00000197548, ENSG00000130449, ENSG00000147202, ENSG00000119522 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000203747, ENSG00000155366, ENSG00000129757, ENSG00000142512, ENSG00000224397, ENSG00000105374, ENSG00000108798, ENSG00000148737, ENSG00000170873, ENSG00000166927 \n",
      "\t   ENSG00000197353, ENSG00000166165, ENSG00000185201, ENSG00000259225, ENSG00000168329, ENSG00000287682, ENSG00000233392, ENSG00000248810, ENSG00000161921, ENSG00000198682 \n",
      "\t   ENSG00000105371, ENSG00000107736, ENSG00000137757, ENSG00000100292, ENSG00000077984, ENSG00000160326, ENSG00000142089, ENSG00000110079, ENSG00000261222, ENSG00000120949 \n",
      "Negative:  ENSG00000038427, ENSG00000170458, ENSG00000163221, ENSG00000143546, ENSG00000100079, ENSG00000110077, ENSG00000110203, ENSG00000135218, ENSG00000119535, ENSG00000163220 \n",
      "\t   ENSG00000257764, ENSG00000150337, ENSG00000042493, ENSG00000008394, ENSG00000010327, ENSG00000124491, ENSG00000171051, ENSG00000169403, ENSG00000166523, ENSG00000158517 \n",
      "\t   ENSG00000141505, ENSG00000169385, ENSG00000124731, ENSG00000146592, ENSG00000106546, ENSG00000109743, ENSG00000120738, ENSG00000090382, ENSG00000178695, ENSG00000177575 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000168685, ENSG00000138795, ENSG00000227507, ENSG00000152495, ENSG00000172005, ENSG00000126353, ENSG00000109452, ENSG00000164300, ENSG00000182463, ENSG00000165272 \n",
      "\t   ENSG00000178562, ENSG00000169756, ENSG00000152270, ENSG00000162804, ENSG00000129757, ENSG00000163519, ENSG00000187109, ENSG00000151150, ENSG00000154229, ENSG00000079335 \n",
      "\t   ENSG00000081059, ENSG00000148737, ENSG00000102245, ENSG00000139193, ENSG00000184613, ENSG00000107736, ENSG00000101162, ENSG00000224397, ENSG00000167286, ENSG00000149212 \n",
      "Negative:  ENSG00000077984, ENSG00000115523, ENSG00000105374, ENSG00000100450, ENSG00000145649, ENSG00000100453, ENSG00000180644, ENSG00000271503, ENSG00000137441, ENSG00000134539 \n",
      "\t   ENSG00000172543, ENSG00000135404, ENSG00000275302, ENSG00000205809, ENSG00000171476, ENSG00000150045, ENSG00000213809, ENSG00000187118, ENSG00000211829, ENSG00000075234 \n",
      "\t   ENSG00000223865, ENSG00000180739, ENSG00000169583, ENSG00000181036, ENSG00000136250, ENSG00000240403, ENSG00000211751, ENSG00000276070, ENSG00000181847, ENSG00000196126 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 19 features requested have zero variance (running reduction without them): ENSG00000282651, ENSG00000211655, ENSG00000184709, ENSG00000095932, ENSG00000163751, ENSG00000286848, ENSG00000188290, ENSG00000248991, ENSG00000127074, ENSG00000017427, ENSG00000253188, ENSG00000129173, ENSG00000196542, ENSG00000113389, ENSG00000168078, ENSG00000164112, ENSG00000229613, ENSG00000184292, ENSG00000224607”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000185885, ENSG00000168685, ENSG00000110848, ENSG00000227507, ENSG00000167286, ENSG00000203896, ENSG00000069667, ENSG00000121966, ENSG00000160654 \n",
      "\t   ENSG00000152495, ENSG00000109452, ENSG00000173762, ENSG00000107742, ENSG00000162894, ENSG00000081059, ENSG00000172183, ENSG00000151150, ENSG00000271503, ENSG00000111716 \n",
      "\t   ENSG00000171791, ENSG00000172673, ENSG00000126353, ENSG00000078596, ENSG00000172005, ENSG00000211751, ENSG00000112182, ENSG00000115523, ENSG00000139193, ENSG00000138795 \n",
      "Negative:  ENSG00000216490, ENSG00000101439, ENSG00000158869, ENSG00000085265, ENSG00000197766, ENSG00000197249, ENSG00000204472, ENSG00000090382, ENSG00000129226, ENSG00000011600 \n",
      "\t   ENSG00000066336, ENSG00000163563, ENSG00000204482, ENSG00000163220, ENSG00000025708, ENSG00000140749, ENSG00000163131, ENSG00000197746, ENSG00000116701, ENSG00000106066 \n",
      "\t   ENSG00000101336, ENSG00000139318, ENSG00000143546, ENSG00000165168, ENSG00000120594, ENSG00000121552, ENSG00000126759, ENSG00000030582, ENSG00000170458, ENSG00000085514 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000156738, ENSG00000105369, ENSG00000153064, ENSG00000247982, ENSG00000167483, ENSG00000116191, ENSG00000144218, ENSG00000211899, ENSG00000204287, ENSG00000196735 \n",
      "\t   ENSG00000100721, ENSG00000007312, ENSG00000163534, ENSG00000179344, ENSG00000042980, ENSG00000104921, ENSG00000196092, ENSG00000136573, ENSG00000176533, ENSG00000241106 \n",
      "\t   ENSG00000019582, ENSG00000122986, ENSG00000081189, ENSG00000177455, ENSG00000082438, ENSG00000269404, ENSG00000082293, ENSG00000211592, ENSG00000012124, ENSG00000133789 \n",
      "Negative:  ENSG00000008517, ENSG00000185885, ENSG00000196154, ENSG00000135046, ENSG00000168685, ENSG00000167286, ENSG00000122862, ENSG00000160654, ENSG00000173762, ENSG00000197747 \n",
      "\t   ENSG00000203896, ENSG00000271503, ENSG00000111716, ENSG00000105374, ENSG00000163191, ENSG00000197956, ENSG00000026025, ENSG00000077984, ENSG00000069667, ENSG00000211751 \n",
      "\t   ENSG00000145649, ENSG00000115523, ENSG00000213145, ENSG00000142546, ENSG00000109452, ENSG00000078596, ENSG00000100450, ENSG00000111640, ENSG00000180644, ENSG00000165272 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000137818, ENSG00000034510, ENSG00000089157, ENSG00000026025, ENSG00000196262, ENSG00000196154, ENSG00000240972, ENSG00000185885, ENSG00000197747, ENSG00000197956 \n",
      "\t   ENSG00000008517, ENSG00000213145, ENSG00000171223, ENSG00000227507, ENSG00000111716, ENSG00000160932, ENSG00000104894, ENSG00000166794, ENSG00000185201, ENSG00000108518 \n",
      "\t   ENSG00000243678, ENSG00000132341, ENSG00000100911, ENSG00000163191, ENSG00000180879, ENSG00000135046, ENSG00000167286, ENSG00000161960, ENSG00000166598, ENSG00000160888 \n",
      "Negative:  ENSG00000101162, ENSG00000169704, ENSG00000168497, ENSG00000163737, ENSG00000127920, ENSG00000088726, ENSG00000111644, ENSG00000166091, ENSG00000113140, ENSG00000204420 \n",
      "\t   ENSG00000171611, ENSG00000163736, ENSG00000149564, ENSG00000102804, ENSG00000161911, ENSG00000187699, ENSG00000154146, ENSG00000167644, ENSG00000120885, ENSG00000095303 \n",
      "\t   ENSG00000108960, ENSG00000162722, ENSG00000150681, ENSG00000156265, ENSG00000130830, ENSG00000109272, ENSG00000176783, ENSG00000158856, ENSG00000162614, ENSG00000278828 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000169554, ENSG00000136250, ENSG00000145819, ENSG00000269028, ENSG00000188641, ENSG00000245532, ENSG00000149177, ENSG00000123066, ENSG00000176463, ENSG00000075420 \n",
      "\t   ENSG00000119522, ENSG00000007237, ENSG00000139083, ENSG00000182621, ENSG00000130449, ENSG00000147202, ENSG00000008083, ENSG00000197548, ENSG00000183023, ENSG00000148655 \n",
      "\t   ENSG00000163694, ENSG00000132334, ENSG00000104447, ENSG00000120594, ENSG00000184384, ENSG00000143669, ENSG00000038427, ENSG00000198695, ENSG00000186153, ENSG00000121101 \n",
      "Negative:  ENSG00000137818, ENSG00000108518, ENSG00000075624, ENSG00000196262, ENSG00000089157, ENSG00000184009, ENSG00000034510, ENSG00000104904, ENSG00000167996, ENSG00000087086 \n",
      "\t   ENSG00000158710, ENSG00000163041, ENSG00000104894, ENSG00000227507, ENSG00000111640, ENSG00000185885, ENSG00000170315, ENSG00000160932, ENSG00000185201, ENSG00000240972 \n",
      "\t   ENSG00000026025, ENSG00000111716, ENSG00000130429, ENSG00000170296, ENSG00000196154, ENSG00000132341, ENSG00000243678, ENSG00000180879, ENSG00000166794, ENSG00000213145 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000203747, ENSG00000129757, ENSG00000148737, ENSG00000155366, ENSG00000224397, ENSG00000142512, ENSG00000166927, ENSG00000108798, ENSG00000170873, ENSG00000166165 \n",
      "\t   ENSG00000197353, ENSG00000105374, ENSG00000105371, ENSG00000259225, ENSG00000125898, ENSG00000287682, ENSG00000233392, ENSG00000248810, ENSG00000204634, ENSG00000137757 \n",
      "\t   ENSG00000261371, ENSG00000110079, ENSG00000255398, ENSG00000104972, ENSG00000175489, ENSG00000168329, ENSG00000160326, ENSG00000107736, ENSG00000100292, ENSG00000261222 \n",
      "Negative:  ENSG00000163221, ENSG00000170458, ENSG00000038427, ENSG00000143546, ENSG00000110077, ENSG00000042493, ENSG00000100079, ENSG00000110203, ENSG00000158517, ENSG00000163220 \n",
      "\t   ENSG00000257764, ENSG00000119535, ENSG00000059804, ENSG00000171051, ENSG00000135218, ENSG00000008394, ENSG00000188404, ENSG00000124731, ENSG00000150337, ENSG00000137818 \n",
      "\t   ENSG00000111716, ENSG00000089157, ENSG00000166523, ENSG00000168685, ENSG00000169403, ENSG00000115828, ENSG00000169385, ENSG00000121552, ENSG00000109743, ENSG00000026025 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 35 features requested have zero variance (running reduction without them): ENSG00000211720, ENSG00000211821, ENSG00000253291, ENSG00000211664, ENSG00000211654, ENSG00000256590, ENSG00000253818, ENSG00000211632, ENSG00000198178, ENSG00000275243, ENSG00000225698, ENSG00000174059, ENSG00000204866, ENSG00000248991, ENSG00000129195, ENSG00000112984, ENSG00000136457, ENSG00000276778, ENSG00000184661, ENSG00000115718, ENSG00000163518, ENSG00000130720, ENSG00000164112, ENSG00000162692, ENSG00000105697, ENSG00000231772, ENSG00000167601, ENSG00000229613, ENSG00000100448, ENSG00000225523, ENSG00000184292, ENSG00000253755, ENSG00000211961, ENSG00000211667, ENSG00000169575”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000244734, ENSG00000183508, ENSG00000188536, ENSG00000206172, ENSG00000158578, ENSG00000136732, ENSG00000223609, ENSG00000169877, ENSG00000145335, ENSG00000158856 \n",
      "\t   ENSG00000206177, ENSG00000143416, ENSG00000133742, ENSG00000162722, ENSG00000171552, ENSG00000013306, ENSG00000004939, ENSG00000082146, ENSG00000198876, ENSG00000110848 \n",
      "\t   ENSG00000172331, ENSG00000066926, ENSG00000166947, ENSG00000167815, ENSG00000069667, ENSG00000235169, ENSG00000168685, ENSG00000152495, ENSG00000086506, ENSG00000008517 \n",
      "Negative:  ENSG00000085265, ENSG00000216490, ENSG00000011600, ENSG00000163131, ENSG00000204472, ENSG00000163220, ENSG00000197249, ENSG00000090382, ENSG00000143546, ENSG00000163563 \n",
      "\t   ENSG00000025708, ENSG00000101439, ENSG00000066336, ENSG00000197956, ENSG00000163191, ENSG00000204482, ENSG00000197746, ENSG00000038427, ENSG00000197766, ENSG00000100097 \n",
      "\t   ENSG00000106565, ENSG00000158869, ENSG00000105835, ENSG00000131042, ENSG00000165168, ENSG00000170458, ENSG00000101336, ENSG00000172243, ENSG00000100300, ENSG00000120594 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000169704, ENSG00000127920, ENSG00000101162, ENSG00000168497, ENSG00000204420, ENSG00000163737, ENSG00000111644, ENSG00000171611, ENSG00000113140, ENSG00000154146 \n",
      "\t   ENSG00000166091, ENSG00000163736, ENSG00000102804, ENSG00000005961, ENSG00000161911, ENSG00000088726, ENSG00000150681, ENSG00000109272, ENSG00000180573, ENSG00000120885 \n",
      "\t   ENSG00000166681, ENSG00000095303, ENSG00000259207, ENSG00000176783, ENSG00000127824, ENSG00000156265, ENSG00000101335, ENSG00000278828, ENSG00000140022, ENSG00000064601 \n",
      "Negative:  ENSG00000227507, ENSG00000089157, ENSG00000110848, ENSG00000121966, ENSG00000177606, ENSG00000034510, ENSG00000137818, ENSG00000104894, ENSG00000171223, ENSG00000112182 \n",
      "\t   ENSG00000172183, ENSG00000156738, ENSG00000162894, ENSG00000105369, ENSG00000188404, ENSG00000116191, ENSG00000185885, ENSG00000240972, ENSG00000144218, ENSG00000111716 \n",
      "\t   ENSG00000196735, ENSG00000008517, ENSG00000153064, ENSG00000168685, ENSG00000163534, ENSG00000171791, ENSG00000152495, ENSG00000126353, ENSG00000167483, ENSG00000247982 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000104894, ENSG00000227507, ENSG00000105369, ENSG00000156738, ENSG00000121966, ENSG00000089157, ENSG00000153064, ENSG00000196735, ENSG00000144218, ENSG00000116191 \n",
      "\t   ENSG00000104921, ENSG00000196262, ENSG00000137818, ENSG00000247982, ENSG00000223865, ENSG00000007312, ENSG00000034510, ENSG00000211899, ENSG00000231389, ENSG00000167483 \n",
      "\t   ENSG00000100721, ENSG00000162894, ENSG00000075624, ENSG00000019582, ENSG00000163534, ENSG00000189403, ENSG00000108518, ENSG00000196126, ENSG00000012124, ENSG00000240972 \n",
      "Negative:  ENSG00000244734, ENSG00000188536, ENSG00000158578, ENSG00000206172, ENSG00000147454, ENSG00000223609, ENSG00000169877, ENSG00000143416, ENSG00000206177, ENSG00000198876 \n",
      "\t   ENSG00000133742, ENSG00000004939, ENSG00000013306, ENSG00000145335, ENSG00000066926, ENSG00000090013, ENSG00000104765, ENSG00000172331, ENSG00000166947, ENSG00000171552 \n",
      "\t   ENSG00000204010, ENSG00000235169, ENSG00000131981, ENSG00000082146, ENSG00000008441, ENSG00000060138, ENSG00000198336, ENSG00000086506, ENSG00000158856, ENSG00000250361 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000153064, ENSG00000144218, ENSG00000156738, ENSG00000116191, ENSG00000163534, ENSG00000167483, ENSG00000247982, ENSG00000196735, ENSG00000138639, ENSG00000104921 \n",
      "\t   ENSG00000105369, ENSG00000196092, ENSG00000081189, ENSG00000211899, ENSG00000012124, ENSG00000100721, ENSG00000082293, ENSG00000102445, ENSG00000164330, ENSG00000042980 \n",
      "\t   ENSG00000136573, ENSG00000082438, ENSG00000112232, ENSG00000145416, ENSG00000058091, ENSG00000254087, ENSG00000112149, ENSG00000204287, ENSG00000211898, ENSG00000120278 \n",
      "Negative:  ENSG00000008517, ENSG00000168685, ENSG00000185885, ENSG00000167286, ENSG00000160654, ENSG00000111716, ENSG00000173762, ENSG00000203896, ENSG00000172005, ENSG00000196154 \n",
      "\t   ENSG00000081059, ENSG00000142546, ENSG00000152495, ENSG00000213145, ENSG00000107742, ENSG00000109452, ENSG00000211751, ENSG00000026025, ENSG00000134333, ENSG00000165272 \n",
      "\t   ENSG00000163519, ENSG00000138795, ENSG00000078596, ENSG00000237943, ENSG00000184009, ENSG00000240972, ENSG00000108518, ENSG00000111640, ENSG00000197747, ENSG00000135046 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000170315, ENSG00000206172, ENSG00000188536, ENSG00000169877, ENSG00000223609, ENSG00000158578, ENSG00000013306, ENSG00000104904, ENSG00000143416, ENSG00000185883 \n",
      "\t   ENSG00000198876, ENSG00000133742, ENSG00000206177, ENSG00000087086, ENSG00000147454, ENSG00000066926, ENSG00000145335, ENSG00000004939, ENSG00000090013, ENSG00000172331 \n",
      "\t   ENSG00000131981, ENSG00000136732, ENSG00000082146, ENSG00000104765, ENSG00000166947, ENSG00000266412, ENSG00000170296, ENSG00000244734, ENSG00000235169, ENSG00000204010 \n",
      "Negative:  ENSG00000269028, ENSG00000198695, ENSG00000188641, ENSG00000184384, ENSG00000245532, ENSG00000154229, ENSG00000145819, ENSG00000139083, ENSG00000075420, ENSG00000136250 \n",
      "\t   ENSG00000147202, ENSG00000123066, ENSG00000059804, ENSG00000176463, ENSG00000149177, ENSG00000152270, ENSG00000130449, ENSG00000069667, ENSG00000008083, ENSG00000007237 \n",
      "\t   ENSG00000095015, ENSG00000169554, ENSG00000164300, ENSG00000162946, ENSG00000182621, ENSG00000156052, ENSG00000119522, ENSG00000197548, ENSG00000112531, ENSG00000148655 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 14 features requested have zero variance (running reduction without them): ENSG00000211968, ENSG00000275243, ENSG00000222037, ENSG00000204866, ENSG00000244116, ENSG00000211715, ENSG00000211731, ENSG00000254056, ENSG00000249790, ENSG00000163518, ENSG00000269927, ENSG00000229613, ENSG00000130203, ENSG00000224607”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000269028, ENSG00000101162, ENSG00000163736, ENSG00000182463, ENSG00000227240, ENSG00000244575, ENSG00000211937, ENSG00000168497, ENSG00000198846, ENSG00000211934 \n",
      "\t   ENSG00000211598, ENSG00000163737, ENSG00000239951, ENSG00000211662, ENSG00000204420, ENSG00000211935, ENSG00000172543, ENSG00000113140, ENSG00000211746, ENSG00000143995 \n",
      "\t   ENSG00000130844, ENSG00000127920, ENSG00000255733, ENSG00000211753, ENSG00000211816, ENSG00000253438, ENSG00000249738, ENSG00000180644, ENSG00000242534, ENSG00000143184 \n",
      "Negative:  ENSG00000087086, ENSG00000090382, ENSG00000143546, ENSG00000163220, ENSG00000216490, ENSG00000167996, ENSG00000101439, ENSG00000019582, ENSG00000197746, ENSG00000075624 \n",
      "\t   ENSG00000197956, ENSG00000196154, ENSG00000011600, ENSG00000170345, ENSG00000163131, ENSG00000204287, ENSG00000204472, ENSG00000034510, ENSG00000158869, ENSG00000085265 \n",
      "\t   ENSG00000196126, ENSG00000026025, ENSG00000108518, ENSG00000130066, ENSG00000197747, ENSG00000163191, ENSG00000128016, ENSG00000111640, ENSG00000231389, ENSG00000197249 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000163220, ENSG00000038427, ENSG00000269028, ENSG00000143546, ENSG00000090382, ENSG00000163221, ENSG00000085265, ENSG00000145819, ENSG00000169554, ENSG00000170458 \n",
      "\t   ENSG00000135218, ENSG00000188641, ENSG00000163563, ENSG00000148655, ENSG00000198695, ENSG00000245532, ENSG00000084234, ENSG00000125538, ENSG00000249437, ENSG00000100906 \n",
      "\t   ENSG00000120594, ENSG00000276085, ENSG00000105835, ENSG00000185215, ENSG00000257764, ENSG00000157168, ENSG00000121316, ENSG00000136250, ENSG00000163131, ENSG00000104093 \n",
      "Negative:  ENSG00000231389, ENSG00000019582, ENSG00000204287, ENSG00000196735, ENSG00000223865, ENSG00000253239, ENSG00000196126, ENSG00000244734, ENSG00000132465, ENSG00000243238 \n",
      "\t   ENSG00000278196, ENSG00000089157, ENSG00000156738, ENSG00000211938, ENSG00000270550, ENSG00000211666, ENSG00000179344, ENSG00000211592, ENSG00000101439, ENSG00000137818 \n",
      "\t   ENSG00000227507, ENSG00000104894, ENSG00000105369, ENSG00000198502, ENSG00000239951, ENSG00000211895, ENSG00000206172, ENSG00000153064, ENSG00000075624, ENSG00000184009 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000253239, ENSG00000143546, ENSG00000270550, ENSG00000278196, ENSG00000132465, ENSG00000163220, ENSG00000244734, ENSG00000090382, ENSG00000137818, ENSG00000038427 \n",
      "\t   ENSG00000275063, ENSG00000211895, ENSG00000089157, ENSG00000111640, ENSG00000206172, ENSG00000163221, ENSG00000008517, ENSG00000168685, ENSG00000243238, ENSG00000211938 \n",
      "\t   ENSG00000211677, ENSG00000188536, ENSG00000135046, ENSG00000273149, ENSG00000166794, ENSG00000170458, ENSG00000170476, ENSG00000177606, ENSG00000170345, ENSG00000135218 \n",
      "Negative:  ENSG00000203747, ENSG00000142089, ENSG00000231389, ENSG00000223865, ENSG00000196735, ENSG00000204482, ENSG00000130066, ENSG00000196126, ENSG00000148737, ENSG00000166927 \n",
      "\t   ENSG00000158869, ENSG00000204287, ENSG00000140105, ENSG00000019582, ENSG00000261371, ENSG00000155366, ENSG00000179344, ENSG00000224397, ENSG00000185201, ENSG00000101439 \n",
      "\t   ENSG00000142512, ENSG00000108798, ENSG00000204472, ENSG00000170873, ENSG00000197746, ENSG00000168329, ENSG00000197249, ENSG00000182578, ENSG00000104972, ENSG00000131042 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000196735, ENSG00000223865, ENSG00000179344, ENSG00000204287, ENSG00000019582, ENSG00000196126, ENSG00000231389, ENSG00000106066, ENSG00000273149, ENSG00000090382 \n",
      "\t   ENSG00000128815, ENSG00000198502, ENSG00000140968, ENSG00000101439, ENSG00000144218, ENSG00000204257, ENSG00000038427, ENSG00000160307, ENSG00000163534, ENSG00000269028 \n",
      "\t   ENSG00000156738, ENSG00000131400, ENSG00000211598, ENSG00000125740, ENSG00000089157, ENSG00000242574, ENSG00000158710, ENSG00000137818, ENSG00000131238, ENSG00000048052 \n",
      "Negative:  ENSG00000244734, ENSG00000206172, ENSG00000243238, ENSG00000211938, ENSG00000188536, ENSG00000203747, ENSG00000142089, ENSG00000211895, ENSG00000132465, ENSG00000158869 \n",
      "\t   ENSG00000204482, ENSG00000087086, ENSG00000204472, ENSG00000216490, ENSG00000167996, ENSG00000166927, ENSG00000197249, ENSG00000185201, ENSG00000197746, ENSG00000224397 \n",
      "\t   ENSG00000253239, ENSG00000170873, ENSG00000131042, ENSG00000011600, ENSG00000163131, ENSG00000130066, ENSG00000148737, ENSG00000278196, ENSG00000261371, ENSG00000155366 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000198695, ENSG00000243238, ENSG00000211938, ENSG00000269028, ENSG00000273149, ENSG00000211592, ENSG00000177606, ENSG00000245532, ENSG00000211895, ENSG00000132465 \n",
      "\t   ENSG00000254087, ENSG00000110848, ENSG00000112182, ENSG00000156738, ENSG00000227507, ENSG00000239951, ENSG00000162894, ENSG00000267598, ENSG00000170873, ENSG00000125740 \n",
      "\t   ENSG00000189283, ENSG00000154229, ENSG00000123066, ENSG00000188641, ENSG00000144802, ENSG00000103489, ENSG00000105369, ENSG00000211666, ENSG00000109452, ENSG00000168685 \n",
      "Negative:  ENSG00000101439, ENSG00000253239, ENSG00000090382, ENSG00000196126, ENSG00000270550, ENSG00000278196, ENSG00000231389, ENSG00000223865, ENSG00000019582, ENSG00000204287 \n",
      "\t   ENSG00000106066, ENSG00000143546, ENSG00000075624, ENSG00000084207, ENSG00000244734, ENSG00000111640, ENSG00000198502, ENSG00000163220, ENSG00000167996, ENSG00000163563 \n",
      "\t   ENSG00000170296, ENSG00000179344, ENSG00000197747, ENSG00000206172, ENSG00000163221, ENSG00000188536, ENSG00000085265, ENSG00000160307, ENSG00000216490, ENSG00000087086 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 15 features requested have zero variance (running reduction without them): ENSG00000204010, ENSG00000138755, ENSG00000198336, ENSG00000184709, ENSG00000163554, ENSG00000248991, ENSG00000186185, ENSG00000164825, ENSG00000229613, ENSG00000100448, ENSG00000112077, ENSG00000166265, ENSG00000112212, ENSG00000186081, ENSG00000169575”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000185885, ENSG00000111716, ENSG00000168685, ENSG00000167286, ENSG00000227507, ENSG00000121966, ENSG00000089157, ENSG00000160654, ENSG00000203896 \n",
      "\t   ENSG00000110848, ENSG00000152495, ENSG00000173762, ENSG00000081059, ENSG00000107742, ENSG00000197061, ENSG00000142546, ENSG00000172005, ENSG00000240972, ENSG00000078596 \n",
      "\t   ENSG00000109452, ENSG00000138795, ENSG00000162894, ENSG00000139193, ENSG00000126353, ENSG00000136732, ENSG00000172183, ENSG00000069667, ENSG00000167815, ENSG00000165272 \n",
      "Negative:  ENSG00000216490, ENSG00000090382, ENSG00000101439, ENSG00000085265, ENSG00000163220, ENSG00000066336, ENSG00000143546, ENSG00000163563, ENSG00000116701, ENSG00000120594 \n",
      "\t   ENSG00000197249, ENSG00000038427, ENSG00000204472, ENSG00000172243, ENSG00000158869, ENSG00000121552, ENSG00000170458, ENSG00000183023, ENSG00000204482, ENSG00000105835 \n",
      "\t   ENSG00000185215, ENSG00000129226, ENSG00000165168, ENSG00000254087, ENSG00000101336, ENSG00000163131, ENSG00000119535, ENSG00000025708, ENSG00000106066, ENSG00000204103 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000008517, ENSG00000196154, ENSG00000185885, ENSG00000197956, ENSG00000122862, ENSG00000197747, ENSG00000135046, ENSG00000160654, ENSG00000168685, ENSG00000167286 \n",
      "\t   ENSG00000163191, ENSG00000026025, ENSG00000173762, ENSG00000172543, ENSG00000184009, ENSG00000115738, ENSG00000145649, ENSG00000271503, ENSG00000108518, ENSG00000203896 \n",
      "\t   ENSG00000118503, ENSG00000111716, ENSG00000105374, ENSG00000077984, ENSG00000211751, ENSG00000180644, ENSG00000115523, ENSG00000185201, ENSG00000136250, ENSG00000100450 \n",
      "Negative:  ENSG00000105369, ENSG00000156738, ENSG00000153064, ENSG00000116191, ENSG00000196735, ENSG00000144218, ENSG00000211899, ENSG00000007312, ENSG00000204287, ENSG00000247982 \n",
      "\t   ENSG00000167483, ENSG00000179344, ENSG00000132185, ENSG00000196092, ENSG00000136573, ENSG00000163534, ENSG00000269404, ENSG00000177455, ENSG00000042980, ENSG00000128218 \n",
      "\t   ENSG00000019582, ENSG00000012124, ENSG00000196126, ENSG00000110777, ENSG00000211592, ENSG00000272398, ENSG00000198502, ENSG00000223865, ENSG00000104921, ENSG00000102445 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000168685, ENSG00000227507, ENSG00000172005, ENSG00000081059, ENSG00000152495, ENSG00000111716, ENSG00000137818, ENSG00000138795, ENSG00000167286, ENSG00000089157 \n",
      "\t   ENSG00000126353, ENSG00000142546, ENSG00000164300, ENSG00000109452, ENSG00000188404, ENSG00000139193, ENSG00000107742, ENSG00000165272, ENSG00000189283, ENSG00000182463 \n",
      "\t   ENSG00000163519, ENSG00000010610, ENSG00000184384, ENSG00000184613, ENSG00000178562, ENSG00000026025, ENSG00000102760, ENSG00000171223, ENSG00000148908, ENSG00000151150 \n",
      "Negative:  ENSG00000105374, ENSG00000115523, ENSG00000100453, ENSG00000100450, ENSG00000077984, ENSG00000180644, ENSG00000137441, ENSG00000134539, ENSG00000203747, ENSG00000205809 \n",
      "\t   ENSG00000145649, ENSG00000271503, ENSG00000205336, ENSG00000150045, ENSG00000172543, ENSG00000075234, ENSG00000171476, ENSG00000181036, ENSG00000275302, ENSG00000168329 \n",
      "\t   ENSG00000169583, ENSG00000150687, ENSG00000007350, ENSG00000159674, ENSG00000180739, ENSG00000187118, ENSG00000155366, ENSG00000075624, ENSG00000108518, ENSG00000213809 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000038427, ENSG00000163221, ENSG00000119535, ENSG00000145819, ENSG00000143546, ENSG00000110077, ENSG00000182621, ENSG00000135218, ENSG00000100079, ENSG00000170458 \n",
      "\t   ENSG00000132334, ENSG00000203804, ENSG00000257764, ENSG00000186431, ENSG00000110203, ENSG00000267519, ENSG00000146592, ENSG00000163563, ENSG00000158517, ENSG00000249437 \n",
      "\t   ENSG00000148655, ENSG00000075420, ENSG00000171051, ENSG00000007237, ENSG00000136250, ENSG00000137331, ENSG00000008083, ENSG00000162711, ENSG00000125538, ENSG00000169896 \n",
      "Negative:  ENSG00000129757, ENSG00000166927, ENSG00000188290, ENSG00000182853, ENSG00000224397, ENSG00000197353, ENSG00000166165, ENSG00000287682, ENSG00000233392, ENSG00000100292 \n",
      "\t   ENSG00000142512, ENSG00000175489, ENSG00000259225, ENSG00000123685, ENSG00000148737, ENSG00000170956, ENSG00000104974, ENSG00000135047, ENSG00000198682, ENSG00000179921 \n",
      "\t   ENSG00000120949, ENSG00000105371, ENSG00000171954, ENSG00000131042, ENSG00000125898, ENSG00000261371, ENSG00000129007, ENSG00000248810, ENSG00000110079, ENSG00000182578 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000163736, ENSG00000113140, ENSG00000101162, ENSG00000168497, ENSG00000163737, ENSG00000101335, ENSG00000127920, ENSG00000204420, ENSG00000120885, ENSG00000154146 \n",
      "\t   ENSG00000169704, ENSG00000236304, ENSG00000180573, ENSG00000109272, ENSG00000259207, ENSG00000005961, ENSG00000111644, ENSG00000161911, ENSG00000250334, ENSG00000203618 \n",
      "\t   ENSG00000187699, ENSG00000166091, ENSG00000149564, ENSG00000102804, ENSG00000088726, ENSG00000254614, ENSG00000156265, ENSG00000124491, ENSG00000171611, ENSG00000150681 \n",
      "Negative:  ENSG00000034510, ENSG00000137818, ENSG00000089157, ENSG00000120129, ENSG00000171223, ENSG00000170345, ENSG00000026025, ENSG00000177606, ENSG00000197956, ENSG00000196154 \n",
      "\t   ENSG00000160888, ENSG00000213145, ENSG00000197747, ENSG00000163191, ENSG00000100906, ENSG00000160932, ENSG00000185201, ENSG00000128016, ENSG00000161960, ENSG00000104894 \n",
      "\t   ENSG00000167552, ENSG00000243678, ENSG00000240972, ENSG00000175489, ENSG00000166927, ENSG00000197249, ENSG00000224397, ENSG00000142089, ENSG00000204482, ENSG00000163131 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 14 features requested have zero variance (running reduction without them): ENSG00000211720, ENSG00000275243, ENSG00000204866, ENSG00000231912, ENSG00000145147, ENSG00000164825, ENSG00000164112, ENSG00000236481, ENSG00000165152, ENSG00000085831, ENSG00000100448, ENSG00000251363, ENSG00000112077, ENSG00000169575”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000216490, ENSG00000101439, ENSG00000204472, ENSG00000011600, ENSG00000158869, ENSG00000197249, ENSG00000204482, ENSG00000090382, ENSG00000163131, ENSG00000066336 \n",
      "\t   ENSG00000163220, ENSG00000197766, ENSG00000197746, ENSG00000085265, ENSG00000129226, ENSG00000025708, ENSG00000224397, ENSG00000139318, ENSG00000116701, ENSG00000126759 \n",
      "\t   ENSG00000101336, ENSG00000165168, ENSG00000130066, ENSG00000143546, ENSG00000131042, ENSG00000163931, ENSG00000085514, ENSG00000103187, ENSG00000187116, ENSG00000166927 \n",
      "Negative:  ENSG00000008517, ENSG00000168685, ENSG00000152495, ENSG00000185885, ENSG00000227507, ENSG00000167286, ENSG00000109452, ENSG00000111716, ENSG00000203896, ENSG00000069667 \n",
      "\t   ENSG00000081059, ENSG00000110848, ENSG00000160654, ENSG00000138795, ENSG00000121966, ENSG00000171791, ENSG00000107742, ENSG00000162894, ENSG00000173762, ENSG00000112182 \n",
      "\t   ENSG00000172673, ENSG00000126353, ENSG00000172183, ENSG00000151150, ENSG00000172005, ENSG00000139193, ENSG00000078596, ENSG00000142546, ENSG00000248905, ENSG00000172543 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000104894, ENSG00000105369, ENSG00000231389, ENSG00000137818, ENSG00000089157, ENSG00000223865, ENSG00000196262, ENSG00000156738, ENSG00000196126, ENSG00000204287 \n",
      "\t   ENSG00000196735, ENSG00000153064, ENSG00000019582, ENSG00000179344, ENSG00000198502, ENSG00000007312, ENSG00000034510, ENSG00000204257, ENSG00000116191, ENSG00000108518 \n",
      "\t   ENSG00000087086, ENSG00000247982, ENSG00000170296, ENSG00000110777, ENSG00000136573, ENSG00000243678, ENSG00000184009, ENSG00000108639, ENSG00000167483, ENSG00000111640 \n",
      "Negative:  ENSG00000149177, ENSG00000269028, ENSG00000245532, ENSG00000148737, ENSG00000136250, ENSG00000169554, ENSG00000103489, ENSG00000183023, ENSG00000147459, ENSG00000133812 \n",
      "\t   ENSG00000188641, ENSG00000059377, ENSG00000204634, ENSG00000163590, ENSG00000143669, ENSG00000254087, ENSG00000261371, ENSG00000233392, ENSG00000197321, ENSG00000120594 \n",
      "\t   ENSG00000137486, ENSG00000100077, ENSG00000198695, ENSG00000073910, ENSG00000170873, ENSG00000147202, ENSG00000176463, ENSG00000156052, ENSG00000140678, ENSG00000130449 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000153064, ENSG00000156738, ENSG00000144218, ENSG00000116191, ENSG00000105369, ENSG00000247982, ENSG00000167483, ENSG00000254087, ENSG00000196735, ENSG00000081189 \n",
      "\t   ENSG00000163534, ENSG00000196092, ENSG00000145416, ENSG00000136573, ENSG00000138639, ENSG00000128815, ENSG00000082438, ENSG00000042980, ENSG00000164330, ENSG00000012124 \n",
      "\t   ENSG00000211899, ENSG00000176533, ENSG00000058091, ENSG00000082293, ENSG00000137478, ENSG00000179344, ENSG00000152217, ENSG00000133789, ENSG00000269028, ENSG00000102445 \n",
      "Negative:  ENSG00000196154, ENSG00000008517, ENSG00000185885, ENSG00000108518, ENSG00000184009, ENSG00000168685, ENSG00000075624, ENSG00000197747, ENSG00000197956, ENSG00000111640 \n",
      "\t   ENSG00000167286, ENSG00000163191, ENSG00000137818, ENSG00000111716, ENSG00000026025, ENSG00000087086, ENSG00000034510, ENSG00000185201, ENSG00000160654, ENSG00000102265 \n",
      "\t   ENSG00000196262, ENSG00000089157, ENSG00000148908, ENSG00000142546, ENSG00000213145, ENSG00000100300, ENSG00000173762, ENSG00000167996, ENSG00000130429, ENSG00000104904 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000008517, ENSG00000168685, ENSG00000185885, ENSG00000034510, ENSG00000167286, ENSG00000026025, ENSG00000196154, ENSG00000160654, ENSG00000089157, ENSG00000197747 \n",
      "\t   ENSG00000111716, ENSG00000197956, ENSG00000173762, ENSG00000142546, ENSG00000152495, ENSG00000109452, ENSG00000135046, ENSG00000203896, ENSG00000081059, ENSG00000172005 \n",
      "\t   ENSG00000240972, ENSG00000163191, ENSG00000138795, ENSG00000107742, ENSG00000185201, ENSG00000213145, ENSG00000069667, ENSG00000137818, ENSG00000227507, ENSG00000152270 \n",
      "Negative:  ENSG00000169704, ENSG00000101162, ENSG00000204420, ENSG00000127920, ENSG00000166091, ENSG00000168497, ENSG00000163737, ENSG00000113140, ENSG00000111644, ENSG00000154146 \n",
      "\t   ENSG00000101335, ENSG00000088726, ENSG00000171611, ENSG00000005961, ENSG00000120885, ENSG00000259207, ENSG00000149564, ENSG00000124491, ENSG00000163736, ENSG00000162722 \n",
      "\t   ENSG00000161911, ENSG00000095303, ENSG00000137801, ENSG00000158856, ENSG00000223855, ENSG00000256235, ENSG00000187699, ENSG00000109272, ENSG00000180573, ENSG00000236304 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000203747, ENSG00000129757, ENSG00000166927, ENSG00000224397, ENSG00000182853, ENSG00000155366, ENSG00000148737, ENSG00000197353, ENSG00000170873, ENSG00000142512 \n",
      "\t   ENSG00000287682, ENSG00000166165, ENSG00000100292, ENSG00000261371, ENSG00000125898, ENSG00000233392, ENSG00000104972, ENSG00000137393, ENSG00000259225, ENSG00000089127 \n",
      "\t   ENSG00000160326, ENSG00000135047, ENSG00000175489, ENSG00000131042, ENSG00000142089, ENSG00000168329, ENSG00000123685, ENSG00000204482, ENSG00000108798, ENSG00000170956 \n",
      "Negative:  ENSG00000038427, ENSG00000143546, ENSG00000163221, ENSG00000110077, ENSG00000100079, ENSG00000119535, ENSG00000170458, ENSG00000163220, ENSG00000085265, ENSG00000182621 \n",
      "\t   ENSG00000090382, ENSG00000135218, ENSG00000163563, ENSG00000110203, ENSG00000137331, ENSG00000277632, ENSG00000153823, ENSG00000158517, ENSG00000124731, ENSG00000146592 \n",
      "\t   ENSG00000141505, ENSG00000132334, ENSG00000145819, ENSG00000100906, ENSG00000249437, ENSG00000111275, ENSG00000007237, ENSG00000125538, ENSG00000162711, ENSG00000276085 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 15 features requested have zero variance (running reduction without them): ENSG00000211976, ENSG00000253818, ENSG00000211891, ENSG00000222037, ENSG00000211665, ENSG00000163751, ENSG00000174059, ENSG00000129596, ENSG00000102854, ENSG00000077522, ENSG00000113389, ENSG00000225523, ENSG00000253755, ENSG00000211961, ENSG00000224607”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000105374, ENSG00000271503, ENSG00000077984, ENSG00000185885, ENSG00000172543, ENSG00000145649, ENSG00000180644, ENSG00000100450, ENSG00000213809 \n",
      "\t   ENSG00000137441, ENSG00000100453, ENSG00000173762, ENSG00000134539, ENSG00000115523, ENSG00000078596, ENSG00000167286, ENSG00000153563, ENSG00000160654, ENSG00000205336 \n",
      "\t   ENSG00000203896, ENSG00000181036, ENSG00000276241, ENSG00000075234, ENSG00000181847, ENSG00000275302, ENSG00000172116, ENSG00000211751, ENSG00000139187, ENSG00000168329 \n",
      "Negative:  ENSG00000085265, ENSG00000101439, ENSG00000204472, ENSG00000090382, ENSG00000216490, ENSG00000197249, ENSG00000158869, ENSG00000163220, ENSG00000204482, ENSG00000066336 \n",
      "\t   ENSG00000165168, ENSG00000143546, ENSG00000163563, ENSG00000120594, ENSG00000116701, ENSG00000038427, ENSG00000129226, ENSG00000105835, ENSG00000170458, ENSG00000197766 \n",
      "\t   ENSG00000183023, ENSG00000140749, ENSG00000172322, ENSG00000113269, ENSG00000100079, ENSG00000104093, ENSG00000121552, ENSG00000106066, ENSG00000163131, ENSG00000110077 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000122862, ENSG00000135046, ENSG00000196154, ENSG00000105374, ENSG00000008517, ENSG00000271503, ENSG00000077984, ENSG00000172543, ENSG00000180644, ENSG00000197747 \n",
      "\t   ENSG00000145649, ENSG00000100450, ENSG00000137441, ENSG00000108518, ENSG00000100453, ENSG00000136250, ENSG00000197956, ENSG00000185885, ENSG00000075624, ENSG00000134539 \n",
      "\t   ENSG00000163191, ENSG00000115523, ENSG00000185201, ENSG00000168329, ENSG00000203747, ENSG00000111640, ENSG00000117984, ENSG00000173762, ENSG00000205336, ENSG00000213809 \n",
      "Negative:  ENSG00000156738, ENSG00000105369, ENSG00000153064, ENSG00000116191, ENSG00000211899, ENSG00000144218, ENSG00000136573, ENSG00000007312, ENSG00000132185, ENSG00000247982 \n",
      "\t   ENSG00000177455, ENSG00000269404, ENSG00000012124, ENSG00000196092, ENSG00000163534, ENSG00000100721, ENSG00000167483, ENSG00000164330, ENSG00000112182, ENSG00000095585 \n",
      "\t   ENSG00000211898, ENSG00000110777, ENSG00000211592, ENSG00000241106, ENSG00000227507, ENSG00000128218, ENSG00000083454, ENSG00000176533, ENSG00000145088, ENSG00000143297 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000163221, ENSG00000038427, ENSG00000143546, ENSG00000170458, ENSG00000135218, ENSG00000119535, ENSG00000182621, ENSG00000163220, ENSG00000110077, ENSG00000166523 \n",
      "\t   ENSG00000177606, ENSG00000249437, ENSG00000145819, ENSG00000059804, ENSG00000072110, ENSG00000123405, ENSG00000146592, ENSG00000124491, ENSG00000008394, ENSG00000162444 \n",
      "\t   ENSG00000163563, ENSG00000163736, ENSG00000267519, ENSG00000186431, ENSG00000101162, ENSG00000121316, ENSG00000104918, ENSG00000124731, ENSG00000169429, ENSG00000151726 \n",
      "Negative:  ENSG00000129757, ENSG00000203747, ENSG00000182853, ENSG00000135047, ENSG00000142512, ENSG00000104972, ENSG00000185201, ENSG00000166927, ENSG00000182578, ENSG00000161921 \n",
      "\t   ENSG00000224397, ENSG00000174837, ENSG00000100292, ENSG00000175489, ENSG00000155366, ENSG00000148737, ENSG00000142089, ENSG00000104974, ENSG00000131042, ENSG00000173372 \n",
      "\t   ENSG00000170873, ENSG00000231389, ENSG00000287682, ENSG00000188290, ENSG00000160326, ENSG00000108798, ENSG00000157827, ENSG00000197353, ENSG00000137757, ENSG00000108518 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000101162, ENSG00000163736, ENSG00000168497, ENSG00000127920, ENSG00000169704, ENSG00000113140, ENSG00000204420, ENSG00000111644, ENSG00000163737, ENSG00000101335 \n",
      "\t   ENSG00000154146, ENSG00000166091, ENSG00000161911, ENSG00000171611, ENSG00000120885, ENSG00000088726, ENSG00000180573, ENSG00000095303, ENSG00000166681, ENSG00000005961 \n",
      "\t   ENSG00000109272, ENSG00000108960, ENSG00000259207, ENSG00000124491, ENSG00000158856, ENSG00000150681, ENSG00000102804, ENSG00000187699, ENSG00000138449, ENSG00000162722 \n",
      "Negative:  ENSG00000137818, ENSG00000034510, ENSG00000163221, ENSG00000038427, ENSG00000177606, ENSG00000089157, ENSG00000143546, ENSG00000170458, ENSG00000163220, ENSG00000135046 \n",
      "\t   ENSG00000119535, ENSG00000163563, ENSG00000249437, ENSG00000110077, ENSG00000166523, ENSG00000182621, ENSG00000151726, ENSG00000104918, ENSG00000162444, ENSG00000197956 \n",
      "\t   ENSG00000121552, ENSG00000019582, ENSG00000146592, ENSG00000026025, ENSG00000121316, ENSG00000158517, ENSG00000125740, ENSG00000197747, ENSG00000148655, ENSG00000090382 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000168685, ENSG00000167286, ENSG00000107742, ENSG00000152495, ENSG00000113088, ENSG00000165272, ENSG00000139193, ENSG00000160654, ENSG00000109452, ENSG00000138795 \n",
      "\t   ENSG00000227507, ENSG00000178562, ENSG00000172116, ENSG00000172673, ENSG00000081059, ENSG00000172005, ENSG00000163519, ENSG00000182463, ENSG00000153563, ENSG00000102245 \n",
      "\t   ENSG00000163600, ENSG00000129757, ENSG00000121966, ENSG00000139187, ENSG00000151150, ENSG00000165591, ENSG00000010610, ENSG00000171223, ENSG00000126353, ENSG00000182853 \n",
      "Negative:  ENSG00000205809, ENSG00000240403, ENSG00000075234, ENSG00000100453, ENSG00000115523, ENSG00000137441, ENSG00000134539, ENSG00000203747, ENSG00000125498, ENSG00000211829 \n",
      "\t   ENSG00000150045, ENSG00000198574, ENSG00000171476, ENSG00000180644, ENSG00000205336, ENSG00000276241, ENSG00000211751, ENSG00000172543, ENSG00000243772, ENSG00000011600 \n",
      "\t   ENSG00000159674, ENSG00000169583, ENSG00000181036, ENSG00000174482, ENSG00000100450, ENSG00000150687, ENSG00000180739, ENSG00000169554, ENSG00000136250, ENSG00000163453 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 52 features requested have zero variance (running reduction without them): ENSG00000211720, ENSG00000282651, ENSG00000276819, ENSG00000224650, ENSG00000211655, ENSG00000242766, ENSG00000253291, ENSG00000211717, ENSG00000211816, ENSG00000211638, ENSG00000282639, ENSG00000253818, ENSG00000211658, ENSG00000211813, ENSG00000211611, ENSG00000275243, ENSG00000126787, ENSG00000211672, ENSG00000211665, ENSG00000241566, ENSG00000211623, ENSG00000211974, ENSG00000225698, ENSG00000254395, ENSG00000204866, ENSG00000282600, ENSG00000211731, ENSG00000129195, ENSG00000254056, ENSG00000102854, ENSG00000125144, ENSG00000127074, ENSG00000017427, ENSG00000276778, ENSG00000065328, ENSG00000186185, ENSG00000115718, ENSG00000162576, ENSG00000163518, ENSG00000130720, ENSG00000164825, ENSG00000102024, ENSG00000168078, ENSG00000164112, ENSG00000242736, ENSG00000026559, ENSG00000100448, ENSG00000251363, ENSG00000225523, ENSG00000138135, ENSG00000253755, ENSG00000211961”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000269028, ENSG00000115523, ENSG00000198846, ENSG00000104490, ENSG00000172673, ENSG00000069667, ENSG00000276241, ENSG00000271503, ENSG00000100351, ENSG00000180644 \n",
      "\t   ENSG00000205336, ENSG00000109452, ENSG00000198695, ENSG00000152495, ENSG00000205809, ENSG00000172543, ENSG00000171791, ENSG00000183508, ENSG00000174482, ENSG00000008517 \n",
      "\t   ENSG00000151150, ENSG00000244734, ENSG00000234663, ENSG00000100453, ENSG00000134539, ENSG00000168685, ENSG00000155657, ENSG00000170962, ENSG00000173068, ENSG00000206172 \n",
      "Negative:  ENSG00000163220, ENSG00000090382, ENSG00000085265, ENSG00000216490, ENSG00000143546, ENSG00000011600, ENSG00000204472, ENSG00000197249, ENSG00000101439, ENSG00000170458 \n",
      "\t   ENSG00000158869, ENSG00000163131, ENSG00000163563, ENSG00000163191, ENSG00000204482, ENSG00000100300, ENSG00000140749, ENSG00000038427, ENSG00000100079, ENSG00000197746 \n",
      "\t   ENSG00000197956, ENSG00000170345, ENSG00000196154, ENSG00000165168, ENSG00000110077, ENSG00000121552, ENSG00000163221, ENSG00000030582, ENSG00000119655, ENSG00000104918 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000169704, ENSG00000111644, ENSG00000101162, ENSG00000168497, ENSG00000204420, ENSG00000166091, ENSG00000163737, ENSG00000113140, ENSG00000127920, ENSG00000171611 \n",
      "\t   ENSG00000154146, ENSG00000163736, ENSG00000124491, ENSG00000120885, ENSG00000005961, ENSG00000088726, ENSG00000161911, ENSG00000095303, ENSG00000150681, ENSG00000259207 \n",
      "\t   ENSG00000137801, ENSG00000149564, ENSG00000064601, ENSG00000166681, ENSG00000108960, ENSG00000156265, ENSG00000187699, ENSG00000101335, ENSG00000223855, ENSG00000127824 \n",
      "Negative:  ENSG00000177606, ENSG00000136250, ENSG00000269028, ENSG00000245532, ENSG00000169554, ENSG00000198695, ENSG00000123066, ENSG00000137478, ENSG00000147202, ENSG00000197555 \n",
      "\t   ENSG00000160888, ENSG00000184384, ENSG00000123146, ENSG00000019582, ENSG00000121101, ENSG00000070961, ENSG00000188641, ENSG00000095015, ENSG00000110848, ENSG00000145819 \n",
      "\t   ENSG00000152270, ENSG00000170345, ENSG00000038382, ENSG00000178209, ENSG00000171223, ENSG00000075420, ENSG00000186153, ENSG00000034510, ENSG00000198846, ENSG00000119522 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000156738, ENSG00000105369, ENSG00000153064, ENSG00000007312, ENSG00000116191, ENSG00000136573, ENSG00000196735, ENSG00000104894, ENSG00000211899, ENSG00000231389 \n",
      "\t   ENSG00000247982, ENSG00000144218, ENSG00000163534, ENSG00000196092, ENSG00000167483, ENSG00000100721, ENSG00000227507, ENSG00000164330, ENSG00000128218, ENSG00000269404 \n",
      "\t   ENSG00000132185, ENSG00000176533, ENSG00000177455, ENSG00000241106, ENSG00000089157, ENSG00000110777, ENSG00000082438, ENSG00000188404, ENSG00000012124, ENSG00000104921 \n",
      "Negative:  ENSG00000136250, ENSG00000245532, ENSG00000169554, ENSG00000183023, ENSG00000145819, ENSG00000120594, ENSG00000188641, ENSG00000148655, ENSG00000038427, ENSG00000104093 \n",
      "\t   ENSG00000059377, ENSG00000185215, ENSG00000075420, ENSG00000182621, ENSG00000163694, ENSG00000147459, ENSG00000119535, ENSG00000149177, ENSG00000148737, ENSG00000176463 \n",
      "\t   ENSG00000267519, ENSG00000132334, ENSG00000153823, ENSG00000018280, ENSG00000123384, ENSG00000186350, ENSG00000123146, ENSG00000196730, ENSG00000175471, ENSG00000105835 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000269028, ENSG00000254087, ENSG00000198695, ENSG00000245532, ENSG00000145416, ENSG00000183023, ENSG00000100077, ENSG00000182247, ENSG00000138639, ENSG00000130066 \n",
      "\t   ENSG00000153064, ENSG00000081189, ENSG00000143669, ENSG00000170873, ENSG00000116191, ENSG00000144218, ENSG00000169554, ENSG00000197555, ENSG00000095015, ENSG00000148655 \n",
      "\t   ENSG00000112149, ENSG00000137478, ENSG00000197872, ENSG00000156738, ENSG00000070961, ENSG00000179583, ENSG00000096968, ENSG00000128815, ENSG00000164330, ENSG00000119522 \n",
      "Negative:  ENSG00000158578, ENSG00000223609, ENSG00000188536, ENSG00000169877, ENSG00000206172, ENSG00000143416, ENSG00000004939, ENSG00000170315, ENSG00000244734, ENSG00000206177 \n",
      "\t   ENSG00000013306, ENSG00000133742, ENSG00000198876, ENSG00000166947, ENSG00000136732, ENSG00000090013, ENSG00000145335, ENSG00000066926, ENSG00000131981, ENSG00000167768 \n",
      "\t   ENSG00000167815, ENSG00000250361, ENSG00000172331, ENSG00000147454, ENSG00000082146, ENSG00000104765, ENSG00000204010, ENSG00000185883, ENSG00000170296, ENSG00000086506 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000008517, ENSG00000185885, ENSG00000108518, ENSG00000034510, ENSG00000167286, ENSG00000168685, ENSG00000196154, ENSG00000185201, ENSG00000105374, ENSG00000184009 \n",
      "\t   ENSG00000115523, ENSG00000160654, ENSG00000240972, ENSG00000111716, ENSG00000271503, ENSG00000173762, ENSG00000137818, ENSG00000196262, ENSG00000089157, ENSG00000111640 \n",
      "\t   ENSG00000166794, ENSG00000269028, ENSG00000077984, ENSG00000109452, ENSG00000197747, ENSG00000180879, ENSG00000197956, ENSG00000145649, ENSG00000172673, ENSG00000152495 \n",
      "Negative:  ENSG00000158578, ENSG00000223609, ENSG00000188536, ENSG00000169877, ENSG00000206172, ENSG00000143416, ENSG00000004939, ENSG00000244734, ENSG00000206177, ENSG00000145335 \n",
      "\t   ENSG00000133742, ENSG00000198876, ENSG00000147454, ENSG00000166947, ENSG00000066926, ENSG00000013306, ENSG00000250361, ENSG00000082146, ENSG00000167768, ENSG00000008441 \n",
      "\t   ENSG00000172331, ENSG00000086506, ENSG00000204010, ENSG00000198336, ENSG00000104765, ENSG00000090013, ENSG00000171552, ENSG00000158856, ENSG00000105610, ENSG00000162722 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 17 features requested have zero variance (running reduction without them): ENSG00000282651, ENSG00000242766, ENSG00000253291, ENSG00000211638, ENSG00000211968, ENSG00000138755, ENSG00000211731, ENSG00000119865, ENSG00000070159, ENSG00000251129, ENSG00000100448, ENSG00000138135, ENSG00000253755, ENSG00000112077, ENSG00000211961, ENSG00000112212, ENSG00000204710”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000269028, ENSG00000172673, ENSG00000198846, ENSG00000069667, ENSG00000104490, ENSG00000109452, ENSG00000115523, ENSG00000152495, ENSG00000100351, ENSG00000110848 \n",
      "\t   ENSG00000234663, ENSG00000276241, ENSG00000198695, ENSG00000151150, ENSG00000106415, ENSG00000152270, ENSG00000171791, ENSG00000112182, ENSG00000185760, ENSG00000139187 \n",
      "\t   ENSG00000213809, ENSG00000168685, ENSG00000205809, ENSG00000033170, ENSG00000152217, ENSG00000234323, ENSG00000079335, ENSG00000170962, ENSG00000174482, ENSG00000144218 \n",
      "Negative:  ENSG00000085265, ENSG00000216490, ENSG00000090382, ENSG00000163220, ENSG00000101439, ENSG00000204472, ENSG00000011600, ENSG00000197249, ENSG00000143546, ENSG00000158869 \n",
      "\t   ENSG00000170458, ENSG00000204482, ENSG00000163563, ENSG00000163131, ENSG00000197746, ENSG00000197766, ENSG00000100079, ENSG00000100300, ENSG00000129226, ENSG00000163191 \n",
      "\t   ENSG00000197956, ENSG00000038427, ENSG00000030582, ENSG00000163931, ENSG00000110077, ENSG00000140749, ENSG00000131981, ENSG00000087086, ENSG00000163221, ENSG00000121552 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000156738, ENSG00000153064, ENSG00000116191, ENSG00000144218, ENSG00000196735, ENSG00000105369, ENSG00000019582, ENSG00000204287, ENSG00000112182, ENSG00000136573 \n",
      "\t   ENSG00000163534, ENSG00000167483, ENSG00000179344, ENSG00000196092, ENSG00000007312, ENSG00000247982, ENSG00000104894, ENSG00000164330, ENSG00000223865, ENSG00000196126 \n",
      "\t   ENSG00000176533, ENSG00000081189, ENSG00000012124, ENSG00000231389, ENSG00000082438, ENSG00000138639, ENSG00000058091, ENSG00000211899, ENSG00000042980, ENSG00000145416 \n",
      "Negative:  ENSG00000169704, ENSG00000101162, ENSG00000168497, ENSG00000204420, ENSG00000111644, ENSG00000163737, ENSG00000113140, ENSG00000127920, ENSG00000171611, ENSG00000163736 \n",
      "\t   ENSG00000166091, ENSG00000120885, ENSG00000154146, ENSG00000101335, ENSG00000005961, ENSG00000161911, ENSG00000124491, ENSG00000088726, ENSG00000095303, ENSG00000109272 \n",
      "\t   ENSG00000259207, ENSG00000158856, ENSG00000150681, ENSG00000137801, ENSG00000149564, ENSG00000184113, ENSG00000138449, ENSG00000065534, ENSG00000101856, ENSG00000064601 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000136250, ENSG00000245532, ENSG00000169554, ENSG00000145819, ENSG00000188641, ENSG00000269028, ENSG00000183023, ENSG00000176463, ENSG00000120594, ENSG00000149177 \n",
      "\t   ENSG00000148655, ENSG00000075420, ENSG00000182621, ENSG00000104093, ENSG00000163694, ENSG00000119522, ENSG00000059377, ENSG00000147459, ENSG00000038427, ENSG00000007237 \n",
      "\t   ENSG00000133812, ENSG00000147202, ENSG00000062716, ENSG00000153823, ENSG00000197548, ENSG00000073910, ENSG00000132334, ENSG00000196730, ENSG00000198695, ENSG00000104490 \n",
      "Negative:  ENSG00000105369, ENSG00000156738, ENSG00000104894, ENSG00000007312, ENSG00000231389, ENSG00000227507, ENSG00000196735, ENSG00000089157, ENSG00000158710, ENSG00000137818 \n",
      "\t   ENSG00000196262, ENSG00000153064, ENSG00000104904, ENSG00000223865, ENSG00000075624, ENSG00000116191, ENSG00000128218, ENSG00000211899, ENSG00000184009, ENSG00000100721 \n",
      "\t   ENSG00000019582, ENSG00000163041, ENSG00000170315, ENSG00000108518, ENSG00000188404, ENSG00000247982, ENSG00000107438, ENSG00000136573, ENSG00000269404, ENSG00000204287 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000153064, ENSG00000144218, ENSG00000081189, ENSG00000116191, ENSG00000138639, ENSG00000145416, ENSG00000254087, ENSG00000164330, ENSG00000058091, ENSG00000182247 \n",
      "\t   ENSG00000163534, ENSG00000156738, ENSG00000196092, ENSG00000128815, ENSG00000197555, ENSG00000137478, ENSG00000112149, ENSG00000144645, ENSG00000100077, ENSG00000136573 \n",
      "\t   ENSG00000167483, ENSG00000082438, ENSG00000112182, ENSG00000082293, ENSG00000048052, ENSG00000176533, ENSG00000042980, ENSG00000138756, ENSG00000120278, ENSG00000112232 \n",
      "Negative:  ENSG00000008517, ENSG00000185885, ENSG00000185201, ENSG00000167286, ENSG00000108518, ENSG00000196154, ENSG00000105374, ENSG00000240972, ENSG00000137818, ENSG00000089157 \n",
      "\t   ENSG00000077984, ENSG00000034510, ENSG00000197747, ENSG00000213145, ENSG00000173762, ENSG00000196262, ENSG00000160654, ENSG00000111716, ENSG00000203896, ENSG00000078596 \n",
      "\t   ENSG00000145649, ENSG00000184009, ENSG00000197956, ENSG00000166794, ENSG00000172543, ENSG00000111640, ENSG00000163191, ENSG00000100911, ENSG00000243678, ENSG00000168685 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000163221, ENSG00000143546, ENSG00000170458, ENSG00000110077, ENSG00000038427, ENSG00000163220, ENSG00000100079, ENSG00000166523, ENSG00000137331, ENSG00000163563 \n",
      "\t   ENSG00000162444, ENSG00000008394, ENSG00000121552, ENSG00000115828, ENSG00000119535, ENSG00000124731, ENSG00000090382, ENSG00000171051, ENSG00000042493, ENSG00000104918 \n",
      "\t   ENSG00000059804, ENSG00000150337, ENSG00000182621, ENSG00000277632, ENSG00000111275, ENSG00000135218, ENSG00000165092, ENSG00000276085, ENSG00000109743, ENSG00000141505 \n",
      "Negative:  ENSG00000203747, ENSG00000148737, ENSG00000129757, ENSG00000155366, ENSG00000233392, ENSG00000170873, ENSG00000142512, ENSG00000182578, ENSG00000224397, ENSG00000204634 \n",
      "\t   ENSG00000287682, ENSG00000133687, ENSG00000198682, ENSG00000182853, ENSG00000166927, ENSG00000104972, ENSG00000197353, ENSG00000261371, ENSG00000166068, ENSG00000138119 \n",
      "\t   ENSG00000185201, ENSG00000157827, ENSG00000229140, ENSG00000131042, ENSG00000135047, ENSG00000160326, ENSG00000166165, ENSG00000108405, ENSG00000104974, ENSG00000137393 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 20 features requested have zero variance (running reduction without them): ENSG00000133742, ENSG00000204010, ENSG00000143416, ENSG00000206177, ENSG00000211655, ENSG00000004939, ENSG00000196565, ENSG00000250361, ENSG00000254046, ENSG00000106341, ENSG00000158485, ENSG00000162631, ENSG00000056736, ENSG00000251363, ENSG00000231412, ENSG00000253755, ENSG00000112077, ENSG00000166342, ENSG00000112212, ENSG00000204710”\n",
      "PC_ 1 \n",
      "Positive:  ENSG00000008517, ENSG00000185885, ENSG00000110848, ENSG00000077984, ENSG00000271503, ENSG00000145649, ENSG00000173762, ENSG00000105374, ENSG00000172543, ENSG00000180644 \n",
      "\t   ENSG00000100450, ENSG00000213809, ENSG00000121966, ENSG00000167286, ENSG00000203896, ENSG00000160654, ENSG00000115523, ENSG00000134539, ENSG00000177606, ENSG00000078596 \n",
      "\t   ENSG00000137441, ENSG00000100453, ENSG00000153563, ENSG00000211751, ENSG00000172116, ENSG00000181847, ENSG00000172183, ENSG00000111716, ENSG00000256039, ENSG00000162894 \n",
      "Negative:  ENSG00000216490, ENSG00000197249, ENSG00000101439, ENSG00000090382, ENSG00000085265, ENSG00000204472, ENSG00000066336, ENSG00000129226, ENSG00000204482, ENSG00000158869 \n",
      "\t   ENSG00000204103, ENSG00000197766, ENSG00000163220, ENSG00000105835, ENSG00000140749, ENSG00000025708, ENSG00000163563, ENSG00000165168, ENSG00000128383, ENSG00000120594 \n",
      "\t   ENSG00000187116, ENSG00000126759, ENSG00000018280, ENSG00000183023, ENSG00000197746, ENSG00000116701, ENSG00000101336, ENSG00000143546, ENSG00000254087, ENSG00000011422 \n",
      "PC_ 2 \n",
      "Positive:  ENSG00000105369, ENSG00000156738, ENSG00000153064, ENSG00000116191, ENSG00000144218, ENSG00000211899, ENSG00000247982, ENSG00000136573, ENSG00000269404, ENSG00000167483 \n",
      "\t   ENSG00000007312, ENSG00000177455, ENSG00000176533, ENSG00000132185, ENSG00000163534, ENSG00000179344, ENSG00000012124, ENSG00000167641, ENSG00000095585, ENSG00000211592 \n",
      "\t   ENSG00000164330, ENSG00000081189, ENSG00000082438, ENSG00000196628, ENSG00000211898, ENSG00000196735, ENSG00000196092, ENSG00000134532, ENSG00000110777, ENSG00000133789 \n",
      "Negative:  ENSG00000105374, ENSG00000008517, ENSG00000271503, ENSG00000077984, ENSG00000122862, ENSG00000185885, ENSG00000196154, ENSG00000145649, ENSG00000180644, ENSG00000100450 \n",
      "\t   ENSG00000115523, ENSG00000172543, ENSG00000108518, ENSG00000134539, ENSG00000135046, ENSG00000075624, ENSG00000137441, ENSG00000100453, ENSG00000011600, ENSG00000203747 \n",
      "\t   ENSG00000136250, ENSG00000213809, ENSG00000173762, ENSG00000197956, ENSG00000115738, ENSG00000197747, ENSG00000142089, ENSG00000185201, ENSG00000117984, ENSG00000169583 \n",
      "PC_ 3 \n",
      "Positive:  ENSG00000168685, ENSG00000138795, ENSG00000227507, ENSG00000126353, ENSG00000081059, ENSG00000172005, ENSG00000111716, ENSG00000167286, ENSG00000152495, ENSG00000107742 \n",
      "\t   ENSG00000165272, ENSG00000160654, ENSG00000139193, ENSG00000142546, ENSG00000164300, ENSG00000089157, ENSG00000109452, ENSG00000137818, ENSG00000038427, ENSG00000182463 \n",
      "\t   ENSG00000059804, ENSG00000188404, ENSG00000072110, ENSG00000143546, ENSG00000163221, ENSG00000178562, ENSG00000166681, ENSG00000148908, ENSG00000170458, ENSG00000152270 \n",
      "Negative:  ENSG00000203747, ENSG00000105374, ENSG00000115523, ENSG00000100453, ENSG00000180644, ENSG00000100450, ENSG00000134539, ENSG00000137441, ENSG00000077984, ENSG00000172543 \n",
      "\t   ENSG00000155366, ENSG00000171476, ENSG00000145649, ENSG00000011600, ENSG00000185201, ENSG00000271503, ENSG00000075234, ENSG00000231389, ENSG00000223865, ENSG00000150045 \n",
      "\t   ENSG00000169583, ENSG00000170873, ENSG00000159674, ENSG00000168329, ENSG00000169554, ENSG00000108798, ENSG00000181036, ENSG00000198574, ENSG00000143878, ENSG00000180739 \n",
      "PC_ 4 \n",
      "Positive:  ENSG00000129757, ENSG00000224397, ENSG00000148737, ENSG00000166927, ENSG00000287682, ENSG00000176890, ENSG00000142512, ENSG00000197353, ENSG00000135047, ENSG00000233392 \n",
      "\t   ENSG00000248810, ENSG00000157827, ENSG00000170956, ENSG00000166803, ENSG00000100292, ENSG00000125898, ENSG00000171848, ENSG00000173372, ENSG00000198682, ENSG00000187109 \n",
      "\t   ENSG00000175489, ENSG00000166165, ENSG00000104974, ENSG00000204634, ENSG00000227507, ENSG00000167900, ENSG00000089685, ENSG00000261371, ENSG00000148773, ENSG00000182853 \n",
      "Negative:  ENSG00000038427, ENSG00000163221, ENSG00000143546, ENSG00000170458, ENSG00000110077, ENSG00000115523, ENSG00000119535, ENSG00000163220, ENSG00000100079, ENSG00000100453 \n",
      "\t   ENSG00000182621, ENSG00000134539, ENSG00000166523, ENSG00000151726, ENSG00000180644, ENSG00000158517, ENSG00000135218, ENSG00000137441, ENSG00000249437, ENSG00000177575 \n",
      "\t   ENSG00000100450, ENSG00000124731, ENSG00000137331, ENSG00000136250, ENSG00000172543, ENSG00000077984, ENSG00000145819, ENSG00000171051, ENSG00000271503, ENSG00000257764 \n",
      "PC_ 5 \n",
      "Positive:  ENSG00000227507, ENSG00000171223, ENSG00000129757, ENSG00000224397, ENSG00000148737, ENSG00000166927, ENSG00000126353, ENSG00000168685, ENSG00000142512, ENSG00000287682 \n",
      "\t   ENSG00000197353, ENSG00000233392, ENSG00000135047, ENSG00000172005, ENSG00000157827, ENSG00000138795, ENSG00000248810, ENSG00000081059, ENSG00000170956, ENSG00000204634 \n",
      "\t   ENSG00000198682, ENSG00000112182, ENSG00000100292, ENSG00000175489, ENSG00000173372, ENSG00000104974, ENSG00000170873, ENSG00000166165, ENSG00000182853, ENSG00000137393 \n",
      "Negative:  ENSG00000176890, ENSG00000171848, ENSG00000166803, ENSG00000167900, ENSG00000148773, ENSG00000089685, ENSG00000137804, ENSG00000286522, ENSG00000127564, ENSG00000145386 \n",
      "\t   ENSG00000117632, ENSG00000175063, ENSG00000092853, ENSG00000178999, ENSG00000122952, ENSG00000117724, ENSG00000066279, ENSG00000088325, ENSG00000131153, ENSG00000184357 \n",
      "\t   ENSG00000139734, ENSG00000131747, ENSG00000189057, ENSG00000237649, ENSG00000276043, ENSG00000170312, ENSG00000157456, ENSG00000075218, ENSG00000126787, ENSG00000105011 \n",
      "\n",
      "Warning message:\n",
      "“Key ‘PC_’ taken, using ‘pca_’ instead”\n",
      "Scaling features for provided objects\n",
      "\n",
      "Computing within dataset neighborhoods\n",
      "\n",
      "Finding all pairwise anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 5674 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 5067 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 4057 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 6973 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 5770 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 15625 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3003 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3245 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2470 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2784 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2673 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2022 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2134 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2243 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 4422 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2744 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2380 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 5065 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 5039 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 4607 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3831 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2111 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1719 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2072 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2533 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1868 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1607 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2056 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1850 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2025 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1826 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2077 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3054 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1784 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2011 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3145 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2194 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1563 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2199 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2233 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2042 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1921 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2439 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1421 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1767 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2415 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2006 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2212 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2467 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2154 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1797 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2096 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1577 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1624 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2965 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2619 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2064 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2707 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3134 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2407 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2012 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2557 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1593 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1737 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3269 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 4700 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1424 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1144 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1577 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1647 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1306 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1281 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1616 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 986 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1129 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2124 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2061 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2507 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 300 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 353 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 577 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 605 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 339 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 400 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 558 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 316 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 376 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 488 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 447 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 563 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 560 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3048 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2564 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3587 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3861 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2671 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2228 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2727 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1789 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1721 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1946 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2756 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3650 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1817 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 418 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2018 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1437 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2554 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2493 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1681 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1506 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2063 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1223 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1219 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2621 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2794 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3436 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2450 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 554 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3057 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3622 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2351 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3186 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3558 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2937 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2747 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3397 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2314 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2137 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2476 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2437 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2715 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1591 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 401 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2791 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1997 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1221 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1004 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1463 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1405 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1165 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1161 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1572 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 930 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 935 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2066 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1665 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1995 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1768 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 500 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1426 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2163 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1587 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1840 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1362 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2078 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2070 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1653 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1686 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2142 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1407 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1413 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2842 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2656 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3139 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2125 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 485 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2066 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3270 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2509 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2623 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3121 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2281 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3261 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3697 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3258 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2917 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3619 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1935 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2068 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2446 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2459 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2760 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1455 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 441 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3088 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2189 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 3763 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 1391 anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 2149 anchors\n",
      "\n",
      "Merging dataset 6 into 5\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 1 into 4\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 11 into 12\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 13 into 16\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 18 into 19\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 2 into 4 1\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 16 13 into 19 18\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 10 into 12 11\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 9 into 8\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 3 into 4 1 2\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 20 into 17\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 12 11 10 into 19 18 16 13\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 5 6 into 7\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 17 20 into 19 18 16 13 12 11 10\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 15 into 4 1 2 3\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 7 5 6 into 19 18 16 13 12 11 10 17 20\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 8 9 into 4 1 2 3 15\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 4 1 2 3 15 8 9 into 19 18 16 13 12 11 10 17 20 7 5 6\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 14 into 19 18 16 13 12 11 10 17 20 7 5 6 4 1 2 3 15 8 9\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Warning message:\n",
      "“The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session”\n",
      "03:45:06 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "03:45:07 Read 246521 rows and found 30 numeric columns\n",
      "\n",
      "03:45:07 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "03:45:07 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "03:45:35 Writing NN index file to temp file /loc/scratch/33598905/RtmpftM7Ky/file43f7551e25ac\n",
      "\n",
      "03:45:35 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "03:47:22 Annoy recall = 100%\n",
      "\n",
      "03:47:36 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "03:48:08 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "03:48:34 Commencing optimization for 200 epochs, with 11977114 positive edges\n",
      "\n",
      "03:51:07 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 246521\n",
      "Number of edges: 9418092\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9311\n",
      "Number of communities: 25\n",
      "Elapsed time: 167 seconds\n"
     ]
    }
   ],
   "source": [
    "pbmc_objects <- list.files(path = seurat_input_path,\n",
    "  pattern = \"V\\\\d+\", full.names = TRUE) |>\n",
    "  map(LoadH5Seurat)|>\n",
    "  map(~subset(.x, subset = discard == 0 & scDblFinder.class == \"singlet\")) |>\n",
    "  map(NormalizeData) |>\n",
    "  map(~FindVariableFeatures(.x, selection.method = \"vst\", nfeatures = 2000)) \n",
    "  \n",
    "pbmc_features <- SelectIntegrationFeatures(object.list = pbmc_objects)\n",
    "\n",
    "pbmc_objects <- pbmc_objects |>\n",
    "  map(~ScaleData(.x, features = pbmc_features)) |>\n",
    "  map(~RunPCA(.x, features = pbmc_features))\n",
    "\n",
    "pbmc_anchors <- FindIntegrationAnchors(object.list = pbmc_objects, \n",
    "  anchor.features = pbmc_features, reduction = \"rpca\")\n",
    "pbmc_integrated <- IntegrateData(anchorset = pbmc_anchors)\n",
    "DefaultAssay(pbmc_integrated) <- \"integrated\"\n",
    "pbmc_integrated <- ScaleData(pbmc_integrated, verbose = FALSE)\n",
    "pbmc_integrated <- RunPCA(pbmc_integrated, npcs = 30, verbose = FALSE)\n",
    "pbmc_integrated <- RunUMAP(pbmc_integrated, reduction = \"pca\", dims = 1:30)\n",
    "pbmc_integrated <- RunTSNE(pbmc_integrated, reduction = \"pca\", dims = 1:30)\n",
    "pbmc_integrated <- FindNeighbors(pbmc_integrated, reduction = \"pca\", dims = 1:30)\n",
    "pbmc_integrated <- FindClusters(pbmc_integrated, resolution = 0.5)\n",
    "DefaultAssay(pbmc_integrated) <- \"RNA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m682\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m42\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (38): cohort, patient_id, repertoire_id, tra_repertoire_id, trb_repertoi...\n",
      "\u001b[32mdbl\u001b[39m  (4): sc_library_count, lession_number, age, fudy\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "pbsample_table <- pbmc_integrated@meta.data |>\n",
    "  as_tibble() |>\n",
    "  dplyr::select(Sample) |>\n",
    "  distinct() \n",
    "\n",
    "pbcohort_table <- read_csv(kstme_metadata_path) |> \n",
    "  filter(!is.na(gex_library)) |> \n",
    "  dplyr::select(patient_id, repertoire_id, hiv_status,response) |> \n",
    "  mutate(cohort = if_else(hiv_status == \"Positive\", \"Epidemic KS\", \"Endemic KS\")) |> \n",
    "  dplyr::select(patient_id, repertoire_id, cohort, response) |>\n",
    "  filter(repertoire_id %in% pbsample_table$Sample) |>\n",
    "  dplyr::select(repertoire_id, cohort, response) |>\n",
    "  mutate(cohort = if_else(is.na(cohort), \"Epidemic KS\", cohort),\n",
    "    response = factor(response, levels = c(\"CR\", \"PR\" ,\"SD\", \"PD\")))\n",
    "\n",
    "cohort_dict <- pbcohort_table |>\n",
    "  pull(cohort) \n",
    "names(cohort_dict) <- pbcohort_table |>\n",
    "  pull(repertoire_id)\n",
    "\n",
    "pbmc_integrated@meta.data$cohort <- cohort_dict[pbmc_integrated@meta.data$Sample]\n",
    "pbmc_tcells <- pbmc_integrated@meta.data |>\n",
    "  as_tibble() |>\n",
    "  mutate(isT = if_else(CTaa == \"NA\", \"Other\", \"Tcell\")) |>\n",
    "  pull(isT)\n",
    "pbmc_integrated@meta.data$is_tcell <- pbmc_tcells\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "14:47:35 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "14:47:35 Read 246521 rows and found 30 numeric columns\n",
      "\n",
      "14:47:35 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'BiocGenerics'\n",
      "\n",
      "Also defined by ‘spam’\n",
      "\n",
      "14:47:35 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "14:47:59 Writing NN index file to temp file /loc/scratch/33598905/RtmpftM7Ky/file43f76579f75\n",
      "\n",
      "14:47:59 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "14:49:32 Annoy recall = 100%\n",
      "\n",
      "14:49:45 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "14:50:13 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "14:50:30 Commencing optimization for 200 epochs, with 11540566 positive edges\n",
      "\n",
      "14:52:58 Optimization finished\n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 246521\n",
      "Number of edges: 6572981\n",
      "\n",
      "Running Louvain algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9421\n",
      "Number of communities: 38\n",
      "Elapsed time: 119 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10 singletons identified. 28 final clusters.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Convert all ensemble IDs to Gene symbols\n",
    "\n",
    "feature_table <- pbmc_integrated[[\"RNA\"]][[]]\n",
    "feature_table$ensembl_id <- rownames(feature_table)\n",
    "master_gene_table <- feature_table |>\n",
    "  dplyr::mutate(symbol = mapIds(org.Hs.eg.db, keys = feature_table$ensembl_id, keytype = \"ENSEMBL\", column = \"SYMBOL\")) |>\n",
    "  filter(!is.na(symbol)) |>\n",
    "  group_by(symbol) |>\n",
    "  slice_head(n =1) |>\n",
    "  ungroup() \n",
    "master_gene_dict <- master_gene_table |>\n",
    "  pull(symbol)\n",
    "names(master_gene_dict) <- master_gene_table |>\n",
    "  pull(ensembl_id)\n",
    "pbmc_gs <- pbmc_integrated[rownames(pbmc_integrated) %in% master_gene_table$ensembl_id, ]\n",
    "pbmc_gsc <- as.SingleCellExperiment(pbmc_integrated)\n",
    "pbmc_gsc <- pbmc_gsc[rownames(pbmc_gsc) %in% master_gene_table$ensembl_id, ]\n",
    "pbmc_gsc <- scuttle::logNormCounts(pbmc_gsc) \n",
    "rownames(pbmc_gsc) <- master_gene_dict[rownames(pbmc_gsc)]\n",
    "pbmc_gs <- CreateSeuratObject(counts = assay(pbmc_gsc, \"counts\"), data = assay(pbmc_gsc, \"logcounts\"))\n",
    "pbmc_gs@meta.data <- colData(pbmc_gsc) |> \n",
    "  as.data.frame()\n",
    "pbmc_gs <- NormalizeData(pbmc_gs)\n",
    "pbmc_gs <- FindVariableFeatures(pbmc_gs, selection.method = \"vst\", nfeatures = 2000)\n",
    "pbmc_gs <- ScaleData(pbmc_gs, verbose = FALSE)\n",
    "pbmc_gs <- RunPCA(pbmc_gs, npcs = 30, verbose = FALSE)\n",
    "pbmc_gs <- RunUMAP(pbmc_gs, reduction = \"pca\", dims = 1:30)\n",
    "pbmc_gs <- RunTSNE(pbmc_gs, reduction = \"pca\", dims = 1:30)\n",
    "pbmc_gs <- FindNeighbors(pbmc_gs, reduction = \"pca\", dims = 1:30)\n",
    "pbmc_gs <- FindClusters(pbmc_gs, resolution = 0.5)\n",
    "DefaultAssay(pbmc_gs) <- \"RNA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Validating h5Seurat file\n",
      "\n",
      "Initializing ADT with data\n",
      "\n",
      "Adding counts for ADT\n",
      "\n",
      "Adding variable feature information for ADT\n",
      "\n",
      "Adding miscellaneous information for ADT\n",
      "\n",
      "Initializing SCT with data\n",
      "\n",
      "Adding counts for SCT\n",
      "\n",
      "Adding variable feature information for SCT\n",
      "\n",
      "Adding miscellaneous information for SCT\n",
      "\n",
      "Adding reduction apca\n",
      "\n",
      "Adding cell embeddings for apca\n",
      "\n",
      "Adding feature loadings for apca\n",
      "\n",
      "Adding miscellaneous information for apca\n",
      "\n",
      "Adding reduction aumap\n",
      "\n",
      "Adding cell embeddings for aumap\n",
      "\n",
      "Adding miscellaneous information for aumap\n",
      "\n",
      "Adding reduction pca\n",
      "\n",
      "Adding cell embeddings for pca\n",
      "\n",
      "Adding feature loadings for pca\n",
      "\n",
      "Adding miscellaneous information for pca\n",
      "\n",
      "Adding reduction spca\n",
      "\n",
      "Adding cell embeddings for spca\n",
      "\n",
      "Adding feature loadings for spca\n",
      "\n",
      "Adding miscellaneous information for spca\n",
      "\n",
      "Adding reduction umap\n",
      "\n",
      "Adding cell embeddings for umap\n",
      "\n",
      "Adding miscellaneous information for umap\n",
      "\n",
      "Adding reduction wnn.umap\n",
      "\n",
      "Adding cell embeddings for wnn.umap\n",
      "\n",
      "Adding miscellaneous information for wnn.umap\n",
      "\n",
      "Adding graph wknn\n",
      "\n",
      "Adding graph wsnn\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Normalizing query using reference SCT model\n",
      "\n",
      "Projecting cell embeddings\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 44679 anchors\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Predicting cell labels\n",
      "\n",
      "Predicting cell labels\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n",
      "Transfering 228 features onto reference data\n",
      "\n",
      "\n",
      "Integrating dataset 2 with reference dataset\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Computing nearest neighbors\n",
      "\n",
      "Running UMAP projection\n",
      "\n",
      "16:33:06 Read 246521 rows\n",
      "\n",
      "16:33:06 Processing block 1 of 1\n",
      "\n",
      "16:33:06 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 15\n",
      "\n",
      "16:33:07 Initializing by weighted average of neighbor coordinates using 1 thread\n",
      "\n",
      "16:33:08 Commencing optimization for 67 epochs, with 3697815 positive edges\n",
      "\n",
      "16:33:32 Finished\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Azimuth classification\n",
    "pbmc_gs <- SCTransform(pbmc_gs, verbose = FALSE)\n",
    "reference <- LoadH5Seurat(azimuth_reference_path)\n",
    "\n",
    "anchors <- FindTransferAnchors(\n",
    "  reference = reference,\n",
    "  query = pbmc_gs,\n",
    "  normalization.method = \"SCT\",\n",
    "  reference.reduction = \"spca\",\n",
    "  dims = 1:50\n",
    ")\n",
    "              \n",
    "pbmc_gs <-  MapQuery(\n",
    "  anchorset = anchors,\n",
    "  query = pbmc_gs,\n",
    "  reference = reference,\n",
    "  refdata = list(\n",
    "    celltype.l1 = \"celltype.l1\",\n",
    "    celltype.l2 = \"celltype.l2\",\n",
    "    predicted_ADT = \"ADT\"\n",
    "  ),\n",
    "  reference.reduction = \"spca\", \n",
    "  reduction.model = \"wnn.umap\"\n",
    ")\n",
    "\n",
    "pbmc_gs$cloneType <- pbmc_gs@meta.data |> \n",
    "  as_tibble() |>\n",
    "  mutate(cloneType = if_else(cloneType == \"NA\", NA_character_, cloneType)) |>\n",
    "  pull(cloneType)\n",
    "\n",
    "clones <- pbmc_gs@meta.data |> \n",
    "  as_tibble(rownames = \"cell_id\") |>\n",
    "  filter(!is.na(cloneType)) |>\n",
    "  pull(cell_id)\n",
    "DefaultAssay(pbmc_gs) <- \"RNA\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 5:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 421,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "pbmc_gs$celltype <- pbmc_gs@meta.data |>\n",
    "  as_tibble() |>\n",
    "  mutate(celltype = case_when(\n",
    "    predicted.celltype.l2 %in% c(\"ASDC\", \"cDC1\", \"cDC2\", \"pDC\") ~ \"DC\",\n",
    "    predicted.celltype.l2 %in% c(\"B intermediate\", \"B memory\", \"B naive\") ~ \"B cell\",\n",
    "    predicted.celltype.l2 %in% c(\"Doublet\", \"HSPC\", \"ILC\", \"Plasmablast\", \"Platelet\") ~ \"Other\",\n",
    "    TRUE ~ predicted.celltype.l2\n",
    "  )) |>\n",
    "  pull(celltype)\n",
    "\n",
    "palette <- c(\"B cell\" = \"#e41a1c\", \"CD14 Mono\" = \"#377eb8\", \"CD16 Mono\" = \"#a6cee3\", \n",
    "  \"CD4 CTL\" = \"#ff7f00\", \"CD4 Naive\" = \"#fdbf6f\", \n",
    "  \"CD4 Proliferating\" = \"#fdae6b\", \"CD4 TCM\" = \"#a63603\", \"CD4 TEM\" = \"#f16913\", \n",
    "  \"CD8 Naive\"= \"#984ea3\", \n",
    "  \"CD8 Proliferating\" = \"#bcbddc\", \"CD8 TCM\" = \"#3f007d\", \"CD8 TEM\" = \"#807dba\", \"DC\" = \"#999999\", \n",
    "  \"dnT\" = \"#ffff33\", \"MAIT\" = \"#ffeda0\", \"gdT\" = \"#a65628\", \"NK\" = \"#f781bf\",\n",
    "  \"NK Proliferating\", \"NK_CD56bright\", \"Treg\" = \"#4daf4a\", \"Other\" = \"grey\")\n",
    "\n",
    "\n",
    "celltype_umap <- DimPlot(pbmc_gs, reduction = \"ref.umap\", raster = FALSE, \n",
    "  group.by = \"celltype\", label = TRUE, cols = palette,\n",
    "  label.size = 8 ,repel = TRUE) + theme_classic(base_size = 24) +\n",
    "  guides(fill = guide_legend(override.aes = list(size = 16))) +\n",
    "  theme(plot.title = element_blank())\n",
    "\n",
    "\n",
    "persample_alluvial <- function(celltype_table) {\n",
    "  ptid <- celltype_table |>\n",
    "    pull(patient_id) |>\n",
    "    unique()\n",
    "  palette <- c(\"B cell\" = \"#e41a1c\", \"CD14 Mono\" = \"#377eb8\", \"CD16 Mono\" = \"#a6cee3\", \n",
    "  \"CD4 CTL\" = \"#ff7f00\", \"CD4 Naive\" = \"#fdbf6f\", \n",
    "  \"CD4 Proliferating\" = \"#fdae6b\", \"CD4 TCM\" = \"#a63603\", \"CD4 TEM\" = \"#f16913\", \n",
    "  \"CD8 Naive\"= \"#984ea3\",\n",
    "  \"CD8 Proliferating\" = \"#bcbddc\", \"CD8 TCM\" = \"#3f007d\", \"CD8 TEM\" = \"#807dba\", \"DC\" = \"#999999\", \n",
    "  \"dnT\" = \"#ffff33\", \"MAIT\" = \"#ffeda0\", \"gdT\" = \"#a65628\", \"NK\" = \"#f781bf\",\n",
    "  \"NK Proliferating\", \"NK_CD56bright\", \"Treg\" = \"#4daf4a\", \"Other\" = \"grey\")\n",
    "  celltype_table <- celltype_table |>\n",
    "    mutate(Sample = str_extract(Sample, \"V\\\\d+$\"))\n",
    "  celltype_alluvial <-  ggplot2::ggplot(\n",
    "  celltype_table,\n",
    "  aes(\n",
    "      x = Sample, y = duplicate_frequency, stratum = celltype,\n",
    "      alluvium = celltype, fill = celltype, \n",
    "      label = celltype\n",
    "    )\n",
    "  ) +\n",
    "  ggalluvial::geom_alluvium() +\n",
    "  ggalluvial::geom_stratum(ggplot2::aes(y = duplicate_frequency)) +\n",
    "  ggplot2::xlab(\"Repertoire ID\") +\n",
    "  ggplot2::ylab(\"Frequency of cell type\") +\n",
    "  ggplot2::ggtitle(ptid) +\n",
    "  ggplot2::scale_fill_manual(values = palette) +\n",
    "  ggplot2::theme_classic(base_size =  24) +\n",
    "  ggplot2::theme(\n",
    "    legend.position = \"none\",\n",
    "    axis.text.x = ggplot2::element_text(angle = -90)\n",
    "  )\n",
    "  return(celltype_alluvial)\n",
    "}\n",
    "\n",
    "celltype_table <- pbmc_gs@meta.data |>\n",
    "  as_tibble() |>\n",
    "  mutate(celltype = case_when(\n",
    "    predicted.celltype.l2 %in% c(\"ASDC\", \"cDC1\", \"cDC2\", \"pDC\") ~ \"DC\",\n",
    "    predicted.celltype.l2 %in% c(\"B intermediate\", \"B memory\", \"B naive\") ~ \"B cell\",\n",
    "    predicted.celltype.l2 %in% c(\"Doublet\", \"HSPC\", \"ILC\", \"Plasmablast\", \"Platelet\", \"Eryth\") ~ \"Other\",\n",
    "    TRUE ~ predicted.celltype.l2\n",
    "  )) |>\n",
    "  group_by(Sample, celltype) |>\n",
    "  summarize(duplicate_count = length(unique(Barcode))) |>\n",
    "  group_by(Sample) |>\n",
    "  mutate(duplicate_frequency = (duplicate_count*100) / sum(duplicate_count)) |>\n",
    "  ungroup() |>\n",
    "  mutate(patient_id = str_extract(Sample, \"008_\\\\d+\"))\n",
    "\n",
    "celltype_alluvials <- celltype_table |>\n",
    "  group_by(patient_id) |>\n",
    "  group_split() |>\n",
    "  map(persample_alluvial)\n",
    "\n",
    "figure_one_layout <- \"000\n",
    "000\n",
    "000\n",
    "123\n",
    "456\n",
    "789\"\n",
    "\n",
    "figure_five <- celltype_umap + celltype_alluvials[[1]] +\n",
    "  celltype_alluvials[[2]] + celltype_alluvials[[3]] +\n",
    "  celltype_alluvials[[4]] + celltype_alluvials[[5]] +\n",
    "  celltype_alluvials[[6]] + celltype_alluvials[[7]] +\n",
    "  celltype_alluvials[[8]] + celltype_alluvials[[9]] +\n",
    "  plot_layout(design = figure_one_layout, guides = \"collect\") +\n",
    "  plot_annotation(tag_levels = 'A') & theme(text = element_text('NimbusSan'))\n",
    "\n",
    "\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Figure_five.pdf\", sep = \"/\"), figure_five, \n",
    "  width = 20, height = 30, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Figure_five.svg\", sep = \"/\"), figure_five, \n",
    "  width = 20, height = 30)\n",
    "ggsave(str_c(figures_path, \"png\", \"Figure_five.png\", sep = \"/\"), figure_five, \n",
    "  width = 20, height = 30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 6: Identify T-cells from peripheral blood that map to CLusteredUnknowns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading AIRR-Seq files [=>---------------------------------------] 2/41 (  5%) eta:  9s elapsed:  0s\n",
      "\n",
      "Reading AIRR-Seq files [==>--------------------------------------] 3/41 (  7%) eta: 14s elapsed:  1s\n",
      "\n",
      "Reading AIRR-Seq files [===>-------------------------------------] 4/41 ( 10%) eta: 18s elapsed:  2s\n",
      "\n",
      "Reading AIRR-Seq files [====>------------------------------------] 5/41 ( 12%) eta: 22s elapsed:  3s\n",
      "\n",
      "Reading AIRR-Seq files [=====>-----------------------------------] 6/41 ( 15%) eta: 22s elapsed:  4s\n",
      "\n",
      "Reading AIRR-Seq files [======>----------------------------------] 7/41 ( 17%) eta: 21s elapsed:  4s\n",
      "\n",
      "Reading AIRR-Seq files [=======>---------------------------------] 8/41 ( 20%) eta: 20s elapsed:  5s\n",
      "\n",
      "Reading AIRR-Seq files [========>--------------------------------] 9/41 ( 22%) eta: 19s elapsed:  5s\n",
      "\n",
      "Reading AIRR-Seq files [=========>------------------------------] 10/41 ( 24%) eta: 18s elapsed:  6s\n",
      "\n",
      "Reading AIRR-Seq files [==========>-----------------------------] 11/41 ( 27%) eta: 18s elapsed:  7s\n",
      "\n",
      "Reading AIRR-Seq files [===========>----------------------------] 12/41 ( 29%) eta: 19s elapsed:  8s\n",
      "\n",
      "Reading AIRR-Seq files [============>---------------------------] 13/41 ( 32%) eta: 19s elapsed:  9s\n",
      "\n",
      "Reading AIRR-Seq files [=============>--------------------------] 14/41 ( 34%) eta: 18s elapsed:  9s\n",
      "\n",
      "Reading AIRR-Seq files [==============>-------------------------] 15/41 ( 37%) eta: 17s elapsed: 10s\n",
      "\n",
      "Reading AIRR-Seq files [===============>------------------------] 16/41 ( 39%) eta: 16s elapsed: 10s\n",
      "\n",
      "Reading AIRR-Seq files [================>-----------------------] 17/41 ( 41%) eta: 15s elapsed: 11s\n",
      "\n",
      "Reading AIRR-Seq files [=================>----------------------] 18/41 ( 44%) eta: 15s elapsed: 11s\n",
      "\n",
      "Reading AIRR-Seq files [==================>---------------------] 19/41 ( 46%) eta: 14s elapsed: 12s\n",
      "\n",
      "Reading AIRR-Seq files [===================>--------------------] 20/41 ( 49%) eta: 13s elapsed: 12s\n",
      "\n",
      "Reading AIRR-Seq files [===================>--------------------] 21/41 ( 51%) eta: 12s elapsed: 13s\n",
      "\n",
      "Reading AIRR-Seq files [====================>-------------------] 22/41 ( 54%) eta: 12s elapsed: 14s\n",
      "\n",
      "Reading AIRR-Seq files [=====================>------------------] 23/41 ( 56%) eta: 11s elapsed: 14s\n",
      "\n",
      "Reading AIRR-Seq files [======================>-----------------] 24/41 ( 59%) eta: 11s elapsed: 16s\n",
      "\n",
      "Reading AIRR-Seq files [=======================>----------------] 25/41 ( 61%) eta: 10s elapsed: 16s\n",
      "\n",
      "Reading AIRR-Seq files [========================>---------------] 26/41 ( 63%) eta: 10s elapsed: 17s\n",
      "\n",
      "Reading AIRR-Seq files [=========================>--------------] 27/41 ( 66%) eta:  9s elapsed: 18s\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-------------] 28/41 ( 68%) eta:  8s elapsed: 18s\n",
      "\n",
      "Reading AIRR-Seq files [===========================>------------] 29/41 ( 71%) eta:  8s elapsed: 20s\n",
      "\n",
      "Reading AIRR-Seq files [============================>-----------] 30/41 ( 73%) eta:  8s elapsed: 21s\n",
      "\n",
      "Reading AIRR-Seq files [=============================>----------] 31/41 ( 76%) eta:  7s elapsed: 23s\n",
      "\n",
      "Reading AIRR-Seq files [==============================>---------] 32/41 ( 78%) eta:  8s elapsed: 28s\n",
      "\n",
      "Reading AIRR-Seq files [===============================>--------] 33/41 ( 80%) eta:  7s elapsed: 29s\n",
      "\n",
      "Reading AIRR-Seq files [================================>-------] 34/41 ( 83%) eta:  6s elapsed: 30s\n",
      "\n",
      "Reading AIRR-Seq files [=================================>------] 35/41 ( 85%) eta:  6s elapsed: 32s\n",
      "\n",
      "Reading AIRR-Seq files [==================================>-----] 36/41 ( 88%) eta:  5s elapsed: 33s\n",
      "\n",
      "Reading AIRR-Seq files [===================================>----] 37/41 ( 90%) eta:  4s elapsed: 33s\n",
      "\n",
      "Reading AIRR-Seq files [====================================>---] 38/41 ( 93%) eta:  3s elapsed: 34s\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>--] 39/41 ( 95%) eta:  2s elapsed: 34s\n",
      "\n",
      "Reading AIRR-Seq files [======================================>-] 40/41 ( 98%) eta:  1s elapsed: 35s\n",
      "\n",
      "Reading AIRR-Seq files [========================================] 41/41 (100%) eta:  0s elapsed: 36s\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 28s\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 23s\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 28s\n",
      "\n",
      "Subsetting productive sequences [>-----------]  10% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  12% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  15% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  17% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  20% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  22% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  24% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  27% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  29% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  32% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  34% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  37% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  39% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  41% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  44% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  46% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  49% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  51% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  54% eta: 50s\n",
      "\n",
      "Subsetting productive sequences [======>-----]  56% eta: 45s\n",
      "\n",
      "Subsetting productive sequences [======>-----]  59% eta: 49s\n",
      "\n",
      "Subsetting productive sequences [======>-----]  61% eta: 47s\n",
      "\n",
      "Subsetting productive sequences [=======>----]  63% eta: 43s\n",
      "\n",
      "Subsetting productive sequences [=======>----]  66% eta: 40s\n",
      "\n",
      "Subsetting productive sequences [=======>----]  68% eta: 36s\n",
      "\n",
      "Subsetting productive sequences [=======>----]  71% eta: 39s\n",
      "\n",
      "Subsetting productive sequences [========>---]  73% eta: 37s\n",
      "\n",
      "Subsetting productive sequences [========>---]  76% eta: 35s\n",
      "\n",
      "Subsetting productive sequences [========>---]  78% eta: 33s\n",
      "\n",
      "Subsetting productive sequences [=========>--]  80% eta: 31s\n",
      "\n",
      "Subsetting productive sequences [=========>--]  83% eta: 27s\n",
      "\n",
      "Subsetting productive sequences [=========>--]  85% eta: 25s\n",
      "\n",
      "Subsetting productive sequences [==========>-]  88% eta: 20s\n",
      "\n",
      "Subsetting productive sequences [==========>-]  90% eta: 16s\n",
      "\n",
      "Subsetting productive sequences [==========>-]  93% eta: 12s\n",
      "\n",
      "Subsetting productive sequences [==========>-]  95% eta:  8s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta:  4s\n",
      "\n",
      "Subsetting productive sequences [============] 100% eta:  0s\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "kstil_data <- list.files(str_c(tcr_sequencing_path, \"TRB/KS/\", sep = \"/\"), \n",
    "  pattern = \"008_216|008_217|008_220|008_235|008_241|008_236|008_252|008_255|008_256\",\n",
    "  full.names = TRUE) |>\n",
    "  readImmunoSeq() |>\n",
    "  productiveSeq() |> \n",
    "  dplyr::rename(trb_repertoire_id = repertoire_id) |>\n",
    "  inner_join(study_metadata, by = \"trb_repertoire_id\") |>\n",
    "  filter(tissue_type == \"Tumor\") |>\n",
    "  dplyr::select(patient_id, junction_aa, duplicate_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kstme_clustered_unknown_tcrs <- kstme_clustered_unknown_table |>\n",
    "    dplyr::select(pattern, junction_aa) |>\n",
    "    distinct()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mExpected 2 pieces. Missing pieces filled with `NA` in 184758 rows [2, 5, 6, 8, 9, 12, 13, 15, 18, 20, 24, 29, 30, 31, 34, 39, 41, 42, 44, 45, ...].”\n"
     ]
    }
   ],
   "source": [
    "pbmc_tcells_type <- pbmc_gs@meta.data |>\n",
    "  as_tibble(rownames = \"cell_id\") |>\n",
    "  separate(CTaa, into = c(\"tra\", \"trb\"), sep = \"_\", remove = F) |>\n",
    "  mutate(tra = if_else(tra == \"NA\", str_replace_all(tra, \"NA\", NA_character_), tra),\n",
    "    trb = if_else(trb == \"NA\", str_replace_all(trb, \"NA\", NA_character_), trb)) |>\n",
    "  filter(!is.na(trb) ) |>\n",
    "  inner_join(kstme_clustered_unknown_tcrs, by = c(\"trb\" = \"junction_aa\"), relationship = \"many-to-many\") |>\n",
    "  filter(trb %in% kstil_data$junction_aa) |>\n",
    "  dplyr::rename(repertoire_id = Sample, \n",
    "    cell_type = predicted.celltype.l2) |>\n",
    "  mutate(tcr_id = str_c(repertoire_id, trb,cell_type, sep = \";\")) |>\n",
    "  dplyr::select(tcr_id, pattern, repertoire_id, trb, cell_id, Barcode, Frequency)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNon-leaf weights ignored\n",
      "\u001b[1m\u001b[22mNon-leaf weights ignored\n",
      "\u001b[1m\u001b[22mNon-leaf weights ignored\n",
      "\u001b[1m\u001b[22mNon-leaf weights ignored\n"
     ]
    }
   ],
   "source": [
    "\n",
    "packEmUp <- function(pbmc_tcells_type, gsize_filter = 3) {\n",
    "    pbmc_pack_plot_table <- pbmc_tcells_type |>\n",
    "        group_by(pattern, tcr_id) |>\n",
    "        summarize(cell_count = length(unique(cell_id)),\n",
    "            frequency_count = sum(Frequency)) |>\n",
    "        group_by(pattern) |>\n",
    "        mutate(ptid = str_extract(tcr_id, \"008_\\\\d+\"),\n",
    "            gsize = length(unique(ptid))) |>\n",
    "        filter(gsize >= gsize_filter )\n",
    "    gcount_table <- pbmc_pack_plot_table |>\n",
    "        group_by(pattern) |>\n",
    "        summarize(count = sum(cell_count)) |>\n",
    "        dplyr::rename(name = pattern)   \n",
    "    tcount_table <- pbmc_pack_plot_table |>\n",
    "        ungroup() |>\n",
    "        dplyr::select(tcr_id, cell_count) |>\n",
    "        dplyr::rename(count = cell_count,\n",
    "            name = tcr_id)  \n",
    "    count_table <- bind_rows(gcount_table, tcount_table)\n",
    "    count_dict <- count_table |> pull(count)\n",
    "    names(count_dict) <- count_table |> pull(name)\n",
    "    sample_pal <- c(\"008_216\" = \"#08306b\", \"008_241\" = \"#4292c6\", \n",
    "        \"008_252\" = \"#c6dbef\", \"008_236\" = \"#08519c\", \n",
    "        \"008_217\" = \"#67000d\", \"008_220\" = \"#a50f15\", \n",
    "        \"008_255\" = \"#fb6a4a\" , \"008_256\" = \"#fcbba1\",\n",
    "        \"008_235\" = \"#ef3b2c\")\n",
    "    gtable <- pbmc_pack_plot_table |>\n",
    "        dplyr::select(pattern, tcr_id) |>\n",
    "        dplyr::rename(from = pattern, to = tcr_id)  |>\n",
    "        dplyr::select(from, to) \n",
    "    graph <- gtable |>\n",
    "        group_by(from) |>\n",
    "        group_split() |>\n",
    "        map(buildGliphNetworks) |>\n",
    "        bind_graphs() |> \n",
    "        mutate(ptid = str_extract(name, \"008_\\\\d+\"),\n",
    "            ptid = factor(ptid, levels = c(\"008_216\", \"008_236\", \"008_241\", \"008_252\",\n",
    "            \"008_217\", \"008_220\", \"008_235\", \"008_255\", \"008_256\")),\n",
    "            cell_type = str_remove(str_extract(name, \";\\\\w+\\\\s?\\\\w+$\"), \";\"),\n",
    "            count = count_dict[name]) |>\n",
    "        mutate(group_size = if_else(is.na(cell_type), NA_integer_, count))\n",
    "\n",
    "    palette <- c(\"B cell\" = \"#e41a1c\", \"CD14 Mono\" = \"#377eb8\", \"CD16 Mono\" = \"#a6cee3\", \n",
    "    \"CD4 CTL\" = \"#ff7f00\", \"CD4 Naive\" = \"#fdbf6f\", \n",
    "    \"CD4 Proliferating\" = \"#fdae6b\", \"CD4 TCM\" = \"#a63603\", \"CD4 TEM\" = \"#f16913\", \n",
    "    \"CD8 Naive\"= \"#984ea3\", \n",
    "    \"CD8 Proliferating\" = \"#bcbddc\", \"CD8 TCM\" = \"#3f007d\", \"CD8 TEM\" = \"#807dba\", \"DC\" = \"#999999\", \n",
    "    \"dnT\" = \"#ffff33\", \"MAIT\" = \"#ffeda0\", \"gdT\" = \"#a65628\", \"NK\" = \"#f781bf\",\n",
    "    \"NK Proliferating\", \"NK_CD56bright\", \"Treg\" = \"#4daf4a\", \"Other\" = \"grey\")\n",
    "    set.seed(12357)\n",
    "    ptid_plot <- ggraph(graph, 'circlepack', weight = count) + \n",
    "        geom_node_circle(aes(fill = ptid), show.legend = TRUE) + \n",
    "        coord_fixed() +\n",
    "        scale_fill_manual(values = sample_pal, na.value = alpha(\"#99d8c9\", 0.2)) + \n",
    "        geom_text(aes(x = x, y = y, label = group_size)) +\n",
    "        labs(fill = \"Patient ID\", size = \"Cell count\") +\n",
    "        theme_void(base_size = 24) +\n",
    "        theme(legend.position = \"right\")\n",
    "    set.seed(12357)\n",
    "    ctype_plot <- ggraph(graph, 'circlepack', weight = count) + \n",
    "        geom_node_circle(aes(fill = cell_type)) + \n",
    "        coord_fixed() +\n",
    "        scale_fill_manual(values = palette, na.value = alpha(\"#99d8c9\", 0.2)) +\n",
    "        scale_size_binned() + \n",
    "        labs(fill = \"Cell type\", size = \"Cell count\") +\n",
    "        theme_void(base_size = 24) \n",
    "    return(list(ptid_plot, ctype_plot))\n",
    "}\n",
    "\n",
    "buildGliphNetworks <- function(gtable) {\n",
    "    graph <- gtable |>\n",
    "        dplyr::select(from, to, everything()) |>\n",
    "        igraph::graph_from_data_frame() |>\n",
    "        as_tbl_graph()\n",
    "    return(graph)\n",
    "}\n",
    "\n",
    "kspbmc_cell_pack_four <- packEmUp(pbmc_tcells_type, gsize_filter = 4)\n",
    "kspbmc_cell_pack_three <- packEmUp(pbmc_tcells_type, gsize_filter = 3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mExpected 2 pieces. Missing pieces filled with `NA` in 184758 rows [2, 5, 6, 8, 9, 12, 13, 15, 18, 20, 24, 29, 30, 31, 34, 39, 41, 42, 44, 45, ...].”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 24795 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Sample</th><th scope=col>CTaa</th><th scope=col>pattern</th><th scope=col>tra</th><th scope=col>trb</th><th scope=col>cell_id</th><th scope=col>Frequency</th><th scope=col>count</th><th scope=col>phenotype</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_216_V01</td><td>CAVRDNDGYGQNFVF_CASSQDLGQGGDTQYF  </td><td>SQDLGQG%DT</td><td>CAVRDNDGYGQNFVF   </td><td>CASSQDLGQGGDTQYF</td><td>008_216_V01_AAACCTGTCACCGGGT-1</td><td> 67</td><td> 63</td><td>Endemic KS</td></tr>\n",
       "\t<tr><td>008_216_V01</td><td>CALSAPDNYGQNFVF_CASSQPSGGTGELFF   </td><td>S%PSGGTGE </td><td>CALSAPDNYGQNFVF   </td><td>CASSQPSGGTGELFF </td><td>008_216_V01_AAACGGGTCAGGCAAG-1</td><td>161</td><td>151</td><td>Endemic KS</td></tr>\n",
       "\t<tr><td>008_216_V01</td><td>CAASGNVYSGGGADGLTF_CATSLGTGITDTQYF</td><td>SLGTG%TDT </td><td>CAASGNVYSGGGADGLTF</td><td>CATSLGTGITDTQYF </td><td>008_216_V01_AAACGGGTCGTATCAG-1</td><td>152</td><td>138</td><td>Endemic KS</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>008_256_V01</td><td>CAIPSGGGADGLTF_CASSHTGGGSYEQYF</td><td>SH%GGGSYE</td><td>CAIPSGGGADGLTF</td><td>CASSHTGGGSYEQYF</td><td>008_256_V01_TTTGTCATCCAGAGGA-1</td><td>37</td><td>34</td><td>Epidemic KS</td></tr>\n",
       "\t<tr><td>008_256_V01</td><td>CAIPSGGGADGLTF_CASSHTGGGSYEQYF</td><td>S%TGGGSYE</td><td>CAIPSGGGADGLTF</td><td>CASSHTGGGSYEQYF</td><td>008_256_V01_TTTGTCATCCAGAGGA-1</td><td>37</td><td>34</td><td>Epidemic KS</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 24795 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Sample & CTaa & pattern & tra & trb & cell\\_id & Frequency & count & phenotype\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_216\\_V01 & CAVRDNDGYGQNFVF\\_CASSQDLGQGGDTQYF   & SQDLGQG\\%DT & CAVRDNDGYGQNFVF    & CASSQDLGQGGDTQYF & 008\\_216\\_V01\\_AAACCTGTCACCGGGT-1 &  67 &  63 & Endemic KS\\\\\n",
       "\t 008\\_216\\_V01 & CALSAPDNYGQNFVF\\_CASSQPSGGTGELFF    & S\\%PSGGTGE  & CALSAPDNYGQNFVF    & CASSQPSGGTGELFF  & 008\\_216\\_V01\\_AAACGGGTCAGGCAAG-1 & 161 & 151 & Endemic KS\\\\\n",
       "\t 008\\_216\\_V01 & CAASGNVYSGGGADGLTF\\_CATSLGTGITDTQYF & SLGTG\\%TDT  & CAASGNVYSGGGADGLTF & CATSLGTGITDTQYF  & 008\\_216\\_V01\\_AAACGGGTCGTATCAG-1 & 152 & 138 & Endemic KS\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 008\\_256\\_V01 & CAIPSGGGADGLTF\\_CASSHTGGGSYEQYF & SH\\%GGGSYE & CAIPSGGGADGLTF & CASSHTGGGSYEQYF & 008\\_256\\_V01\\_TTTGTCATCCAGAGGA-1 & 37 & 34 & Epidemic KS\\\\\n",
       "\t 008\\_256\\_V01 & CAIPSGGGADGLTF\\_CASSHTGGGSYEQYF & S\\%TGGGSYE & CAIPSGGGADGLTF & CASSHTGGGSYEQYF & 008\\_256\\_V01\\_TTTGTCATCCAGAGGA-1 & 37 & 34 & Epidemic KS\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 24795 × 9\n",
       "\n",
       "| Sample &lt;chr&gt; | CTaa &lt;chr&gt; | pattern &lt;chr&gt; | tra &lt;chr&gt; | trb &lt;chr&gt; | cell_id &lt;chr&gt; | Frequency &lt;int&gt; | count &lt;int&gt; | phenotype &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 008_216_V01 | CAVRDNDGYGQNFVF_CASSQDLGQGGDTQYF   | SQDLGQG%DT | CAVRDNDGYGQNFVF    | CASSQDLGQGGDTQYF | 008_216_V01_AAACCTGTCACCGGGT-1 |  67 |  63 | Endemic KS |\n",
       "| 008_216_V01 | CALSAPDNYGQNFVF_CASSQPSGGTGELFF    | S%PSGGTGE  | CALSAPDNYGQNFVF    | CASSQPSGGTGELFF  | 008_216_V01_AAACGGGTCAGGCAAG-1 | 161 | 151 | Endemic KS |\n",
       "| 008_216_V01 | CAASGNVYSGGGADGLTF_CATSLGTGITDTQYF | SLGTG%TDT  | CAASGNVYSGGGADGLTF | CATSLGTGITDTQYF  | 008_216_V01_AAACGGGTCGTATCAG-1 | 152 | 138 | Endemic KS |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 008_256_V01 | CAIPSGGGADGLTF_CASSHTGGGSYEQYF | SH%GGGSYE | CAIPSGGGADGLTF | CASSHTGGGSYEQYF | 008_256_V01_TTTGTCATCCAGAGGA-1 | 37 | 34 | Epidemic KS |\n",
       "| 008_256_V01 | CAIPSGGGADGLTF_CASSHTGGGSYEQYF | S%TGGGSYE | CAIPSGGGADGLTF | CASSHTGGGSYEQYF | 008_256_V01_TTTGTCATCCAGAGGA-1 | 37 | 34 | Epidemic KS |\n",
       "\n"
      ],
      "text/plain": [
       "      Sample      CTaa                               pattern   \n",
       "1     008_216_V01 CAVRDNDGYGQNFVF_CASSQDLGQGGDTQYF   SQDLGQG%DT\n",
       "2     008_216_V01 CALSAPDNYGQNFVF_CASSQPSGGTGELFF    S%PSGGTGE \n",
       "3     008_216_V01 CAASGNVYSGGGADGLTF_CATSLGTGITDTQYF SLGTG%TDT \n",
       "⋮     ⋮           ⋮                                  ⋮         \n",
       "24794 008_256_V01 CAIPSGGGADGLTF_CASSHTGGGSYEQYF     SH%GGGSYE \n",
       "24795 008_256_V01 CAIPSGGGADGLTF_CASSHTGGGSYEQYF     S%TGGGSYE \n",
       "      tra                trb              cell_id                       \n",
       "1     CAVRDNDGYGQNFVF    CASSQDLGQGGDTQYF 008_216_V01_AAACCTGTCACCGGGT-1\n",
       "2     CALSAPDNYGQNFVF    CASSQPSGGTGELFF  008_216_V01_AAACGGGTCAGGCAAG-1\n",
       "3     CAASGNVYSGGGADGLTF CATSLGTGITDTQYF  008_216_V01_AAACGGGTCGTATCAG-1\n",
       "⋮     ⋮                  ⋮                ⋮                             \n",
       "24794 CAIPSGGGADGLTF     CASSHTGGGSYEQYF  008_256_V01_TTTGTCATCCAGAGGA-1\n",
       "24795 CAIPSGGGADGLTF     CASSHTGGGSYEQYF  008_256_V01_TTTGTCATCCAGAGGA-1\n",
       "      Frequency count phenotype  \n",
       "1      67        63   Endemic KS \n",
       "2     161       151   Endemic KS \n",
       "3     152       138   Endemic KS \n",
       "⋮     ⋮         ⋮     ⋮          \n",
       "24794 37        34    Epidemic KS\n",
       "24795 37        34    Epidemic KS"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pbmc_logo_table <- pbmc_gs@meta.data |>\n",
    "  as_tibble(rownames = \"cell_id\") |>\n",
    "  separate(CTaa, into = c(\"tra\", \"trb\"), sep = \"_\", remove = F) |>\n",
    "  mutate(tra = if_else(tra == \"NA\", str_replace_all(tra, \"NA\", NA_character_), tra),\n",
    "    trb = if_else(trb == \"NA\", str_replace_all(trb, \"NA\", NA_character_), trb)) |>\n",
    "  filter(!is.na(trb) ) |>\n",
    "  inner_join(kstme_clustered_unknown_tcrs, by = c(\"trb\" = \"junction_aa\"), relationship = \"many-to-many\") |>\n",
    "  filter(trb %in% kstil_data$junction_aa) |>\n",
    "  dplyr::select(Sample, CTaa, pattern, tra, trb, cell_id, Frequency) |>\n",
    "  group_by(Sample, pattern, CTaa, tra, trb) |>\n",
    "  mutate(count = length(unique(cell_id))) |>\n",
    "  left_join(study_metadata |> dplyr::select(repertoire_id, phenotype), by = c(\"Sample\" = \"repertoire_id\"))\n",
    "pbmc_logo_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pattern</th><th scope=col>nptids</th><th scope=col>nseqs</th><th scope=col>nunique</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SP%NTGE</td><td>3</td><td>117</td><td>4</td></tr>\n",
       "\t<tr><td>SLG%GE </td><td>4</td><td> 46</td><td>5</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 4\n",
       "\\begin{tabular}{llll}\n",
       " pattern & nptids & nseqs & nunique\\\\\n",
       " <chr> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t SP\\%NTGE & 3 & 117 & 4\\\\\n",
       "\t SLG\\%GE  & 4 &  46 & 5\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 4\n",
       "\n",
       "| pattern &lt;chr&gt; | nptids &lt;int&gt; | nseqs &lt;int&gt; | nunique &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| SP%NTGE | 3 | 117 | 4 |\n",
       "| SLG%GE  | 4 |  46 | 5 |\n",
       "\n"
      ],
      "text/plain": [
       "  pattern nptids nseqs nunique\n",
       "1 SP%NTGE 3      117   4      \n",
       "2 SLG%GE  4       46   5      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pbmc_logo_quant_table <- pbmc_logo_table |>\n",
    "    mutate(Sample = str_extract(Sample, \"008_\\\\d+\")) |>\n",
    "    group_by(pattern) |>\n",
    "    mutate(nptids = length(unique(Sample)),\n",
    "        nseqs = length(CTaa),\n",
    "        nunique = length(unique(CTaa))) |>\n",
    "    dplyr::select(-Sample) |>\n",
    "    distinct() \n",
    "\n",
    "top_pbmc_groups <- pbmc_logo_quant_table |>\n",
    "    filter(nptids > 2) |>\n",
    "    dplyr::select(pattern, nptids, nseqs, nunique) |>\n",
    "    ungroup() |>\n",
    "    distinct() |>\n",
    "    arrange(desc(nseqs)) |>\n",
    "    slice_head(n = 2)\n",
    "\n",
    "top_pbmc_groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "logosGalore <- function(pattern_table) {\n",
    "    pattern <- pattern_table |>\n",
    "        pull(pattern) |>\n",
    "        unique()\n",
    "    tra_list <- pattern_table |>\n",
    "        dplyr::select(tra) |>\n",
    "        distinct() |>\n",
    "        pull(tra)\n",
    "    names(tra_list) <- pattern_table |>\n",
    "        dplyr::select(tra) |>\n",
    "        distinct() |>\n",
    "        mutate(tra_id = str_c(tra, row_number(), sep = \"_\")) |>\n",
    "        pull(tra_id)\n",
    "    tra_logo <- seqlogo(AAStringSet(tra_list), color = \"Chemistry_AA\") +\n",
    "        ggtitle(str_c(\"Specifcity group: \", pattern, sep = \"\")) \n",
    "    trb_list <- pattern_table |>\n",
    "        dplyr::select(trb) |>\n",
    "        distinct() |>\n",
    "        pull(trb) \n",
    "    names(trb_list) <- pattern_table |>\n",
    "        dplyr::select(trb) |>\n",
    "        distinct() |>\n",
    "        mutate(trb_id = str_c(trb, row_number(), sep = \"_\")) |>\n",
    "        pull(trb_id) \n",
    "    trb_logo <- seqlogo(AAStringSet(trb_list), color = \"Chemistry_AA\") \n",
    "    return(list(tra_logo, trb_logo))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 6:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 424,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in rbind(c(\"C\", \"A\", \"L\", \"T\", \"P\", \"H\", \"G\", \"S\", \"A\", \"R\", \"Q\", :\n",
      "“number of columns of result is not a multiple of vector length (arg 1)”\n",
      "Warning message in rbind(c(\"C\", \"A\", \"G\", \"E\", \"R\", \"S\", \"S\", \"N\", \"T\", \"G\", \"K\", :\n",
      "“number of columns of result is not a multiple of vector length (arg 1)”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 105 rows containing missing values (`geom_text()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 105 rows containing missing values (`geom_text()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 105 rows containing missing values (`geom_text()`).”\n"
     ]
    }
   ],
   "source": [
    "pbmc_top_patterns_logos <-  pbmc_logo_quant_table |>\n",
    "    filter(pattern %in% top_pbmc_groups$pattern) |>\n",
    "    group_by(pattern) |>\n",
    "    group_split() |>\n",
    "    map(logosGalore) |>\n",
    "    wrap_plots(nrow = 5, ncol = 2)\n",
    "\n",
    "figure_six <- ((kspbmc_cell_pack_three[[1]] + labs(tag = 'A')) |\n",
    "    (kspbmc_cell_pack_three[[2]] + labs(tag = 'B'))) /\n",
    "    ((pbmc_top_patterns_logos[[1]][[1]] + labs(tag = 'C')) +\n",
    "     pbmc_top_patterns_logos[[1]][[2]] |\n",
    "     pbmc_top_patterns_logos[[2]][[1]] +\n",
    "     pbmc_top_patterns_logos[[2]][[2]]) + \n",
    "     plot_layout(nrow = 2, widths = c(20, 20, 10, 10,10,10),\n",
    "        heights = c(20,20,5,5,5,5), guides = \"collect\") &\n",
    "    theme(text = element_text('NimbusSan')) \n",
    "\n",
    "\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Figure_six.pdf\", sep = \"/\"), figure_six, \n",
    "  width = 35, height = 30, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Figure_six.svg\", sep = \"/\"), figure_six, \n",
    "  width = 35, height = 30)\n",
    "ggsave(str_c(figures_path, \"png\", \"Figure_six.png\", sep = \"/\"), figure_six, \n",
    "  width = 35, height = 30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Section 7: Characterize all the cell from the 10X data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 7:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n",
      "Rasterizing points since number of points exceeds 100,000.\n",
      "To disable this behavior set `raster=FALSE`\n",
      "\n"
     ]
    }
   ],
   "source": [
    "figure_seven <- FeaturePlot(pbmc_gs, \n",
    "  features = c(\"CD4\", \"CD8A\", \"CD8B\", \"NCAM1\", \n",
    "  \"GZMA\", \"GZMB\", \"PRF1\",  \"IFNG\", \n",
    "  \"PDCD1\", \"CTLA4\", \"LAG3\", \"HAVCR2\", \n",
    "   \"MKI67\", \"KLRG1\", \"ENTPD1\", \"CD68\",\n",
    "   \"CD163\", \"IL10\", \"VEGFA\", \"MAF\"), ncol = 4,\n",
    "   keep.scale = \"all\", pt.size = 1, reduction = \"ref.umap\")\n",
    "\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Figure_seven.pdf\", sep = \"/\"), figure_seven, \n",
    "  width = 16, height = 20, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Figure_seven.svg\", sep = \"/\"), figure_seven, \n",
    "  width = 16, height = 20)\n",
    "ggsave(str_c(figures_path, \"png\", \"Figure_seven.png\", sep = \"/\"), figure_seven, \n",
    "  width = 16, height = 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 8: Characterization of clustered unknown in AIRR-Seq data from KSHV+ PBMC datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load KS, KOBS, and U035 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kspbmc_trb_path <- str_c(tcr_v3_sequencing_path, \"ks\", sep = \"/\")\n",
    "kskobs_trb_path <- str_c(tcr_v3_sequencing_path, \"kobs\", sep = \"/\")\n",
    "ksu035_trb_path <- str_c(tcr_sequencing_path, \"TRB/U035_KSHV\", sep \"/\")\n",
    "kspbmc_trb_files <- list.files(kspbmc_trb_path, pattern = \"^008_\\\\d+_V\\\\d+_\\\\d+\",\n",
    "    full.names = TRUE)\n",
    "kskobs_trb_files <- list.files(kskobs_trb_path, pattern = \"tsv\",\n",
    "    full.names = TRUE)\n",
    "ksu035_trb_files <- list.files(ksu035_trb_path, pattern = \"tsv\",\n",
    "    full.names = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 429,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Reading AIRR-Seq files [>---------------------------------------] 2/153 (  1%) eta:  2m elapsed:  1s\n",
      "\n",
      "Reading AIRR-Seq files [>---------------------------------------] 3/153 (  2%) eta:  3m elapsed:  3s\n",
      "\n",
      "Reading AIRR-Seq files [>---------------------------------------] 4/153 (  3%) eta:  2m elapsed:  3s\n",
      "\n",
      "Reading AIRR-Seq files [>---------------------------------------] 5/153 (  3%) eta:  2m elapsed:  4s\n",
      "\n",
      "Reading AIRR-Seq files [=>--------------------------------------] 6/153 (  4%) eta:  2m elapsed:  6s\n",
      "\n",
      "Reading AIRR-Seq files [=>--------------------------------------] 7/153 (  5%) eta:  3m elapsed:  7s\n",
      "\n",
      "Reading AIRR-Seq files [=>--------------------------------------] 8/153 (  5%) eta:  3m elapsed:  9s\n",
      "\n",
      "Reading AIRR-Seq files [=>--------------------------------------] 9/153 (  6%) eta:  2m elapsed:  9s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 10/153 (  7%) eta:  3m elapsed: 11s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 11/153 (  7%) eta:  3m elapsed: 13s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 12/153 (  8%) eta:  3m elapsed: 16s\n",
      "\n",
      "Reading AIRR-Seq files [==>------------------------------------] 13/153 (  8%) eta:  3m elapsed: 16s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 14/153 (  9%) eta:  3m elapsed: 17s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 15/153 ( 10%) eta:  3m elapsed: 18s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 16/153 ( 10%) eta:  3m elapsed: 19s\n",
      "\n",
      "Reading AIRR-Seq files [===>-----------------------------------] 17/153 ( 11%) eta:  3m elapsed: 21s\n",
      "\n",
      "Reading AIRR-Seq files [====>----------------------------------] 18/153 ( 12%) eta:  3m elapsed: 23s\n",
      "\n",
      "Reading AIRR-Seq files [====>----------------------------------] 19/153 ( 12%) eta:  3m elapsed: 24s\n",
      "\n",
      "Reading AIRR-Seq files [====>----------------------------------] 20/153 ( 13%) eta:  3m elapsed: 25s\n",
      "\n",
      "Reading AIRR-Seq files [====>----------------------------------] 21/153 ( 14%) eta:  3m elapsed: 26s\n",
      "\n",
      "Reading AIRR-Seq files [=====>---------------------------------] 22/153 ( 14%) eta:  3m elapsed: 28s\n",
      "\n",
      "Reading AIRR-Seq files [=====>---------------------------------] 23/153 ( 15%) eta:  3m elapsed: 29s\n",
      "\n",
      "Reading AIRR-Seq files [=====>---------------------------------] 24/153 ( 16%) eta:  3m elapsed: 29s\n",
      "\n",
      "Reading AIRR-Seq files [=====>---------------------------------] 25/153 ( 16%) eta:  3m elapsed: 30s\n",
      "\n",
      "Reading AIRR-Seq files [======>--------------------------------] 26/153 ( 17%) eta:  3m elapsed: 31s\n",
      "\n",
      "Reading AIRR-Seq files [======>--------------------------------] 27/153 ( 18%) eta:  2m elapsed: 32s\n",
      "\n",
      "Reading AIRR-Seq files [======>--------------------------------] 28/153 ( 18%) eta:  2m elapsed: 33s\n",
      "\n",
      "Reading AIRR-Seq files [======>--------------------------------] 29/153 ( 19%) eta:  2m elapsed: 34s\n",
      "\n",
      "Reading AIRR-Seq files [=======>-------------------------------] 30/153 ( 20%) eta:  2m elapsed: 35s\n",
      "\n",
      "Reading AIRR-Seq files [=======>-------------------------------] 31/153 ( 20%) eta:  2m elapsed: 37s\n",
      "\n",
      "Reading AIRR-Seq files [=======>-------------------------------] 32/153 ( 21%) eta:  2m elapsed: 39s\n",
      "\n",
      "Reading AIRR-Seq files [=======>-------------------------------] 33/153 ( 22%) eta:  3m elapsed: 46s\n",
      "\n",
      "Reading AIRR-Seq files [========>------------------------------] 34/153 ( 22%) eta:  3m elapsed: 48s\n",
      "\n",
      "Reading AIRR-Seq files [========>------------------------------] 35/153 ( 23%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>------------------------------] 36/153 ( 24%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [========>------------------------------] 37/153 ( 24%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=========>-----------------------------] 38/153 ( 25%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=========>-----------------------------] 39/153 ( 25%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=========>-----------------------------] 40/153 ( 26%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=========>-----------------------------] 41/153 ( 27%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [==========>----------------------------] 42/153 ( 27%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [==========>----------------------------] 43/153 ( 28%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [==========>----------------------------] 44/153 ( 29%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [==========>----------------------------] 45/153 ( 29%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [===========>---------------------------] 46/153 ( 30%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [===========>---------------------------] 47/153 ( 31%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [===========>---------------------------] 48/153 ( 31%) eta:  3m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [===========>---------------------------] 49/153 ( 32%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [============>--------------------------] 50/153 ( 33%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [============>--------------------------] 51/153 ( 33%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [============>--------------------------] 52/153 ( 34%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=============>-------------------------] 53/153 ( 35%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=============>-------------------------] 54/153 ( 35%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=============>-------------------------] 55/153 ( 36%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [=============>-------------------------] 56/153 ( 37%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [==============>------------------------] 57/153 ( 37%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [==============>------------------------] 58/153 ( 38%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [==============>------------------------] 59/153 ( 39%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [==============>------------------------] 60/153 ( 39%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [===============>-----------------------] 61/153 ( 40%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [===============>-----------------------] 62/153 ( 41%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [===============>-----------------------] 63/153 ( 41%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [===============>-----------------------] 64/153 ( 42%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [================>----------------------] 65/153 ( 42%) eta:  2m elapsed:  1m\n",
      "\n",
      "Reading AIRR-Seq files [================>----------------------] 66/153 ( 43%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [================>----------------------] 67/153 ( 44%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [================>----------------------] 68/153 ( 44%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=================>---------------------] 69/153 ( 45%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=================>---------------------] 70/153 ( 46%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=================>---------------------] 71/153 ( 46%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=================>---------------------] 72/153 ( 47%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==================>--------------------] 73/153 ( 48%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==================>--------------------] 74/153 ( 48%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==================>--------------------] 75/153 ( 49%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [==================>--------------------] 76/153 ( 50%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===================>-------------------] 77/153 ( 50%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===================>-------------------] 78/153 ( 51%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===================>-------------------] 79/153 ( 52%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [===================>-------------------] 80/153 ( 52%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [====================>------------------] 81/153 ( 53%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [====================>------------------] 82/153 ( 54%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [====================>------------------] 83/153 ( 54%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [====================>------------------] 84/153 ( 55%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>-----------------] 85/153 ( 56%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>-----------------] 86/153 ( 56%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>-----------------] 87/153 ( 57%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=====================>-----------------] 88/153 ( 58%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [======================>----------------] 89/153 ( 58%) eta:  2m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [======================>----------------] 90/153 ( 59%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [======================>----------------] 91/153 ( 59%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [======================>----------------] 92/153 ( 60%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>---------------] 93/153 ( 61%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>---------------] 94/153 ( 61%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>---------------] 95/153 ( 62%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=======================>---------------] 96/153 ( 63%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========================>--------------] 97/153 ( 63%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========================>--------------] 98/153 ( 64%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========================>--------------] 99/153 ( 65%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 100/153 ( 65%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 101/153 ( 66%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [========================>-------------] 102/153 ( 67%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 103/153 ( 67%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 104/153 ( 68%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 105/153 ( 69%) eta:  1m elapsed:  2m\n",
      "\n",
      "Reading AIRR-Seq files [=========================>------------] 106/153 ( 69%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 107/153 ( 70%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 108/153 ( 71%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 109/153 ( 71%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==========================>-----------] 110/153 ( 72%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 111/153 ( 73%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 112/153 ( 73%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 113/153 ( 74%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===========================>----------] 114/153 ( 75%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 115/153 ( 75%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 116/153 ( 76%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 117/153 ( 76%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [============================>---------] 118/153 ( 77%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 119/153 ( 78%) eta:  1m elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 120/153 ( 78%) eta: 50s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 121/153 ( 79%) eta: 48s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [=============================>--------] 122/153 ( 80%) eta: 47s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 123/153 ( 80%) eta: 46s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 124/153 ( 81%) eta: 45s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 125/153 ( 82%) eta: 44s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [==============================>-------] 126/153 ( 82%) eta: 43s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 127/153 ( 83%) eta: 42s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 128/153 ( 84%) eta: 40s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 129/153 ( 84%) eta: 39s elapsed:  3m\n",
      "\n",
      "Reading AIRR-Seq files [===============================>------] 130/153 ( 85%) eta: 39s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 131/153 ( 86%) eta: 37s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 132/153 ( 86%) eta: 36s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 133/153 ( 87%) eta: 34s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [================================>-----] 134/153 ( 88%) eta: 33s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 135/153 ( 88%) eta: 31s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 136/153 ( 89%) eta: 30s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 137/153 ( 90%) eta: 28s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [=================================>----] 138/153 ( 90%) eta: 27s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 139/153 ( 91%) eta: 25s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 140/153 ( 92%) eta: 23s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 141/153 ( 92%) eta: 22s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [==================================>---] 142/153 ( 93%) eta: 20s elapsed:  4m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 143/153 ( 93%) eta: 19s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 144/153 ( 94%) eta: 17s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 145/153 ( 95%) eta: 16s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [===================================>--] 146/153 ( 95%) eta: 15s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 147/153 ( 96%) eta: 13s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 148/153 ( 97%) eta: 10s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 149/153 ( 97%) eta:  8s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [====================================>-] 150/153 ( 98%) eta:  6s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 151/153 ( 99%) eta:  4s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [=====================================>] 152/153 ( 99%) eta:  2s elapsed:  5m\n",
      "\n",
      "Reading AIRR-Seq files [======================================] 153/153 (100%) eta:  0s elapsed:  5m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 15m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   6% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   8% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   8% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   9% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  10% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  11% eta: 14m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  12% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  12% eta: 14m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  13% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  14% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  15% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  16% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  17% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  19% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  20% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  21% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  22% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  23% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  24% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  25% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  25% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  26% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  27% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  27% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  28% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  29% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  29% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  30% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  31% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  32% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  33% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  34% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  35% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  36% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  37% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  37% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  38% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  39% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  40% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  41% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  42% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  43% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  44% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  45% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  46% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  46% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  47% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  48% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  49% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  50% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  51% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  52% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  53% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  54% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  54% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  55% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  56% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  57% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  58% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  58% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  59% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  60% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  61% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  62% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  63% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  64% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  65% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  65% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  66% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  67% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  68% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  69% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  70% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  71% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  71% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  72% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  73% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  74% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [========>---]  75% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [========>---]  76% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [========>---]  77% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [========>---]  78% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [========>---]  79% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  80% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  81% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  82% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  83% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  84% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  85% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  86% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  87% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  88% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  89% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  90% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  91% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  92% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  93% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  94% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  95% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  95% eta: 47s\n",
      "\n",
      "Subsetting productive sequences [===========>]  96% eta: 40s\n",
      "\n",
      "Subsetting productive sequences [===========>]  97% eta: 34s\n",
      "\n",
      "Subsetting productive sequences [===========>]  97% eta: 28s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 21s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 14s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta:  7s\n",
      "\n",
      "Subsetting productive sequences [============] 100% eta:  0s\n",
      "\n",
      "\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta: 14m\n",
      "\n",
      "Subsetting productive sequences [------------]   1% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [------------]   2% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [------------]   3% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [------------]   4% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   5% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   6% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   7% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   8% eta: 14m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   8% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [>-----------]   9% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  10% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  11% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  12% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [>-----------]  12% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  13% eta: 13m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  14% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  15% eta: 12m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  16% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  17% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  18% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  19% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [=>----------]  20% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  21% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  22% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  23% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  24% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  25% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  26% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  27% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  28% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [==>---------]  29% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  29% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  30% eta: 11m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  31% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  32% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  33% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  34% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  35% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  35% eta: 10m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  36% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [===>--------]  37% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  38% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  39% eta:  9m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  39% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  40% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  41% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  42% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  43% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  44% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  45% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [====>-------]  46% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  46% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  47% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  48% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  49% eta:  8m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  50% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  51% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  52% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  53% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [=====>------]  54% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  54% eta:  7m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  55% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  56% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  57% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  58% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  59% eta:  6m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  60% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  61% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [======>-----]  62% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  63% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  64% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  65% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  66% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  67% eta:  5m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  68% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  69% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  70% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [=======>----]  71% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  71% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  72% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  73% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  74% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  75% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  76% eta:  4m\n",
      "\n",
      "Subsetting productive sequences [========>---]  76% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [========>---]  77% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [========>---]  78% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [========>---]  79% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  80% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  81% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  82% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  83% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  84% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  85% eta:  3m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  86% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [=========>--]  87% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  88% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  89% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  90% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  91% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  92% eta:  2m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  92% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  93% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  94% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [==========>-]  95% eta:  1m\n",
      "\n",
      "Subsetting productive sequences [===========>]  96% eta: 46s\n",
      "\n",
      "Subsetting productive sequences [===========>]  97% eta: 38s\n",
      "\n",
      "Subsetting productive sequences [===========>]  97% eta: 31s\n",
      "\n",
      "Subsetting productive sequences [===========>]  98% eta: 24s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta: 16s\n",
      "\n",
      "Subsetting productive sequences [===========>]  99% eta:  8s\n",
      "\n",
      "Subsetting productive sequences [============] 100% eta:  0s\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "kspbmc_study_table <- readImmunoSeq(c(kspbmc_trb_files, kskobs_trb_files)) \n",
    "kspbmc_amino_acid_table <- productiveSeq(kspbmc_study_table)\n",
    "kspbmc_nprod_table <- productiveSeq(kspbmc_study_table, aggregate = \"junction\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate PBMC metadata with HLA information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 430,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Data : KS PBMC\"\n",
      "[1] \"Number of PTIDs : 48\"\n",
      "[1] \"Number of repertoires : 118\"\n"
     ]
    }
   ],
   "source": [
    "kspbmc_metadata <- study_metadata |>\n",
    "    filter(tissue_type == \"PBMC\") |>\n",
    "    dplyr::select(-c('tra_repertoire_id','rna_libraries',\n",
    "        'nanostring_libraries','gex_library','vdj_library','sc_library_count'))\n",
    "print(str_c(\"Data : KS PBMC\"))\n",
    "print(str_c(\"Number of PTIDs : \", kspbmc_metadata |> pull(patient_id) |> unique() |> length(), sep = \"\"))\n",
    "print(str_c(\"Number of repertoires : \", kspbmc_metadata |> pull(repertoire_id) |> unique() |> length(), sep = \"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 431,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 34 × 35</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ptid</th><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>trb_repertoire_id</th><th scope=col>tissue_type</th><th scope=col>visit_code</th><th scope=col>lession_number</th><th scope=col>tumor_code</th><th scope=col>age</th><th scope=col>gender</th><th scope=col>⋯</th><th scope=col>DPB1allele2</th><th scope=col>DQA1allele1</th><th scope=col>DQA1allele2</th><th scope=col>DQB1allele1</th><th scope=col>DQB1allele2</th><th scope=col>DRB1allele1</th><th scope=col>DRB1allele2</th><th scope=col>DRB345allele1</th><th scope=col>DRB345allele2</th><th scope=col>timestamp</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>011</td><td>KOBSC_11</td><td>KOBSC_11</td><td>KOBSC_11</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>23</td><td>Female</td><td>⋯</td><td>DPB1*105:01:01</td><td>DQA1*04:01:02</td><td>DQA1*04:01:02</td><td>DQB1*03:19:01</td><td>DQB1*03:19:01</td><td>DRB1*08:04:01</td><td>DRB1*08:04:01</td><td>DRB345*Not_Present</td><td>DRB345*Not_Present</td><td>1/12/23</td></tr>\n",
       "\t<tr><td>014</td><td>KOBSC_14</td><td>KOBSC_14</td><td>KOBSC_14</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>25</td><td>Female</td><td>⋯</td><td>DPB1*01:01:01 </td><td>DQA1*01:01:02</td><td>DQA1*02:01:01</td><td>DQB1*02:02:01</td><td>DQB1*05:01:01</td><td>DRB1*07:01:01</td><td>DRB1*11:01:02</td><td>DRB3*03:01:01     </td><td>DRB4*01:03:01     </td><td>1/12/23</td></tr>\n",
       "\t<tr><td>015</td><td>KOBSC_15</td><td>KOBSC_15</td><td>KOBSC_15</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>23</td><td>Female</td><td>⋯</td><td>DPB1*02:01:02 </td><td>DQA1*01:02:01</td><td>DQA1*01:02:01</td><td>DQB1*05:01:01</td><td>DQB1*06:02:01</td><td>DRB1*13:01:01</td><td>DRB1*15:03:01</td><td>DRB3*03:01:01     </td><td>DRB5*01:01:01     </td><td>1/12/23</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>034</td><td>KOBSC_34</td><td>KOBSC_34</td><td>KOBSC_34</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>24</td><td>Female</td><td>⋯</td><td>DPB1*01:01:01</td><td>DQA1*02:01:01</td><td>DQA1*05:05:01</td><td>DQB1*02:02:01</td><td>DQB1*03:01:01</td><td>DRB1*07:01:01</td><td>DRB1*11:02:01</td><td>DRB3*02:02:01</td><td>DRB4*01:01:01</td><td>1/12/23</td></tr>\n",
       "\t<tr><td>078</td><td>KOBSC_78</td><td>KOBSC_78</td><td>KOBSC_78</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>51</td><td>Male  </td><td>⋯</td><td>DPB1*18:01:01</td><td>DQA1*05:01:01</td><td>DQA1*05:11   </td><td>DQB1*02:01:01</td><td>DQB1*03:01:01</td><td>DRB1*03:01:01</td><td>DRB1*11:01:02</td><td>DRB3*02:02:01</td><td>DRB3*02:02:01</td><td>1/12/23</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 34 × 35\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " ptid & patient\\_id & repertoire\\_id & trb\\_repertoire\\_id & tissue\\_type & visit\\_code & lession\\_number & tumor\\_code & age & gender & ⋯ & DPB1allele2 & DQA1allele1 & DQA1allele2 & DQB1allele1 & DQB1allele2 & DRB1allele1 & DRB1allele2 & DRB345allele1 & DRB345allele2 & timestamp\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <lgl> & <lgl> & <lgl> & <dbl> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 011 & KOBSC\\_11 & KOBSC\\_11 & KOBSC\\_11 & PBMC & NA & NA & NA & 23 & Female & ⋯ & DPB1*105:01:01 & DQA1*04:01:02 & DQA1*04:01:02 & DQB1*03:19:01 & DQB1*03:19:01 & DRB1*08:04:01 & DRB1*08:04:01 & DRB345*Not\\_Present & DRB345*Not\\_Present & 1/12/23\\\\\n",
       "\t 014 & KOBSC\\_14 & KOBSC\\_14 & KOBSC\\_14 & PBMC & NA & NA & NA & 25 & Female & ⋯ & DPB1*01:01:01  & DQA1*01:01:02 & DQA1*02:01:01 & DQB1*02:02:01 & DQB1*05:01:01 & DRB1*07:01:01 & DRB1*11:01:02 & DRB3*03:01:01      & DRB4*01:03:01      & 1/12/23\\\\\n",
       "\t 015 & KOBSC\\_15 & KOBSC\\_15 & KOBSC\\_15 & PBMC & NA & NA & NA & 23 & Female & ⋯ & DPB1*02:01:02  & DQA1*01:02:01 & DQA1*01:02:01 & DQB1*05:01:01 & DQB1*06:02:01 & DRB1*13:01:01 & DRB1*15:03:01 & DRB3*03:01:01      & DRB5*01:01:01      & 1/12/23\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 034 & KOBSC\\_34 & KOBSC\\_34 & KOBSC\\_34 & PBMC & NA & NA & NA & 24 & Female & ⋯ & DPB1*01:01:01 & DQA1*02:01:01 & DQA1*05:05:01 & DQB1*02:02:01 & DQB1*03:01:01 & DRB1*07:01:01 & DRB1*11:02:01 & DRB3*02:02:01 & DRB4*01:01:01 & 1/12/23\\\\\n",
       "\t 078 & KOBSC\\_78 & KOBSC\\_78 & KOBSC\\_78 & PBMC & NA & NA & NA & 51 & Male   & ⋯ & DPB1*18:01:01 & DQA1*05:01:01 & DQA1*05:11    & DQB1*02:01:01 & DQB1*03:01:01 & DRB1*03:01:01 & DRB1*11:01:02 & DRB3*02:02:01 & DRB3*02:02:01 & 1/12/23\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 34 × 35\n",
       "\n",
       "| ptid &lt;chr&gt; | patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | trb_repertoire_id &lt;chr&gt; | tissue_type &lt;chr&gt; | visit_code &lt;lgl&gt; | lession_number &lt;lgl&gt; | tumor_code &lt;lgl&gt; | age &lt;dbl&gt; | gender &lt;chr&gt; | ⋯ ⋯ | DPB1allele2 &lt;chr&gt; | DQA1allele1 &lt;chr&gt; | DQA1allele2 &lt;chr&gt; | DQB1allele1 &lt;chr&gt; | DQB1allele2 &lt;chr&gt; | DRB1allele1 &lt;chr&gt; | DRB1allele2 &lt;chr&gt; | DRB345allele1 &lt;chr&gt; | DRB345allele2 &lt;chr&gt; | timestamp &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 011 | KOBSC_11 | KOBSC_11 | KOBSC_11 | PBMC | NA | NA | NA | 23 | Female | ⋯ | DPB1*105:01:01 | DQA1*04:01:02 | DQA1*04:01:02 | DQB1*03:19:01 | DQB1*03:19:01 | DRB1*08:04:01 | DRB1*08:04:01 | DRB345*Not_Present | DRB345*Not_Present | 1/12/23 |\n",
       "| 014 | KOBSC_14 | KOBSC_14 | KOBSC_14 | PBMC | NA | NA | NA | 25 | Female | ⋯ | DPB1*01:01:01  | DQA1*01:01:02 | DQA1*02:01:01 | DQB1*02:02:01 | DQB1*05:01:01 | DRB1*07:01:01 | DRB1*11:01:02 | DRB3*03:01:01      | DRB4*01:03:01      | 1/12/23 |\n",
       "| 015 | KOBSC_15 | KOBSC_15 | KOBSC_15 | PBMC | NA | NA | NA | 23 | Female | ⋯ | DPB1*02:01:02  | DQA1*01:02:01 | DQA1*01:02:01 | DQB1*05:01:01 | DQB1*06:02:01 | DRB1*13:01:01 | DRB1*15:03:01 | DRB3*03:01:01      | DRB5*01:01:01      | 1/12/23 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 034 | KOBSC_34 | KOBSC_34 | KOBSC_34 | PBMC | NA | NA | NA | 24 | Female | ⋯ | DPB1*01:01:01 | DQA1*02:01:01 | DQA1*05:05:01 | DQB1*02:02:01 | DQB1*03:01:01 | DRB1*07:01:01 | DRB1*11:02:01 | DRB3*02:02:01 | DRB4*01:01:01 | 1/12/23 |\n",
       "| 078 | KOBSC_78 | KOBSC_78 | KOBSC_78 | PBMC | NA | NA | NA | 51 | Male   | ⋯ | DPB1*18:01:01 | DQA1*05:01:01 | DQA1*05:11    | DQB1*02:01:01 | DQB1*03:01:01 | DRB1*03:01:01 | DRB1*11:01:02 | DRB3*02:02:01 | DRB3*02:02:01 | 1/12/23 |\n",
       "\n"
      ],
      "text/plain": [
       "   ptid patient_id repertoire_id trb_repertoire_id tissue_type visit_code\n",
       "1  011  KOBSC_11   KOBSC_11      KOBSC_11          PBMC        NA        \n",
       "2  014  KOBSC_14   KOBSC_14      KOBSC_14          PBMC        NA        \n",
       "3  015  KOBSC_15   KOBSC_15      KOBSC_15          PBMC        NA        \n",
       "⋮  ⋮    ⋮          ⋮             ⋮                 ⋮           ⋮         \n",
       "33 034  KOBSC_34   KOBSC_34      KOBSC_34          PBMC        NA        \n",
       "34 078  KOBSC_78   KOBSC_78      KOBSC_78          PBMC        NA        \n",
       "   lession_number tumor_code age gender ⋯ DPB1allele2    DQA1allele1  \n",
       "1  NA             NA         23  Female ⋯ DPB1*105:01:01 DQA1*04:01:02\n",
       "2  NA             NA         25  Female ⋯ DPB1*01:01:01  DQA1*01:01:02\n",
       "3  NA             NA         23  Female ⋯ DPB1*02:01:02  DQA1*01:02:01\n",
       "⋮  ⋮              ⋮          ⋮   ⋮      ⋱ ⋮              ⋮            \n",
       "33 NA             NA         24  Female ⋯ DPB1*01:01:01  DQA1*02:01:01\n",
       "34 NA             NA         51  Male   ⋯ DPB1*18:01:01  DQA1*05:01:01\n",
       "   DQA1allele2   DQB1allele1   DQB1allele2   DRB1allele1   DRB1allele2  \n",
       "1  DQA1*04:01:02 DQB1*03:19:01 DQB1*03:19:01 DRB1*08:04:01 DRB1*08:04:01\n",
       "2  DQA1*02:01:01 DQB1*02:02:01 DQB1*05:01:01 DRB1*07:01:01 DRB1*11:01:02\n",
       "3  DQA1*01:02:01 DQB1*05:01:01 DQB1*06:02:01 DRB1*13:01:01 DRB1*15:03:01\n",
       "⋮  ⋮             ⋮             ⋮             ⋮             ⋮            \n",
       "33 DQA1*05:05:01 DQB1*02:02:01 DQB1*03:01:01 DRB1*07:01:01 DRB1*11:02:01\n",
       "34 DQA1*05:11    DQB1*02:01:01 DQB1*03:01:01 DRB1*03:01:01 DRB1*11:01:02\n",
       "   DRB345allele1      DRB345allele2      timestamp\n",
       "1  DRB345*Not_Present DRB345*Not_Present 1/12/23  \n",
       "2  DRB3*03:01:01      DRB4*01:03:01      1/12/23  \n",
       "3  DRB3*03:01:01      DRB5*01:01:01      1/12/23  \n",
       "⋮  ⋮                  ⋮                  ⋮        \n",
       "33 DRB3*02:02:01      DRB4*01:01:01      1/12/23  \n",
       "34 DRB3*02:02:01      DRB3*02:02:01      1/12/23  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load KOBS sequencing metadata\n",
    "kskobs_metadata_path <- str_c(analysis_path, \"metadata/Control_metadata.csv\", sep = \"/\")\n",
    "kskobs_metadata <- read_csv(kskobs_metadata_path, show_col_types = FALSE) |>\n",
    "    dplyr::select(-c('tra_repertoire_id','rna_libraries','nanostring_libraries',\n",
    "        'gex_library','vdj_library','sc_library_count')) |>\n",
    "    filter(!is.na(trb_repertoire_id)) |>\n",
    "    mutate(ptid = str_pad(str_extract(patient_id, \"\\\\d+$\"), width = 3, side = \"left\", pad = \"0\")) |>\n",
    "    dplyr::select(ptid, everything())\n",
    "kskobs_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 432,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Load KOBS clinical metadata\n",
    "kskobs_clinical_path <- str_c(analysis_path, \"metadata/KOBS_metadata.RData\", sep = \"/\")\n",
    "load(kskobs_clinical_path)\n",
    "kskobs_clinical_metadata <- perperson |>\n",
    "    dplyr::select(ptid, med_hhv8_logcopy_oral, CD4, rna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 433,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(ptid)`\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 34 × 38</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>trb_repertoire_id</th><th scope=col>tissue_type</th><th scope=col>visit_code</th><th scope=col>lession_number</th><th scope=col>tumor_code</th><th scope=col>age</th><th scope=col>gender</th><th scope=col>response</th><th scope=col>⋯</th><th scope=col>DQB1allele2</th><th scope=col>DRB1allele1</th><th scope=col>DRB1allele2</th><th scope=col>DRB345allele1</th><th scope=col>DRB345allele2</th><th scope=col>timestamp</th><th scope=col>KSHV_Oral</th><th scope=col>CD4_count</th><th scope=col>HIV</th><th scope=col>phenotype</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>KOBSC_11</td><td>KOBSC_11</td><td>KOBSC_11</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>23</td><td>Female</td><td>NA</td><td>⋯</td><td>DQB1*03:19:01</td><td>DRB1*08:04:01</td><td>DRB1*08:04:01</td><td>DRB345*Not_Present</td><td>DRB345*Not_Present</td><td>1/12/23</td><td>NA</td><td>NA</td><td>NA</td><td>Control</td></tr>\n",
       "\t<tr><td>KOBSC_14</td><td>KOBSC_14</td><td>KOBSC_14</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>25</td><td>Female</td><td>NA</td><td>⋯</td><td>DQB1*05:01:01</td><td>DRB1*07:01:01</td><td>DRB1*11:01:02</td><td>DRB3*03:01:01     </td><td>DRB4*01:03:01     </td><td>1/12/23</td><td>NA</td><td>NA</td><td>NA</td><td>Control</td></tr>\n",
       "\t<tr><td>KOBSC_15</td><td>KOBSC_15</td><td>KOBSC_15</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>23</td><td>Female</td><td>NA</td><td>⋯</td><td>DQB1*06:02:01</td><td>DRB1*13:01:01</td><td>DRB1*15:03:01</td><td>DRB3*03:01:01     </td><td>DRB5*01:01:01     </td><td>1/12/23</td><td>NA</td><td>NA</td><td>NA</td><td>Control</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>KOBSC_34</td><td>KOBSC_34</td><td>KOBSC_34</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>24</td><td>Female</td><td>NA</td><td>⋯</td><td>DQB1*03:01:01</td><td>DRB1*07:01:01</td><td>DRB1*11:02:01</td><td>DRB3*02:02:01</td><td>DRB4*01:01:01</td><td>1/12/23</td><td>5.129206</td><td>616</td><td>194023</td><td>HIV &amp; KSHV Positive</td></tr>\n",
       "\t<tr><td>KOBSC_78</td><td>KOBSC_78</td><td>KOBSC_78</td><td>PBMC</td><td>NA</td><td>NA</td><td>NA</td><td>51</td><td><span style=white-space:pre-wrap>Male  </span></td><td>NA</td><td>⋯</td><td>DQB1*03:01:01</td><td>DRB1*03:01:01</td><td>DRB1*11:01:02</td><td>DRB3*02:02:01</td><td>DRB3*02:02:01</td><td>1/12/23</td><td>2.867577</td><td>176</td><td>750000</td><td>HIV &amp; KSHV Positive</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 34 × 38\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " patient\\_id & repertoire\\_id & trb\\_repertoire\\_id & tissue\\_type & visit\\_code & lession\\_number & tumor\\_code & age & gender & response & ⋯ & DQB1allele2 & DRB1allele1 & DRB1allele2 & DRB345allele1 & DRB345allele2 & timestamp & KSHV\\_Oral & CD4\\_count & HIV & phenotype\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <lgl> & <lgl> & <lgl> & <dbl> & <chr> & <lgl> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <int> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t KOBSC\\_11 & KOBSC\\_11 & KOBSC\\_11 & PBMC & NA & NA & NA & 23 & Female & NA & ⋯ & DQB1*03:19:01 & DRB1*08:04:01 & DRB1*08:04:01 & DRB345*Not\\_Present & DRB345*Not\\_Present & 1/12/23 & NA & NA & NA & Control\\\\\n",
       "\t KOBSC\\_14 & KOBSC\\_14 & KOBSC\\_14 & PBMC & NA & NA & NA & 25 & Female & NA & ⋯ & DQB1*05:01:01 & DRB1*07:01:01 & DRB1*11:01:02 & DRB3*03:01:01      & DRB4*01:03:01      & 1/12/23 & NA & NA & NA & Control\\\\\n",
       "\t KOBSC\\_15 & KOBSC\\_15 & KOBSC\\_15 & PBMC & NA & NA & NA & 23 & Female & NA & ⋯ & DQB1*06:02:01 & DRB1*13:01:01 & DRB1*15:03:01 & DRB3*03:01:01      & DRB5*01:01:01      & 1/12/23 & NA & NA & NA & Control\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t KOBSC\\_34 & KOBSC\\_34 & KOBSC\\_34 & PBMC & NA & NA & NA & 24 & Female & NA & ⋯ & DQB1*03:01:01 & DRB1*07:01:01 & DRB1*11:02:01 & DRB3*02:02:01 & DRB4*01:01:01 & 1/12/23 & 5.129206 & 616 & 194023 & HIV \\& KSHV Positive\\\\\n",
       "\t KOBSC\\_78 & KOBSC\\_78 & KOBSC\\_78 & PBMC & NA & NA & NA & 51 & Male   & NA & ⋯ & DQB1*03:01:01 & DRB1*03:01:01 & DRB1*11:01:02 & DRB3*02:02:01 & DRB3*02:02:01 & 1/12/23 & 2.867577 & 176 & 750000 & HIV \\& KSHV Positive\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 34 × 38\n",
       "\n",
       "| patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | trb_repertoire_id &lt;chr&gt; | tissue_type &lt;chr&gt; | visit_code &lt;lgl&gt; | lession_number &lt;lgl&gt; | tumor_code &lt;lgl&gt; | age &lt;dbl&gt; | gender &lt;chr&gt; | response &lt;lgl&gt; | ⋯ ⋯ | DQB1allele2 &lt;chr&gt; | DRB1allele1 &lt;chr&gt; | DRB1allele2 &lt;chr&gt; | DRB345allele1 &lt;chr&gt; | DRB345allele2 &lt;chr&gt; | timestamp &lt;chr&gt; | KSHV_Oral &lt;dbl&gt; | CD4_count &lt;int&gt; | HIV &lt;int&gt; | phenotype &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| KOBSC_11 | KOBSC_11 | KOBSC_11 | PBMC | NA | NA | NA | 23 | Female | NA | ⋯ | DQB1*03:19:01 | DRB1*08:04:01 | DRB1*08:04:01 | DRB345*Not_Present | DRB345*Not_Present | 1/12/23 | NA | NA | NA | Control |\n",
       "| KOBSC_14 | KOBSC_14 | KOBSC_14 | PBMC | NA | NA | NA | 25 | Female | NA | ⋯ | DQB1*05:01:01 | DRB1*07:01:01 | DRB1*11:01:02 | DRB3*03:01:01      | DRB4*01:03:01      | 1/12/23 | NA | NA | NA | Control |\n",
       "| KOBSC_15 | KOBSC_15 | KOBSC_15 | PBMC | NA | NA | NA | 23 | Female | NA | ⋯ | DQB1*06:02:01 | DRB1*13:01:01 | DRB1*15:03:01 | DRB3*03:01:01      | DRB5*01:01:01      | 1/12/23 | NA | NA | NA | Control |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| KOBSC_34 | KOBSC_34 | KOBSC_34 | PBMC | NA | NA | NA | 24 | Female | NA | ⋯ | DQB1*03:01:01 | DRB1*07:01:01 | DRB1*11:02:01 | DRB3*02:02:01 | DRB4*01:01:01 | 1/12/23 | 5.129206 | 616 | 194023 | HIV &amp; KSHV Positive |\n",
       "| KOBSC_78 | KOBSC_78 | KOBSC_78 | PBMC | NA | NA | NA | 51 | Male   | NA | ⋯ | DQB1*03:01:01 | DRB1*03:01:01 | DRB1*11:01:02 | DRB3*02:02:01 | DRB3*02:02:01 | 1/12/23 | 2.867577 | 176 | 750000 | HIV &amp; KSHV Positive |\n",
       "\n"
      ],
      "text/plain": [
       "   patient_id repertoire_id trb_repertoire_id tissue_type visit_code\n",
       "1  KOBSC_11   KOBSC_11      KOBSC_11          PBMC        NA        \n",
       "2  KOBSC_14   KOBSC_14      KOBSC_14          PBMC        NA        \n",
       "3  KOBSC_15   KOBSC_15      KOBSC_15          PBMC        NA        \n",
       "⋮  ⋮          ⋮             ⋮                 ⋮           ⋮         \n",
       "33 KOBSC_34   KOBSC_34      KOBSC_34          PBMC        NA        \n",
       "34 KOBSC_78   KOBSC_78      KOBSC_78          PBMC        NA        \n",
       "   lession_number tumor_code age gender response ⋯ DQB1allele2   DRB1allele1  \n",
       "1  NA             NA         23  Female NA       ⋯ DQB1*03:19:01 DRB1*08:04:01\n",
       "2  NA             NA         25  Female NA       ⋯ DQB1*05:01:01 DRB1*07:01:01\n",
       "3  NA             NA         23  Female NA       ⋯ DQB1*06:02:01 DRB1*13:01:01\n",
       "⋮  ⋮              ⋮          ⋮   ⋮      ⋮        ⋱ ⋮             ⋮            \n",
       "33 NA             NA         24  Female NA       ⋯ DQB1*03:01:01 DRB1*07:01:01\n",
       "34 NA             NA         51  Male   NA       ⋯ DQB1*03:01:01 DRB1*03:01:01\n",
       "   DRB1allele2   DRB345allele1      DRB345allele2      timestamp KSHV_Oral\n",
       "1  DRB1*08:04:01 DRB345*Not_Present DRB345*Not_Present 1/12/23   NA       \n",
       "2  DRB1*11:01:02 DRB3*03:01:01      DRB4*01:03:01      1/12/23   NA       \n",
       "3  DRB1*15:03:01 DRB3*03:01:01      DRB5*01:01:01      1/12/23   NA       \n",
       "⋮  ⋮             ⋮                  ⋮                  ⋮         ⋮        \n",
       "33 DRB1*11:02:01 DRB3*02:02:01      DRB4*01:01:01      1/12/23   5.129206 \n",
       "34 DRB1*11:01:02 DRB3*02:02:01      DRB3*02:02:01      1/12/23   2.867577 \n",
       "   CD4_count HIV    phenotype          \n",
       "1  NA        NA     Control            \n",
       "2  NA        NA     Control            \n",
       "3  NA        NA     Control            \n",
       "⋮  ⋮         ⋮      ⋮                  \n",
       "33 616       194023 HIV & KSHV Positive\n",
       "34 176       750000 HIV & KSHV Positive"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kskobs_metadata <- kskobs_metadata |> left_join(kskobs_clinical_metadata) |>\n",
    "    dplyr::mutate(phenotype = case_when(hiv_status == \"Negative\" & kshv_status == \"Negative\" ~ \"Control\",\n",
    "        hiv_status == \"Positive\" & kshv_status == \"Negative\" ~ \"HIV Positive\",\n",
    "        hiv_status == \"Negative\" & kshv_status == \"Positive\" ~ \"KSHV Positive\",\n",
    "        hiv_status == \"Positive\" & kshv_status == \"Positive\" ~ \"HIV & KSHV Positive\")) |>\n",
    "    dplyr::rename(KSHV_Oral = med_hhv8_logcopy_oral, \n",
    "        CD4_count = CD4, HIV = rna) |>\n",
    "    dplyr::select(-ptid)\n",
    "kskobs_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kspbmc_study_metadata <- bind_rows(kspbmc_metadata, kskobs_metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prep GLIPH2 inputs for KOBS and KS PBMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 435,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 4108445 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>junction_aa</th><th scope=col>v_call</th><th scope=col>j_call</th><th scope=col>CDR3a</th><th scope=col>gliph_id</th><th scope=col>duplicate_count</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>CAAAADSYNSPLHF</td><td>TRBV10-3</td><td>TRBJ1-6</td><td>NA</td><td>008_216_V01:Endemic KS</td><td>2</td></tr>\n",
       "\t<tr><td>CAAGLGRYYGYTF </td><td>TRBV30-1</td><td>TRBJ1-2</td><td>NA</td><td>008_216_V01:Endemic KS</td><td>1</td></tr>\n",
       "\t<tr><td>CAAGPGQNTEAFF </td><td>TRBV7-5 </td><td>TRBJ1-1</td><td>NA</td><td>008_216_V01:Endemic KS</td><td>1</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td><span style=white-space:pre-wrap>YDSTDTQYF    </span></td><td>TRBV5-1</td><td>TRBJ2-3</td><td>NA</td><td>KOBSC_78:HIV &amp; KSHV Positive</td><td>1</td></tr>\n",
       "\t<tr><td>YKDRVRINQPQHF</td><td>TRBV6-4</td><td>TRBJ1-5</td><td>NA</td><td>KOBSC_78:HIV &amp; KSHV Positive</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 4108445 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " junction\\_aa & v\\_call & j\\_call & CDR3a & gliph\\_id & duplicate\\_count\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t CAAAADSYNSPLHF & TRBV10-3 & TRBJ1-6 & NA & 008\\_216\\_V01:Endemic KS & 2\\\\\n",
       "\t CAAGLGRYYGYTF  & TRBV30-1 & TRBJ1-2 & NA & 008\\_216\\_V01:Endemic KS & 1\\\\\n",
       "\t CAAGPGQNTEAFF  & TRBV7-5  & TRBJ1-1 & NA & 008\\_216\\_V01:Endemic KS & 1\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t YDSTDTQYF     & TRBV5-1 & TRBJ2-3 & NA & KOBSC\\_78:HIV \\& KSHV Positive & 1\\\\\n",
       "\t YKDRVRINQPQHF & TRBV6-4 & TRBJ1-5 & NA & KOBSC\\_78:HIV \\& KSHV Positive & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 4108445 × 6\n",
       "\n",
       "| junction_aa &lt;chr&gt; | v_call &lt;chr&gt; | j_call &lt;chr&gt; | CDR3a &lt;chr&gt; | gliph_id &lt;chr&gt; | duplicate_count &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| CAAAADSYNSPLHF | TRBV10-3 | TRBJ1-6 | NA | 008_216_V01:Endemic KS | 2 |\n",
       "| CAAGLGRYYGYTF  | TRBV30-1 | TRBJ1-2 | NA | 008_216_V01:Endemic KS | 1 |\n",
       "| CAAGPGQNTEAFF  | TRBV7-5  | TRBJ1-1 | NA | 008_216_V01:Endemic KS | 1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| YDSTDTQYF     | TRBV5-1 | TRBJ2-3 | NA | KOBSC_78:HIV &amp; KSHV Positive | 1 |\n",
       "| YKDRVRINQPQHF | TRBV6-4 | TRBJ1-5 | NA | KOBSC_78:HIV &amp; KSHV Positive | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "        junction_aa    v_call   j_call  CDR3a gliph_id                    \n",
       "1       CAAAADSYNSPLHF TRBV10-3 TRBJ1-6 NA    008_216_V01:Endemic KS      \n",
       "2       CAAGLGRYYGYTF  TRBV30-1 TRBJ1-2 NA    008_216_V01:Endemic KS      \n",
       "3       CAAGPGQNTEAFF  TRBV7-5  TRBJ1-1 NA    008_216_V01:Endemic KS      \n",
       "⋮       ⋮              ⋮        ⋮       ⋮     ⋮                           \n",
       "4108444 YDSTDTQYF      TRBV5-1  TRBJ2-3 NA    KOBSC_78:HIV & KSHV Positive\n",
       "4108445 YKDRVRINQPQHF  TRBV6-4  TRBJ1-5 NA    KOBSC_78:HIV & KSHV Positive\n",
       "        duplicate_count\n",
       "1       2              \n",
       "2       1              \n",
       "3       1              \n",
       "⋮       ⋮              \n",
       "4108444 1              \n",
       "4108445 1              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kspbmc_gliph_trb_table <- kspbmc_study_metadata |>\n",
    "    left_join(kspbmc_amino_acid_table, by = c(\"trb_repertoire_id\" = \"repertoire_id\")) |>\n",
    "    mutate(gliph_id = str_c(repertoire_id, phenotype, sep = \":\"),\n",
    "        CDR3a = NA_character_) |>\n",
    "    dplyr::select(junction_aa, v_call, j_call, CDR3a, gliph_id, duplicate_count) \n",
    "kspbmc_gliph_trb_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 152 × 19</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>repertoire_id</th><th scope=col>aAllele1</th><th scope=col>aAllele2</th><th scope=col>bAllele1</th><th scope=col>bAllele2</th><th scope=col>cAllele1</th><th scope=col>cAllele2</th><th scope=col>DPA1allele1</th><th scope=col>DPA1allele2</th><th scope=col>DPB1allele1</th><th scope=col>DPB1allele2</th><th scope=col>DQA1allele1</th><th scope=col>DQA1allele2</th><th scope=col>DQB1allele1</th><th scope=col>DQB1allele2</th><th scope=col>DRB1allele1</th><th scope=col>DRB1allele2</th><th scope=col>DRB345allele1</th><th scope=col>DRB345allele2</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_216_V01</td><td>A*29:02:01</td><td>A*34:02:01</td><td>B*41:01:01</td><td>B*82:01:01</td><td>C*03:02:02</td><td>C*07:01:01</td><td>DPA1*01:03:01</td><td>DPA1*03:01:01</td><td>DPB1*04:01:01</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*01:05:01</td><td>DQB1*05:01:01</td><td>DQB1*06:02:01</td><td>DRB1*10:01:01</td><td>DRB1*15:03:01</td><td>DRB345*Not_Present</td><td>DRB5*01:01:01</td></tr>\n",
       "\t<tr><td>008_216_V02</td><td>A*29:02:01</td><td>A*34:02:01</td><td>B*41:01:01</td><td>B*82:01:01</td><td>C*03:02:02</td><td>C*07:01:01</td><td>DPA1*01:03:01</td><td>DPA1*03:01:01</td><td>DPB1*04:01:01</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*01:05:01</td><td>DQB1*05:01:01</td><td>DQB1*06:02:01</td><td>DRB1*10:01:01</td><td>DRB1*15:03:01</td><td>DRB345*Not_Present</td><td>DRB5*01:01:01</td></tr>\n",
       "\t<tr><td>008_216_V09</td><td>A*29:02:01</td><td>A*34:02:01</td><td>B*41:01:01</td><td>B*82:01:01</td><td>C*03:02:02</td><td>C*07:01:01</td><td>DPA1*01:03:01</td><td>DPA1*03:01:01</td><td>DPB1*04:01:01</td><td>DPB1*105:01:01</td><td>DQA1*01:02:01</td><td>DQA1*01:05:01</td><td>DQB1*05:01:01</td><td>DQB1*06:02:01</td><td>DRB1*10:01:01</td><td>DRB1*15:03:01</td><td>DRB345*Not_Present</td><td>DRB5*01:01:01</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>KOBSC_34</td><td>A*34:02:01</td><td>A*80:01:01</td><td>B*18:01:01</td><td>B*58:02:01</td><td>C*02:02:02</td><td>C*06:02:01</td><td>DPA1*02:01:08</td><td>DPA1*02:02:02</td><td>DPB1*01:01:01</td><td>DPB1*01:01:01</td><td>DQA1*02:01:01</td><td>DQA1*05:05:01</td><td>DQB1*02:02:01</td><td>DQB1*03:01:01</td><td>DRB1*07:01:01</td><td>DRB1*11:02:01</td><td>DRB3*02:02:01</td><td>DRB4*01:01:01</td></tr>\n",
       "\t<tr><td>KOBSC_78</td><td>A*02:01:01</td><td>A*30:02:01</td><td>B*08:01:01</td><td>B*45:01:01</td><td>C*07:01:01</td><td>C*16:01:01</td><td>DPA1*01:03:01</td><td>DPA1*02:01:08</td><td>DPB1*01:01:01</td><td>DPB1*18:01:01</td><td>DQA1*05:01:01</td><td>DQA1*05:11   </td><td>DQB1*02:01:01</td><td>DQB1*03:01:01</td><td>DRB1*03:01:01</td><td>DRB1*11:01:02</td><td>DRB3*02:02:01</td><td>DRB3*02:02:01</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 152 × 19\n",
       "\\begin{tabular}{lllllllllllllllllll}\n",
       " repertoire\\_id & aAllele1 & aAllele2 & bAllele1 & bAllele2 & cAllele1 & cAllele2 & DPA1allele1 & DPA1allele2 & DPB1allele1 & DPB1allele2 & DQA1allele1 & DQA1allele2 & DQB1allele1 & DQB1allele2 & DRB1allele1 & DRB1allele2 & DRB345allele1 & DRB345allele2\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_216\\_V01 & A*29:02:01 & A*34:02:01 & B*41:01:01 & B*82:01:01 & C*03:02:02 & C*07:01:01 & DPA1*01:03:01 & DPA1*03:01:01 & DPB1*04:01:01 & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*01:05:01 & DQB1*05:01:01 & DQB1*06:02:01 & DRB1*10:01:01 & DRB1*15:03:01 & DRB345*Not\\_Present & DRB5*01:01:01\\\\\n",
       "\t 008\\_216\\_V02 & A*29:02:01 & A*34:02:01 & B*41:01:01 & B*82:01:01 & C*03:02:02 & C*07:01:01 & DPA1*01:03:01 & DPA1*03:01:01 & DPB1*04:01:01 & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*01:05:01 & DQB1*05:01:01 & DQB1*06:02:01 & DRB1*10:01:01 & DRB1*15:03:01 & DRB345*Not\\_Present & DRB5*01:01:01\\\\\n",
       "\t 008\\_216\\_V09 & A*29:02:01 & A*34:02:01 & B*41:01:01 & B*82:01:01 & C*03:02:02 & C*07:01:01 & DPA1*01:03:01 & DPA1*03:01:01 & DPB1*04:01:01 & DPB1*105:01:01 & DQA1*01:02:01 & DQA1*01:05:01 & DQB1*05:01:01 & DQB1*06:02:01 & DRB1*10:01:01 & DRB1*15:03:01 & DRB345*Not\\_Present & DRB5*01:01:01\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t KOBSC\\_34 & A*34:02:01 & A*80:01:01 & B*18:01:01 & B*58:02:01 & C*02:02:02 & C*06:02:01 & DPA1*02:01:08 & DPA1*02:02:02 & DPB1*01:01:01 & DPB1*01:01:01 & DQA1*02:01:01 & DQA1*05:05:01 & DQB1*02:02:01 & DQB1*03:01:01 & DRB1*07:01:01 & DRB1*11:02:01 & DRB3*02:02:01 & DRB4*01:01:01\\\\\n",
       "\t KOBSC\\_78 & A*02:01:01 & A*30:02:01 & B*08:01:01 & B*45:01:01 & C*07:01:01 & C*16:01:01 & DPA1*01:03:01 & DPA1*02:01:08 & DPB1*01:01:01 & DPB1*18:01:01 & DQA1*05:01:01 & DQA1*05:11    & DQB1*02:01:01 & DQB1*03:01:01 & DRB1*03:01:01 & DRB1*11:01:02 & DRB3*02:02:01 & DRB3*02:02:01\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 152 × 19\n",
       "\n",
       "| repertoire_id &lt;chr&gt; | aAllele1 &lt;chr&gt; | aAllele2 &lt;chr&gt; | bAllele1 &lt;chr&gt; | bAllele2 &lt;chr&gt; | cAllele1 &lt;chr&gt; | cAllele2 &lt;chr&gt; | DPA1allele1 &lt;chr&gt; | DPA1allele2 &lt;chr&gt; | DPB1allele1 &lt;chr&gt; | DPB1allele2 &lt;chr&gt; | DQA1allele1 &lt;chr&gt; | DQA1allele2 &lt;chr&gt; | DQB1allele1 &lt;chr&gt; | DQB1allele2 &lt;chr&gt; | DRB1allele1 &lt;chr&gt; | DRB1allele2 &lt;chr&gt; | DRB345allele1 &lt;chr&gt; | DRB345allele2 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 008_216_V01 | A*29:02:01 | A*34:02:01 | B*41:01:01 | B*82:01:01 | C*03:02:02 | C*07:01:01 | DPA1*01:03:01 | DPA1*03:01:01 | DPB1*04:01:01 | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*01:05:01 | DQB1*05:01:01 | DQB1*06:02:01 | DRB1*10:01:01 | DRB1*15:03:01 | DRB345*Not_Present | DRB5*01:01:01 |\n",
       "| 008_216_V02 | A*29:02:01 | A*34:02:01 | B*41:01:01 | B*82:01:01 | C*03:02:02 | C*07:01:01 | DPA1*01:03:01 | DPA1*03:01:01 | DPB1*04:01:01 | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*01:05:01 | DQB1*05:01:01 | DQB1*06:02:01 | DRB1*10:01:01 | DRB1*15:03:01 | DRB345*Not_Present | DRB5*01:01:01 |\n",
       "| 008_216_V09 | A*29:02:01 | A*34:02:01 | B*41:01:01 | B*82:01:01 | C*03:02:02 | C*07:01:01 | DPA1*01:03:01 | DPA1*03:01:01 | DPB1*04:01:01 | DPB1*105:01:01 | DQA1*01:02:01 | DQA1*01:05:01 | DQB1*05:01:01 | DQB1*06:02:01 | DRB1*10:01:01 | DRB1*15:03:01 | DRB345*Not_Present | DRB5*01:01:01 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| KOBSC_34 | A*34:02:01 | A*80:01:01 | B*18:01:01 | B*58:02:01 | C*02:02:02 | C*06:02:01 | DPA1*02:01:08 | DPA1*02:02:02 | DPB1*01:01:01 | DPB1*01:01:01 | DQA1*02:01:01 | DQA1*05:05:01 | DQB1*02:02:01 | DQB1*03:01:01 | DRB1*07:01:01 | DRB1*11:02:01 | DRB3*02:02:01 | DRB4*01:01:01 |\n",
       "| KOBSC_78 | A*02:01:01 | A*30:02:01 | B*08:01:01 | B*45:01:01 | C*07:01:01 | C*16:01:01 | DPA1*01:03:01 | DPA1*02:01:08 | DPB1*01:01:01 | DPB1*18:01:01 | DQA1*05:01:01 | DQA1*05:11    | DQB1*02:01:01 | DQB1*03:01:01 | DRB1*03:01:01 | DRB1*11:01:02 | DRB3*02:02:01 | DRB3*02:02:01 |\n",
       "\n"
      ],
      "text/plain": [
       "    repertoire_id aAllele1   aAllele2   bAllele1   bAllele2   cAllele1  \n",
       "1   008_216_V01   A*29:02:01 A*34:02:01 B*41:01:01 B*82:01:01 C*03:02:02\n",
       "2   008_216_V02   A*29:02:01 A*34:02:01 B*41:01:01 B*82:01:01 C*03:02:02\n",
       "3   008_216_V09   A*29:02:01 A*34:02:01 B*41:01:01 B*82:01:01 C*03:02:02\n",
       "⋮   ⋮             ⋮          ⋮          ⋮          ⋮          ⋮         \n",
       "151 KOBSC_34      A*34:02:01 A*80:01:01 B*18:01:01 B*58:02:01 C*02:02:02\n",
       "152 KOBSC_78      A*02:01:01 A*30:02:01 B*08:01:01 B*45:01:01 C*07:01:01\n",
       "    cAllele2   DPA1allele1   DPA1allele2   DPB1allele1   DPB1allele2   \n",
       "1   C*07:01:01 DPA1*01:03:01 DPA1*03:01:01 DPB1*04:01:01 DPB1*105:01:01\n",
       "2   C*07:01:01 DPA1*01:03:01 DPA1*03:01:01 DPB1*04:01:01 DPB1*105:01:01\n",
       "3   C*07:01:01 DPA1*01:03:01 DPA1*03:01:01 DPB1*04:01:01 DPB1*105:01:01\n",
       "⋮   ⋮          ⋮             ⋮             ⋮             ⋮             \n",
       "151 C*06:02:01 DPA1*02:01:08 DPA1*02:02:02 DPB1*01:01:01 DPB1*01:01:01 \n",
       "152 C*16:01:01 DPA1*01:03:01 DPA1*02:01:08 DPB1*01:01:01 DPB1*18:01:01 \n",
       "    DQA1allele1   DQA1allele2   DQB1allele1   DQB1allele2   DRB1allele1  \n",
       "1   DQA1*01:02:01 DQA1*01:05:01 DQB1*05:01:01 DQB1*06:02:01 DRB1*10:01:01\n",
       "2   DQA1*01:02:01 DQA1*01:05:01 DQB1*05:01:01 DQB1*06:02:01 DRB1*10:01:01\n",
       "3   DQA1*01:02:01 DQA1*01:05:01 DQB1*05:01:01 DQB1*06:02:01 DRB1*10:01:01\n",
       "⋮   ⋮             ⋮             ⋮             ⋮             ⋮            \n",
       "151 DQA1*02:01:01 DQA1*05:05:01 DQB1*02:02:01 DQB1*03:01:01 DRB1*07:01:01\n",
       "152 DQA1*05:01:01 DQA1*05:11    DQB1*02:01:01 DQB1*03:01:01 DRB1*03:01:01\n",
       "    DRB1allele2   DRB345allele1      DRB345allele2\n",
       "1   DRB1*15:03:01 DRB345*Not_Present DRB5*01:01:01\n",
       "2   DRB1*15:03:01 DRB345*Not_Present DRB5*01:01:01\n",
       "3   DRB1*15:03:01 DRB345*Not_Present DRB5*01:01:01\n",
       "⋮   ⋮             ⋮                  ⋮            \n",
       "151 DRB1*11:02:01 DRB3*02:02:01      DRB4*01:01:01\n",
       "152 DRB1*11:01:02 DRB3*02:02:01      DRB3*02:02:01"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kspbmc_gliph_hla_table <- kspbmc_study_metadata |>\n",
    "    dplyr::select(repertoire_id, aAllele1:DRB345allele2) |>\n",
    "    distinct()\n",
    "kspbmc_gliph_hla_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 437,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kspbmc_gliph_trb_input_table <- str_c(table_path, \"kspbmc_gliph\", \"gliph_trb_table.tsv\", sep = \"/\")\n",
    "kspbmc_gliph_hla_input_table <- str_c(table_path, \"kspbmc_gliph\", \"gliph_hla_table.tsv\", sep = \"/\")\n",
    "write_tsv(kspbmc_gliph_trb_table, kspbmc_gliph_trb_input_table)\n",
    "write_tsv(kspbmc_gliph_hla_table, kspbmc_gliph_hla_input_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read GLIPH2 outputs from PBMC data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 440,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kspbmc_gliph_clusters_table <- read_csv(kspbmc_gliph_clusters_path, show_col_types = FALSE)\n",
    "kspbmc_gliph_hlapreds_table <- read_csv(kspbmc_gliph_hlapreds_path, show_col_types = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get HLA matched TCRs from RAW PBMC AIRR-Seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 4169961 × 55</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>trb_repertoire_id</th><th scope=col>junction</th><th scope=col>junction_aa</th><th scope=col>v_call</th><th scope=col>d_call</th><th scope=col>j_call</th><th scope=col>v_family</th><th scope=col>d_family</th><th scope=col>j_family</th><th scope=col>reading_frame</th><th scope=col>⋯</th><th scope=col>DRB345allele2</th><th scope=col>timestamp</th><th scope=col>HIV</th><th scope=col>CD4_count</th><th scope=col>KSHV_plasma</th><th scope=col>phenotype</th><th scope=col>DPBallele2</th><th scope=col>ebv_status</th><th scope=col>bl_status</th><th scope=col>KSHV_Oral</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_216_V10_11_7_19_r</td><td>ATCGGACAGGCTCACTATGGCTACACCTTC      </td><td>IGQAHYGYTF  </td><td>TRBV21-1</td><td>TRBD1-1</td><td>TRBJ1-2</td><td>TRBV21</td><td>TRBD1</td><td>TRBJ1</td><td>in-frame</td><td>⋯</td><td>DRB5*01:01:01</td><td>1/12/23</td><td>NA</td><td>776</td><td>3854.72</td><td>Endemic KS</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>008_216_V10_11_7_19_r</td><td>ATCGGGGGGCAGGGGGCATACTATGGCTACACCTTC</td><td>IGGQGAYYGYTF</td><td>NA      </td><td>TRBD1-1</td><td>TRBJ1-2</td><td>NA    </td><td>TRBD1</td><td>TRBJ1</td><td>in-frame</td><td>⋯</td><td>DRB5*01:01:01</td><td>1/12/23</td><td>NA</td><td>776</td><td>3854.72</td><td>Endemic KS</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>008_216_V10_11_7_19_r</td><td>CCCCCAGGGACCTATGGCTACACCTTC         </td><td>PPGTYGYTF   </td><td>TRBV21-1</td><td>NA     </td><td>TRBJ1-2</td><td>TRBV21</td><td>NA   </td><td>TRBJ1</td><td>in-frame</td><td>⋯</td><td>DRB5*01:01:01</td><td>1/12/23</td><td>NA</td><td>776</td><td>3854.72</td><td>Endemic KS</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>KOBSC_9</td><td>TTTAGGGCCGCAGATACGCAGTATTTT         </td><td>FRAADTQYF   </td><td>TRBV29-1</td><td>NA</td><td>TRBJ2-3</td><td>TRBV29</td><td>NA</td><td>TRBJ2</td><td>in-frame</td><td>⋯</td><td>DRB5*01:01:01</td><td>1/12/23</td><td>NA</td><td>NA</td><td>NA</td><td>KSHV Positive</td><td>NA</td><td>NA</td><td>NA</td><td>4.4323</td></tr>\n",
       "\t<tr><td>KOBSC_9</td><td>TTTATAACAAGGGAAAACGTTAGTGGCTACACCTTC</td><td>FITRENVSGYTF</td><td>TRBV7-3 </td><td>NA</td><td>TRBJ1-2</td><td>TRBV7 </td><td>NA</td><td>TRBJ1</td><td>in-frame</td><td>⋯</td><td>DRB5*01:01:01</td><td>1/12/23</td><td>NA</td><td>NA</td><td>NA</td><td>KSHV Positive</td><td>NA</td><td>NA</td><td>NA</td><td>4.4323</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 4169961 × 55\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " trb\\_repertoire\\_id & junction & junction\\_aa & v\\_call & d\\_call & j\\_call & v\\_family & d\\_family & j\\_family & reading\\_frame & ⋯ & DRB345allele2 & timestamp & HIV & CD4\\_count & KSHV\\_plasma & phenotype & DPBallele2 & ebv\\_status & bl\\_status & KSHV\\_Oral\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 008\\_216\\_V10\\_11\\_7\\_19\\_r & ATCGGACAGGCTCACTATGGCTACACCTTC       & IGQAHYGYTF   & TRBV21-1 & TRBD1-1 & TRBJ1-2 & TRBV21 & TRBD1 & TRBJ1 & in-frame & ⋯ & DRB5*01:01:01 & 1/12/23 & NA & 776 & 3854.72 & Endemic KS & NA & NA & NA & NA\\\\\n",
       "\t 008\\_216\\_V10\\_11\\_7\\_19\\_r & ATCGGGGGGCAGGGGGCATACTATGGCTACACCTTC & IGGQGAYYGYTF & NA       & TRBD1-1 & TRBJ1-2 & NA     & TRBD1 & TRBJ1 & in-frame & ⋯ & DRB5*01:01:01 & 1/12/23 & NA & 776 & 3854.72 & Endemic KS & NA & NA & NA & NA\\\\\n",
       "\t 008\\_216\\_V10\\_11\\_7\\_19\\_r & CCCCCAGGGACCTATGGCTACACCTTC          & PPGTYGYTF    & TRBV21-1 & NA      & TRBJ1-2 & TRBV21 & NA    & TRBJ1 & in-frame & ⋯ & DRB5*01:01:01 & 1/12/23 & NA & 776 & 3854.72 & Endemic KS & NA & NA & NA & NA\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t KOBSC\\_9 & TTTAGGGCCGCAGATACGCAGTATTTT          & FRAADTQYF    & TRBV29-1 & NA & TRBJ2-3 & TRBV29 & NA & TRBJ2 & in-frame & ⋯ & DRB5*01:01:01 & 1/12/23 & NA & NA & NA & KSHV Positive & NA & NA & NA & 4.4323\\\\\n",
       "\t KOBSC\\_9 & TTTATAACAAGGGAAAACGTTAGTGGCTACACCTTC & FITRENVSGYTF & TRBV7-3  & NA & TRBJ1-2 & TRBV7  & NA & TRBJ1 & in-frame & ⋯ & DRB5*01:01:01 & 1/12/23 & NA & NA & NA & KSHV Positive & NA & NA & NA & 4.4323\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 4169961 × 55\n",
       "\n",
       "| trb_repertoire_id &lt;chr&gt; | junction &lt;chr&gt; | junction_aa &lt;chr&gt; | v_call &lt;chr&gt; | d_call &lt;chr&gt; | j_call &lt;chr&gt; | v_family &lt;chr&gt; | d_family &lt;chr&gt; | j_family &lt;chr&gt; | reading_frame &lt;chr&gt; | ⋯ ⋯ | DRB345allele2 &lt;chr&gt; | timestamp &lt;chr&gt; | HIV &lt;dbl&gt; | CD4_count &lt;dbl&gt; | KSHV_plasma &lt;dbl&gt; | phenotype &lt;chr&gt; | DPBallele2 &lt;chr&gt; | ebv_status &lt;chr&gt; | bl_status &lt;chr&gt; | KSHV_Oral &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 008_216_V10_11_7_19_r | ATCGGACAGGCTCACTATGGCTACACCTTC       | IGQAHYGYTF   | TRBV21-1 | TRBD1-1 | TRBJ1-2 | TRBV21 | TRBD1 | TRBJ1 | in-frame | ⋯ | DRB5*01:01:01 | 1/12/23 | NA | 776 | 3854.72 | Endemic KS | NA | NA | NA | NA |\n",
       "| 008_216_V10_11_7_19_r | ATCGGGGGGCAGGGGGCATACTATGGCTACACCTTC | IGGQGAYYGYTF | NA       | TRBD1-1 | TRBJ1-2 | NA     | TRBD1 | TRBJ1 | in-frame | ⋯ | DRB5*01:01:01 | 1/12/23 | NA | 776 | 3854.72 | Endemic KS | NA | NA | NA | NA |\n",
       "| 008_216_V10_11_7_19_r | CCCCCAGGGACCTATGGCTACACCTTC          | PPGTYGYTF    | TRBV21-1 | NA      | TRBJ1-2 | TRBV21 | NA    | TRBJ1 | in-frame | ⋯ | DRB5*01:01:01 | 1/12/23 | NA | 776 | 3854.72 | Endemic KS | NA | NA | NA | NA |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| KOBSC_9 | TTTAGGGCCGCAGATACGCAGTATTTT          | FRAADTQYF    | TRBV29-1 | NA | TRBJ2-3 | TRBV29 | NA | TRBJ2 | in-frame | ⋯ | DRB5*01:01:01 | 1/12/23 | NA | NA | NA | KSHV Positive | NA | NA | NA | 4.4323 |\n",
       "| KOBSC_9 | TTTATAACAAGGGAAAACGTTAGTGGCTACACCTTC | FITRENVSGYTF | TRBV7-3  | NA | TRBJ1-2 | TRBV7  | NA | TRBJ1 | in-frame | ⋯ | DRB5*01:01:01 | 1/12/23 | NA | NA | NA | KSHV Positive | NA | NA | NA | 4.4323 |\n",
       "\n"
      ],
      "text/plain": [
       "        trb_repertoire_id     junction                             junction_aa \n",
       "1       008_216_V10_11_7_19_r ATCGGACAGGCTCACTATGGCTACACCTTC       IGQAHYGYTF  \n",
       "2       008_216_V10_11_7_19_r ATCGGGGGGCAGGGGGCATACTATGGCTACACCTTC IGGQGAYYGYTF\n",
       "3       008_216_V10_11_7_19_r CCCCCAGGGACCTATGGCTACACCTTC          PPGTYGYTF   \n",
       "⋮       ⋮                     ⋮                                    ⋮           \n",
       "4169960 KOBSC_9               TTTAGGGCCGCAGATACGCAGTATTTT          FRAADTQYF   \n",
       "4169961 KOBSC_9               TTTATAACAAGGGAAAACGTTAGTGGCTACACCTTC FITRENVSGYTF\n",
       "        v_call   d_call  j_call  v_family d_family j_family reading_frame ⋯\n",
       "1       TRBV21-1 TRBD1-1 TRBJ1-2 TRBV21   TRBD1    TRBJ1    in-frame      ⋯\n",
       "2       NA       TRBD1-1 TRBJ1-2 NA       TRBD1    TRBJ1    in-frame      ⋯\n",
       "3       TRBV21-1 NA      TRBJ1-2 TRBV21   NA       TRBJ1    in-frame      ⋯\n",
       "⋮       ⋮        ⋮       ⋮       ⋮        ⋮        ⋮        ⋮             ⋱\n",
       "4169960 TRBV29-1 NA      TRBJ2-3 TRBV29   NA       TRBJ2    in-frame      ⋯\n",
       "4169961 TRBV7-3  NA      TRBJ1-2 TRBV7    NA       TRBJ1    in-frame      ⋯\n",
       "        DRB345allele2 timestamp HIV CD4_count KSHV_plasma phenotype    \n",
       "1       DRB5*01:01:01 1/12/23   NA  776       3854.72     Endemic KS   \n",
       "2       DRB5*01:01:01 1/12/23   NA  776       3854.72     Endemic KS   \n",
       "3       DRB5*01:01:01 1/12/23   NA  776       3854.72     Endemic KS   \n",
       "⋮       ⋮             ⋮         ⋮   ⋮         ⋮           ⋮            \n",
       "4169960 DRB5*01:01:01 1/12/23   NA  NA        NA          KSHV Positive\n",
       "4169961 DRB5*01:01:01 1/12/23   NA  NA        NA          KSHV Positive\n",
       "        DPBallele2 ebv_status bl_status KSHV_Oral\n",
       "1       NA         NA         NA        NA       \n",
       "2       NA         NA         NA        NA       \n",
       "3       NA         NA         NA        NA       \n",
       "⋮       ⋮          ⋮          ⋮         ⋮        \n",
       "4169960 NA         NA         NA        4.4323   \n",
       "4169961 NA         NA         NA        4.4323   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kspbmc_annotated_nprod_table <- kspbmc_nprod_table |>\n",
    "    dplyr::rename(trb_repertoire_id = repertoire_id) |>\n",
    "    inner_join(kspbmc_study_metadata, by = c(\"trb_repertoire_id\"))\n",
    "kspbmc_annotated_nprod_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 5364 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>junction</th><th scope=col>junction_aa</th><th scope=col>pathology</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_216</td><td>008_216_V01</td><td>TGTGCCAGCAGCTTAGAAGAGACCCAGTACTTC   </td><td>CASSLEETQYF </td><td>M.tuberculosis</td></tr>\n",
       "\t<tr><td>008_216</td><td>008_216_V01</td><td>TGTGCCAGCAGTTTAGAAGAGACCCAGTACTTC   </td><td>CASSLEETQYF </td><td>M.tuberculosis</td></tr>\n",
       "\t<tr><td>008_216</td><td>008_216_V10</td><td>TGTGCCAGCAGTTCCCGGGACTACGAGCAGTACTTC</td><td>CASSSRDYEQYF</td><td>Other         </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>KOBSC_9</td><td>KOBSC_9</td><td>TGTGCCAGTAGTAGGACGCTGAACACTGAAGCTTTCTTT</td><td>CASSRTLNTEAFF</td><td>CMV</td></tr>\n",
       "\t<tr><td>KOBSC_9</td><td>KOBSC_9</td><td>TGTGCCAGTAGTTTTATCACAGATACGCAGTATTTT   </td><td>CASSFITDTQYF </td><td>CMV</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 5364 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " patient\\_id & repertoire\\_id & junction & junction\\_aa & pathology\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_216 & 008\\_216\\_V01 & TGTGCCAGCAGCTTAGAAGAGACCCAGTACTTC    & CASSLEETQYF  & M.tuberculosis\\\\\n",
       "\t 008\\_216 & 008\\_216\\_V01 & TGTGCCAGCAGTTTAGAAGAGACCCAGTACTTC    & CASSLEETQYF  & M.tuberculosis\\\\\n",
       "\t 008\\_216 & 008\\_216\\_V10 & TGTGCCAGCAGTTCCCGGGACTACGAGCAGTACTTC & CASSSRDYEQYF & Other         \\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t KOBSC\\_9 & KOBSC\\_9 & TGTGCCAGTAGTAGGACGCTGAACACTGAAGCTTTCTTT & CASSRTLNTEAFF & CMV\\\\\n",
       "\t KOBSC\\_9 & KOBSC\\_9 & TGTGCCAGTAGTTTTATCACAGATACGCAGTATTTT    & CASSFITDTQYF  & CMV\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 5364 × 5\n",
       "\n",
       "| patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | junction &lt;chr&gt; | junction_aa &lt;chr&gt; | pathology &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 008_216 | 008_216_V01 | TGTGCCAGCAGCTTAGAAGAGACCCAGTACTTC    | CASSLEETQYF  | M.tuberculosis |\n",
       "| 008_216 | 008_216_V01 | TGTGCCAGCAGTTTAGAAGAGACCCAGTACTTC    | CASSLEETQYF  | M.tuberculosis |\n",
       "| 008_216 | 008_216_V10 | TGTGCCAGCAGTTCCCGGGACTACGAGCAGTACTTC | CASSSRDYEQYF | Other          |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| KOBSC_9 | KOBSC_9 | TGTGCCAGTAGTAGGACGCTGAACACTGAAGCTTTCTTT | CASSRTLNTEAFF | CMV |\n",
       "| KOBSC_9 | KOBSC_9 | TGTGCCAGTAGTTTTATCACAGATACGCAGTATTTT    | CASSFITDTQYF  | CMV |\n",
       "\n"
      ],
      "text/plain": [
       "     patient_id repertoire_id junction                               \n",
       "1    008_216    008_216_V01   TGTGCCAGCAGCTTAGAAGAGACCCAGTACTTC      \n",
       "2    008_216    008_216_V01   TGTGCCAGCAGTTTAGAAGAGACCCAGTACTTC      \n",
       "3    008_216    008_216_V10   TGTGCCAGCAGTTCCCGGGACTACGAGCAGTACTTC   \n",
       "⋮    ⋮          ⋮             ⋮                                      \n",
       "5363 KOBSC_9    KOBSC_9       TGTGCCAGTAGTAGGACGCTGAACACTGAAGCTTTCTTT\n",
       "5364 KOBSC_9    KOBSC_9       TGTGCCAGTAGTTTTATCACAGATACGCAGTATTTT   \n",
       "     junction_aa   pathology     \n",
       "1    CASSLEETQYF   M.tuberculosis\n",
       "2    CASSLEETQYF   M.tuberculosis\n",
       "3    CASSSRDYEQYF  Other         \n",
       "⋮    ⋮             ⋮             \n",
       "5363 CASSRTLNTEAFF CMV           \n",
       "5364 CASSFITDTQYF  CMV           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kspbmc_hla_matched_tcr <- kspbmc_annotated_nprod_table |>\n",
    "    mutate(visit_code = if_else(is.na(visit_code), \"V01\", visit_code)) |>\n",
    "    inner_join(antigen_db, by = c(\"junction_aa\" = \"trb_cdr3_aa\"), \n",
    "        relationship = \"many-to-many\") |>\n",
    "    dplyr::select(patient_id, repertoire_id, junction, junction_aa, epitope, pathology,\n",
    "        antigen, mhc_allele, aAllele1:DRB345allele2) |> \n",
    "    pivot_longer(cols = aAllele1:DRB345allele2, names_to = \"allele\",\n",
    "        values_to = \"mhc_value\") |> \n",
    "    mutate(mhc_allele = str_remove(mhc_allele, \"HLA-\"),\n",
    "        hla_matched = if_else(mhc_allele == mhc_value |\n",
    "            mhc_allele == str_extract(mhc_value, \"\\\\w+\\\\*\\\\d+:\\\\d+\") |\n",
    "            mhc_allele == str_extract(mhc_value, \"\\\\w+\\\\*\\\\d+\"), TRUE, FALSE)) |>\n",
    "    filter(hla_matched) |> \n",
    "    dplyr::select(patient_id:junction_aa, epitope:antigen) |>\n",
    "    distinct() |> \n",
    "    group_by(patient_id, repertoire_id, junction, junction_aa) |>\n",
    "    summarise(npath = length(unique(pathology)),\n",
    "        pathology = str_c(unique(pathology), collapse = \";\")) |>\n",
    "    ungroup() |>\n",
    "    mutate(pathology = if_else(npath > 1, \"Multi-pathogen\", pathology)) |>\n",
    "    dplyr::select(patient_id, repertoire_id, junction, junction_aa, pathology) |>\n",
    "    distinct() |>\n",
    "    mutate(pathology = recode(pathology, \n",
    "        `HomoSapiens` = \"Other\", \"Homo sapiens\" = \"Other\", \n",
    "        \"SelaginellaMoellendorffii\" = \"Other\", \"TriticumAestivum\" = \"Other\", \n",
    "        \"synthetic\" = \"Other\", \"MCPyV\" = \"Other\"))\n",
    "kspbmc_hla_matched_tcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mJoining with `by = join_by(junction, junction_aa, patient_id, repertoire_id)`\n"
     ]
    }
   ],
   "source": [
    "kspbmc_path_annotated_nprod_table <- kspbmc_annotated_nprod_table |> \n",
    "    left_join(kspbmc_hla_matched_tcr) |>\n",
    "    filter(str_detect(junction_aa, \"^C\") & str_detect(junction_aa, \"F$\")) |>\n",
    "    dplyr::select(cohort, patient_id, repertoire_id, tissue_type, phenotype, pathology, junction_aa, duplicate_frequency) |>\n",
    "    mutate(pathology = str_replace_na(pathology, \"Unknown\"),\n",
    "        cohort = if_else(is.na(cohort), \"KOBS\", cohort)) |>\n",
    "    group_by(cohort, repertoire_id, pathology, phenotype, tissue_type) |>\n",
    "    summarize(frequency = sum(duplicate_frequency)) |>\n",
    "    ungroup()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get HLA matched GLIPH groups with known antigenic specificity "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kspbmc_gliph_clusters_path <- str_c(analysis_path, \"gliph/kspbmc_gliph_run/gliph_clusters_cluster.csv\", sep = \"/\")\n",
    "kspbmc_gliph_hlapreds_path <-str_c(analysis_path, \"gliph/kspbmc_gliph_run/gliph_clusters_HLA.csv\", sep = \"/\")\n",
    "# KSPBMC GLIPH table\n",
    "kspbmc_gliph_cluster_pred_table <- read_csv(kspbmc_gliph_clusters_path,\n",
    "        show_col_types = FALSE) \n",
    "kspbmc_gliph_hla_pred_table <- read_csv(kspbmc_gliph_hlapreds_path, \n",
    "    show_col_types = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Total number of GLIPH groups in KSPBMC datasets : 2153193\"\n",
      "[1] \"Number of GLIPH groups of length >= 3 : 2153152\"\n",
      "[1] \"Number of GLIPH groups of length >= 3, vb_score <= 0.01, unique CDR3s >= 3, and unique PTIDs >= 3 : 315964\"\n"
     ]
    }
   ],
   "source": [
    "options(repr.matrix.max.rows=5, repr.matrix.max.columns=5)\n",
    "# Total number of GLIPH groups identified from the KSPBMC dataset\n",
    "print(str_c(\"Total number of GLIPH groups in KSPBMC datasets\",\n",
    "    length(unique(kspbmc_gliph_cluster_pred_table$pattern)), sep = \" : \"))\n",
    "print(str_c(\"Number of GLIPH groups of length >= 3\",\n",
    "    kspbmc_gliph_cluster_pred_table |> \n",
    "        filter(pattern != \"single\" & nchar(pattern) > 2) |>\n",
    "        pull(pattern) |>\n",
    "        unique() |>\n",
    "        length(), sep = \" : \")) \n",
    "print(str_c(\"Number of GLIPH groups of length >= 3, vb_score <= 0.01, unique CDR3s >= 3, and unique PTIDs >= 3\",\n",
    "    kspbmc_gliph_cluster_pred_table |> \n",
    "        filter(pattern != \"single\" & nchar(pattern) > 2) |>\n",
    "        separate(Sample, into = c(\"repertoire_id\", \"cohort\"), sep = \":\") |>\n",
    "        mutate(patient_id = if_else(str_detect(repertoire_id, \"KOBS\"), repertoire_id,\n",
    "            str_extract(repertoire_id, \"008_\\\\d+\")),\n",
    "            pattern = if_else(pattern == \"single\", TcRb, pattern)) |>\n",
    "        group_by(pattern) |>\n",
    "        mutate(number_of_unique_cdr = length(unique(TcRb)),\n",
    "                number_of_unique_ptid = length(unique(patient_id))) |>\n",
    "        filter(vb_score <= 0.01 & number_of_unique_cdr >= 3 &\n",
    "                number_of_unique_ptid >= 3) |>\n",
    "        pull(pattern) |>\n",
    "        unique() |>\n",
    "        length(), sep = \" : \")) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 4063885 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pattern</th><th scope=col>vb_score</th><th scope=col>hla_score</th><th scope=col>expansion_score</th><th scope=col>TcRb</th><th scope=col>repertoire_id</th><th scope=col>cohort</th><th scope=col>V</th><th scope=col>J</th><th scope=col>Freq</th><th scope=col>patient_id</th><th scope=col>number_of_unique_cdr</th><th scope=col>number_of_unique_ptid</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>RG%AGE</td><td>0.006</td><td>0.018</td><td>0.76</td><td>CASRGEAGELFF</td><td>008_229_V01</td><td>Epidemic KS</td><td>TRBV5-1</td><td>TRBJ2-2</td><td>1</td><td>008_229</td><td>52</td><td>47</td></tr>\n",
       "\t<tr><td>RG%AGE</td><td>0.006</td><td>0.018</td><td>0.76</td><td>CASRGEAGELFF</td><td>008_229_V02</td><td>Epidemic KS</td><td>TRBV5-1</td><td>TRBJ2-2</td><td>3</td><td>008_229</td><td>52</td><td>47</td></tr>\n",
       "\t<tr><td>RG%AGE</td><td>0.006</td><td>0.018</td><td>0.76</td><td>CASRGEAGELFF</td><td>008_229_V09</td><td>Epidemic KS</td><td>TRBV5-1</td><td>TRBJ2-2</td><td>3</td><td>008_229</td><td>52</td><td>47</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>SV%GGMNTE</td><td>0.001</td><td>0.0017</td><td>0.91</td><td>CASSVVGGMNTEAFF</td><td>KOBSC_37</td><td>Control</td><td>TRBV9-1 </td><td>TRBJ1-1</td><td>1</td><td>KOBSC_37</td><td>15</td><td>25</td></tr>\n",
       "\t<tr><td>SV%GGMNTE</td><td>0.001</td><td>0.0017</td><td>0.91</td><td>CASSVPGGMNTEAFF</td><td>KOBSC_26</td><td>Control</td><td>TRBV19-1</td><td>TRBJ1-1</td><td>1</td><td>KOBSC_26</td><td>15</td><td>25</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 4063885 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " pattern & vb\\_score & hla\\_score & expansion\\_score & TcRb & repertoire\\_id & cohort & V & J & Freq & patient\\_id & number\\_of\\_unique\\_cdr & number\\_of\\_unique\\_ptid\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t RG\\%AGE & 0.006 & 0.018 & 0.76 & CASRGEAGELFF & 008\\_229\\_V01 & Epidemic KS & TRBV5-1 & TRBJ2-2 & 1 & 008\\_229 & 52 & 47\\\\\n",
       "\t RG\\%AGE & 0.006 & 0.018 & 0.76 & CASRGEAGELFF & 008\\_229\\_V02 & Epidemic KS & TRBV5-1 & TRBJ2-2 & 3 & 008\\_229 & 52 & 47\\\\\n",
       "\t RG\\%AGE & 0.006 & 0.018 & 0.76 & CASRGEAGELFF & 008\\_229\\_V09 & Epidemic KS & TRBV5-1 & TRBJ2-2 & 3 & 008\\_229 & 52 & 47\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t SV\\%GGMNTE & 0.001 & 0.0017 & 0.91 & CASSVVGGMNTEAFF & KOBSC\\_37 & Control & TRBV9-1  & TRBJ1-1 & 1 & KOBSC\\_37 & 15 & 25\\\\\n",
       "\t SV\\%GGMNTE & 0.001 & 0.0017 & 0.91 & CASSVPGGMNTEAFF & KOBSC\\_26 & Control & TRBV19-1 & TRBJ1-1 & 1 & KOBSC\\_26 & 15 & 25\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 4063885 × 13\n",
       "\n",
       "| pattern &lt;chr&gt; | vb_score &lt;dbl&gt; | hla_score &lt;dbl&gt; | expansion_score &lt;dbl&gt; | TcRb &lt;chr&gt; | repertoire_id &lt;chr&gt; | cohort &lt;chr&gt; | V &lt;chr&gt; | J &lt;chr&gt; | Freq &lt;dbl&gt; | patient_id &lt;chr&gt; | number_of_unique_cdr &lt;int&gt; | number_of_unique_ptid &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| RG%AGE | 0.006 | 0.018 | 0.76 | CASRGEAGELFF | 008_229_V01 | Epidemic KS | TRBV5-1 | TRBJ2-2 | 1 | 008_229 | 52 | 47 |\n",
       "| RG%AGE | 0.006 | 0.018 | 0.76 | CASRGEAGELFF | 008_229_V02 | Epidemic KS | TRBV5-1 | TRBJ2-2 | 3 | 008_229 | 52 | 47 |\n",
       "| RG%AGE | 0.006 | 0.018 | 0.76 | CASRGEAGELFF | 008_229_V09 | Epidemic KS | TRBV5-1 | TRBJ2-2 | 3 | 008_229 | 52 | 47 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| SV%GGMNTE | 0.001 | 0.0017 | 0.91 | CASSVVGGMNTEAFF | KOBSC_37 | Control | TRBV9-1  | TRBJ1-1 | 1 | KOBSC_37 | 15 | 25 |\n",
       "| SV%GGMNTE | 0.001 | 0.0017 | 0.91 | CASSVPGGMNTEAFF | KOBSC_26 | Control | TRBV19-1 | TRBJ1-1 | 1 | KOBSC_26 | 15 | 25 |\n",
       "\n"
      ],
      "text/plain": [
       "        pattern   vb_score hla_score expansion_score TcRb           \n",
       "1       RG%AGE    0.006    0.018     0.76            CASRGEAGELFF   \n",
       "2       RG%AGE    0.006    0.018     0.76            CASRGEAGELFF   \n",
       "3       RG%AGE    0.006    0.018     0.76            CASRGEAGELFF   \n",
       "⋮       ⋮         ⋮        ⋮         ⋮               ⋮              \n",
       "4063884 SV%GGMNTE 0.001    0.0017    0.91            CASSVVGGMNTEAFF\n",
       "4063885 SV%GGMNTE 0.001    0.0017    0.91            CASSVPGGMNTEAFF\n",
       "        repertoire_id cohort      V        J       Freq patient_id\n",
       "1       008_229_V01   Epidemic KS TRBV5-1  TRBJ2-2 1    008_229   \n",
       "2       008_229_V02   Epidemic KS TRBV5-1  TRBJ2-2 3    008_229   \n",
       "3       008_229_V09   Epidemic KS TRBV5-1  TRBJ2-2 3    008_229   \n",
       "⋮       ⋮             ⋮           ⋮        ⋮       ⋮    ⋮         \n",
       "4063884 KOBSC_37      Control     TRBV9-1  TRBJ1-1 1    KOBSC_37  \n",
       "4063885 KOBSC_26      Control     TRBV19-1 TRBJ1-1 1    KOBSC_26  \n",
       "        number_of_unique_cdr number_of_unique_ptid\n",
       "1       52                   47                   \n",
       "2       52                   47                   \n",
       "3       52                   47                   \n",
       "⋮       ⋮                    ⋮                    \n",
       "4063884 15                   25                   \n",
       "4063885 15                   25                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Filter out high confidence GLIPH groups with a pattern length of >2 where, \n",
    "# the pattern has at 3 unique CDR3s associated with it, at least 3 PTIDs \n",
    "# associated with it and with a vb_score <= 0.01\n",
    "kspbmc_hc_gliph_cluster_table <- kspbmc_gliph_cluster_pred_table |>\n",
    "  dplyr::select(pattern, vb_score, hla_score, expansion_score, TcRb, Sample, V, J,\n",
    "    Freq) |>\n",
    "  separate(Sample, into = c(\"repertoire_id\", \"cohort\"), sep = \":\") |>\n",
    "  mutate(patient_id = if_else(str_detect(repertoire_id, \"KOBS\"), repertoire_id,\n",
    "    str_extract(repertoire_id, \"008_\\\\d+\"))) |>\n",
    "  group_by(pattern) |>\n",
    "  mutate(number_of_unique_cdr = length(unique(TcRb)),\n",
    "    number_of_unique_ptid = length(unique(patient_id))) |>\n",
    "  ungroup() |>\n",
    "  filter(vb_score <= 0.01 & number_of_unique_cdr >= 3 &\n",
    "    number_of_unique_ptid >= 3 & pattern != \"single\" & nchar(pattern) > 2)\n",
    "kspbmc_hc_gliph_cluster_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 910503 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pattern</th><th scope=col>sig_allele</th><th scope=col>Pvalue</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>RG%AGE</td><td>DPB1*105</td><td>0.036</td></tr>\n",
       "\t<tr><td>RG%AGE</td><td>DRB3*02 </td><td>0.049</td></tr>\n",
       "\t<tr><td>RG%AGE</td><td>B*53    </td><td>0.018</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>SV%GGMNTE</td><td>C*14</td><td>0.0280</td></tr>\n",
       "\t<tr><td>SV%GGMNTE</td><td>B*44</td><td>0.0081</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 910503 × 3\n",
       "\\begin{tabular}{lll}\n",
       " pattern & sig\\_allele & Pvalue\\\\\n",
       " <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t RG\\%AGE & DPB1*105 & 0.036\\\\\n",
       "\t RG\\%AGE & DRB3*02  & 0.049\\\\\n",
       "\t RG\\%AGE & B*53     & 0.018\\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t SV\\%GGMNTE & C*14 & 0.0280\\\\\n",
       "\t SV\\%GGMNTE & B*44 & 0.0081\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 910503 × 3\n",
       "\n",
       "| pattern &lt;chr&gt; | sig_allele &lt;chr&gt; | Pvalue &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| RG%AGE | DPB1*105 | 0.036 |\n",
       "| RG%AGE | DRB3*02  | 0.049 |\n",
       "| RG%AGE | B*53     | 0.018 |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| SV%GGMNTE | C*14 | 0.0280 |\n",
       "| SV%GGMNTE | B*44 | 0.0081 |\n",
       "\n"
      ],
      "text/plain": [
       "       pattern   sig_allele Pvalue\n",
       "1      RG%AGE    DPB1*105   0.036 \n",
       "2      RG%AGE    DRB3*02    0.049 \n",
       "3      RG%AGE    B*53       0.018 \n",
       "⋮      ⋮         ⋮          ⋮     \n",
       "910502 SV%GGMNTE C*14       0.0280\n",
       "910503 SV%GGMNTE B*44       0.0081"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Filter out HLA predictions for high confidence GLIPH groups identified in the\n",
    "# previous step\n",
    "kspbmc_hc_gliph_hla_table <- kspbmc_gliph_hla_pred_table |>\n",
    "  dplyr::rename(sig_allele = `'HLA allele with lowest Fisher Score'`) |>\n",
    "  mutate(pattern = str_remove(pattern, \"^[lg]\")) |>\n",
    "  filter(pattern %in% kspbmc_hc_gliph_cluster_table$pattern & Pvalue <= 0.05) |>\n",
    "  dplyr::select(pattern, sig_allele, Pvalue)\n",
    "kspbmc_hc_gliph_hla_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Total number of HLA restrictions predicted by GLIPH : 7584840\"\n",
      "[1] \"Total number of HLA restrictions for GLIPH gorups of length >= 3 : 7237204\"\n",
      "[1] \"Total number of significant HLA restrictions for GLIPH gorups of length >= 3 : 3892975\"\n"
     ]
    }
   ],
   "source": [
    "print(str_c(\"Total number of HLA restrictions predicted by GLIPH\", \n",
    "    nrow(kspbmc_gliph_hla_pred_table), sep = \" : \"))\n",
    "print(str_c(\"Total number of HLA restrictions for GLIPH gorups of length >= 3\",\n",
    "    kspbmc_gliph_hla_pred_table |> \n",
    "        mutate(pattern = str_remove(pattern, \"^[lg]\")) |>\n",
    "        filter(pattern != \"single\" & nchar(pattern) > 2) |>\n",
    "        nrow(), sep = \" : \")) \n",
    "print(str_c(\"Total number of significant HLA restrictions for GLIPH gorups of length >= 3\",\n",
    "    kspbmc_gliph_hla_pred_table |> \n",
    "        mutate(pattern = str_remove(pattern, \"^[lg]\")) |>\n",
    "        filter(pattern != \"single\" & nchar(pattern) > 2 & Pvalue <= 0.05) |>\n",
    "        nrow(), sep = \" : \")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 15720246 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>pattern</th><th scope=col>vb_score</th><th scope=col>hla_score</th><th scope=col>expansion_score</th><th scope=col>TcRb</th><th scope=col>repertoire_id</th><th scope=col>cohort</th><th scope=col>V</th><th scope=col>J</th><th scope=col>Freq</th><th scope=col>patient_id</th><th scope=col>number_of_unique_cdr</th><th scope=col>number_of_unique_ptid</th><th scope=col>sig_allele</th><th scope=col>Pvalue</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>RG%AGE</td><td>0.006</td><td>0.018</td><td>0.76</td><td>CASRGEAGELFF</td><td>008_229_V01</td><td>Epidemic KS</td><td>TRBV5-1</td><td>TRBJ2-2</td><td>1</td><td>008_229</td><td>52</td><td>47</td><td>DPB1*105</td><td>0.036</td></tr>\n",
       "\t<tr><td>RG%AGE</td><td>0.006</td><td>0.018</td><td>0.76</td><td>CASRGEAGELFF</td><td>008_229_V01</td><td>Epidemic KS</td><td>TRBV5-1</td><td>TRBJ2-2</td><td>1</td><td>008_229</td><td>52</td><td>47</td><td>DRB3*02 </td><td>0.049</td></tr>\n",
       "\t<tr><td>RG%AGE</td><td>0.006</td><td>0.018</td><td>0.76</td><td>CASRGEAGELFF</td><td>008_229_V01</td><td>Epidemic KS</td><td>TRBV5-1</td><td>TRBJ2-2</td><td>1</td><td>008_229</td><td>52</td><td>47</td><td>B*53    </td><td>0.018</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>SV%GGMNTE</td><td>0.001</td><td>0.0017</td><td>0.91</td><td>CASSVPGGMNTEAFF</td><td>KOBSC_26</td><td>Control</td><td>TRBV19-1</td><td>TRBJ1-1</td><td>1</td><td>KOBSC_26</td><td>15</td><td>25</td><td>C*14</td><td>0.0280</td></tr>\n",
       "\t<tr><td>SV%GGMNTE</td><td>0.001</td><td>0.0017</td><td>0.91</td><td>CASSVPGGMNTEAFF</td><td>KOBSC_26</td><td>Control</td><td>TRBV19-1</td><td>TRBJ1-1</td><td>1</td><td>KOBSC_26</td><td>15</td><td>25</td><td>B*44</td><td>0.0081</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 15720246 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " pattern & vb\\_score & hla\\_score & expansion\\_score & TcRb & repertoire\\_id & cohort & V & J & Freq & patient\\_id & number\\_of\\_unique\\_cdr & number\\_of\\_unique\\_ptid & sig\\_allele & Pvalue\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <chr> & <int> & <int> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t RG\\%AGE & 0.006 & 0.018 & 0.76 & CASRGEAGELFF & 008\\_229\\_V01 & Epidemic KS & TRBV5-1 & TRBJ2-2 & 1 & 008\\_229 & 52 & 47 & DPB1*105 & 0.036\\\\\n",
       "\t RG\\%AGE & 0.006 & 0.018 & 0.76 & CASRGEAGELFF & 008\\_229\\_V01 & Epidemic KS & TRBV5-1 & TRBJ2-2 & 1 & 008\\_229 & 52 & 47 & DRB3*02  & 0.049\\\\\n",
       "\t RG\\%AGE & 0.006 & 0.018 & 0.76 & CASRGEAGELFF & 008\\_229\\_V01 & Epidemic KS & TRBV5-1 & TRBJ2-2 & 1 & 008\\_229 & 52 & 47 & B*53     & 0.018\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t SV\\%GGMNTE & 0.001 & 0.0017 & 0.91 & CASSVPGGMNTEAFF & KOBSC\\_26 & Control & TRBV19-1 & TRBJ1-1 & 1 & KOBSC\\_26 & 15 & 25 & C*14 & 0.0280\\\\\n",
       "\t SV\\%GGMNTE & 0.001 & 0.0017 & 0.91 & CASSVPGGMNTEAFF & KOBSC\\_26 & Control & TRBV19-1 & TRBJ1-1 & 1 & KOBSC\\_26 & 15 & 25 & B*44 & 0.0081\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 15720246 × 15\n",
       "\n",
       "| pattern &lt;chr&gt; | vb_score &lt;dbl&gt; | hla_score &lt;dbl&gt; | expansion_score &lt;dbl&gt; | TcRb &lt;chr&gt; | repertoire_id &lt;chr&gt; | cohort &lt;chr&gt; | V &lt;chr&gt; | J &lt;chr&gt; | Freq &lt;dbl&gt; | patient_id &lt;chr&gt; | number_of_unique_cdr &lt;int&gt; | number_of_unique_ptid &lt;int&gt; | sig_allele &lt;chr&gt; | Pvalue &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| RG%AGE | 0.006 | 0.018 | 0.76 | CASRGEAGELFF | 008_229_V01 | Epidemic KS | TRBV5-1 | TRBJ2-2 | 1 | 008_229 | 52 | 47 | DPB1*105 | 0.036 |\n",
       "| RG%AGE | 0.006 | 0.018 | 0.76 | CASRGEAGELFF | 008_229_V01 | Epidemic KS | TRBV5-1 | TRBJ2-2 | 1 | 008_229 | 52 | 47 | DRB3*02  | 0.049 |\n",
       "| RG%AGE | 0.006 | 0.018 | 0.76 | CASRGEAGELFF | 008_229_V01 | Epidemic KS | TRBV5-1 | TRBJ2-2 | 1 | 008_229 | 52 | 47 | B*53     | 0.018 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| SV%GGMNTE | 0.001 | 0.0017 | 0.91 | CASSVPGGMNTEAFF | KOBSC_26 | Control | TRBV19-1 | TRBJ1-1 | 1 | KOBSC_26 | 15 | 25 | C*14 | 0.0280 |\n",
       "| SV%GGMNTE | 0.001 | 0.0017 | 0.91 | CASSVPGGMNTEAFF | KOBSC_26 | Control | TRBV19-1 | TRBJ1-1 | 1 | KOBSC_26 | 15 | 25 | B*44 | 0.0081 |\n",
       "\n"
      ],
      "text/plain": [
       "         pattern   vb_score hla_score expansion_score TcRb           \n",
       "1        RG%AGE    0.006    0.018     0.76            CASRGEAGELFF   \n",
       "2        RG%AGE    0.006    0.018     0.76            CASRGEAGELFF   \n",
       "3        RG%AGE    0.006    0.018     0.76            CASRGEAGELFF   \n",
       "⋮        ⋮         ⋮        ⋮         ⋮               ⋮              \n",
       "15720245 SV%GGMNTE 0.001    0.0017    0.91            CASSVPGGMNTEAFF\n",
       "15720246 SV%GGMNTE 0.001    0.0017    0.91            CASSVPGGMNTEAFF\n",
       "         repertoire_id cohort      V        J       Freq patient_id\n",
       "1        008_229_V01   Epidemic KS TRBV5-1  TRBJ2-2 1    008_229   \n",
       "2        008_229_V01   Epidemic KS TRBV5-1  TRBJ2-2 1    008_229   \n",
       "3        008_229_V01   Epidemic KS TRBV5-1  TRBJ2-2 1    008_229   \n",
       "⋮        ⋮             ⋮           ⋮        ⋮       ⋮    ⋮         \n",
       "15720245 KOBSC_26      Control     TRBV19-1 TRBJ1-1 1    KOBSC_26  \n",
       "15720246 KOBSC_26      Control     TRBV19-1 TRBJ1-1 1    KOBSC_26  \n",
       "         number_of_unique_cdr number_of_unique_ptid sig_allele Pvalue\n",
       "1        52                   47                    DPB1*105   0.036 \n",
       "2        52                   47                    DRB3*02    0.049 \n",
       "3        52                   47                    B*53       0.018 \n",
       "⋮        ⋮                    ⋮                     ⋮          ⋮     \n",
       "15720245 15                   25                    C*14       0.0280\n",
       "15720246 15                   25                    B*44       0.0081"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Merge GLIPH clusters with significant HLA association\n",
    "kspbmc_high_confidence_gliph_groups <- kspbmc_hc_gliph_cluster_table |>\n",
    "  left_join(kspbmc_hc_gliph_hla_table, by = \"pattern\",\n",
    "    relationship = \"many-to-many\") \n",
    "kspbmc_high_confidence_gliph_groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 451,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 14799 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>repertoire_id</th><th scope=col>pathology</th><th scope=col>pattern</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>008_216_V01</td><td>M.tuberculosis</td><td>SLEE%</td></tr>\n",
       "\t<tr><td>008_216_V01</td><td>M.tuberculosis</td><td>%LEET</td></tr>\n",
       "\t<tr><td>008_216_V01</td><td>M.tuberculosis</td><td>SLE%T</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>KOBSC_9</td><td>CMV</td><td>SRT%NTE</td></tr>\n",
       "\t<tr><td>KOBSC_9</td><td>CMV</td><td>SF%TDT </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 14799 × 3\n",
       "\\begin{tabular}{lll}\n",
       " repertoire\\_id & pathology & pattern\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 008\\_216\\_V01 & M.tuberculosis & SLEE\\%\\\\\n",
       "\t 008\\_216\\_V01 & M.tuberculosis & \\%LEET\\\\\n",
       "\t 008\\_216\\_V01 & M.tuberculosis & SLE\\%T\\\\\n",
       "\t ⋮ & ⋮ & ⋮\\\\\n",
       "\t KOBSC\\_9 & CMV & SRT\\%NTE\\\\\n",
       "\t KOBSC\\_9 & CMV & SF\\%TDT \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 14799 × 3\n",
       "\n",
       "| repertoire_id &lt;chr&gt; | pathology &lt;chr&gt; | pattern &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 008_216_V01 | M.tuberculosis | SLEE% |\n",
       "| 008_216_V01 | M.tuberculosis | %LEET |\n",
       "| 008_216_V01 | M.tuberculosis | SLE%T |\n",
       "| ⋮ | ⋮ | ⋮ |\n",
       "| KOBSC_9 | CMV | SRT%NTE |\n",
       "| KOBSC_9 | CMV | SF%TDT  |\n",
       "\n"
      ],
      "text/plain": [
       "      repertoire_id pathology      pattern\n",
       "1     008_216_V01   M.tuberculosis SLEE%  \n",
       "2     008_216_V01   M.tuberculosis %LEET  \n",
       "3     008_216_V01   M.tuberculosis SLE%T  \n",
       "⋮     ⋮             ⋮              ⋮      \n",
       "14798 KOBSC_9       CMV            SRT%NTE\n",
       "14799 KOBSC_9       CMV            SF%TDT "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Annotate GLIPH groups with antigenic specificity based on the CDR3 sequences\n",
    "# they contain\n",
    "kspbmc_path_annotated_gliph_clusters <- kspbmc_hla_matched_tcr |>\n",
    "  distinct() |>\n",
    "  inner_join(kspbmc_high_confidence_gliph_groups, by = c(\"repertoire_id\", \"junction_aa\" = \"TcRb\"),\n",
    "    relationship = \"many-to-many\") |>\n",
    "  dplyr::select(repertoire_id, pathology, pattern) |>\n",
    "  distinct()\n",
    "kspbmc_path_annotated_gliph_clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90m# A tibble: 11 × 2\u001b[39m\n",
      "   pathology              count\n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m\n",
      "\u001b[90m 1\u001b[39m \u001b[90m\"\u001b[39mCMV\u001b[90m\"\u001b[39m                    194\n",
      "\u001b[90m 2\u001b[39m \u001b[90m\"\u001b[39mClustered\\nUnknown\u001b[90m\"\u001b[39m 1\u001b[4m0\u001b[24m\u001b[4m8\u001b[24m\u001b[4m9\u001b[24m540\n",
      "\u001b[90m 3\u001b[39m \u001b[90m\"\u001b[39mEBV\u001b[90m\"\u001b[39m                     26\n",
      "\u001b[90m 4\u001b[39m \u001b[90m\"\u001b[39mHCV\u001b[90m\"\u001b[39m                     60\n",
      "\u001b[90m 5\u001b[39m \u001b[90m\"\u001b[39mHIV-1\u001b[90m\"\u001b[39m                    2\n",
      "\u001b[90m 6\u001b[39m \u001b[90m\"\u001b[39mHSV-2\u001b[90m\"\u001b[39m                    4\n",
      "\u001b[90m 7\u001b[39m \u001b[90m\"\u001b[39mInfluenzaA\u001b[90m\"\u001b[39m             645\n",
      "\u001b[90m 8\u001b[39m \u001b[90m\"\u001b[39mM.tuberculosis\u001b[90m\"\u001b[39m          28\n",
      "\u001b[90m 9\u001b[39m \u001b[90m\"\u001b[39mMulti-pathogen\u001b[90m\"\u001b[39m        \u001b[4m8\u001b[24m471\n",
      "\u001b[90m10\u001b[39m \u001b[90m\"\u001b[39mOther\u001b[90m\"\u001b[39m                  797\n",
      "\u001b[90m11\u001b[39m \u001b[90m\"\u001b[39mSARS-CoV-2\u001b[90m\"\u001b[39m              14\n"
     ]
    }
   ],
   "source": [
    "# Annotate GLIPH groups by antigenic specificity of CDR3s goruped within\n",
    "kspbmc_path_annotated_gliph_groups <- kspbmc_high_confidence_gliph_groups |>\n",
    "  dplyr::select(repertoire_id, pattern, TcRb, V, J) |>\n",
    "  left_join(kspbmc_path_annotated_gliph_clusters, by = c(\"repertoire_id\", \"pattern\"), relationship = \"many-to-many\") |>\n",
    "  distinct() |>\n",
    "  mutate(pathology = if_else(is.na(pathology), \"Clustered\\nUnknown\", pathology)) |>\n",
    "  group_by(TcRb, V, J) |>\n",
    "  mutate(npath = length(unique(pathology)),\n",
    "    pathology = str_c(unique(pathology), collapse =\";\")) |>\n",
    "  ungroup() |>\n",
    "  mutate(pathology = if_else(npath > 1, \"Multi-pathogen\", pathology)) \n",
    "kspbmc_path_annotated_gliph_groups |>\n",
    "  group_by(pathology) |> \n",
    "  summarize(count = length(unique(TcRb))) |>\n",
    "  print(n = 11)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 4156529 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>patient_id</th><th scope=col>repertoire_id</th><th scope=col>tissue_type</th><th scope=col>phenotype</th><th scope=col>pathology</th><th scope=col>junction_aa</th><th scope=col>v_call</th><th scope=col>j_call</th><th scope=col>duplicate_frequency</th><th scope=col>npath</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>008_216</td><td>008_216_V10</td><td>PBMC</td><td>Endemic KS</td><td>Clustered\n",
       "Unknown  </td><td>IGQAHYGYTF  </td><td>TRBV21-1</td><td>TRBJ1-2</td><td>3.365813e-05</td><td>1</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_216</td><td>008_216_V10</td><td>PBMC</td><td>Endemic KS</td><td>Unclustered\n",
       "Unknown</td><td>IGGQGAYYGYTF</td><td>NA      </td><td>TRBJ1-2</td><td>1.682907e-05</td><td>1</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_216</td><td>008_216_V10</td><td>PBMC</td><td>Endemic KS</td><td>Clustered\n",
       "Unknown  </td><td>PPGTYGYTF   </td><td>TRBV21-1</td><td>TRBJ1-2</td><td>1.682907e-05</td><td>1</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>KOBS</td><td>KOBSC_9</td><td>KOBSC_9</td><td>PBMC</td><td>KSHV Positive</td><td>Clustered\n",
       "Unknown  </td><td>FRAADTQYF   </td><td>TRBV29-1</td><td>TRBJ2-3</td><td>9.970786e-06</td><td>1</td></tr>\n",
       "\t<tr><td>KOBS</td><td>KOBSC_9</td><td>KOBSC_9</td><td>PBMC</td><td>KSHV Positive</td><td>Unclustered\n",
       "Unknown</td><td>FITRENVSGYTF</td><td>TRBV7-3 </td><td>TRBJ1-2</td><td>9.970786e-06</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 4156529 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " cohort & patient\\_id & repertoire\\_id & tissue\\_type & phenotype & pathology & junction\\_aa & v\\_call & j\\_call & duplicate\\_frequency & npath\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t Hippos & 008\\_216 & 008\\_216\\_V10 & PBMC & Endemic KS & Clustered\n",
       "Unknown   & IGQAHYGYTF   & TRBV21-1 & TRBJ1-2 & 3.365813e-05 & 1\\\\\n",
       "\t Hippos & 008\\_216 & 008\\_216\\_V10 & PBMC & Endemic KS & Unclustered\n",
       "Unknown & IGGQGAYYGYTF & NA       & TRBJ1-2 & 1.682907e-05 & 1\\\\\n",
       "\t Hippos & 008\\_216 & 008\\_216\\_V10 & PBMC & Endemic KS & Clustered\n",
       "Unknown   & PPGTYGYTF    & TRBV21-1 & TRBJ1-2 & 1.682907e-05 & 1\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t KOBS & KOBSC\\_9 & KOBSC\\_9 & PBMC & KSHV Positive & Clustered\n",
       "Unknown   & FRAADTQYF    & TRBV29-1 & TRBJ2-3 & 9.970786e-06 & 1\\\\\n",
       "\t KOBS & KOBSC\\_9 & KOBSC\\_9 & PBMC & KSHV Positive & Unclustered\n",
       "Unknown & FITRENVSGYTF & TRBV7-3  & TRBJ1-2 & 9.970786e-06 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 4156529 × 11\n",
       "\n",
       "| cohort &lt;chr&gt; | patient_id &lt;chr&gt; | repertoire_id &lt;chr&gt; | tissue_type &lt;chr&gt; | phenotype &lt;chr&gt; | pathology &lt;chr&gt; | junction_aa &lt;chr&gt; | v_call &lt;chr&gt; | j_call &lt;chr&gt; | duplicate_frequency &lt;dbl&gt; | npath &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Hippos | 008_216 | 008_216_V10 | PBMC | Endemic KS | Clustered\n",
       "Unknown   | IGQAHYGYTF   | TRBV21-1 | TRBJ1-2 | 3.365813e-05 | 1 |\n",
       "| Hippos | 008_216 | 008_216_V10 | PBMC | Endemic KS | Unclustered\n",
       "Unknown | IGGQGAYYGYTF | NA       | TRBJ1-2 | 1.682907e-05 | 1 |\n",
       "| Hippos | 008_216 | 008_216_V10 | PBMC | Endemic KS | Clustered\n",
       "Unknown   | PPGTYGYTF    | TRBV21-1 | TRBJ1-2 | 1.682907e-05 | 1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| KOBS | KOBSC_9 | KOBSC_9 | PBMC | KSHV Positive | Clustered\n",
       "Unknown   | FRAADTQYF    | TRBV29-1 | TRBJ2-3 | 9.970786e-06 | 1 |\n",
       "| KOBS | KOBSC_9 | KOBSC_9 | PBMC | KSHV Positive | Unclustered\n",
       "Unknown | FITRENVSGYTF | TRBV7-3  | TRBJ1-2 | 9.970786e-06 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "        cohort patient_id repertoire_id tissue_type phenotype    \n",
       "1       Hippos 008_216    008_216_V10   PBMC        Endemic KS   \n",
       "2       Hippos 008_216    008_216_V10   PBMC        Endemic KS   \n",
       "3       Hippos 008_216    008_216_V10   PBMC        Endemic KS   \n",
       "⋮       ⋮      ⋮          ⋮             ⋮           ⋮            \n",
       "4156528 KOBS   KOBSC_9    KOBSC_9       PBMC        KSHV Positive\n",
       "4156529 KOBS   KOBSC_9    KOBSC_9       PBMC        KSHV Positive\n",
       "        pathology            junction_aa  v_call   j_call  duplicate_frequency\n",
       "1       Clustered\\nUnknown   IGQAHYGYTF   TRBV21-1 TRBJ1-2 3.365813e-05       \n",
       "2       Unclustered\\nUnknown IGGQGAYYGYTF NA       TRBJ1-2 1.682907e-05       \n",
       "3       Clustered\\nUnknown   PPGTYGYTF    TRBV21-1 TRBJ1-2 1.682907e-05       \n",
       "⋮       ⋮                    ⋮            ⋮        ⋮       ⋮                  \n",
       "4156528 Clustered\\nUnknown   FRAADTQYF    TRBV29-1 TRBJ2-3 9.970786e-06       \n",
       "4156529 Unclustered\\nUnknown FITRENVSGYTF TRBV7-3  TRBJ1-2 9.970786e-06       \n",
       "        npath\n",
       "1       1    \n",
       "2       1    \n",
       "3       1    \n",
       "⋮       ⋮    \n",
       "4156528 1    \n",
       "4156529 1    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kspbmc_gliph_nprod_table <- kspbmc_annotated_nprod_table |>\n",
    "    left_join(kspbmc_path_annotated_gliph_groups, by = c(\"repertoire_id\", \n",
    "        \"junction_aa\" = \"TcRb\", \"v_call\" = \"V\",\n",
    "        \"j_call\" = \"J\"), relationship = \"many-to-many\") |>\n",
    "    distinct() |>\n",
    "    dplyr::select(cohort, patient_id, repertoire_id, tissue_type, phenotype, \n",
    "        pathology, junction_aa, v_call, j_call, duplicate_frequency) |>\n",
    "    mutate(pathology = if_else(is.na(pathology), \"Unclustered\\nUnknown\", pathology),\n",
    "        phenotype = if_else(is.na(phenotype) , \"Epidemic KS\", phenotype)) |>\n",
    "    group_by(junction_aa, v_call, j_call ) |>\n",
    "    mutate(npath = length(unique(pathology)),\n",
    "        pathology = str_c(unique(pathology), collapse =\";\")) |>\n",
    "    ungroup() |>\n",
    "    mutate(pathology = if_else(npath > 1, \"Multi-pathogen\", pathology)) |>\n",
    "    distinct() |>\n",
    "    mutate(cohort = if_else(is.na(cohort), \"KOBS\", cohort))\n",
    "kspbmc_gliph_nprod_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 689 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cohort</th><th scope=col>repertoire_id</th><th scope=col>pathology</th><th scope=col>phenotype</th><th scope=col>tissue_type</th><th scope=col>frequency</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Hippos</td><td>008_216_V01</td><td>Clustered\n",
       "Unknown  </td><td>Endemic KS</td><td>PBMC</td><td>0.592600906</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_216_V01</td><td>Multi-pathogen      </td><td>Endemic KS</td><td>PBMC</td><td>0.008366322</td></tr>\n",
       "\t<tr><td>Hippos</td><td>008_216_V01</td><td>Unclustered\n",
       "Unknown</td><td>Endemic KS</td><td>PBMC</td><td>0.398461413</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>KOBS</td><td>KOBSC_9</td><td>Multi-pathogen      </td><td>KSHV Positive</td><td>PBMC</td><td>0.01916385</td></tr>\n",
       "\t<tr><td>KOBS</td><td>KOBSC_9</td><td>Unclustered\n",
       "Unknown</td><td>KSHV Positive</td><td>PBMC</td><td>0.58731915</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 689 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " cohort & repertoire\\_id & pathology & phenotype & tissue\\_type & frequency\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Hippos & 008\\_216\\_V01 & Clustered\n",
       "Unknown   & Endemic KS & PBMC & 0.592600906\\\\\n",
       "\t Hippos & 008\\_216\\_V01 & Multi-pathogen       & Endemic KS & PBMC & 0.008366322\\\\\n",
       "\t Hippos & 008\\_216\\_V01 & Unclustered\n",
       "Unknown & Endemic KS & PBMC & 0.398461413\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t KOBS & KOBSC\\_9 & Multi-pathogen       & KSHV Positive & PBMC & 0.01916385\\\\\n",
       "\t KOBS & KOBSC\\_9 & Unclustered\n",
       "Unknown & KSHV Positive & PBMC & 0.58731915\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 689 × 6\n",
       "\n",
       "| cohort &lt;chr&gt; | repertoire_id &lt;chr&gt; | pathology &lt;chr&gt; | phenotype &lt;chr&gt; | tissue_type &lt;chr&gt; | frequency &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| Hippos | 008_216_V01 | Clustered\n",
       "Unknown   | Endemic KS | PBMC | 0.592600906 |\n",
       "| Hippos | 008_216_V01 | Multi-pathogen       | Endemic KS | PBMC | 0.008366322 |\n",
       "| Hippos | 008_216_V01 | Unclustered\n",
       "Unknown | Endemic KS | PBMC | 0.398461413 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| KOBS | KOBSC_9 | Multi-pathogen       | KSHV Positive | PBMC | 0.01916385 |\n",
       "| KOBS | KOBSC_9 | Unclustered\n",
       "Unknown | KSHV Positive | PBMC | 0.58731915 |\n",
       "\n"
      ],
      "text/plain": [
       "    cohort repertoire_id pathology            phenotype     tissue_type\n",
       "1   Hippos 008_216_V01   Clustered\\nUnknown   Endemic KS    PBMC       \n",
       "2   Hippos 008_216_V01   Multi-pathogen       Endemic KS    PBMC       \n",
       "3   Hippos 008_216_V01   Unclustered\\nUnknown Endemic KS    PBMC       \n",
       "⋮   ⋮      ⋮             ⋮                    ⋮             ⋮          \n",
       "688 KOBS   KOBSC_9       Multi-pathogen       KSHV Positive PBMC       \n",
       "689 KOBS   KOBSC_9       Unclustered\\nUnknown KSHV Positive PBMC       \n",
       "    frequency  \n",
       "1   0.592600906\n",
       "2   0.008366322\n",
       "3   0.398461413\n",
       "⋮   ⋮          \n",
       "688 0.01916385 \n",
       "689 0.58731915 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kspbmc_path_annotated_gliph_nprod_table <- kspbmc_gliph_nprod_table |>\n",
    "    group_by(cohort, repertoire_id, pathology, phenotype, tissue_type) |>\n",
    "    summarize(frequency = sum(duplicate_frequency)) |>\n",
    "    ungroup() \n",
    "kspbmc_path_annotated_gliph_nprod_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "kspbmc_public_tcrs_plot <- kspbmc_path_annotated_nprod_table |>\n",
    "  mutate(frequency = frequency * 100) |>\n",
    "  ggboxplot(x = \"pathology\", y = \"frequency\", color = \"phenotype\", \n",
    "    add = \"jitter\", order = c(\"CMV\", \"DENV1\", \"DENV3/4\", \"EBV\", \"HCV\", \"HIV-1\",\n",
    "       \"HSV-2\",\n",
    "      \"InfluenzaA\", \"M.tuberculosis\", \"SARS-CoV-2\", \"Other\", \"Multi-pathogen\",\n",
    "      \"Unknown\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Cummulative frequency\", \n",
    "    legend.title = \"Cohort\", \n",
    "    format.scale = T, legend = \"right\",\n",
    "    ggtheme = theme_classic(base_size = 32),\n",
    "    palette = c(\"HIV & KSHV Positive\" = \"#fb8072\", \n",
    "  \"Epidemic KS\" = \"#e41a1c\", \"KSHV Positive\" = \"#80b1d3\",\n",
    "  \"Endemic KS\" = \"#377eb8\", \"HIV Positive\" = \"#756bb1\", \"Control\" = \"#addd8e\")) +\n",
    "  geom_hline(yintercept = 1, linetype = 2) + \n",
    "  geom_hline(yintercept = 25, linetype = 2) + \n",
    "  geom_hline(yintercept = 75, linetype = 2) + \n",
    "  geom_hline(yintercept = 90, linetype = 2) + \n",
    "  scale_y_log10(breaks = scales::trans_breaks(\"log10\", function(x) 10^x),\n",
    "    labels = scales::trans_format(\"log10\", scales::math_format(10^.x))) +\n",
    "  annotation_logticks(sides = \"l\", short = unit(2,\"mm\"),\n",
    "    mid = unit(4,\"mm\"),long = unit(8,\"mm\"), size = 1) +\n",
    "  theme_classic(base_size = 26) +\n",
    "  theme(axis.text = element_text(face = \"bold\"))\n",
    "  \n",
    "kspbmc_public_gliph_plot <- kspbmc_path_annotated_gliph_nprod_table |>\n",
    "  mutate(frequency = frequency * 100) |>\n",
    "  ggboxplot(x = \"pathology\", y = \"frequency\", color = \"phenotype\", \n",
    "    add = \"jitter\", order = c(\"CMV\", \"EBV\", \"HCV\", \"HIV-1\",\n",
    "       \"HSV-2\", \"InfluenzaA\", \"M.tuberculosis\", \"SARS-CoV-2\", \"Other\", \n",
    "       \"Multi-pathogen\",   \"Clustered\\nUnknown\", \"Unclustered\\nUnknown\")) |>\n",
    "  ggpar(xlab = FALSE, ylab = \"Cummulative frequency\", \n",
    "    legend.title = \"Cohort\", \n",
    "    format.scale = T, legend = \"right\",\n",
    "    ggtheme = theme_classic(base_size = 32),\n",
    "    palette = c(\"HIV & KSHV Positive\" = \"#fb8072\", \n",
    "  \"Epidemic KS\" = \"#e41a1c\", \"KSHV Positive\" = \"#80b1d3\",\n",
    "  \"Endemic KS\" = \"#377eb8\", \"HIV Positive\" = \"#756bb1\", \"Control\" = \"#429f03\")) +\n",
    "  geom_hline(yintercept = 1, linetype = 2) +\n",
    "  geom_hline(yintercept = 25, linetype = 2) +\n",
    "  geom_hline(yintercept = 75, linetype = 2) +  \n",
    "  geom_hline(yintercept = 90, linetype = 2) +  \n",
    "  scale_y_log10(breaks = scales::trans_breaks(\"log10\", function(x) 10^x),\n",
    "    labels = scales::trans_format(\"log10\", scales::math_format(10^.x))) +\n",
    "  annotation_logticks(sides = \"l\", short = unit(2,\"mm\"),\n",
    "    mid = unit(4,\"mm\"),long = unit(8,\"mm\"), size = 1) + \n",
    "  theme_classic(base_size = 26) +\n",
    "  theme(axis.text = element_text(face = \"bold\"))\n",
    "\n",
    "\n",
    "figure_eight_layout = \"\n",
    "1111\n",
    "1111\n",
    "2222\n",
    "2222\"\n",
    "figure_eight <- kspbmc_public_tcrs_plot +  \n",
    "    kspbmc_public_gliph_plot + \n",
    "  plot_layout(design = figure_eight_layout, guides = \"collect\") +\n",
    "  plot_annotation(tag_levels = 'A') & theme(text = element_text('NimbusSan'),\n",
    "    legend.position = \"right\",\n",
    "    axis.title.x = element_blank())\n",
    "ggsave(str_c(figures_path, \"pdf\", \"Figure_eight.pdf\", sep = \"/\"), figure_eight, width = 36, \n",
    "    height = 12, device = cairo_pdf, family = \"Arial Unicode MS\")\n",
    "ggsave(str_c(figures_path, \"svg\", \"Figure_eight.svg\", sep = \"/\"), figure_eight, width = 36, \n",
    "    height = 12)\n",
    "ggsave(str_c(figures_path, \"png\", \"Figure_eight.png\", sep = \"/\"), figure_eight, width = 36, \n",
    "    height = 12)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
